# YAML Configuration Files {#editing-yaml}

All YAML files which ruin the workflow are located in the `project_home/config` directory.
The standard YAML structure is used in all files with the primary objective being passing workflow parameters to the various steps of the workflow.
There are four files which control various aspects: `config.yml`, `colours.yml`, `params.yml` and `rmarkdown.yml`

## The Main Configuration: `config.yml` {#config-yml}

This file sets many of the primary parameters and is the file which *will need editing for any new dataset.*
Many settings should remain unchanged as changing default locations of files may lead to unexpected instability in the workflow, whilst other setting **should** be changed, such as those which determine which comparisons to perform.
This is the only file parsed directly by `snakemake` and subsequent rules, whilst all others are used to pass parameters to `R` environments.

An example layout of `config.yml` might be:

```
samples:
  file: "config/samples.tsv"

paths:
  bam: "data/aligned"
  bigwig: "data/bigwig"
  macs2: "data/macs2"

genome:
  build: "GRCh37"
  gencode: "33"

external:
  rnaseq: "data/external/results.tsv"
  features: "data/external/h3k27ac_features.gtf"
  hic: "data/external/encode_hic.bedpe"
  coverage:
    H3K27Ac:
      control: "data/external/H3K27Ac_control.bw"
      treat: "data/external/H3K27Ac_treat.bw"

comparisons:
  fc: 1.2
  fdr: 0.05
  paired: false
  filter_q: 0.6
  contrasts:
    - ["control", "treat"]
  ihw: "regions"

peaks:
  macs2:
    gsize: "hs"
    fdr: 0.05
    keep_duplicates: "all"
  qc:
    min_prop_peaks: 0.1
    min_prop_reps: 0.3
```


### Settings Which Don't Need To Be Modified {.unlisted .unnumbered}

In general, the paths to key files don't need to be changed and default configurations are well tested.
Whilst varying these has been intermittently attempted successfully, unexpected instability may occur and as such, is discouraged.

`samples:` (Default: "config/samples.tsv")

By default, the file which contains all sample-level information is defined as `config/samples.tsv`.
In theory, this can be changed but this may adversely impact the pipeline.

`paths:`

- `bam:` (Default: "data/aligned") Alignments should be placed in `data/aligned` as advised in section \@ref(alignments), although this can be changed to `data/bam` or any other relative path as desired. Again, changes to the default layout may adversely affect the pipeline stability
- `macs2:` (Default: "data/macs2") Output from `macs2 callpeak` will be placed in this directory, mirroring the input structure from `data/aligned` where each ChIP target (e.g. TF1, TF2) will have results written to separate directories.
- `bigwig:` (Default: "data/bigwig") After running `macs2 callpeak`, `bedGraph` files will be converted to the more space efficient `bigwig` files. The directory structure from both `data/aligned` and `data/macs2` will again be mirrored such that each ChIP target has all samples written to the same directory

### Settings Which Should Be Modified {.unlisted .unnumbered}

`genome:`

Specify the genome build used for alignments and for gene annotations.
Build must match that used exactly for performing alignments.
The gtf corresponding to the specified [Gencode release](https://www.gencodegenes.org/) will be downloaded as part of the workflow

`external:`

Provide paths to all optional data files here.
Only files provided will be included in the workflow.

- `rnaseq` should be the results as output by `limma::topTable()` or similar. Gene IDs should match those provided in the Gencode GTF (e.g. Ensembl IDs) and these should be in a column names `gene_id`. Columns such as `logFC` and `FDR` will be searched for during the workflow using regular expressions to find the best match. Can be in `csv` or `tsv` format. Excel-specific (`xls`, `xlsx`) formats are not supported.
- `features` can be determined by any method, with common choices being relevant histone marks, or promoters, enhancers and super-enhancers determined by H3K27ac marks. Features should be non-overlapping with the field `feature` defining the different feature types. Must be provided in GTF format.
- `hic` HiC interactions must be provided as a `bedpe` file
- `coverage` Tracks provided in this argument will be added to all genomic plots showing binding peaks or differential binding. Multiple files provided within each YAML list element will be overlaid as a single track. There is no upper limit to the number of tracks, however more tracks generally detract from an informative figure.

`comparisons:`

These settings determine how the differential binding analysis is performed.

- `fc` (Default: 1.2) The setting of 1.2 indicates a 20% change in binding as the threshold below which we are not interested, or below which we consider binding changes to be inconsequential. This parameters is passed to `limma::treat()` [@treat] in all differential binding analyses.
- `fdr` (Default: 0.05) Windows with significance below this threshold are considered to provide supporting evidence of differential binding.
- `paired` (Default: `false`) If set to `true` the values in the optional column (e.g. replicate, passage etc.) are used to perform a paired analysis as described in the `limma` manual
- `filter_q` (Default: 0.6) Passed to `extraChIPs::dualFilter()`. When filtering (i.e. discarding) genomic sliding windows which are unlikely to contain true binding signal, determine thresholds which will retain this proportion of windows which overlap a peak identified by `macs2 callpeak`.
- `contrasts`: Define all contrasts for differential binding. Any ChIP target containing both treatment groups will be included for differential binding. Values must match those in the `treat` column of `samples.tsv`. Each differential binding analysis will be performed using the limma-trend method in the context of A vs B, such that complex models are not supported. Use new YAML list elements to define additional contrasts
- `ihw` (Default: "regions") Options used to stratify p-values for Independent Hypothesis Weighting[@ihw] of differential binding results. Can take values `"regions"`, `"features"`, `"targets"` or `"none"`
    + `"regions"` P-values will be stratified by annotated genomic regions as determined in the initial steps of the workflow
    + `"features"` P-values will be stratified by provided external features
    + `"targets"` P-values will be stratified by the presence of a `macs2` consensus peak using all other ChIP targets in combination
    + `"none"` No Independent Hypothesis Weighting will be performed on the results from differential binding

`peaks:`

`macs2` settings are passed to `macs2 callpeak`. Only the arguments `gsize`, `fdr` and `keep_duplicates` are accepted.
Please see the [`macs2` manual](https://macs3-project.github.io/MACS/) for more detailed explanations.

`qc` parameters are used for determining if samples are of a high enough quality, and how to determine consensus/oracle peaks for each target and treatment group.

- `min_prop_peaks:` (Default: 0.1) Samples within a treatment group must contain more than this proportion of peaks which overlap peaks from the "best" sample within the relevant treatment group, i.e. the sample with the most number of peaks identified by `macs2 callpeak`. This is an inclusive threshold which will only discard clearly poor samples, which themselves are more likely to occur in tissue/organoid than in cell-line derived data.
- `min_prop_reps` (Default: 0.3) When forming **oracle** peaks within each *target and treatment group*, peaks must be represented in at least this proportion of samples. This defaults to 0.3 which would equate to 1 of 3 samples, 2 of 4 samples etc. This value may need to be altered pending the results of a complete run after Quality Assessment has performed.

## Colour Schemes: `colours.yml` [#colours-yml]

Defines all plotting colour schemes for consistency throughout the entire workflow.
Colours can be any standard colours able to be interpreted by `R`, such as `'blue'` or `'#0000FF'`.
Recommended YAML list elements are `qc`, `regions`, `direction` and `treat`. 
Any unspecified colours will be automatically assigned and will propagate through the workflow.
As is standard across most programming languages, names are case-sensitive.
An example file is given below:

```
qc:
  pass: "#0571B0"
  fail: "#CA0020"
direction:
  up: "#CA0020"
  down: "#0571B0"
  unchanged: "#7F7F7F"
  undetected: "#E5E5E5"
regions:
  promoters: '#FF3300'
  upstream: '#E1EE05'
  exons: '#7EDD57'
  introns: '#006600'
  proximal: '#000066'
  distal: '#551A8B'
treat:
  Input: "#33333380"
  control: "#4D4D4D"
  treat: "#C53270"
features:
  promoter: "#FF4500"
  enhancer: "#FFFF00"
  super_enhancer: "#FFC34D"
  no_feature: "#E5E5E5"
```

## Additional Parameters: `params.yml` {#params-yml}

Default settings for defining initial annotations and enrichment testing. 
In general, these will not need to be changed, but can be if required.

Parameters for `msigdb`[@msigdb] are passed to `msigdbr`[@msigdbr] and fields should match this layout.
Any categories passed to `gs_cat` will lead to all subcategories being used from that category.
Specific sub-categories of larger databases can be passed using `gs_subcat` element.

```
gene_regions:
  promoters:
    upstream: 1500
    downstream: 500
  upstream: 5000
  intergenic: 10000

msigdb:
  gs_cat: "H"
  gs_subcat:
    - "CP:KEGG"
    - "CP:REACTOME"
    - "CP:WIKIPATHWAYS"
    - "TFT:GTRD"
```


## HTML Settings: `rmarkdown.yml` {#rmarkdown-yml}

The main workflow will produce a compiled set of HTML pages using `rmarkdown::render_site()`[@R-rmarkdown].
The two available fields to supply here are:

1. `knitr_opts` which are passed to `knitr::opts_chunk$set()`[@R-knitr] at the beginning of every compiled Rmarkdown document, and 
2. `rmarkdown_site` which determines the layout and style of the final HTML report.
All *left* elements of the `navbar` are determined automatically during the workflow and will be ignored if supplied here, whilst all other parameters are passed via the file `_site.yaml` which will be generated during the workflow.

```
knitr_opts:
  echo: TRUE
  message: FALSE
  warning: FALSE
  dev: ["png", "pdf"]
  fig.align: "center"
  fig.width: 10
  fig.height: 8

rmarkdown_site:
  name: "GRAVI: Gene Regulatory Analaysis"
  output_dir: "../docs"
  navbar:
    title: "GRAVI"
    right:
      - icon: fa-github
        href: "https://github.com/steveped/GRAVI"
  output:
    html_document:
      toc: yes
      toc_float: yes
      code_folding: hide
      self_contained: false
      theme: sandstone
      highlight: textmate
      includes:
        after_body: footer.html
```
