<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 YAML Configuration Files | GRAVI User Guide (v0.1.6)</title>
  <meta name="description" content="This is the main documentation for the GRAVI workflow" />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 YAML Configuration Files | GRAVI User Guide (v0.1.6)" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is the main documentation for the GRAVI workflow" />
  <meta name="github-repo" content="smped/GRAVI" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 YAML Configuration Files | GRAVI User Guide (v0.1.6)" />
  
  <meta name="twitter:description" content="This is the main documentation for the GRAVI workflow" />
  

<meta name="author" content="Stephen Pederson" />


<meta name="date" content="2023-07-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="input-files.html"/>
<link rel="next" href="references.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.6.2/htmlwidgets.js"></script>
<script src="libs/viz-1.8.2/viz.js"></script>
<link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
<script src="libs/grViz-binding-1.0.10/grViz.js"></script>



<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">GRAVI User Guide</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="workflow.html"><a href="workflow.html"><i class="fa fa-check"></i><b>2</b> Workflow Description</a>
<ul>
<li class="chapter" data-level="2.1" data-path="workflow.html"><a href="workflow.html#annotation"><i class="fa fa-check"></i><b>2.1</b> Annotation Setup</a></li>
<li class="chapter" data-level="2.2" data-path="workflow.html"><a href="workflow.html#peak-calling"><i class="fa fa-check"></i><b>2.2</b> Peak Calling</a></li>
<li class="chapter" data-level="2.3" data-path="workflow.html"><a href="workflow.html#differential-signal"><i class="fa fa-check"></i><b>2.3</b> Differential Signal</a></li>
<li class="chapter" data-level="2.4" data-path="workflow.html"><a href="workflow.html#pairwise-comparisons"><i class="fa fa-check"></i><b>2.4</b> Pairwise Comparisons</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="workflow.html"><a href="workflow.html#key-outputs-2"><i class="fa fa-check"></i><b>2.4.1</b> Key Outputs</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>3</b> Quick Start Guide</a>
<ul>
<li class="chapter" data-level="3.1" data-path="quick-start.html"><a href="quick-start.html#snakemake"><i class="fa fa-check"></i><b>3.1</b> Install Snakemake</a></li>
<li class="chapter" data-level="3.2" data-path="quick-start.html"><a href="quick-start.html#quick-directories"><i class="fa fa-check"></i><b>3.2</b> Create the Directory Structure</a></li>
<li class="chapter" data-level="3.3" data-path="quick-start.html"><a href="quick-start.html#quick-run"><i class="fa fa-check"></i><b>3.3</b> Run the Pipeline</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="quick-start.html"><a href="quick-start.html#run-on-a-local-server"><i class="fa fa-check"></i><b>3.3.1</b> Run On A Local Server</a></li>
<li class="chapter" data-level="3.3.2" data-path="quick-start.html"><a href="quick-start.html#run-on-an-hpc"><i class="fa fa-check"></i><b>3.3.2</b> Run On An HPC</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="quick-start.html"><a href="quick-start.html#tips-and-tricks"><i class="fa fa-check"></i><b>3.4</b> Tips And Tricks</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="quick-start.html"><a href="quick-start.html#removing-large-files"><i class="fa fa-check"></i><b>3.4.1</b> Removing Large files</a></li>
<li class="chapter" data-level="3.4.2" data-path="quick-start.html"><a href="quick-start.html#shared-conda-environments"><i class="fa fa-check"></i><b>3.4.2</b> Shared Conda Environments</a></li>
<li class="chapter" data-level="3.4.3" data-path="quick-start.html"><a href="quick-start.html#running-restricted-sections-of-the-worklow"><i class="fa fa-check"></i><b>3.4.3</b> Running Restricted Sections of the Worklow</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="input-files.html"><a href="input-files.html"><i class="fa fa-check"></i><b>4</b> Files and Directories</a>
<ul>
<li class="chapter" data-level="4.1" data-path="input-files.html"><a href="input-files.html#directories"><i class="fa fa-check"></i><b>4.1</b> Directory Structure</a></li>
<li class="chapter" data-level="4.2" data-path="input-files.html"><a href="input-files.html#alignments"><i class="fa fa-check"></i><b>4.2</b> Alignments</a></li>
<li class="chapter" data-level="4.3" data-path="input-files.html"><a href="input-files.html#sample-descriptions"><i class="fa fa-check"></i><b>4.3</b> Sample Descriptions</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="input-files.html"><a href="input-files.html#required-columns"><i class="fa fa-check"></i><b>4.3.1</b> Required columns</a></li>
<li class="chapter" data-level="4.3.2" data-path="input-files.html"><a href="input-files.html#optional-columns"><i class="fa fa-check"></i><b>4.3.2</b> Optional Columns</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="input-files.html"><a href="input-files.html#additional-files"><i class="fa fa-check"></i><b>4.4</b> Additional Files</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="input-files.html"><a href="input-files.html#rna-seq"><i class="fa fa-check"></i><b>4.4.1</b> RNA-Seq</a></li>
<li class="chapter" data-level="4.4.2" data-path="input-files.html"><a href="input-files.html#external-features"><i class="fa fa-check"></i><b>4.4.2</b> External Features</a></li>
<li class="chapter" data-level="4.4.3" data-path="input-files.html"><a href="input-files.html#hic-interactions"><i class="fa fa-check"></i><b>4.4.3</b> HiC Interactions</a></li>
<li class="chapter" data-level="4.4.4" data-path="input-files.html"><a href="input-files.html#external-coverage"><i class="fa fa-check"></i><b>4.4.4</b> External Coverage</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="editing-yaml.html"><a href="editing-yaml.html"><i class="fa fa-check"></i><b>5</b> YAML Configuration Files</a>
<ul>
<li class="chapter" data-level="5.1" data-path="editing-yaml.html"><a href="editing-yaml.html#config-yml"><i class="fa fa-check"></i><b>5.1</b> The Main Configuration: <code>config.yml</code></a></li>
<li class="chapter" data-level="5.2" data-path="editing-yaml.html"><a href="editing-yaml.html#colour-schemes-colours.yml-colours-yml"><i class="fa fa-check"></i><b>5.2</b> Colour Schemes: <code>colours.yml</code> [#colours-yml]</a></li>
<li class="chapter" data-level="5.3" data-path="editing-yaml.html"><a href="editing-yaml.html#params-yml"><i class="fa fa-check"></i><b>5.3</b> Additional Parameters: <code>params.yml</code></a></li>
<li class="chapter" data-level="5.4" data-path="editing-yaml.html"><a href="editing-yaml.html#rmarkdown-yml"><i class="fa fa-check"></i><b>5.4</b> HTML Settings: <code>rmarkdown.yml</code></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">GRAVI User Guide (v0.1.6)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="editing-yaml" class="section level1 hasAnchor" number="5">
<h1><span class="header-section-number">Chapter 5</span> YAML Configuration Files<a href="editing-yaml.html#editing-yaml" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>All YAML files which ruin the workflow are located in the <code>project_home/config</code> directory.
The standard YAML structure is used in all files with the primary objective being passing workflow parameters to the various steps of the workflow.
There are four files which control various aspects: <code>config.yml</code>, <code>colours.yml</code>, <code>params.yml</code> and <code>rmarkdown.yml</code></p>
<div id="config-yml" class="section level2 hasAnchor" number="5.1">
<h2><span class="header-section-number">5.1</span> The Main Configuration: <code>config.yml</code><a href="editing-yaml.html#config-yml" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This file sets many of the primary parameters and is the file which <em>will need editing for any new dataset.</em>
Many settings should remain unchanged as changing default locations of files may lead to unexpected instability in the workflow, whilst other setting <strong>should</strong> be changed, such as those which determine which comparisons to perform.
This is the only file parsed directly by <code>snakemake</code> and subsequent rules, whilst all others are used to pass parameters to <code>R</code> environments.</p>
<p>An example layout of <code>config.yml</code> might be:</p>
<pre><code>samples:
  file: &quot;config/samples.tsv&quot;

paths:
  bam: &quot;data/bam&quot;

genome:
  build: &quot;GRCh37&quot;

external:
  blacklist: &quot;data/external/blacklist.bed.gz&quot;
  gtf: &quot;data/external/gencode.v33lift37.annotation.gtf.gz&quot;
  rnaseq: &quot;data/external/some_results.tsv&quot;
  features: &quot;data/external/custom_features.gtf.gz&quot;
  hic: &quot;data/external/hic_interactions.bedpe&quot;
  coverage:
    H3K27Ac:
      control: &quot;data/external/H3K27Ac_control.bw&quot;
      treat: &quot;data/external/H3K27Ac_treat.bw&quot;

comparisons:
  method: &#39;sq-lt&#39;
  fc: 1.2
  fdr: 0.05
  paired: false
  filter_q: 0.6
  contrasts:
    - [&quot;control&quot;, &quot;treat&quot;]
  ihw: &quot;regions&quot;

peaks:
  macs2:
    gsize: &quot;hs&quot;
    fdr: 0.05
    keep_duplicates: &quot;all&quot;
  qc:
    outlier_threshold: 10
    allow_zero: true
    min_prop_reps: 0.5</code></pre>
<div id="settings-which-dont-need-to-be-modified" class="section level3 unlisted unnumbered hasAnchor">
<h3>Settings Which Don’t Need To Be Modified<a href="editing-yaml.html#settings-which-dont-need-to-be-modified" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In general, the paths to key files don’t need to be changed and default configurations are well tested.
Whilst varying these has been intermittently attempted successfully, unexpected instability may occur and as such, is discouraged.</p>
<p><code>samples:</code> (Default: “config/samples.tsv”)</p>
<p>By default, the file which contains all sample-level information is defined as <code>config/samples.tsv</code>.
This can be changed to reflect any requirements after inspecting QC reports.</p>
<p><code>paths:</code></p>
<ul>
<li><code>bam:</code> (Default: “data/bam”) Alignments should be placed in <code>data/bam</code> as advised in section <a href="input-files.html#alignments">4.2</a>, although this can be changed to any other path as desired.</li>
</ul>
</div>
<div id="settings-which-should-be-modified" class="section level3 unlisted unnumbered hasAnchor">
<h3>Settings Which Should Be Modified<a href="editing-yaml.html#settings-which-should-be-modified" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><code>genome:</code></p>
<p>Specify the genome build used for alignments and for gene annotations.</p>
<p><code>external:</code></p>
<p>Provide paths to all additional data files here.
Only files provided will be included in the workflow.</p>
<ul>
<li><code>blacklist</code> Should be obtained from <a href="https://github.com/Boyle-Lab/Blacklist/tree/master/lists">here</a></li>
<li><code>gtf</code>: Should ideally be obtained from <a href="https://www.gencodegenes.org/">Gencode</a></li>
<li>(Optional) <code>rnaseq</code> should be the results as output by <code>limma::topTable()</code> or similar. Gene IDs should match those provided in the Gencode GTF (e.g. Ensembl IDs) and these should be in a column names <code>gene_id</code>. Columns such as <code>logFC</code> and <code>FDR</code> will be searched for during the workflow using regular expressions to find the best match. Can <strong>only</strong> be in <code>csv</code> or <code>tsv</code> format. Excel-specific (<code>xls</code>, <code>xlsx</code>) formats are not supported.</li>
<li>(Optional) <code>features</code> can be determined by any method, with common choices being relevant histone marks, or promoters, enhancers and super-enhancers determined by H3K27ac marks. Features should be non-overlapping with the field <code>feature</code> defining the different feature types. Must be provided in GTF format.</li>
<li>(Optional) <code>hic</code> HiC interactions must be provided as a <code>bedpe</code> file</li>
<li>(Optional) <code>coverage</code> Tracks provided in this argument will be added to all genomic plots showing binding peaks or differential binding. Multiple files provided within each YAML list element will be overlaid as a single track. There is no upper limit to the number of tracks, however more tracks generally detract from an informative figure.</li>
</ul>
<p><code>comparisons:</code></p>
<p>These settings determine how the differential binding analysis is performed.</p>
<ul>
<li><code>method</code> (Default: “sq-lt”) Defines the strategy used for differential binding. The currently implemented options are “sq-lt” and “ls-ql” as described previously</li>
<li><code>fc</code> (Default: 1.2) The setting of 1.2 indicates a 20% change in binding as the threshold below which we are not interested, or below which we consider binding changes to be inconsequential. This parameters is passed to <code>limma::treat()</code> <span class="citation">(<a href="#ref-treat">McCarthy and Smyth 2009</a>)</span> in all differential binding analyses.</li>
<li><code>fdr</code> (Default: 0.05) Windows with significance below this threshold are considered to provide supporting evidence of differential binding.</li>
<li><code>paired</code> (Default: <code>false</code>) If set to <code>true</code> the values in the optional column (e.g. replicate, passage etc.) are used to perform a paired analysis as described in the <code>limma</code> manual</li>
<li><code>filter_q</code> (Default: 0.6) Passed to <code>extraChIPs::dualFilter()</code>. When filtering (i.e. discarding) genomic sliding windows which are unlikely to contain true binding signal, determine thresholds which will retain this proportion of windows which overlap a peak identified by <code>macs2 callpeak</code>.</li>
<li><code>contrasts</code>: Define all contrasts for differential binding. Any ChIP target containing both treatment groups will be included for differential binding. Values must match those in the <code>treat</code> column of <code>samples.tsv</code>. Each differential binding analysis will be performed using the limma-trend method in the context of A vs B, such that complex models are not supported. Use new YAML list elements to define additional contrasts</li>
<li><code>ihw</code> (Default: “regions”) Options used to stratify p-values for Independent Hypothesis Weighting<span class="citation">(<a href="#ref-ihw">Ignatiadis et al. 2016</a>)</span> of differential binding results. Can take values <code>"regions"</code>, <code>"features"</code>, <code>"targets"</code> or <code>"none"</code>
<ul>
<li><code>"regions"</code> P-values will be stratified by annotated genomic regions as determined in the initial steps of the workflow</li>
<li><code>"features"</code> P-values will be stratified by provided external features</li>
<li><code>"targets"</code> P-values will be stratified by the presence of a union peak using all other ChIP targets</li>
<li><code>"none"</code> No Independent Hypothesis Weighting will be performed on the results from differential binding</li>
</ul></li>
</ul>
<p><code>peaks:</code></p>
<p><code>macs2</code> settings are passed to <code>macs2 callpeak</code>. Only the arguments <code>gsize</code>, <code>fdr</code> and <code>keep_duplicates</code> are accepted.
Please see the <a href="https://macs3-project.github.io/MACS/"><code>macs2</code> manual</a> for more detailed explanations.</p>
<p><code>qc</code> parameters are used for determining if samples are of a high enough quality, and how to determine union/treatment peaks for each target and treatment group.</p>
<ul>
<li><code>outlier_threshold</code> (Default: 10) As previously described, samples with peak numbers beyond this value relative to the median peak numbers in each treatment group are marked for removal. To ignore this parameter, set this to <code>.inf</code> which effectively sets the threshold to <span class="math inline">\(\infty\)</span></li>
<li><code>allow_zero</code> (Default: <code>true</code>) Allow samples with no identified peaks. If a ChIP target is expected to be cytoplasmic in one condition, this can allow samples with no peaks to be retained.</li>
<li><code>min_prop_reps</code> (Default: 0.5) When forming <strong>treatment-specific</strong> peaks within each <em>target and treatment group</em>, peaks must be represented in at least this proportion of samples. This defaults to 0.5 which would equate to 2 of 3 samples, 2 of 4 samples etc. This value may need to be altered pending the results of a complete run after Quality Assessment has performed.</li>
</ul>
</div>
</div>
<div id="colour-schemes-colours.yml-colours-yml" class="section level2 hasAnchor" number="5.2">
<h2><span class="header-section-number">5.2</span> Colour Schemes: <code>colours.yml</code> [#colours-yml]<a href="editing-yaml.html#colour-schemes-colours.yml-colours-yml" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Defines all plotting colour schemes for consistency throughout the entire workflow.
Colours can be any standard colours able to be interpreted by <code>R</code>, such as <code>'blue'</code> or <code>'#0000FF'</code>.
Recommended YAML list elements are <code>qc</code>, <code>regions</code>, <code>direction</code> and <code>treat</code>.
A gradient through al colours provided in the <code>heatmaps</code> element will be generated during generation of heatmaps.
Any unspecified colours will be automatically assigned and will propagate through the workflow.
As is standard across most programming languages, names are case-sensitive.
An example file is given below:</p>
<pre><code>qc:
  pass: &quot;#0571B0&quot;
  fail: &quot;#CA0020&quot;
regions:
  promoter: &#39;#FF3300&#39;
  upstream_promoter: &#39;#E1EE05&#39;
  exon: &#39;#7EDD57&#39;
  intron: &#39;#006600&#39;
  proximal_intergenic: &#39;#000066&#39;
  distal_intergenic: &#39;#551A8B&#39;
treat:
  Input: &quot;#33333380&quot;
  E2: &quot;#3333CC&quot;
  E2DHT: &quot;#C52240&quot;
features:
  enhancer: &quot;#FFFF00&quot;
  no_feature: &quot;#E5E5E5&quot;
direction:
  up: &quot;#CA0020&quot;
  down: &quot;#0571B0&quot;
  unchanged: &quot;#7F7F7F&quot;
  undetected: &quot;#E5E5E5&quot;
heatmaps: [&#39;white&#39;, &#39;red&#39;]</code></pre>
</div>
<div id="params-yml" class="section level2 hasAnchor" number="5.3">
<h2><span class="header-section-number">5.3</span> Additional Parameters: <code>params.yml</code><a href="editing-yaml.html#params-yml" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Default settings for defining initial annotations, mapping of peaks to genes and enrichment testing.
In general, these will not need to be changed, but can be if required.
In particular, the value <strong>enh2gene</strong> can be set to zero if using data such as H3K27ac HiChIP to determine long-range interactions.</p>
<p>For enrichment testing, the preferred p-value adjustment method can be specified using any value that match <code>p.adjust.methods</code> with an appropriate threshold also set in this section.
Default settings are relatively inclusive and more stringent setting may be preferred by some users.</p>
<p>Parameters for <code>msigdb</code><span class="citation">(<a href="#ref-msigdb">Liberzon et al. 2015</a>)</span> are passed to <code>msigdbr</code><span class="citation">(<a href="#ref-msigdbr">Dolgalev 2022</a>)</span> and fields should match this layout.
Any categories passed to <code>gs_cat</code> will lead to all subcategories being used from that category.
Specific sub-categories of larger databases can be passed using <code>gs_subcat</code> element.</p>
<p>Values which determine the minimum and maximum size of any network plots are also defined here.
Node-pair distances above the given distance will be removed from the plot and these edges excluded.</p>
<pre><code>gene_regions:
  promoters:
    upstream: 1500
    downstream: 500
  upstream: 5000
  intergenic: 10000

## The values used when mapping peaks to genes.
## Passed to `extraChIPs::mapByFeature()`
## If including H3K27ac HiChIP for long range-interactions, it is advised to
## set `enh2gene` as zero, given that long range interactions in this case
## will more accurately map long-range enhancer interactions than using all
## genes within a given distance
mapping:
  gr2gene: 100000
  prom2gene: 0
  enh2gene: 100000
  gi2gene: 0

enrichment:
  adj: &quot;fdr&quot;
  alpha: 0.05
  ## Only gene-sets between these two values will be retained before testing
  min_size: 5
  max_size: 1000
  ## Only gene-sets above this size will be shown in the results
  min_sig: 3
  ## The categories to use from MSigDB. These are passed to msigdbr in the
  ## columns of the same name in an &#39;OR&#39; approach
  species: &quot;Homo sapiens&quot;
  msigdb:
    gs_cat: &quot;H&quot;
    gs_subcat:
      - &quot;CP:KEGG&quot;
      - &quot;CP:REACTOME&quot;
      - &quot;CP:WIKIPATHWAYS&quot;
      - &quot;TFT:GTRD&quot;

## Used for network plots
networks:
  min_size: 4
  max_size: 80
  max_distance: 0.9
  layout: &#39;fr&#39;</code></pre>
</div>
<div id="rmarkdown-yml" class="section level2 hasAnchor" number="5.4">
<h2><span class="header-section-number">5.4</span> HTML Settings: <code>rmarkdown.yml</code><a href="editing-yaml.html#rmarkdown-yml" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The main workflow will produce a compiled set of HTML pages using <code>rmarkdown::render_site()</code><span class="citation">(<a href="#ref-R-rmarkdown">Allaire et al. 2022</a>)</span>.
The two available fields to supply here are:</p>
<ol style="list-style-type: decimal">
<li><code>knitr_opts</code> which are passed to <code>knitr::opts_chunk$set()</code><span class="citation">(<a href="#ref-R-knitr">Xie 2022</a>)</span> at the beginning of every compiled Rmarkdown document, and</li>
<li><code>rmarkdown_site</code> which determines the layout and style of the final HTML report.
All <em>left</em> elements of the <code>navbar</code> are determined automatically during the workflow and will be ignored if supplied here, whilst all other parameters are passed via the file <code>_site.yaml</code> which will be generated during the workflow.</li>
</ol>
<p>Figure height and width parameters must be passed to <code>knitr::opts_set()</code> in inches, and the workflow defaults to <code>svg</code> output for most figures.
Some figures which can become excessively large under this setting, such as those with many thousands of overlapping points, are set to always produce <code>png</code> output.
Multiple output formats can also be set using values such as <code>['png', 'pdf']</code> if preferred.</p>
<pre><code>knitr_opts:
  echo: TRUE
  message: FALSE
  warning: FALSE
  dev: [&quot;png&quot;, &quot;pdf&quot;]
  fig.align: &quot;center&quot;
  fig.width: 10
  fig.height: 8

rmarkdown_site:
  name: &quot;GRAVI: Gene Regulatory Analysis&quot;
  output_dir: &quot;../docs&quot;
  navbar:
    title: &quot;GRAVI&quot;
    right:
      - icon: fa-github
        href: &quot;https://github.com/smped/GRAVI&quot;
  output:
    html_document:
      toc: yes
      toc_float: yes
      code_folding: hide
      self_contained: false
      theme: sandstone
      highlight: textmate
      css: gravi.css
      includes:
        after_body: footer.html</code></pre>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-R-rmarkdown" class="csl-entry">
Allaire, JJ, Yihui Xie, Jonathan McPherson, Javier Luraschi, Kevin Ushey, Aron Atkins, Hadley Wickham, Joe Cheng, Winston Chang, and Richard Iannone. 2022. <em>Rmarkdown: Dynamic Documents for r</em>. <a href="https://CRAN.R-project.org/package=rmarkdown">https://CRAN.R-project.org/package=rmarkdown</a>.
</div>
<div id="ref-msigdbr" class="csl-entry">
Dolgalev, Igor. 2022. <em>Msigdbr: MSigDB Gene Sets for Multiple Organisms in a Tidy Data Format</em>. <a href="https://CRAN.R-project.org/package=msigdbr">https://CRAN.R-project.org/package=msigdbr</a>.
</div>
<div id="ref-ihw" class="csl-entry">
Ignatiadis, N., B. Klaus, J. B. Zaugg, and W. Huber. 2016. <span>“<span class="nocase"><span>D</span>ata-driven hypothesis weighting increases detection power in genome-scale multiple testing</span>.”</span> <em>Nat Methods</em> 13 (7): 577–80.
</div>
<div id="ref-msigdb" class="csl-entry">
Liberzon, A., C. Birger, H. Thorvaldsdóttir, M. Ghandi, J. P. Mesirov, and P. Tamayo. 2015. <span>“<span class="nocase"><span>T</span>he <span>M</span>olecular <span>S</span>ignatures <span>D</span>atabase (<span>M</span><span>S</span>ig<span>D</span><span>B</span>) hallmark gene set collection</span>.”</span> <em>Cell Syst</em> 1 (6): 417–25.
</div>
<div id="ref-treat" class="csl-entry">
McCarthy, Davis J., and Gordon K. Smyth. 2009. <span>“<span class="nocase">Testing significance relative to a fold-change threshold is a TREAT</span>.”</span> <em>Bioinformatics</em> 25 (6): 765–71. <a href="https://doi.org/10.1093/bioinformatics/btp053">https://doi.org/10.1093/bioinformatics/btp053</a>.
</div>
<div id="ref-R-knitr" class="csl-entry">
Xie, Yihui. 2022. <em>Knitr: A General-Purpose Package for Dynamic Report Generation in r</em>. <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="input-files.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/steveped/GRAVI/edit/gh-pages/editing_yaml.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
