<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Editing The YAML Configuration Files</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GRAVI User Guide<br>(Development version)</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Getting Started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="quickstart.html">Quick Start Guide</a>
    </li>
    <li>
      <a href="input_files.html">Input Files and Directories</a>
    </li>
    <li>
      <a href="editing_yaml.html">Editing YAML Configuration Files</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Workflow Description
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="annotations.html">Annotations</a>
    </li>
    <li>
      <a href="peak_calling.html">Peak Calling</a>
    </li>
    <li>
      <a href="differential_binding.html">Differential Binding</a>
    </li>
    <li>
      <a href="pairwise_comparisons.html">Pairwise Comparisons</a>
    </li>
    <li>
      <a href="motif_detection.html">Motif Detection</a>
    </li>
  </ul>
</li>
<li>
  <a href="output.html">Output Description</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steveped/GRAVI">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Editing The YAML Configuration Files</h1>

</div>


<p>All YAML files which ruin the workflow are located in the
<code>project_home/config</code> directory.</p>
<div id="config.yml-the-main-configuration" class="section level2">
<h2><code>config.yml</code>: The Main Configuration</h2>
<p>This file sets many of the primary parameters and is the file which
<em>will need editing for any new dataset.</em></p>
<div id="settings-which-shouldnt-change" class="section level3">
<h3>Settings Which Shouldn’t Change</h3>
<p><code>samples:</code> (Default: “config/samples.tsv”)</p>
<p>By default, the file which contains all sample-level information is
defined as <code>config/samples.tsv</code>. In theory, this can be
changed but this may adversely impact the pipeline.</p>
<p><code>paths:</code></p>
<ul>
<li><code>bam:</code> (Default: “data/aligned”) Alignments should be
placed in <code>data/aligned</code> <a href="input_files.html">as
advised</a>, although this can be changed to <code>data/bam</code> or
any other relative path as desired. Again, changes to the default layout
may adversely affect the pipeline</li>
<li><code>macs2:</code> (Default: “data/macs2”) Output from
<code>macs2 callpeak</code> will be placed in this directory, mirroring
the input structure from <code>data/aligned</code> where each ChIP
target (e.g. TF1, TF2) will have results written to separate
directories.</li>
<li><code>bigwig:</code> (Default: “data/bigwig”) After running
<code>macs2 callpeak</code>, <code>bedGraph</code> files will be
converted to the more space efficient <code>bigwig</code> files. The
directory structure from both <code>data/aligned</code> and
<code>data/macs2</code> will again be mirrored such that each ChIP
target has all samples written to the same directory</li>
</ul>
<p><code>comparisons:</code></p>
<p>These settings determine how the differential binding analysis is
performed.</p>
<ul>
<li><code>fc</code> (Default: 1.2) The setting of 1.2 indicates a 20%
change in binding as the threshold below which we are not interested, or
below which we consider binding changes to be inconsequential. This
parameters is passed to <code>limma::treat()</code> in all differential
binding analyses.</li>
<li><code>fdr</code> (Default: 0.05) Windows with significance below
this threshold are considered to represent evidence of differential
binding.</li>
<li><code>paired</code> (Default: <code>false</code>) If set to
<code>true</code> the values in the optional column (e.g. replicate,
passage etc.) are used to perform a paired analysis as described in the
<code>limma</code> manual</li>
<li><code>filter_q</code> (Default: 0.6) Passed to
<code>extraChIPs::dualFilter()</code>. When filtering (i.e. discarding)
genomic sliding windows which are unlikely to contain true binding
signal, determine thresholds which will retain this proportion of
windows which overlap a peak identified by
<code>macs2 callpeak</code>.</li>
<li><code>contrasts</code>: Define all contrasts for differential
binding. Any ChIP target containing both treatment groups will be
included for differential binding. Values must match those in the
<code>treat</code> column of <code>samples.tsv</code></li>
<li><code>ihw</code> (Default: “regions”) Used to stratify p-values for
Independent Hypothesis Weighting for differential binding results. Can
take values <code>"regions"</code>, <code>"features"</code>,
<code>"targets"</code> or <code>"none"</code>
<ul>
<li><code>"regions"</code> P-values will be stratified by annotated
genomic regions as determined in the initial steps of the workflow</li>
<li><code>"features"</code> P-values will be stratified by provided
external features</li>
<li><code>"targets"</code> P-values will be stratified by the presence
of a <code>macs2</code> consensus peak using all other ChIP targets in
combination</li>
<li><code>"none"</code> No Independent Hypothesis Weighting will be
performed on the results from differential binding</li>
</ul></li>
</ul>
<p><code>genome:</code></p>
<p>Specify the genome build used for alignments and for gene
annotations. Build must match that used exactly for performing
alignments. The gtf corresponding to the specified <a
href="https://www.gencodegenes.org/">Gencode release</a> will be
downloaded as part of the workflow</p>
<p><code>external:</code></p>
<p>Provide paths to all optional data files here. Only files provided
will be included in the workflow.</p>
<ul>
<li><code>rnaseq</code> should be the results as output by
<code>limma::topTable()</code> or similar. Gene IDs should match those
provided in the Gencode GTF. Should be in <code>csv</code> or
<code>tsv</code> format.</li>
<li><code>features</code> can be determined by any method, with common
choices being relevant histone marks, or promoters, enhancers and
super-enhancers determined by H3K27ac marks. Features should be
non-overlapping. Must be provided in GTF format.</li>
<li><code>hic</code> HiC interactions must be provided as a
<code>bedpe</code> file</li>
<li><code>coverage</code> Tracks provided in this argument will be added
to all genomic plots showing binding peaks or differential binding.
Multiple files provided within each YAML list element will be overlaid
as a single track. There is no upper limit to the number of tracks,
however more tracks generally detract from an informative figure.</li>
</ul>
<p><code>peaks</code></p>
<p><code>macs2</code> settings are passed to
<code>macs2 callpeak</code>. Only the arguments <code>gsize</code>,
<code>fdr</code> and <code>keep_duplicates</code> are accepted. Please
see the <a
href="https://macs3-project.github.io/MACS/"><code>macs2</code>
manual</a> for more detailed explanations.</p>
<p><code>qc</code> parameters are used for determining if samples are of
a high enough quality, and how to determine consensus/oracle peaks for
each target and treatment group.</p>
<ul>
<li><code>min_prop_peaks:</code> (Default: 0.1) Samples within a
treatment group must contain more than this proportion of peaks which
overlap peaks from the “best” sample within the relevant treatment
group, i.e. the sample with the most number of peaks identified by
<code>macs2 callpeak</code></li>
<li><code>min_prop_reps</code> (Default: 0.3) When forming
<em>oracle</em> peaks within each target and treatment group, peaks must
be represented in at least this proportion of samples. This defaults to
0.3 or 1 of 2, 1 of 3, 2 of 4 etc.</li>
</ul>
</div>
</div>
<div id="colours.yml-visualisation-settings" class="section level2">
<h2><code>colours.yml</code>: Visualisation Settings</h2>
<p>Defines all colour schemes for plots. Required YAML list elements are
<code>qc</code>, <code>regions</code>, <code>direction</code> and
<code>treat</code>. Optional colours can be provided for
<code>features</code>.</p>
</div>
<div id="params.yml-additional-parameters" class="section level2">
<h2><code>params.yml</code>: Additional Parameters</h2>
<p>Default settings for annotations and enrichment testing</p>
</div>
<div id="rmarkdown.yml-html-settings" class="section level2">
<h2><code>rmarkdown.yml</code>: HTML Settings</h2>
<p>Default settings for rmarkdown documents</p>
</div>

<div class="footer">
  Workflow developed by:<br>
  Dr Stephen Pederson<br>
  Dame Roma Mitchell Cancer Research Laboratories<br>
  Adelaide Medical School, University of Adelaide, SA, Australia
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
