--- 
title: "GRAVI User Guide (v0.1.3)"
author: "Stephen Pederson"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
documentclass: book
bibliography: [book.bib, packages.bib]
url: https://steveped.github.io/GRAVI/
# cover-image: path to the social sharing image like images/cover.jpg
description: |
  This is the main documentation for the GRAVI workflow
link-citations: yes
github-repo: steveped/GRAVI
---

# Introduction

This book is the primary documentation for running the GRAVI (Gene Regulatory Analysis using Variable Inputs) workflow.
This workflow is managed using `snakemake` for running locally or on any HPC, and is designed to minimally take one ChIP target under at least two conditions.
There is no theoretical upper limit to the number of ChIP targets which can be analysed, although the practicalities of interpretation will dictate this.


In addition to $\geq 1$ ChIP targets, **optional** input includes:

- Results from a single RNA-Seq experiment
- Any type of genomic feature derived or obtained externally (`*gtf`)
- HiC Interactions (`*.bedpe`)
- Additional coverage tracks for visualisation, such as those produced by key histone marks (`*bigwig`)

The GRAVI workflow itself will

- Annotate the genome using a custom, transcript-focussed approach
- Identify peaks using `macs2 callpeak`
- Perform differential binding analysis for each ChIP target & requested comparisons
- Compare differential binding results across ChIP targets or samples (Pair-wise comparisons)
- Perform enrichment analyses at all steps of the workflow
<!-- - *TO BE COMPLETED*: Perform motif analysis based on key binding patterns within and across ChIP targets -->
- Add Rmd and html files to any local `git` repository

The primary output is a series of `html` pages generated from `rmarkdown` files, along with key figures and tables able to be shared amongst collaborators and incorporated directly into publications.

