# Files and Directories {#input-files}


```{r set-knitr-opts, echo=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, echo = FALSE)
```

```{r packages}
library(tidyverse)
library(pander)
```

## Directory Structure {#directories}


The GRAVI workflow requires a set directory structure. 
If using the template repository, as advised,  this will be mostly taken care of.
The required directory structure is

```
project_home/
 ├── analysis
 ├── config
 ├── data
 ├── docs
 ├── output
 └── workflow
```

- Rmarkdown scripts will be added and executed from the `analysis` directory
- Key configuration files are provided in the `config` directory
- *Your data* should be placed in the `data` directory as described below
- The `html` output summarising all results will be produced in the `docs` directory
- Additional output files will be placed in `output` with all figures written within the respective folder for each html page, or with `docs/assets`.
- The workflow itself is run by all code supplied in the `workflow` directory
- The complete `R Environment` for each compiled RMarkdown document is also saved in `output/envs`, and given their larger sizes, these can be deleted using `snakemake --delete-temp-output --cores 1` to conserve storage space.

## Alignments {#alignments}

The GRAVI workflow currently takes `bam` files as the primary input.
Multiple workflows exist for quality control, adapter removal and de-duplication and it is assumed that supplied reads will have been pre-processed with the above steps, then aligned to the genome of interest.

Files should be placed in the `data/bam` directory as set in `config.yml`, although this can be changed if desired.
Each ChIP target should be placed in a separate directory using the example layout as given below.
As IgG input may be shared between ChIP targets, these can all be placed in the directory `data/Input` and this is hard-wired into the workflow.

```
project_home/
 └── data
      └── bam 
           ├── TF1 
           │    ├── control_rep1.bam
           │    ├── control_rep2.bam
           │    ├── control_rep3.bam
           │    ├── treat_rep1.bam
           │    ├── treat_rep2.bam
           │    └── treat_rep3.bam
           ├── TF2 
           │    ├── control_rep1.bam
           │    ├── control_rep2.bam
           │    ├── control_rep3.bam
           │    ├── treat_rep1.bam
           │    ├── treat_rep2.bam
           │    └── treat_rep3.bam           
           └── Input
                └── pooled_input.bam 
```

## Sample Descriptions {#sample-descriptions}

The file `samples.tsv` defines the set of files which the workflow will be applied to.
Any files placed in the `data/aligned` directory, but not specified in this file will be ignored.
The desired layout should be a *tab-delimited file* (i.e. tsv).
These can be generated using Excel, Notepad++, R, Visual Studio, or any other software you are comfortable with.
A brief example would follow the layout

```{r}
tribble(
    ~sample, ~target, ~treat, ~replicate, ~input,
    "control_rep1", "TF1", "Control", 1, "pooled_input",
    "control_rep2", "TF1", "Control", 2, "pooled_input",
    "control_rep3", "TF1", "Control", 3, "pooled_input",
    "treat_rep1", "TF1", "Treatment", 1, "pooled_input",
    "treat_rep2", "TF1", "Treatment", 2, "pooled_input",
    "treat_rep3", "TF1", "Treatment", 3, "pooled_input"
) %>% 
    pander()
```


### Required columns

This file must contain all four of the columns `sample`, `target`, `treat`, `input`, in any order.
If supplied, optional columns such as `replicate`, `passage` etc can be referenced in the workflow.
As well as defining all required steps for the workflow, labels for plots will be generated from combinations of these columns.

- `sample`: This **must** be identical to the filename, but without the `bam` extension.
- `target`: This **must** be the directory name within `data/aligned` which contains the sample
- `treat`: This is used to define all comparisons 
- `input`: All files must correspond to a file in `data/aligned/Input` but without the `bam` suffix. Each sample can have a separate input sample, or input samples can be shared across all or some of the samples.

### Optional Columns

Any additional columns can be used to denote batches, or passages if running a nested/paired model.
These column names will be automatically detected at the appropriate steps of the workflow and incorporated into figures and tables.
Common column names may be `replicate` or `passage` (for cell lines)

## Additional Files {#additional-files}

Additional, optional files can also be supplied and is it customary to place these in `data/external` with paths (relative to `project_home`) added to `config.yml`.
Full details are demonstrated in Section \@ref(config-yml)
Names can be any informative name chosen by the user.

```
project_home/
 └── data
      ├── aligned 
      │    ├── TF1 
      │    ├── TF2 
      │    └── Input
      └── external
           ├── rnaseq_topTable.tsv
           ├── external_features.gtf
           ├── hic_interactions.bedpe
           ├── additional_coverage_control.bw
           └── additional_coverage_treat.bw
```

### RNA-Seq

Files provided with differential expression analysis results from a relevant RNA-Seq experiment should follow the layout as produced by `topTable()` from the `limma` package[@limma].
Gene IDs should match those in the Gencode GTF (Ensembl IDs) and should be contained in a column called `gene_id`.
Additional expected columns will be `logFC` and `FDR` or similar names which could be reasonably found by `regex` matching within the workflow.

### External Features

These must be provided as a GTF which can be prepared by any method.
The feature types should be defined in a field named `feature`.
Non-overlapping features are optimal but not essential, and this is left to the users discretion.
For example, if providing features such as enhancers and super-enhancers[@rose], it may be more sensible to provide these as mutually exclusive groups.

### HiC Interactions

Significant interactions can be sourced using any methodology, however these must be provided in `bedpe` format.

### External Coverage

Additional coverage files should be provided in `bigwig` format.
