```{r example, echo = FALSE, eval=TRUE}
## This module requires the following to be set during preparation
# target <- "ER"
# threads <- 6
# treat_levels <- c("Veh", "DHT")
```

```{r packages}
library(tidyverse)
library(glue)
library(pander)
library(BiocParallel)
library(yaml)
library(plyranges)
library(rtracklayer)
library(GenomeInfoDb)
library(Rsamtools)
library(cowplot)
library(ngsReports)
library(csaw)
library(edgeR)
library(scales)
library(quantro)
library(qsmooth)
library(magrittr)
library(statmod)
library(IHW)
library(ggrepel)
library(rlang)
library(ggside)
library(InteractionSet)
library(GenomicInteractions)
library(reactable)
library(htmltools)
library(msigdbr)
library(goseq)
```

```{r remotes, results = 'hide'}
if (!"extraChIPs" %in% rownames(installed.packages()))
  BiocManager::install("steveped/extraChIPs")

stopifnot(library(extraChIPs, logical.return = TRUE))
```

```{r options}
panderOptions("big.mark", ",")
panderOptions("missing", "")
panderOptions("table.split.table", Inf)
theme_set(
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5)
  )
)
register(MulticoreParam(workers = threads))
source(here::here("workflow", "scripts", "custom_functions.R"))
```

```{r load-config}
params <- read_yaml(here::here("config", "params.yml"))
config <- read_yaml(here::here("config", "config.yml"))
macs2_path <- here::here("output", "macs2", target)
samples <- file.path(macs2_path, "qc_samples.tsv") %>% 
  read_tsv() %>% 
  dplyr::filter(
    qc == "pass",
    treat %in% treat_levels
  ) %>% 
  mutate(treat = factor(treat, levels = treat_levels)) %>% 
  droplevels()
stopifnot(nrow(samples) > 0)
rep_col <- setdiff(
  colnames(samples), c("sample", "treat", "target", "input", "label", "qc")
)
samples[[rep_col]] <- as.factor(samples[[rep_col]])
fdr_alpha <- config$comparisons$fdr
```

```{r set-paths}
bam_path <- here::here(config$paths$bam)
stopifnot(dir.exists(bam_path))
annotation_path <- here::here("output", "annotations")
stopifnot(dir.exists(annotation_path))
out_path <- here::here("output", "differential_binding", target)
if (!dir.exists(out_path)) dir.create(out_path)
```


```{r load-annotations}
sq <- file.path(annotation_path, "seqinfo.rds") %>% 
  read_rds()
blacklist <- here::here(annotation_path, "blacklist.bed.gz") %>%
  import.bed(seqinfo = sq) %>%
  sort()

cb <- config$genome$build %>% 
  str_to_lower() %>% 
  paste0(".cytobands") 
data(list = cb)
bands_df <- get(cb)

stopifnot(
  colnames(bands_df) == c("chrom", "chromStart", "chromEnd", "name", "gieStain")
)

all_gr <- file.path(annotation_path, "all_gr.rds") %>% 
  read_rds() 
genes_gr <- all_gr$gene
id2gene <- structure(genes_gr$gene_name, names = genes_gr$gene_id)
trans_models <- file.path(annotation_path, "trans_models.rds") %>% 
  read_rds() 

external_features <- GRanges(seqinfo = sq)
if (!is.null(config$external$features)) {
  external_features <- suppressWarnings(
    import.gff(here::here(config$external$features), genome = sq)
  )
  keep_cols <- !vapply(
    mcols(external_features), function(x) all(is.na(x)), logical(1)
  )
  mcols(external_features) <- mcols(external_features)[keep_cols]
}
has_features <- length(external_features) > 0
gene_regions <- file.path(annotation_path, "gene_regions.rds") %>% 
  read_rds()
regions <- vapply(gene_regions, function(x) unique(x$region), character(1))
any_detected <- gene_regions %>% 
  vapply(function(x) any(x$detected), logical(1)) %>% 
  any()

rna_path <- here::here(config$external$rnaseq)
rnaseq <- tibble(gene_id = character())
if (length(rna_path) > 0) {
  stopifnot(file.exists(rna_path))
  if (str_detect(rna_path, "tsv$")) rnaseq <- read_tsv(rna_path)
  if (str_detect(rna_path, "csv$")) rnaseq <- read_csv(rna_path)
  if (!"gene_id" %in% colnames(rnaseq)) stop("Supplied RNA-Seq data must contain the column 'gene_id'")
  genes_gr <- subset(genes_gr, gene_id %in% rnaseq$gene_id)
}
tx_col <- intersect(c("tx_id", "transcript_id"), colnames(rnaseq))
rna_gr_col <- ifelse(length(tx_col) > 0, "transcript_id", "gene_id")
rna_col <- c(tx_col, "gene_id")[[1]]
```

```{r set-colours}
colours <- read_rds(here::here("output", "annotations", "colours.rds")) %>% 
  lapply(unlist)
region_colours <- setNames(colours$regions, regions[names(colours$regions)])
if (has_features)
  feature_colours <- setNames(
    colours$features, str_sep_to_title(names(colours$features))
  )
direction_colours <- colours$direction %>% 
  setNames(str_to_title(names(.))) %>% 
  .[names(.) %in% c("Up", "Down", "Unchanged", "Ambiguous")]
treat_colours <- colours$treat[treat_levels]
```


```{r load-hic}
hic <- GInteractions()
hic_path <- here::here(config$external$hic)
has_hic <- FALSE
if (length(hic_path) > 0)
  if (file.exists(hic_path)) {
    has_hic <- TRUE
    hic <- makeGenomicInteractionsFromFile(hic_path, type = "bedpe")
    reg_combs <- expand.grid(regions, regions) %>% 
      as.matrix() %>% 
      apply(
        MARGIN = 1, 
        function(x) {
          x <- sort(factor(x, levels = regions))
          paste(as.character(x), collapse = " - ")
        }
      ) %>% 
      unique()
    hic$regions <- anchors(hic) %>% 
      vapply(
        bestOverlap,
        y = GRangesList(lapply(gene_regions, granges)),
        character(length(hic))
      ) %>% 
      apply(MARGIN = 2, function(x) regions[x]) %>% 
      apply(
        MARGIN = 1, 
        function(x) {
          x <- sort(factor(x, levels = regions))
          paste(as.character(x), collapse = " - ")
        }
      ) %>% 
      factor(levels = reg_combs) %>%
      fct_relabel(
        str_replace_all,
        pattern = "Promoter \\([0-9kbp/\\+-]+\\)", replacement = "Promoter"
      )
    if (has_features) {
      feat_combs <- expand.grid(names(feature_colours), names(feature_colours)) %>% 
        as.matrix() %>% 
        apply(
          MARGIN = 1, 
          function(x) {
            x <- sort(factor(x, levels = names(feature_colours)))
            paste(as.character(x), collapse = " - ")
          }
        ) %>% 
        unique()
      hic$features <- vapply(
        anchors(hic),
        function(x) bestOverlap(
          x, external_features, var = "feature", missing = "no_feature"
        ),
        character(length(hic))
      )  %>% 
        apply(
          MARGIN = 1, 
          function(x) {
            x <- sort(factor(x, levels = names(feature_colours)))
            paste(as.character(x), collapse = " - ")
          }
        ) %>% 
        factor(levels = feat_combs) %>%
        fct_relabel(str_sep_to_title, pattern = "_")
    }
  }
stopifnot(is(hic, "GInteractions"))
seqlevels(hic) <- seqlevels(sq)
seqinfo(hic) <- sq
hic <- hic[!overlapsAny(hic, blacklist)]
```


```{r labeller}
facet_labeller <- as_labeller(
  c(
    c(
      `TRUE` = "Consensus Peak",
      `FALSE` = "No Consensus Peak"
    ),
    structure(treat_levels, names = treat_levels),
    c(
      Up = glue("Increased {target} Binding"),
      Down = glue("Decreased {target} Binding")
    ),
    c(
      AveExpr = "logCPM", logFC = "logFC"
    )
  )
)
```

```{r define-outputs}
all_out <- list(
  results = file.path(
    out_path, 
    glue("{target}_{treat_levels[[1]]}_{treat_levels[[2]]}-differential_binding.rds")
  ),
  csv = file.path(
    out_path, 
    glue("{target}_{treat_levels[[1]]}_{treat_levels[[2]]}-differential_binding.csv.gz")
  ),
  windows  = file.path(
    out_path,
    glue("{target}_{treat_levels[[1]]}_{treat_levels[[2]]}-filtered_windows.rds")
  ),
  up_regions = file.path(
    out_path, glue("{target}_{treat_levels[[1]]}_{treat_levels[[2]]}-up.bed")
  ),
  down_regions = file.path(
    out_path, glue("{target}_{treat_levels[[1]]}_{treat_levels[[2]]}-down.bed")
  ),
  de_genes = file.path(
    out_path, glue("{target}_{treat_levels[1]}_{treat_levels[2]}-DE_genes.csv")
  ),
  renv = here::here(
    "output/envs",
    glue(
      "{target}_{treat_levels[[1]]}_{treat_levels[[2]]}-differential_binding.RData"
    )
  )
)
```

# Outline

## Differential Binding

This step of the workflow uses a sliding window approach to differential binding, as a logical extension to the methods suggested and implemented in the `csaw` package [@LunSmythF1000].
Several differences are used in the approach here:

1. Consensus Peaks derived from `macs2 callpeak` [@Zhang18798982] results are used to determine inclusion/exclusion values for sliding windows
2. Smooth Quantile Normalisation [@HicksSQN2017] is used on the set of logCPM values obtained from retained sliding windows. This was chosen given that many transcription factors can show vastly different cytoplasmic/nuclear distributions across treatments
3. The *limma-trend* method [@LawVoom2014] is used alongside `treat` [@McCarthyTreat2009] for detection of differential binding. This should correctly control the FDR, and allows the specification as a percentage of a suitable threshold for the estimated change in binding, below which we are not interested in any changed signal.
3. Independent Hypothesis Weighting (IHW) [@IgnatiadisIHW2016] is additionally used to improve the power of the results. Under this strategy, p-values are partitioned based on an independent variable, such as genomic region (`config:comparisons:ihw: "regions"`), external features (`config:comparisons:ihw: "features"`) or the presence of additional ChIP targets (`config:comparisons:ihw: "targets"`).

The workflow also depends heavily on the function implemented in the Bioconductor package [extraChIPs](https://bioconductor.org/packages/release/bioc/html/extraChIPs.html)

## Enrichment Analysis

Beyond the simple analysis of differential binding, peaks are mapped to genes and enrichment testing is performed on the following:

1. Genes mapped to any window with detected `r target` are compared to all genes not mapped to any window
2. Genes mapped to *all differentially bound windows* are compared to genes mapped to windows which are not differentially bound
3. Genes mapped to windows with *increasing* `r target` binding are compared to genes mapped to windows which are not differentially bound
4. Genes mapped to windows with *decreasing* `r target` binding are compared to genes mapped to windows which are not differentially bound

Enrichment testing is performed using `goseq` [@YoungGoseq2010] using the number of peaks mapped to each gene as the source of any sampling bias, except when comparing genes mapped to any window.
For this case, gene width is used to capture any sampling bias.
If RNA-seq data is provided, only genes contained within this dataset are considered as detected, and all gene-level analysis is restricted to this set of genes.

## Incorporation with RNA-Seq

If RNA-Seq data is provided, candidate genes for mapping will be restricted to the set of genes within the RNA-seq dataset as the set of genes expressed, or detected within the relevant tissue or samples.
Any association between differentially expressed genes and differentially bound sites will be assessed using Gene Set Enrichment Analysis [@SubramanianGsea2005], as implemented in the `fgsea` package [@KorotkevichFgsea2019].
The sets of genes associated with changed binding will be subset by regions and any provided external features, and these novel gene-sets will be used to test for enrichment within the RNA-Seq results.
ChIP-seq derived gene-sets will be tested for differential expression using genes ranked directionally and by significance alone.

# Data Preparation

```{r import-peaks}
consensus_peaks <- here::here(macs2_path, "consensus_peaks.bed") %>%
	import.bed(seqinfo = sq)
```


```{r bfl}
bfl <- bam_path %>%
  file.path(target, glue("{samples$sample}.bam")) %>%
  c(
    file.path(bam_path, "Input", glue("{unique(samples$input)}.bam"))
  ) %>% 
  BamFileList() %>%
  setNames(c(samples$sample, unique(samples$input)))
```


```{r get-bfl-summary}
ys <- 1000
bfl_summary <- bfl %>%
	bplapply(
		function(x) {
			yieldSize(x) <- ys
			sbp <- ScanBamParam(what = c("qwidth", "mapq"))
			open(x)
			vals <- scanBam(x, param = sbp)[[1]]
			close(x)
			list(as_tibble(vals))
		}
	) %>% 
	as_tibble() %>%
	pivot_longer(everything(), names_to = "sample") %>%
   	unnest(value) %>% 
	pivot_longer(
		cols = c("qwidth", "mapq"),
		names_to = "stat",
		values_to = "value"
	) %>%
	left_join(samples, by = "sample") %>% 
	mutate(
		across(
			all_of(c("target", "label", "treat")),
			str_replace_na, replacement = "Input"
		),
		treat = factor(treat, levels = c("Input", treat_levels)),
		label = fct_inorder(label)
	) %>% 
	split(f = .$stat)
anyDups <- bplapply(
	bfl,
	function(x) {
		sbp <- ScanBamParam(
			flag = scanBamFlag(isDuplicate = TRUE),
			which = GRanges(sq)[which.min(seqlengths(sq))],
			what = "qname"
			)
		length(scanBam(x, param = sbp)[[1]]$qname)  > 0
	}
) %>%
  unlist()
anyPE <- bplapply(
	bfl,
	function(x){
		yieldSize(x) <- ys
		open(x)
		flag <- scanBam(x, param=ScanBamParam(what="flag"))[[1]]$flag
		close(x)
		any(bamFlagTest(flag, "isPaired"))
	}
) %>% 
  unlist()
```


Taking the first `r comma(ys)` alignments, a brief inspection of the Bam Files revealed:

- Aligned read lengths ranged between `r glue_collapse(range(bfl_summary$qwidth$value), last = " and ")`. The median length was `r median(bfl_summary$qwidth$value)`.
- MAPQ scores of aligned reads ranged between `r glue_collapse(range(bfl_summary$mapq$value), last = " and ")`. The median score was `r median(bfl_summary$mapq$value)`.
- `r ifelse(any(anyDups), glue("Duplicate alignments were detected in ", glue_collapse(names(anyDups), sep = ", ", last = " & ")), "All bam files have been de-duplicated")`
- `r case_when(!any(anyPE) ~ "All alignments were single-end", all(anyPE) ~ "All Alignments were paired-end", TRUE ~ "Bam files appear to be a mixture of single-end and paired-end samples")`


## Sliding Window Counts

```{r window-counts}
macs2_merged_logs <- file.path(
  macs2_path, glue("{treat_levels}_merged_callpeak.log")
	) %>%
	importNgsLogs()
fl <- max(macs2_merged_logs$fragment_length)
rp <- readParam(
  pe = ifelse(any(anyPE), "both", "none"),
  dedup = any(anyDups),
  restrict = seqnames(sq),
  discard = blacklist,
)
win_step <- 10*(1 + fl %/% 30)
win_size <- 3*win_step
window_counts <- windowCounts(
  bam.files = bfl,
  spacing = win_step,
  width = win_size,
  ext = fl,
  filter = length(bfl) - 1,
  param = rp,
  BPPARAM = bpparam()
)
colData(window_counts) <- colData(window_counts) %>% 
  as_tibble(rownames = "sample") %>% 
  dplyr::select(all_of(c("sample", "bam.files", "totals", "ext", "rlen"))) %>% 
  left_join(samples, by = "sample") %>% 
  mutate(
    treat = fct_explicit_na(treat, "Input"), 
    target = str_replace_na(target, "Input")
  ) %>% 
  as.data.frame() %>% 
  column_to_rownames("sample") %>% 
  DataFrame()
window_counts <- sortSeqlevels(window_counts)
seqinfo(window_counts) <- sq
```

Using the macs2-estimated fragment length of `r fl`nt, a set of genomic sliding windows were defined using a window size of `r win_size`bp, sliding in increments of `r win_step`bp.
With the exclusion of black-listed regions, all alignments within each window were counted for each `r target`-associated sample, and all relevant input samples.
Any windows with fewer than `r length(bfl) - 1` alignments (i.e.1 read/sample) across all samples were discarded, leaving a total of `r comma(nrow(window_counts))` sliding windows, covering `r percent(sum(width(reduce(granges(window_counts)))) / sum(width(GRanges(sq))))` of the reference genome.

These windows were then filtered using the `dualFilter()` function from `extraChIPs`, discarding windows with low counts.
Under this approach, two thresholds are determined above which windows are retained, and with values chosen to return `r percent(config$comparisons$filter_q)` of sliding windows which overlap a macs2-derived consensus peak.
These values are set to filter based on 1) signal relative to input over an extended range and, 2) overall signal level.
Both filtering strategies rely on the infrastructure provided by `csaw` [@LunSmythCsaw2014].


```{r filtered-counts}
filtered_counts <- dualFilter(
  x = window_counts[, samples$sample],
  bg = window_counts[, samples$input],
  ref = consensus_peaks,
  keep.totals = TRUE,
  q = config$comparisons$filter_q
)
colData(filtered_counts) <- droplevels(colData(filtered_counts))
n_max <- min(5e4, nrow(filtered_counts))
```

```{r tab-filtered-windows}
list(
  `Genomic Windows` = window_counts,
  `Retained Windows` = filtered_counts,
  `Consensus Peaks` = consensus_peaks
) %>% 
  lapply(granges) %>% 
  lapply(
    function(x) {
      tibble(
        N = comma(length(x)),
        `Total Width (kb)` = comma(sum(width(reduce(x))) / 1e3),
        `Median Width (bp)` = median(width(x))
      )
    }
  ) %>% 
  lapply(list) %>% 
  as_tibble() %>% 
  pivot_longer(everything(), names_to = "Dataset") %>% 
  unnest(everything()) %>% 
  left_join(
    tibble(
      Dataset = c("Retained Windows", "Consensus Peaks"),
      `Unique (kb)` = c(
        round(sum(width(setdiff(granges(filtered_counts), consensus_peaks))) / 1e3, 1),
        round(sum(width(setdiff(consensus_peaks, granges(filtered_counts)))) / 1e3, 1)
      ),
      `% Unique` = percent(
        1e3*`Unique (kb)` / c(
          sum(width(reduce(granges(filtered_counts)))),
          sum(width(consensus_peaks))
        ),
        0.1
      )
    )
  ) %>% 
  pander(
    justify = "lrrrrr",
    caption = paste(
      "A dual filtering strategy was used based on retaining", 
      percent(config$comparisons$filter_q, 0.1), 
      "of genomic windows overlapping the consensus peaks identified by ", 
      "`macs2 callpeak` on merged samples. This approach combined both ", 
      "1) expression percentiles, and 2) signal strength in relation to the input sample.",
      "The complete set of sliding windows covered the majority of the genome,",
      "whilst those retained after filtering were focussed on strong binding",
      "signal. Consensus peaks were as identified by `macs2 callpeak` in a", 
      "previous step of the workflow.",
      "Importantly, consensus peaks are non-overlapping, whilst the other two",
      "datasets are derived from overlapping sliding windows.",
      "This strategy of filtering the set of initial sliding windows retained",
      percent(
        1 - sum(width(setdiff(consensus_peaks, rowRanges(filtered_counts)))) / sum(width(consensus_peaks))
      ), 
      "of the genomic regions covered by consensus peaks, with", 
      percent(
        sum(width(setdiff(rowRanges(filtered_counts), consensus_peaks))) / sum(width(reduce(granges(filtered_counts)))))
      , 
      "of the retained windows being outside genomic regions covered by consensus peaks.",
      ifelse(
        n_max == 5e4, 
        "For subsequent density and RLE plots, a subsample of 50,000 genomic regions will be used for speed.", 
        ""
      )
    )
  )
```


## Normalisation {.tabset}

The `quantro` test was first applied [@HicksQuantro2015] to determine if treatment-specific binding distributions were found in the data.
Whilst this may not always be the case, the robustness of smooth-quantile normalisation (SQN) [@HicksSQN2017] will be applicable if data is drawn from different or near-identical distributions, and this method was applied grouping data by treatment.

```{r qtest, results='asis'}
qtest <- quantro(
  assay(filtered_counts, "logCPM"),
  groupFactor = filtered_counts$treat
)
pander(
  anova(qtest), 
  caption = paste(
    "*Results from qtest suggesting that the two treatment groups are drawn from",
    ifelse(
      qtest@anova$`Pr(>F)`[[1]] < 0.05,
      "different distributions.*",
      "the same distribution.*"
    )
  )
)
```


```{r add-qsmooth}
qs <-qsmooth(
  assay(filtered_counts, "logCPM"), group_factor = filtered_counts$treat
)
assay(filtered_counts, "qsmooth") <- qsmoothData(qs)
```


### QSmooth Weights

```{r plot-qs-weights, fig.height = 6, fig.cap = "*Quantile-specific weights used by the Smooth-Quantile normalisation. Low weights indicate signal quantiles which appear to be more specific within a group, whilst higher weights indicate similarity between groups.*"}
qsmoothPlotWeights(qs, xLab = "Quantiles", yLab = "Weights", mainLab = "QSmooth Weights")
```


### Density Plot: All Sliding Windows

```{r plot-density-windows, fig.cap = "*Density plot for __all windows prior to the selection of windows__ more likely to contain true signal. Retained windows will be those at the upper end, whilst discarded windows will be at the lower end.*"}
plotAssayDensities(
  window_counts, trans = "log1p", n_max = n_max, colour = "treat",
  linetype = "target"
) +
  scale_colour_manual(values = colours$treat) + 
  labs(
    x = "log(counts + 1)", y = "Density", colour = "Treat", linetype = "Target"
  )
```

### logCPM Densities

```{r plot-logcpm-densities, fig.height = 6, fig.cap = "*logCPM distributions for all retained windows A) prior to smooth quantile normalisation, and 2) after SQ normalisations*"}
a <- filtered_counts %>% 
  plotAssayDensities("logCPM", colour = "treat", n_max = n_max) +
  scale_colour_manual(values = treat_colours) +
  labs(colour = "Treat")
b <- filtered_counts %>% 
  plotAssayDensities("qsmooth", colour = "treat", n_max = n_max) +
  scale_colour_manual(values = treat_colours) +
  labs(
    x = "Normalised logCPM", colour = "Treat"
  )
plot_grid(
  a + theme(legend.position = "none"), 
  b + theme(legend.position = "none"),
  get_legend(b),
  nrow = 1,
  rel_widths = c(5, 5, 1),
  labels = c("A", "B")
)
```

### RLE (Post-SQN)

```{r plot-rle-normalised, fig.height=6, fig.cap = glue("*RLE plot showing logCPM values after Smooth Quantile normalisation. RLE values were calculated within each treatment group to account for the potentially different binding dynamics of {target}.*")}
filtered_counts %>% 
  plotAssayRle(
  "qsmooth", fill = "treat", rle_group = "treat", x_col = "label", n_max = n_max
) +
  geom_hline(yintercept = 0) + 
  facet_grid(.~treat, scales = "free_x", space = "free_x") +
  scale_fill_manual(values = treat_colours) +
  labs(x = "Sample", fill = "Treat")
```


### PCA

```{r plot-pca, fig.height = 5, fig.cap = "*PCA plots for logCPM values A) before and B) after Smooth Quantile normalisation*"}
a <- plotAssayPCA(filtered_counts, "logCPM", colour = "treat", label = "label") +
  scale_colour_manual(values = treat_colours)
b <- plotAssayPCA(filtered_counts, "qsmooth", colour = "treat", label = "label") +
  scale_colour_manual(values = treat_colours) 
plot_grid(
  a + theme(legend.position = "none"), b + theme(legend.position = "none"), 
  get_legend(b + labs(colour = "Treat")), rel_widths = c(1, 1, 0.2), 
  nrow = 1, labels = c("A", "B")
  )
```


# Differential Binding Analysis

## Primary Analysis {.tabset}

```{r fit-model}
X <- model.matrix(~treat, data = colData(filtered_counts)) %>%
  set_colnames(str_remove(colnames(.), "treat"))
colData(filtered_counts)$design <- X
paired_cors <- block <- NULL
if (config$comparisons$paired) {
  block <- colData(filtered_counts)[[rep_col]]
  set.seed(1e6)
  ind <- sample.int(nrow(filtered_counts), n_max, replace = FALSE)
  paired_cors <- duplicateCorrelation(
    object = assay(filtered_counts, "qsmooth")[ind, ],
    design = X,
    block = block
  )$consensus.correlation
}
fc <- ifelse(is.null(config$comparisons$fc), 1, config$comparisons$fc)
fit <- assay(filtered_counts, "qsmooth") %>%
  lmFit(design = X, block = block, correlation = paired_cors) %>% 
  treat(fc = fc, trend = TRUE, robust = FALSE)
fit_mu0 <- assay(filtered_counts, "qsmooth") %>%
  lmFit(design = X, block = block, correlation = paired_cors) %>% 
  treat(fc = 1, trend = TRUE, robust = FALSE)
res_cols <- c("logFC", "AveExpr", "t", "P.Value", "fdr")
rowData(filtered_counts) <- rowData(filtered_counts) %>% 
  .[!colnames(.) %in% res_cols] %>% 
  cbind(
    fit %>% 
      topTable(sort.by = "none", number = Inf) %>% 
      as.list %>% 
      setNames(res_cols) %>% 
      c(
        list(
          p_mu0 = topTable(fit_mu0, sort.by = "none", number = Inf)$P.Value
        )
      )
  ) 
```


After SQ-normalisation of logCPM values, the `limma-trend` method [@LawVoom2014] was applied to all retained windows.
A simple linear model was fitted taking `r treat_levels[[1]]` as the baseline and estimating the effects of `r treat_levels[[2]]` on `r target` binding within each sliding window, incorporating a trended prior-variance.
After fitting, hypothesis testing was performed testing:

$$
H_0: -\lambda < \mu < \lambda
$$
against
$$
H_A: |\mu| > \lambda
$$
where $\mu$ represents the true mean change in `r target` binding.
The value $\lambda =$ `r round(log2(fc), 2)` was chosen as this corresponds to a `r percent(fc - 1)` change in detected signal on the log~2~ scale.
This is known as the `treat` method [@McCarthyTreat2009], and p-values were obtained for each initial window, before merging adjoining windows.

For subsequent classification when combining across multiple differential binding results, an additional p-value testing $\mu = 0$ was added, however this is not used for detection of sites showing evidence of altered `r target` binding.


```{r merged-results}
merged_results <- mergeByCol(
  filtered_counts, col = "AveExpr", merge_within = 2 * win_step,
  inc_cols = "p_mu0"
) %>% 
  mutate(
    overlaps_ref = overlapsAny(., consensus_peaks),
    direction = case_when(
      n_up == 0 & n_down == 0 ~ "Unchanged",
      logFC > 0 & n_up > n_down ~ "Up",
      logFC < 0 & n_up < n_down ~ "Down",
      TRUE ~ "Ambiguous"
    ) %>% 
      factor(levels = c("Up", "Down", "Ambiguous", "Unchanged")) %>% 
      droplevels()
  ) 
```

After selection of the `r comma(nrow(filtered_counts))` sliding windows (`r win_size`bp), these were merged into `r comma(length(merged_results))` genomic regions with size ranging from `r glue_collapse(range(width(merged_results)), last = " to ")`bp, with the median width being `r median(width(merged_results))`bp.
Representative estimates of differential binding (i.e. logFC) were taken from the sliding window with the *highest average signal* across all samples, within each set of windows to be merged.
Similarly, p-values from the above tests were also selected from the same sliding window as representative of the merged region [@LunSmythF1000].
The number of windows showing increased or decreased binding within each merged region were also included, by counting those within each set of windows with p-values lower than the selected window.
FDR-adjustment was performed with `r comma(sum(merged_results$P.Value_fdr < fdr_alpha))` regions (`r percent(mean(merged_results$P.Value_fdr < fdr_alpha), accuracy = 0.1)`) showing significance, based on an FDR-adjusted p-value alone.

### Merged Windows

```{r plot-merged-size, fig.height=7, fig.cap = glue("*Comparison of {target} signal intensity (logCPM) and merged window size (kb) for windows which overlap a macs2 consensus peak and those which don't. A mall amount of jitter has been added to the y-axis in the lower-left panel.*")}
merged_results %>% 
  mutate(kb = width / 1e3) %>% 
  as_tibble() %>% 
  mutate(
    overlaps_ref = unlist(facet_labeller(overlaps_ref)) %>% 
      str_replace_all(" Peak", "\nPeak")
  ) %>% 
  ggplot(aes(AveExpr, kb)) +
  geom_jitter(
    aes(colour = overlaps_ref), 
    alpha = 0.6, width = 0, height = 0.04
  ) +
  geom_xsideboxplot(
    aes(y = overlaps_ref, fill = overlaps_ref), orientation = "y"
  ) +
  geom_ysideboxplot(
    aes(x = overlaps_ref, fill = overlaps_ref), orientation = "x"
  ) +
  scale_xsidey_discrete() +
  scale_ysidex_discrete() +
  scale_colour_brewer(palette = "Set1") +
  scale_fill_brewer(palette = "Set1") +
  labs(
    x = "logCPM", y = "Merged Window Size (kb)",
    fill = glue("{target} Macs2 Peak"), colour = glue("{target} Macs2 Peak")
  ) +
  theme(
    axis.title.x = element_text(hjust = 0.45),
    axis.title.y = element_text(hjust = 0.4),
    ggside.axis.text.y = element_text(hjust = 0.5),
    ggside.panel.scale.x = 0.3,
    ggside.panel.scale.y = 0.25,
    legend.position = c(0.9, 0.9)
  )
```

### Mean-Variance Relationship

```{r plot-mean-var, fig.height = 6, fig.cap = "*Mean-variance for all retained windows. Black dots indicate the estimated standard deviation for each window as a function of signal. Red dots are the shrunken estimates after applying limma-trend [@LawVoom2014], whilst the loess curve through the initial estimates is shown in blue.*"}
tibble(A = fit$Amean, pre = fit$sigma, post = sqrt(fit$s2.post)) %>% 
  sample_n(n_max) %>% 
  pivot_longer(c("pre", "post"), names_to = "step", values_to = "sigma") %>%
  mutate(
    step = factor(step, levels = c("pre", "post"))
  ) %>% 
  ggplot(aes(A, sigma, colour = step)) +
  geom_point(size = 1, alpha = 0.5) +
  geom_smooth(
    se = FALSE, method = "loess", formula = y ~ x,
    data = . %>% dplyr::filter(step == "pre"),
    colour = "blue",
    show.legend = FALSE
  ) +
  labs(
    x = "logCPM", y = expression(hat(sigma)), colour = "limma-trend"
  ) +
  scale_colour_manual(values = c(pre = "black", post = "red")) +
  theme(axis.title.y = element_text(angle = 0, vjust = 0.5))
```


## Independent Hypothesis Weighting {.tabset}

```{r set-ihw}
ihw_opts <- c("regions", "features", "targets")
ihw_method <- c(
  intersect(str_to_lower(config$comparisons$ihw), ihw_opts), "none"
)
ihw_method <- ihw_method[[1]]
fdr_column <- str_subset(names(mcols(merged_results)), "fdr|FDR")
## Add status (direction) for the ihw = "none" possibility
merged_results <- merged_results %>% 
  mutate(
    status = case_when(
      !!sym(fdr_column) < fdr_alpha ~ as.character(direction),
      !!sym(fdr_column) >= fdr_alpha ~ "Unchanged"
    ) %>% 
      factor(levels = c(levels(direction), "Unchanged")) %>% 
      droplevels()
  )
```

`r ifelse(ihw_method == "none", "No independent Hypothesis Weighting was specified", "")`

```{r ihw-regions, eval = ihw_method == "regions", echo=TRUE, child = here::here('workflow/modules/ihw_regions.Rmd')}
```

```{r ihw-features, eval = ihw_method == "features", echo=TRUE, child = here::here('workflow/modules/ihw_features.Rmd')}
```

```{r ihw-targets, eval = ihw_method == "targets", echo=TRUE, child = here::here('workflow/modules/ihw_targets.Rmd')}
```

## Mapping Windows

```{r map-to-regions}
merged_results <- merged_results %>% 
  mutate(
    region = bestOverlap(
      ., 
      gene_regions %>% 
        lapply(select, region) %>% 
        GRangesList() %>% 
        unlist(),
      var = "region"
    ) %>% 
      factor(levels = regions)
  )
```

```{r map-to-features, eval = has_features, echo = has_features}
merged_results <- merged_results %>% 
  mutate(
    feature = bestOverlap(
      ., external_features, "feature", missing = "no_feature"
    ) %>% 
      factor(levels = names(colours$features)) %>% 
      fct_relabel(str_sep_to_title)
  )
```

Merged windows were first mapped to the gene-centric region with the largest overlap.
`r ifelse(has_features, "Windows were then mapped to the external features provided in the same manner", "")`.

```{r map-to-genes}
merged_results <- merged_results %>%
  mapByFeature(
    genes = genes_gr, 
    prom = GRangesList(
      ## Add any external features which appear to be promoters
      ifelse(
        "feature" %in% colnames(mcols(external_features)), 
        list(granges(subset(external_features, str_detect(feature, "[Pp]rom")))), 
        list(GRanges())
      )[[1]], 
      granges(gene_regions$promoters)) %>% 
      unlist() %>% 
      reduce(),
    enh = ifelse(
      "feature" %in% colnames(mcols(external_features)), 
      list(granges(subset(external_features, str_detect(feature, "[Ee]nhanc")))), 
      list(GRanges())
    )[[1]],
    gi = hic, 
    gr2gene = params$mapping$gr2gene,
    prom2gene = params$mapping$prom2gene,
    enh2gene = params$mapping$enh2gene,
    gi2gene = params$mapping$gi2gene
  )
```

In addition to statistical analysis, the regions produced by merging windows were mapped to all annotated genes.
Promoters were the union of all regions defined based on transcript-level annotation, and any external features which were defined as promoters.
`r ifelse(has_features, glue("Enhancers were any regions defined in {basename(config$external$features)} as enhancers."), glue(""))`
`r ifelse(length(hic) > 0, glue("A set of {length(hic)} HiC-interactions was also included to enable the use of long-range interactions during the mappig process."), "")`

These features were used to map merged windows to features using the process defined in the function `extraChIPs::mapByFeature()`:

1. Ranges *overlapping a promoter* are assigned to genes `r ifelse(params$mapping$prom2gene == 0, "directly overlapping", paste("within", params$mapping$prom2gene, "bp of"))` that specific promoter
2. Ranges *overlapping an enhancer* are assigned to all genes `r ifelse(params$mapping$enh2gene == 0, "directly overlapping", paste("within", round(params$mapping$enh2gene/1e3, 1), "kb of"))` the enhancer
3. Ranges *overlapping a long-range interaction* are assigned to all genes `r ifelse(params$mapping$gi2gene == 0, "directly overlapping", paste("within", round(params$mapping$gi2gene/1e3, 1), "kb of"))` either end of the interaction
4. Ranges with *no gene assignment from the previous steps* are assigned to all overlapping genes, or the nearest gene within `r round(params$mapping$gr2gene/1e3, 1)`kb

Notably, genes are only passed to step 4 if no gene assignment has been made in steps 1, 2 or 3.
For visualisation purposes, only genes which were considered as detected in any provided RNA-Seq data will be shown as the mapping targets.

# Results 

## Summary Plots {.tabset}

```{r make-heatmaps}
profile_width <- 5e3
n_bins <- 100
bwfl <- file.path(macs2_path, glue("{treat_levels}_merged_treat_pileup.bw")) %>% 
  BigWigFileList() %>% 
  setNames(treat_levels) 
sig_ranges <- merged_results %>% 
  filter(!!sym(fdr_column) < fdr_alpha) %>% 
  colToRanges("keyval_range") %>% 
  splitAsList(f = .$direction) %>% 
  .[vapply(., length, integer(1)) > 0] 
sig_profiles <- bplapply(
  sig_ranges, 
  function(x) getProfileData(
    bwfl, x, upstream = profile_width / 2, bins = n_bins
  ),
  BPPARAM = MulticoreParam(length(sig_ranges))
)
## Currently, this is set to CPM as the only possible output of 
## extraChIP::getProfileData. Until this is rectified, manually log-transform
sig_profiles <- sig_profiles %>% 
  bplapply(
    function(x) {
      temp_gr <- unlist(x)
      temp_gr$profile_data <- lapply(
        temp_gr$profile_data,
        function(pd) {
          pd$score <- log2(pd$score + 1)
          pd
        }
      )
      temp_gr$profile_data <- as(temp_gr$profile_data, "SplitDataFrameList")
      out <- split(temp_gr, names(temp_gr))
      out <- lapply(out, setNames, NULL)[names(x)]
      GRangesList(out)
    },
    BPPARAM = MulticoreParam(length(sig_ranges))
  ) 
profile_heatmaps <- sig_profiles %>% 
  bplapply(
    plotProfileHeatmap,
    profileCol = "profile_data", 
    colour = "name",
    xLab = "Distance from Centre (bp)",
    fillLab = "logCPM",
    BPPARAM = MulticoreParam(length(sig_ranges))
  ) 
fill_range <- profile_heatmaps %>% 
  lapply(function(x) x$data[,"score"]) %>% 
  unlist() %>%
  range()
sidey_range <- profile_heatmaps %>% 
  lapply(function(x) x$layers[[3]]$data$y) %>% 
  unlist() %>% 
  range()
profile_heatmaps <- profile_heatmaps %>% 
  lapply(
    function(x) {
      x + 
        scale_fill_viridis_c(
          option = "inferno", direction = -1, limits = fill_range
        ) +
        scale_xsidey_continuous(limits = sidey_range) +
        scale_colour_manual(values = treat_colours) +
        labs(
          x = "Distance from Centre (bp)",
          fill = "logCPM", 
          colour = "Treat"
        ) +
        theme(
          strip.text.y = element_text(angle = 0)
        )
    }
  )
```


### MA Plot

```{r plot-ma, fig.cap = glue("*MA plot showing the status of each window under consideration. The two most extreme regions are labelled by region and any associated genes, whilst the overall pattern of association between signal level (logCPM) and changed signal (logFC) is shown as the blue curve. Windows are plotted in separate panels based on those which overlap a `macs2` consensus peak.*")}
if (!"Ambiguous" %in% levels(merged_results$status)) 
  direction_colours <- direction_colours[names(direction_colours) != "Ambiguous"]
merged_results %>% 
  as_tibble() %>%
  ggplot(aes(AveExpr, logFC)) +
  geom_point(aes(colour = status), alpha = 0.5) +
  geom_smooth(se = FALSE, method = "gam", formula = y ~ s(x, bs = "cs")) +
  geom_hline(yintercept = 0) +
  geom_label_repel(
    aes(y = 1.1*logFC, label = annot, colour = status),
    data= . %>%
      dplyr::filter(logFC == max(logFC) | logFC == min(logFC)) %>%
      mutate(
        detected = gene_name %>%
          lapply(paste, collapse = ", ") %>%
          unlist() %>%
          str_replace("^NA$", "") %>%
          str_wrap(width = 40),
        range = ifelse(detected == "", range, paste0(range, "\n")),
        annot = glue("{range}{detected}")
      ),
    fill = rgb(1, 1, 1, 0.2),
    size = 3,
    show.legend = FALSE
  ) +
  facet_grid(
    .~overlaps_ref, 
    scales = "free_x", space = "free_x",
    labeller = facet_labeller
  ) +
  scale_colour_manual(values = direction_colours) +
  labs(
    x = "Ave Signal (logCPM)",
    colour = "Status"
  ) 
```


### Volcano Plot

```{r plot-volcano, fig.cap = glue("*Volcano plot showing regions with evidence of differential {target} binding. The most significant regions are labelled along with any genes these regions are mapped to.*")}
merged_results %>% 
  as_tibble() %>% 
  ggplot(aes(logFC, -log10(P.Value))) +
  geom_point(aes(colour = status), alpha = 0.6) +
  geom_label_repel(
    aes(label = annot, colour = status),
    data= . %>%
      arrange(sign(logFC)*log10(P.Value)) %>% 
      dplyr::slice(1:2) %>% 
      dplyr::filter(logFC > 1) %>%
      mutate(
        detected = gene_name %>%
          lapply(paste, collapse = ", ") %>%
          unlist() %>%
          str_replace("^NA$", "") %>%
          str_wrap(width = 40),
        range = ifelse(detected == "", range, paste0(range, "\n")),
        annot = glue("{range}{detected}")
      ),
    fill = rgb(1, 1, 1, 0.2),
    size = 3,
    show.legend = FALSE
  ) +
  geom_label_repel(
    aes(label = annot, colour = status),
    data= . %>%
      arrange(-sign(logFC)*log10(P.Value)) %>% 
      dplyr::slice(1:2) %>% 
      dplyr::filter(logFC < -1) %>% 
      mutate(
        detected = gene_name %>%
          lapply(paste, collapse = ", ") %>%
          unlist() %>%
          str_replace("^NA$", "") %>%
          str_wrap(width = 40),
        range = ifelse(detected == "", range, paste0(range, "\n")),
        annot = glue("{range}{detected}")
      ),
    fill = rgb(1, 1, 1, 0.2),
    size = 3,
    show.legend = FALSE
  ) +
  scale_colour_manual(values = direction_colours) +
  labs(
    x = "logFC",
    y = expression(paste(-log[10], "p")),
    colour = "Status"
  )
```

`r ifelse("Up" %in% names(profile_heatmaps), "### Heatmap: Increased Binding\n", "")`

```{r plot-up-heatmap, eval = "Up" %in% names(profile_heatmaps), echo = "Up" %in% names(profile_heatmaps), fig.cap = glue("*Heatmap and histogram for all regions considered to show evidence of increased {target} binding in the {treat_levels[[2]]} group. A total of {comma(length(sig_ranges[['Up']]))} regions were in this group.*"), fig.height = min(16, length(sig_ranges$Up)/1.5e3 + 6)}
profile_heatmaps$Up
```


`r ifelse("Down" %in% names(profile_heatmaps), "### Heatmap: Decreased Binding\n", "")`

```{r plot-down-heatmap, eval = "Down" %in% names(profile_heatmaps), echo = "Down" %in% names(profile_heatmaps), fig.cap = glue("*Heatmap and histogram for all regions considered to show evidence of decreased {target} binding in the {treat_levels[[2]]} group. A total of {comma(length(sig_ranges[['Down']]))} regions were in this group.*"), fig.height = min(16, length(sig_ranges$Down)/1.5e3 + 6)}
profile_heatmaps$Down
```

### Results By Chromosome

```{r plot-res-by-chromosome, fig.cap = "*Results for differential binding separated by chromosome*"}
merged_results %>% 
  select(overlaps_ref, status) %>% 
  as.data.frame() %>% 
  mutate(
    merge_status = fct_collapse(
      status,
      Up = "Up",
      Down = "Down",
      `Ambiguous/Unchanged` = c("Ambiguous", "Unchanged")
    )
  ) %>% 
  ggplot(aes(seqnames, fill = status, alpha = overlaps_ref)) +
  geom_bar() +
  facet_grid(merge_status~., scales = "free_y") +
  scale_y_continuous(expand = expansion(c(0, 0.05)), labels = comma) +
  scale_fill_manual(values =  direction_colours) +
  scale_alpha_manual(values = c(0.5, 1)) +
  labs(
    x = "Chromosome", y = "Number of Windows", 
    fill = "Status", alpha = "Macs2 Peak"
  )
```

### Signal By Region

```{r boxplot-regions, fig.height = 7, fig.cap = "*Distributions of overall signal (logCPM) and changed signal (logFC) for each genomic region.*"}
merged_results %>% 
  select(AveExpr, logFC, region) %>% 
  as_tibble() %>% 
  mutate(
    x = fct_relabel(
      region, str_replace_all, pattern = " \\(", replacement = "\n("
    ),
  ) %>% 
  pivot_longer(cols = c("AveExpr", "logFC")) %>% 
  ggplot(aes(x, value, fill = region)) +
  geom_boxplot() +
  facet_grid(
    name~., scales = "free", labeller = facet_labeller, switch = "y"
  ) +
  scale_fill_manual(values = region_colours) +
  labs(x = "Region", y = c(), fill = "Overlaps") 
```

### Differential Binding By Region

```{r plot-db-regions, fig.height = 7, fig.cap = glue("*Merged windows considered as showing differential {target} binding across all genomic regions.*")}
merged_results %>% 
  select(status, region) %>% 
  as_tibble() %>% 
  mutate(
    x = fct_relabel(
      region, str_replace_all, pattern = " \\(", replacement = "\n("
    )
  ) %>% 
  group_by(x, region, status) %>% 
  tally() %>% 
  arrange(region, desc(status)) %>% 
  mutate(y = cumsum(n)) %>% 
  ggplot(
    aes(x, n, fill = status)
  ) +
  geom_col() +
  geom_label(
    aes(y = y - 0.5*n, label = comma(n, 1), colour = status),
    data = . %>% dplyr::filter(n > 0.2*max(y)),
    fill = "white", alpha = 0.8,
    show.legend = FALSE
  ) +
  geom_text(
    aes(y = y1, label = comma(y, 1)),
    data = . %>% 
      dplyr::filter(y == max(y)) %>% 
      ungroup() %>% 
      mutate(y1 = y + 0.03*max(y))
  ) +
  scale_y_continuous(expand = expansion(c(0, 0.5)), labels = comma) +
  scale_fill_manual(values = direction_colours) +
  scale_colour_manual(values = direction_colours) +
  labs(x = "Region", y = "Merged Windows", fill = "Status")
```

`r ifelse(has_features, "### Signal By Feature", "")`

```{r boxplot-features, eval = has_features, echo = has_features, fig.height = 7, fig.cap = "*Distributions of overall signal (logCPM) and changed signal (logFC) for each external features.*"}
merged_results %>% 
  select(AveExpr, logFC, feature) %>% 
  as_tibble() %>% 
  pivot_longer(cols = c("AveExpr", "logFC")) %>% 
  ggplot(aes(feature, value, fill = feature)) +
  geom_boxplot() +
  facet_grid(
    name~., scales = "free", labeller = facet_labeller, switch = "y"
  ) +
  scale_fill_manual(values = feature_colours) +
  labs(x = "Region", y = c(), fill = "Overlaps") 
```

`r ifelse(has_features, "### Differential Binding By Feature", "")`

```{r plot-db-features, eval = has_features, echo = has_features, fig.height = 7, fig.cap = glue("*Merged windows considered as showing differential {target} binding across all external features.*")}
merged_results %>% 
  select(status, feature) %>% 
  as_tibble() %>% 
  group_by(feature, status) %>% 
  tally() %>% 
  arrange(feature, desc(status)) %>% 
  mutate(y = cumsum(n)) %>% 
  ggplot(
    aes(feature, n, fill = status)
  ) +
  geom_col() +
  geom_label(
    aes(y = y - 0.5*n, label = comma(n, 1), colour = status),
    data = . %>% dplyr::filter(n > 0.2*max(y)),
    fill = "white", alpha = 0.8,
    show.legend = FALSE
  ) +
  geom_text(
    aes(y = y1, label = comma(y, 1)),
    data = . %>% 
      dplyr::filter(y == max(y)) %>% 
      ungroup() %>% 
      mutate(y1 = y + 0.03*max(y))
  ) +
  scale_y_continuous(expand = expansion(c(0, 0.5)), labels = comma) +
  scale_fill_manual(values = direction_colours) +
  scale_colour_manual(values = direction_colours) +
  labs(x = "Feature", y = "Merged Windows", fill = "Status")
```

`r ifelse(has_features, "### Differential Binding By Region And Feature", "")`

```{r scatterpie-region-feature, eval = has_features, echo = has_features, fig.cap = glue("*Comparison of changes in {target} binding separated by gene-centric region and external features provided in {basename(config$external$features)}*")}
df <- merged_results %>% 
  as_tibble() %>% 
  group_by(region, feature, status) %>% 
  summarise(n = dplyr::n(), .groups = "drop_last") %>% 
  mutate(total = sum(n)) %>% 
  ungroup() %>% 
  mutate(
    region = fct_relabel(region, str_wrap, width = 20),
    x0 = as.integer(region),
    y0 = as.integer(feature),
    r = total / sum(total),
    r = 0.5* r / max(r)
  ) 
df %>% 
  ggplot() +
  ggforce::stat_pie(
    aes(x0 = x0, y0 = y0, r0 = 0, r = r, fill = status, amount = n)
  ) +
  coord_equal() +
  geom_label(
    aes(x0, y0, label = comma(total)),
    data = . %>% 
      distinct(x0, y0, total) %>% 
      dplyr::filter(total > 0.01 * sum(total)),
    alpha = 0.6,
    inherit.aes = FALSE,
    show.legend = FALSE
  ) +
  scale_x_continuous(
    breaks = seq_along(levels(df$region)), labels = levels(df$region)
  ) +
  scale_y_continuous(
    breaks = seq_along(levels(df$feature)), labels = levels(df$feature)
  ) +
  scale_fill_manual(values = direction_colours) +
  labs(
    x = "Region", y = "Feature", fill = "Status"
  ) +
  theme(legend.position = "bottom")
```


## Result Tables {.tabset}

```{r tbl-funs}
up_col <- function(x) {
  if (is.na(x) | is.nan(x)) return("#ffffff")
  rgb(
    colorRamp(c("#ffffff", colours$direction[["up"]]))(x), maxColorValue = 255
  )
}
down_col <- function(x) {
  if (is.na(x) | is.nan(x)) return("#ffffff")
  rgb(
    colorRamp(c("#ffffff", colours$direction[["down"]]))(x), maxColorValue = 255
  )
}
unch_col <- function(x) {
  if (is.na(x) | is.nan(x)) return("#ffffff")
  rgb(
    colorRamp(c("#ffffff", colours$direction[["unchanged"]]))(x), 
    maxColorValue = 255
  )
}
lfc_col <- function(x){
  if (is.na(x) | is.nan(x)) return("#ffffff")
  rgb(
    colorRamp(c(colours$direction[["down"]], "#ffffff", colours$direction[["up"]]))(x), 
    maxColorValue = 255
  )
}
expr_col <- function(x){
  if (is.na(x) | is.nan(x)) return("#ffffff")
  rgb(colorRamp(hcl.colors(9, "TealRose"))(x), maxColorValue = 255)
}
bar_style <- function(width = 1, fill = "#e6e6e6", height = "75%", align = c("left", "right"), color = NULL, fontSize = c()) {
  align <- match.arg(align)
  if (align == "left") {
    position <- paste0(width * 100, "%")
    image <- sprintf("linear-gradient(90deg, %1$s %2$s, transparent %2$s)", fill, position)
  } else {
    position <- paste0(100 - width * 100, "%")
    image <- sprintf("linear-gradient(90deg, transparent %1$s, %2$s %1$s)", position, fill)
  }
  styles <- list(
    backgroundImage = image,
    backgroundSize = paste("100%", height),
    backgroundRepeat = "no-repeat",
    backgroundPosition = "center",
    color = color
  )
  if (!is.null(fontSize)) styles$fontSize = fontSize
  styles
}
with_tooltip <- function(value, width = 30) {
  tags$span(title = value, str_trunc(value, width))
}
tbl_cols <- list(
  `Macs2 Peak` = colDef(footer = htmltools::tags$b("Total")),
  Region = colDef(
    footer = htmltools::tags$b("Total"), minWidth = 200
  ),
  Feature = colDef(
    footer = htmltools::tags$b("Total"), minWidth = 200
  ),
  Up = colDef(
    cell = function(value) comma(value, 1),
    style = function(value) {
      normalized <- value / (length(merged_results) - sum(df$Unchanged))
      color <- up_col(normalized)
      list(background = color)
    },
    footer = function(values) htmltools::tags$b(comma(sum(values)))
  ),
  Down = colDef(
    cell = function(value) comma(value, 1),
    style = function(value) {
      normalized <- value / (length(merged_results) - sum(df$Unchanged))
      color <- down_col(normalized)
      list(background = color)
    },
    footer = function(values) htmltools::tags$b(comma(sum(values)))
  ),
  Ambiguous = colDef(
    cell = function(value) comma(value, 1),
    style = function(value) {
      normalized <- value / (length(merged_results) - sum(df$Unchanged))
      color <- unch_col(normalized)
      list(background = color)
    },
    footer = function(values) htmltools::tags$b(comma(sum(values)))
  ),
  Unchanged = colDef(
    cell = function(value) comma(value, 1),
    style = function(value) {
      normalized <- value / length(merged_results)#sum(df$Unchanged)
      color <- unch_col(normalized)
      list(background = color)
    },
    footer = function(values) htmltools::tags$b(comma(sum(values)))
  ),
  `% Changed` = colDef(
    cell = function(value) percent(value, 0.1),
    style = function(value) bar_style(width = value, fill = "#B3B3B3", align = "right"),
    align = "right",
    footer = htmltools::tags$b(
      percent(sum(1 - sum(df$Unchanged) / sum(df$Total)), 0.1)
    )
  ),
  Total = colDef(
    cell = function(value) comma(value, 1),
    style = function(value) {
      bar_style(width = value / length(merged_results), fill = "#B3B3B3", align = "right")
    },
    align = "right",
    footer = function(values) htmltools::tags$b(comma(sum(values)))
  ),
  `% Of All Windows` = colDef(
    style = function(value) bar_style(width = value, fill = "#B3B3B3", align = "right"),
    format = colFormat(digits = 2, percent = TRUE)
  )
)
```


### Overall Results

```{r tab-overall}
df <- merged_results %>% 
  as_tibble() %>% 
  group_by(overlaps_ref, status) %>% 
  summarise(n = dplyr::n(), .groups = "drop") %>% 
  complete(overlaps_ref, status, fill = list(n = 0)) %>% 
  group_by(overlaps_ref) %>% 
  mutate(Total = sum(n)) %>%
  ungroup() %>% 
  pivot_wider(names_from = "status", values_from = "n", values_fill = 0) %>% 
  mutate(
    overlaps_ref = unlist(facet_labeller(overlaps_ref)),
    `% Changed` = 1 - Unchanged / Total
  ) %>% 
  dplyr::rename(`Macs2 Peak` = overlaps_ref) 
cp <- htmltools::em(
  glue(
    "Overall results for changed {target} binding after {treat_levels[[2]]} ",
    "treament, broken down by whether each merged window overlaps a macs2-",
    "derived consensus peak."
  )
)
tbl_cols$`% Changed`$footer <- htmltools::tags$b(
      percent(sum(1 - sum(df$Unchanged) / sum(df$Total)), 0.1)
    )
tbl <- reactable(df, columns = tbl_cols[colnames(df)])
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```

### Summary By Region

```{r tab-res-by-region}
df <- merged_results %>% 
  select(
    overlaps_ref, status, all_of(fdr_column), region
  ) %>% 
  as_tibble() %>% 
  group_by(region, status) %>% 
  summarise(n = dplyr::n(), .groups = "drop") %>% 
  complete(region, status, fill = list(n = 0)) %>% 
  group_by(region) %>% 
  mutate(Total = sum(n)) %>% 
  ungroup() %>% 
  pivot_wider(names_from = "status", values_from = "n") %>% 
  mutate(
    `% Of All Windows` = Total / length(merged_results),
    `% Changed` = 1 - Unchanged / Total
  ) %>% 
  arrange(region) %>% 
  dplyr::select(
    Region = region,
    any_of(names(direction_colours)),
    `% Changed`, Total, `% Of All Windows`
  )
cp <- htmltools::em(
  glue(
    "Overall results for changed {target} binding after {treat_levels[[2]]} ",
    "treament, broken down by which genomic region contains the largest ", 
    "overlap with each merged window."
  )
)
tbl <- df %>%
  reactable(
    columns = tbl_cols[colnames(df)],
    defaultColDef = colDef(
      footer = function(values) htmltools::tags$b(comma(sum(values)))
    ),
    fullWidth = TRUE
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```

`r ifelse(has_features, "### Summary By Feature", "")`

```{r tab-res-by-feature, echo = has_features, eval = has_features}
df <- merged_results %>% 
  select(
    overlaps_ref, status, all_of(fdr_column), feature
  ) %>% 
  as_tibble() %>% 
  group_by(feature, status) %>% 
  tally() %>% 
  ungroup() %>% 
  complete(feature, status, fill = list(n = 0)) %>% 
  group_by(feature) %>% 
  mutate(Total = sum(n)) %>% 
  ungroup() %>% 
  pivot_wider(names_from = "status", values_from = "n") %>% 
  mutate(
    `% Of All Windows` = Total / length(merged_results),
    `% Changed` = 1 - Unchanged / Total
  ) %>% 
  arrange(feature) %>% 
  dplyr::select(
    Feature = feature,
    any_of(names(direction_colours)),
    `% Changed`, Total, `% Of All Windows`
  )
cp <- htmltools::em(
  glue(
    "Overall results for changed {target} binding after {treat_levels[[2]]} ",
    "treament, broken down by which external feature contains the largest ", 
    "overlap with each merged window."
  )
)
tbl <- df %>%
  reactable(
    columns = tbl_cols[colnames(df)],
    defaultColDef = colDef(
      footer = function(values) htmltools::tags$b(comma(sum(values)))
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```


### Most Highly Ranked

```{r tab-highy-ranked}
show_n <- min(200, length(merged_results))
scaling_vals <- list(
  logFC = c(-1, 1)*max(abs(merged_results$logFC)),
  AveExpr = range(merged_results$AveExpr),
  Width = max(width(merged_results))
)
cp <- htmltools::em(
  glue(
    "The {show_n} most highly-ranked merged windows by FDR. ",
    "Regions were assigned based on which genomic region showed the largest ",
    "overlap with the merged window. ",
    ifelse(
      has_features, 
      glue("Features are as provided in the file {basename(config$external$features)}. "),
      glue("")
    ),
    "For windows mapped to large numbers of genes, hovering a mouse over the ", 
    "cell will reveal the full set of genes. ",
    ifelse(
      any_detected,
      "Only genes considered as detected in the RNA-Seq data are shown. ",
      ""
    ),
    "The Macs2 Peak column is filterable using T/F values"
  )
)
fs <- 12
tbl <- merged_results %>%
  mutate(w = width) %>% 
  arrange(!!sym(fdr_column)) %>% 
  plyranges::slice(seq_len(show_n)) %>% 
  select(
    w, AveExpr, logFC, FDR = !!sym(fdr_column), 
    overlaps_ref, Region = region, any_of("feature"), Genes = gene_name
  ) %>% 
  as_tibble() %>% 
  dplyr::rename(
    Range = range, `Width (kb)` = w, `Macs2 Peak` = overlaps_ref
  ) %>% 
  rename_all(str_replace_all, "feature", "Feature") %>% 
  mutate(Genes = vapply(Genes, paste, character(1), collapse = "; ")) %>% 
  reactable(
    filterable = TRUE,
    showPageSizeOptions = TRUE,
    pageSizeOptions = c(10, 20, 50, show_n), defaultPageSize = 10,
    borderless  = TRUE,
    columns = list(
      Range = colDef(
        minWidth = 10 * fs,
        cell = function(value) {
          str_replace_all(value, ":", ": ")
        }
      ),
      `Width (kb)` = colDef(
        cell = function(value) round(value / 1e3, 1),
        style = function(value) {
          x <- value / scaling_vals$Width
          colour <- expr_col(x)
          list(
            background = colour, 
            borderRight = "1px solid rgba(0, 0, 0, 0.1)"
          )
        },
        maxWidth = 5 * fs
      ),
      AveExpr = colDef(
        cell = function(value) round(value, 2),
        style = function(value) {
          x <- (value - min(scaling_vals$AveExpr)) / diff(scaling_vals$AveExpr)
          colour = expr_col(x)
          list(background = colour)
        },
        maxWidth = 5.5 * fs
      ),
      logFC = colDef(
        cell = function(value) round(value, 2),
        style = function(value) {
          x <- (value - min(scaling_vals$logFC)) / diff(scaling_vals$logFC)
          colour <- lfc_col(x)
          list(background = colour)
        },
        maxWidth = 5.5 * fs
      ),
      FDR = colDef(
        cell = function(value) sprintf("%.2e", value),
        style = list(borderRight = "1px solid rgba(0, 0, 0, 0.1)"),
        maxWidth = 5.5 * fs
      ),
      `Macs2 Peak` = colDef(
        cell = function(value) ifelse(value, "\u2714 Yes", "\u2716 No"),
        style = function(value) {
          color <- ifelse(value, "#008000", "#e00000")
          list(color = color, borderRight = "1px solid rgba(0, 0, 0, 0.1)")
        },
        maxWidth = 5 * fs
      ),
      Region = colDef(maxWidth = 150),
      Genes = colDef(
        cell = function(value) with_tooltip(value),
        minWidth = 11 * fs
      )
    ),
    theme = reactableTheme(style = list(fontSize = fs))
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```


## Inspection of Top-Ranked Regions {.tabset}

```{r grl-to-plot}
## Define a GRL with the key ranges.
## Then we can step through it & make all of the requisite plots
grl_to_plot <- GRangesList(
  top_up_by_fdr = merged_results %>% 
    filter(logFC > 0) %>% 
    filter(!!sym(fdr_column) == min(!!sym(fdr_column))),
  top_up_by_lfc = merged_results %>% 
    filter(logFC == max(logFC)),
  top_up_by_expr = merged_results %>% 
    filter(logFC > 0, !!sym(fdr_column) < fdr_alpha) %>% 
    filter(AveExpr == max(AveExpr)),
  top_up_no_macs2 = merged_results %>% 
    filter(logFC > 0, !!sym(fdr_column) < fdr_alpha, !overlaps_ref) %>% 
    arrange(!!sym(fdr_column)),
  top_down_by_fdr = merged_results %>% 
    filter(logFC < 0) %>% 
    filter(!!sym(fdr_column) == min(!!sym(fdr_column))),
  top_down_by_lfc = merged_results %>% 
    filter(logFC == min(logFC)),
  top_down_by_expr = merged_results %>% 
    filter(logFC < 0, !!sym(fdr_column) < fdr_alpha) %>% 
    filter(AveExpr == max(AveExpr)),
  top_up_no_macs2 = merged_results %>% 
    filter(logFC < 0, !!sym(fdr_column) < fdr_alpha, !overlaps_ref) %>% 
    arrange(!!sym(fdr_column)) 
) %>% 
  unlist() %>% 
  .[!duplicated(.)] %>% 
  splitAsList(names(.)) %>% 
  lapply(setNames, c()) %>% 
  lapply(function(x) x[1]) %>% 
  GRangesList()
```


```{r prepare-plot-objects}
## The coverage
bwfl <- list2(
  "{target}" := file.path(
    macs2_path, glue("{treat_levels}_merged_treat_pileup.bw")
  ) %>% 
    BigWigFileList() %>% 
    setNames(treat_levels)
)
line_col <- list2("{target}" := treat_colours)
## Coverage annotations
annot <- merged_results %>% 
  splitAsList(.$status) %>% 
  .[vapply(., length, integer(1)) > 0] %>% 
  endoapply(granges) %>% 
  list() %>% 
  setNames(target)
## Coverage y-limits
ind <- filtered_counts %>% 
  assay("counts") %>% 
  apply(MARGIN = 2, which.max) %>%
  unique()
max_ranges <- rowRanges(filtered_counts[ind])
y_lim <- bwfl[[target]] %>%
  lapply(import.bw, which = max_ranges) %>%
  lapply(function(x) c(0, max(x$score))) %>% 
  unlist() %>% 
  range() %>% 
  list() %>% 
  setNames(target)

## The features track
feat_gr <- gene_regions %>% 
  lapply(granges) %>% 
  GRangesList()
feature_colours <- colours$regions
if (has_features) {
  feat_gr <- feat_gr %>% 
    c(
      splitAsList(external_features, external_features$feature)
    )
  feature_colours <- c(feature_colours, colours$features)
}

## The genes track
hfgc_genes <- trans_models
gene_col <- "grey"
if (any_detected){
  rna_lfc_col <- colnames(rnaseq)[str_detect(str_to_lower(colnames(rnaseq)), "logfc")][1]
  rna_fdr_col <- colnames(rnaseq)[str_detect(str_to_lower(colnames(rnaseq)), "fdr|adjp")][1]
  if (!is.na(rna_lfc_col) & !is.na(rna_fdr_col)) {
    hfgc_genes <- trans_models %>% 
      mutate(
        status = case_when(
          !gene %in% rnaseq$gene_id ~ "Undetected",
          gene %in% dplyr::filter(
            rnaseq, !!sym(rna_lfc_col) > 0, !!sym(rna_fdr_col) < fdr_alpha
          )$gene_id ~ "Up",
          gene %in% dplyr::filter(
            rnaseq, !!sym(rna_lfc_col) < 0, !!sym(rna_fdr_col) < fdr_alpha
          )$gene_id ~ "Down",
          gene %in% dplyr::filter(
            rnaseq, !!sym(rna_fdr_col) >= fdr_alpha
          )$gene_id ~ "Unchanged",
        )
      ) %>% 
      splitAsList(.$status) %>% 
      lapply(select, -status) %>% 
      GRangesList()
    gene_col <- colours$direction %>% 
      setNames(str_to_title(names(.)))
  }
}

## External Coverage (Optional)
if (!is.null(config$external$coverage)) {
  ext_cov_path <- config$external$coverage %>% 
    lapply(unlist) %>% 
    lapply(function(x) setNames(here::here(x), names(x)))
  bwfl <- c(
    bwfl[target],
    lapply(ext_cov_path, function(x) BigWigFileList(x) %>% setNames(names(x)))
  )
  line_col <- c(
    line_col[target],
    ext_cov_path %>% 
      lapply(
        function(x) {
          missing <- setdiff(names(x), names(treat_colours))
          cmn <- intersect(names(x), names(treat_colours))
          col <- setNames(character(length(names(x))), names(x))
          if (length(cmn) > 0) col[cmn] <- treat_colours[cmn]
          if (length(missing) > 0) 
            col[missing] <- hcl.colors(
              max(5, length(missing)), "Zissou 1")[seq_along(missing)]
          col
        }
      )
  )
  
  y_ranges <- grl_to_plot %>% 
    unlist() %>% 
    granges() %>% 
    resize(w = 10 * width(.), fix = 'center')
  
  y_lim <- c(
    y_lim[target],
    bwfl[names(bwfl) != target] %>% 
      lapply(
        function(x) {
          GRangesList(lapply(x, import.bw, which = y_ranges)) %>% 
            unlist() %>% 
            filter(score == max(score)) %>% 
            mcols() %>% 
            .[["score"]] %>% 
            c(0) %>% 
            range()
        }
      )
  )
  
}
```


```{r makeCaption}
makeCaption <- function(.gr) {
  if (is.null(.gr)) return(NULL)
  dir <- ifelse(.gr$direction == "Down", 'decreased', 'increased')
  reg <- case_when(
    str_detect(.gr$region, "Inter") ~ paste("an", .gr$region, "region"),
    str_detect(.gr$region, "Upstream") ~ paste("an", .gr$region),
    str_detect(.gr$region, "(Ex|Intr)on") ~ paste("an", .gr$region),
    str_detect(.gr$region, "^Prom") ~ paste("a", .gr$region)
  )
  feat <- c()
  if (has_features) feat <- case_when(
    str_detect(.gr$feature, "^[AEIOU]") ~ paste("an", .gr$feature),
    !str_detect(.gr$feature, "^[AEIOU]") ~ paste("a", .gr$feature)
  )
  gn <- unlist(.gr$gene_name)
  fdr <- mcols(.gr)[[fdr_column]]
  fdr <- ifelse(
    fdr < 0.001, sprintf('%.2e', fdr), sprintf('%.3f', fdr)
  )
  cp <- c(
    glue(
      "*The {width(.gr)}bp region showing {dir} {target} binding in response to ", 
      "{treat_levels[[2]]} treatment (FDR = {fdr}). ",
      "The range mostly overlapped with {reg}, with all ",
      "defined regions shown as a contiguous bar. ",
      ifelse(
        has_features,
        glue(
          "Using the features supplied in {basename(config$external$features)}, ",
          "this mostly overlapped {feat}, shown as a separate block ",
          "with the gene-centric regions. "
        ),
        glue("")
      ),
    ),
    ifelse(
      .gr$overlaps_ref,
      paste(
        "A consensus peak overlapping this region was identified by",
        "`macs2 callpeak` when using merged samples."
      ),
      "No consensus peak was identified using `macs2 callpeak`."
    ),
    ifelse(
      length(gn) > 0,
      paste0(
        "Using the above mapping strategy, this range is likely to regulate ",
        collapseGenes(gn, format = "")
      ),
      "No genes were able to be assigned to this region."
    ),
    paste(
      "For each sample, the y-axis limits represent the values from the window", 
      "with the highest signal.*"
    )
  )
  paste(cp, collapse = " ")
}
```


### Normalised Sliding Windows

```{r plot-qsmooth-top-ranges, fig.cap = glue("*Most highly ranked ranges for both gained and decreased {target} binding in repsonse to {treat_levels[[2]]} treatment. The smooth-quantile normalised values are shown across the initial set of sliding windows before merging. Ranges were chosen as the most extreme for FDR, Binding Strength (Signal) and logFC. Windows are shown free of the genomic context.*")}
grl_to_plot %>% 
  lapply(
    function(x){
      filtered_counts %>% 
        subsetByOverlaps(x) %>% 
        assay("qsmooth") %>% 
        as_tibble(rownames = "window") %>% 
        pivot_longer(
          cols = all_of(colnames(filtered_counts)),
          names_to = "sample",
          values_to = "logCPM"
        ) %>% 
        mutate(range = as.character(x)) %>% 
        list()
    }
  ) %>% 
  as_tibble() %>% 
  pivot_longer(everything()) %>% 
  unnest(everything()) %>% 
  mutate(
    direction = str_extract(name, "down|up"),
    by = str_extract(name, "by_.+"),
    across(
      all_of(c("name", "direction", "by")), str_sep_to_title
    ),
    name = str_remove_all(name, "Top ") %>% 
      str_replace_all("Lfc", "logFC") %>% 
      str_replace_all("Fdr", "FDR") %>% 
      str_replace_all("Expr", "Highest Signal") %>% 
      str_replace_all("By (.+)", "(\\1)") %>% 
      str_replace_all("No Macs2", "(No Macs2 Peak)"),
    window = as.integer(window)
  ) %>% 
  left_join(samples, by = "sample") %>% 
  ggplot(
    aes(window, logCPM, colour = treat, linetype = !!sym(rep_col))
  ) +
  geom_line(
    data = . %>% 
      group_by(name) %>% 
      dplyr::filter(max(window) > 1)
  ) + 
  geom_point(
    data = . %>% 
      group_by(name) %>% 
      dplyr::filter(max(window) == 1),
    show.legend = FALSE
  ) + 
  facet_wrap(~name + range, scales = "free_x") +
  scale_colour_manual(values = treat_colours) + 
  labs(
    x = "Sliding Window\n(Unmerged)", y = "SQ-Normalised logCPM", colour = "Treat",
    linetype = str_sep_to_title(rep_col)
  )
```

`r ifelse("top_down_by_expr" %in% names(grl_to_plot), "### Decreased Binding (Strongest Signal)", "")`

```{r plot-top-down-by-expr, eval = "top_down_by_expr" %in% names(grl_to_plot), echo = "top_down_by_expr" %in% names(grl_to_plot), fig.cap = makeCaption(grl_to_plot$top_down_by_expr)}
plotHFGC(
  grl_to_plot$top_down_by_expr,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = direction_colours,
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10,
  max = 10 * width(grl_to_plot$top_down_by_expr),
  ylim = y_lim,
  featsize = 2,
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
```

`r ifelse("top_down_by_fdr" %in% names(grl_to_plot), "### Decreased Binding (Lowest FDR)", "")`

```{r plot-top_down_by_fdr, eval = "top_down_by_fdr" %in% names(grl_to_plot), echo = "top_down_by_fdr" %in% names(grl_to_plot), fig.cap = makeCaption(grl_to_plot$top_down_by_fdr)}
plotHFGC(
  grl_to_plot$top_down_by_fdr,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = direction_colours,
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10,
  max = 10 * width(grl_to_plot$top_down_by_fdr),
  ylim = y_lim,
  featsize = 2,
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
```

`r ifelse("top_down_by_lfc" %in% names(grl_to_plot), "### Decreased Binding (Most Extreme)", "")`

```{r plot-top_down_by_lfc, eval = "top_down_by_lfc" %in% names(grl_to_plot), echo = "top_down_by_lfc" %in% names(grl_to_plot), fig.cap = makeCaption(grl_to_plot$top_down_by_lfc)}
plotHFGC(
  grl_to_plot$top_down_by_lfc,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = direction_colours,
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10, 
  max = 10 * width(grl_to_plot$top_down_by_lfc),
  ylim = y_lim,
  featsize = 2,
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
```

`r ifelse("top_down_no_macs2" %in% names(grl_to_plot), "### Significant Decreased Binding (No Consensus Peak)", "")`

```{r plot-top_down_no_macs2, eval = "top_down_no_macs2" %in% names(grl_to_plot), echo = "top_down_no_macs2" %in% names(grl_to_plot), fig.cap = makeCaption(grl_to_plot$top_down_no_macs2)}
plotHFGC(
  grl_to_plot$top_down_no_macs2,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = direction_colours,
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10,
  max = 10 * width(grl_to_plot$top_down_no_macs2),
  ylim = y_lim,
  featsize = 2
)
```


`r ifelse("top_up_by_expr" %in% names(grl_to_plot), "### Increased Binding (Strongest Signal)", "")`

```{r plot-top-up-by-expr, eval = "top_up_by_expr" %in% names(grl_to_plot), echo = "top_up_by_expr" %in% names(grl_to_plot), fig.cap = makeCaption(grl_to_plot$top_up_by_expr)}
plotHFGC(
  grl_to_plot$top_up_by_expr,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = direction_colours,
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10,
  max = 10 * width(grl_to_plot$top_up_by_expr),
  ylim = y_lim,
  featsize = 2,
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
```

`r ifelse("top_up_by_fdr" %in% names(grl_to_plot), "### Increased Binding (Lowest FDR)", "")`

```{r plot-top_up_by_fdr, eval = "top_up_by_fdr" %in% names(grl_to_plot), echo = "top_up_by_fdr" %in% names(grl_to_plot), fig.cap = makeCaption(grl_to_plot$top_up_by_fdr)}
plotHFGC(
  grl_to_plot$top_up_by_fdr,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = direction_colours,
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10,
  max = 10 * width(grl_to_plot$top_up_by_fdr),
  ylim = y_lim,
  featsize = 2,
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
```

`r ifelse("top_up_by_lfc" %in% names(grl_to_plot), "### Increased Binding (Most Extreme)", "")`

```{r plot-top_up_by_lfc, eval = "top_up_by_lfc" %in% names(grl_to_plot), echo = "top_up_by_lfc" %in% names(grl_to_plot), fig.cap = makeCaption(grl_to_plot$top_up_by_lfc)}
plotHFGC(
  grl_to_plot$top_up_by_lfc,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = direction_colours,
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10,
  max = 10 * width(grl_to_plot$top_up_by_lfc),
  ylim = y_lim,
  featsize = 2,
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
```

`r ifelse("top_up_no_macs2" %in% names(grl_to_plot), "### Significant Increased Binding (No Consensus Peak)", "")`

```{r plot-top_up_no_macs2, eval = "top_up_no_macs2" %in% names(grl_to_plot), echo = "top_up_no_macs2" %in% names(grl_to_plot), fig.cap = makeCaption(grl_to_plot$top_up_no_macs2)}
plotHFGC(
  grl_to_plot$top_up_no_macs2,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = direction_colours,  
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10,
  max = 10 * width(grl_to_plot$top_up_no_macs2),
  ylim = y_lim,
  featsize = 2,
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
```

# Enrichment Testing {.tabset}

```{r msigdb}
msigdb <- msigdbr(species = "Homo sapiens") %>% 
  dplyr::filter(
    gs_cat %in% unlist(params$enrichment$msigdb$gs_cat) |
      gs_subcat %in% unlist(params$enrichment$msigdb$gs_subcat),
    str_detect(ensembl_gene, "^E")
  ) %>% 
  dplyr::rename(gene_id = ensembl_gene, gene_name = gene_symbol) %>% # For easier integration 
  dplyr::select(-starts_with("human"), -contains("entrez")) %>% 
  dplyr::filter(gene_id %in% genes_gr$gene_id) %>% 
  mutate(gs_url = str_extract(gs_url, "^[^|]+"))
gs_by_gsid <- msigdb %>% 
  split(.$gs_name) %>% 
  lapply(pull, "gene_id") %>% 
  lapply(unique)
gs_by_geneid <- msigdb %>% 
  split(.$gene_id) %>% 
  lapply(pull, "gs_name") %>% 
  lapply(unique)
gs_url <- msigdb %>% 
  distinct(gs_name, gs_url) %>%
  mutate(
    gs_url = ifelse(
      gs_url == "", "
      http://www.gsea-msigdb.org/gsea/msigdb/collections.jsp",
      gs_url
    )
  ) %>% 
  with(setNames(gs_url, gs_name))
min_genes <- params$enrichment$min_prop_goseq * length(genes_gr)
max_genes <- params$enrichment$max_prop_goseq * length(genes_gr)
enrich_alpha <- params$enrichment$alpha
adj_method <- match.arg(params$enrichment$adj, p.adjust.methods)
adj_desc <- case_when(
  p.adjust.methods %in% c("fdr", "BH") ~ "the Benjamini-Hochberg FDR",
  p.adjust.methods %in% c("BY") ~ "the Benjamini-Yekutieli FDR",
  p.adjust.methods %in% c("bonferroni") ~ "the Bonferroni",
  p.adjust.methods %in% c("holm") ~ "Holm's",
  p.adjust.methods %in% c("hommel") ~ "Hommel's",
  p.adjust.methods %in% c("hochberg") ~ "Hochberg's",
  p.adjust.methods %in% c("none") ~ "no"
) %>% 
  setNames(p.adjust.methods)
mapped_ids <- merged_results %>%
  as_tibble() %>% 
  dplyr::select(status, gene_id) %>%
  unnest(gene_id) %>%
  group_by(gene_id) %>%
  summarise(
    up = any(status == "Up"), 
    down = any(status == "Down"), 
    n_peaks = dplyr::n(),
    .groups = "drop"
  )
```

As an initial exploration, the retained windows were compared to pre-defined gene-sets.
These were taken from the `MSigDB` database and the gene-sets used for enrichment testing were:

```{r tbl-msigdb}
msigdb %>% 
  distinct(gs_cat, gs_subcat, gs_name) %>%
  group_by(gs_cat, gs_subcat) %>% 
  summarise(`Gene Sets` = dplyr::n(),.groups = "drop") %>% 
  dplyr::rename(Category = gs_cat, `Sub-Category` = gs_subcat) %>% 
  mutate(
    Description = case_when(
      Category == "H" ~ "Hallmark Gene-Sets",
      Category == "C1" ~ "Positional Gene Sets",
      Category == "C2" ~ str_replace_all(`Sub-Category`, "CP:(.+)", "Curated Gene-Sets: \\1"),
      str_detect(`Sub-Category`, "TFT") ~ str_replace_all(`Sub-Category`, "TFT:(.+)", "Transcription Factor Target Prediction Gene-Sets: \\1"),
      str_detect(`Sub-Category`, "MIR") ~ str_replace_all(`Sub-Category`, "MIR:(.+)", "microRNA Target Gene-Sets: \\1"),
      str_detect(`Sub-Category`, "CGN") ~ "Cancer Gene Neighbourhoods",
      str_detect(`Sub-Category`, "CM") ~ "Cancer Modules",
      str_detect(`Sub-Category`, "GO:BP") ~ "Gene Ontology: Bological Process",
      str_detect(`Sub-Category`, "GO:MF") ~ "Gene Ontology: Molecular Function",
      str_detect(`Sub-Category`, "GO:CC") ~ "Gene Ontology: Cellular Component",
      str_detect(`Sub-Category`, "HPO") ~ "Human Phenotype Ontology",
      Category == "C6" ~ "Oncogenic Signature Gene Sets",
      `Sub-Category` == "IMMUNESIGDB" ~ "ImmuneSigDB Gene Sets",
      `Sub-Category` == "VAX" ~ "Vaccine Response Gene Sets",
      Category == "C8" ~ "Cell Type Signature Gene Sets"
    ),
    Category = factor(Category, levels = c("H", paste0("C", 1:8)))
  ) %>% 
  droplevels() %>% 
  arrange(Category) %>% 
  pander(
    justify = "llrl", 
    caption = "*Summary of Gene-Sets used for Enrichment Testing*"
  )
```

Retained windows were tested for enrichment of these gene-sets using:

1. Any window mapped to a gene from these gene-sets
2. Any window with evidence of differential binding mapped to a gene from these gene-sets

In the first case, mapped genes were tested for enrichment against all annotated genes, or all detected genes if RNA-Seq data was provided.
In the second case, i.e. Differentially Bound Windows, the control set of genes were those mapped to a window (i.e. those test set from step 1).
Testing was only performed if the number of genes mapped to the specific treatment group fell between
`r comma(min_genes, 1)` (`r percent(params$enrichment$min_prop_goseq, 0.1)` of included genes) and `r comma(max_genes, 1)` (`r percent(params$enrichment$max_prop_goseq)`).
All adjusted p-values below are calculated using `r adj_desc[[adj_method]]` adjustment.

## Windows With Detected `r target`

```{r goseq-mapped, fig.show='hide'}
pwf_mapped <- genes_gr %>% 
  mutate(w = width(.)) %>% 
  as_tibble() %>% 
  dplyr::select(gene_id, w) %>% 
  arrange(desc(w)) %>% 
  distinct(gene_id, .keep_all = TRUE) %>% 
  mutate(mapped = gene_id %in% unlist(merged_results$gene_id)) %>% 
  with(
    nullp(
      structure(mapped, names = gene_id),
      genome = config$genome$build,
      bias.data = log10(w),
      plot.fit = TRUE
    )
  )
goseq_mapped <- tibble(gs_name = character(), adj_p = numeric())
if (sum(pwf_mapped$DEgenes) > min_genes & sum(pwf_mapped$DEgenes) < max_genes) {
  goseq_mapped <- goseq(
    pwf_mapped, config$genome$build, gene2cat = gs_by_geneid
  ) %>% 
    dplyr::rename(gs_name = category, pval = over_represented_pvalue) %>% 
    dplyr::select(-starts_with("under")) %>% 
    as_tibble() %>% 
    arrange(pval) %>% 
    dplyr::filter(numDEInCat > 0) %>% 
    mutate(adj_p = p.adjust(pval, adj_method))
}
```

`r ifelse(sum(goseq_mapped$adj_p < enrich_alpha) == 0, "No enrichment was found amongst genes mapped to any window", "")`

```{r tbl-goseq-mapped, eval = sum(goseq_mapped$adj_p < enrich_alpha) > 0, echo = sum(goseq_mapped$adj_p < enrich_alpha) > 0}
cp <- htmltools::tags$caption(
  htmltools::em(
    glue(
      "
    All {sum(goseq_mapped$adj_p < enrich_alpha)} gene sets considered as 
    significantly enriched (p
    "
    ),
    htmltools::tags$sub("adj"),
    htmltools::em(
      glue(
        "
         < {enrich_alpha}) amongst the merged windows with detectable {target} 
        binding. Genes mapped to a merged window were compared to those not 
        mapped to any merged windows. Gene width was used to capture any 
        gene-level sampling bias.
        "
      )
    )
  )
)
tbl <- goseq_mapped %>%
  dplyr::filter(adj_p < enrich_alpha) %>% 
  left_join(
    dplyr::select(msigdb, gs_name, gene_name, gene_id, gs_url, gs_description),
    by = "gs_name"
  ) %>% 
  dplyr::filter(gene_id %in% unlist(merged_results$gene_id)) %>% 
  dplyr::select(-gene_id) %>% 
  chop(gene_name) %>% 
  mutate(
    gene_name = vapply(gene_name, paste, character(1), collapse = "; "),
    `%` = numDEInCat / numInCat
  ) %>% 
  dplyr::select(
    `Gene Set` = gs_name, Description = gs_description,
    `%`, numDEInCat, numInCat, 
    p = pval, adj_p, 
    `Mapped Genes` = gene_name
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list(
      "Gene Set" = colDef(
        minWidth = 150,
        cell = function(value) htmltools::tags$a(
          href = gs_url[[value]], 
          target = "_blank", 
          str_replace_all(value, "_", " ")
        ),
        html = TRUE
      ),
      Description = colDef(
        minWidth = 150,
        cell = function(value) with_tooltip(value, width = 100)
      ),
      numDEInCat = colDef(name = "Total Mapped", maxWidth = 80),
      numInCat = colDef(
        name = "Gene Set Size",
        maxWidth = 80,
        cell = function(value) comma(value, 1)
      ),
      "%" = colDef(
        name = "% Mapped",
        cell = function(value) percent(value, 0.1),
        maxWidth = 80
      ),
      p = colDef(
        cell = function(value) {
          ifelse(
            value < 0.001, sprintf("%.2e", value), sprintf("%.3f", value)
          )
        },
        maxWidth = 80
      ),
      adj_p = colDef(
        name = "p<sub>adj</sub>", html = TRUE,
        cell = function(value) {
          ifelse(
            value < 0.001, sprintf("%.2e", value), sprintf("%.3f", value)
          )
        },
        maxWidth = 80
      ),
      "Mapped Genes" = colDef(
        cell = function(value) with_tooltip(value, width = 100),
        minWidth = 150
      )
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```

## Windows With Differentially Bound `r target`

```{r goseq-diff, fig.show='hide'}
pwf_diff <- genes_gr %>% 
  as_tibble() %>% 
  distinct(gene_id) %>% 
  left_join(mapped_ids, by = "gene_id") %>% 
  mutate(
    n_peaks = ifelse(is.na(n_peaks), 0, n_peaks),
    diff = ifelse(n_peaks > 0, up|down, FALSE),
    diff = ifelse(is.na(diff), FALSE, diff)
  ) %>% 
  with(
    nullp(
      structure(diff, names = gene_id),
      genome = config$genome$build,
      bias.data = n_peaks,
      plot.fit = TRUE
    )
  )
goseq_diff <- tibble(gs_name = character(), adj_p = numeric())
if (sum(pwf_diff$DEgenes) > min_genes & sum(pwf_diff$DEgenes) < max_genes) {
  goseq_diff <- goseq(
    pwf_diff, config$genome$build, gene2cat = gs_by_geneid
  ) %>% 
    dplyr::rename(gs_name = category, pval = over_represented_pvalue) %>% 
    dplyr::select(-starts_with("under")) %>% 
    as_tibble() %>% 
    arrange(pval) %>% 
    dplyr::filter(numDEInCat > 0) %>% 
    mutate(adj_p = p.adjust(pval, adj_method))
}
```


`r ifelse(sum(goseq_diff$adj_p < enrich_alpha) == 0, "No enrichment was found amongst genes mapped to sites showing any differential binding.", "")`

```{r tbl-goseq-diff, eval = sum(goseq_diff$adj_p < enrich_alpha) > 0, echo = sum(goseq_diff$adj_p < enrich_alpha) > 0}
cp <- htmltools::tags$caption(
  htmltools::em(
    glue(
      "
    All {sum(goseq_diff$adj_p < enrich_alpha)} gene sets considered as 
    significantly enriched (p
    "
    ),
    htmltools::tags$sub("adj"),
    htmltools::em(
      glue(
        "
         < {enrich_alpha}) amongst the windows showing evidence of changed 
        {target} binding. Genes mapped to a differentially-bound window were 
        compared to those mapped to any merged window.
        "
      )
    )
  )
)
tbl <- goseq_diff %>% 
  dplyr::filter(adj_p < enrich_alpha) %>%
  left_join(
    dplyr::select(msigdb, gs_name, gene_name, gene_id, gs_url, gs_description),
    by = "gs_name"
  ) %>% 
  dplyr::filter(
    gene_id %in% unlist(subset(merged_results, status != "Unchanged")$gene_id)
  ) %>% 
  dplyr::select(-gene_id) %>% 
  chop(gene_name) %>% 
  mutate(
    gene_name = vapply(gene_name, paste, character(1), collapse = ": "),
    `%` = numDEInCat / numInCat
  ) %>% 
  dplyr::select(
    `Gene Set` = gs_name, Description = gs_description,
    `%`, numDEInCat, numInCat, p = pval, adj_p, 
    `Genes Mapped to DB Windows` = gene_name
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list(
      "Gene Set" = colDef(
        minWidth = 150,
        cell = function(value) htmltools::tags$a(
          href = gs_url[[value]], 
          target = "_blank", 
          str_replace_all(value, "_", " ")
        ),
        html = TRUE
      ),
      Description = colDef(
        minWidth = 150,
        cell = function(value) with_tooltip(value, width = 100)
      ),
      "%" = colDef(
        name = "% DB",
        maxWidth = 80,
        cell = function(value) percent(value, 0.1)
      ),
      numDEInCat = colDef(name = "Total Mapped To DB", maxWidth = 80),
      numInCat = colDef(
        name = "Total Mapped",
        maxWidth = 80,
        cell = function(value) comma(value, 1)
      ),
      p = colDef(
        cell = function(value) {
          ifelse(
            value < 0.001, sprintf("%.2e", value), sprintf("%.3f", value)
          )
        },
        maxWidth = 70
      ),
      adj_p = colDef(
        name = "p<sub>adj</sub>", html = TRUE,
        cell = function(value) {
          ifelse(
            value < 0.001, sprintf("%.2e", value), sprintf("%.3f", value)
          )
        },
        maxWidth = 70
      ),
      "Genes Mapped to DB Windows" = colDef(
        cell = function(value) with_tooltip(value, width = 100),
        minWidth = 200
      )
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```


## Windows With Increased `r target` Binding

```{r goseq-up, fig.show='hide'}
pwf_up <- genes_gr %>% 
  as_tibble() %>% 
  distinct(gene_id) %>% 
  left_join(mapped_ids, by = "gene_id") %>% 
  mutate(
    n_peaks = ifelse(is.na(n_peaks), 0, n_peaks),
    diff = ifelse(n_peaks > 0, up, FALSE),
    diff = ifelse(is.na(diff), FALSE, diff)
  ) %>% 
  with(
    nullp(
      structure(diff, names = gene_id),
      genome = config$genome$build,
      bias.data = n_peaks,
      plot.fit = TRUE
    )
  )
goseq_up <- tibble(gs_name = character(), adj_p = numeric())
if (sum(pwf_up$DEgenes) > min_genes & sum(pwf_up$DEgenes) < max_genes) {
  goseq_up <- goseq(
    pwf_up, config$genome$build, gene2cat = gs_by_geneid
  ) %>% 
    dplyr::rename(gs_name = category, pval = over_represented_pvalue) %>% 
    dplyr::select(-starts_with("under")) %>% 
    as_tibble() %>% 
    arrange(pval) %>% 
    dplyr::filter(numDEInCat > 0) %>% 
    mutate(adj_p = p.adjust(pval, adj_method))
}
```


`r ifelse(sum(goseq_up$adj_p < enrich_alpha) == 0, "No enrichment was found amongst genes mapped to sites with increased binding", "")`

```{r tbl-goseq-up, eval = sum(goseq_up$adj_p < enrich_alpha) > 0, echo = sum(goseq_up$adj_p < enrich_alpha) > 0}
cp <- htmltools::tags$caption(
  htmltools::em(
    glue(
      "
    All {sum(goseq_up$adj_p < enrich_alpha)} gene sets considered as 
    significantly enriched (p
    "
    ),
    htmltools::tags$sub("adj"),
    htmltools::em(
      glue(
        "
         < {enrich_alpha}) amongst the windows showing evidence of increased 
        {target} binding. Genes mapped to an increasing window were 
        compared to those mapped to any merged window.
        "
      )
    )
  )
)
tbl <- goseq_up %>% 
  dplyr::filter(adj_p < enrich_alpha) %>%
  left_join(
    dplyr::select(msigdb, gs_name, gene_name, gene_id, gs_url, gs_description),
    by = "gs_name"
  ) %>% 
  dplyr::filter(
    gene_id %in% unlist(subset(merged_results, status == "Up")$gene_id)
  ) %>% 
  dplyr::select(-gene_id) %>% 
  chop(gene_name) %>% 
  mutate(
    gene_name = vapply(gene_name, paste, character(1), collapse = ": "),
    `%` = numDEInCat / numInCat
  ) %>% 
  dplyr::select(
    `Gene Set` = gs_name, Description = gs_description,
    `%`, numDEInCat, numInCat, p = pval, adj_p, 
    `Genes Mapped to Gained Windows` = gene_name
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list(
      "Gene Set" = colDef(
        minWidth = 150,
        cell = function(value) htmltools::tags$a(
          href = gs_url[[value]], 
          target = "_blank", 
          str_replace_all(value, "_", " ")
        ),
        html = TRUE
      ),
      Description = colDef(
        minWidth = 150,
        cell = function(value) with_tooltip(value, width = 100)
      ),
      "%" = colDef(
        name = "% Gained",
        cell = function(value) percent(value, 0.1),
        maxWidth = 80
      ),
      numDEInCat = colDef(name = "Mapped to Gained Windows", maxWidth = 80),
      numInCat = colDef(
        name = "Total Mapped",
        maxWidth = 80,
        cell = function(value) comma(value, 1)
      ),
      p = colDef(
        cell = function(value) {
          ifelse(
            value < 0.001, sprintf("%.2e", value), sprintf("%.3f", value)
          )
        },
        maxWidth = 70
      ),
      adj_p = colDef(
        name = "p<sub>adj</sub>", html = TRUE,
        cell = function(value) {
          ifelse(
            value < 0.001, sprintf("%.2e", value), sprintf("%.3f", value)
          )
        },
        maxWidth = 70
      ),
      "Genes Mapped to Gained Windows" = colDef(
        cell = function(value) with_tooltip(value, width = 100),
        minWidth = 200
      )
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```

## Windows With Decreased `r target` Binding

```{r goseq-down, fig.show='hide'}
pwf_down <- genes_gr %>% 
  as_tibble() %>% 
  distinct(gene_id) %>% 
  left_join(mapped_ids, by = "gene_id") %>% 
  mutate(
    n_peaks = ifelse(is.na(n_peaks), 0, n_peaks),
    diff = ifelse(n_peaks > 0, down, FALSE),
    diff = ifelse(is.na(diff), FALSE, diff)
  ) %>% 
  with(
    nullp(
      structure(diff, names = gene_id),
      genome = config$genome$build,
      bias.data = n_peaks,
      plot.fit = TRUE
    )
  )
goseq_down <- tibble(gs_name = character(), adj_p = numeric())
if (sum(pwf_down$DEgenes) > min_genes & sum(pwf_down$DEgenes) < max_genes) {
  goseq_down <- goseq(
    pwf_down, config$genome$build, gene2cat = gs_by_geneid
  ) %>% 
    dplyr::rename(gs_name = category, pval = over_represented_pvalue) %>% 
    dplyr::select(-starts_with("under")) %>% 
    as_tibble() %>% 
    arrange(pval) %>% 
    dplyr::filter(numDEInCat > 0) %>% 
    mutate(adj_p = p.adjust(pval, adj_method))
}
```


`r ifelse(sum(goseq_down$adj_p < enrich_alpha) == 0, "No enrichment was found amongst genes mapped to sites with decreased binding", "")`

```{r tbl-goseq-down, eval = sum(goseq_down$adj_p < enrich_alpha) > 0, echo = sum(goseq_down$adj_p < enrich_alpha) > 0}
cp <- htmltools::tags$caption(
  htmltools::em(
    glue(
      "
    All {sum(goseq_down$adj_p < enrich_alpha)} gene sets considered as 
    significantly enriched (p
    "
    ),
    htmltools::tags$sub("adj"),
    htmltools::em(
      glue(
        "
         < {enrich_alpha}) amongst the windows showing evidence of decreased 
        {target} binding. Genes mapped to a decreasing window were 
        compared to those mapped to any merged window.
        "
      )
    )
  )
)
tbl <- goseq_down %>% 
  dplyr::filter(adj_p < enrich_alpha) %>%
  left_join(
    dplyr::select(msigdb, gs_name, gene_name, gene_id, gs_url, gs_description),
    by = "gs_name"
  ) %>% 
  dplyr::filter(
    gene_id %in% unlist(subset(merged_results, status == "Down")$gene_id)
  ) %>% 
  dplyr::select(-gene_id) %>% 
  chop(gene_name) %>% 
  mutate(
    gene_name = vapply(gene_name, paste, character(1), collapse = ": "),
    `%` = numDEInCat / numInCat
  ) %>% 
  dplyr::select(
    `Gene Set` = gs_name, Description = gs_description,
    `%`, numDEInCat, numInCat, p = pval, adj_p, 
    `Genes Mapped to Decreased Windows` = gene_name
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list(
      "Gene Set" = colDef(
        minWidth = 150,
        cell = function(value) htmltools::tags$a(
          href = gs_url[[value]], 
          target = "_blank", 
          str_replace_all(value, "_", " ")
        ),
        html = TRUE
      ),
      Description = colDef(
        minWidth = 150,
        cell = function(value) with_tooltip(value, width = 100)
      ),
      "%" = colDef(
        name = "% Decreased",
        cell = function(value) percent(value, 0.1),
        maxWidth = 80
      ),
      numDEInCat = colDef(name = "Mapped to Decreased Windows", maxWidth = 80),
      numInCat = colDef(
        name = "Total Mapped",
        maxWidth = 80,
        cell = function(value) comma(value, 1)
      ),
      p = colDef(
        cell = function(value) {
          ifelse(
            value < 0.001, sprintf("%.2e", value), sprintf("%.3f", value)
          )
        },
        maxWidth = 70
      ),
      adj_p = colDef(
        name = "p<sub>adj</sub>", html = TRUE,
        cell = function(value) {
          ifelse(
            value < 0.001, sprintf("%.2e", value), sprintf("%.3f", value)
          )
        },
        maxWidth = 70
      ),
      "Genes Mapped to Decreased Windows" = colDef(
        cell = function(value) with_tooltip(value, width = 100),
        minWidth = 200
      )
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```

```{r add-rna, eval = any_detected, child = here::here('workflow/modules/rnaseq_differential_binding.Rmd')}
```


# Data Export

```{r export-diff-by-gene}
merged_results %>% 
  as_tibble() %>%
  unnest(all_of("gene_id")) %>%
  mutate(gene_name = id2gene[gene_id]) %>% 
  dplyr::select(
    gene_id, gene_name, range, AveExpr, logFC, P.Value, FDR = !!sym(fdr_column), 
    status, any_of(c("region", "feature")), macs2_peak = overlaps_ref,
  ) %>% 
  mutate(
    distance_to_gene = distance(
      GRanges(range),
      setNames(genes_gr, genes_gr$gene_id)[gene_id]
    )
  ) %>% 
  write_csv(
    gzfile(all_out$csv)
  )
```


```{r data-export, results='hide'}
write_rds(merged_results, all_out$results, compress = "gz")
write_rds(filtered_counts, all_out$windows, compress = "gz")
file.create(all_out$de_genes)
if (any_detected) write_csv(de_genes_db_regions, all_out$de_genes)

file.create(all_out$up_regions)
if (sum(merged_results$status == "Up") > 0) {
  merged_results %>% 
    filter(status == "Up") %>% 
    granges() %>% 
    export.bed(all_out$up_regions)
}

file.create(all_out$down_regions)
if (sum(merged_results$status == "Down") > 0) {
  merged_results %>% 
    filter(status == "Down") %>% 
    granges() %>% 
    export.bed(all_out$up_regions)
}

save.image(all_out$renv)
```

During this workflow, the following files were exported:

`r pander(lapply(all_out, str_extract, "output.+"))`

## References

<div id="refs"></div>
<br>
<button type="button" class="btn btn-default btn-sessioninfo" data-toggle="collapse" data-target="#sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Session information
</button>
</p>
<div id="sessioninfo" class="collapse">

```{r session-info, echo=FALSE}
pander::pander(sessionInfo())
```

</div>
