```{r get-params}
target <- params$target
macs2_fdr <- params$macs2_fdr
outlier_thresh <- params$outlier_thresh
min_prop <- params$min_prop
threads <- parallel::detectCores() - 1
```


```{r packages}
library(tidyverse)
library(magrittr)
library(rtracklayer)
library(glue)
library(pander)
library(scales)
library(plyranges)
library(yaml)
library(ngsReports)
library(ComplexUpset)
library(VennDiagram)
library(rlang)
library(BiocParallel)
library(parallel)
library(Rsamtools)
library(Biostrings)
library(ggside)
library(extraChIPs)
library(patchwork)
library(reactable)
library(htmltools)
library(rGREAT)
library(ggraph)
library(tidygraph)
library(patchwork)
library(msigdbr)
panderOptions("big.mark", ",")
panderOptions("missing", "")
panderOptions("table.split.table", Inf)
theme_set(
  theme_bw() +
    theme(plot.title = element_text(hjust = 0.5))
)
register(MulticoreParam(workers = threads))
source(here::here("workflow/scripts/custom_functions.R"))
source(here::here("workflow/scripts/table_funs.R"))
```

```{r config-paths}
config <- read_yaml(here::here("config", "config.yml"))
fdr_alpha <- config$comparisons$fdr
extra_params <- read_yaml(here::here("config", "params.yml"))
bam_path <- here::here(config$paths$bam)
macs2_path <- here::here("output", "macs2")
annotation_path <- here::here("output", "annotations")
colours <- file.path(annotation_path, "colours.rds") %>% 
  read_rds() %>% 
  lapply(unlist)
```

```{r read-samples}
samples <- macs2_path %>%
  file.path(target, glue("{target}_qc_samples.tsv")) %>%
  read_tsv()
treat_levels <- unique(samples$treat)
if (!is.null(config$comparisons$contrasts)) {
  ## Ensure levels respect those provided in contrasts
  treat_levels <- config$comparisons$contrasts %>%
    unlist() %>%
    intersect(samples$treat) %>% 
    unique()
}
rep_col <- setdiff(
  colnames(samples), c("sample", "treat", "target", "input", "label", "qc")
)
samples <- samples %>%
  mutate(
    treat = factor(treat, levels = treat_levels)
  ) %>% 
  dplyr::filter(treat %in% treat_levels)
stopifnot(nrow(samples) > 0)
ip <- unique(samples$input)
```


```{r annotations}
## Seqinfo
sq <- read_rds(file.path(annotation_path, "seqinfo.rds"))

## Gene Annotations
gtf_gene <- read_rds(file.path(annotation_path, "gtf_gene.rds"))
gtf_transcript <- read_rds(file.path(annotation_path, "gtf_transcript.rds"))
id2gene <- setNames(gtf_gene$gene_name, gtf_gene$gene_id)

external_features <- c()
has_features <- FALSE
if (!is.null(config$external$features)) {
  external_features <- suppressWarnings(
    import.gff(here::here(config$external$features), genome = sq)
  )
  keep_cols <- !vapply(
    mcols(external_features), function(x) all(is.na(x)), logical(1)
  )
  mcols(external_features) <- mcols(external_features)[keep_cols]
  has_features <- TRUE
}
gene_regions <- read_rds(file.path(annotation_path, "gene_regions.rds"))
regions <- vapply(gene_regions, function(x) unique(x$region), character(1))
rna_path <- here::here(config$external$rnaseq)
rnaseq <- tibble(gene_id = character())
if (length(rna_path) > 0) {
  stopifnot(file.exists(rna_path))
  if (str_detect(rna_path, "tsv$")) rnaseq <- read_tsv(rna_path)
  if (str_detect(rna_path, "csv$")) rnaseq <- read_csv(rna_path)
  if (!"gene_id" %in% colnames(rnaseq)) stop("Supplied RNA-Seq data must contain the column 'gene_id'")
  gtf_gene <- subset(gtf_gene, gene_id %in% rnaseq$gene_id)
}
tx_col <- intersect(c("tx_id", "transcript_id"), colnames(rnaseq))
rna_gr_col <- ifelse(length(tx_col) > 0, "transcript_id", "gene_id")
rna_col <- c(tx_col, "gene_id")[[1]]
tss <- read_rds(file.path(annotation_path, "tss.rds"))
## bands_df
cb <- config$genome$build %>%
  str_to_lower() %>% 
  paste0(".cytobands") 
data(list = cb)
bands_df <- get(cb)
```

```{r figure-params}
fig_path <- here::here("docs", "assets", target)
if (!dir.exists(fig_path)) dir.create(fig_path, recursive = TRUE)
fig_dev <- knitr::opts_chunk$get("dev")
fig_type <- fig_dev[[1]]
if (is.null(fig_type)) stop("Couldn't detect figure type")
fig_fun <- match.fun(fig_type)
if (fig_type %in% c("bmp", "jpeg", "png", "tiff")) {
  ## These figure types require dpi & resetting to be inches
  formals(fig_fun)$units <- "in"
  formals(fig_fun)$res <- 300
}
```

```{r bfl}
bfl <- bam_path %>%
  file.path(glue("{samples$sample}.bam")) %>%
  BamFileList() %>%
  setNames(samples$sample)
```

```{r load-peaks}
individual_peaks <- file.path(
  macs2_path, samples$sample, glue("{samples$sample}_peaks.narrowPeak")
) %>%
  importPeaks(seqinfo = sq) %>%
  endoapply(names_to_column, var = "sample") %>% 
  endoapply(mutate, sample = str_remove(sample, "_peak_[0-9]+$")) %>% 
  endoapply(
    mutate, 
    treat = left_join(tibble(sample = sample), samples, by = "sample")$treat
  ) %>% 
  setNames(samples$sample)
merged_peaks <- macs2_path %>% 
  file.path(target, glue("{target}_{treat_levels}_merged_peaks.narrowPeak")) %>% 
  importPeaks(seqinfo = sq) %>% 
  setNames(treat_levels)
filtered_peaks <- macs2_path %>% 
  file.path(target, glue("{target}_{treat_levels}_filtered_peaks.narrowPeak")) %>% 
  importPeaks(seqinfo = sq) %>% 
  setNames(treat_levels)
consensus_peaks <- macs2_path %>% 
  file.path(target, glue("{target}_consensus_peaks.bed.gz")) %>% 
  importPeaks(type = "bed", seqinfo = sq, setNames = FALSE) %>% 
  unlist() %>% 
  mutate(
    region = bestOverlap(., gene_regions) %>% 
      fct(levels = names(gene_regions)) %>% 
      fct_relabel(\(x) regions[x])
  )
if (has_features) {
  consensus_peaks$Feature <- bestOverlap(
    consensus_peaks, external_features, var = "feature", missing = "no_feature"
  ) %>%
    factor(levels = names(colours$features)) %>%
    fct_relabel(str_sep_to_title)
}
```


```{r macs2-logs}
macs2_logs <- macs2_path %>%
  file.path(samples$sample, glue("{samples$sample}_callpeak.log")  ) %>%
  importNgsLogs() %>%
  dplyr::select(
    -contains("file"), -outputs, -n_reads, -alt_fragment_length
  ) %>%
  left_join(samples, by = c("name" = "sample")) %>%
  mutate(
    total_peaks = map_int(
      name, function(x) length(individual_peaks[[x]])
    )
  ) 
```

## QC {.tabset}

This section provides a simple series of visualisations to enable detection of any problematic samples.

- Library Sizes: These are the total number of alignments contained in each `bam` file, as passed to `macs2 callpeak` [@Zhang18798982]
- GC Content: Most variation in GC-content should have been identified prior to performing alignments, using common tools such as [FastQC](https://github.com/s-andrews/FastQC), MultiQC [@EwelsMultiQC2016] or `ngsReports` [@WardNgsReports2019]. However, these plots may still be informative for detection of potential sequencing issues not previously addressed
- Peaks Detected: The number of peaks detected within each individual replicate are shown here, and provide clear guidance towards any samples where the IP may have been less successful, or there may be possible sample mis-labelling. Using the settings provided in `config.yml` (i.e. `peaks:qc:outlier_threshold`), any replicates where the number of peaks falls outside the range defined by $\pm$ `r outlier_thresh`-fold of the median peak number within each treatment group **will be marked as failing QC**. Whilst most cell-line generated data-sets are consistent, organoid or solid-tissue samples are far more prone to high variability in the success of the IP step.
- Cross Correlations: Shows the cross-correlation coefficients between read positions across a series of intervals [@LunSmythCsaw2014]. Weak cross-correlations can also indicate low-quality samples. These values are also used to estimate fragment length within each sample, as the peak value of the cross-correlations
- Fraction Of Reads In Peaks (FRIP): This plot shows the proportion of the alignments which fall within peaks identified by `macs2 callpeak`, with the remainder of alignments being assumed to be background [@Landt01092012]. This can provide guidance as to the success of the IP protocol, and the common-use threshold of 1% is indicated as a horizontal line. This value is not enforced as a hard QC criteria, but may be used to manually exclude samples from the file `samples.tsv` of deemed to be appropriate.


```{r macs2-logs-individual}
emphasize.italics.rows(NULL)
any_fail <- any(macs2_logs$qc == "fail")
if (any_fail) emphasize.italics.rows(which(macs2_logs$qc == "fail"))
macs2_logs %>%
  dplyr::select(
    sample = name, label,
    total_peaks, 
    reads = n_tags_treatment, read_length = tag_length,
    fragment_length
  ) %>%
  rename_all(str_sep_to_title )%>%
  pander(
    justify = "llrrrr",
    caption = glue(
      "*Summary of results for `macs2 callpeak` on individual {target} samples.",
      "Total peaks indicates the number retained after applying the FDR ", 
      "threshold of {percent(macs2_fdr)} during peak calling.",
      ifelse(
        any_fail,
        glue(
          "Samples shown in italics were marked for exclusion from ",
          "downstream analysis as they identified a number of peaks beyond ",
          "+/- {outlier_thresh}-fold the median number of ",
          "peaks amongst all samples within the relevant treatment group.",
        ),
        glue(
          "No samples were identified as failing QC based on the number of ",
          "peaks identified relative to the highest quality sample."
        )
      ),
      "Any peaks passing the FDR cutoff, but which overlapped any black-listed",
      "regions were additionally excluded. The fragment length as estimated by",
      "`macs2 predictd` is given in the final column.",
      case_when(
        all(macs2_logs$paired_end) ~
          "All input files contained paired-end reads.*",
        all(!macs2_logs$paired_end) ~
          "All input files contained single-end reads.*",
        TRUE ~
          "Input files were a mixture of paired and single-end reads*"
      ),
      .sep = " "
    )
  )
```

### Library Sizes

```{r plot-macs2-libsize-individual, fig.height=6, fig.cap = glue("*Library sizes for each {target} sample. The horizontal line indicates the mean library size for each treatment group. Any samples marked for exclusion as described above will be indicated with an (F)*")}
macs2_logs %>%
  ggplot(aes(label, n_tags_treatment, fill = qc)) +
  geom_col(position = "dodge") +
  geom_hline(
    aes(yintercept = mn),
    data = . %>%
      group_by(treat) %>%
      summarise(mn = mean(n_tags_treatment)),
    linetype = 2,
    col = "grey"
  ) +
  facet_grid(~treat, scales = "free_x", space = "free_x") +
  scale_y_continuous(expand = expansion(c(0, 0.05)), labels = comma) +
  scale_fill_manual(values = colours$qc) +
  labs(
    x = "Sample", y = "Library Size", fill = "QC"
  ) +
  ggtitle(glue("{target}: Library Sizes"))
```


### Peaks Detected

```{r plot-macs2-peaks-individual, fig.height=6, fig.cap = glue("*Peaks identified for each {target} sample. The number of peaks passing the inclusion criteria for `macs2 callpeak` (FDR < {macs2_fdr}) are provided. Any samples marked for exclusion are coloured as indicated in the figure legend.*")}
suppressWarnings(
  macs2_logs %>%
    ggplot(aes(label, total_peaks, fill = qc)) +
    geom_col() +
    geom_label(
      aes(x = label, y = total_peaks, label = lab, colour = qc),
      data = . %>%
        dplyr::filter(total_peaks > 0) %>% 
        mutate(
          lab = comma(total_peaks, accuracy = 1), total = total_peaks
        ),
      nudge_y = 0.03 * max(macs2_logs$total_peaks),
      inherit.aes = FALSE, show.legend = FALSE
    ) +
    facet_grid(~treat, scales = "free_x", space = "free_x") +
    scale_y_continuous(expand = expansion(c(0, 0.05)), labels = comma) +
    scale_fill_manual(values = colours$qc) +
    scale_colour_manual(values = colours$qc) +
    labs(
      x = "Sample", y = "Total Peaks", fill = "QC"
    ) +
    ggtitle(glue("{target}: Number of Peaks"))
)
```


### FRIP

```{r plot-frip, fig.height=6, fig.cap = glue("*Fraction of Reads In Peaks for each sample. Higher values indicate more reads specifically associated with the ChIP target ({target}). The common-use minimum value for an acceptable sample (1%) is shown as a dashed horizontal line*")}
samples$sample %>%
  bplapply(
    function(x) {
      gr <- individual_peaks[[x]]
      rip <- 0
      if (length(gr) > 0) {
        sbp <- ScanBamParam(which = gr)
        rip <- sum(countBam(bfl[[x]], param = sbp)$records)
      }
      tibble(name = x, reads_in_peaks = rip)
    }
  ) %>%
  bind_rows() %>%
  left_join(macs2_logs, by = "name") %>%
  mutate(frip = reads_in_peaks / n_tags_treatment) %>%
  ggplot(aes(label, frip, fill = qc)) +
  geom_col() +
  geom_hline(yintercept = 0.01, colour = "grey", linetype = 2) +
  facet_grid(cols = vars(treat), scales = "free_x", space = "free") +
  scale_y_continuous(labels = percent, expand = expansion(c(0, 0.05))) +
  scale_fill_manual(values = colours$qc) +
  labs(x = "Sample", y = "Fraction of Reads In Peaks", fill = "QC") +
  ggtitle(glue("{target}: Fraction Of Reads In Peaks"))
```


### Cross Correlations

```{r plot_correlation, fig.height=6, fig.cap = glue("*Cross Correlaton between alignments up to 1kb apart. The dashed, grey, vertical line is the fragment length estimated by `macs2 callpeak` for each sample, with labels indicating the approximate point of the highest correlation, as representative of the average fragment length. For speed, only the first 5 chromosomes were used for sample-specific estimates.*")}
macs2_path %>%
  file.path(target, glue("{target}_cross_correlations.tsv")) %>% 
  read_tsv() %>% 
  left_join(samples, by = "sample") %>% 
  ggplot(aes(fl, correlation, colour = treat)) +
  geom_point(alpha = 0.1) +
  geom_smooth(se = FALSE, method = 'gam', formula = y ~ s(x, bs = "cs")) +
  geom_vline(
    aes(xintercept = fragment_length),
    data = macs2_logs,
    colour = "grey40", linetype = 2
  ) +
  geom_label(
    aes(label = fl),
    data = . %>% 
      mutate(correlation = correlation + 0.03 * max(correlation)) %>% 
      dplyr::filter(correlation == max(correlation), .by = sample),
    show.legend = FALSE, alpha = 0.7
  ) +
  facet_wrap(~label) +
  scale_colour_manual(values = colours$treat[treat_levels]) +
  scale_x_continuous(
    breaks = seq(0, 5*max(macs2_logs$fragment_length), by = 200)
  ) +
  labs(
    x = "Distance (bp)", y = "Cross Correlation", colour = "Treat"
  ) +
  ggtitle(glue("{target}: Cross Correlations"))
```


### GC Content

```{r set-yield-size}
ys <- 5e5
yieldSize(bfl) <- ys
```


```{r plot-gc-content, dev = 'png', fig.height = 6, fig.cap = glue("*GC content for each bam file, taking the first {comma(ys)} alignments from each sample. QC status is based on the number of peaks identified (see table above)*")}
bfl %>% 
  bplapply(
    function(x){
      seq <- scanBam(x, param = ScanBamParam(what = "seq"))[[1]]$seq
      freq <- letterFrequency(seq, letters = "GC") / width(seq)
      list(freq[,1])
    }
  ) %>% 
  as_tibble() %>% 
  pivot_longer(
    cols = everything(), names_to = "name", values_to = "freq"
  ) %>% 
  left_join(macs2_logs, by = "name") %>% 
  unnest(freq) %>% 
  ggplot(aes(label, freq, fill = qc)) +
  geom_boxplot(alpha = 0.8) +
  geom_hline(
    aes(yintercept = med),
    data = . %>% 
      group_by(treat) %>% 
      summarise(med = median(freq)),
    linetype = 2,
    colour = rgb(0.2, 0.2, 0.8)
  ) +
  facet_grid(~treat, scales = "free_x", space = "free_x") +
  scale_y_continuous(labels = percent) +
  scale_fill_manual(values = colours$qc) +
  labs(x = "Sample", y = "GC content", fill = "QC") +
  ggtitle(glue("{target}: GC Content"))
```



## Results 

### Individual Replicates {.tabset}

Replicates within treatment groups are only shown where at least one peak was detected.
For two or three replicates, Euler (Venn) Diagrams were generated using the `python` package `matplotlib-venn`.
Peak numbers may differ from above as those shown below are overlapping peaks which have been merged across replicates

In the case of four or more replicates, the package `ComplexUpset` was used to create UpSet plots [@Upset2014], with colours indicating QC status.

#### All Samples

```{r all-reps-upset, fig.cap = "*UpSet plot showing all samples including those which failed prior QC steps. Any potential sample mislabelling will show up clearly here as samples from each group should show a preference to overlap other samples within the same treatment group. Intersections are only included if 10 or more sites are present. The top panel shows a boxplot of the median $q$-values produced by `macs2 callpeak` for each peak in the intersection. The y-axis for this panel is truncated at the 99^th^ percentile of values. Only intersections with 10 or more peaks are shown.*"}
sample2label <- setNames(samples$label, samples$sample)
sets <- list(all = paste(samples$label, collapse = "; ")) %>% 
  c(
    samples %>% 
      split(.$treat) %>% 
      lapply(function(x) paste(x$label, collapse = "; "))
  ) %>% 
  unlist()
valid_sets <- makeConsensus(individual_peaks) %>% 
  as_tibble() %>% 
  pivot_longer(
    cols = all_of(samples$sample), names_to = "sample", values_to = "has_peak"
  ) %>% 
  dplyr::filter(has_peak) %>% 
  mutate(sample = sample2label[sample]) %>% 
  summarise(
    samples = paste(sample, collapse = "; "), .by = range
  ) %>% 
  distinct(samples) %>% 
  dplyr::filter(samples %in% sets) %>% 
  pull(samples) %>% 
  setNames(vapply(., function(x){names(which(sets == x))}, character(1))) %>% 
  lapply(function(x) str_split(x, "; ")[[1]])
## Use upset_query objects to fill key intersections and sets
ql <- samples$label %>% 
  lapply(
    function(x) upset_query(
      set = x, 
      fill = colours$treat[as.character(dplyr::filter(samples, label == x)$treat)]
    )
  )
if (length(valid_sets) > 0) {
  ql <- c(
    ql,
    names(valid_sets) %>% 
      lapply(
        function(x) {
          col <- ifelse(x == "all", "darkorange", colours$treat[[x]])
          upset_query(
            intersect = valid_sets[[x]], 
            only_components = "intersections_matrix", 
            color = col, fill = col
          )
        }
      )
  )
}
lb <- rev(arrange(samples, treat)$label)
size <- get_size_mode('exclusive_intersection')
individual_peaks %>% 
  setNames(sample2label[names(.)]) %>% 
  .[lb] %>% 
  plotOverlaps(
    var = "qValue", f = "median",
    .sort_sets = FALSE, queries = ql,
    base_annotations = list(
      `Peaks in Intersection` = intersection_size(
        text_mapping = aes(label = comma(!!size)),
        bar_number_threshold = 1, text_colors = "black", 
        text = list(size = 3.5, angle = 90, vjust = 0.5, hjust = -0.1)
      ) +
        scale_y_continuous(expand = expansion(c(0, 0.25)), label = comma) +
        theme(
          panel.grid = element_blank(), 
          axis.line = element_line(colour = "grey20"),
          panel.border = element_rect(colour = "grey20", fill =  NA)
        )
    ),
    annotations = list(
      qValue = ggplot(mapping = aes(y = qValue)) +
        geom_boxplot(na.rm = TRUE, outlier.colour = rgb(0, 0, 0, 0.2)) +
        coord_cartesian(
          ylim = c(0, quantile(unlist(individual_peaks)$qValue, 0.99))
        ) +
        scale_y_continuous(expand = expansion(c(0, 0.05))) +
        ylab(expr(paste("Macs2 ", q[med]))) +
        theme(
          panel.grid = element_blank(), 
          axis.line = element_line(colour = "grey20"),
          panel.border = element_rect(colour = "grey20", fill =  NA)
        )
    ),
    set_sizes = (
      upset_set_size() +
        geom_text(
          aes(label = comma(after_stat(count))), 
          hjust = 1.1, stat = 'count', size = 3.5
        ) +
        scale_y_reverse(expand = expansion(c(0.3, 0)), label = comma) +
        ylab(glue("Macs2 Peaks (FDR < {macs2_fdr})")) +
        theme(
          panel.grid = element_blank(), 
          axis.line = element_line(colour = "grey20"),
          panel.border = element_rect(colour = "grey20", fill =  NA)
        )
    ),
    n_intersections = 30
  ) +
  theme(
    panel.grid = element_blank(), 
    panel.border = element_rect(colour = "grey20", fill =  NA),
    axis.line = element_line(colour = "grey20")
  ) +
  labs(x = "Intersection") +
  plot_layout(heights = c(2, 3, 2))
```


```{r plot-replicates}
ranges_by_rep_treat <- individual_peaks %>%
  unlist() %>% 
  splitAsList(.$treat) %>% 
  .[vapply(., length, integer(1)) > 1] %>% 
  lapply(select, sample) %>% 
  lapply(reduceMC) %>% 
  lapply(as_tibble) %>% 
  lapply(unnest, sample) %>% 
  lapply(mutate, detected = 1) %>% 
  lapply(left_join, dplyr::select(samples, sample, label), by = "sample") %>%
  lapply(dplyr::select, -sample) %>% 
  lapply(function(x) {
    split(x, x$label) %>% 
      lapply(pull, "range")
  }
  )
## First find the python executable
which_py <- c(
  file.path(Sys.getenv("CONDA_PREFIX"), "bin", "python3"),
  Sys.which("python3")
) %>% 
  .[file.exists(.)] %>% 
  .[1]
stopifnot(length(which_py) == 1)
rep_col <- hcl.colors(3, palette = "Zissou", rev = TRUE)
htmltools::tagList(
  names(ranges_by_rep_treat) %>% 
    lapply(
      function(i) {
        x <- ranges_by_rep_treat[[i]]
        fig_out <- file.path(fig_path, glue("{i}_replicates.{fig_type}"))
        if (length(x) == 1) {
          fig_fun(
            fig_out, 
            width = knitr::opts_chunk$get("fig.width"),
            height = ifelse(
              length(x) < 3,
              knitr::opts_chunk$get("fig.height") * 0.8,
              knitr::opts_chunk$get("fig.height") * 0.8
            )
          )
          grid.newpage()
          draw.single.venn(length(x[[1]]), category = names(x))
          dev.off()
        }
        ## Euler diagrams for 2-3 replicates.
        if (length(x) == 2) {
          a12 <- sum(duplicated(unlist(x)))
          a1 <- length(x[[1]]) - a12 
          a2 <- length(x[[2]]) - a12
          args <- glue(
            here::here("workflow", "scripts", "plot_venn.py "),
            "-a1 {a1} -a2 {a2} -a12 {a12} ",
            "-s1 '{names(x)[[1]]}' -s2 '{names(x)[[2]]}' ", 
            "-c1 '{rep_col[[1]]}' -c2 '{rep_col[[3]]}' ",
            "-ht {knitr::opts_chunk$get('fig.height')} ", 
            "-w {knitr::opts_chunk$get('fig.width')} ", 
            "-o '{fig_out}'"
          )
          system2(which_py, args)
        }
        if (length(x) == 3) {
          a123 <- sum(table(unlist(x)) == 3)
          a12 <- sum(table(unlist(x[1:2])) == 2) - a123
          a13 <- sum(table(unlist(x[c(1, 3)])) == 2) - a123
          a23 <- sum(table(unlist(x[c(2, 3)])) == 2) - a123
          a1 <- length(x[[1]]) - (a12 + a13 + a123)
          a2 <- length(x[[2]]) - (a12 + a23 + a123)
          a3 <- length(x[[3]]) - (a13 + a23 + a123)
          args <- glue(
            here::here("workflow", "scripts", "plot_venn.py "),
            "-a1 {a1} -a2 {a2} -a12 {a12} -a3 {a3} ",
            "-a13 {a13} -a23 {a23} -a123 {a123} ",
            "-s1 '{names(x)[[1]]}' -s2 '{names(x)[[2]]}' -s3 '{names(x)[[3]]}' ", 
            "-c1 '{rep_col[[1]]}' -c2 '{rep_col[[2]]}' -c3 '{rep_col[[3]]}' ",
            "-ht {knitr::opts_chunk$get('fig.height')} ", 
            "-w {knitr::opts_chunk$get('fig.width')} ", 
            "-o '{fig_out}'"
          )
          system2(which_py, args)
        }
        ## An UpSet plot for > 3 replicates
        if (length(x) > 3) {
          fig_fun(
            fig_out, 
            width = knitr::opts_chunk$get("fig.width"),
            height = knitr::opts_chunk$get("fig.height")
          )
          grid.newpage()
          ## Prep the data for ComplexUpset
          df <- individual_peaks[dplyr::filter(samples, treat == i)$sample] %>% 
            makeConsensus(var = "qValue") %>% 
            mutate(qValue = vapply(qValue, median, numeric(1))) %>% 
            as_tibble() %>% 
            dplyr::rename_with(
              function(x) setNames(samples$label, samples$sample)[x],
              any_of(samples$sample)
            )
          ## Highlight samples by QC
          ql <- samples %>% 
            dplyr::filter(treat == i) %>% 
            pull('label') %>% 
            lapply(
              function(x) {
                i <- as.character(
                  dplyr::filter(samples, treat == i, label == x)$qc
                )
                upset_query(set = x, fill = colours$qc[i])
              }
            )
          ## And plot
          p <- df %>% 
            upset(
              intersect = rev(dplyr::filter(samples, treat == i)$label),
              base_annotations = list(
                `Peaks in Intersection` = intersection_size(
                  text_mapping = aes(label = comma(!!size)),
                  bar_number_threshold = 1, text_colors = "black", 
                  text = list(size = 4, hjust = 0.5)
                ) +
                  scale_y_continuous(expand = expansion(c(0, 0.2)), label = comma) +
                  theme(
                    text = element_text(size = 14),
                    panel.grid = element_blank(), 
                    axis.line = element_line(colour = "grey20"),
                    panel.border = element_rect(colour = "grey20", fill =  NA)
                  )
              ),
              annotations = list(
                qValue = ggplot(mapping = aes(y = qValue)) +
                  geom_boxplot(na.rm = TRUE, outlier.colour = rgb(0, 0, 0, 0.2)) +
                  coord_cartesian(ylim = c(0, quantile(df$qValue, 0.99))) +
                  scale_y_continuous(expand = expansion(c(0, 0.05))) +
                  ylab(expr(paste("Macs2 ", q[med]))) +
                  theme(
                    text = element_text(size = 14),
                    panel.grid = element_blank(), 
                    axis.line = element_line(colour = "grey20"),
                    panel.border = element_rect(colour = "grey20", fill =  NA)
                  )
              ),
              set_sizes = (
                upset_set_size() +
                  geom_text(
                    aes(label = comma(after_stat(count))), 
                    hjust = 1.1, stat = 'count', size = 4
                  ) +
                  scale_y_reverse(expand = expansion(c(0.2, 0)), label = comma) +
                  ylab(glue("Macs2 Peaks (FDR < {macs2_fdr})")) +
                  theme(
                    text = element_text(size = 14),
                    panel.grid = element_blank(), 
                    axis.line = element_line(colour = "grey20"),
                    panel.border = element_rect(colour = "grey20", fill =  NA)
                  )
              ),
              queries = ql,
              min_size = 10,
              sort_sets = FALSE
            ) +
            theme(text = element_text(size = 14)) +
            patchwork::plot_layout(heights = c(2, 3, 2))
          print(p)
          dev.off()
        }
        
        ## Define the caption
        cp <- htmltools::tags$em(
          glue(
            "
            Number of {target} peaks detected by macs2 callpeak in each {i} 
            replicate and the number of peaks shared between replicates. 
            Replicates which passed/failed QC are show in the specified colours.
            "
          )
        )
        
        ## Create html tags
        fig_link <- str_extract(fig_out, "assets.+")
        htmltools::div(
          htmltools::div(
            id = glue("{i}-replicates"),
            class = "section level4",
            htmltools::h4(i),
            htmltools::div(
              class = "figure", style = "text-align: center",
              htmltools::img(src = fig_link, width = 960),
              htmltools::p(
                class = "caption", htmltools::tags$em(cp)
              )
            )
          )
        )
      }
    )
)
```

### Individual Peak Widths

```{r plot-peak-width, fig.cap = glue("*Range of peak widths across all samples. The median value for all peaks ({median(width(unlist(individual_peaks)))}bp) is shown as the dashed horizontal line*")}
individual_peaks %>% 
  width() %>% 
  as.list() %>% 
  lapply(list) %>% 
  as_tibble() %>% 
  pivot_longer(cols = everything(), names_to = "sample", values_to = "width") %>% 
  unnest(width) %>% 
  left_join(samples, by = "sample") %>% 
  ggplot(aes(label, width, fill = treat)) +
  geom_boxplot(alpha = 0.9) +
  geom_hline(
    yintercept = median(width(unlist(individual_peaks))),
    linetype = 2
  ) +
  facet_wrap(~treat, nrow = 1, scales = "free_x") +
  scale_y_log10() +
  scale_fill_manual(values = colours$treat) +
  labs(x = "Sample", y = "Peak Width (bp)", fill = "Treatment")
```


### Filtered Peaks 

A set of treatment-specific `r target` peaks was defined for each of `r pander(treat_levels)` by comparing the peaks detected when merging all treatment-specific samples, against those detected within each replicate.
Replicates which failed the previous QC steps were omitted during peak-calling.

Peaks detected when merging samples, and which overlapped a peak in more than `r percent(min_prop)` of the *individual replicates passing QC* in each treatment group, were retained.
Those additionally overlapping any black- or grey-listed regions were also removed as viable peaks.

```{r tab-treatment-peak}
treat_levels %>% 
  lapply(
    \(i) tibble(
      treat = i,
      detected_peaks = length(merged_peaks[[i]]),
      retained = length(filtered_peaks[[i]]),
      `% retained` = percent(retained / detected_peaks, 0.1),
      median_width = median(width(filtered_peaks[[i]]))
    )
  ) %>% 
  bind_rows() %>% 
  mutate(treat = factor(treat, levels = treat_levels)) %>% 
  group_by(treat) %>% 
  rename_all(str_sep_to_title) %>% 
  dplyr::rename(`% Retained` = ` Retained`) %>% 
  pander(
    justify = "lrrrr",
    caption = glue(
      "Treatment-specific peaks detected by merging samples within each",
      "treatment group. Peaks were only retained if detected in at least",
      "{percent(min_prop)} of the qc-passing samples for each treatment group,",
      "as described above.",
      .sep = " "
    )
  )
```


### Consensus Peaks {.tabset}

After filtering peaks, a set of `r comma(length(consensus_peaks))` treatment-agnostic `r target` **consensus peaks** were defined.
These ranges represent the *union* of all overlapping ranges defined and retained within any treatment.

Resulting values for the `score`, `signalValue`, `pValue` and `qValue` were calculated as the *median* across all treatments.
*Peak summits* were taken as the median position of all summits which comprise the union peak.

#### `r ifelse(length(treat_levels) < 3, "Venn Diagram", "Upset Plot")`

```{r plot-common-peaks, fig.width=7, fig.height=7, fig.cap = glue("*Number of {target} consensus peaks which overlap the filtered peaks defined within each condition.*")}
plotOverlaps(
  filtered_peaks, 
  type = ifelse(length(treat_levels) > 2, "upset", "venn"),
  set_col = colours$treat[treat_levels]
)
```


#### Distance to TSS 

```{r plot-union-dist-to-tss-hist, fig.height=6, fig.cap = glue("*Distances from the centre of the {target} consensus peak to the transcription start-site shown as A) a histogram, and B) as a cumulative distribution. The vertical dashed line indicates the range considered to be a promoter during annotation preparation. {comma(sum(overlapsAny(consensus_peaks, tss)))} of the {comma(length(consensus_peaks))} {target} consensus peaks ({percent(mean(overlapsAny(consensus_peaks, tss)), 0.1)}) directly overlapped a TSS.*")}
a <- consensus_peaks %>% 
  join_nearest(mutate(tss, tss = start)) %>% 
  mutate(d = start + width/2 - tss) %>% 
  as_tibble() %>% 
  ggplot(aes(d / 1e3)) +
  geom_density() +
  coord_cartesian(xlim = c(-100, 100)) +
  labs(
    x = "Distance Around Nearest TSS (kb)",
    y = "Density"
  )
b <- consensus_peaks %>% 
  join_nearest(
    mutate(tss, tss = start)
  ) %>% 
  mutate(d = start + width/2 - tss) %>%
  as_tibble() %>% 
  select(d) %>% 
  arrange(abs(d)) %>% 
  mutate(
    q = seq_along(d) / nrow(.)
  ) %>% 
  ggplot(aes(abs(d) / 1e3, q)) +
  geom_line() +
  geom_vline(
    xintercept = max(unlist(extra_params$gene_regions$promoters)) / 1e3, 
    linetype = 2, colour = "grey40"
  ) +
  coord_cartesian(xlim = c(0, 100)) +
  labs(
    x = "Absolute Distance to Nearest TSS (kb)",
    y = "Percentile"
  ) +
  scale_y_continuous(labels = percent, breaks = seq(0, 1, by = 0.2)) +
  scale_x_continuous(breaks = seq(0, 100, by = 20))
a + b + plot_annotation(tag_levels = "A")
```



#### Gene-Centric Regions

```{r plot-region-overlap, fig.height=8, fig.width=8, fig.cap = glue("*Proportions of {target} consensus peaks which overlap gene-centric features.*")}
consensus_peaks %>% 
  plotPie(
    fill = "region", total_size = 4,
    cat_glue = "{str_wrap(.data[[fill]], 15)}\n{comma(n, 1)}\n({percent(p, 0.1)})",
    cat_alpha = 0.9, cat_adj = 0.05, min_p = 0.025
  ) +
  scale_fill_manual(
    values = colours$regions %>% setNames(regions[names(.)])
  ) +
  theme(legend.position = "none")
```

`r ifelse(has_features, "#### External Features", "")`

```{r plot-feature-overlap, eval = has_features, echo = has_features, fig.height=8, fig.width=8, fig.cap = glue("*The total number of {target} consensus peaks which overlap external features provided in {basename(config$external$features)}. If peaks map to multiple features, they are assigned to the one with the largest overlap.*")}
consensus_peaks %>% 
  plotPie(
    fill = "Feature", total_size = 4,
    cat_glue = "{str_wrap(.data[[fill]], 15)}\n{comma(n, 1)}\n({percent(p, 0.1)})",
    cat_alpha = 0.9, cat_adj = 0.05
  ) +
  scale_fill_manual(
    values = colours$features %>% setNames(str_sep_to_title(names(.)))
  ) +
  theme(legend.position = "none")
```

`r ifelse(has_features, "#### External Features And Gene-Centric Regions", "")`

```{r plot-feature-region-overlap, eval = has_features, echo = has_features, fig.cap = glue("*The total number of {target} consensus peaks overlapping external features and gene-centric regions. If a peak overlaps multiple features or regions, it is assigned to be the one with the largest overlap. Any peaks which don't overlap a feature have been excluded.*")}
consensus_peaks %>% 
  plotSplitDonut(
    inner = "Feature", outer = "region",
    inner_palette = colours$features %>% setNames(str_to_title(names(.))),
    outer_palette = colours$regions %>% setNames(regions[names(.)]),
    inner_glue = "{str_wrap(.data[[inner]], 15)}\nn = {comma(n, 1)}\n{percent(p,0.1)}",
    outer_glue = "{str_wrap(.data[[outer]], 15)}\nn = {comma(n, 1)}\n{percent(p,0.1)}",
    min_p = 0.03, outer_alpha = 0.8
  )
```


## Highly Ranked Peaks {.tabset}

```{r grl-to-plot}
grl_to_plot <- GRangesList()
grl_to_plot$consensus <- consensus_peaks %>% 
  subset(score == max(score)) %>% 
  granges()
for (i in treat_levels) {
  grl_to_plot[[i]] <- filtered_peaks[[i]] %>% 
    filter_by_non_overlaps(unlist(grl_to_plot)) %>% 
    subset(score == max(score)) %>% 
    granges() %>% 
    unname()
}
```

```{r prepare-plot-objects}
## The coverage
bwfl <- list2(
  "{target}" := macs2_path %>%
    file.path(target, glue("{target}_{treat_levels}_merged_treat_pileup.bw")) %>% 
    BigWigFileList() %>% 
    setNames(treat_levels)
)
## The features track
feat_gr <- gene_regions %>% 
  lapply(granges) %>% 
  GRangesList()
feature_colours <- colours$regions
if (has_features) {
  feat_gr <- list(Regions = feat_gr)
  feat_gr$Features <- splitAsList(external_features, external_features$feature)
  feature_colours <- list(
    Regions = unlist(colours$regions),
    Features = unlist(colours$features)
  )
}
## The genes track defaults to transcript models
hfgc_genes <- read_rds(
  here::here("output", "annotations", "trans_models.rds")
)
gene_col <- "grey"
## If RNA-Seq data is present, the genes track switches to Up/Down etc
if (nrow(rnaseq)) {
  rna_lfc_col <- colnames(rnaseq)[str_detect(str_to_lower(colnames(rnaseq)), "logfc")][1]
  rna_fdr_col <- colnames(rnaseq)[str_detect(str_to_lower(colnames(rnaseq)), "fdr|adjp")][1]
  if (!is.na(rna_lfc_col) & !is.na(rna_fdr_col)) {
    hfgc_genes <- hfgc_genes %>%
      mutate(
        status = case_when(
          !gene %in% rnaseq$gene_id ~ "Undetected",
          gene %in% dplyr::filter(
            rnaseq, !!sym(rna_lfc_col) > 0, !!sym(rna_fdr_col) < fdr_alpha
          )$gene_id ~ "Up",
          gene %in% dplyr::filter(
            rnaseq, !!sym(rna_lfc_col) < 0, !!sym(rna_fdr_col) < fdr_alpha
          )$gene_id ~ "Down",
          gene %in% dplyr::filter(
            rnaseq, !!sym(rna_fdr_col) >= fdr_alpha
          )$gene_id ~ "Unchanged"
        )
      ) %>%
      splitAsList(.$status) %>%
      lapply(select, -status) %>%
      GRangesList()
    gene_col <- colours$direction %>%
      c(
        c(
          up = colours$direction[["increased"]], 
          down = colours$direction[["decreased"]]
        )
      ) %>% 
      setNames(str_to_title(names(.))) 
      
  }
}
## External Coverage (Optional)
if (!is.null(config$external$coverage)) {
  ext_cov_path <- config$external$coverage %>% 
    lapply(unlist) %>% 
    lapply(function(x) setNames(here::here(x), names(x)))
  bwfl <- c(
    bwfl[target],
    config$external$coverage %>% 
      lapply(
        function(x) {
          BigWigFileList(here::here(unlist(x))) %>% 
            setNames(names(x))
        }
      )
  )
}
line_col <- lapply(bwfl, function(x) colours$treat[names(x)])
y_lim <- bwfl %>%
  lapply(
    function(x) {
      gr <- unlist(grl_to_plot) %>% 
        resize(width = 30 * width(.), fix = 'center')
      x %>% 
        lapply(import.bw, which = gr) %>%
        lapply(function(rng) c(0, max(rng$score))) %>%
        unlist() %>%
        range()
    }
  )
```

Coverage for a small set of highly ranked peaks are shown below.
These are the most highly ranked Consensus Peak (by `score`) and the most highly ranked peak for each of `r pander(treat_levels)` after incrementally stepping through and excluding any already selected peak.


```{r plot-highly-ranked}
htmltools::tagList(
  mclapply(
    seq_along(grl_to_plot),
    function(x) {
      nm <- names(grl_to_plot)[[x]]
      ## Export the figure
      fig_out <- file.path(
        fig_path,
        nm %>% 
          str_replace_all(" ", "_") %>% 
          paste0("_topranked.", fig_type)
      )
      fig_fun(
        filename = fig_out,
        width = knitr::opts_current$get("fig.width"), 
        height = knitr::opts_current$get("fig.height")
      )
      ## Automatically collapse Transcripts if more than 10
      ct <- FALSE
      gh <- 1
      if (length(subsetByOverlaps(gtf_transcript, grl_to_plot[[x]])) > 20) {
        ct <- "meta"
        gh <- 0.5
      }
      ## Generate the plot
      plotHFGC(
        grl_to_plot[[x]],
        features = feat_gr, featcol = feature_colours, featsize = 1 + has_features,
        genes = hfgc_genes, genecol = gene_col,
        coverage = bwfl, linecol = line_col,
        cytobands = bands_df,
        rotation.title = 90,
        zoom = 30,
        ylim = y_lim,
        collapseTranscripts = ct, genesize = gh,
        col.title = "black", background.title = "white", 
        showAxis = FALSE
      )
      dev.off()
      
      ## Define the caption
      gr <- join_nearest(grl_to_plot[[x]], gtf_gene, distance = TRUE)
      d <- gr$distance
      gn <- gr$gene_name
      peak_desc <- ifelse(
        nm == "union",
        "union peak by combined score across all treatments.",
        paste(
          "treatment-peak unique to the merged", nm, "samples."
        )
      )
      cp <- htmltools::tags$em(
        glue(
          "The most highly ranked {peak_desc} ",
          ifelse(
            d == 0,
            paste('The peak directly overlaps', gn),
            paste0("The nearest gene was ", gn, ", ", round(d/1e3, 1), "kb away")
          ),
          ifelse(
            (nrow(rnaseq) > 0) & (gn %in% gtf_gene$gene_name),
            glue(" which was detected in the RNA-Seq data."),
            glue(".")
          ),
          "
          Y-axes for each coverage track are set to the most highly ranked peak
          across all conditions.
          "
        )
      )
      
      ## Create html tags
      fig_link <- str_extract(fig_out, "assets.+")
      htmltools::div(
        htmltools::div(
          id = nm %>% 
            str_replace_all(" ", "-") %>% 
            str_to_lower() %>% 
            paste0("-topranked"),
          class = "section level3",
          htmltools::h3(
            ifelse(
              nm == "union",
              "Union Peaks",
              paste("Treatment Peaks:", nm)
            )
          ),
          htmltools::div(
            class = "figure", style = "text-align: center",
            htmltools::img(src = fig_link, width = 960),
            htmltools::p(
              class = "caption", htmltools::tags$em(cp)
            )
          )
        )
      )
    },
    mc.cores = min(length(grl_to_plot), threads)
  )
)  
```

## Enrichment Analysis

### GREAT

As a brief overview of pathways and genes likely to be regulated by `r target`, the package `rGREAT` [@Gu2023-rp] was used to test for enrichment of pathways and genesets.
This approach takes a set of 'background' genomic ranges and assesses relative enrichment of matches to a given pathway amongst the provided peaks.
For this specific anaysis, the background ranges were set to be the whole genome, excluding the ranges defined as gaps by UCSC.


```{r setup-great}
## Set parameters for enabling networks
min_network_size <- extra_params$networks$min_size
if (is.null(min_network_size) | is.na(min_network_size))
  min_network_size <- 2
max_network_size <- extra_params$networks$max_size
if (is.null(max_network_size) | is.na(max_network_size))
  max_network_size <- Inf
max_network_dist <- extra_params$networks$max_distance
if (is.null(max_network_dist) | is.na(max_network_dist))
  max_network_dist <- 1
net_layout <- extra_params$networks$layout
enrich_params <- extra_params$enrichment
adj_method <- match.arg(enrich_params$adj, p.adjust.methods)
map_great_refs <- function(x = NULL){
  map <- c(
    hg19 = "hg19", hg38 = "hg38", grch37 = "hg19", grch38 = "hg38",
    mm10 = "mm10", mm39 = "mm39", grcm38 = "mm10", grcm39 = "mm39",
    rn7 = "rn7", mratbn7.2 = "rn7", galgal6 = "galGal6", rhemac10 = "rheMac10",
    canfam5 = "canFam5", susscr11 = "susScr11", pantro6 = "panTro6", dm6 = "dm6"
  )
  x <- match.arg(tolower(x), names(map))
  map[[x]]
}
ucsc_ref <- map_great_refs(config$genome$build)
sp <- c(
  hg19 = "Homo sapiens", hg38 = "Homo sapiens", mm10 = "Mus musculus",
  mm39 = "Mus musculus", rn7 = "Rattus norvegicus", susScr11 = "Sus scrofa",
  canFam5 = "Canis lupus familiaris", dm6 = "Drosophila melanogaster",
  galGal6 = "Gallus gallus", panTro6 = "Pan troglodytes"
)[[ucsc_ref]]
ext_tss <- gtf_gene %>% 
  select(gene_id) %>% 
  extendTSS(
    genome = ucsc_ref, gene_id_type = 'ENSEMBL', 
    basal_upstream = extra_params$gene_regions$promoters$upstream,
    basal_downstream = extra_params$gene_regions$promoters$downstream
  )
ext_tss_ol <- subsetByOverlaps(ext_tss, consensus_peaks)
msigdb <- msigdbr(sp) %>% 
  dplyr::filter(
    gs_cat %in% enrich_params$msigdb$gs_cat |
      gs_subcat %in% enrich_params$msigdb$gs_subcat,
    ensembl_gene %in% ext_tss$gene_id
  )
gs_list <- msigdb %>% 
  split(.$gs_name) %>% 
  lapply(pull, ensembl_gene) %>% 
  .[map_int(., length) <= enrich_params$max_size] %>% 
  .[map_int(., length) >= enrich_params$min_size] 
```

Genes from the supplied GTF were used to define the TSS, giving only one TSS for each gene.
Whilst unrealistic, this ensures model assumptions are not violated.
This gave a total of `r comma(length(ext_tss))` TSS regions were defined for analysis.

The `r length(gs_list)` pathways tested were `r pander(unname(unlist(enrich_params$msigdb)))` as defined by MSigDB v`r packageVersion("msigdbr")`.
Pathways will be considered as enriched relative to background if returned an adjusted p-value (`r enrich_params$adj`) below `r enrich_params$alpha`, using the `rGREAT` results from the *hyper-geometric test*, instead of the *binomial test*.

```{r run-great}
great_res <- great(
    set_genome_info(consensus_peaks, genome = ucsc_ref), 
  gene_sets = gs_list, extended_tss = ext_tss, 
  min_gene_set_size = enrich_params$min_size, 
  cores = threads
)
great_df <- getEnrichmentTable(great_res) %>% 
  dplyr::select(-all_of(c("fold_enrichment", "p_adjust", "p_value"))) %>% 
  as_tibble() %>% 
  arrange(p_value_hyper) %>% 
  mutate(
    p_adjust_hyper = p.adjust(p_value_hyper, adj_method),
    gene_id  = gs_list[id],
    genes_with_hits = map_chr(
      gene_id, 
      \(x) {
        ids <- subset(ext_tss_ol, gene_id %in% x)$gene_id
        paste(unname(id2gene[unique(ids)]), collapse = "; ")
      }
    )
  )
n_great <- sum(great_df$p_adjust_hyper < enrich_params$alpha)
```

`r if (n_great == 0) "No enrichment of gene-sets was found"`

```{r tbl-great-res, eval = n_great > 0, echo = n_great > 0}
tbl <- great_df %>% 
  dplyr::filter(p_adjust_hyper < enrich_params$alpha) %>% 
  dplyr::slice(seq_len(100)) %>%
  dplyr::select(-all_of(c("p_value_hyper", "gene_id"))) %>% 
  dplyr::select(id, fold_enrichment_hyper, p_adjust_hyper, everything()) %>% 
  reactable(
    sortable = TRUE, filterable = TRUE,
    columns = list(
      id = colDef(
        name = "Gene Set",
        cell = \(value) str_replace_all(value, "_", " "),
        minWidth = 180
      ),
      genome_fraction = colDef(show = FALSE),
      observed_region_hits = colDef(show = FALSE),
      fold_enrichment_hyper = colDef(
        name = "Fold Enrichment", format = colFormat(digits = 3),
        filterable = FALSE, minWidth = 110
      ),
      p_adjust_hyper = colDef(
        name = glue("P<sub>adj</sub> ({enrich_params$adj})"),
        html = TRUE, filterable = FALSE,
        cell = \(value) sprint_pval(value),
        maxWidth = 110
      ),
      mean_tss_dist = colDef(
        "Mean TSS Distance (kb)", 
        cell = \(value) round(value / 1e3, 2),
        filterable = FALSE
      ),
      observed_gene_hits = colDef(
        "Gene Hits", maxWidth = 100, filterable = FALSE
      ),
      gene_set_size = colDef(
        "Gene Set Size", maxWidth = 100, filterable = FALSE
      ),
      genes_with_hits = colDef(
        name = "Genes",
        cell = \(value) with_tooltip(value, width = 60),
        minWidth = 200
      )
    )
  )
cp <- htmltools::em(
  glue(
    "
    Enrichment results for regions showing consistently detecable signal for 
    {target} as considered significant when using rGREAT. Backround regions to
    determine relative enrichment were taken as the entire genome. In the case
    of large numbers of significant gene-sets, results are limited to the 100 
    most highly ranked.
    "
  )
)
div(class = "table",
    div(class = "table-header",
        div(class = "caption", cp), tbl
    )
)
```

```{r great-network, echo = n_great > min_network_size, eval = n_great > min_network_size, fig.cap = glue("*The {n_great} most highly ranked gene sets from rGREAT analysis. Edge thickness denotes simiarity between connected nodes. Lighter node colours indicate increased significance. Label transparency and the outline thickness of each node also represent the proportion of each gene set mapped to a range with detected {target}. Node size indicates the number of genes mapped to a range. Networks are capped at a maximum of {max_network_size} gene sets*")}
tg_great <- great_df %>% 
  dplyr::filter(p_adjust_hyper < enrich_params$alpha) %>%
  make_tbl_graph(
    gs = great_df[["gene_id"]], 
    alpha = enrich_params$alpha, p_col = "p_adjust_hyper", gs_col = "id", 
    hit_col = "observed_gene_hits", size_col = "gene_set_size", 
    min_dist = max_network_dist, max_gs = max_network_size
  )
if (length(tg_great) >= min_network_size) {
  tg_great %>% 
    mutate(
      ## The lowest p-value obtainable from pbinom is 1e-16
      p_value_hyper = ifelse(p_value_hyper > 0, p_value_hyper, 1e-16),
      prop = observed_gene_hits / gene_set_size 
    ) %>% 
    ggraph(layout = net_layout, weights = oc) +
    geom_edge_link(aes(width = oc^2, alpha = oc^2)) +
    geom_node_point(
      aes(
        fill = -log10(p_value_hyper), size = observed_gene_hits, stroke = 1.5 * prop
      ),
      shape = 21
    ) +
    geom_node_text(
      aes(label = label, alpha = prop),
      colour = "black", size = 3, 
      data = . %>%
        mutate(
          label = str_replace_all(label, "_", " ") %>% 
            str_trunc(60) %>% 
            str_wrap(width = 18)
        ),
      repel = TRUE, max.overlaps = max(10, round(length(tg_great) / 4, 0)),
      bg.color = "white", bg.r = 0.1, 
    ) +
    scale_x_continuous(expand = expansion(c(0.1, 0.1))) +
    scale_y_continuous(expand = expansion(c(0.1, 0.1))) +
    scale_fill_viridis_c(option = "inferno", begin = 0.2) +
    scale_alpha(range = c(0.5, 1), trans = "sqrt") +
    scale_size_continuous(range = c(1, 10)) +
    scale_edge_width(range = c(1, 6), limits = c(0, 1)) +
    scale_edge_alpha(range = c(0.1, 0.4), limits = c(0, 1))  +
    guides(edge_alpha = "none", edge_width = "none", alpha = "none") +
    labs(size = "Mapped\nGenes", fill = expr(paste("-", log[10], "p"))) +
    theme_void() 
}
```


```{r plot-highest-great, eval = n_great > 0, echo = n_great > 0, fig.cap = glue("*Distribution of {target} signal around TSS regions for {top_id}, which was the most highly enriched of the test gene sets.*")}
top_id <- great_df$id[1]
plotRegionGeneAssociations(great_res, top_id, which = 2:3)
```



```{r export, results='hide', echo=FALSE}
all_out <- list(
  great = here::here(macs2_path, target, glue("{target}_great_results.tsv.gz")),
  renv = file.path(
    here::here("output/envs"), glue("{target}_macs2_summary.RData")
  )
)
great_df %>% 
  mutate(gene_id = map_chr(gene_id, paste, collapse = "; ")) %>% 
  write_tsv(all_out$great)
if (!dir.exists(dirname(all_out$renv))) dir.create(dirname(all_out$renv))
save.image(all_out$renv)
```

Files exported during preparation of this document are:

`r pander(all_out)`

## References

<div id="refs"></div>

<br>
<button type="button" class="btn btn-default btn-sessioninfo" data-toggle="collapse" data-target="#sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Session information
</button>
</p>
<div id="sessioninfo" class="collapse">
```{r session-info, echo=FALSE}
pander::pander(sessionInfo())
```
</div>
