```{r example, echo = FALSE, eval = FALSE}
## This module requires the following to be set during preparation
# target <- "AR"
# threads <- 6
```


```{r packages}
library(tidyverse)
library(magrittr)
library(rtracklayer)
library(glue)
library(pander)
library(scales)
library(plyranges)
library(yaml)
library(ngsReports)
library(UpSetR)
library(VennDiagram)
library(cowplot)
library(rlang)
library(BiocParallel)
library(parallel)
library(Rsamtools)
library(Biostrings)
library(ggside)
```

```{r remotes, results = 'hide'}
## Deal with github packages
if (!"extraChIPs" %in% rownames(installed.packages()))
  BiocManager::install("steveped/extraChIPs", ask = FALSE)
stopifnot(library(extraChIPs, logical.return = TRUE))
```

```{r options}
panderOptions("big.mark", ",")
panderOptions("missing", "")
panderOptions("table.split.table", Inf)
theme_set(
  theme_bw() +
    theme(
      plot.title = element_text(hjust = 0.5)
    )
)
register(MulticoreParam(workers = threads))
source(here::here("workflow/scripts/custom_functions.R"))
## This has been pushed to the main repo of ggside, but is not yet
## incorporated into the package. The functions are geom_*sidelabel
## Once ggside 0.2.1 is released, this can be deleted
source(here::here("workflow/scripts/geom_sidelabel.R"))
```


```{r config-paths}
config <- read_yaml(
  here::here("config", "config.yml")
)
fdr_alpha <- config$comparisons$fdr
params <- read_yaml(here::here("config", "params.yml"))
bam_path <- here::here(config$paths$bam, target)
macs2_path <- here::here("output", "macs2", target)
annotation_path <- here::here("output", "annotations")
colours <- read_rds(
  file.path(annotation_path, "colours.rds")
) %>% 
  lapply(unlist)
```

```{r read-samples}
samples <-read_tsv(file.path(macs2_path, "qc_samples.tsv"))
treat_levels <- unique(samples$treat)
if (!is.null(config$comparisons$contrasts)) {
  ## Ensure levels respect those provided in contrasts
  treat_levels <- config$comparisons$contrasts %>%
    unlist() %>%
    intersect(samples$treat) %>% 
    unique()
}
rep_col <- setdiff(
  colnames(samples), c("sample", "treat", "target", "input", "label", "qc")
)
samples <- samples %>%
  unite(label, treat, !!sym(rep_col), remove = FALSE) %>%
  mutate(
    treat = factor(treat, levels = treat_levels),
    "{rep_col}" := as.factor(!!sym(rep_col))
  )
```


```{r annotations}
sq <- read_rds(file.path(annotation_path, "seqinfo.rds"))
blacklist <-  file.path(annotation_path, "blacklist.bed.gz") %>%
  import.bed(seqinfo = sq) %>%
  sort()
all_gr <- file.path(annotation_path, "all_gr.rds") %>% 
  read_rds()
genes_gr <- all_gr$gene
external_features <- c()
has_features <- FALSE
if (!is.null(config$external$features)) {
  external_features <- suppressWarnings(
    import.gff(here::here(config$external$features), genome = sq)
  )
  keep_cols <- !vapply(
    mcols(external_features), function(x) all(is.na(x)), logical(1)
  )
  mcols(external_features) <- mcols(external_features)[keep_cols]
  has_features <- TRUE
}
gene_regions <- read_rds(file.path(annotation_path, "gene_regions.rds"))
regions <- vapply(gene_regions, function(x) unique(x$region), character(1))
any_detected <- gene_regions %>% 
  vapply(function(x) any(x$detected), logical(1)) %>% 
  any()
rna_path <- here::here(config$external$rnaseq)
rnaseq <- tibble(gene_id = character())
if (length(rna_path) > 0) {
  stopifnot(file.exists(rna_path))
  if (str_detect(rna_path, "tsv$")) rnaseq <- read_tsv(rna_path)
  if (str_detect(rna_path, "csv$")) rnaseq <- read_csv(rna_path)
  if (!"gene_id" %in% colnames(rnaseq)) stop("Supplied RNA-Seq data must contain the column 'gene_id'")
  genes_gr <- subset(genes_gr, gene_id %in% rnaseq$gene_id)
}
tx_col <- intersect(c("tx_id", "transcript_id"), colnames(rnaseq))
rna_gr_col <- ifelse(length(tx_col) > 0, "transcript_id", "gene_id")
rna_col <- c(tx_col, "gene_id")[[1]]
tss <- read_rds(file.path(annotation_path, "tss.rds"))
## bands_df
cb <- config$genome$build %>%
  str_to_lower() %>% 
  paste0(".cytobands") 
data(list = cb)
bands_df <- get(cb)
```

```{r bfl}
bfl <- bam_path %>%
  file.path(glue("{samples$sample}.bam")) %>%
  BamFileList() %>%
  setNames(samples$sample)
```

```{r individual-peaks}
individual_peaks <- file.path(
  macs2_path, glue("{samples$sample}_peaks.narrowPeak")
) %>%
  importPeaks(seqinfo = sq, blacklist = blacklist) %>%
  setNames(samples$sample)
```


```{r macs2-logs}
macs2_logs <- file.path(macs2_path, glue("{samples$sample}_callpeak.log")) %>%
  importNgsLogs() %>%
  dplyr::select(
    -contains("file"), -outputs, -n_reads, -alt_fragment_length
  ) %>%
  left_join(samples, by = c("name" = "sample")) %>%
  mutate(
    total_peaks = map_int(
      name,
      function(x) {
        length(individual_peaks[[x]])
      }
    )
  ) 
n_reps <- macs2_logs %>% 
  group_by(treat) %>%
  summarise(n = sum(qc == "pass"))
```

## QC {.tabset}

This section provides a simple series of visualisations to enable detection of any problematic samples.

- Library Sizes: These are the total number of alignments contained in each `bam` file, as passed to `macs2 callpeak` [@Zhang18798982]
- GC Content: Most variation in GC-content should have been identified prior to performing alignments, using common tools such as [FastQC](https://github.com/s-andrews/FastQC), MultiQC [@EwelsMultiQC2016] or `ngsReports` [@WardNgsReports2019]. However, these plots may still be informative for detection of potential sequencing issues not previously addressed
- Peaks Detected: The number of peaks detected within each individual replicate are shown here, and provide clear guidance towards any samples where the IP may have been less successful. Using the settings provided in `config.yml` (i.e. `peaks:qc:min_prop_peaks`), any replicates where the number of peaks falls below `r percent(config$peaks$qc$min_prop_peaks)` of the replicate with the highest number of peaks (i.e. the 'best' sample), **will be marked as failing QC**. Whilst most cell-line generated data-sets are consistent, organoid or solid-tissue samples are far more prone to high variability in the success of the IP step.
- Cross Correlations: Shows the cross-correlation coefficients between read positions across a series of intervals [@LunSmythCsaw2014]. Weak cross-correlations can also indicate low-quality samples. These values are also used to estimate fragment length within each sample, as the peak value of the cross-correlations
- Fraction Of Reads In Peaks (FRIP): This plot shows the proportion of the alignments which fall within peaks identified by `macs2 callpeak`, with the remainder of alignments being assumed to be background [@Landt01092012]. This can provide guidance as to the success of the IP protocol, and the common-use threshold of 1% is indicated as a horizontal line. This value is not enforced as a hard QC criteria, but may be used to manually exclude samples from the file `samples.tsv` of deemed to be appropriate.


```{r macs2-logs-individual}
emphasize.italics.rows(NULL)
any_fail <- any(macs2_logs$qc == "fail")
qc_prop <- config$peaks$qc$min_prop_peaks
if (any_fail) emphasize.italics.rows(which(macs2_logs$qc == "fail"))
macs2_logs %>%
  dplyr::select(
    sample = name, label,
    total_peaks, 
    reads = n_tags_treatment, read_length = tag_length,
    fragment_length
  ) %>%
  rename_all(str_sep_to_title )%>%
  pander(
    justify = "llrrrr",
    caption = glue(
      "*Summary of results for `macs2 callpeak` on individual {target} samples.",
      "Total peaks indicates the number retained after applying the FDR ", 
      "threshold of {percent(config$peaks$macs2$fdr)} during the peak calling ", 
      "process.",
      ifelse(
        any_fail,
        glue(
          "Samples shown in italics were marked (F) for exclusion from ",
          "downstream analysis as they identified fewer than ",
          "{percent(qc_prop)} of the peaks identified in the sample returning ",
          "the maximum number of peaks."
        ),
        glue(
          "No samples were identified as failing QC based on the number of ",
          "peaks identified relative to the highest quality sample."
        )
      ),
      "Any peaks passing the FDR cutoff, but which overlapped any black-listed",
      "regions were additionally excluded. The fragment length as estimated by",
      "`macs2 predictd` is given in the final column.",
      case_when(
        all(macs2_logs$paired_end) ~
          "All input files contained paired-end reads.*",
        all(!macs2_logs$paired_end) ~
          "All input files contained single-end reads.*",
        TRUE ~
          "Input files were a mixture of paired and single-end reads*"
      ),
      .sep = " "
    )
  )
```

### Library Sizes

```{r plot-macs2-libsize-individual, fig.height=6, fig.cap = glue("*Library sizes for each {target} sample. The horizontal line indicates the mean library size for each treatment group. Any samples marked for exclusion as described above will be indicated with an (F)*")}
macs2_logs %>%
  ggplot(
    aes(label, n_tags_treatment, fill = qc)
  ) +
  geom_col(position = "dodge") +
  geom_hline(
    aes(yintercept = mn),
    data = . %>%
      group_by(treat) %>%
      summarise(mn = mean(n_tags_treatment)),
    linetype = 2,
    col = "grey"
  ) +
  facet_grid(~treat, scales = "free_x", space = "free_x") +
  scale_y_continuous(expand = expansion(c(0, 0.05)), labels = comma) +
  scale_fill_manual(values = colours$qc) +
  labs(
    x = "Sample",
    y = "Library Size",
    fill = "QC"
  ) +
  ggtitle(
    glue("{target}: Library Sizes")
  )
```

### GC Content

```{r set-yield-size}
ys <- 5e5
yieldSize(bfl) <- ys
```


```{r plot-gc-content, fig.height = 6, fig.cap = glue("*GC content for each bam file, taking the first {comma(ys)} alignments from each sample. QC status is based on the number of peaks identified (see table above)*")}
bfl %>% 
  bplapply(
    function(x){
      seq <- scanBam(x, param = ScanBamParam(what = "seq"))[[1]]$seq
      freq <- letterFrequency(seq, letters = "GC") / width(seq)
      list(freq[,1])
    }
  ) %>% 
  as_tibble() %>% 
  pivot_longer(
    cols = everything(), 
    names_to = "name",
    values_to = "freq"
  ) %>% 
  left_join(macs2_logs, by = "name") %>% 
  unnest(freq) %>% 
  ggplot(aes(label, freq, fill = qc)) +
  geom_boxplot(alpha = 0.8) +
  geom_hline(
    aes(yintercept = med),
    data = . %>% 
      group_by(treat) %>% 
      summarise(med = median(freq)),
    linetype = 2,
    colour = rgb(0.2, 0.2, 0.8)
  ) +
  facet_grid(~treat, scales = "free_x", space = "free_x") +
  scale_y_continuous(labels = percent) +
  scale_fill_manual(values = colours$qc) +
  labs(
    x = "Sample",
    y = "GC content",
    fill = "QC"
  ) +
  ggtitle(
    glue("{target}: GC Content")
  )
```


### Peaks Detected

```{r plot-macs2-peaks-individual, fig.height=6, fig.cap = glue("*Peaks identified for each {target} sample. The number of peaks passing the inclusion criteria for `macs2 callpeak` (FDR < {config$peaks$macs2$fdr}) are provided. The number of retained peaks required for inclusion is shown for each treatment as the dashed grey line, with any samples marked for exclusion marked with an F.*")}
suppressWarnings(
  macs2_logs %>%
    ggplot(
      aes(label, total_peaks, fill = qc)
    ) +
    geom_col() +
    geom_label(
      aes(x = label, y = total_peaks, label = lab, colour = qc),
      data = . %>%
        mutate(
          lab = comma(total_peaks, accuracy = 1),
          total = total_peaks
        ),
      inherit.aes = FALSE,
      show.legend = FALSE
    ) +
    geom_hline(
      aes(yintercept = min),
      data = . %>%
        group_by(treat) %>%
        summarise(min = config$peaks$qc$min_prop_peaks*max(total_peaks)),
      linetype = 2,
      col = "grey"
    ) +
    facet_grid(~treat, scales = "free_x", space = "free_x") +
    scale_y_continuous(expand = expansion(c(0, 0.05)), labels = comma) +
    scale_fill_manual(values = colours$qc) +
    scale_colour_manual(values = colours$qc) +
    labs(
      x = "Sample",
      y = "Total Peaks",
      fill = "QC"
    ) +
    ggtitle(
      glue("{target}: Number of Peaks")
    )
)
```

### Cross Correlations

```{r plot_correlation, fig.height=6, fig.cap = glue("*Cross Correlaton between alignments up to 1kb apart. The dashed, grey, vertical line is the fragment length estimated by `macs2 callpeak` for each sample. For speed, only the first 5 chromosomes were used for sample-specific estimates.*")}
file.path(macs2_path, "cross_correlations.tsv") %>% 
  read_tsv() %>% 
  left_join(samples, by = "sample") %>% 
  ggplot(aes(fl, correlation, colour = treat)) +
  geom_point(alpha = 0.1) +
  geom_smooth(se = FALSE, method = 'gam', formula = y ~ s(x, bs = "cs")) +
  geom_vline(
    aes(xintercept = fragment_length),
    data = macs2_logs,
    colour = "grey40", linetype = 2
  ) +
  facet_grid(as.formula(paste("treat ~", rep_col))) +
  scale_colour_manual(values = colours$treat[treat_levels]) +
  scale_x_continuous(
    breaks = seq(0, 5*max(macs2_logs$fragment_length), by = 200)
  ) +
  labs(
    x = "Distance (bp)",
    y = "Cross Correlation",
    colour = "Treat"
  ) +
  ggtitle(
    glue("{target}: Cross Correlations")
  )
```


### FRIP

```{r plot-frip, fig.height=6, fig.cap = glue("*Fraction of Reads In Peaks for each sample. Higher values indicate more reads specifically associated with the ChIP target ({target}). The common-use minimum value for an acceptable sample (1%) is shown as a dashed horizontal line*")}
samples$sample %>%
  bplapply(
    function(x) {
      gr <- individual_peaks[[x]]
      rip <- 0
      if (length(gr) > 0) {
        sbp <- ScanBamParam(which = gr)
        rip <- sum(countBam(bfl[[x]], param = sbp)$records)
      }
      tibble(
        name = x,
        reads_in_peaks = rip
      )
    }
  ) %>%
  bind_rows() %>%
  left_join(macs2_logs, by = "name") %>%
  mutate(
    frip = reads_in_peaks / n_tags_treatment
  ) %>%
  ggplot(
    aes(label, frip, fill = qc)
  ) +
  geom_col() +
  geom_hline(yintercept = 0.01, colour = "grey", linetype = 2) +
  facet_grid(~treat, scales = "free_x", space = "free") +
  scale_y_continuous(labels = percent, expand = expansion(c(0, 0.05))) +
  scale_fill_manual(values = colours$qc) +
  labs(
    x = "Sample",
    y = "Fraction of Reads In Peaks",
    fill = "QC"
  ) +
  ggtitle(
    glue("{target}: Fraction Of Reads In Peaks")
  )
```


## Results 

### Oracle Peaks {.tabset}

```{r oracle-peaks}
oracle_peaks <- treat_levels %>%
  sapply(
    function(x) {
      gr <- file.path(macs2_path, glue("{x}_merged_peaks.narrowPeak")) %>%
        importPeaks(seqinfo = sq, blacklist = blacklist) %>% 
        unlist()
      k <- dplyr::filter(n_reps, treat == x)$n * config$peaks$qc$min_prop_reps
      if (k > 0) {
        samp <- dplyr::filter(macs2_logs, treat == x, qc != "fail")$name
        gr$n_reps <- countOverlaps(gr, individual_peaks[samp]) 
        gr$keep <- gr$n_reps >= k
      } else {
        gr <- GRanges(seqinfo = sq)
      }
      gr
    },
    simplify = FALSE
  ) %>%
  GRangesList() 
consensus_peaks <- oracle_peaks %>%
  unlist() %>%
  subset(keep) %>%
  mutate(peak_position = start + peak) %>% 
  reduceMC() %>% 
  setNames(as.character(.)) %>% 
  mutate(
    score = vapply(score, max, numeric(1)),
    signalValue = vapply(signalValue, max, numeric(1)),
    pValue = vapply(pValue, max, numeric(1)),
    qValue = vapply(qValue, max, numeric(1)),
    n_reps = vapply(n_reps, sum, integer(1)),
    peak = round(vapply(peak_position, median, numeric(1)) - start, 0),
    region = bestOverlap(
      ., 
      lapply(gene_regions, select, region) %>% 
        GRangesList() %>% 
        unlist(),
      var = "region"
    ) %>% 
      factor(levels = regions)
  ) %>% 
  select(-keep, -peak_position)
if (has_features) {
  consensus_peaks$Feature <- bestOverlap(
    consensus_peaks, external_features, 
    var = "feature", missing = "no_feature"
  ) %>% 
    factor(levels = names(colours$features)) %>% 
    fct_relabel(str_sep_to_title)
}
if (any_detected) {
  consensus_peaks$detected <- bestOverlap(
    consensus_peaks, 
    gene_regions %>% 
      lapply(select, any_of("detected")) %>% 
      GRangesList() %>% 
      unlist(), 
    var = "detected"
  ) %>% 
    as.logical()
}
```

A set of treatment-specific `r target` *oracle peaks* was defined for each condition by comparing the peaks detected when merging all treatment-specific samples, against those detected within each replicate.
Replicates which failed the previous QC steps are omitted from this step.
Oracle peaks which overlapped a peak in more than `r percent(config$peaks$qc$min_prop_reps)` of the *individual replicates passing QC* in each treatment group, were retained.

```{r tab-oracle-peak}
oracle_peaks %>%
  lapply(mutate, w = width) %>% 
  lapply(as_tibble) %>% 
  bind_rows(.id = "treat") %>% 
  mutate(treat = factor(treat, levels = treat_levels)) %>% 
  group_by(treat) %>% 
  summarise(
    detected_peaks = dplyr::n(),
    retained = sum(keep),
    `% retained` = percent(retained / detected_peaks, 0.1),
    median_width = median(w)
  ) %>% 
  rename_all(str_sep_to_title) %>% 
  dplyr::rename(`% Retained` = ` Retained`) %>% 
  pander(
    justify = "lrrrr",
    caption = glue(
      "Oracle peaks detected by merging samples within each treatment group.",
      "Peaks were only retained if detected in at least",
      "{percent(config$peaks$qc$min_prop_reps)} of the retained samples for each", "treatment group, as described above.",
      .sep = " "
    )
  )
```

#### Peak Widths

```{r plot-peak-widths, fig.height=6, fig.cap = "Widths of all Oracle Peaks as a function of the number of replicates the peak was detected in. The y-axis is shown on the log scale."}
oracle_peaks %>%
  lapply(mutate, w = width) %>%
  lapply(as_tibble) %>%
  bind_rows(.id = "treat") %>%
  dplyr::filter(keep) %>%
  ggplot(aes(as.factor(n_reps), w, fill = treat)) +
  geom_boxplot(alpha = 0.8) +
  scale_y_log10() +
  scale_fill_manual(values = colours$treat[treat_levels]) +
  labs(x = "Replicates", y = "Peak Width", fill = "Treatment")
```

#### Distance to TSS

```{r plot-oracle-dist-to-tss-hist, fig.height=6, fig.cap = glue("*Distances from the centre of each {target} oracle peak to the nearest transcription start-site shown as A) a histogram, and B) as a cumulative distribution. The vertical dashed line indicates the range considered to be a promoter during annotation preparation. Between ", lapply(oracle_peaks, overlapsAny, tss) %>% vapply(mean, numeric(1)) %>% range(na.rm = TRUE) %>% percent(0.1) %>% glue_collapse(last = " and "), " of the oracle peaks directly overlapped a TSS.*")}
df <- oracle_peaks %>% 
  lapply(join_nearest, mutate(tss, tss = start)) %>% 
  lapply(mutate, d = (start + width/2 - tss)/1e3) %>%
  lapply(as_tibble) %>%
  bind_rows(.id = "treat") %>% 
  mutate(treat = factor(treat, treat_levels))
a <- df %>%
  ggplot(aes(d, colour = treat)) +
  geom_density() +
  scale_colour_manual(values = colours$treat[treat_levels]) +
  coord_cartesian(xlim = c(-100, 100)) +
  labs(
    x = "Distance to Nearest TSS (kb)",
    y = "Density",
    colour = "Treat"
  ) +
  guides(colour = "none")
b <- df %>% 
  select(treat, d) %>% 
  arrange(abs(d)) %>% 
  mutate(
    q = seq_along(d) / nrow(.)
  ) %>% 
  ggplot(aes(abs(d), q, colour = treat)) +
  geom_line() +
  geom_vline(
    xintercept = max(unlist(params$gene_regions$promoters)) / 1e3, 
    linetype = 2, colour = "grey40"
  ) +
  coord_cartesian(xlim = c(0, 100)) +
  labs(
    x = "Distance to Nearest TSS (kb)",
    y = "Quantile",
    colour = "Treat"
  ) +
  scale_y_continuous(labels = percent, breaks = seq(0, 1, by = 0.2)) +
  scale_x_continuous(breaks = seq(0, 100, by = 20)) +
  scale_colour_manual(values = colours$treat[treat_levels]) 
plot_grid(
  a, b + guides(colour = "none"), get_legend(b),
  rel_widths = c(0.5, 0.5, 0.1),  nrow = 1,
  labels = c("A", "B")
)
```


### Consensus Peaks {.tabset}

In addition to the oracle peaks, a set of `r comma(length(consensus_peaks))` treatment-agnostic `r target` **consensus peaks** were defined.
Consensus ranges were the *union* of all overlapping ranges defined and retained in one or more sets of oracle peaks.
Resulting values for the `score`, `signalValue`, `pValue` and `qValue` were calculated as the *maximum* across all treatments.
*Peak summits* were taken as the median position of all summits which comprise the consensus peak.
The number of replicates generating each peak was defined as the total across all treatment groups.

#### `r ifelse(length(treat_levels) <= 3, "Venn Diagram", "Upset Plot")`

```{r plot-common-peaks, results = 'hide', echo = length(treat_levels) <= 3}
fig_path <- here::here("docs", "assets", target)
if (!dir.exists(fig_path)) dir.create(fig_path, recursive = TRUE)
fig_name <- glue("{target}_common_peaks.png")
## An empty file to keep snakemake happy if > 3 treatments
file.create(file.path(fig_path, fig_name)) 
if (length(treat_levels) <= 3) {
  vd <- oracle_peaks %>%
    lapply(
      function(x) {
        names(
          subsetByOverlaps(
            consensus_peaks, 
            subset(x, keep)
          )
        )
      }
    ) %>%
    setNames(treat_levels) %>%
    venn.diagram(
      filename = file.path(fig_path, fig_name),
      imagetype = "png",
      units = "in",
      cat.cex = 1.4,
      height = 9, 
      width = 10,
      fill = colours$treat[treat_levels],
      alpha = 0.3
    )
  file.remove(list.files(fig_path, pattern = "log$", full.names = TRUE))
}
```

`r ifelse(length(treat_levels) <= 3, glue("![*Number of {target} consensus peaks which overlap oracle peaks defined within each condition.*](assets/{target}/{target}_common_peaks.png)"), "")`

```{r plot-common-upset, eval = length(treat_levels) > 3, echo = length(treat_levels) > 3, fig.height = 7, fig.cap = glue("*Number of {target} consensus peaks which overlap the oracle peaks defined within each condition.*")}
oracle_peaks %>%
  lapply(
    function(x) {
        names(
          subsetByOverlaps(
            consensus_peaks, 
            subset(x, keep)
          )
        )
    }
  ) %>%
  setNames(treat_levels) %>%
  fromList() %>%
  upset(
    sets = treat_levels,
    keep.order = TRUE,
    order.by = "freq",
    sets.bar.color = colours$treat[treat_levels],
    set_size.show = TRUE,
    set_size.scale_max = nrow(.) * 1.2,
    text.scale = 1.2
  )
```


#### Distance to TSS 

```{r plot-consensus-dist-to-tss-hist, fig.height=6, fig.cap = glue("*Distances from the centre of the {target} consensus peak to the transcription start-site shown as A) a histogram, and B) as a cumulative distribution. The vertical dashed line indicates the range considered to be a promoter during annotation preparation. {comma(sum(overlapsAny(consensus_peaks, tss)))} of the {comma(length(consensus_peaks))} {target} consensus peaks ({percent(mean(overlapsAny(consensus_peaks, tss)), 0.1)}) directly overlapped a TSS.*")}
a <- consensus_peaks %>% 
  join_nearest(
    mutate(tss, tss = start)
  ) %>% 
  mutate(d = start + width/2 - tss) %>% 
  mcols() %>% 
  as.data.frame() %>%
  ggplot(aes(d / 1e3)) +
  geom_density() +
  coord_cartesian(xlim = c(-100, 100)) +
  labs(
    x = "Distance to Nearest TSS (kb)",
    y = "Density"
  )
b <- consensus_peaks %>% 
  join_nearest(
    mutate(tss, tss = start)
  ) %>% 
  mutate(d = start + width/2 - tss) %>%
  as.data.frame() %>% 
  select(d) %>% 
  arrange(abs(d)) %>% 
  mutate(
    q = seq_along(d) / nrow(.)
  ) %>% 
  ggplot(aes(abs(d) / 1e3, q)) +
  geom_line() +
  geom_vline(
    xintercept = max(unlist(params$gene_regions$promoters)) / 1e3, 
    linetype = 2, colour = "grey40"
  ) +
  coord_cartesian(xlim = c(0, 100)) +
  labs(
    x = "Distance to Nearest TSS (kb)",
    y = "Quantile"
  ) +
  scale_y_continuous(labels = percent, breaks = seq(0, 1, by = 0.2)) +
  scale_x_continuous(breaks = seq(0, 100, by = 20))
plot_grid(a, b, labels = c("A", "B"))
```



#### Gene-Centric Regions

```{r plot-region-overlap, fig.height=8, fig.width=8, fig.cap = glue("*Proportions of {target} consensus peaks which overlap gene-centric features.*")}
consensus_peaks %>% 
  plotPie(fill = "region") +
  scale_fill_manual(
    values = colours$regions %>%
      setNames(regions[names(.)])
  ) +
  theme(legend.position = "none")
```

`r ifelse(has_features, "#### External Features", "")`

```{r plot-feature-overlap, eval = has_features, echo = has_features, fig.height=8, fig.width=8, fig.cap = glue("*The total number of {target} consensus peaks which overlap external features provided in {basename(config$external$features)}. If peaks map to multiple features, they are assigned to the one with the largest overlap.*")}
consensus_peaks %>% 
  plotPie(fill = "Feature", label_size = 4) +
  scale_fill_manual(
    values = colours$features %>% setNames(str_sep_to_title(names(.)))
  ) +
  theme(legend.position = "none")
```

`r ifelse(has_features, "#### External Features And Gene-Centric Regions", "")`

```{r plot-feature-region-overlap, eval = has_features, echo = has_features, fig.cap = glue("*The total number of {target} consensus peaks overlapping external features and gene-centric regions. If a peak overlaps multiple features or regions, it is assigned to be the one with the largest overlap.*")}
consensus_peaks %>% 
  plotPie(fill = "region", x = "Feature", label_size = 4, label_alpha = 0.7) +
  scale_fill_manual(
    values = colours$regions %>%
      setNames(regions[names(.)])
  ) +
  labs(fill = "Region") +
  theme(
    panel.grid = element_blank(),
    panel.border = element_blank(),
    axis.line.x = element_line(),
    legend.position = "top"
  )
```

`r ifelse(has_features & any_detected, "#### Detected Genes, External Features And Gene-Centric Regions", "")`

```{r plot-pie-feature-region-detected, eval = has_features & any_detected, echo = has_features & any_detected, fig.cap = glue("*The total number of {target} consensus peaks overlapping external features and gene-centric regions, separated by whether each peak is mapped to a detected gene. If a peak overlaps multiple features or regions, it is assigned to be the one with the largest overlap. In the absence of HiC an excess of Distal Intergenic regions would be expected to remain unmapped.*")}
df <- consensus_peaks %>% 
  as_tibble() %>% 
  mutate(
     region = fct_relabel(region, str_wrap, width = 20)
  ) %>% 
  group_by(region, Feature) %>% 
  summarise(
    total = dplyr::n(),
    detected = sum(detected),
    .groups = "drop"
  ) %>% 
  mutate(
    not_detected = total - detected,
    r = total / sum(total)
  ) %>% 
  pivot_longer(
    cols = ends_with("detected"), names_to = "status", values_to = "n"
  ) %>% 
  mutate(
    status = str_sep_to_title(status),
    x0 = as.integer(region),
    y0 = as.integer(Feature)
  ) 
df %>% 
  ggplot() +
  ggforce::stat_pie(
    aes(
      x0 = x0, y0 = y0, r0 = 0, r = 0.5 * r / max(r), fill = status, amount = n
    )
  ) +
  geom_label(
    aes(x0, y0, label = comma(total)),
    data = . %>% 
      distinct(region, Feature, .keep_all = TRUE) %>% 
      dplyr::filter(total > 0.05*sum(total)),
    alpha = 0.8,
    size = 4
  ) +
  scale_x_continuous(
    breaks = seq_along(levels(df$region)), labels = levels(df$region)
  ) +
  scale_y_continuous(
    breaks = seq_along(levels(df$Feature)), labels = levels(df$Feature)
  ) +
  scale_fill_manual(values = c(Detected = "red", "Not Detected" = "grey70") ) +
  labs(
    x = "Region", y = "External Feature",
    fill = "Mapped To Any\nDetected Genes"
  ) +
  coord_equal() +
  theme(
    panel.grid = element_blank(),
    axis.line.x = element_line(),
    legend.position = "top"
  )
```

## Highly Ranked Peaks {.tabset}

```{r grl-to-plot}
grl_to_plot <- vector("list", length(treat_levels) + 1) %>%
  setNames(c("consensus", treat_levels))
grl_to_plot$consensus <- consensus_peaks %>%
  as_tibble() %>%
  arrange(desc(score)) %>%
  dplyr::slice(1) %>%
  colToRanges("range", seqinfo = sq)
grl_to_plot[treat_levels] <- treat_levels %>% 
  lapply(
    function(x) {
      if (length(oracle_peaks[[x]]) == 0) return(NULL)
      tbl <- oracle_peaks[[x]] %>% 
        filter(keep) %>% 
        filter_by_non_overlaps(
          unlist(oracle_peaks[setdiff(treat_levels, x)])
        ) %>% 
        filter_by_non_overlaps(grl_to_plot$consensus) %>% 
        as_tibble() 
      if (nrow(tbl) == 0) return(NULL)
      tbl %>% 
        arrange(desc(score)) %>% 
        dplyr::slice(1) %>% 
        colToRanges("range", seqinfo = sq)
    }
  )
grl_to_plot <- grl_to_plot %>% 
  .[vapply(., length, integer(1)) > 0] %>% 
  lapply(setNames, NULL) %>% 
  GRangesList() %>% 
  unlist() %>% 
  distinctMC(.keep_all = TRUE) %>% 
  splitAsList(names(.)) %>% 
  endoapply(function(x) x[1,]) %>% 
  .[intersect(c("consensus", treat_levels), names(.))]
```

```{r prepare-plot-objects}
## The coverage
bwfl <- list2(
  "{target}" := file.path(macs2_path, glue("{treat_levels}_merged_treat_pileup.bw")) %>% 
  BigWigFileList() %>% 
  setNames(treat_levels)
)
## The features track
feat_gr <- gene_regions %>% 
  lapply(granges) %>% 
  GRangesList()
feature_colours <- colours$regions
if (has_features) {
  feat_gr <- feat_gr %>% 
    c(
      splitAsList(external_features, external_features$feature)
    )
  feature_colours <- c(feature_colours, colours$features)
}
## The genes track
hfgc_genes <- read_rds(
  here::here("output", "annotations", "trans_models.rds")
)
gene_col <- "grey"
if (any_detected){
  rna_lfc_col <- colnames(rnaseq)[str_detect(str_to_lower(colnames(rnaseq)), "logfc")][1]
  rna_fdr_col <- colnames(rnaseq)[str_detect(str_to_lower(colnames(rnaseq)), "fdr|adjp")][1]
  if (!is.na(rna_lfc_col) & !is.na(rna_fdr_col)) {
    hfgc_genes <- hfgc_genes %>%
      mutate(
        status = case_when(
          !gene %in% rnaseq$gene_id ~ "Undetected",
          gene %in% dplyr::filter(
            rnaseq, !!sym(rna_lfc_col) > 0, !!sym(rna_fdr_col) < fdr_alpha
          )$gene_id ~ "Up",
          gene %in% dplyr::filter(
            rnaseq, !!sym(rna_lfc_col) < 0, !!sym(rna_fdr_col) < fdr_alpha
          )$gene_id ~ "Down",
          gene %in% dplyr::filter(
            rnaseq, !!sym(rna_fdr_col) >= fdr_alpha
          )$gene_id ~ "Unchanged",
        )
      ) %>%
      splitAsList(.$status) %>%
      lapply(select, -status) %>%
      GRangesList()
    gene_col <- colours$direction %>%
      setNames(str_to_title(names(.)))
  }
}
## External Coverage (Optional)
if (!is.null(config$external$coverage)) {
  ext_cov_path <- config$external$coverage %>% 
    lapply(unlist) %>% 
    lapply(function(x) setNames(here::here(x), names(x)))
  bwfl <- c(
    bwfl[target],
    config$external$coverage %>% 
      lapply(
        function(x) {
          BigWigFileList(here::here(unlist(x))) %>% 
            setNames(names(x))
        }
      )
  )
}
line_col <- lapply(bwfl, function(x) colours$treat[names(x)])
y_lim <- bwfl %>%
  lapply(
    function(x) {
      lapply(x, import.bw, which = unlist(grl_to_plot)) %>%
        lapply(function(gr) c(0, max(gr$score))) %>%
        unlist() %>%
        range()
    }
  )
```

Coverage for a small set of highly ranked peaks are shown below.
These are the most highly ranked Consensus Peak (by `score`) and any oracle peaks which are unique to each treatment group, after excluding the most highly ranked consensus peak.


```{r plot-highly-ranked}
htmltools::tagList(
  mclapply(
    seq_along(grl_to_plot),
    function(x) {
      nm <- names(grl_to_plot)[[x]]
      ## Export the png
      png_out <- file.path(
        fig_path,
        nm %>% 
          str_replace_all(" ", "_") %>% 
          paste0("_topranked.png")
      )
      png(
        filename = png_out,
        width = knitr::opts_current$get("fig.width"), 
        height = knitr::opts_current$get("fig.height"), 
        units = "in", res = 300
      )
      ## Automatically collapse Transcripts if more than 10
      ct <- FALSE
      gh <- 1
      if (length(subsetByOverlaps(all_gr$transcript, grl_to_plot[[x]])) > 20) {
        ct <- "meta"
        gh <- 0.5
      }
      ## Generate the plot
      plotHFGC(
        grl_to_plot[[x]],
        features = feat_gr, featcol = feature_colours,
        genes = hfgc_genes, genecol = gene_col,
        coverage = bwfl, linecol = line_col,
        cytobands = bands_df,
        rotation.title = 90,
        zoom = 30,
        ylim = y_lim,
        collapseTranscripts = ct, genesize = gh,
        col.title = "black", background.title = "white", 
        showAxis = FALSE
      )
      dev.off()
      
      ## Define the caption
      gr <- join_nearest(grl_to_plot[[x]], all_gr$gene, distance = TRUE)
      d <- gr$distance
      gn <- gr$gene_name
      peak_desc <- ifelse(
        nm == "consensus",
        "consensus peak by combined score across all treatments.",
        paste(
          "oracle-peak unique to the merged", nm, "samples."
        )
      )
      cp <- htmltools::tags$em(
        glue(
          "The most highly ranked {peak_desc} ",
          ifelse(
            d == 0,
            paste('The peak directly overlaps', gn),
            paste0("The nearest gene was ", gn, ", ", round(d/1e3, 1), "kb away")
          ),
          ifelse(
            any_detected & gn %in% genes_gr$gene_name,
            glue(" which was detected in the RNA-Seq data."),
            glue(".")
          )
        )
      )
      
      ## Create html tags
      fig_link <- str_extract(png_out, "assets.+")
      htmltools::div(
        htmltools::div(
          id = nm %>% 
            str_replace_all(" ", "-") %>% 
            str_to_lower() %>% 
            paste0("-topranked"),
          class="section level3",
          htmltools::h3(
            ifelse(
              nm == "consensus",
              "Consensus Peaks",
              paste("Oracle Peaks:", nm)
            )
          ),
          htmltools::div(
            class = "figure", style = "text-align: center",
            htmltools::img(src = fig_link, width = 960),
            htmltools::p(
              class = "caption", htmltools::tags$em(cp)
            )
          )
        )
      )
    },
    mc.cores = min(length(grl_to_plot), threads)
  )
)  
```



## Data Export


```{r export, results='hide'}
all_out <- list(
  consensus_peaks_bed = file.path(macs2_path, "consensus_peaks.bed"),
  oracle_peaks_rds = file.path(macs2_path, "oracle_peaks.rds"),
  renv = file.path(
    here::here("output/envs"), 
    glue("{target}_macs2_summary.RData")
  )
) %>% 
  c(
    sapply(
      names(oracle_peaks), 
      function(x) {
        file.path(macs2_path, glue("{x}_oracle_peaks.bed"))
      },
      simplify = FALSE
    ) %>% 
      setNames(
        glue("{names(.)}_oracle_peaks_bed")
      )
  )
export(consensus_peaks, all_out$consensus_peaks)
oracle_peaks %>%
  lapply(subset, keep) %>%
  lapply(select, -keep) %>%
  GRangesList() %>%
  write_rds(all_out$oracle_peaks, compress = "gz")
names(oracle_peaks) %>% 
  lapply(
    function(x) {
      id <- glue("{x}_oracle_peaks_bed")
      oracle_peaks[[x]] %>% 
        subset(keep) %>% 
        select(score, signalValue, pValue, qValue, peak) %>% 
        export(all_out[[id]], format = "narrowPeak")
    }
  )
if (!dir.exists(dirname(all_out$renv))) dir.create(dirname(all_out$renv))
save.image(all_out$renv)
```

During this workflow the following files were exported:

`r pander(lapply(all_out, str_extract, pattern = "output.+"))`

## References

<div id="refs"></div>

<br>
<button type="button" class="btn btn-default btn-sessioninfo" data-toggle="collapse" data-target="#sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Session information
</button>
</p>
<div id="sessioninfo" class="collapse">
```{r session-info, echo=FALSE}
pander::pander(sessionInfo())
```
</div>
