# Comparison With RNA Seq

```{r define-objects}
library(fgsea)
library(metap)
rna_status <- rnaseq %>% 
  mutate(
    de_status = case_when(
      !!sym(rna_fdr_col) < fdr_alpha & !!sym(rna_lfc_col) < 0 ~ "Down",
      !!sym(rna_fdr_col) < fdr_alpha & !!sym(rna_lfc_col) > 0 ~ "Up",
      !!sym(rna_fdr_col) >= fdr_alpha ~ "Unchanged"
    )
  ) %>% 
  dplyr::select(gene_id, gene_name, de_status) %>% 
  bind_rows(
    genes_gr %>% 
      subset(!gene_id %in% rnaseq$gene_id) %>% 
      select(gene_id, gene_name) %>% 
      mcols() %>% 
      as_tibble() %>% 
      mutate(de_status = "Undetected")
  ) %>% 
  mutate(
    de_status = factor(
      de_status, levels = str_to_title(names(colours$direction))
    )
  ) %>% 
  arrange(gene_id)
rna_p_col <- colnames(rnaseq)[
  str_detect(str_to_lower(colnames(rnaseq)), "pval|p\\.val")
][1]
stopifnot(length(rna_p_col) == 1)
diff_windows <- merged_results %>% 
  splitAsList(.$status) %>% 
  lapply(select, gene_id) %>% 
  lapply(mcols) %>% 
  lapply(as_tibble) %>% 
  lapply(pull, "gene_id") %>% 
  lapply(unlist) %>% 
  lapply(unique)
```

## Distribution Of Detected Genes {.tabset}

### Detected Genes By `r target` Binding Region

The distribution of merged windows, where `r target` was considered as present, was firstly compared to genes within the RNA-Seq dataset.
Genes were simply considered as detected if present in the RNA-Seq data, or undetected if not present.
Gene-centric regions were as defined previously.

```{r plot-region-detected, fig.height = 6, fig.cap = glue("*Distribution of {target}-bound windows by region, according to whether the window is mapped to a detected gene in the RNA-Seq dataset.*")}
w <- 0.8
min_p <- 0.025
merged_results %>%
  mutate(
    any_detected = ifelse(
      vapply(gene_name, length, integer(1)) > 1,
      "Mapped to Detected Genes",
      "Mapped to Undetected Genes"
    )
  ) %>% 
  plotPie(fill = "region", x = "any_detected", label_size = 4, width = w) +
  geom_text(
    aes(xlab, ylab, label =lab),
    data = . %>% 
      mutate(
        xlab = x + w*(r + 0.1)*sin(label_radians),
        ylab = 1 + w*(r + 0.1)*cos(label_radians),
        lab = glue("{region}\n{percent(p, 0.1)}") %>% 
          str_wrap(10)
      ) %>% 
      dplyr::filter(p * N > min_p * max(N)),
    size = 3
  ) +
  geom_segment(
    aes(x_inner, y_inner, xend = x_outer, yend = y_outer),
     data = . %>% 
      mutate(
        x_inner = x + w*r*sin(label_radians),
        y_inner = 1 + w*r*cos(label_radians),
        x_outer = x + w*(r + 0.01)*sin(label_radians),
        y_outer = 1 + w*(r + 0.01)*cos(label_radians),
      ) %>% 
      dplyr::filter(p * N > min_p * max(N))
  ) +
  scale_fill_manual(
    values = colours$regions %>% 
      setNames(regions[names(.)]) 
  ) +
  labs(x = "", fill = "Region") +
  theme(panel.grid = element_blank())
```

`r ifelse(has_features, "### Detected Genes By Feature", "")`

```{r plot-feature-detected, eval = has_features, echo = has_features, fig.height = 6, fig.cap = glue("*Distribution of {target}-bound windows by externally-defined features from {basename(config$external$features)}, according to whether the window is mapped to a detected gene in the RNA-Seq dataset.*")}
merged_results %>% 
  mutate(
    any_detected = ifelse(
      vapply(gene_name, length, integer(1)) > 1,
      "Mapped to Detected Genes",
      "Mapped to Undetected Genes"
    )
  ) %>% 
  plotPie(fill = "feature", x = "any_detected", label_size = 4, width = w) +
  geom_text(
    aes(xlab, ylab, label =lab),
    data = . %>% 
      mutate(
        xlab = x + w*(r + 0.1)*sin(label_radians),
        ylab = 1 + w*(r + 0.1)*cos(label_radians),
        lab = glue("{feature}\n{percent(p, 0.1)}") %>% 
          str_wrap(10)
      ) %>% 
      dplyr::filter(p * N > min_p * max(N)),
    size = 3
  ) +
  geom_segment(
    aes(x_inner, y_inner, xend = x_outer, yend = y_outer),
     data = . %>% 
      mutate(
        x_inner = x + w*r*sin(label_radians),
        y_inner = 1 + w*r*cos(label_radians),
        x_outer = x + w*(r + 0.01)*sin(label_radians),
        y_outer = 1 + w*(r + 0.01)*cos(label_radians),
      ) %>% 
      dplyr::filter(p * N > min_p * max(N))
  ) +
  scale_fill_manual(
    values = colours$features %>% 
      setNames(str_sep_to_title(names(.)))
  ) +
  labs(x = "", fill = "External Feature") +
  theme(panel.grid = element_blank())
```


## Relationship With Differentially Expressed Genes {.tabset}


```{r define-genesets}
rna_dir_ranks <- rnaseq %>% 
  mutate(
    ranking_stat = -sign(!!sym(rna_lfc_col))*log10(!!sym(rna_p_col)) %>% 
      setNames(gene_id)
  ) %>% 
  arrange(ranking_stat) %>% 
  dplyr::filter(!!sym(rna_p_col) < 1) %>% 
  pull("ranking_stat") 
rna_sig_ranks <- rna_dir_ranks %>% 
  abs() %>% 
  sort()
dir_gs <- merged_results %>% 
  select(status, starts_with("gene")) %>% 
  as_tibble() %>% 
  mutate(group = glue("{target} {status}")) %>% 
  split(.$group)%>% 
  lapply(pull, "gene_id") %>% 
  lapply(unlist) %>% 
  lapply(unique) %>% 
  .[vapply(., length, integer(1)) >= min_gs_size]
reg_dir_gs <- merged_results %>% 
  select(
    overlaps_ref, status, any_of(c("region", "feature")), starts_with("gene")
  ) %>% 
  as_tibble() %>% 
  mutate(group = glue("{region} / {target} {status}")) %>% 
  split(.$group)%>% 
  lapply(pull, "gene_id") %>% 
  lapply(unlist) %>% 
  lapply(unique) %>% 
  .[vapply(., length, integer(1)) >= min_gs_size]
if (has_features) {
  feat_dir_gs <- merged_results %>% 
  select(
    overlaps_ref, status, any_of(c("feature")), starts_with("gene")
  ) %>% 
  as_tibble() %>% 
  mutate(group = glue("{feature} / {target} {status}")) %>% 
  split(.$group)%>% 
  lapply(pull, "gene_id") %>% 
  lapply(unlist) %>% 
  lapply(unique) %>% 
  .[vapply(., length, integer(1)) >= min_gs_size]
}
```

### DE Genes Associated with Changed `r target` Binding

```{r any_diff_sites}
de_genes_db_regions <- tibble()
any_diff_sites <- sum(mcols(merged_results)[[fdr_column]] < fdr_alpha) > 0
any_de <- any(rnaseq$DE)
para <- ""
if (any_de & any_diff_sites) {
  ft_data <- rnaseq %>%
    dplyr::filter(gene_id %in% unlist(diff_windows)) %>% 
    mutate(
      window = case_when(
        gene_id %in% unlist(diff_windows[c("Up", "Down")]) ~ "Diff",
        TRUE ~ "Unchanged"
      ) %>%
        fct_infreq()
    ) %>% 
    group_by(DE, window) %>% 
    tally() %>% 
    pivot_wider(
      names_from = "DE", values_from = "n", 
      names_prefix = "DE_", values_fill = 0
    ) %>% 
    as.data.frame() %>% 
    column_to_rownames("window") %>% 
    as.matrix()
  ft <- fisher.test(ft_data)
  para <- glue(
    "
    Using the {comma(length(unique(unlist(diff_windows))))} genes mapped to a 
    {target} binding site, genes were tested for any association between sites 
    showing evidence of diferential binding and differential expression.
    {comma(length(unique(unlist(diff_windows[c('Up', 'Down')]))))} genes were 
    mapped to at least one site showing differential {target} binding with 
    {comma(ft_data['Diff', 'DE_TRUE'])} 
    ({percent(ft_data['Diff', 'DE_TRUE'] / sum(ft_data['Diff',]))}) of these
    being considered differentially expressed. This compares to 
    {percent(ft_data['Unchanged', 'DE_TRUE'] / sum(ft_data['Unchanged', ]), 0.1)}
    of genes where no differential {target} binding was detected. Fisher's 
    Exact test for any association gave this a p-value of 
    {ifelse(ft$p.value < 0.01, sprintf('%.2e', ft$p.value), round(ft$p.value, 3))}.
    "
  )
}
```

`r if (!any_diff_sites) "No differentially bound sites were detected."`

`r if (any_de & any_diff_sites) para`

```{r de-genes-db-regions, eval = any_diff_sites, echo = any_diff_sites}
cmn_genes <- intersect(
  rnaseq %>% dplyr::filter(!!sym(rna_fdr_col) < fdr_alpha) %>% pull("gene_id"),
  merged_results %>% 
    filter(!!sym(fdr_column) < fdr_alpha) %>% 
    as_tibble() %>% 
    pull("gene_id") %>% 
    unlist()
)
de_genes_db_regions <- merged_results %>% 
  filter(!!sym(fdr_column) < fdr_alpha) %>% 
  filter(vapply(gene_id, function(x) any(x %in% cmn_genes), logical(1))) %>% 
  as_tibble() %>% 
  dplyr::select(
    range, ChIP_logCPM = AveExpr, ChIP_logFC = logFC, ChIP_FDR := !!sym(fdr_column), overlaps_ref,
    status, any_of(c("region", "feature")), gene_id
  ) %>% 
  unnest(gene_id) %>% 
  dplyr::filter(gene_id %in% cmn_genes) %>% 
  left_join(
    rnaseq %>% 
      dplyr::filter(gene_id %in% cmn_genes) %>% 
      dplyr::select(
        all_of(c("gene_id", "gene_name")), 
        RNA_logFC := !!sym(rna_lfc_col), RNA_FDR := !!sym(rna_fdr_col)
      ),
    by = "gene_id"
  ) %>% 
  arrange(RNA_FDR) %>% 
  # dplyr::slice(1:20) %>%
  mutate(
    dist2Gene = mapply(
      function(x, y) {
        distance(GRanges(x), subset(genes_gr, gene_id == y)) 
      },
      x = range,
      y = gene_id
    )
  ) %>% 
  mutate(
    region = case_when(
      dist2Gene > params$gene_regions$upstream & str_detect(region, "Promoter|Exon|Intron") ~ paste(
        str_extract(region, ".*(Promoter|Exon|Intron)"), "(Other Gene)"),
      TRUE ~ as.character(region)
    ),
    region = factor(
      region, 
      levels = c(
        levels(merged_results$region), 
        paste(
          str_extract(levels(merged_results$region), ".*(Promoter|Exon|Intron)"), 
          "(Other Gene)"
        )
      ) %>% 
        unique()
    )
  ) %>% 
  dplyr::select(
    gene_name, gene_id, starts_with("RNA"), range, dist2Gene, 
    starts_with("ChIP"), everything()
  ) %>% 
  droplevels()
```

```{r tab-de-genes-db-regions, eval = nrow(de_genes_db_regions) > 0, echo = nrow(de_genes_db_regions) > 0}
range_js <- JS(
  "function(values) {
    var min_val = Math.round(100 * Math.min(...values)) / 100
    var max_val = Math.round(100 * Math.max(...values)) / 100
    if (min_val == max_val) {
      var ret_val = min_val.toString()
    } else {
      var ret_val = '[' + min_val.toString() + ', ' + max_val.toString() + ']'
    }
    return ret_val
  }"
)
max_signal <- max(merged_results$AveExpr)
feat_def <- NULL
if (has_features) feat_def = list(
  feature = colDef (
    name = "Feature",
    aggregate = "unique"
  )
)
cp <-  htmltools::em(
  glue(
    "
    Differentially Expressed (DE) genes were checked for {target} binding
    regions which showed evidence of changed binding. 
    {comma(length(unique(de_genes_db_regions$range)), 1)} unique regions were 
    mapped to {comma(length(unique(de_genes_db_regions$gene_id)), 1)} DE genes. 
    Data are presented in a gene-centric manner and ranges may map to multiple
    genes. The full set of ranges considered to be showing evidence of altered 
    {target} binding are shown for each gene. The summarised range containing 
    all mapped ranges is given in the collapsed row, along with the range of 
    distances to gene and the range of signal levels. In the collapsed row, the 
    value for logFC represents the average across all mapped regions, with the 
    maximum FDR also shown. If {target} binding occurs within another gene this 
    is indicated in the 'Region' column.
    All sites and genes were exported as {basename(all_out$de_genes)}.
    "
  )
)
fs <- 12
tbl <- de_genes_db_regions %>% 
  reactable(
    groupBy = "gene_name",
    filterable = TRUE,
    borderless = TRUE,
    showPageSizeOptions = TRUE,
    columns = c(
      list(
        gene_name = colDef(
          name = "Gene", maxWidth = fs*11,
          style = list(
            borderRight = "1px solid rgba(0, 0, 0, 0.1)",
            fontSize = fs
          )
        ),
        gene_id = colDef(
          name = "ID", aggregate = "unique", cell = function(value) "",
          show = FALSE
        ),
        RNA_logFC = colDef(
          name = "logFC",
          aggregate = "mean", 
          style = JS(
            glue(
            "function(rowInfo) {
              var value = rowInfo.row['RNA_logFC']
              if (value < 0) {
                var color = '{{direction_colours[['Down']]}}'
              } else if (value > 0) {
                var color = '{{direction_colours[['Up']]}}'
              } else {
                var color = '#000000'
              }
              return { color: color, fontSize: {{fs}} }
            }",
            .open = "{{", .close = "}}"
            )
          ),
          format = colFormat(digits= 2),
          maxWidth = fs*6,
          cell = function(value) ""
        ),
        RNA_FDR = colDef(
          name = "FDR",
          aggregate = "min",
          format = colFormat(digits = 4),
          maxWidth = fs*6,
          cell = function(value) "",
          style = list(
            borderRight = "1px solid rgba(0, 0, 0, 0.1)", 
            fontSize = fs
          )
        ),
        range = colDef(
          name = "Range",
          minWidth = 160,
          aggregate = JS(
            "function(values){
              var chrom = [];
              var rng = [];
              var start = [];
              var end = [];
              for (i = 0; i < values.length; i++) {
                chrom[i] = values[i].split(':')[0];
                rng[i] = values[i].split(':')[1];
                start[i] = rng[i].split('-')[0];
                end[i] = rng[i].split('-')[1];
              }
              
              min_start = Math.min(...start);
              max_end = Math.max(...end);
              var ret_val = [...new Set(chrom)] + ':' + min_start.toString() + '-' + max_end.toString();
              
              return ret_val
              
            }"
          )
        ),
        dist2Gene = colDef(
          name = "Distance to Gene",
          aggregate = range_js,
          format = colFormat(digits = 0, separators = TRUE),
          maxWidth = fs*8,
          style = list(
            borderRight = "1px solid rgba(0, 0, 0, 0.1)",
            fontSize = fs
          ),
        ),
        ChIP_logCPM = colDef(
          name = "logCPM",
          maxWidth = fs*6,
          aggregate = range_js,
          format = colFormat(digits = 2),
          style = function(value) bar_style(
            width = value / max_signal, fontSize = fs
          )
        ),
        ChIP_logFC = colDef(
          name = "logFC",
          aggregate = "mean",
          maxWidth = fs*5,
          format = colFormat(digits = 2),
          cell = function(value) sprintf("%.2f", value),
          style =  JS(
            glue(
            "function(rowInfo) {
              var value = rowInfo.row['ChIP_logFC']
              if (value < 0) {
                var color = '{{direction_colours[['Down']]}}'
              } else if (value > 0) {
                var color = '{{direction_colours[['Up']]}}'
              } else {
                var color = '#000000'
              }
              return { color: color, fontSize: {{fs}} }
            }",
            .open = "{{", .close = "}}"
            )
          )
        ),
        ChIP_FDR = colDef(
          name = "FDR",
          aggregate = "max",
          maxWidth = 10 + fs*5,
          format = colFormat(prefix = "< ", digits = 4),
          cell = function(value) ifelse(
            value < 0.001,
            sprintf("%.2e", value),
            sprintf("%.3f", value)
          )
        ),
        overlaps_ref = colDef(
          name = "Macs Peak",
          show = FALSE,
          aggregate = "frequency",
          cell = function(value) ifelse(value, "\u2714 Yes", "\u2716 No"),
          style = function(value) {
            color <- ifelse(value, "#008000", "#e00000")
            list(color = color, fontSize = fs)
          },
          maxWidth = fs*5
        ),
        status = colDef(
          name = "Status",
          aggregate = "frequency",
          maxWidth = fs*6,
          cell = function(value) {
            html_symbol <- ""
            if (str_detect(value, "Up")) html_symbol <- "\u21E7"
            if (str_detect(value, "Down")) html_symbol <- "\u21E9"
            paste(html_symbol, value)
          },
          style = JS(
            glue(
            "function(rowInfo) {
              var value = rowInfo.row['ChIP_logFC']
              if (value < 0) {
                var color = '{{direction_colours[['Down']]}}'
              } else if (value > 0) {
                var color = '{{direction_colours[['Up']]}}'
              } else {
                var color = '#000000'
              }
              return { color: color, fontSize: {{fs}} }
            }",
            .open = "{{", .close = "}}"
            )
          )
        ),
        region = colDef(
          name = "Binding Region",
          aggregate = "unique"
        )
      ),
      feat_def
    ),
    columnGroups = list(
      colGroup(name = "RNA-Seq", columns = c("RNA_logFC", "RNA_FDR")),
      colGroup(
        name = paste(target, "ChIP-Seq"), 
        columns = c("ChIP_logCPM", "ChIP_logFC", "ChIP_FDR", "status")
      )
    ),
    theme = reactableTheme(style = list(fontSize = fs))
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```


### Volcano Plots By `r target` Binding

```{r plot-volcano-by-target, fig.cap = glue("*Volcano plot showing gene expression patterns separated by {target} status. The two most highly ranked genes for up and down regulation are labelled in each panel. Given that some genes may be mapped to multiple {target} binding events which include different binding patterns, genes may appear in multiple panels.*")}
diff_tbl <- diff_windows %>% 
  lapply(function(x) tibble(gene_id = x)) %>% 
  bind_rows(.id = "status")
rnaseq %>% 
  left_join(diff_tbl) %>% 
  mutate(
    status = str_replace_na(status, "Undetected"),
    status = glue("{target} {status}") %>%
      factor(levels = paste(target, c("Up", "Down", "Ambiguous", "Unchanged", "Undetected")))
  ) %>% 
  left_join(rna_status) %>% 
  droplevels() %>% 
  ggplot(
    aes(!!sym(rna_lfc_col), -log10(!!sym(rna_p_col)), colour = de_status)
  ) +
  geom_point() +
  geom_text_repel(
    aes(label = gene_name),
    data = . %>% 
      dplyr::arrange(!!sym(rna_p_col)) %>% 
      dplyr::filter(!!sym(rna_lfc_col) > 0, DE) %>% 
      split(.$status) %>% 
      lapply(dplyr::slice, 1:2) %>% 
      bind_rows(),
    show.legend = FALSE
  ) +
  geom_text_repel(
    aes(label = gene_name),
    data = . %>% 
      dplyr::arrange(!!sym(rna_p_col)) %>% 
      dplyr::filter(!!sym(rna_lfc_col) < 0, DE) %>% 
      split(.$status) %>% 
      lapply(dplyr::slice, 1:2) %>% 
      bind_rows(),
    show.legend = FALSE
  ) +
  geom_text(
    aes(label = lab),
    data = . %>% 
      group_by(status) %>% 
      summarise(n = dplyr::n()) %>% 
      mutate(lab = glue("n = {comma(n)}")),
    x = min(rnaseq[[rna_lfc_col]]) + 0.9 * diff(range(rnaseq[[rna_lfc_col]])),
    y = 1,
    colour = "black",
    inherit.aes = FALSE,
    show.legend = FALSE
  ) +
  facet_wrap(~status, ncol = 2) +
  scale_colour_manual(
    values = direction_colours
  ) +
  labs(colour = "DE Status") 
```


### P-Values By `r target` Status

```{r plot-p-target, fig.height = 5, fig.cap = glue("*P-Values for differential expression partitioned by {target} ChIP peak status. Genes were considered as being mapped to a differentially bound peak if one or more peaks was considered as differentially bound. No IHW procedure was undertaken using this data.*")}
p <- rnaseq %>% 
  mutate(
    mapped = case_when(
      gene_id %in% unlist(diff_windows[names(diff_windows) %in% c("Up", "Down")]) ~ paste(target, "Differentially Bound"),
      gene_id %in% unlist(diff_windows[names(diff_windows) %in% c("Unchanged")]) ~ paste(target, "Unchanged"),
      TRUE ~ paste("No", target, "Detected")
    )
  ) %>% 
  ggplot(
    aes(!!sym(rna_p_col), stat(density))
  ) +
  geom_histogram(
    colour = "black", fill = "grey", bins = 100
  ) +
  facet_wrap(~mapped, nrow = 1) +
  scale_y_continuous(expand = expansion(c(0, 0.05))) +
  labs(y = "Density", x = "P (Differential Expression)")
y <- 0.95 * max(layer_scales(p)$y$range$range)
p  +
  geom_text(
    aes(0.8, y, label = lab),
    data = . %>% 
      group_by(mapped) %>% 
      tally() %>% 
      mutate(lab = glue("N = {comma(n, 1)}"))
  ) 
```



`r if (nrow(de_genes_db_regions) > 0) "### ChIP logFC Vs RNA logFC"`

```{r plot-chip-rna-logfc, eval = nrow(de_genes_db_regions) > 0, echo = nrow(de_genes_db_regions) > 0,  fig.cap = "*logFC values for differentially bound ChIP-Seq peaks mapped to DE genes. The most highly ranked peak for each quandrant is labelled at showing both the gene and centre of the binding region. Whilst binding sites may be mapped to multiple genes, only the most highly ranked binding site is shown mapped to the mosthighly ranked DE genes. Points are coloured by differential expression status.*"}
de_genes_db_regions %>% 
  dplyr::filter(ChIP_FDR < fdr_alpha) %>% 
  group_by(gene_name) %>%
  dplyr::filter(ChIP_FDR == min(ChIP_FDR)) %>%
  mutate(DE = ifelse(RNA_logFC > 0, "Up", "Down")) %>% 
  ungroup() %>% 
  ggplot(
    aes(ChIP_logFC, RNA_logFC, colour = DE)
  ) +
  geom_point() +
  geom_label_repel(
    aes(label = lab),
    data = . %>% 
      mutate(grp = paste(DE, status)) %>% 
      # arrange(desc(abs(ChIP_logFC))) %>% 
      arrange(ChIP_FDR, RNA_FDR) %>% 
      split(.$grp) %>% 
      lapply(dplyr::slice, 1) %>% 
      bind_rows() %>% 
      mutate(
        range = GRanges(range) %>% 
          resize(width = 1, fix = "center") %>% 
          as.character(),
        lab = paste(gene_name, "\n", range)
      ),
    show.legend = FALSE,
    alpha = 0.7
  ) +
  geom_text(
    aes(x, y, label = lab),
    data = . %>% 
      group_by(DE, status) %>% 
      summarise(
        n = dplyr::n(),
        lab = glue("n = {comma(n)}"),
        x = max(abs(ChIP_logFC)) * sign(min(ChIP_logFC)) * 1.05,
        y = max(abs(RNA_logFC)) * sign(min(RNA_logFC)) * 1.05,
        .groups = "drop"
      ) %>% 
      group_by(DE) %>% 
      mutate(y = max(abs(y)) * sign(y)) %>% 
      ungroup() %>% 
      group_by(status) %>% 
      mutate(x = max(abs(x)) * sign(x)),
    colour = "black"
  ) +
  geom_hline(yintercept = 0, colour = "grey20") +
  geom_vline(xintercept = 0, colour = "grey20") +
  scale_colour_manual(values = direction_colours[c("Up", "Down")]) +
  labs(
    x = glue("logFC ({target} ChIP-Seq)"),
    y = "logFC (RNA-Seq)",
    colour = "Differential\nExpression"
  )
```


## Binding Pattern GSEA 

```{r dev-to-png, echo = FALSE}
knitr::opts_chunk$set(dev = 'png')
```

### GSEA By Direction of Regulation {.tabset}

GSEA analysis was first performed by taking the genes mapped to various sets of peaks, and checking for enrichment amongst the RNA-Seq results, ranking genes from up- to down-regulated, by significance of the logFC estimates.

#### All Windows

```{r gsea-dir-sig}
gsea_dir <- fgsea(dir_gs, rna_dir_ranks)
gsea_dir_sig <- gsea_dir %>% 
  arrange(pval) %>% 
  mutate(padj = p.adjust(pval, adj_method)) %>% 
  dplyr::filter(padj < enrich_alpha, size >= min_sig) %>% 
  as_tibble() 
p <- gsea_dir_sig %>% 
  dplyr::slice(1:9) %>% 
  pull("pathway") %>% 
  lapply(
    function(x) {
      plotEnrichment(dir_gs[[x]], rna_dir_ranks) +
        ggtitle(x) +
        theme(
          plot.title = element_text(hjust = 0.5, size = 10)
        )
    }
  )
```

`r if (nrow(gsea_dir_sig) == 0) glue("No association was found between {target} binding and *directional* significance in differentially expressed genes.")`

```{r tbl-gsea-dir-sig, eval = nrow(gsea_dir_sig) > 0, echo = nrow(gsea_dir_sig) > 0}
cp <-  htmltools::em(
  glue(
    "Merged windows were mapped to genes and their position amongst the ",
    "RNA-Seq results was assessed. Windows were classified based on  ",
    "the direction of {target} binding with ",
    "{treat_levels[[2]]} treatment. {nrow(gsea_dir_sig)} sets of windows were ",
    "associated with changes in gene expression, using the sign of ", 
    "fold-change and ranking statistic to initially rank the ", 
    "{comma(nrow(rnaseq), 1)} genes considered as detected."
  )
)
tbl <- gsea_dir_sig %>%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %>% 
      vapply(paste, character(1), collapse = "; "),
    Direction = ifelse(NES > 0, "\u21E7 Up-regulated", "\u21E9 Down-regulated"),
    "{target}" := str_extract(pathway, "(Up|Down|Unchanged)$"),
    pathway = str_remove_all(pathway, " / [^/]+$")
  ) %>% 
  dplyr::select(
    !!sym(target), Windows = size, Direction, 
    p = pval, padj, `Edge Size`, `Leading Edge` = leadingEdge
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list2(
      "{target}" := colDef(
        maxWidth = 90,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      Windows = colDef(maxWidth = 80, format = colFormat(separators = TRUE)),
      Direction = colDef(
        name = "Gene Direction",
        maxWidth = 150,
        style = function(value) {
          colour <- ifelse(
            str_detect(value, "Up"), 
            colours$direction[["up"]], 
            colours$direction[["down"]]
          )
          list(color = colour)
        },
      ),
      p = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 80
      ),
      padj = colDef(
        name = glue("p<sub>{adj_method}</sub>"), html = TRUE,
        cell = function(value) {
          ifelse(
            value < 0.001, sprintf("%.2e", value), sprintf("%.3f", value)
          )
        }
      ),
      "Edge Size" = colDef(maxWidth = 80),
     "Leading Edge" = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```

```{r, echo=FALSE}
## Set the default device to be png for barcode plots as they can become
## very large for some datasets
knitr::opts_chunk$set(dev = 'png')
```


```{r barcode-gsea-dir-sig, eval = nrow(gsea_dir_sig) > 0, echo = nrow(gsea_dir_sig) > 0, fig.height= max(3, 3 * ceiling( length(p) / 3)), fig.cap = glue("*Barcode plots for the top {length(p)} sets of windows associated with __directional__ changes in gene expression.*")}
plot_grid(plotlist = p, nrow = ceiling(length(p) / 3))
```

#### Windows By Region

```{r gsea-reg-dir-sig}
gsea_reg_dir <- fgsea(reg_dir_gs, rna_dir_ranks)
gsea_reg_dir_sig <- gsea_reg_dir %>% 
  arrange(pval) %>% 
  mutate(padj = p.adjust(pval, adj_method)) %>% 
  dplyr::filter(padj < enrich_alpha, size >= min_sig) %>% 
  as_tibble() 
p <- gsea_reg_dir_sig %>% 
  dplyr::slice(1:9) %>% 
  pull("pathway") %>% 
  lapply(
    function(x) {
      plotEnrichment(reg_dir_gs[[x]], rna_dir_ranks) +
        ggtitle(x) +
        theme(
          plot.title = element_text(hjust = 0.5, size = 10)
        )
    }
  )
```

`r if (nrow(gsea_reg_dir_sig) == 0) glue("No association was found between {target} binding and *directional* significance in differentially expressed genes, when incorporating genomic region.")`

```{r tbl-gsea-reg-dir_sig, eval = nrow(gsea_reg_dir_sig) > 0, echo = nrow(gsea_reg_dir_sig) > 0}
cp <-  htmltools::em(
  glue(
    "Merged windows were mapped to genes and their position amongst the ",
    "RNA-Seq results was assessed. Windows were classified based on which ",
    "region was the best overlap, and the direction of {target} binding with ",
    "{treat_levels[[2]]} treatment. {nrow(gsea_reg_dir_sig)} sets of windows were ",
    "associated with changes in gene expression, using the sign of ", 
    "fold-change and ranking statistic to initially rank the ", 
    "{comma(nrow(rnaseq), 1)} genes considered as detected."
  )
)
tbl <- gsea_reg_dir_sig %>%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %>% 
      vapply(paste, character(1), collapse = "; "),
    Direction = ifelse(NES > 0, "\u21E7 Up-regulated", "\u21E9 Down-regulated"),
    "{target}" := str_extract(pathway, "(Up|Down|Unchanged)$"),
    pathway = str_remove_all(pathway, " / [^/]+$")
  ) %>% 
  dplyr::select(
    pathway, !!sym(target), Windows = size, Direction, 
    p = pval, padj, `Edge Size`, `Leading Edge` = leadingEdge
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list2(
      pathway = colDef(
        minWidth = 100, name = "Region"
      ),
      "{target}" := colDef(
        maxWidth = 90,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      Windows = colDef(maxWidth = 80, format = colFormat(separators = TRUE)),
      Direction = colDef(
        name = "Gene Direction",
        maxWidth = 150,
        style = function(value) {
          colour <- ifelse(
            str_detect(value, "Up"), 
            colours$direction[["up"]], 
            colours$direction[["down"]]
          )
          list(color = colour)
        },
      ),
      p = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 80
      ),
      padj = colDef(
        name = glue("p<sub>{adj_method}</sub>"), html = TRUE,
        cell = function(value) {
          ifelse(
            value < 0.001, sprintf("%.2e", value), sprintf("%.3f", value)
          )
        }
      ),
      "Edge Size" = colDef(maxWidth = 80),
     "Leading Edge" = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```



```{r barcode-gsea-reg-dir-sig, eval = nrow(gsea_reg_dir_sig) > 0, echo = nrow(gsea_reg_dir_sig) > 0, fig.height= max(3, 3 * ceiling( length(p) / 3)), fig.cap = glue("*Barcode plots for the top {length(p)} sets of windows associated with __directional__ changes in gene expression.*")}
plot_grid(plotlist = p, nrow = ceiling(length(p) / 3))
```


`r if (has_features) "#### Windows By Feature"`



```{r gsea_feat_dir_sig_empty, eval = TRUE, echo = FALSE}
gsea_feat_dir_sig <- tibble()
```


```{r gsea-feat-dir-sig, eval = has_features, echo = has_features}
gsea_feat_dir <- fgsea(feat_dir_gs, rna_dir_ranks)
gsea_feat_dir_sig <- gsea_feat_dir %>% 
  arrange(pval) %>% 
  dplyr::filter(padj < enrich_alpha, size >= min_sig) %>% 
  as_tibble() 
p <- gsea_feat_dir_sig %>% 
  dplyr::slice(1:9) %>% 
  pull("pathway") %>% 
  lapply(
    function(x) {
      plotEnrichment(feat_dir_gs[[x]], rna_dir_ranks) +
        ggtitle(x) +
        theme(
          plot.title = element_text(hjust = 0.5, size = 10)
        )
    }
  )
```

`r if (has_features){ if (nrow(gsea_feat_dir_sig) == 0) glue("No association was found between {target} binding and *directional* significance in differentially expressed genes, when incorporating external features.")}`

```{r tbl-gsea-feat-dir-sig, eval = has_features & nrow(gsea_feat_dir_sig) > 0, echo = has_features & nrow(gsea_feat_dir_sig) > 0}
cp <-  htmltools::em(
  glue(
    "Merged windows were mapped to genes and their position amongst the ",
    "RNA-Seq results was assessed. Windows were classified based on which ",
    "feature was the best overlap, and the direction of {target} binding with ",
    "{treat_levels[[2]]} treatment. {nrow(gsea_feat_dir_sig)} sets of windows were ",
    "associated with changes in gene expression, using the sign of ", 
    "fold-change and ranking statistic to initially rank the ", 
    "{comma(nrow(rnaseq), 1)} genes considered as detected."
  )
)
tbl <- gsea_feat_dir_sig %>%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %>% 
      vapply(paste, character(1), collapse = "; "),
    Direction = ifelse(NES > 0, "\u21E7 Up-regulated", "\u21E9 Down-regulated"),
    "{target}" := str_extract(pathway, "(Up|Down|Unchanged)$"),
    pathway = str_remove_all(pathway, " / [^/]+$")
  ) %>% 
  dplyr::select(
    pathway, !!sym(target), Windows = size, Direction, 
    p = pval, FDR = padj, `Edge Size`, `Leading Edge` = leadingEdge
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list2(
      pathway = colDef(
        minWidth = 100, name = "Region"
      ),
      "{target}" := colDef(
        maxWidth = 120,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      Windows = colDef(maxWidth = 80, format = colFormat(separators = TRUE)),
      Direction = colDef(
        name = "Gene Direction",
        maxWidth = 120,
        style = function(value) {
          colour <- ifelse(
            str_detect(value, "Up"), 
            colours$direction[["up"]], 
            colours$direction[["down"]]
          )
          list(color = colour)
        },
      ),
      p = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 80
      ),
      FDR = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 80
      ),
      "Edge Size" = colDef(maxWidth = 80),
     "Leading Edge" = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```


```{r barcode-gsea-feat-dir-sig, eval = has_features & nrow(gsea_feat_dir_sig) > 0, echo = has_features & nrow(gsea_feat_dir_sig) > 0,fig.height= max(3, 3 * ceiling( length(p) / 3)), fig.cap = glue("*Barcode plots for the top {length(p)} sets of windows associated with __directional__ changes in gene expression.*")}
plot_grid(plotlist = p, nrow = ceiling(length(p) / 3))
```


### GSEA By Significance Only {.tabset}

GSEA analysis was then performed by taking the genes mapped to various sets of peaks, and checking for enrichment amongst the RNA-Seq results, ranking genes only from least to most significant, which effectively treats up- and down-regulation as indistinguishable.

#### All Windows

```{r gsea-nondir-sig}
gsea_nondir <- fgsea(dir_gs, rna_sig_ranks, scoreType = "pos")
gsea_nondir_sig <- gsea_nondir %>% 
  arrange(pval) %>% 
  dplyr::filter(padj < enrich_alpha, size >= min_sig) %>% 
  as_tibble() 
p <- gsea_nondir_sig %>% 
  dplyr::slice(1:9) %>% 
  pull("pathway") %>% 
  lapply(
    function(x) {
      plotEnrichment(dir_gs[[x]], rna_sig_ranks) +
        ggtitle(x) +
        theme(plot.title = element_text(hjust = 0.5, size = 10))
    }
  )
```

`r if (nrow(gsea_nondir_sig) == 0) glue("No association was found between {target} binding and *overall* significance in differentially expressed genes.")`

```{r tbl-gsea-nondir-sig, eval = nrow(gsea_nondir_sig) > 0, echo = nrow(gsea_nondir_sig) > 0}
cp <-  htmltools::em(
  glue(
    "Merged windows were mapped to genes and their position amongst the ",
    "RNA-Seq results was assessed. Windows were classified based on  ",
    "the direction of {target} binding with ",
    "{treat_levels[[2]]} treatment. {nrow(gsea_nondir_sig)} sets of windows were ",
    "associated with changes in gene expression, using only the p-value to ", 
    "rank the {comma(nrow(rnaseq), 1)} genes considered as detected."
  )
)
tbl <- gsea_nondir_sig %>%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %>% 
      vapply(paste, character(1), collapse = "; "),
    "{target}" := str_extract(pathway, "(Up|Down|Unchanged)$"),
    pathway = str_remove_all(pathway, " / [^/]+$")
  ) %>% 
  dplyr::select(
    !!sym(target), Windows = size, p = pval, FDR = padj, 
    `Edge Size`, `Leading Edge` = leadingEdge
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list2(
      "{target}" := colDef(
        maxWidth = 150,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
           colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      Windows = colDef(maxWidth = 80, format = colFormat(separators = TRUE)),
      p = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 80
      ),
      FDR = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 80
      ),
      "Edge Size" = colDef(maxWidth = 80),
     "Leading Edge" = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```

```{r barcode-gsea-nondir-sig, eval = nrow(gsea_nondir_sig) > 0, echo = nrow(gsea_nondir_sig) > 0, fig.height= max(3, 3 * ceiling( length(p) / 3)), fig.cap = glue("*Barcode plots for the top {length(p)} sets of windows associated with __non-directional__ changes in gene expression.*")}
plot_grid(plotlist = p, nrow = ceiling(length(p) / 3))
```


#### Windows By Region

```{r gsea-reg-nondir-sig}
gsea_reg_nondir <- fgsea(reg_dir_gs, rna_sig_ranks, scoreType = "pos")
gsea_reg_nondir_sig <- gsea_reg_nondir %>% 
  arrange(pval) %>% 
  dplyr::filter(padj < enrich_alpha, size >= min_sig) %>% 
  as_tibble() 
p <- gsea_reg_nondir_sig %>% 
  dplyr::slice(1:9) %>% 
  pull("pathway") %>% 
  lapply(
    function(x) {
      plotEnrichment(reg_dir_gs[[x]], rna_sig_ranks) +
        ggtitle(x) +
        theme(plot.title = element_text(hjust = 0.5, size = 10))
    }
  )
```

`r if (nrow(gsea_reg_nondir_sig) == 0) glue("No association was found between {target} binding and *overall* significance in differentially expressed genes.")`

```{r tbl-gsea-reg-nondir-sig, eval = nrow(gsea_reg_nondir_sig) > 0, echo = nrow(gsea_reg_nondir_sig) > 0}
cp <-  htmltools::em(
  glue(
    "Merged windows were mapped to genes and their position amongst the ",
    "RNA-Seq results was assessed. Windows were classified based on which ",
    "region was the best overlap, and the direction of {target} binding with ",
    "{treat_levels[[2]]} treatment. {nrow(gsea_reg_nondir_sig)} sets of windows were ",
    "associated with changes in gene expression, using only the p-value to ", 
    "rank the {comma(nrow(rnaseq), 1)} genes considered as detected."
  )
)
tbl <- gsea_reg_nondir_sig %>%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %>% 
      vapply(paste, character(1), collapse = "; "),
    "{target}" := str_extract(pathway, "(Up|Down|Unchanged)$"),
    pathway = str_remove_all(pathway, " / [^/]+$")
  ) %>% 
  dplyr::select(
    pathway, !!sym(target), Windows = size, p = pval, FDR = padj, 
    `Edge Size`, `Leading Edge` = leadingEdge
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list2(
      pathway = colDef(
        minWidth = 100, name = "Region"
      ),
      "{target}" := colDef(
        maxWidth = 90,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
           colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      Windows = colDef(maxWidth = 80, format = colFormat(separators = TRUE)),
      p = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 80
      ),
      FDR = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 80
      ),
      "Edge Size" = colDef(maxWidth = 80),
     "Leading Edge" = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```

```{r barcode-gsea-reg-nondir-sig, eval = nrow(gsea_reg_nondir_sig) > 0, echo = nrow(gsea_reg_nondir_sig) > 0, fig.height= max(3, 3 * ceiling( length(p) / 3)), fig.cap = glue("*Barcode plots for the top {length(p)} sets of windows associated with __non-directional__ changes in gene expression.*")}
plot_grid(plotlist = p, nrow = ceiling(length(p) / 3))
```


`r if (has_features) "#### Windows By Feature"`


```{r gsea_feat_nondir_sig_empty, eval = TRUE, echo = FALSE}
gsea_feat_nondir_sig <- tibble()
```

```{r gsea-feat-nondir-sig, eval = has_features, echo = has_features}
gsea_feat_nondir <- fgsea(feat_dir_gs, rna_sig_ranks, scoreType = "pos")
gsea_feat_nondir_sig <- gsea_feat_nondir %>% 
  arrange(pval) %>% 
  dplyr::filter(padj < enrich_alpha, size >= min_sig) %>% 
  as_tibble() 
p <- gsea_feat_nondir_sig %>% 
  dplyr::slice(1:9) %>% 
  pull("pathway") %>% 
  lapply(
    function(x) {
      plotEnrichment(feat_dir_gs[[x]], rna_sig_ranks) +
        ggtitle(x) +
        theme(
          plot.title = element_text(hjust = 0.5, size = 10)
        )
    }
  )
```

`r if (has_features){ if (nrow(gsea_feat_nondir_sig) == 0) glue("No association was found between {target} binding and *directional* significance in differentially expressed genes, when incorporating external features.")}`

```{r tbl-gsea-feat-nondir-sig, eval = has_features & nrow(gsea_feat_nondir_sig) > 0, echo = has_features & nrow(gsea_feat_nondir_sig) > 0}
cp <-  htmltools::em(
  glue(
    "Merged windows were mapped to genes and their position amongst the ",
    "RNA-Seq results was assessed. Windows were classified based on which ",
    "feature was the best overlap, and the direction of {target} binding with ",
    "{treat_levels[[2]]} treatment. {nrow(gsea_feat_nondir_sig)} sets of windows were ",
    "associated with overall changes in gene expression, using only the ", 
    "test statistic to rank the ", 
    "{comma(nrow(rnaseq), 1)} genes considered as detected."
  )
)
tbl <- gsea_feat_nondir_sig %>%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %>% 
      vapply(paste, character(1), collapse = "; "),
    Direction = ifelse(NES > 0, "\u21E7 Up-regulated", "\u21E9 Down-regulated"),
    "{target}" := str_extract(pathway, "(Up|Down|Unchanged)$"),
    pathway = str_remove_all(pathway, " / [^/]+$")
  ) %>% 
  dplyr::select(
    pathway, !!sym(target), Windows = size, Direction, 
    p = pval, FDR = padj, `Edge Size`, `Leading Edge` = leadingEdge
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list2(
      pathway = colDef(
        minWidth = 100, name = "Region"
      ),
      "{target}" := colDef(
        maxWidth = 90,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      Windows = colDef(maxWidth = 80, format = colFormat(separators = TRUE)),
      Direction = colDef(
        name = "Gene Direction",
        maxWidth = 120,
        style = function(value) {
          colour <- ifelse(
            str_detect(value, "Up"), 
            colours$direction[["up"]], 
            colours$direction[["down"]]
          )
          list(color = colour)
        },
      ),
      p = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 80
      ),
      FDR = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 80
      ),
      "Edge Size" = colDef(maxWidth = 80),
     "Leading Edge" = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```


```{r barcode-gsea-feat-nondir-sig, eval = nrow(gsea_feat_nondir_sig) > 0, echo = nrow(gsea_feat_nondir_sig) > 0, fig.height= max(3, 3 * ceiling( length(p) / 3)), fig.cap = glue("*Barcode plots for the top {length(p)} sets of windows associated with __nondirectional__ changes in gene expression.*")}
plot_grid(plotlist = p, nrow = ceiling(length(p) / 3))
```


```{r dev-to-fig-type, echo = FALSE}
knitr::opts_chunk$set(dev = fig_dev)
```


<!-- Could be done regardless of changes in expression -->

## Integration of RNA and ChIP Enrichment Analyses

```{r rnaseq-gsea}
rnaseq_gsea <- fgsea(
  pathways = gs_by_gsid,
  stats = rna_dir_ranks, 
  minSize = min_gs_size, 
  maxSize = max_gs_size
) %>% 
  dplyr::filter(!is.na(pval))
status_cols <- setNames(
  hcl.colors(4, "Viridis", rev = TRUE),
  c("Both", "ChIP Only", "RNA Only", "Neither")
)
```

The enrichment testing results from the `r target` differential binding analysis were then
compared to GSEA results obtained from the RNA-Seq data set.
Given the orthogonal nature of the two dataset, p-values from each analysis were combined using Wilkinson's maximum p-value method and the resulting p-values adjusted as previously.
Distances between nodes (gene-sets) for all network plots were determined by using ChIP target genes within the leading edge only.

### Differentially Bound Windows {.tabset}

```{r compare-alldiff-gsea}
cmn_diff <- rnaseq_gsea %>%
  dplyr::select(-log2err, -ES, -size) %>% 
  left_join(
    goseq_diff, by = c("pathway" = "gs_name"), suffix = c(".rna", ".chip")
  ) %>%
  dplyr::filter(!is.na(pval.rna), !is.na(pval.chip)) %>% 
  nest(p = starts_with("pval")) %>% 
  mutate(
    pval = vapply(p, function(x) metap::maximump(unlist(x))$p, numeric(1))
  ) %>%
  unnest(p) %>% 
  dplyr::rename(padj_chip = adj_p, padj_rna = padj) %>% 
  mutate(adj_p = p.adjust(pval, params$enrichment$adj)) %>% 
  arrange(pval) %>% 
  dplyr::filter(adj_p < enrich_alpha) %>% 
  mutate(
    leadingEdge = lapply(
      leadingEdge, intersect, dplyr::filter(mapped_ids, up | down)$gene_id
    ),
    numDEInCat = vapply(leadingEdge, length, integer(1)),
    Status = case_when(
      padj_rna < enrich_alpha & padj_chip < enrich_alpha ~ "Both",
      padj_rna < enrich_alpha & !padj_chip < enrich_alpha ~ "RNA Only",
      !padj_rna < enrich_alpha & padj_chip < enrich_alpha ~ "ChIP Only",
      !padj_rna < enrich_alpha & !padj_chip < enrich_alpha ~ "Neither"
    ) %>% 
      factor(levels = names(status_cols))
  ) %>% 
  dplyr::filter(numDEInCat >= min_sig) %>% 
  dplyr::select(
    gs_name = pathway, adj_p, Status, NES, starts_with("padj"),
    starts_with("num"), leadingEdge
  ) 
tg_diff <- cmn_diff %>% 
  make_tbl_graph(setNames(.$leadingEdge, .$gs_name))
plot_net <- length(tg_diff) >= min_network_size
txt <- ifelse(
  nrow(cmn_diff) == 0,
  glue("No gene sets were jointly enriched between all {target} differentially bound windows and the RNA-Seq data"),
  "#### Results Table"
)
```

`r txt`

```{r tbl-cmn-alldiff-gs, eval = nrow(cmn_diff) > 0, echo = nrow(cmn_diff) > 0}
cp <- htmltools::tags$em(
  glue(
    "
    Genes mapped to all windows showing evidence of differential binding were 
    tested for enrichment using goseq, and results were compared to GSEA results
    for the RNA-Seq dataset. Wilkinson's method was used to integrate the two
    p-values with a combined FDR-adjusted p-value < {enrich_alpha} being taken
    as evidence of potential regulatory targets from differentially bound 
    {target} binding sites. Of the {nrow(cmn_diff)} gene-sets of interest, 
    {sum(cmn_diff$Status == 'Both')} were individually enriched in both datasets,
    as indicated in the 'Status' column
    "
  )
)
tbl <- cmn_diff %>%
  mutate(
    leadingEdge = vapply(
      leadingEdge,
      function(x) paste(id2gene[x], collapse = "; "),
      character(1)
    )
  ) %>% 
  reactable(
    filterable = TRUE, showPageSizeOptions = TRUE,
    columns = list(
      gs_name = colDef(
        name = "Gene Set",
        cell = function(value) htmltools::tags$a(
          href = gs_url[[value]], 
          target = "_blank", 
          str_replace_all(value, "_", " ")
        ),
        minWidth = 120
      ),
      adj_p = colDef(
        name = glue("p<sub>{params$enrichment$adj}</sub>"), html = TRUE,
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ),
        maxWidth = 80
      ),
      Status = colDef(name = "Prior Status", maxWidth = 100),
      NES = colDef(
        format = colFormat(digits = 2),
        style = function(value) {
          col <- ifelse(value > 0, colours$direction["up"], colours$direction["down"])
          list(color = col)
        },
        maxWidth = 80
      ),
      padj_rna = colDef(
        name =  glue("p<sub>{params$enrichment$adj}</sub> (RNA)"), html = TRUE,
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ),
        maxWidth = 80
      ),
      padj_chip = colDef(
        name =  glue("p<sub>{params$enrichment$adj}</sub> (ChIP)"), html = TRUE,
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ),
        maxWidth = 80
      ),
      numDEInCat = colDef(name = "ChIP Targets in Leading Edge", maxWidth = 100),
      numInCat = colDef(name = "Mapped Genes", maxWidth = 100),
      leadingEdge = colDef(
        name = "Leading Edge ChIP Targets",
        cell = function(value) with_tooltip(value, width = 50),
        minWidth = 120
      )
    )
  )
div(
  class = "table", 
  div(
    class = "table-header",
    div(class = "caption", cp)
  ),
  tbl
)
```

`r if (plot_net) "#### Network Plot"`

```{r network-cmn-alldiff, eval = plot_net, fig.cap = glue("*Network plot for pathways considered significant when assessing all differentially bound {target} sites along with GSEA on the provided RNA-Seq data. Distances are derived from all detected genes within the pathway. Label colours indicate the direction of enrichment in GSEA results for genes in the pathway.*")}
tg_diff %>% 
  activate(nodes) %>% 
  mutate(direction = ifelse(NES > 0, "Up", "Down"))%>% 
  ggraph(layout = net_layout, weights = oc^2) +
  geom_edge_link(aes(width = oc^2, alpha = oc^2)) +
  geom_node_point(
    aes(fill = Status, size = numDEInCat),
    shape = 21
  ) +
  geom_node_text(
    aes(label = label, colour = direction),
    size = 3, 
    data = . %>%
      mutate(
        label = str_replace_all(label, "_", " ") %>% str_trunc(80) %>% str_wrap(width = 18)
      ),
    repel = TRUE, max.overlaps = max(10, round(length(tg_diff) / 4, 0)),
    bg.color = "white", bg.r = 0.1, 
  ) +
  scale_x_continuous(expand = expansion(c(0.1, 0.1))) +
  scale_y_continuous(expand = expansion(c(0.1, 0.1))) +
  scale_fill_manual(values = status_cols) +
  scale_size_continuous(range = c(1, 10)) +
  scale_edge_width(range = c(1, 6), limits = c(0, 1)) +
  scale_edge_alpha(range = c(0.1, 0.4), limits = c(0, 1))  +
  scale_colour_manual(
    values = setNames(direction_colours, str_to_title(names(direction_colours)))
  ) +
  guides(edge_alpha= "none") +
  labs(
    size = glue("{target} Targets In Leading Edge") %>% str_wrap(15),
    colour = "GSEA\nDirection",
    fill = "Prior\nSignificance",
    edge_width = expr(paste(OC^2))
  ) +
  theme_void() +
  theme(legend.text = element_text(hjust = 0.5))
```


### Windows With Increased `r target` {.tabset}

```{r compare-allup-gsea}
cmn_up <- rnaseq_gsea %>%
  dplyr::select(-log2err, -ES, -size) %>% 
  left_join(
    goseq_up, by = c("pathway" = "gs_name"), suffix = c(".rna", ".chip")
  ) %>%
  dplyr::filter(!is.na(pval.rna), !is.na(pval.chip)) %>% 
  nest(p = starts_with("pval")) %>% 
  mutate(
    pval = vapply(p, function(x) metap::maximump(unlist(x))$p, numeric(1))
    ) %>%
  unnest(p) %>% 
  dplyr::rename(padj_chip = adj_p, padj_rna = padj) %>% 
  mutate(adj_p = p.adjust(pval, params$enrichment$adj)) %>% 
  arrange(pval) %>% 
  dplyr::filter(adj_p < enrich_alpha) %>% 
  mutate(
    leadingEdge = lapply(
      leadingEdge, intersect, dplyr::filter(mapped_ids, up)$gene_id
    ),
    numDEInCat = vapply(leadingEdge, length, integer(1)),
    Status = case_when(
      padj_rna < enrich_alpha & padj_chip < enrich_alpha ~ "Both",
      padj_rna < enrich_alpha & !padj_chip < enrich_alpha ~ "RNA Only",
      !padj_rna < enrich_alpha & padj_chip < enrich_alpha ~ "ChIP Only",
      !padj_rna < enrich_alpha & !padj_chip < enrich_alpha ~ "Neither"
    ) %>% 
      factor(levels = names(status_cols))
  ) %>% 
  dplyr::filter(numDEInCat >= min_sig) %>% 
  dplyr::select(
    gs_name = pathway, adj_p, Status, NES, starts_with("padj"),
    starts_with("num"), leadingEdge
  ) 
tg_up <- cmn_up %>% 
  make_tbl_graph(
    setNames(.$leadingEdge, .$gs_name)
  )
plot_net <- length(tg_up) >= min_network_size
txt <- ifelse(
  nrow(cmn_up) == 0,
  glue("No gene sets were jointly enriched between all increasing {target} bound windows and the RNA-Seq data"),
  "#### Results Table"
)
```

`r txt`

```{r tbl-cmn-allup-gs, eval = nrow(cmn_up) > 0, echo = nrow(cmn_up) > 0}
cp <- htmltools::tags$em(
  glue(
    "
    Genes mapped to all windows showing evidence of increased {target} binding were 
    tested for enrichment using goseq, and results were compared to GSEA results
    for the RNA-Seq dataset. Wilkinson's method was used to integrate the two
    p-values with a combined FDR-adjusted p-value < {enrich_alpha} being taken
    as evidence of potential regulatory targets from increasingly bound 
    {target} binding sites. Of the {nrow(cmn_up)} gene-sets of interest, 
    {sum(cmn_up$Status == 'Both')} were individually enriched in both datasets,
    as indicated in the 'Status' column
    "
  )
)
tbl <- cmn_up %>%
  mutate(
    leadingEdge = vapply(
      leadingEdge,
      function(x) paste(id2gene[x], collapse = "; "),
      character(1)
    )
  ) %>% 
  reactable(
    filterable = TRUE, showPageSizeOptions = TRUE,
    columns = list(
      gs_name = colDef(
        name = "Gene Set",
        cell = function(value) htmltools::tags$a(
          href = gs_url[[value]], 
          target = "_blank", 
          str_replace_all(value, "_", " ")
        ),
        minWidth = 120
      ),
      adj_p = colDef(
        name = glue("p<sub>{params$enrichment$adj}</sub>"), html = TRUE,
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ),
        maxWidth = 80
      ),
      Status = colDef(name = "Prior Status", maxWidth = 100),
      NES = colDef(
        format = colFormat(digits = 2),
        style = function(value) {
          col <- ifelse(value > 0, colours$direction["up"], colours$direction["down"])
          list(color = col)
        },
        maxWidth = 80
      ),
      padj_rna = colDef(
        name =  glue("p<sub>{params$enrichment$adj}</sub> (RNA)"), html = TRUE,
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ),
        maxWidth = 80
      ),
      padj_chip = colDef(
        name =  glue("p<sub>{params$enrichment$adj}</sub> (ChIP)"), html = TRUE,
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ),
        maxWidth = 80
      ),
      numDEInCat = colDef(name = "ChIP Targets in Leading Edge", maxWidth = 100),
      numInCat = colDef(name = "Mapped Genes", maxWidth = 100),
      leadingEdge = colDef(
        name = "Leading Edge ChIP Targets",
        cell = function(value) with_tooltip(value, width = 50),
        minWidth = 120
      )
    )
  )
div(
  class = "table", 
  div(
    class = "table-header",
    div(class = "caption", cp)
  ),
  tbl
)
```

`r if (plot_net) "#### Network Plot"`

```{r network-cmn-allup, eval = plot_net, fig.cap = glue("*Network plot for pathways considered significant when assessing all increasingly bound {target} sites along with GSEA on the provided RNA-Seq data. Distances are derived from all detected genes within the pathway. Label colours indicate the direction of enrichment in GSEA results for genes in the pathway.*")}
tg_up %>% 
  activate(nodes) %>% 
  mutate(direction = ifelse(NES > 0, "Up", "Down")) %>% 
  ggraph(layout = net_layout, weights = oc^2) +
  geom_edge_link(aes(width = oc^2, alpha = oc^2)) +
  geom_node_point(
    aes(fill = Status, size = numDEInCat),
    shape = 21
  ) +
  geom_node_text(
    aes(label = label, colour = direction),
    size = 3, 
    data = . %>%
      mutate(
        label = str_replace_all(label, "_", " ") %>% str_trunc(80) %>% str_wrap(width = 18)
      ),
    repel = TRUE, max.overlaps = max(10, round(length(tg_up) / 4, 0)),
    bg.color = "white", bg.r = 0.1, 
  ) +
  scale_x_continuous(expand = expansion(c(0.1, 0.1))) +
  scale_y_continuous(expand = expansion(c(0.1, 0.1))) +
  scale_fill_manual(values = status_cols) +
  scale_size_continuous(range = c(1, 10)) +
  scale_edge_width(range = c(1, 6), limits = c(0, 1)) +
  scale_edge_alpha(range = c(0.1, 0.4), limits = c(0, 1))  +
  scale_colour_manual(
    values = setNames(direction_colours, str_to_title(names(direction_colours)))
  ) +
  guides(edge_alpha= "none") +
  labs(
    size = glue("{target} Targets In Leading Edge") %>% str_wrap(15),
    colour = "GSEA\nDirection",
    fill = "Prior\nSignificance",
    edge_width = expr(paste(OC^2))
  ) +
  theme_void() +
  theme(legend.text = element_text(hjust = 0.5))
```


### Windows With Decreased `r target` {.tabset}

```{r compare-alldown-gsea}
cmn_down <- rnaseq_gsea %>%
  dplyr::select(-log2err, -ES, -size) %>% 
  left_join(
    goseq_down, by = c("pathway" = "gs_name"), suffix = c(".rna", ".chip")
  ) %>%
  dplyr::filter(!is.na(pval.rna), !is.na(pval.chip)) %>% 
  nest(p = starts_with("pval")) %>% 
  mutate(
    pval = vapply(p, function(x) metap::maximump(unlist(x))$p, numeric(1))
    ) %>%
  dplyr::rename(padj_chip = adj_p, padj_rna = padj) %>% 
  mutate(adj_p = p.adjust(pval, params$enrichment$adj)) %>% 
  arrange(pval) %>% 
  dplyr::filter(adj_p < enrich_alpha) %>% 
  mutate(
    leadingEdge = lapply(
      leadingEdge, intersect, dplyr::filter(mapped_ids, down)$gene_id
    ),
    numDEInCat = vapply(leadingEdge, length, integer(1)),
    Status = case_when(
      padj_rna < enrich_alpha & padj_chip < enrich_alpha ~ "Both",
      padj_rna < enrich_alpha & !padj_chip < enrich_alpha ~ "RNA Only",
      !padj_rna < enrich_alpha & padj_chip < enrich_alpha ~ "ChIP Only",
      !padj_rna < enrich_alpha & !padj_chip < enrich_alpha ~ "Neither"
    ) %>% 
      factor(levels = names(status_cols))
  ) %>% 
  dplyr::filter(numDEInCat >= min_sig) %>% 
  dplyr::select(
    gs_name = pathway, adj_p, Status, NES, starts_with("padj"),
    starts_with("num"), leadingEdge
  ) 
tg_down <- cmn_down %>% 
  make_tbl_graph(
    setNames(.$leadingEdge, .$gs_name)
  )
plot_net <- length(tg_down) >= min_network_size
txt <- ifelse(
  nrow(cmn_down) == 0,
  glue("No gene sets were jointly enriched between all decreasing {target} bound windows and the RNA-Seq data"),
  "#### Results Table"
)
```

`r txt`

```{r tbl-cmn-alldown-gs, eval = nrow(cmn_down) > 0, echo = nrow(cmn_down) > 0}
cp <- htmltools::tags$em(
  glue(
    "
    Genes mapped to all windows showing evidence of decreased {target} binding were 
    tested for enrichment using goseq, and results were compared to GSEA results
    for the RNA-Seq dataset. Wilkinson's method was used to integrate the two
    p-values with a combined FDR-adjusted p-value < {enrich_alpha} being taken
    as evidence of potential regulatory targets from decreasingly bound 
    {target} binding sites. Of the {nrow(cmn_down)} gene-sets of interest, 
    {sum(cmn_down$Status == 'Both')} were individually enriched in both datasets,
    as indicated in the 'Status' column
    "
  )
)
tbl <- cmn_down %>%
  mutate(
    leadingEdge = vapply(
      leadingEdge,
      function(x) paste(id2gene[x], collapse = "; "),
      character(1)
    )
  ) %>% 
  reactable(
    filterable = TRUE, showPageSizeOptions = TRUE,
    columns = list(
      gs_name = colDef(
        name = "Gene Set",
        cell = function(value) htmltools::tags$a(
          href = gs_url[[value]], 
          target = "_blank", 
          str_replace_all(value, "_", " ")
        ),
        minWidth = 120
      ),
      adj_p = colDef(
        name = glue("p<sub>{params$enrichment$adj}</sub>"), html = TRUE,
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ),
        maxWidth = 80
      ),
      Status = colDef(name = "Prior Status", maxWidth = 100),
      NES = colDef(
        format = colFormat(digits = 2),
        style = function(value) {
          col <- ifelse(value > 0, colours$direction["up"], colours$direction["down"])
          list(color = col)
        },
        maxWidth = 80
      ),
      padj_rna = colDef(
        name =  glue("p<sub>{params$enrichment$adj}</sub> (RNA)"), html = TRUE,
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ),
        maxWidth = 80
      ),
      padj_chip = colDef(
        name =  glue("p<sub>{params$enrichment$adj}</sub> (ChIP)"), html = TRUE,
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ),
        maxWidth = 80
      ),
      numDEInCat = colDef(name = "ChIP Targets in Leading Edge", maxWidth = 100),
      numInCat = colDef(name = "Mapped Genes", maxWidth = 100),
      leadingEdge = colDef(
        name = "Leading Edge ChIP Targets",
        cell = function(value) with_tooltip(value, width = 50),
        minWidth = 120
      )
    )
  )
div(
  class = "table", 
  div(
    class = "table-header",
    div(class = "caption", cp)
  ),
  tbl
)
```

`r if (plot_net) "#### Network Plot"`

```{r network-cmn-alldown, eval = plot_net, fig.cap = glue("*Network plot for pathways considered significant when assessing all decreasingly bound {target} sites along with GSEA on the provided RNA-Seq data. Distances are derived from all detected genes within the pathway. Label colours indicate the direction of enrichment in GSEA results for genes in the pathway.*")}
tg_down %>% 
  activate(nodes) %>% 
  mutate(direction = ifelse(NES > 0, "Up", "Down")) %>% 
  ggraph(layout = net_layout, weights = oc^2) +
  geom_edge_link(aes(width = oc^2, alpha = oc^2)) +
  geom_node_point(
    aes(fill = Status, size = numDEInCat),
    shape = 21
  ) +
  geom_node_text(
    aes(label = label, colour = direction),
    size = 3, 
    data = . %>%
      mutate(
        label = str_replace_all(label, "_", " ") %>% str_trunc(80) %>% str_wrap(width = 18)
      ),
    repel = TRUE, max.overlaps = max(10, round(length(tg_down) / 4, 0)),
    bg.color = "white", bg.r = 0.1, 
  ) +
  scale_x_continuous(expand = expansion(c(0.1, 0.1))) +
  scale_y_continuous(expand = expansion(c(0.1, 0.1))) +
  scale_fill_manual(values = status_cols) +
  scale_size_continuous(range = c(1, 10)) +
  scale_edge_width(range = c(1, 6), limits = c(0, 1)) +
  scale_edge_alpha(range = c(0.1, 0.4), limits = c(0, 1))  +
  scale_colour_manual(
    values = setNames(direction_colours, str_to_title(names(direction_colours)))
  ) +
  guides(edge_alpha= "none") +
  labs(
    size = glue("{target} Targets In Leading Edge") %>% str_wrap(15),
    colour = "GSEA\nDirection",
    fill = "Prior\nSignificance",
    edge_width = expr(paste(OC^2))
  ) +
  theme_void() +
  theme(legend.text = element_text(hjust = 0.5))
```


## Highly Ranked Genes {.tabset}

```{r setup-genes-to-plot}
knitr::opts_chunk$set(dev = fig_type)
all_ids <- unlist(merged_results$gene_id)
db_ids <- unlist(filter(merged_results, !!sym(fdr_column) < fdr_alpha)$gene_id)
top_diff_win <- rnaseq %>% 
  mutate(
    n_windows = vapply(
      gene_id,
      function(x) sum(str_detect(all_ids, x)),
      integer(1)
    ),
    n_db_windows = vapply(
      gene_id,
      function(x) sum(str_detect(db_ids, x)),
      integer(1)
    )
  ) %>% 
  arrange(!!sym(rna_p_col)) %>% 
  dplyr::filter(n_windows > 1) %>% 
  dplyr::slice(1:5)
grl_to_plot <- top_diff_win$gene_id %>% 
  sapply(
    function(x) {
      subset(
        merged_results, 
        vapply(gene_id, function(id) x %in% id, logical(1))
      ) %>% 
        granges() %>% 
        ## Restrict this to only sites within 5Mb of a gene
        subsetByOverlaps(
          resize(subset(genes_gr, gene_id == x), width = 1e7, fix = 'center')
        ) %>% 
        c(
          granges(subset(genes_gr, gene_id == x))
        ) %>% 
        unstrand() %>% 
        sort()
    },
    simplify = FALSE
  ) %>% 
  lapply(range) %>% 
  GRangesList()
hfgc_genes <- hfgc_genes[names(hfgc_genes) != "Undetected"]
if (!is.null(config$external$coverage)) {
  y_ranges <- granges(unlist(grl_to_plot))
  y_lim <- c(
    y_lim[target],
    bwfl[names(bwfl) != target] %>% 
      lapply(
        function(x) {
          GRangesList(lapply(x, import.bw, which = y_ranges)) %>% 
            unlist() %>% 
            filter(score == max(score)) %>% 
            mcols() %>% 
            .[["score"]] %>% 
            c(0) %>% 
            range()
        }
      )
  )
}
i <- 1
```

Using the ranked list of genes from the RNA-Seq data, the 5 genes most highly ranked for differential expression, with more than one `r target`-bound window mapped to it were selected for visualisation.

```{r make-deg-caption, echo = FALSE}
.makeDEGCaption <- function(id) {
  nm <- id2gene[id]
  df <- dplyr::filter(top_diff_win, gene_id == id)
  fdr <- ifelse(
    df[[rna_fdr_col]] < 0.001,
    sprintf("%.2e", df[[rna_fdr_col]]),
    sprintf("%.3f", df[[rna_fdr_col]])
  )
  glue(
    "*Region showing all merged {target}-bound windows mapped to ", 
    "{nm}. Windows considered to be unchanged for {target} are ", 
    "annotated in grey, with other colours indicating {target} gain or loss. ",
    "Mapped windows are only shown if within 5Mb of {nm}. ",
    ifelse(
      has_hic,
      glue(
        "HiC Interactions are only shown which directly connect {target} to {nm}. "
      ),
      glue("")
    ),
    "Gene-centric regions {ifelse(has_features, 'and external features', '')} ",
    "are shown in the top panel. The estimated logFC for {nm} is ",
    sprintf("%.2f ", df$logFC),
    "with an FDR of {fdr}. Undetected and unchanged genes are shown on ",
    "separate tracks.*"
  )
}
```


### `r id2gene[names(grl_to_plot)[i]]`


```{r plot-gene1, fig.cap = .makeDEGCaption(names(grl_to_plot)[i])}
ct <- trans_models %>% 
  subsetByOverlaps(grl_to_plot[[i]]) %>% 
  as_tibble() %>% 
  pull("transcript") %>% 
  unique() %>% 
  length() %>% 
  is_greater_than(10) %>% 
  ifelse("meta", FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic %>% 
     subsetByOverlaps(
       subset(genes_gr, gene_id == names(grl_to_plot)[[i]])
     ) %>% 
    subsetByOverlaps(
      subsetByOverlaps(merged_results, grl_to_plot[[i]])
    ),
  features = feat_gr, featcol = feature_colours, featstack = "dense", 
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %>%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  fontsize = 10,
  max = width(grl_to_plot[[i]]),
  ylim = y_lim,
  highlight = c(),
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
i <- i + 1
```

### `r id2gene[names(grl_to_plot)[i]]`


```{r plot-gene2, fig.cap = .makeDEGCaption(names(grl_to_plot)[i])}
ct <- trans_models %>% 
  subsetByOverlaps(grl_to_plot[[i]]) %>% 
  as_tibble() %>% 
  pull("transcript") %>% 
  unique() %>% 
  length() %>% 
  is_greater_than(10) %>% 
  ifelse("meta", FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic %>% 
     subsetByOverlaps(
       subset(genes_gr, gene_id == names(grl_to_plot)[[i]])
     ) %>% 
    subsetByOverlaps(
      subsetByOverlaps(merged_results, grl_to_plot[[i]])
    ),
  features = feat_gr, featcol = feature_colours, featstack = "dense", 
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %>%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  max = width(grl_to_plot[[i]]),
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
i <- i + 1
```

### `r id2gene[names(grl_to_plot)[i]]`


```{r plot-gene3, fig.cap = .makeDEGCaption(names(grl_to_plot)[i])}
ct <- trans_models %>% 
  subsetByOverlaps(grl_to_plot[[i]]) %>% 
  as_tibble() %>% 
  pull("transcript") %>% 
  unique() %>% 
  length() %>% 
  is_greater_than(10) %>% 
  ifelse("meta", FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic %>% 
     subsetByOverlaps(
       subset(genes_gr, gene_id == names(grl_to_plot)[[i]])
     ) %>% 
    subsetByOverlaps(
      subsetByOverlaps(merged_results, grl_to_plot[[i]])
    ),
  features = feat_gr, featcol = feature_colours, featstack = "dense", 
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %>%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  max = width(grl_to_plot[[i]]),
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
i <- i + 1
```

### `r id2gene[names(grl_to_plot)[i]]`


```{r plot-gene4, fig.cap = .makeDEGCaption(names(grl_to_plot)[i])}
ct <- trans_models %>% 
  subsetByOverlaps(grl_to_plot[[i]]) %>% 
  as_tibble() %>% 
  pull("transcript") %>% 
  unique() %>% 
  length() %>% 
  is_greater_than(10) %>% 
  ifelse("meta", FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic %>% 
     subsetByOverlaps(
       subset(genes_gr, gene_id == names(grl_to_plot)[[i]])
     ) %>% 
    subsetByOverlaps(
      subsetByOverlaps(merged_results, grl_to_plot[[i]])
    ),
  features = feat_gr, featcol = feature_colours, featstack = "dense", 
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %>%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  max = width(grl_to_plot[[i]]),
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
i <- i + 1
```

### `r id2gene[names(grl_to_plot)[i]]`


```{r plot-gene5, fig.cap = .makeDEGCaption(names(grl_to_plot)[i])}
ct <- trans_models %>% 
  subsetByOverlaps(grl_to_plot[[i]]) %>% 
  as_tibble() %>% 
  pull("transcript") %>% 
  unique() %>% 
  length() %>% 
  is_greater_than(10) %>% 
  ifelse("meta", FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic %>% 
     subsetByOverlaps(
       subset(genes_gr, gene_id == names(grl_to_plot)[[i]])
     ) %>% 
    subsetByOverlaps(
      subsetByOverlaps(merged_results, grl_to_plot[[i]])
    ),
  features = feat_gr, featcol = feature_colours, featstack = "dense", 
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %>%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  max = width(grl_to_plot[[i]]),
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
```

