This will be to integrate RNA-Seq data with DB peaks.
Key questions that should be addressed

- Are DE genes mapped to peaks
    + Which region of the gene for detected?
    + Which region of the gene for DE?
- If no DE genes, what is the best GSEA-type approach
- Pathway analysis for peaks
    + All genes mapped to peaks
    + DE genes mapped to peaks
    
We already have these values:
"rna_col" - Can be tx_id, transcript_id or gene_id
"rna_fdr_col" - Just finds the first column matching 'FDR'
"rna_gr_col" - Can be transcript_id or gene_id
"rnaseq" - This is the data

Write this workflow for the case: `rna_col == 'gene_id'`


```{r plot-p-target, fig.height = 7, fig.cap = glue("*P-Values for differential expression partitioned by whether a {target} ChIP peak was mapped to each gene.*")}
rna_p_col <- colnames(rnaseq)[
  str_detect(str_to_lower(colnames(rnaseq)), "pval|p\\.val")
][1]
stopifnot(length(rna_p_col) == 1)
y <- ggplot_build(
  tibble(p = rnaseq[[rna_p_col]]) %>% 
    ggplot(aes(p, stat(density))) + 
    geom_histogram(bins = 100) 
) %>% .[["data"]] %>% 
  .[[1]] %>% 
  .[["y"]] %>% 
  max()
rnaseq %>% 
  mutate(
    mapped = gene_id %in% unlist(merged_results$gene_id),
    mapped = ifelse(mapped, glue("{target} Peak"), glue("No {target} Peak"))
  ) %>%
  ggplot(
    aes(!!sym(rna_p_col), stat(density))
  ) +
  geom_histogram(
    colour = "black", fill = "grey", bins = 100
  ) +
  geom_text(
    aes(0.85, y, label = lab),
    data = . %>% 
      group_by(mapped) %>% 
      tally() %>% 
      mutate(lab = glue("N = {comma(n, 1)}"))
  ) +
  facet_wrap(~mapped)
```


```{r}
rna_status <- rnaseq %>% 
  mutate(
    de_status = case_when(
      !!sym(rna_fdr_col) < fdr_alpha & !!sym(lfc_col) < 0 ~ "Down",
      !!sym(rna_fdr_col) < fdr_alpha & !!sym(lfc_col) > 0 ~ "Up",
      !!sym(rna_fdr_col) >= fdr_alpha ~ "Unchanged"
    )
  ) %>% 
  dplyr::select(gene_id, gene_name, de_status) %>% 
  bind_rows(
    genes_gr %>% 
      subset(!gene_id %in% rnaseq$gene_id) %>% 
      select(gene_id, gene_name) %>% 
      mcols() %>% 
      as_tibble() %>% 
      mutate(de_status = "Undetected")
  ) %>% 
  mutate(
    de_status = factor(
      de_status, levels = str_to_title(names(colours$direction))
    )
  ) %>% 
  arrange(gene_id)
```


```{r plot-region-detected, fig.height = 6, fig.cap = "*Distribution of merged windows by region, according to whether the window is mapped to a detected gene in the RNA-Seq dataset.*"}
merged_results %>% 
  mutate(
    any_detected = vapply(detected, length, integer(1)) > 1
  ) %>% 
  group_by(any_detected, region) %>% 
  summarise(n = n()) %>% 
  as_tibble() %>% 
  group_by(any_detected) %>% 
  mutate(
    p = n / sum(n),
    x = cumsum(p),
    any_detected = ifelse(any_detected, "Detected", "Not Detected"),
    any_detected = glue("{any_detected}\n(N = {comma(sum(n), 1)})")
  ) %>% 
  ungroup() %>% 
  ggplot(
    aes(p, any_detected, fill = fct_rev(region))
  ) +
  geom_col() +
  geom_label(
    aes(x - 0.5*p, any_detected, label = percent(p, 0.1)),
    data = . %>% dplyr::filter(p > 0.03),
    alpha = 0.6, fill = "white"
  ) +
  scale_fill_manual(
    values = colours$regions %>% 
      setNames(regions[names(.)]) %>% 
      setNames(str_sep_to_title(names(.)))
  ) +
  scale_x_continuous(expand = expansion(0), labels = percent) +
  labs(
    x = c(), y = "Mapped To Detected Genes", fill = "Region"
  ) +
  theme(
    axis.text.y = element_text(hjust = 0.5)
  )
```


```{r plot-feature-detected, eval = has_features, echo = has_features, fig.height = 6, fig.cap = "*Distribution of merged windows by feature, according to whether the window is mapped to a detected gene in the RNA-Seq dataset.*"}
merged_results %>% 
  mutate(
    any_detected = vapply(detected, length, integer(1)) > 1
  ) %>% 
  group_by(any_detected, feature) %>% 
  summarise(n = n()) %>% 
  as_tibble() %>% 
  group_by(any_detected) %>% 
  mutate(
    p = n / sum(n),
    x = cumsum(p),
    any_detected = ifelse(any_detected, "Detected", "Not Detected"),
    any_detected = glue("{any_detected}\n(N = {comma(sum(n), 1)})")
  ) %>% 
  ungroup() %>% 
  ggplot(
    aes(p, any_detected, fill = fct_rev(feature))
  ) +
  geom_col() +
  geom_label(
    aes(x - 0.5*p, any_detected, label = percent(p, 0.1)),
    data = . %>% dplyr::filter(p > 0.03),
    alpha = 0.6, fill = "white"
  ) +
  scale_fill_manual(
    values = colours$feature %>% 
      setNames(str_sep_to_title(names(.)))
  ) +
  scale_x_continuous(expand = expansion(0), labels = percent) +
  labs(
    x = c(), y = "Mapped To Detected Genes", fill = "Feature"
  ) +
  theme(
    axis.text.y = element_text(hjust = 0.5)
  )
```

```{r plot-feature-region-detected, eval = has_features, echo = has_features, fig.cap = glue("*Detected/Undetected genes mapped by region and feature for merged windows*")}
merged_results %>%
  mutate(
    any_detected = ifelse(
      vapply(detected, length, integer(1)) > 1,
      "Detected", "Not Detected"
    ) %>% 
      factor(levels = c("Not Detected", "Detected"))
  ) %>% 
  plotPie(fill = "any_detected", x = "region", y = "feature") +
  scale_fill_manual(values = c("grey", "forestgreen"))
```


```{r}
db_labeller <- as_labeller(
  c(
    Up = glue("{target} Increase"),
    Down = glue("{target} Decrease"),
    Unchanged = glue("{target} Unchanged")
  )
)
merged_results %>% 
  mutate(
    any_detected = vapply(detected, length, integer(1)) > 1
  ) %>% 
  group_by(any_detected, region, status) %>% 
  summarise(n = n()) %>% 
  as_tibble() %>% 
  group_by(any_detected, status) %>% 
  mutate(
    p = n / sum(n),
    x = cumsum(p),
    any_detected = ifelse(any_detected, "Detected", "Not Detected"),
    any_detected = glue("{any_detected}\n(N = {comma(sum(n), 1)})")
  ) %>% 
  ungroup() %>% 
  ggplot(
    aes(p, any_detected, fill = fct_rev(region))
  ) +
  geom_col() +
  geom_label(
    aes(x - 0.5*p, any_detected, label = percent(p, 0.1)),
    data = . %>% dplyr::filter(p > 0.03),
    alpha = 0.6, fill = "white", size = 3.5
  ) +
  facet_grid(status~., scales = "free_y", space = "free", labeller = db_labeller) +
  scale_fill_manual(
    values = colours$regions %>% 
      setNames(regions[names(.)]) %>% 
      setNames(str_sep_to_title(names(.)))
  ) +
  scale_x_continuous(expand = expansion(0), labels = percent) +
  labs(
    x = c(), y = "Mapped To Detected Genes", fill = "Region"
  ) +
  theme(
    axis.text.y = element_text(hjust = 0.5)
  )
## Not really convincing
```

```{r explore-scatterpie}
df <- merged_results %>%
  select(status, region, gene_id) %>% 
  as_tibble() %>% 
  unnest(gene_id) %>% 
  left_join(rna_status, by = "gene_id") %>% 
  dplyr::filter(de_status %in% c("Up", "Down")) %>% 
  group_by(de_status, status, region) %>% 
  summarise(n = dplyr::n(), .groups = "drop_last") %>%
  mutate(N = sum(n)) %>% 
  ungroup() %>% 
  mutate(
    x = as.integer(status),
    y = as.integer(de_status),
    r = N / sum(N)
  ) %>% 
  pivot_wider(names_from = "region", values_from = "n", values_fill = 0)

ggplot() +
  geom_scatterpie(
    aes(x, y, r = 2*r),
    data = df,
    cols = regions
  ) +
  geom_label(
    aes(x, y, label = lab),
    data = df %>% 
      # dplyr::filter(N > 0.025*sum(N)) %>% 
      mutate(lab = comma(N, 1)),
    alpha = 0.7, size = 3.5
  ) +
  coord_equal() +
  scale_y_continuous(
    breaks = seq_along(levels(df$de_status)), labels = levels(df$de_status)
  ) +
  scale_x_continuous(
    breaks = seq_along(levels(df$status)), labels = levels(df$status)
  ) +
  scale_fill_manual(
    values = colours$regions %>% 
      setNames(regions[names(.)])
  ) +
  labs(
    x = glue("Changed {target} Binding"),
    y = "DE Genes",
    fill = "Region"
  )
# Maybe
```

```{r explore-gsea}
library(fgsea)
rna_ranks <- rnaseq %>% 
  mutate(
    ranking_stat = -sign(!!sym(lfc_col))*log10(!!sym(rna_p_col)) %>% 
      setNames(gene_name)
  ) %>% 
  arrange(ranking_stat) %>% 
  dplyr::filter(!!sym(rna_p_col) < 1) %>% 
  pull("ranking_stat") 
reg_dir_gs <- merged_results %>% 
  select(
    overlaps_ref, status, any_of(c("region", "feature")), starts_with("gene")
  ) %>% 
  as_tibble() %>% 
  mutate(group = glue("{region} / {target} {status}")) %>% 
  split(.$group)%>% 
  lapply(pull, "gene_name") %>% 
  lapply(unlist) %>% 
  lapply(unique)
if (has_features) {
  reg_dir_gs <- merged_results %>% 
      select(
        overlaps_ref, status, any_of(c("region", "feature")), starts_with("gene")
      ) %>% 
      as_tibble() %>% 
      mutate(group = glue("{region} / {feature} / {target} {status}")) %>% 
      split(.$group)%>% 
      lapply(pull, "gene_name") %>% 
      lapply(unlist) %>% 
      lapply(unique)
}

## The full dataset
## Directional
fgsea_dir <- fgsea(reg_dir_gs, rna_ranks, minSize = 20) 
fgsea_dir %>% 
  as_tibble()%>% 
  arrange(pval) %>% 
  dplyr::filter(padj < 0.05) %>% 
  # dplyr::slice(1:10) %>% 
  mutate(leadingEdge = vapply(leadingEdge, paste, character(1), collapse = "; "))

reg_dir_gs[
  fgsea_dir %>% arrange(pval) %>% dplyr::slice(1:5) %>% pull("pathway")
] %>% 
  plotGseaTable(rna_ranks, fgseaRes = fgsea_dir)

## Non-directional
fgsea_non <- fgsea(reg_dir_gs, sort(abs(rna_ranks)), minSize = 20, scoreType = "pos") 
fgsea_non %>% 
  as_tibble() %>% 
  arrange(pval) %>% 
  mutate(
    gene_name = vapply(leadingEdge, paste, character(1), collapse = ": ")
  )

grid::grid.newpage()
reg_dir_gs[
  fgsea_non %>% arrange(pval) %>% dplyr::filter(padj < 0.05) %>% pull("pathway")
] %>% 
  plotGseaTable(sort(abs(rna_ranks)), fgseaRes = fgsea_dir)


plotEnrichment(
  reg_dir_gs[["Intron / Enhancer / ER Unchanged"]],
  sort(abs(rna_ranks))
) +
  ggtitle("Intron / Enhancer / ER Unchanged") # Probs co-factors


```


```{r}
all_ids <- unlist(merged_results$gene_id)
db_ids <- unlist(filter(merged_results, !!sym(fdr_column) < fdr_alpha)$gene_id)
top_diff_win <- rnaseq %>% 
  mutate(
    n_windows = vapply(
      gene_id,
      function(x) sum(str_detect(all_ids, x)),
      integer(1)
    ),
    n_db_windows = vapply(
      gene_id,
      function(x) sum(str_detect(db_ids, x)),
      integer(1)
    )
  ) %>% 
  dplyr::filter(!!sym(rna_fdr_col) < fdr_alpha, n_db_windows > 1) %>% 
  dplyr::slice(1:5)

grl_to_plot <- top_diff_win$gene_id %>% 
  sapply(
    function(x) {
      subset(
        merged_results, 
        vapply(gene_id, function(id) x %in% id, logical(1))
      ) %>% 
        granges() %>% 
        c(
          granges(subset(genes_gr, gene_id == x))
        ) %>% 
        unstrand()
    },
    simplify = FALSE
  ) %>% 
  lapply(range) %>% 
  GRangesList()
```





```{r}
if (!is.null(config$external$coverage)) {
  y_ranges <- granges(unlist(grl_to_plot))
  y_lim <- c(
    y_lim[target],
    bwfl[names(bwfl) != target] %>% 
      lapply(
        function(x) {
          GRangesList(lapply(x, import.bw, which = y_ranges)) %>% 
            unlist() %>% 
            filter(score == max(score)) %>% 
            mcols() %>% 
            .[["score"]] %>% 
            c(0) %>% 
            range()
        }
      )
  )
}

## Let's just check one of those
## For these, we should use the merged bw, add the DB annotations and include
## any external coverage tracks
plotHFGC(
  grl_to_plot[[5]],
  hic = hic,
  features = feat_gr, featcol = feat_col,
  genes = hfgc_genes, genecol = colours$direction, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %>%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts =  "meta", # Set this as a function of n_transcripts
  zoom = 1.1,
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  showAxis = FALSE, rotation.title = 90
)
```

