# Comparison With RNA Seq

```{r define-objects}
library(fgsea)
library(cowplot)
library(BiocParallel)
register(MulticoreParam(workers = threads))
id2gene <- structure(all_gr$gene$gene_name, names = all_gr$gene$gene_id)
rna_status <- rnaseq %>% 
  mutate(
    de_status = case_when(
      !!sym(rna_fdr_col) < fdr_alpha & !!sym(rna_lfc_col) < 0 ~ "Down",
      !!sym(rna_fdr_col) < fdr_alpha & !!sym(rna_lfc_col) > 0 ~ "Up",
      !!sym(rna_fdr_col) >= fdr_alpha ~ "Unchanged"
    )
  ) %>% 
  dplyr::select(gene_id, gene_name, de_status) %>% 
  bind_rows(
    all_gr$gene %>% 
      subset(!gene_id %in% rnaseq$gene_id) %>% 
      select(gene_id, gene_name) %>% 
      mcols() %>% 
      as_tibble() %>% 
      mutate(de_status = "Undetected")
  ) %>% 
  mutate(
    de_status = factor(
      de_status, levels = str_to_title(names(colours$direction))
    )
  ) %>% 
  arrange(gene_id)
rna_p_col <- colnames(rnaseq)[
  str_detect(str_to_lower(colnames(rnaseq)), "pval|p\\.val")
][1]
stopifnot(length(rna_p_col) == 1)
diff_windows <- all_windows %>% 
  as_tibble() 
diff_windows[["r1"]] <- rank(mcols(all_windows)[[c1_fdr]], ties.method = "min")
diff_windows[["r2"]] <- rank(mcols(all_windows)[[c2_fdr]], ties.method = "min")
# %>% 
#   mutate(
#     r1 = rank(!!sym(c1_fdr), ties.method = "min"),
#     r2 = rank(!!sym(c2_fdr), ties.method = "min")
#   ) %>% 
diff_windows <- diff_windows %>% 
  dplyr::arrange(r1 + r2) %>% 
  dplyr::select(gene_id, status, r1, r2) %>% 
  unnest(everything()) %>% 
  distinct(gene_id, .keep_all = TRUE) %>% 
  droplevels() %>% 
  split(.$status) %>% 
  lapply(pull, "gene_id") 
any_de <- sum(rnaseq[[rna_fdr_col]] < fdr_alpha) > 0
```

## Volcano Plots

```{r plot-volcano, fig.cap = glue("*Volcano plots for differential expression separated by target binding patterns. Many genes will have multiple target binding regions mapped to them, and as such only the mapping with the most highly-ranked region for binding across both comparisons was retained. In subsequent enrichment analyses, these genes will appear in multiple groups given that multiple binding patterns may be associated with them.*")}
names(diff_windows) %>% 
  lapply(
    function(x) {
      rnaseq %>% 
        dplyr::filter(gene_id %in% diff_windows[[x]]) %>% 
        mutate(binding_status = x)
    }
  ) %>% 
  bind_rows() %>% 
  bind_rows(
    rnaseq %>% 
      dplyr::filter(!gene_id %in% unlist(diff_windows)) %>% 
      mutate(binding_status = combs[16])
  ) %>% 
  left_join(rna_status) %>% 
  mutate(
    binding_status = binding_status %>% 
      factor(levels = combs) %>% 
      fct_lump_prop(0.01)
  ) %>% 
  ggplot(
    aes(!!sym(rna_lfc_col), -log10(!!sym(rna_p_col)), colour = de_status)
  ) +
  geom_point() +
  geom_text_repel(
    aes(label = gene_name),
    data = . %>% 
      arrange(!!sym(rna_p_col)) %>% 
      split(.$binding_status) %>% 
      lapply(dplyr::slice, 1:2) %>% 
      bind_rows(),
    show.legend = FALSE
  ) +
  geom_text(
    aes(label = lab),
    data = . %>% 
      group_by(binding_status) %>% 
      summarise(n = dplyr::n()) %>% 
      mutate(lab = glue("n = {comma(n)}")),
    x = min(rnaseq[[rna_lfc_col]]) + 0.9 * diff(range(rnaseq[[rna_lfc_col]])),
    y = 1,
    colour = "black",
    inherit.aes = FALSE,
    show.legend = FALSE
  ) +
  facet_wrap(~binding_status) +
  scale_colour_manual(
    values = colours$direction[c("up", "down", "unchanged")] %>% 
      unlist() %>% 
      setNames(str_to_title(names(.))) 
  ) +
  labs(colour = "Status") +
  theme(legend.position = "bottom")
```


`r if (any_de) "## DE Genes And Pairwise Binding"`

```{r tab-de-both, eval = any_de, echo=any_de}
col_defs <- list2(
      de = colDef(
        name = "Gene", maxWidth = 120
      ),
      logFC = colDef(
        name = "logFC<br>(RNA-Seq)",
        html = TRUE,
        aggregate = "mean",
        maxWidth = 100,
        format = colFormat(digits = 2),
        style = JS(
            glue(
            "function(rowInfo) {
              var value = rowInfo.row['logFC']
              if (value < 0) {
                var color = '{{colours$direction[['down']]}}'
              } else if (value > 0) {
                var color = '{{colours$direction[['up']]}}'
              } 
              return { color: color }
            }",
            .open = "{{", .close = "}}"
            )
        )
      ),
      range = colDef(
        name = "Range",
        aggregate = JS(
            "function(values){
              var chrom = [];
              var rng = [];
              var start = [];
              var end = [];
              for (i = 0; i < values.length; i++) {
                chrom[i] = values[i].split(':')[0];
                rng[i] = values[i].split(':')[1];
                start[i] = rng[i].split('-')[0];
                end[i] = rng[i].split('-')[1];
              }
              
              min_start = Math.min(...start);
              max_end = Math.max(...end);
              var ret_val = [...new Set(chrom)] + ':' + min_start.toString() + '-' + max_end.toString();
              
              return ret_val
              
            }"
          )
      ),
      region = colDef(
        aggregate = "unique",
        name = "Region"
      ),
      "{c1_logfc}" := colDef(
        name = comps[[1]],
        maxWidth = 100,
        format = colFormat(digits = 2),
        style = function(value) {
          cl <- ifelse(
            value > 0, colours$direction[["up"]], colours$direction[["down"]]
          )
          list(color = cl)
        }
      ),
      "{c2_logfc}" := colDef(
        name = comps[[2]],
        maxWidth = 100,
        format = colFormat(digits = 2),
        style = function(value) {
          cl <- ifelse(
            value > 0, colours$direction[["up"]], colours$direction[["down"]]
          )
          list(color = cl)
        }
      ),
      "{c1_status}" := colDef(
        aggregate = "unique",
        name = comps[[1]],
        maxWidth = 150
      ),
      "{c2_status}" := colDef(
        aggregate = "unique",
        name = comps[[2]],
        maxWidth = 150
      )
    ) 
if (has_features) col_defs$feature <- colDef(
  name = "Feature",
  aggregate = "unique"   
)
htmltools::tags$caption(
  htmltools::em(
    glue(
    "
    All differentially expressed genes where binding was detected in both 
    comparisons.
    "
    )
  )
)
de_genes_both_comps <- all_windows %>% 
  as_tibble()
## Placeholder until conda-forge sort out R4.2.1
colnames(de_genes_both_comps) <- c("range", colnames(mcols(all_windows)))
de_genes_both_comps <- de_genes_both_comps %>% 
  dplyr::select(
    any_of(c("range", "region", "feature")), ends_with("logFC"), ends_with("status"), 
    any_of(c("detected"))
  ) %>% 
  mutate(
    de = lapply(detected, intersect, dplyr::filter(rnaseq, !!sym(rna_fdr_col) < fdr_alpha)$gene_name)
  ) %>% 
  dplyr::filter(
    vapply(de, length, integer(1)) > 0, 
    str_detect(status, "Up|Down"),
    !str_detect(status, "Undetected")
  ) %>% 
  unnest(de) %>% 
  left_join(rnaseq, by = c("de" = "gene_name")) %>% 
  mutate(direction = ifelse(logFC > 0, "Up", "Down")) %>% 
  droplevels() %>% 
  dplyr::select(
    de, logFC, any_of(c("range", "region", "feature")), ends_with("logFC"),
    ends_with("_status")
  ) 
de_genes_both_comps %>% 
  reactable(
    filterable = TRUE,
    groupBy = "de",
    columns = col_defs,
    columnGroups = list(
      colGroup(
        name = "ChIP logFC", 
        columns = c(as.character(c1_logfc), as.character(c2_logfc))
      ),
      colGroup(
        name = "ChIP Status",
        columns =  c(as.character(c1_status), as.character(c2_status))
      )
    ),
    theme = reactableTheme(style = list(fontSize = 14))
  )
```



## Enrichment Analysis {.tabset}


```{r define_genesets}
rna_dir_ranks <- rnaseq %>% 
  mutate(
    ranking_stat = -sign(!!sym(rna_lfc_col))*log10(!!sym(rna_p_col)) %>% 
      setNames(gene_id)
  ) %>% 
  arrange(ranking_stat) %>% 
  dplyr::filter(!!sym(rna_p_col) < 1) %>% 
  pull("ranking_stat") 
rna_sig_ranks <- rna_dir_ranks %>% 
  abs() %>% 
  sort()
status_gs <- all_windows %>% 
  select(status, starts_with("gene")) %>% 
  as_tibble() %>% 
  split(.$status) %>% 
  lapply(pull, "gene_id") %>% 
  lapply(unlist) %>% 
  lapply(unique) %>% 
  .[vapply(., length, integer(1)) > 20]
```


### Directional GSEA

```{r gsea-dir-sig}
gsea_dir <- fgsea(status_gs, rna_dir_ranks)
gsea_dir_sig <- gsea_dir %>%
  arrange(pval) %>%
  dplyr::filter(padj < enrich_alpha) %>%
  as_tibble()
p <- gsea_dir_sig %>% 
  dplyr::slice(1:9) %>% 
  pull("pathway") %>% 
  lapply(
    function(x) {
      plotEnrichment(status_gs[[x]], rna_dir_ranks) +
        ggtitle(x) +
        theme(
          plot.title = element_text(hjust = 0.5, size = 10)
        )
    }
  )
```

`r if (nrow(gsea_dir_sig) == 0) glue("No association was found between joint binding patterns and *directional* significance in differentially expressed genes.")`

```{r tbl-gsea-dir_sig, eval = nrow(gsea_dir_sig) > 0, echo = nrow(gsea_dir_sig) > 0}
cp <-  htmltools::em(
  glue(
    "Combined windows were mapped to genes, and their position amongst the ",
    "RNA-Seq results was assessed. {nrow(gsea_dir_sig)} sets of windows were ",
    "associated with changes in gene expression, using the sign of ", 
    "fold-change and ranking statistic to initially rank the ", 
    "{comma(nrow(rnaseq), 1)} genes considered as detected."
  )
)
tbl <- gsea_dir_sig %>%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %>% 
      vapply(paste, character(1), collapse = "; "),
    Direction = ifelse(NES > 0, "\u21E7 Up-regulated", "\u21E9 Down-regulated")
  ) %>%
  separate(pathway, comps, sep = " - ") %>% 
  dplyr::select(
    all_of(comps), Windows = size, Direction, 
    p = pval, FDR = padj, `Edge Size`, `Leading Edge` = leadingEdge
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list2(
      "{comps[[1]]}" := colDef(
        maxWidth = 150,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      "{comps[[2]]}" := colDef(
        maxWidth = 150,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      Windows = colDef(maxWidth = 80),
      Direction = colDef(
        name = "Gene Direction",
        maxWidth = 120,
        style = function(value) {
          colour <- ifelse(
            str_detect(value, "Up"), 
            colours$direction[["up"]], 
            colours$direction[["down"]]
          )
          list(color = colour)
        },
      ),
      p = colDef(
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ), 
        maxWidth = 80
      ),
      FDR = colDef(
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ), 
        maxWidth = 80
      ),
      "Edge Size" = colDef(maxWidth = 80),
     "Leading Edge" = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```


```{r, echo=FALSE}
## Set the default device to be png for barcode plots as they can become
## very large for some datasets
knitr::opts_chunk$set(dev = 'png')
```


```{r barcode-gsea-dir-sig, eval = nrow(gsea_dir_sig) > 0, echo = nrow(gsea_dir_sig) > 0, fig.height= 3 * ceiling(length(p) / 3), fig.cap = glue("*Barcode plots for the top {length(p)} sets of windows associated with __directional__ changes in gene expression.*")}
plot_grid(plotlist = p, nrow = ceiling(length(p) / 3))
```


### Non-Directional GSEA

```{r gsea-nondir-sig}
gsea_nondir <- fgsea(status_gs, rna_sig_ranks)
gsea_nondir_sig <- gsea_nondir %>% 
  arrange(pval) %>% 
  dplyr::filter(padj < enrich_alpha) %>% 
  as_tibble() 
p <- gsea_nondir_sig %>% 
  dplyr::slice(1:9) %>% 
  pull("pathway") %>% 
  lapply(
    function(x) {
      plotEnrichment(status_gs[[x]], rna_sig_ranks) +
        ggtitle(x) +
        theme(plot.title = element_text(hjust = 0.5, size = 10))
    }
  )
```

`r if (nrow(gsea_nondir_sig) == 0) glue("No association was found between {target} binding and *overall* significance in differentially expressed genes.")`

```{r tbl-gsea_nondir_sig, eval = nrow(gsea_nondir_sig) > 0, echo = nrow(gsea_nondir_sig) > 0}
cp <-  htmltools::em(
  glue(
    "Combined windows were mapped to genes, and their position amongst the ",
    "RNA-Seq results was assessed. {nrow(gsea_nondir_sig)} sets of windows were ",
    "associated with changes in gene expression, using only the p-value to ", 
    "rank the {comma(nrow(rnaseq), 1)} genes considered as detected."
  )
)
tbl <- gsea_nondir_sig %>%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %>% 
      vapply(paste, character(1), collapse = "; ")
  ) %>% 
  separate(pathway, comps, sep = " - ") %>% 
  dplyr::select(
    all_of(comps), Windows = size, 
    p = pval, FDR = padj, `Edge Size`, `Leading Edge` = leadingEdge
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list2(
      "{comps[[1]]}" := colDef(
        maxWidth = 150,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      "{comps[[2]]}" := colDef(
        maxWidth = 150,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      Windows = colDef(maxWidth = 80),
      p = colDef(
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ), 
        maxWidth = 80
      ),
      FDR = colDef(
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ), 
        maxWidth = 80
      ),
      "Edge Size" = colDef(maxWidth = 80),
     "Leading Edge" = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```

```{r barcode-gsea-nondir-sig, eval = nrow(gsea_nondir_sig) > 0, echo = nrow(gsea_nondir_sig) > 0, fig.height= 3 * ceiling(length(p) / 3), fig.cap = glue("*Barcode plots for the top {length(p)} sets of windows associated with __non-directional__ changes in gene expression.*")}
plot_grid(plotlist = p, nrow = ceiling(length(p) / 3))
```

```{r, echo=FALSE}
## Reset the default device 
knitr::opts_chunk$set(dev = fig_type)
```

