# Comparison With RNA Seq

```{r define-objects}
library(fgsea)
library(cowplot)
library(BiocParallel)
library(metap)
register(MulticoreParam(workers = threads))
id2gene <- structure(all_gr$gene$gene_name, names = all_gr$gene$gene_id)
rna_status <- rnaseq %>% 
  mutate(
    de_status = case_when(
      !!sym(rna_fdr_col) < fdr_alpha & !!sym(rna_lfc_col) < 0 ~ "Down",
      !!sym(rna_fdr_col) < fdr_alpha & !!sym(rna_lfc_col) > 0 ~ "Up",
      !!sym(rna_fdr_col) >= fdr_alpha ~ "Unchanged"
    )
  ) %>% 
  dplyr::select(gene_id, gene_name, de_status) %>% 
  bind_rows(
    all_gr$gene %>% 
      subset(!gene_id %in% rnaseq$gene_id) %>% 
      select(gene_id, gene_name) %>% 
      mcols() %>% 
      as_tibble() %>% 
      mutate(de_status = "Undetected")
  ) %>% 
  mutate(
    de_status = factor(
      de_status, levels = str_to_title(names(colours$direction))
    )
  ) %>% 
  arrange(gene_id)
rna_p_col <- colnames(rnaseq)[
  str_detect(str_to_lower(colnames(rnaseq)), "pval|p\\.val")
][1]
stopifnot(length(rna_p_col) == 1)
diff_windows <- all_windows %>% 
  as_tibble() 
diff_windows[["r1"]] <- rank(mcols(all_windows)[[c1_fdr]], ties.method = "min")
diff_windows[["r2"]] <- rank(mcols(all_windows)[[c2_fdr]], ties.method = "min")
diff_windows <- diff_windows %>% 
  dplyr::arrange(r1 + r2) %>% 
  dplyr::select(gene_id, status, r1, r2) %>% 
  unnest(everything()) %>% 
  distinct(gene_id, .keep_all = TRUE) %>% 
  droplevels() %>% 
  split(.$status) %>% 
  lapply(pull, "gene_id") 
any_de <- sum(rnaseq[[rna_fdr_col]] < fdr_alpha) > 0
n_mapped <- all_windows %>% 
  dplyr::select(status, detected) %>% 
  as_tibble() %>% 
  unnest(everything()) %>% 
  distinct(status, detected) %>% 
  group_by(status) %>% 
  summarise(detected = dplyr::n()) %>% 
  right_join(n_mapped, by = "status")
```

## Volcano Plots

```{r plot-volcano, fig.cap = glue("*Volcano plots for differential expression separated by target binding patterns. Some genes will have multiple target binding regions mapped to them, and may appear in multiple panels. The two most highly ranked genes for differential expression are labelled in each panel. All DE genes are listed with full binding patterns in the table below.*")}
diff_windows %>% 
  lapply(function(x) tibble(gene_id = x)) %>% 
  bind_rows(.id = "binding_status") %>% 
  mutate(
    binding_status = factor(binding_status, levels = combs) %>% 
      fct_lump_prop(0.01)
  ) %>% 
  full_join(rnaseq) %>% 
  dplyr::filter(!is.na(gene_name)) %>% 
  mutate(
    binding_status = fct_explicit_na(binding_status, na_level = "No Targets"),
    de_status = case_when(
      DE & !!sym(rna_lfc_col) > 0 ~ "Up",
      DE & !!sym(rna_lfc_col) < 0 ~ "Down",
      TRUE ~ "Unchanged"
    ) %>% 
      factor(levels = c("Up", "Down", "Unchanged"))
  ) %>% 
  ggplot(
    aes(!!sym(rna_lfc_col), -log10(!!sym(rna_p_col)), colour = de_status)
  ) +
  geom_point() +
  geom_text_repel(
    aes(label = gene_name),
    data = . %>% 
      arrange(!!sym(rna_p_col)) %>% 
      dplyr::filter(DE) %>% 
      split(.$binding_status) %>% 
      lapply(dplyr::slice, 1:2) %>% 
      bind_rows(),
    show.legend = FALSE
  ) +
  geom_text(
    aes(label = lab),
    data = . %>% 
      group_by(binding_status) %>% 
      summarise(n = dplyr::n()) %>% 
      mutate(lab = glue("n = {comma(n)}")),
    x = min(rnaseq[[rna_lfc_col]]) + 0.9 * diff(range(rnaseq[[rna_lfc_col]])),
    y = 1,
    colour = "black",
    inherit.aes = FALSE,
    show.legend = FALSE
  ) +
  facet_wrap(~binding_status) +
  scale_colour_manual(
    values = colours$direction[c("up", "down", "unchanged")] %>% 
      unlist() %>% 
      setNames(str_to_title(names(.))) 
  ) +
  labs(colour = "DE Status") +
  theme(legend.position = "right")
```


`r if (any_de) "## DE Genes And Pairwise Binding"`

```{r tab-de-both, eval = any_de, echo=any_de}
col_defs <- list2(
      de = colDef(
        name = "Gene", maxWidth = 120
      ),
      logFC = colDef(
        name = "logFC<br>(RNA-Seq)",
        html = TRUE,
        aggregate = "mean",
        maxWidth = 100,
        format = colFormat(digits = 2),
        style = JS(
            glue(
            "function(rowInfo) {
              var value = rowInfo.row['logFC']
              if (value < 0) {
                var color = '{{colours$direction[['down']]}}'
              } else if (value > 0) {
                var color = '{{colours$direction[['up']]}}'
              } 
              return { color: color }
            }",
            .open = "{{", .close = "}}"
            )
        )
      ),
      range = colDef(
        name = "Range",
        aggregate = JS(
            "function(values){
              var chrom = [];
              var rng = [];
              var start = [];
              var end = [];
              for (i = 0; i < values.length; i++) {
                chrom[i] = values[i].split(':')[0];
                rng[i] = values[i].split(':')[1];
                start[i] = rng[i].split('-')[0];
                end[i] = rng[i].split('-')[1];
              }
              
              min_start = Math.min(...start);
              max_end = Math.max(...end);
              var ret_val = [...new Set(chrom)] + ':' + min_start.toString() + '-' + max_end.toString();
              
              return ret_val
              
            }"
          )
      ),
      region = colDef(
        aggregate = "unique",
        name = "Region"
      ),
      "{c1_logfc}" := colDef(
        name = comps[[1]],
        maxWidth = 100,
        format = colFormat(digits = 2),
        style = function(value) {
          cl <- ifelse(
            value > 0, colours$direction[["up"]], colours$direction[["down"]]
          )
          list(color = cl)
        }
      ),
      "{c2_logfc}" := colDef(
        name = comps[[2]],
        maxWidth = 100,
        format = colFormat(digits = 2),
        style = function(value) {
          cl <- ifelse(
            value > 0, colours$direction[["up"]], colours$direction[["down"]]
          )
          list(color = cl)
        }
      ),
      "{c1_status}" := colDef(
        aggregate = "unique",
        name = comps[[1]],
        maxWidth = 150
      ),
      "{c2_status}" := colDef(
        aggregate = "unique",
        name = comps[[2]],
        maxWidth = 150
      )
    ) 
if (has_features) col_defs$feature <- colDef(
  name = "Feature",
  aggregate = "unique"   
)
htmltools::tags$caption(
  htmltools::em(
    glue(
    "
    All differentially expressed genes where binding was detected in both 
    comparisons.
    "
    )
  )
)
de_genes_both_comps <- all_windows %>% 
  as_tibble()
## Placeholder until conda-forge sort out R4.2.1
colnames(de_genes_both_comps) <- c("range", colnames(mcols(all_windows)))
de_genes_both_comps <- de_genes_both_comps %>% 
  dplyr::select(
    any_of(c("range", "region", "feature")), ends_with("logFC"), ends_with("status"), 
    any_of(c("detected"))
  ) %>% 
  mutate(
    de = lapply(detected, intersect, dplyr::filter(rnaseq, !!sym(rna_fdr_col) < fdr_alpha)$gene_name)
  ) %>% 
  dplyr::filter(
    vapply(de, length, integer(1)) > 0, 
    str_detect(status, "Up|Down"),
    !str_detect(status, "Undetected")
  ) %>% 
  unnest(de) %>% 
  left_join(rnaseq, by = c("de" = "gene_name")) %>% 
  mutate(direction = ifelse(logFC > 0, "Up", "Down")) %>% 
  droplevels() %>% 
  dplyr::select(
    de, logFC, any_of(c("range", "region", "feature")), ends_with("logFC"),
    ends_with("_status")
  ) 
de_genes_both_comps %>% 
  reactable(
    filterable = TRUE,
    groupBy = "de",
    columns = col_defs,
    columnGroups = list(
      colGroup(
        name = "ChIP logFC", 
        columns = c(as.character(c1_logfc), as.character(c2_logfc))
      ),
      colGroup(
        name = "ChIP Status",
        columns =  c(as.character(c1_status), as.character(c2_status))
      )
    ),
    theme = reactableTheme(style = list(fontSize = 14))
  )
```



## Binding Pattern GSEA {.tabset}


```{r define_genesets}
rna_dir_ranks <- rnaseq %>% 
  mutate(
    ranking_stat = -sign(!!sym(rna_lfc_col))*log10(!!sym(rna_p_col)) %>% 
      setNames(gene_id)
  ) %>% 
  arrange(ranking_stat) %>% 
  dplyr::filter(!!sym(rna_p_col) < 1) %>% 
  pull("ranking_stat") 
rna_sig_ranks <- rna_dir_ranks %>% 
  abs() %>% 
  sort()
status_gs <- all_windows %>% 
  select(status, starts_with("gene")) %>% 
  as_tibble() %>% 
  split(.$status) %>% 
  lapply(pull, "gene_id") %>% 
  lapply(unlist) %>% 
  lapply(unique) %>% 
  .[vapply(., length, integer(1)) > min_gs_size]
```

GSEA analysis was performed taking the set of genes mapped to each binding pattern.
This was initially performed ranking the RNA-Seq results in order of significance, with most significantly up-regulated and one end and the most significantly down-regulated at the other, i.e. Directional GSEA
A secondary analysis was performed taking only statistical significance into account, without including the direction of any change in expression, i.e. Non_Directional GSEA.
This second analysis allows for detecting regulatory changes due to altered binding, but which are spread between both up- and down-regulation, as may be induced by the presence of other co-factors not under direct investigation here.

### Directional GSEA

```{r gsea-dir-sig}
gsea_dir <- fgsea(status_gs, rna_dir_ranks)
gsea_dir_sig <- gsea_dir %>%
  arrange(pval) %>%
  mutate(padj = p.adjust(pval, adj_method)) %>% 
  dplyr::filter(padj < enrich_alpha, size >= min_sig) %>%
  as_tibble()
p <- gsea_dir_sig %>% 
  dplyr::slice(1:9) %>% 
  pull("pathway") %>% 
  lapply(
    function(x) {
      plotEnrichment(status_gs[[x]], rna_dir_ranks) +
        ggtitle(x) +
        theme(
          plot.title = element_text(hjust = 0.5, size = 10)
        )
    }
  )
```

`r if (nrow(gsea_dir_sig) == 0) glue("No association was found between joint binding patterns and *directional* significance in differentially expressed genes.")`

```{r tbl-gsea-dir_sig, eval = nrow(gsea_dir_sig) > 0, echo = nrow(gsea_dir_sig) > 0}
cp <-  htmltools::em(
  glue(
    "Combined windows were mapped to genes, and their position amongst the ",
    "RNA-Seq results was assessed. {nrow(gsea_dir_sig)} sets of windows were ",
    "associated with changes in gene expression, using the sign of ", 
    "fold-change and ranking statistic to initially rank the ", 
    "{comma(nrow(rnaseq), 1)} genes considered as detected."
  )
)
tbl <- gsea_dir_sig %>%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %>% 
      vapply(paste, character(1), collapse = "; "),
    Direction = ifelse(NES > 0, "\u21E7 Up-regulated", "\u21E9 Down-regulated")
  ) %>%
  separate(pathway, comps, sep = " - ") %>% 
  dplyr::select(
    all_of(comps), Windows = size, Direction, 
    p = pval, padj, `Edge Size`, `Leading Edge` = leadingEdge
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list2(
      "{comps[[1]]}" := colDef(
        maxWidth = 150,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      "{comps[[2]]}" := colDef(
        maxWidth = 150,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      Windows = colDef(maxWidth = 80),
      Direction = colDef(
        name = "Gene Direction",
        maxWidth = 120,
        style = function(value) {
          colour <- ifelse(
            str_detect(value, "Up"), 
            colours$direction[["up"]], 
            colours$direction[["down"]]
          )
          list(color = colour)
        },
      ),
      p = colDef(
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ), 
        maxWidth = 80
      ),
      padj = colDef(
        name = glue("p<sub>{adj_method}</sub>"), html = TRUE,
        cell = function(value) {
          ifelse(
            value < 0.001, sprintf("%.2e", value), sprintf("%.3f", value)
          )
        }
      ),
      "Edge Size" = colDef(maxWidth = 80),
     "Leading Edge" = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```



```{r barcode-gsea-dir-sig, dev = 'png', eval = length(p) > 0, echo = length(p) > 0, fig.height= max(3, 3 * ceiling(length(p) / 3)), fig.cap = glue("*Barcode plots for the top {length(p)} sets of windows associated with __directional__ changes in gene expression.*")}
plot_grid(plotlist = p, ncol = min(3, length(p)))
```


### Non-Directional GSEA

```{r gsea-nondir-sig}
gsea_nondir <- suppressWarnings(
  fgsea(status_gs, rna_sig_ranks, scoreType = "pos")
)
gsea_nondir_sig <- gsea_nondir %>% 
  arrange(pval) %>% 
  mutate(padj = p.adjust(pval, adj_method)) %>% 
  dplyr::filter(padj < enrich_alpha, size >= min_sig) %>% 
  as_tibble() 
p <- gsea_nondir_sig %>% 
  dplyr::slice(1:9) %>% 
  pull("pathway") %>% 
  lapply(
    function(x) {
      plotEnrichment(status_gs[[x]], rna_sig_ranks) +
        ggtitle(x) +
        theme(plot.title = element_text(hjust = 0.5, size = 10))
    }
  )
txt <- ifelse(
  length(p) > 0,
  glue(""),
  glue("No association was found between changed binding across {comps[[1]]} and {comps[[2]]} when combined with *overall* significance in differentially expressed genes.")
)
```

`r txt`

```{r tbl-gsea_nondir_sig, eval = length(p) > 0, echo = length(p) > 0}
cp <-  htmltools::em(
  glue(
    "Combined windows were mapped to genes, and their position amongst the ",
    "RNA-Seq results was assessed. {nrow(gsea_nondir_sig)} sets of windows were ",
    "associated with changes in gene expression, using only the p-value to ", 
    "rank the {comma(nrow(rnaseq), 1)} genes considered as detected."
  )
)
tbl <- gsea_nondir_sig %>%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %>% 
      vapply(paste, character(1), collapse = "; ")
  ) %>% 
  separate(pathway, comps, sep = " - ") %>% 
  dplyr::select(
    all_of(comps), Windows = size, 
    p = pval, padj, `Edge Size`, `Leading Edge` = leadingEdge
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list2(
      "{comps[[1]]}" := colDef(
        maxWidth = 150,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      "{comps[[2]]}" := colDef(
        maxWidth = 150,
        cell = function(value) {
          html_symbol <- ""
          if (str_detect(value, "Up")) html_symbol <- "\u21E7"
          if (str_detect(value, "Down")) html_symbol <- "\u21E9"
          paste(html_symbol, value)
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            TRUE ~ colours$direction[["unchanged"]]
          )
          list(color = colour)
        }
      ),
      Windows = colDef(maxWidth = 80),
      p = colDef(
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ), 
        maxWidth = 80
      ),
      padj = colDef(
        name = glue("p<sub>{adj_method}</sub>"), html = TRUE,
        cell = function(value) {
          ifelse(
            value < 0.001, sprintf("%.2e", value), sprintf("%.3f", value)
          )
        }
      ),
      "Edge Size" = colDef(maxWidth = 80),
     "Leading Edge" = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```

```{r barcode-gsea-nondir-sig, dev = 'png', eval = length(p) > 0, echo = length(p) > 0, fig.height= max(3, 3 * ceiling(length(p) / 3)), fig.cap = glue("*Barcode plots for the top {length(p)} sets of windows associated with __non-directional__ changes in gene expression.*")}
plot_grid(plotlist = p, ncol = min(3, length(p)))
```


## Integration of RNA and ChIP Enrichment Analyses

Directional GSEA was performed on the set of RNA-Seq results and given that enrichment for differentially bound target is an orthogonal measure, enrichment for a set of potential pathways for each pairwise set of regions was performed.
The p-values from binding enrichment and GSEA were combined using Wilkinson's method for meta-analysis, taking the maximum p-value and adjusting the resulting p-value using the same strategy as earlier.
Any pathways are annotated as being significant in *both, either or neither* dataset when being analysed individually.

Distances between nodes (gene-sets) for any network plots were estimated by using the similarity of genes within the leading edge which were also mapped to binding regions showing the respective pattern.
If the genes of interest from any two pathways were identical, only the most highly-ranked pathway was included in the network, and similarly if one gene-set was a subset of another, the most highly-ranked was retained for visualisation.

```{r cmn-sig}
rnaseq_gsea <- suppressWarnings(
  fgsea(
    pathways = gs_by_gsid,
    stats = rna_dir_ranks, 
    minSize = min_gs_size, 
    maxSize = max_gs_size
  )
) %>% 
  dplyr::filter(!is.na(pval))
status_cols <- setNames(
  hcl.colors(4, "Viridis", rev = TRUE),
  c("Both", "ChIP Only", "RNA Only", "Neither")
)
cmn_res <- goseq_group_res %>% 
  lapply(dplyr::select, group, gs_name, pval_goseq = pval, adj_p, starts_with("num")) %>% 
  lapply(
    left_join, 
    dplyr::select(rnaseq_gsea, gs_name = pathway, NES, pval_gsea = pval, padj, leadingEdge), 
    by = "gs_name"
  ) %>% 
  lapply(
    nest, p = starts_with("pval")
  ) %>% 
  lapply(
    dplyr::rename, padj_chip = adj_p, padj_rna = padj
  ) %>% 
  lapply(
    mutate, 
    pval = vapply(
      p, function(x) metap::maximump(unlist(x))$p, numeric(1)
    ),
    adj_p = p.adjust(pval, adj_method)
  ) %>% 
  lapply(dplyr::filter, adj_p < enrich_alpha) %>% 
  lapply(
    mutate,
    status = case_when(
      padj_chip < enrich_alpha & padj_rna < enrich_alpha ~ "Both",
      !padj_chip < enrich_alpha & padj_rna < enrich_alpha ~ "RNA Only",
      padj_chip < enrich_alpha & !padj_rna < enrich_alpha ~ "ChIP Only",
      !padj_chip < enrich_alpha & !padj_rna < enrich_alpha ~ "Neither"
    ) %>% 
      factor(levels = names(status_cols)),
    leadingEdge = lapply(leadingEdge, intersect, group_ids[[unique(group)]]),
    numDEInCat = vapply(leadingEdge, length, integer(1)),
    genes = vapply(
      leadingEdge, function(x) paste(id2gene[x], collapse = "; "), character(1)
    )
  ) %>% 
  lapply(dplyr::filter, numDEInCat >= min_sig) %>% 
  lapply(
    select,
    group, gs_name, adj_p, NES, status, starts_with("padj"),
    starts_with("num"), leadingEdge, genes
  ) %>% 
  lapply(dplyr::filter, numDEInCat > 0) %>% 
  .[vapply(., nrow, integer(1)) > 0] 
tg_cmn <- names(cmn_res) %>% 
  lapply(
    function(x) {
      make_tbl_graph(
        cmn_res[[x]],
        setNames(cmn_res[[x]]$leadingEdge, cmn_res[[x]]$gs_name)
      )
    }
  ) %>% 
  setNames(names(cmn_res)) %>% 
  .[vapply(., length, integer(1)) >= min_network_size]
any_sig <- length(cmn_res) > 0
plot_net <- length(tg_cmn) > 0
txt <- ifelse(any_sig, "### Results {.tabset}", "No shared enrichment was found")
```

`r txt`

```{r common-result-tables, eval = any_sig, echo = any_sig}
htmltools::tagList(
  names(cmn_res) %>% 
    lapply(
      function(x) {
        df <- cmn_res[[x]] %>% 
          dplyr::select(
            gs_name, adj_p, NES, status, starts_with("padj"), numDEInCat, genes
          )
        tbl <- df %>% 
          arrange(adj_p) %>% 
          reactable(
            filterable = TRUE, showPageSizeOptions = TRUE,
            columns = list(
              gs_name = colDef(
                name = "Gene Set",
                cell = function(value) htmltools::tags$a(
                  href = gs_url[[value]], 
                  target = "_blank", 
                  str_replace_all(value, "_", " ")
                ),
                minWidth = 120
              ),
              NES = colDef(
                format = colFormat(digits = 2),
                style = function(value) {
                  col <- ifelse(value > 0, colours$direction$up, colours$direction$down)
                  list(color = col)
                },
                maxWidth = 80
              ),
              adj_p = colDef(
                name = "p<sub>adj</sub>", html = TRUE,
                cell = function(value) ifelse(
                  value < 0.001,
                  sprintf("%.2e", value),
                  sprintf("%.3f", value)
                ),
                maxWidth = 80
              ),
              status = colDef(name = "Individual Status", maxWidth = 100),
              padj_rna = colDef(
                name =  glue("p<sub>{params$enrichment$adj}</sub> (RNA)"), html = TRUE,
                cell = function(value) ifelse(
                  value < 0.001,
                  sprintf("%.2e", value),
                  sprintf("%.3f", value)
                ),
                maxWidth = 80
              ),
              padj_chip = colDef(
                name =  glue("p<sub>{params$enrichment$adj}</sub> (ChIP)"), html = TRUE,
                cell = function(value) ifelse(
                  value < 0.001,
                  sprintf("%.2e", value),
                  sprintf("%.3f", value)
                ),
                maxWidth = 80
              ),
              numDEInCat = colDef(name = "# Targets in Leading Edge", maxWidth = 100),
              genes = colDef(
                name = "Targets in Leading Edge",
                cell = function(value) with_tooltip(value, width = 50)
              )
            )
          )
        cp <- glue(
          "The {nrow(df)} Gene sets considered significant when combining GSEA 
          from RNA-Seq and enrichment anaylsis from binding sites corresponding 
          to {x}. The direction of the NES indicates the enriched direction in 
          the RNA-Seq data.
        "
        )
        htmltools::div(
          htmltools::div(
            id = x %>% 
              str_to_lower %>% 
              str_replace_all(" ", "-") %>% 
              str_replace_all("-+", "-") %>% 
              paste("gsea", sep = "-"),
            class="section level4",
            htmltools::h4(class = "tabset", x),
            htmltools::tags$em(cp),
            tbl
          )
        )
      }
    ) %>% 
    setNames(names(cmn_res))
)
```

`r if (any_sig) "#### All Groups"`

```{r plot-all-enriched-rna, eval = any_sig, echo = any_sig, fig.width = min(5 + length(cmn_res), knitr::opts_chunk$get("fig.width")), fig.cap = glue("*Combined enrichment across all groups, incorporating RNA expression data, showing only significant results for enrichment. The top panel shows how many detected genes were mapped to sites for each group, whilst the right panel shows gene set size, using detected genes only. Point sizes indicate the proportion of mapped and detected genes which are from each pathway.*")}
cmn_res %>%  
  bind_rows() %>% 
  arrange(adj_p) %>% 
  left_join(n_mapped, by = c("group" = "status")) %>% 
  mutate(
    gs_name = fct_inorder(gs_name) %>% 
      fct_relabel(str_replace_all, "_", " ") %>% 
      fct_relabel(str_trunc, width = 60),
    prop = numDEInCat / detected,
    group = factor(group, levels = arrange(n_mapped, detected)$status)
  ) %>%
  droplevels() %>% 
  ggplot(aes(fct_rev(group), fct_rev(gs_name))) + 
  geom_point(aes(size = prop, fill = -log10(adj_p)), shape = 21) +
  geom_ysidecol(
    aes(x = numInCat), data = . %>% distinct(gs_name, numInCat),
  ) +
  geom_xsidecol(
    aes(y = detected),
    data = n_mapped %>% 
      dplyr::filter(status %in% names(cmn_res)) %>% 
      dplyr::rename(group = status),
    width = 0.5
  ) +
  scale_x_discrete(labels = label_wrap(10)) +
  scale_ysidex_continuous(expand = expansion(c(0, 0.15))) +
  scale_xsidey_continuous(
    expand = expansion(c(0, 0.15))
  ) +
  scale_fill_viridis_c(option = "inferno", begin = 0.2) +
  scale_size(range = c(0, 5), labels = percent) + 
  labs(
    x = c(), y = c(), size = "% Mapped\n& Detected\nGenes", 
    fill = expr(paste(-log[10], p[!!sym(adj_method)]))
  ) +
  theme(
    panel.grid = element_blank(), 
    axis.text = element_text(size = 8),
    legend.position = "right",
    ggside.panel.scale.x = 0.2,
    ggside.panel.scale.y = 0.3,
    ggside.axis.text.x.bottom = element_text(angle = 270, hjust = 0, vjust = 0.5)
  ) 
```


`r if (plot_net) "### Network Plots {.tabset}"`

```{r common-result-networks, eval = plot_net, echo = plot_net}
htmltools::tagList(
  names(tg_cmn) %>% 
    mclapply(
      function(x) {
        ## Export the image
        img_out <- file.path(
          fig_path,
          x %>% 
            str_replace_all(" ", "_") %>% 
            str_replace_all("_-_", "-") %>% 
            paste0("_rnaseq_network.", fig_type)
        )
        fig_fun(
          filename = img_out,
          width = knitr::opts_current$get("fig.width"),
          height = knitr::opts_current$get("fig.height")
        )
        p <- tg_cmn[[x]] %>%
          activate(nodes) %>% 
          mutate(direction = ifelse(NES > 0, "Up", "Down")) %>% 
          ggraph(layout = net_layout, weights = oc^2) +
          geom_edge_link(aes(width = oc^2, alpha = oc^2)) +
          geom_node_point(
            aes(fill = status, size = numDEInCat),
            shape = 21
          ) +
          geom_node_text(
            aes(label = label, colour = direction),
            size = 3, 
            data = . %>%
              mutate(
                label = str_replace_all(label, "_", " ") %>% str_trunc(60) %>% str_wrap(width = 18)
              ),
            repel = TRUE,
            max.overlaps = max(10, round(length(tg_cmn[[x]]) / 4, 0)),
            bg.color = "white", bg.r = 0.1
          ) +
          scale_x_continuous(expand = expansion(c(0.1, 0.1))) +
          scale_y_continuous(expand = expansion(c(0.1, 0.1))) +
          scale_fill_manual(values = status_cols) +
          scale_size_continuous(range = c(1, 10)) +
          scale_edge_width(range = c(1, 6), limits = c(0, 1)) +
          scale_edge_alpha(range = c(0.1, 0.4), limits = c(0, 1)) +
          scale_colour_manual(
            values = setNames(
              unlist(colours$direction), 
              str_to_title(names( unlist(colours$direction)))
            )
          ) +
          guides(edge_alpha= "none") +
          labs(
            size = glue("{x} Targets In Leading Edge") %>% str_wrap(15),
            colour = "GSEA\nDirection",
            fill = "Prior\nStatus",
            edge_width = expr(paste(OC^2))
          ) +
          theme_void()
        print(p)
        dev.off()
        ## Create html tags
        fig_link <- str_extract(img_out, "assets.+")
        cp <- htmltools::tags$em(
          glue(
            "
            Network plot showing enriched pathways mapped to genes 
            associated with {x} in combination with GSEA results from 
            the RNA-Seq data. 
            "
          )
        )
        htmltools::div(
          htmltools::div(
            id = img_out %>%
              basename() %>%
              str_remove_all(glue(".{fig_type}$")) %>%
              str_to_lower() %>%
              str_replace_all("_", "-"),
            class="section level4",
            htmltools::h4(x),
            htmltools::div(
              class = "figure", style = "text-align: center",
              htmltools::img(src = fig_link, width = "100%"),
              htmltools::tags$caption(cp)
            )
          )
        )
      },
      mc.cores = threads
    )
)
```

