# Comparison With RNA Seq

```{r define-objects}
library(fgsea)
id2gene <- structure(genes_gr$gene_name, names = genes_gr$gene_id)
rna_status <- rnaseq %>% 
  mutate(
    de_status = case_when(
      !!sym(rna_fdr_col) < fdr_alpha & !!sym(lfc_col) < 0 ~ "Down",
      !!sym(rna_fdr_col) < fdr_alpha & !!sym(lfc_col) > 0 ~ "Up",
      !!sym(rna_fdr_col) >= fdr_alpha ~ "Unchanged"
    )
  ) %>% 
  dplyr::select(gene_id, gene_name, de_status) %>% 
  bind_rows(
    genes_gr %>% 
      subset(!gene_id %in% rnaseq$gene_id) %>% 
      select(gene_id, gene_name) %>% 
      mcols() %>% 
      as_tibble() %>% 
      mutate(de_status = "Undetected")
  ) %>% 
  mutate(
    de_status = factor(
      de_status, levels = str_to_title(names(colours$direction))
    )
  ) %>% 
  arrange(gene_id)
rna_p_col <- colnames(rnaseq)[
  str_detect(str_to_lower(colnames(rnaseq)), "pval|p\\.val")
][1]
stopifnot(length(rna_p_col) == 1)
diff_windows <- merged_results %>% 
  splitAsList(.$status) %>% 
  lapply(select, gene_id) %>% 
  lapply(mcols) %>% 
  lapply(as_tibble) %>% 
  lapply(pull, "gene_id") %>% 
  lapply(unlist) %>% 
  lapply(unique)
```

## Relationship To Detected Genes {.tabset}

### Detected Genes By Binding Region

Firstly, the distribution of merged windows was compared to genes within the RNA-Seq dataset.
Genes were simply considered as detected if present in the RNA-Seq data, or undetected if not present.
Gene-centric regions were as defined previously.

```{r plot-region-detected, fig.height = 6, fig.cap = "*Distribution of merged windows by region, according to whether the window is mapped to a detected gene in the RNA-Seq dataset.*"}
w <- 0.8
min_p <- 0.05
merged_results %>% 
  mutate(
    any_detected = ifelse(
      vapply(detected, length, integer(1)) > 1,
      "Mapped to Detected Genes",
      "Mapped to Undetected Genes"
    )
  ) %>% 
  plotPie(fill = "region", x = "any_detected", label_size = 5, width = w) +
  geom_text(
    aes(xlab, ylab, label =lab),
    data = . %>% 
      mutate(
        xlab = x + w*(r + 0.1)*sin(label_radians),
        ylab = 1 + w*(r + 0.1)*cos(label_radians),
        lab = glue("{region}\n{percent(p, 0.1)}") %>% 
          str_wrap(10)
      ) %>% 
      dplyr::filter(p > min_p),
    size = 3
  ) +
  geom_segment(
    aes(x_inner, y_inner, xend = x_outer, yend = y_outer),
     data = . %>% 
      mutate(
        x_inner = x + w*r*sin(label_radians),
        y_inner = 1 + w*r*cos(label_radians),
        x_outer = x + w*(r + 0.01)*sin(label_radians),
        y_outer = 1 + w*(r + 0.01)*cos(label_radians),
      ) %>% 
      dplyr::filter(p > min_p)
  ) +
  scale_fill_manual(
    values = colours$regions %>% 
      setNames(regions[names(.)]) 
  ) +
  labs(x = "", fill = "Region") +
  theme(panel.grid = element_blank())
```

`r ifelse(has_features, "### Detected Genes By Feature", "")`

```{r plot-feature-detected, eval = has_features, echo = has_features, fig.height = 6, fig.cap = "*Distribution of merged windows by externally-provided features, according to whether the window is mapped to a detected gene in the RNA-Seq dataset.*"}
merged_results %>% 
  mutate(
    any_detected = ifelse(
      vapply(detected, length, integer(1)) > 1,
      "Mapped to Detected Genes",
      "Mapped to Undetected Genes"
    )
  ) %>% 
  plotPie(fill = "feature", x = "any_detected", label_size = 5, width = w) +
  geom_text(
    aes(xlab, ylab, label =lab),
    data = . %>% 
      mutate(
        xlab = x + w*(r + 0.1)*sin(label_radians),
        ylab = 1 + w*(r + 0.1)*cos(label_radians),
        lab = glue("{feature}\n{percent(p, 0.1)}") %>% 
          str_wrap(10)
      ) %>% 
      dplyr::filter(p > min_p),
    size = 3
  ) +
  geom_segment(
    aes(x_inner, y_inner, xend = x_outer, yend = y_outer),
     data = . %>% 
      mutate(
        x_inner = x + w*r*sin(label_radians),
        y_inner = 1 + w*r*cos(label_radians),
        x_outer = x + w*(r + 0.01)*sin(label_radians),
        y_outer = 1 + w*(r + 0.01)*cos(label_radians),
      ) %>% 
      dplyr::filter(p > min_p)
  ) +
  scale_fill_manual(
    values = colours$features %>% 
      setNames(str_sep_to_title(names(.)))
  ) +
  labs(x = "", fill = "External Feature") +
  theme(panel.grid = element_blank())
```


`r ifelse(has_features, "### Detected Genes By Feature and Binding Region", "")`

```{r plot-region-feature-detected, eval = has_features, echo = has_features, fig.cap = "*Distribution of merged windows mapped to detected/undetected genes separated by gene-centric-region and external fatures.*" }
merged_results %>% 
  mutate(
    any_detected = ifelse(
      vapply(detected, length, integer(1)) > 1,
      "Mapped to Detected Genes",
      "Mapped to Undetected Genes"
    )
  ) %>% 
  plotPie(
    fill = "region", y = "any_detected", x = "feature", 
    label_alpha = 0.7, label_size = 5, width = w
  ) +
  geom_text(
    aes(xlab, ylab, label =lab),
    data = . %>% 
      mutate(
        xlab = x + w*(r + 0.15)*sin(label_radians),
        ylab = y + w*(r + 0.15)*cos(label_radians),
        lab = glue("{region}\n{percent(p, 0.1)}") %>% 
          str_wrap(12)
      ) %>% 
      dplyr::filter(p > 0.2, N > 0.03 * sum(N)),
    size = 3
  ) +
  geom_segment(
    aes(x_inner, y_inner, xend = x_outer, yend = y_outer),
     data = . %>% 
      mutate(
        x_inner = x + w*r*sin(label_radians),
        y_inner = y + w*r*cos(label_radians),
        x_outer = x + w*(r + 0.02)*sin(label_radians),
        y_outer = y + w*(r + 0.02)*cos(label_radians),
      ) %>% 
      dplyr::filter(p > 0.2, N > 0.03 * sum(N))
  ) +
  scale_fill_manual(
    values = colours$regions %>% 
      setNames(regions[names(.)]) 
  ) +
  labs(x = "External Feature", y = "", fill = "Region") +
  theme(panel.grid = element_blank())
```


## Relationship To Differentially Expressed Genes {.tabset}


```{r define_genesets}
rna_dir_ranks <- rnaseq %>% 
  mutate(
    ranking_stat = -sign(!!sym(lfc_col))*log10(!!sym(rna_p_col)) %>% 
      setNames(gene_id)
  ) %>% 
  arrange(ranking_stat) %>% 
  dplyr::filter(!!sym(rna_p_col) < 1) %>% 
  pull("ranking_stat") 
rna_sig_ranks <- rna_dir_ranks %>% 
  abs() %>% 
  sort()
reg_dir_gs <- merged_results %>% 
  select(
    overlaps_ref, status, any_of(c("region", "feature")), starts_with("gene")
  ) %>% 
  as_tibble() %>% 
  mutate(group = glue("{region} / {target} {status}")) %>% 
  split(.$group)%>% 
  lapply(pull, "gene_id") %>% 
  lapply(unlist) %>% 
  lapply(unique) %>% 
  .[vapply(., length, integer(1)) > 20]
if (has_features) {
  new_reg_dir_gs <- merged_results %>% 
    select(
      overlaps_ref, status, any_of(c("region", "feature")), starts_with("gene")
    ) %>% 
    as_tibble() %>% 
    mutate(group = glue("{region} / {feature} / {target} {status}")) %>% 
    split(.$group)%>% 
    lapply(pull, "gene_id") %>% 
    lapply(unlist) %>% 
    lapply(unique) %>% 
    .[vapply(., length, integer(1)) > 20]
  old2keep <- names(reg_dir_gs) %>% 
    str_replace_all("^(.{4}).+(.{4})$", "\\1.+\\2") %>% 
    vapply(function(x) !any(str_detect(names(new_reg_dir_gs), x)), logical(1))
  reg_dir_gs <- c(reg_dir_gs[old2keep], new_reg_dir_gs) 
}
```


### P-Values By `r target` Status

```{r plot-p-target, fig.height = 6, fig.cap = glue("*P-Values for differential expression partitioned by {target} ChIP peak status. Genes were considered as being mapped to a differentially bound peak if one or more peaks was considered as differentially bound.*")}
y <- ggplot_build(
  tibble(p = rnaseq[[rna_p_col]]) %>% 
    ggplot(aes(p, stat(density))) + 
    geom_histogram(bins = 100) 
) %>% .[["data"]] %>% 
  .[[1]] %>% 
  .[["y"]] %>% 
  max()
rnaseq %>% 
  mutate(
    mapped = case_when(
      gene_id %in% unlist(diff_windows[names(diff_windows) %in% c("Up", "Down")]) ~ paste(target, "Differentially Bound"),
      gene_id %in% unlist(diff_windows[names(diff_windows) %in% c("Unchanged")]) ~ paste(target, "Unchanged"),
      TRUE ~ paste("No", target, "Detected")
    )
  ) %>% 
  ggplot(
    aes(!!sym(rna_p_col), stat(density))
  ) +
  geom_histogram(
    colour = "black", fill = "grey", bins = 100
  ) +
  geom_text(
    aes(0.85, y, label = lab),
    data = . %>% 
      group_by(mapped) %>% 
      tally() %>% 
      mutate(lab = glue("N = {comma(n, 1)}"))
  ) +
  facet_wrap(~mapped, nrow = 1)
```


### Directional GSEA

```{r gsea-dir-sig}
gsea_dir <- fgsea(reg_dir_gs, rna_dir_ranks)
gsea_dir_sig <- gsea_dir %>% 
  arrange(pval) %>% 
  dplyr::filter(padj < fdr_alpha) %>% 
  as_tibble() 
p <- gsea_dir_sig %>% 
  dplyr::slice(1:9) %>% 
  pull("pathway") %>% 
  lapply(
    function(x) {
      plotEnrichment(reg_dir_gs[[x]], rna_dir_ranks) +
        ggtitle(x) +
        theme(plot.title = element_text(hjust = 0.5))
    }
  )
```

```{r tbl-gsea-dir_sig, eval = nrow(gsea_dir_sig) > 0, echo = nrow(gsea_dir_sig) > 0}
cp <-  htmltools::em(
  glue(
    "Merged windows were mapped to genes and their position amongst the ",
    "RNA-Seq results was assessed. Windows were classified based on which ",
    "region was the best overlap, and the direction of {target} binding with ",
    "{treat_levels[[2]]} treatment. {nrow(gsea_dir_sig)} sets of windows were ",
    "associated with changes in gene expression, using the sign of ", 
    "fold-change and ranking statistic to initially rank the ", 
    "{comma(nrow(rnaseq), 1)} genes considered as detected."
  )
)
tbl <- gsea_dir_sig %>% 
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %>% 
      vapply(paste, character(1), collapse = "; "),
    Direction = ifelse(NES > 0, "\u21E7 Up-regulated", "\u21E9 Down-regulated")
  ) %>% 
  dplyr::select(
    Description = pathway, `N Windows` = size, Direction, p = pval, FDR = padj, NES, 
    `Edge Size`, `Leading Edge` = leadingEdge
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list(
      Description = colDef(minWidth = 150),
      "N Windows" = colDef(maxWidth = 100),
      Direction = colDef(
        maxWidth = 120,
        style = function(value) {
          colour <- ifelse(
            str_detect(value, "Up"), 
            colours$direction[["up"]], 
            colours$direction[["down"]]
          )
          list(color = colour)
        },
      ),
      p = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 100
      ),
      FDR = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 100
      ),
      NES = colDef(
        cell = function(value) sprintf("%.2f", value), maxWidth = 100
      ),
      "Edge Size" = colDef(maxWidth = 100),
     "Leading Edge" = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```

```{r barcode-gsea-dir-sig, eval = nrow(gsea_dir_sig) > 0, echo = nrow(gsea_dir_sig) > 0, fig.height= 3 * ceiling(length(p) / 3), fig.cap = "*Barcode plots for sets of windows associated with directional changes in gene expression.*"}
plot_grid(plotlist = p, nrow = ceiling(length(p) / 3))
```


### Non-Directional GSEA

```{r gsea-nondir-sig}
gsea_nondir <- fgsea(reg_dir_gs, rna_sig_ranks)
gsea_nondir_sig <- gsea_nondir %>% 
  arrange(pval) %>% 
  dplyr::filter(padj < fdr_alpha) %>% 
  as_tibble() 
p <- gsea_nondir_sig %>% 
  dplyr::slice(1:9) %>% 
  pull("pathway") %>% 
  lapply(
    function(x) {
      plotEnrichment(reg_dir_gs[[x]], rna_sig_ranks) +
        ggtitle(x) +
        theme(plot.title = element_text(hjust = 0.5))
    }
  )
```


```{r tbl-gsea_nondir_sig, eval = nrow(gsea_nondir_sig) > 0, echo = nrow(gsea_nondir_sig) > 0}
cp <-  htmltools::em(
  glue(
    "Merged windows were mapped to genes and their position amongst the ",
    "RNA-Seq results was assessed. Windows were classified based on which ",
    "region was the best overlap, and the direction of {target} binding with ",
    "{treat_levels[[2]]} treatment. {nrow(gsea_nondir_sig)} sets of windows were ",
    "associated with changes in gene expression, using only the p-value to ", 
    "rank the {comma(nrow(rnaseq), 1)} genes considered as detected."
  )
)
tbl <- gsea_nondir_sig %>%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %>% 
      vapply(paste, character(1), collapse = "; "),
  ) %>% 
  dplyr::select(
    Description = pathway, `N Windows` = size, p = pval, FDR = padj, 
    `Edge Size`, `Leading Edge` = leadingEdge
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list(
      Description = colDef(minWidth = 150),
      "N Windows" = colDef(maxWidth = 100),
      p = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 100
      ),
      FDR = colDef(
        cell = function(value) sprintf("%.2e", value), maxWidth = 100
      ),
      "Edge Size" = colDef(maxWidth = 100),
     "Leading Edge" = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```

```{r barcode-gsea-nondir-sig, eval = nrow(gsea_nondir_sig) > 0, echo = nrow(gsea_nondir_sig) > 0, fig.height= 3 * ceiling(length(p) / 3), fig.cap = "*Barcode plots for sets of windows associated with non-directional changes in gene expression.*"}
plot_grid(plotlist = p, nrow = ceiling(length(p) / 3))
```

<!-- Could be done regardless of changes in expression -->

## Highly Ranked Genes {.tabset}

```{r setup-genes-to-plot}
all_ids <- unlist(merged_results$gene_id)
db_ids <- unlist(filter(merged_results, !!sym(fdr_column) < fdr_alpha)$gene_id)
top_diff_win <- rnaseq %>% 
  mutate(
    n_windows = vapply(
      gene_id,
      function(x) sum(str_detect(all_ids, x)),
      integer(1)
    ),
    n_db_windows = vapply(
      gene_id,
      function(x) sum(str_detect(db_ids, x)),
      integer(1)
    )
  ) %>% 
  arrange(!!sym(rna_p_col)) %>% 
  dplyr::filter(n_windows > 1) %>% 
  dplyr::slice(1:5)
grl_to_plot <- top_diff_win$gene_id %>% 
  sapply(
    function(x) {
      subset(
        merged_results, 
        vapply(gene_id, function(id) x %in% id, logical(1))
      ) %>% 
        granges() %>% 
        c(
          granges(subset(genes_gr, gene_id == x))
        ) %>% 
        unstrand()
    },
    simplify = FALSE
  ) %>% 
  lapply(range) %>% 
  GRangesList()
if (!is.null(config$external$coverage)) {
  y_ranges <- granges(unlist(grl_to_plot))
  y_lim <- c(
    y_lim[target],
    bwfl[names(bwfl) != target] %>% 
      lapply(
        function(x) {
          GRangesList(lapply(x, import.bw, which = y_ranges)) %>% 
            unlist() %>% 
            filter(score == max(score)) %>% 
            mcols() %>% 
            .[["score"]] %>% 
            c(0) %>% 
            range()
        }
      )
  )
}
i <- 1
```

Using the ranked list of gens from the RNA-Seq data, the 5 genes most highly ranked for differential expression, with more than one `r target`-bound window mapped to it were selected for visualisation.

```{r make-deg-caption, echo = FALSE}
.makeDEGCaption <- function(id) {
  nm <- id2gene[id]
  df <- dplyr::filter(top_diff_win, gene_id == id)
  fdr <- ifelse(
    df[[rna_fdr_col]] < 0.001,
    sprintf("%.2e", df[[rna_fdr_col]]),
    sprintf("%.3f", df[[rna_fdr_col]])
  )
  glue(
    "*Region showing all merged {target}-bound windows mapped to ", 
    "{nm}. Windows considered to be unchanged for {target} are ", 
    "annotated in grey, with other colours indicating {target} gain or loss.",
    "Gene-centric regions {ifelse(has_features, 'and external features', '')} ",
    "are shown in the top panel. The estimated logFC for {nm} is ",
    sprintf("%.2f ", df$logFC),
    "with an FDR of {fdr}. Undetected and unchanged genes are shown on ",
    "separate tracks.*"
  )
}
```


### `r id2gene[names(grl_to_plot)[i]]`


```{r plot-gene1, fig.cap = .makeDEGCaption(names(grl_to_plot)[i])}
ct <- trans_models %>% 
  subsetByOverlaps(grl_to_plot[[i]]) %>% 
  as_tibble() %>% 
  pull("transcript") %>% 
  unique() %>% 
  length() %>% 
  is_greater_than(10) %>% 
  ifelse("meta", FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic,
  features = feat_gr, featcol = feat_col,
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %>%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
i <- i + 1
```

### `r id2gene[names(grl_to_plot)[i]]`


```{r plot-gene2, fig.cap = .makeDEGCaption(names(grl_to_plot)[i])}
ct <- trans_models %>% 
  subsetByOverlaps(grl_to_plot[[i]]) %>% 
  as_tibble() %>% 
  pull("transcript") %>% 
  unique() %>% 
  length() %>% 
  is_greater_than(10) %>% 
  ifelse("meta", FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic,
  features = feat_gr, featcol = feat_col,
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %>%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
i <- i + 1
```

### `r id2gene[names(grl_to_plot)[i]]`


```{r plot-gene3, fig.cap = .makeDEGCaption(names(grl_to_plot)[i])}
ct <- trans_models %>% 
  subsetByOverlaps(grl_to_plot[[i]]) %>% 
  as_tibble() %>% 
  pull("transcript") %>% 
  unique() %>% 
  length() %>% 
  is_greater_than(10) %>% 
  ifelse("meta", FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic,
  features = feat_gr, featcol = feat_col,
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %>%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
i <- i + 1
```

### `r id2gene[names(grl_to_plot)[i]]`


```{r plot-gene4, fig.cap = .makeDEGCaption(names(grl_to_plot)[i])}
ct <- trans_models %>% 
  subsetByOverlaps(grl_to_plot[[i]]) %>% 
  as_tibble() %>% 
  pull("transcript") %>% 
  unique() %>% 
  length() %>% 
  is_greater_than(10) %>% 
  ifelse("meta", FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic,
  features = feat_gr, featcol = feat_col,
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %>%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
i <- i + 1
```

### `r id2gene[names(grl_to_plot)[i]]`


```{r plot-gene5, fig.cap = .makeDEGCaption(names(grl_to_plot)[i])}
ct <- trans_models %>% 
  subsetByOverlaps(grl_to_plot[[i]]) %>% 
  as_tibble() %>% 
  pull("transcript") %>% 
  unique() %>% 
  length() %>% 
  is_greater_than(10) %>% 
  ifelse("meta", FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic,
  features = feat_gr, featcol = feat_col,
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %>%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  col.title = "black", background.title = "white", showAxis = FALSE,
  rotation.title = 90
)
i <- i + 1
```
