<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Annotations</title>

<script src="site_libs/header-attrs-2.13/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/core-js-2.5.3/shim.min.js"></script>
<script src="site_libs/react-17.0.0/react.min.js"></script>
<script src="site_libs/react-17.0.0/react-dom.min.js"></script>
<script src="site_libs/reactwidget-1.0.0/react-tools.js"></script>
<script src="site_libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<script src="site_libs/reactable-binding-0.2.3/reactable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GRAVI</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Annotations
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="annotation_description.html">Genome and Transcriptome Annotations</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    MACS2 Peak Calling
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="AR_macs2_summary.html">AR</a>
    </li>
    <li>
      <a href="ER_macs2_summary.html">ER</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Differential Binding
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">AR</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="AR_Veh_DHT_differential_binding.html">DHT Vs. Veh</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">ER</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="ER_Veh_DHT_differential_binding.html">DHT Vs. Veh</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pairwise Comparisons
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">AR-ER</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="AR_Veh_DHT_ER_Veh_DHT_pairwise_comparison.html">DHT Vs. Veh / DHT Vs. Veh</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steveped/GRAVI">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Annotations</h1>
<h4 class="date">04 April, 2022</h4>

</div>


<pre class="r"><code>library(tidyverse)
library(magrittr)
library(rtracklayer)
library(glue)
library(pander)
library(scales)
library(yaml)
library(reactable)
library(htmltools)
library(plyranges)
library(GenomicInteractions)</code></pre>
<pre class="r"><code>## Deal with github packages
# BiocManager::install(&quot;steveped/extraChIPs&quot;, ask = FALSE)

stopifnot(library(extraChIPs, logical.return = TRUE))</code></pre>
<pre class="r"><code>panderOptions(&quot;big.mark&quot;, &quot;,&quot;)
panderOptions(&quot;missing&quot;, &quot;&quot;)
panderOptions(&quot;table.split.table&quot;, Inf)
theme_set(
  theme_bw() +
    theme(plot.title = element_text(hjust = 0.5))
)
annotation_path &lt;- here::here(&quot;output&quot;, &quot;annotations&quot;)
all_out &lt;- list(
  chrom_sizes = file.path(annotation_path, &quot;chrom.sizes&quot;),
  gene_regions = file.path(annotation_path, &quot;gene_regions.rds&quot;),
  gtf  = file.path(annotation_path, &quot;all_gr.rds&quot;),
  seqinfo = file.path(annotation_path, &quot;seqinfo.rds&quot;),
  transcript_models = file.path(annotation_path, &quot;trans_models.rds&quot;),
  tss = file.path(annotation_path, &quot;tss.rds&quot;),
  colours = file.path(annotation_path, &quot;colours.rds&quot;)
)</code></pre>
<pre class="r"><code>source(
  here::here(&quot;workflow&quot;, &quot;scripts&quot;, &quot;custom_functions.R&quot;)
)</code></pre>
<pre class="r"><code>config &lt;- read_yaml(here::here(&quot;config&quot;, &quot;config.yml&quot;))
params &lt;- read_yaml(here::here(&quot;config&quot;, &quot;params.yml&quot;))
colours &lt;- read_yaml(here::here(&quot;config&quot;, &quot;colours.yml&quot;))</code></pre>
<pre class="r"><code>treat_levels &lt;- config$comparisons$contrasts %&gt;% 
  unlist() %&gt;% 
  unique()
samples &lt;- here::here(config$samples$file) %&gt;%
  read_tsv() %&gt;% 
  mutate(
    treat = factor(treat, levels = unique(c(treat_levels, treat))),
    target = as.factor(target)
  )
treat_levels &lt;- levels(samples$treat)</code></pre>
<div id="genome-annotations" class="section level1">
<h1>Genome Annotations</h1>
<p>Before analysing any datasets, annotations for the genome and transcriptome were defined.</p>
<div id="genome-description" class="section level2">
<h2>Genome Description</h2>
<pre class="r"><code>sq &lt;- read_rds(file.path(annotation_path, &quot;seqinfo.rds&quot;))</code></pre>
<p>As the foundation for all annotation objects, the description of the underlying genome is required as a <code>Seqinfo</code> object. For this analysis, both the mitochondrial genome and scaffolds were excluded, giving only the autosomes and the X/Y chromosomes. This <code>Seqinfo</code> object is added to all subsequent objects requiring information about the underlying genome, such as a <code>GenomicRanges</code> object.</p>
</div>
<div id="blacklisted-regions" class="section level2 tabset">
<h2 class="tabset">Blacklisted Regions</h2>
<pre class="r"><code>blacklist &lt;- file.path(annotation_path, &quot;blacklist.bed.gz&quot;) %&gt;%
  import.bed(seqinfo = sq) %&gt;%
  sort()</code></pre>
<div id="summary" class="section level3">
<h3>Summary</h3>
<pre class="r"><code>blacklist %&gt;% 
  group_by(name) %&gt;% 
  summarise(
    n = n(),
    p = percent(sum(width)/ sum(seqlengths(sq))),
    kb = sum(width) / 1e3,
    min = min(width) / 1e3,
    median = median(width) / 1e3,
    max = max(width) / 1e3
  ) %&gt;% 
  as_tibble() %&gt;% 
  dplyr::rename_all(str_to_title) %&gt;% 
  dplyr::rename(
    Type = Name, 
    `Nbr Regions` = N,
    `% Genome` = P, 
    `Total (kb)` = Kb,
    `Min (kb)` = Min,
    `Median (kb)` = Median,
    `Max (kb)` = Max
  ) %&gt;% 
  pander(
    caption = glue(
      &quot;*Summary of genomic regions excluded in the blacklist. &quot;,
      &quot;The default set of blacklisted regions was used, as obtained from &quot;,
      &quot;https://github.com/Boyle-Lab/Blacklist (ENCODE)*&quot;
    ),
    justify = &quot;lrrrrrr&quot;
  )</code></pre>
<table>
<caption><em>Summary of genomic regions excluded in the blacklist. The default set of blacklisted regions was used, as obtained from <a href="https://github.com/Boyle-Lab/Blacklist" class="uri">https://github.com/Boyle-Lab/Blacklist</a> (ENCODE)</em></caption>
<colgroup>
<col width="22%" />
<col width="14%" />
<col width="11%" />
<col width="13%" />
<col width="11%" />
<col width="14%" />
<col width="11%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Type</th>
<th align="right">Nbr Regions</th>
<th align="right">% Genome</th>
<th align="right">Total (kb)</th>
<th align="right">Min (kb)</th>
<th align="right">Median (kb)</th>
<th align="right">Max (kb)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">High Signal Region</td>
<td align="right">583</td>
<td align="right">8.7%</td>
<td align="right">269,778</td>
<td align="right">1.1</td>
<td align="right">13.4</td>
<td align="right">30,590</td>
</tr>
<tr class="even">
<td align="left">Low Mappability</td>
<td align="right">251</td>
<td align="right">0.2%</td>
<td align="right">5,192</td>
<td align="right">1.4</td>
<td align="right">5.8</td>
<td align="right">361.7</td>
</tr>
</tbody>
</table>
</div>
<div id="breakdown" class="section level3">
<h3>Breakdown</h3>
<pre class="r"><code>blacklist %&gt;% 
  as_tibble(rangeAsChar = FALSE) %&gt;%
  group_by(seqnames) %&gt;% 
  summarise(blacklisted = sum(width)) %&gt;% 
  left_join(as_tibble(sq), by = &quot;seqnames&quot;) %&gt;% 
  mutate(
    whitelisted = seqlengths - blacklisted,
    seqnames = factor(seqnames, levels = seqlevels(sq))
  ) %&gt;% 
  pivot_longer(
    ends_with(&quot;listed&quot;), names_to = &quot;category&quot;, values_to = &quot;bp&quot;
  ) %&gt;% 
  ggplot(aes(fct_rev(seqnames), bp/1e6, fill = fct_rev(category))) +
  geom_col() +
  coord_flip() +
  scale_y_continuous(expand = expansion(c(0, 0.05))) +
  scale_fill_manual(
    values = c(rgb(0.1, 0.3, 0.9), rgb(0.9, 0.2, 0.1))
  ) +
  labs(
    x = &quot;Chromosome&quot;, y = &quot;Length (Mb)&quot;, fill = &quot;Status&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="annotation_description_files/figure-html/plot-blacklist-1.png" alt="*Breakdown of blacklisted regions by chromosome*" width="960" />
<p class="caption">
<em>Breakdown of blacklisted regions by chromosome</em>
</p>
</div>
</div>
</div>
</div>
<div id="gene-and-transcript-annotations" class="section level1">
<h1>Gene and Transcript Annotations</h1>
<div id="basic-annotations" class="section level2">
<h2>Basic Annotations</h2>
<pre class="r"><code>all_gr &lt;- read_rds(file.path(annotation_path, &quot;all_gr.rds&quot;))</code></pre>
<ul>
<li>The complete set of genes, transcripts and exons was loaded from the supplied <code>gtf</code>, excluding mitochondrial features.</li>
<li>The previously generated <code>Seqinfo</code> was also placed as the foundation of this annotation object, ensuring this propagates through all subsequent objects</li>
<li>Version numbers were removed from all gene, transcript and exon identifiers for convenience, with the minimal set of columns (<em>type</em>, <em>gene_id</em>, <em>gene_type</em>, <em>gene_name</em>, <em>transcript_id</em>, <em>transcript_type</em>, <em>transcript_name</em> and <em>exon_id</em>) retained.</li>
</ul>
<p>In total this object contained annotations for 62,449 genes, 229,655 transcripts and 1,379,777 exons. Visualisation using the Bioconductor package <code>Gviz</code> also requires a specific <code>GRanges</code> structure for gene and transcript models to be displayed. This object was also created so transcript models could be simply visualised.</p>
</div>
<div id="gene-centric-regions" class="section level2 tabset">
<h2 class="tabset">Gene-Centric Regions</h2>
<pre class="r"><code>gene_regions &lt;- read_rds(file.path(annotation_path, &quot;gene_regions.rds&quot;))
regions &lt;- vapply(gene_regions, function(x) unique(x$region), character(1))
missing_reg_col &lt;- setdiff(names(regions), names(colours$regions))
if (length(missing_reg_col) &gt; 0){
  def_reg_cols &lt;- c(
    promoters = &quot;#FF3300&quot;, upstream = &quot;#E1EE05&quot;, exons = &quot;#7EDD57&quot;, 
    introns = &quot;#006600&quot;, proximal = &quot;#000066&quot;, distal = &quot;#551A8B&quot;
  )
  colours$regions[missing_reg_col] &lt;- def_reg_cols[missing_reg_col]
}
region_cols &lt;- unlist(colours$regions) %&gt;% 
  setNames(regions[names(.)])</code></pre>
<p>Using the provided GTF, <strong>unique</strong> gene and transcript-centric features were also defined as:</p>
<ol style="list-style-type: decimal">
<li>Promoters</li>
<li>Upstream</li>
<li>Exons</li>
<li>Introns</li>
<li>Intergenic regions within 10kb of a gene</li>
<li>Intergenic regions &gt;10kb from any defined genes</li>
</ol>
<p>TSS regions were defined as a separate object given each TSS has single-base width. With the exception of TSS and Promoters, these features were non-overlapping and defined in a hierarchical, unstranded manner.</p>
<ul>
<li><em>TSS</em> regions were the individual start sites for each transcript. Given that many transcripts can originate from the same TSS, this is somewhat smaller than the number of actual transcripts defined in the GTF</li>
<li><em>Promoters</em> were first defined for all individual transcripts as the region -1500/+500bp around each TSS. Overlapping regions were merged to create a single promoter region for all transcripts within that genomic region.</li>
<li><em>Upstream promoters</em> were then defined as the region extending 5kb upstream from a TSS, but excluding any region already defined as a promoter through the previous merging step. Any ranges within these which overlapped other genes were removed.</li>
<li><em>Exons</em> and <em>Introns</em> were considered to be any exon or intron, not previously included in one of the above regions.</li>
<li><em>Intergenic</em> regions were defined as those within 10kb (excluding upstream promoters) or those &gt;10kb from any gene</li>
</ul>
<pre class="r"><code>cp &lt;- em(
  glue(
    &quot;
  Summary of gene-centric regions defined as a key annotation set. 
  Colours used throughout the workflow for each region are indicated in the 
  first column, with other summary statistics making up the rest of the table.
  &quot;
  )
)
tbl &lt;- gene_regions %&gt;% 
  lapply(
    function(x){
      tibble(
        n = length(x),
        width = sum(width(x)), 
        region = unique(x$region)
      )
    }
  ) %&gt;%
  bind_rows() %&gt;% 
  mutate(
    width = width / 1e6,
    mn = 1e3*width/n,
    region = fct_inorder(region),
    `% Genome` = width / sum(width),
  ) %&gt;% 
  rename_all(str_to_title) %&gt;% 
  mutate(Guide = region_cols[Region]) %&gt;% 
  dplyr::select(
    Guide, Region, N, Width, Mn, `% Genome`
  ) %&gt;% 
  reactable(
    searchable = FALSE, filterable = FALSE,
    columns = list(
      Guide = colDef(
        maxWidth = 50,
        style = function(value) list(background = value),
        cell = function(value) &quot;&quot;,
        name = &quot;&quot;
      ),
      N = colDef(
        maxWidth = 150,
        cell = function(value) comma(value, 1)
      ),
      Width = colDef(
        name = &quot;Total Width (Mb)&quot;,
        cell = function(value) sprintf(&quot;%.1f&quot;, value)
      ),
      Mn = colDef(
        name = &quot;Average Width (kb)&quot;,
        cell = function(value) sprintf(&quot;%.2f&quot;, value)
      ),
      &quot;% Genome&quot; = colDef(
        cell = function(value) percent(value, 0.1)
      )
    )
  )
div(class = &quot;table&quot;,
  div(class = &quot;table-header&quot;,
      div(class = &quot;caption&quot;, cp),
      tbl
  )
)</code></pre>
<div class="table">
<div class="table-header">
<div class="caption">
<em>Summary of gene-centric regions defined as a key annotation set. 
Colours used throughout the workflow for each region are indicated in the 
first column, with other summary statistics making up the rest of the table.</em>
</div>
<div id="htmlwidget-f30178288a78ea462de4" class="reactable html-widget" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-f30178288a78ea462de4">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"Guide":["#FF3300","#E1EE05","#7EDD57","#006600","#000066","#551A8B"],"Region":["Promoter (-1500/+500bp)","Upstream Promoter (<5kb)","Exon","Intron","Intergenic (<10kb)","Intergenic (>10kb)"],"N":[88840,61232,184051,204758,36482,12646],"Width":[210.742887,183.116491,91.032264,1394.134642,255.426502,961.224626],"Mn":[2.37216216794237,2.9905358472694,0.494603474037088,6.80869437091591,7.00143912066224,76.0101712794559],"% Genome":[0.068076501182934,0.0591523168047718,0.0294062500333933,0.450348811086005,0.0825106973387704,0.310505423554126]},"columns":[{"accessor":"Guide","name":"","type":"character","cell":["","","","","",""],"maxWidth":50,"style":[{"background":"#FF3300"},{"background":"#E1EE05"},{"background":"#7EDD57"},{"background":"#006600"},{"background":"#000066"},{"background":"#551A8B"}]},{"accessor":"Region","name":"Region","type":"factor"},{"accessor":"N","name":"N","type":"numeric","cell":["88,840","61,232","184,051","204,758","36,482","12,646"],"maxWidth":150},{"accessor":"Width","name":"Total Width (Mb)","type":"numeric","cell":["210.7","183.1","91.0","1394.1","255.4","961.2"]},{"accessor":"Mn","name":"Average Width (kb)","type":"numeric","cell":["2.37","2.99","0.49","6.81","7.00","76.01"]},{"accessor":"% Genome","name":"% Genome","type":"numeric","cell":["6.8%","5.9%","2.9%","45.0%","8.3%","31.1%"]}],"defaultPageSize":10,"paginationType":"numbers","showPageInfo":true,"minRows":1,"dataKey":"ea20cadbfe8cd768b5e8093f1523aab3","key":"ea20cadbfe8cd768b5e8093f1523aab3"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<div id="summary-1" class="section level3">
<h3>Summary</h3>
<pre class="r"><code>gene_regions %&gt;% 
  lapply(select, region) %&gt;% 
  GRangesList() %&gt;% 
  unlist() %&gt;% 
  mutate(region = factor(region, levels = regions)) %&gt;% 
  plotPie(
    fill = &quot;region&quot;, scale_by = &quot;width&quot;, show_total = FALSE, 
    category_size = 3.5
  ) +
  scale_fill_manual(values = region_cols) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="annotation_description_files/figure-html/pie-gene-regions-1.png" alt="*Summary of gene-cetric regions using the above workflow and the supplied GTF. Percentages represent the amount of the genome allocated to each region. Blacklisted regions were not considered for this step of the annotation.*" width="768" />
<p class="caption">
<em>Summary of gene-cetric regions using the above workflow and the supplied GTF. Percentages represent the amount of the genome allocated to each region. Blacklisted regions were not considered for this step of the annotation.</em>
</p>
</div>
</div>
<div id="example" class="section level3">
<h3>Example</h3>
<pre class="r"><code>id &lt;- sort(all_gr$gene$gene_id)[[1]]</code></pre>
<pre class="r"><code>gr &lt;- subset(all_gr$gene, gene_id == id) %&gt;% 
  resize(width = width(.) + 2.4e4, fix = &#39;center&#39;) %&gt;% 
  unstrand()
ft &lt;- gene_regions %&gt;% 
  lapply(subsetByOverlaps, gr) %&gt;% 
  lapply(select, region) %&gt;% 
  lapply(intersectMC, gr) %&gt;% 
  GRangesList() %&gt;% 
  unlist() %&gt;% 
  setNames(c()) %&gt;% 
  subset(region != &quot;TSS&quot;) %&gt;% 
  sort()
df &lt;- list(
  all_gr$transcript %&gt;% 
    subsetByOverlaps(gr) %&gt;% 
    as_tibble(rangeAsChar = FALSE),
  all_gr$exon %&gt;% 
    subsetByOverlaps(gr) %&gt;% 
    as_tibble(rangeAsChar = FALSE) 
) %&gt;% 
  bind_rows() %&gt;% 
  mutate(
    transcript_name = as.factor(transcript_name)
  )
df %&gt;% 
  ggplot(aes(x = start, y = as.integer(transcript_name))) +
  geom_rect(
    aes(
      xmin = start, xmax = end,
      ymin =  0, ymax =  Inf,
      fill = region
    ),
    data = ft %&gt;%
      as.data.frame() %&gt;% 
      mutate(region = fct_inorder(region) ),
    inherit.aes = FALSE,
    alpha = 0.6
  ) +
  geom_segment(
    aes(
      x = start, xend = end, 
      y = as.integer(transcript_name),
      yend = as.integer(transcript_name)
    ),
    data = . %&gt;% 
      dplyr::filter(type == &quot;transcript&quot;)
  ) +
  geom_segment(
    aes(
      x = mid, xend = mid_offset, 
      y = as.integer(transcript_name),
      yend = as.integer(transcript_name)
    ),
    data = all_gr$transcript %&gt;% 
      subsetByOverlaps(gr) %&gt;% 
      select(transcript_name) %&gt;% 
      setdiffMC(all_gr$exon) %&gt;%
      as.data.frame() %&gt;% 
      mutate(transcript_name = vctrs::vec_proxy(transcript_name)) %&gt;% 
      unnest(transcript_name) %&gt;% 
      dplyr::filter(width &gt; 600) %&gt;% 
      mutate(
        mid = end - 0.5*width,
        mid_offset = ifelse(strand == &quot;+&quot;, mid + 50, mid - 50),
        transcript_name = factor(transcript_name, levels = levels(df$transcript_name))
      ),
    arrow = arrow(angle = 40, length = unit(0.015, &quot;npc&quot;))
  ) +
  geom_rect(
    aes(
      xmin = start, xmax = end, 
      ymin = as.integer(transcript_name) - 0.2, 
      ymax = as.integer(transcript_name) + 0.2
    ),
    data = . %&gt;% 
      dplyr::filter(type == &quot;exon&quot;),
    fill = &quot;blue&quot;, colour = &quot;blue&quot;
  ) +
  coord_cartesian(xlim = c(start(gr), end(gr))) +
  scale_x_continuous(
    labels = comma, expand = expansion(c(0, 0))
  ) +
  scale_y_continuous(
    breaks = seq_along(levels(df$transcript_name)),
    labels = levels(df$transcript_name),
    expand = expansion(c(-0.05, 0.05))
  ) +
  scale_fill_manual(values = region_cols) + 
  labs(
    x = as.character(seqnames(gr)), y = &quot;Transcript&quot;, fill = &quot;Feature&quot;
  ) +
  theme(
    panel.grid = element_blank()
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="annotation_description_files/figure-html/plot-example-1.png" alt="*12kb region surrounding _TSPAN6_ showing all annotated regions.*" width="960" />
<p class="caption">
<em>12kb region surrounding <em>TSPAN6</em> showing all annotated regions.</em>
</p>
</div>
</div>
</div>
</div>
<div id="external-features" class="section level1">
<h1>External Features</h1>
<pre class="r"><code>feat_path &lt;- here::here(config$external$features)
external_features &lt;- list(features = c())
has_features &lt;- FALSE
if (length(feat_path)&gt; 0) {
  stopifnot(file.exists(feat_path))
  external_features &lt;- suppressWarnings(
    import.gff(feat_path, genome = sq)
  )
  mcols(external_features) &lt;- cbind(
    mcols(external_features),
    gene_regions %&gt;% 
      lapply(function(x) propOverlap(external_features, x)) %&gt;% 
      as(&quot;DataFrame&quot;),
    gene_regions %&gt;% 
      lapply(
        function(x) propOverlap(external_features, subset(x, detected))
      ) %&gt;% 
      as(&quot;DataFrame&quot;) %&gt;% 
      setNames(paste0(names(.), &quot;_detected&quot;))
  )
  keep_cols &lt;- !vapply(
    mcols(external_features), function(x) all(is.na(x)), logical(1)
  )
  mcols(external_features) &lt;- mcols(external_features)[keep_cols]
  has_features &lt;- TRUE
}</code></pre>
<pre class="r"><code>feat_col &lt;- colours$features
feat_levels &lt;- unique(external_features$feature)
missing_feat_col &lt;- setdiff(feat_levels, names(feat_col))
if (length(missing_feat_col) &gt; 0) {
  if (!&quot;no_feature&quot; %in% feat_levels)
    missing_feat_col &lt;- c(missing_feat_col, &quot;no_feature&quot;)
  n &lt;- length(missing_feat_col) 
  feat_col[missing_feat_col] &lt;- hcl.colors(max(9, n), &quot;Spectral&quot;)[seq_len(n)]
}
colours$features &lt;- feat_col</code></pre>
<p>External features were compared to the gene-centric annotations.</p>
<pre class="r"><code>suppressWarnings(
  external_features %&gt;% 
    mutate(w = width) %&gt;% 
    mcols() %&gt;% 
    as_tibble() %&gt;% 
    pivot_longer(
      cols = c(all_of(names(regions)), ends_with(&quot;detected&quot;)),
      names_to = &quot;region&quot;, values_to = &quot;prop&quot;
    ) %&gt;% 
    group_by(feature, region) %&gt;% 
    summarise(ol = sum(prop*w/1e6), .groups = &quot;drop&quot;) %&gt;% 
    mutate(
      detected = str_detect(region, &quot;detected&quot;),
      region = str_remove_all(region, &quot;_detected&quot;)
    ) %&gt;% 
    pivot_wider(names_from = &quot;detected&quot;, values_from = ol) %&gt;% 
    mutate(`FALSE` = `FALSE` - `TRUE`) %&gt;% 
    pivot_longer(
      cols = all_of(c(&quot;TRUE&quot;, &quot;FALSE&quot;)),
      names_to = &quot;detected&quot;,
      values_to = &quot;ol&quot;
    ) %&gt;% 
    left_join(
      external_features %&gt;%
        group_by(feature) %&gt;% 
        summarise(w = sum(width/1e6)) %&gt;%
        as_tibble(),
      by = &quot;feature&quot;
    ) %&gt;% 
    mutate(
      p = ol / w,
      Feature = factor(feature, levels = names(feat_col)) %&gt;% 
        fct_relabel(str_sep_to_title),
      region = factor(regions[region], levels = regions)
    ) %&gt;% 
    ggplot(
      aes(fct_rev(region), p, fill = feature, alpha = detected)
    ) +
    geom_col(colour = &quot;grey80&quot;) +
    geom_label(
      aes(label = percent(p, accuracy = 1)),
      data = . %&gt;% 
        group_by(region, feature, Feature) %&gt;% 
        summarise(p = sum(p), .groups = &quot;drop&quot;),
      alpha = 0.5,
      nudge_y = 0.03
    ) +
    coord_flip() +
    facet_grid(Feature~., scales = &quot;free_y&quot;, space = &quot;free&quot;) +
    scale_y_continuous(labels = percent, expand = expansion(c(0, 0.05))) +
    scale_alpha_discrete(range = c(0.4, 1)) +
    scale_fill_manual(values = feat_col) +
    labs(x = &quot;&quot;, y = &quot;&quot;) +
    theme(legend.position = &quot;none&quot;)
)</code></pre>
<div class="figure" style="text-align: center">
<img src="annotation_description_files/figure-html/plot-feature-by-region-1.png" alt="*The proportion of the ranges provided as external features in the file h3k27ac_features.gtf, and which overlap the gene-centric regions defined above. Regions associated with detected genes are shown as solid colours, whilst those with no information are shown as semi-transparent colours.*" width="960" />
<p class="caption">
<em>The proportion of the ranges provided as external features in the file h3k27ac_features.gtf, and which overlap the gene-centric regions defined above. Regions associated with detected genes are shown as solid colours, whilst those with no information are shown as semi-transparent colours.</em>
</p>
</div>
<pre class="r"><code>cp &lt;- em(
  glue(
      &quot;*Summary of external features provided in the file &quot;, 
      &quot;{basename(config$external$feature)}. All peaks and windows will be &quot;, 
      &quot;compared to these throughout the workflow. The colours defined for each &quot;,
      &quot;feature is show as a guide on the left.*&quot;
    )
)
tbl &lt;- external_features %&gt;% 
  group_by(feature) %&gt;% 
  summarise(
    N = n(),
    Width = sum(width/1e3),
    med = median(width/1e3),
    range = glue(&quot;[{round(min(width/1e3), 1)}, {round(max(width/1e3), 1)}]&quot;)
  ) %&gt;% 
  as_tibble() %&gt;% 
  mutate(
    guide = unlist(feat_col)[feature],
    feature = factor(feature, levels = names(feat_col)) %&gt;%
      fct_relabel(str_sep_to_title)
  ) %&gt;%
  arrange(feature) %&gt;%
  dplyr::select(guide, everything()) %&gt;% 
  reactable(
    filterable = FALSE, searchable = FALSE,
    columns = list(
      guide = colDef(
        maxWidth = 40,
        style = function(value) list(background= value),
        cell = function(value) &quot;&quot;,
        name = &quot;&quot;
      ),
      feature = colDef(
        name = &quot;Feature&quot;,
        footer = htmltools::tags$b(&quot;Total&quot;)
      ),
      N = colDef(
        cell = function(value) comma(value, 1),
        footer = htmltools::tags$b(comma(sum(.$N)))
      ),
      Width = colDef(
        name = &quot;Total Width (kb)&quot;,
        cell = function(value) comma(value, 1),
        footer = htmltools::tags$b(comma(sum(.$Width), 1))
      ),
      med = colDef(
        name = &quot;Median Width (kb)&quot;,
        cell = function(value) sprintf(&quot;%.2f&quot;, value)
      ),
      range = colDef(name = &quot;Size Range (kb)&quot;, align = &quot;right&quot;)
    )
  )
div(class = &quot;table&quot;,
  div(class = &quot;table-header&quot;,
      div(class = &quot;caption&quot;, cp),
      tbl
  )
)</code></pre>
<div class="table">
<div class="table-header">
<div class="caption">
<em>*Summary of external features provided in the file h3k27ac_features.gtf. All peaks and windows will be compared to these throughout the workflow. The colours defined for each feature is show as a guide on the left.*</em>
</div>
<div id="htmlwidget-5d947841e9d791662cfb" class="reactable html-widget" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-5d947841e9d791662cfb">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"guide":["#FF4500","#FFFF00","#FFC34D"],"feature":["Promoter","Enhancer","Super Enhancer"],"N":[14994,32158,2586],"Width":[38662.181,44357.0079999999,12249.917],"med":[1.809,0.891,2.695],"range":["[0.2, 59.8]","[0.2, 33.3]","[0.2, 70.7]"]},"columns":[{"accessor":"guide","name":"","type":"character","cell":["","",""],"maxWidth":40,"style":[{"background":"#FF4500"},{"background":"#FFFF00"},{"background":"#FFC34D"}]},{"accessor":"feature","name":"Feature","type":"factor","footer":{"name":"b","attribs":{},"children":["Total"]}},{"accessor":"N","name":"N","type":"numeric","cell":["14,994","32,158","2,586"],"footer":{"name":"b","attribs":{},"children":["49,738"]}},{"accessor":"Width","name":"Total Width (kb)","type":"numeric","cell":["38,662","44,357","12,250"],"footer":{"name":"b","attribs":{},"children":["95,269"]}},{"accessor":"med","name":"Median Width (kb)","type":"numeric","cell":["1.81","0.89","2.69"]},{"accessor":"range","name":"Size Range (kb)","type":["glue","character"],"align":"right"}],"defaultPageSize":10,"paginationType":"numbers","showPageInfo":true,"minRows":1,"dataKey":"f0377264931356cb4721030a480c0164","key":"f0377264931356cb4721030a480c0164"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</div>
<div id="hic-data" class="section level1 tabset">
<h1 class="tabset">HiC Data</h1>
<pre class="r"><code>hic &lt;- makeGenomicInteractionsFromFile(hic_path, type = &quot;bedpe&quot;)
reg_combs &lt;- expand.grid(regions, regions) %&gt;% 
  as.matrix() %&gt;% 
  apply(
    MARGIN = 1, 
    function(x) {
      x &lt;- sort(factor(x, levels = regions))
      paste(as.character(x), collapse = &quot; - &quot;)
    }
  ) %&gt;% 
  unique()
hic$regions &lt;- anchors(hic) %&gt;% 
  vapply(
    bestOverlap,
    y = GRangesList(lapply(gene_regions, granges)),
    character(length(hic))
  ) %&gt;% 
  apply(MARGIN = 2, function(x) regions[x]) %&gt;% 
  apply(
    MARGIN = 1, 
    function(x) {
      x &lt;- sort(factor(x, levels = regions))
      paste(as.character(x), collapse = &quot; - &quot;)
    }
  ) %&gt;% 
  factor(levels = reg_combs) %&gt;%
  fct_relabel(
    str_replace_all,
    pattern = &quot;Promoter \\([0-9kbp/\\+-]+\\)&quot;, replacement = &quot;Promoter&quot;
  ) </code></pre>
<p>HiC data with 1,339 interaction pairs was provided and loaded.</p>
<div id="genomic-distribution-of-interaction-pairs" class="section level2">
<h2>Genomic Distribution of Interaction Pairs</h2>
<pre class="r"><code>hic %&gt;% 
  as_tibble() %&gt;% 
  mutate(
    regions = fct_lump_prop(regions, prop = 0.01)
  ) %&gt;% 
  plotPie(fill = &quot;regions&quot;, min_p = 0.05, category_width = 20, label_size = 4) +
  labs(fill = &quot;Interacting\n Regions&quot;) +
  scale_fill_viridis_d(direction = -1)</code></pre>
<div class="figure" style="text-align: center">
<img src="annotation_description_files/figure-html/plot-pie-hic-1.png" alt="*Distribution of interaction pairs using pre-defined genomic regions. Region pairs with an interaction frequency &lt; 1% were grouped into the 'Other' category*" width="960" />
<p class="caption">
<em>Distribution of interaction pairs using pre-defined genomic regions. Region pairs with an interaction frequency &lt; 1% were grouped into the ‘Other’ category</em>
</p>
</div>
</div>
<div id="feature-distribution-of-interaction-pairs" class="section level2">
<h2>Feature Distribution of Interaction Pairs</h2>
<pre class="r"><code>feat_combs &lt;- expand.grid(names(feat_col), names(feat_col)) %&gt;% 
  as.matrix() %&gt;% 
  apply(
    MARGIN = 1, 
    function(x) {
      x &lt;- sort(factor(x, levels = names(feat_col)))
      paste(as.character(x), collapse = &quot; - &quot;)
    }
  ) %&gt;% 
  unique()
hic$features &lt;- vapply(
  anchors(hic),
  function(x) bestOverlap(
    x, external_features, var = &quot;feature&quot;, missing = &quot;no_feature&quot;
  ),
  character(length(hic))
)  %&gt;% 
  apply(
    MARGIN = 1, 
    function(x) {
      x &lt;- sort(factor(x, levels = names(feat_col)))
      paste(as.character(x), collapse = &quot; - &quot;)
    }
  ) %&gt;% 
  factor(levels = feat_combs) %&gt;%
  fct_relabel(str_sep_to_title, pattern = &quot;_&quot;)
hic %&gt;% 
  as_tibble() %&gt;% 
  mutate(
    features = fct_lump_prop(features, prop = 0.01)
  ) %&gt;% 
  plotPie(fill = &quot;features&quot;, min_p = 0.02, category_width = 15, label_size = 4) +
  labs(fill = &quot;Interacting\n Features&quot;) +
  scale_fill_viridis_d(direction = -1)</code></pre>
<div class="figure" style="text-align: center">
<img src="annotation_description_files/figure-html/plot-pie-features-1.png" alt="*Distribution of HiC interaction pairs using the features provided in h3k27ac_features.gtf. Feature pairs with an interaction frequency &lt; 1% were grouped into the 'Other' categor*" width="960" />
<p class="caption">
<em>Distribution of HiC interaction pairs using the features provided in h3k27ac_features.gtf. Feature pairs with an interaction frequency &lt; 1% were grouped into the ‘Other’ categor</em>
</p>
</div>
</div>
</div>
<div id="colour-schemes" class="section level1 tabset">
<h1 class="tabset">Colour Schemes</h1>
<pre class="r"><code>## qc_colours need to have `Pass` and `Fail`
missing_qc_cols &lt;- setdiff(c(&quot;pass&quot;, &quot;fail&quot;), names(colours$qc))
if (&quot;pass&quot; %in% missing_qc_cols) colours$qc$pass &lt;- &quot;#0571B0&quot; # Blue
if (&quot;fail&quot; %in% missing_qc_cols) colours$qc$fail &lt;- &quot;#CA0020&quot; # Red
colours$qc &lt;- colours$qc[c(&quot;pass&quot;, &quot;fail&quot;)]

## The colours specified as treat_colours should contain all treat_levels + Input
## If Input is missing, set to #33333380 (&#39;grey20&#39; + alpha = 50)
## This should be a standard chunk for all workflows
missing_treat_cols &lt;- setdiff(
  c(&quot;Input&quot;, treat_levels), names(colours$treat)
)
if (length(missing_treat_cols) &gt; 0) {
  if (&quot;Input&quot; %in% missing_treat_cols) 
    colours$treat$Input &lt;- &quot;#33333380&quot;
  ## Automatically sample from the viridis palette if no colour is assigned
  colours$treat[setdiff(missing_treat_cols, &quot;Input&quot;)] &lt;- hcl.colors(
    length(setdiff(missing_treat_cols, &quot;Input&quot;))
  )
}

## Direction colours always need up, down, unchanged &amp; undetected
missing_dir_cols &lt;- setdiff(
  c(&quot;up&quot;, &quot;down&quot;, &quot;unchanged&quot;, &quot;undetected&quot;), names(colours$direction)
)
if (length(missing_dir_cols) &gt; 0){
  def_dir_cols &lt;- c(
    up = &quot;#CA0020&quot;, down = &quot;#0571B0&quot;, 
    unchanged = &quot;#7F7F7F&quot;, undetected = &quot;#E5E5E5&quot;
  )
  colours$direction[missing_dir_cols] &lt;- def_dir_cols[missing_dir_cols]
}
write_rds(colours, all_out$colours, compress = &quot;gz&quot;)</code></pre>
<p>Colours were checked where provided and any missing colours were automatically assigned. These colour schemes are shown below and will be propagated through all steps of the workflow. To change any colours, simply add them to <code>config/rmarkdown.yml</code>.</p>
<div id="qc" class="section level2">
<h2>QC</h2>
<pre class="r"><code>.plotScheme(colours$qc , xlab = &quot;QC Category&quot;) </code></pre>
<p><img src="annotation_description_files/figure-html/plot-qc-1.png" width="384" style="display: block; margin: auto;" /></p>
</div>
<div id="treatment-groups" class="section level2">
<h2>Treatment Groups</h2>
<pre class="r"><code>.plotScheme(colours$treat, xlab = &quot;Treatment&quot;)</code></pre>
<p><img src="annotation_description_files/figure-html/plot-treat-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
<div id="regions" class="section level2">
<h2>Regions</h2>
<pre class="r"><code>.plotScheme(colours$regions, xlab = &quot;Regions&quot;)</code></pre>
<p><img src="annotation_description_files/figure-html/plot-regions-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="features" class="section level2">
<h2>Features</h2>
<pre class="r"><code>.plotScheme(feat_col, xlab = &quot;Feature&quot;)</code></pre>
<p><img src="annotation_description_files/figure-html/plot-feat-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div id="direction" class="section level2">
<h2>Direction</h2>
<pre class="r"><code>.plotScheme(colours$direction, xlab = &quot;Direction&quot;)</code></pre>
<p><img src="annotation_description_files/figure-html/plot-dir-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="data-export" class="section level1">
<h1>Data Export</h1>
<p>During the setup of all required annotations, the following files were exported:</p>
<ul>
<li><strong>chrom_sizes</strong>: GRAVI/output/annotations/chrom.sizes</li>
<li><strong>gene_regions</strong>: GRAVI/output/annotations/gene_regions.rds</li>
<li><strong>gtf</strong>: GRAVI/output/annotations/all_gr.rds</li>
<li><strong>seqinfo</strong>: GRAVI/output/annotations/seqinfo.rds</li>
<li><strong>transcript_models</strong>: GRAVI/output/annotations/trans_models.rds</li>
<li><strong>tss</strong>: GRAVI/output/annotations/tss.rds</li>
<li><strong>colours</strong>: GRAVI/output/annotations/colours.rds</li>
</ul>
<!-- end of list -->
<button type="button" class="btn btn-default btn-sessioninfo" data-toggle="collapse" data-target="#sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Session information
</button>
</p>
<div id="sessioninfo" class="collapse">
<p><strong>R version 4.1.1 (2021-08-10)</strong></p>
<p><strong>Platform:</strong> x86_64-conda-linux-gnu (64-bit)</p>
<p><strong>locale:</strong> <em>LC_CTYPE=en_AU.UTF-8</em>, <em>LC_NUMERIC=C</em>, <em>LC_TIME=en_AU.UTF-8</em>, <em>LC_COLLATE=en_AU.UTF-8</em>, <em>LC_MONETARY=en_AU.UTF-8</em>, <em>LC_MESSAGES=en_AU.UTF-8</em>, <em>LC_PAPER=en_AU.UTF-8</em>, <em>LC_NAME=C</em>, <em>LC_ADDRESS=C</em>, <em>LC_TELEPHONE=C</em>, <em>LC_MEASUREMENT=en_AU.UTF-8</em> and <em>LC_IDENTIFICATION=C</em></p>
<p><strong>attached base packages:</strong> <em>stats4</em>, <em>stats</em>, <em>graphics</em>, <em>grDevices</em>, <em>utils</em>, <em>datasets</em>, <em>methods</em> and <em>base</em></p>
<p><strong>other attached packages:</strong> <em>extraChIPs(v.0.99.0)</em>, <em>GenomicInteractions(v.1.28.0)</em>, <em>InteractionSet(v.1.22.0)</em>, <em>SummarizedExperiment(v.1.24.0)</em>, <em>Biobase(v.2.54.0)</em>, <em>MatrixGenerics(v.1.6.0)</em>, <em>matrixStats(v.0.61.0)</em>, <em>plyranges(v.1.14.0)</em>, <em>htmltools(v.0.5.2)</em>, <em>reactable(v.0.2.3)</em>, <em>yaml(v.2.3.5)</em>, <em>scales(v.1.1.1)</em>, <em>pander(v.0.6.5)</em>, <em>glue(v.1.6.2)</em>, <em>rtracklayer(v.1.54.0)</em>, <em>GenomicRanges(v.1.46.1)</em>, <em>GenomeInfoDb(v.1.30.1)</em>, <em>IRanges(v.2.28.0)</em>, <em>S4Vectors(v.0.32.4)</em>, <em>BiocGenerics(v.0.40.0)</em>, <em>magrittr(v.2.0.3)</em>, <em>forcats(v.0.5.1)</em>, <em>stringr(v.1.4.0)</em>, <em>dplyr(v.1.0.8)</em>, <em>purrr(v.0.3.4)</em>, <em>readr(v.2.1.2)</em>, <em>tidyr(v.1.2.0)</em>, <em>tibble(v.3.1.6)</em>, <em>ggplot2(v.3.3.5)</em> and <em>tidyverse(v.1.3.1)</em></p>
<p><strong>loaded via a namespace (and not attached):</strong> <em>utf8(v.1.2.2)</em>, <em>tidyselect(v.1.1.2)</em>, <em>RSQLite(v.2.2.12)</em>, <em>AnnotationDbi(v.1.56.2)</em>, <em>htmlwidgets(v.1.5.4)</em>, <em>grid(v.4.1.1)</em>, <em>BiocParallel(v.1.28.3)</em>, <em>scatterpie(v.0.1.7)</em>, <em>munsell(v.0.5.0)</em>, <em>codetools(v.0.2-18)</em>, <em>withr(v.2.5.0)</em>, <em>colorspace(v.2.0-3)</em>, <em>filelock(v.1.0.2)</em>, <em>highr(v.0.9)</em>, <em>knitr(v.1.38)</em>, <em>rstudioapi(v.0.13)</em>, <em>ggside(v.0.2.0)</em>, <em>labeling(v.0.4.2)</em>, <em>GenomeInfoDbData(v.1.2.7)</em>, <em>polyclip(v.1.10-0)</em>, <em>farver(v.2.1.0)</em>, <em>bit64(v.4.0.5)</em>, <em>rprojroot(v.2.0.3)</em>, <em>vctrs(v.0.4.0)</em>, <em>generics(v.0.1.2)</em>, <em>xfun(v.0.30)</em>, <em>biovizBase(v.1.42.0)</em>, <em>csaw(v.1.28.0)</em>, <em>BiocFileCache(v.2.2.1)</em>, <em>R6(v.2.5.1)</em>, <em>doParallel(v.1.0.17)</em>, <em>clue(v.0.3-60)</em>, <em>locfit(v.1.5-9.5)</em>, <em>AnnotationFilter(v.1.18.0)</em>, <em>bitops(v.1.0-7)</em>, <em>cachem(v.1.0.6)</em>, <em>DelayedArray(v.0.20.0)</em>, <em>assertthat(v.0.2.1)</em>, <em>vroom(v.1.5.7)</em>, <em>BiocIO(v.1.4.0)</em>, <em>nnet(v.7.3-17)</em>, <em>gtable(v.0.3.0)</em>, <em>ensembldb(v.2.18.4)</em>, <em>rlang(v.1.0.2)</em>, <em>GlobalOptions(v.0.1.2)</em>, <em>splines(v.4.1.1)</em>, <em>lazyeval(v.0.2.2)</em>, <em>dichromat(v.2.0-0)</em>, <em>broom(v.0.7.12)</em>, <em>checkmate(v.2.0.0)</em>, <em>modelr(v.0.1.8)</em>, <em>crosstalk(v.1.2.0)</em>, <em>GenomicFeatures(v.1.46.5)</em>, <em>backports(v.1.4.1)</em>, <em>Hmisc(v.4.6-0)</em>, <em>EnrichedHeatmap(v.1.24.0)</em>, <em>tools(v.4.1.1)</em>, <em>ellipsis(v.0.3.2)</em>, <em>jquerylib(v.0.1.4)</em>, <em>RColorBrewer(v.1.1-2)</em>, <em>Rcpp(v.1.0.8.3)</em>, <em>base64enc(v.0.1-3)</em>, <em>progress(v.1.2.2)</em>, <em>zlibbioc(v.1.40.0)</em>, <em>RCurl(v.1.98-1.6)</em>, <em>prettyunits(v.1.1.1)</em>, <em>rpart(v.4.1.16)</em>, <em>GetoptLong(v.1.0.5)</em>, <em>reactR(v.0.4.4)</em>, <em>haven(v.2.4.3)</em>, <em>ggrepel(v.0.9.1)</em>, <em>cluster(v.2.1.3)</em>, <em>fs(v.1.5.2)</em>, <em>here(v.1.0.1)</em>, <em>data.table(v.1.14.2)</em>, <em>circlize(v.0.4.14)</em>, <em>reprex(v.2.0.1)</em>, <em>ProtGenerics(v.1.26.0)</em>, <em>hms(v.1.1.1)</em>, <em>evaluate(v.0.15)</em>, <em>XML(v.3.99-0.9)</em>, <em>jpeg(v.0.1-9)</em>, <em>readxl(v.1.4.0)</em>, <em>gridExtra(v.2.3)</em>, <em>shape(v.1.4.6)</em>, <em>compiler(v.4.1.1)</em>, <em>biomaRt(v.2.50.3)</em>, <em>crayon(v.1.5.1)</em>, <em>ggfun(v.0.0.6)</em>, <em>tzdb(v.0.3.0)</em>, <em>Formula(v.1.2-4)</em>, <em>lubridate(v.1.8.0)</em>, <em>DBI(v.1.1.2)</em>, <em>tweenr(v.1.0.2)</em>, <em>dbplyr(v.2.1.1)</em>, <em>ComplexHeatmap(v.2.10.0)</em>, <em>MASS(v.7.3-56)</em>, <em>rappdirs(v.0.3.3)</em>, <em>Matrix(v.1.4-1)</em>, <em>cli(v.3.2.0)</em>, <em>parallel(v.4.1.1)</em>, <em>Gviz(v.1.38.3)</em>, <em>metapod(v.1.2.0)</em>, <em>igraph(v.1.3.0)</em>, <em>pkgconfig(v.2.0.3)</em>, <em>GenomicAlignments(v.1.30.0)</em>, <em>foreign(v.0.8-82)</em>, <em>xml2(v.1.3.3)</em>, <em>foreach(v.1.5.2)</em>, <em>bslib(v.0.3.1)</em>, <em>XVector(v.0.34.0)</em>, <em>rvest(v.1.0.2)</em>, <em>VariantAnnotation(v.1.40.0)</em>, <em>digest(v.0.6.29)</em>, <em>Biostrings(v.2.62.0)</em>, <em>rmarkdown(v.2.13)</em>, <em>cellranger(v.1.1.0)</em>, <em>htmlTable(v.2.4.0)</em>, <em>edgeR(v.3.36.0)</em>, <em>restfulr(v.0.0.13)</em>, <em>curl(v.4.3.2)</em>, <em>Rsamtools(v.2.10.0)</em>, <em>rjson(v.0.2.21)</em>, <em>lifecycle(v.1.0.1)</em>, <em>jsonlite(v.1.8.0)</em>, <em>viridisLite(v.0.4.0)</em>, <em>limma(v.3.50.1)</em>, <em>BSgenome(v.1.62.0)</em>, <em>fansi(v.1.0.3)</em>, <em>pillar(v.1.7.0)</em>, <em>lattice(v.0.20-45)</em>, <em>KEGGREST(v.1.34.0)</em>, <em>fastmap(v.1.1.0)</em>, <em>httr(v.1.4.2)</em>, <em>survival(v.3.3-1)</em>, <em>png(v.0.1-7)</em>, <em>iterators(v.1.0.14)</em>, <em>bit(v.4.0.4)</em>, <em>ggforce(v.0.3.3)</em>, <em>stringi(v.1.7.6)</em>, <em>sass(v.0.4.1)</em>, <em>blob(v.1.2.2)</em>, <em>latticeExtra(v.0.6-29)</em> and <em>memoise(v.2.0.1)</em></p>
</div>
</div>

<div class="footer">
  Workflow developed by:<br>
  Dr Stephen Pederson<br>
  Dame Roma Mitchell Cancer Research Laboratories<br>
  Adelaide Medical School, University of Adelaide, SA, Australia
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
