<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>ER: MACS2 Summary</title>

<script src="site_libs/header-attrs-2.13/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GRAVI</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Annotations
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="annotation_description.html">Genome and Transcriptome Annotations</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    MACS2 Peak Calling
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="AR_macs2_summary.html">AR</a>
    </li>
    <li>
      <a href="ER_macs2_summary.html">ER</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Differential Binding
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">AR</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="AR_Veh_DHT_differential_binding.html">DHT Vs. Veh</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">ER</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="ER_Veh_DHT_differential_binding.html">DHT Vs. Veh</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pairwise Comparisons
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">AR-ER</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="AR_Veh_DHT_ER_Veh_DHT_pairwise_comparison.html">DHT Vs. Veh / DHT Vs. Veh</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steveped/GRAVI">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">ER: MACS2 Summary</h1>
<h4 class="date">04 April, 2022</h4>

</div>


<pre class="r"><code>target &lt;- &quot;ER&quot;
threads &lt;- 6</code></pre>
<pre class="r"><code>library(tidyverse)
library(magrittr)
library(rtracklayer)
library(glue)
library(pander)
library(scales)
library(plyranges)
library(yaml)
library(ngsReports)
library(UpSetR)
library(VennDiagram)
library(cowplot)
library(rlang)
library(BiocParallel)
library(Rsamtools)
library(Biostrings)
library(ggside)</code></pre>
<pre class="r"><code>## Deal with github packages
if (!&quot;extraChIPs&quot; %in% rownames(installed.packages()))
  BiocManager::install(&quot;steveped/extraChIPs&quot;, ask = FALSE)
stopifnot(library(extraChIPs, logical.return = TRUE))</code></pre>
<pre class="r"><code>panderOptions(&quot;big.mark&quot;, &quot;,&quot;)
panderOptions(&quot;missing&quot;, &quot;&quot;)
panderOptions(&quot;table.split.table&quot;, Inf)
theme_set(
  theme_bw() +
    theme(
      plot.title = element_text(hjust = 0.5)
    )
)
register(MulticoreParam(workers = threads))
source(here::here(&quot;workflow/scripts/custom_functions.R&quot;))
## This has been pushed to the main repo of ggside, but is not yet
## incorporated into the packge. The functions are geom_*sidelabel
source(here::here(&quot;workflow/scripts/geom_sidelabel.R&quot;))</code></pre>
<pre class="r"><code>config &lt;- read_yaml(
  here::here(&quot;config&quot;, &quot;config.yml&quot;)
)
bam_path &lt;- here::here(config$paths$bam, target)
stopifnot(dir.exists(bam_path))
macs2_path &lt;- here::here(config$paths$macs2, target)
stopifnot(dir.exists(macs2_path))
out_path &lt;- here::here(&quot;output&quot;, target)
if (!dir.exists(out_path)) dir.create(out_path)
annotation_path &lt;- here::here(&quot;output&quot;, &quot;annotations&quot;)
colours &lt;- read_rds(
  file.path(annotation_path, &quot;colours.rds&quot;)
) %&gt;% 
  lapply(unlist)</code></pre>
<pre class="r"><code>samples &lt;-read_tsv(file.path(out_path, &quot;qc_samples.tsv&quot;))
treat_levels &lt;- unique(samples$treat)
if (!is.null(config$comparisons$contrasts)) {
  ## Ensure levels respect those provided in contrasts
  treat_levels &lt;- config$comparisons$contrasts %&gt;%
    unlist() %&gt;%
    intersect(samples$treat) %&gt;% 
    unique()
}
rep_col &lt;- setdiff(
  colnames(samples), c(&quot;sample&quot;, &quot;treat&quot;, &quot;target&quot;, &quot;input&quot;, &quot;label&quot;, &quot;qc&quot;)
)
samples &lt;- samples %&gt;%
  unite(label, treat, !!sym(rep_col), remove = FALSE) %&gt;%
  mutate(
    treat = factor(treat, levels = treat_levels),
    &quot;{rep_col}&quot; := as.factor(!!sym(rep_col))
  )</code></pre>
<pre class="r"><code>sq &lt;- read_rds(file.path(annotation_path, &quot;seqinfo.rds&quot;))
blacklist &lt;-  file.path(annotation_path, &quot;blacklist.bed.gz&quot;) %&gt;%
  import.bed(seqinfo = sq) %&gt;%
  sort()
all_gr &lt;- file.path(annotation_path, &quot;all_gr.rds&quot;) %&gt;% 
  read_rds()
external_features &lt;- c()
has_features &lt;- FALSE
if (!is.null(config$external$features)) {
  external_features &lt;- suppressWarnings(
    import.gff(here::here(config$external$features), genome = sq)
  )
  keep_cols &lt;- !vapply(
    mcols(external_features), function(x) all(is.na(x)), logical(1)
  )
  mcols(external_features) &lt;- mcols(external_features)[keep_cols]
  has_features &lt;- TRUE
}
gene_regions &lt;- read_rds(file.path(annotation_path, &quot;gene_regions.rds&quot;))
regions &lt;- vapply(gene_regions, function(x) unique(x$region), character(1))
any_detected &lt;- gene_regions %&gt;% 
  vapply(function(x) any(x$detected), logical(1)) %&gt;% 
  any()
tss &lt;- read_rds(file.path(annotation_path, &quot;tss.rds&quot;))</code></pre>
<pre class="r"><code>bfl &lt;- bam_path %&gt;%
  file.path(glue(&quot;{samples$sample}.bam&quot;)) %&gt;%
  BamFileList() %&gt;%
  setNames(samples$sample)</code></pre>
<pre class="r"><code>individual_peaks &lt;- file.path(
  macs2_path, glue(&quot;{samples$sample}_peaks.narrowPeak&quot;)
) %&gt;%
  importPeaks(seqinfo = sq, blacklist = blacklist) %&gt;%
  setNames(samples$sample)</code></pre>
<pre class="r"><code>macs2_logs &lt;- file.path(macs2_path, glue(&quot;{samples$sample}_callpeak.log&quot;)) %&gt;%
  importNgsLogs() %&gt;%
  dplyr::select(
    -contains(&quot;file&quot;), -outputs, -n_reads, -alt_fragment_length
  ) %&gt;%
  left_join(samples, by = c(&quot;name&quot; = &quot;sample&quot;)) %&gt;%
  mutate(
    total_peaks = map_int(
      name,
      function(x) {
        length(individual_peaks[[x]])
      }
    )
  ) 
n_reps &lt;- macs2_logs %&gt;% 
  group_by(treat) %&gt;%
  summarise(n = sum(qc == &quot;pass&quot;))</code></pre>
<div id="qc" class="section level2 tabset">
<h2 class="tabset">QC</h2>
<pre class="r"><code>emphasize.italics.rows(NULL)
any_fail &lt;- any(macs2_logs$qc == &quot;fail&quot;)
if (any_fail) emphasize.italics.rows(which(macs2_logs$qc == &quot;fail&quot;))
macs2_logs %&gt;%
  dplyr::select(
    sample = name, label,
    total_peaks, 
    reads = n_tags_treatment, read_length = tag_length,
    fragment_length
  ) %&gt;%
  rename_all(str_sep_to_title )%&gt;%
  pander(
    justify = &quot;llrrrr&quot;,
    caption = glue(
      &quot;*Summary of results for `macs2 callpeak` on individual {target} samples.&quot;,
      &quot;Total peaks indicates the number retained after applying the FDR &quot;, 
      &quot;threshold of {percent(config$peaks$macs2$fdr)} during the peak calling &quot;, 
      &quot;process.&quot;,
      ifelse(
        any_fail,
        glue(
          &quot;Samples shown in italics were marked (F) for exclusion from &quot;,
          &quot;downstream analysis as they identified fewer than &quot;,
          &quot;{percent(qc_prop)} of the peaks identified in the sample returning &quot;,
          &quot;the maximum number of peaks.&quot;
        ),
        glue(
          &quot;No samples were identified as failing QC based on the number of &quot;,
          &quot;peaks identified relative to the highest quality sample.&quot;
        )
      ),
      &quot;Any peaks passing the FDR cutoff, but which overlapped any black-listed&quot;,
      &quot;regions were additionally excluded. The fragment length as estimated by&quot;,
      &quot;`macs2 predictd` is given in the final column.&quot;,
      case_when(
        all(macs2_logs$paired_end) ~
          &quot;All input files contained paired-end reads.*&quot;,
        all(!macs2_logs$paired_end) ~
          &quot;All input files contained single-end reads.*&quot;,
        TRUE ~
          &quot;Input files were a mixture of paired and single-end reads*&quot;
      ),
      .sep = &quot; &quot;
    )
  )</code></pre>
<table style="width:100%;">
<caption><em>Summary of results for <code>macs2 callpeak</code> on individual ER samples. Total peaks indicates the number retained after applying the FDR threshold of 5% during the peak calling process. No samples were identified as failing QC based on the number of peaks identified relative to the highest quality sample. Any peaks passing the FDR cutoff, but which overlapped any black-listed regions were additionally excluded. The fragment length as estimated by <code>macs2 predictd</code> is given in the final column. All input files contained single-end reads.</em></caption>
<colgroup>
<col width="14%" />
<col width="10%" />
<col width="18%" />
<col width="15%" />
<col width="18%" />
<col width="23%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Sample</th>
<th align="left">Label</th>
<th align="right">Total Peaks</th>
<th align="right">Reads</th>
<th align="right">Read Length</th>
<th align="right">Fragment Length</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ER_DHT_1</td>
<td align="left">DHT_1</td>
<td align="right">416</td>
<td align="right">971,954</td>
<td align="right">75</td>
<td align="right">299</td>
</tr>
<tr class="even">
<td align="left">ER_DHT_2</td>
<td align="left">DHT_2</td>
<td align="right">71</td>
<td align="right">1,405,009</td>
<td align="right">75</td>
<td align="right">298</td>
</tr>
<tr class="odd">
<td align="left">ER_DHT_3</td>
<td align="left">DHT_3</td>
<td align="right">139</td>
<td align="right">1,297,230</td>
<td align="right">76</td>
<td align="right">298</td>
</tr>
<tr class="even">
<td align="left">ER_Veh_1</td>
<td align="left">Veh_1</td>
<td align="right">1,250</td>
<td align="right">997,408</td>
<td align="right">75</td>
<td align="right">299</td>
</tr>
<tr class="odd">
<td align="left">ER_Veh_2</td>
<td align="left">Veh_2</td>
<td align="right">1,657</td>
<td align="right">958,926</td>
<td align="right">75</td>
<td align="right">299</td>
</tr>
<tr class="even">
<td align="left">ER_Veh_3</td>
<td align="left">Veh_3</td>
<td align="right">1,116</td>
<td align="right">1,074,865</td>
<td align="right">72</td>
<td align="right">298</td>
</tr>
</tbody>
</table>
<div id="library-sizes" class="section level3">
<h3>Library Sizes</h3>
<pre class="r"><code>macs2_logs %&gt;%
  ggplot(
    aes(label, n_tags_treatment, fill = qc)
  ) +
  geom_col(position = &quot;dodge&quot;) +
  geom_hline(
    aes(yintercept = mn),
    data = . %&gt;%
      group_by(treat) %&gt;%
      summarise(mn = mean(n_tags_treatment)),
    linetype = 2,
    col = &quot;grey&quot;
  ) +
  facet_grid(~treat, scales = &quot;free_x&quot;, space = &quot;free_x&quot;) +
  scale_y_continuous(expand = expansion(c(0, 0.05)), labels = comma) +
  scale_fill_manual(values = colours$qc) +
  labs(
    x = &quot;Sample&quot;,
    y = &quot;Library Size&quot;,
    fill = &quot;QC&quot;
  ) +
  ggtitle(
    glue(&quot;{target}: Library Sizes&quot;)
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_macs2_summary_files/figure-html/plot-macs2-libsize-individual-1.png" alt="*Library sizes for each ER sample. The horizontal line indicates the mean library size for each treatment group. Any samples marked for exclusion as described above will be indicated with an (F)*" width="960" />
<p class="caption">
<em>Library sizes for each ER sample. The horizontal line indicates the mean library size for each treatment group. Any samples marked for exclusion as described above will be indicated with an (F)</em>
</p>
</div>
</div>
<div id="gc-content" class="section level3">
<h3>GC Content</h3>
<pre class="r"><code>ys &lt;- 5e5
yieldSize(bfl) &lt;- ys</code></pre>
<pre class="r"><code>bfl %&gt;% 
  bplapply(
    function(x){
      seq &lt;- scanBam(x, param = ScanBamParam(what = &quot;seq&quot;))[[1]]$seq
      freq &lt;- letterFrequency(seq, letters = &quot;GC&quot;) / width(seq)
      list(freq[,1])
    }
  ) %&gt;% 
  as_tibble() %&gt;% 
  pivot_longer(
    cols = everything(), 
    names_to = &quot;name&quot;,
    values_to = &quot;freq&quot;
  ) %&gt;% 
  left_join(macs2_logs, by = &quot;name&quot;) %&gt;% 
  unnest(freq) %&gt;% 
  ggplot(aes(label, freq, fill = qc)) +
  geom_boxplot(alpha = 0.8) +
  geom_hline(
    aes(yintercept = med),
    data = . %&gt;% 
      group_by(treat) %&gt;% 
      summarise(med = median(freq)),
    linetype = 2,
    colour = rgb(0.2, 0.2, 0.8)
  ) +
  facet_grid(~treat, scales = &quot;free_x&quot;, space = &quot;free_x&quot;) +
  scale_y_continuous(labels = percent) +
  scale_fill_manual(values = colours$qc) +
  labs(
    x = &quot;Sample&quot;,
    y = &quot;GC content&quot;,
    fill = &quot;QC&quot;
  ) +
  ggtitle(
    glue(&quot;{target}: GC Content&quot;)
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_macs2_summary_files/figure-html/plot-gc-content-1.png" alt="*GC content for each bam file, taking the first 500,000 alignments from each sample. QC status is based on the number of peaks identified (see table above)*" width="960" />
<p class="caption">
<em>GC content for each bam file, taking the first 500,000 alignments from each sample. QC status is based on the number of peaks identified (see table above)</em>
</p>
</div>
</div>
<div id="peaks-detected" class="section level3">
<h3>Peaks Detected</h3>
<pre class="r"><code>suppressWarnings(
  macs2_logs %&gt;%
    ggplot(
      aes(label, total_peaks, fill = qc)
    ) +
    geom_col() +
    geom_label(
      aes(x = label, y = total_peaks, label = lab, colour = qc),
      data = . %&gt;%
        mutate(
          lab = comma(total_peaks, accuracy = 1),
          total = total_peaks
        ),
      inherit.aes = FALSE,
      show.legend = FALSE
    ) +
    geom_hline(
      aes(yintercept = min),
      data = . %&gt;%
        group_by(treat) %&gt;%
        summarise(min = config$peaks$qc$min_prop_peaks*max(total_peaks)),
      linetype = 2,
      col = &quot;grey&quot;
    ) +
    facet_grid(~treat, scales = &quot;free_x&quot;, space = &quot;free_x&quot;) +
    scale_y_continuous(expand = expansion(c(0, 0.05)), labels = comma) +
    scale_fill_manual(values = colours$qc) +
    scale_colour_manual(values = colours$qc) +
    labs(
      x = &quot;Sample&quot;,
      y = &quot;Total Peaks&quot;,
      fill = &quot;QC&quot;
    ) +
    ggtitle(
      glue(&quot;{target}: Number of Peaks&quot;)
    )
)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_macs2_summary_files/figure-html/plot-macs2-peaks-individual-1.png" alt="*Peaks identified for each ER sample. The number of peaks passing the inclusion criteria for `macs2 callpeak` (FDR &lt; 0.05) are provided. The number of retained peaks required for inclusion is shown for each treatment as the dashed grey line, with any samples marked for exclusion marked with an F.*" width="960" />
<p class="caption">
<em>Peaks identified for each ER sample. The number of peaks passing the inclusion criteria for <code>macs2 callpeak</code> (FDR &lt; 0.05) are provided. The number of retained peaks required for inclusion is shown for each treatment as the dashed grey line, with any samples marked for exclusion marked with an F.</em>
</p>
</div>
</div>
<div id="cross-correlations" class="section level3">
<h3>Cross Correlations</h3>
<pre class="r"><code>file.path(out_path, &quot;cross_correlations.tsv&quot;) %&gt;% 
  read_tsv() %&gt;% 
  left_join(samples, by = &quot;sample&quot;) %&gt;% 
  ggplot(aes(fl, correlation, colour = treat)) +
  geom_point(alpha = 0.1) +
  geom_smooth(se = FALSE, method = &#39;gam&#39;, formula = y ~ s(x, bs = &quot;cs&quot;)) +
  geom_vline(
    aes(xintercept = fragment_length),
    data = macs2_logs,
    colour = &quot;grey40&quot;, linetype = 2
  ) +
  facet_grid(as.formula(paste(&quot;treat ~&quot;, rep_col))) +
  scale_colour_manual(values = colours$treat[treat_levels]) +
  scale_x_continuous(
    breaks = seq(0, 5*max(macs2_logs$fragment_length), by = 200)
  ) +
  labs(
    x = &quot;Distance (bp)&quot;,
    y = &quot;Cross Correlation&quot;,
    colour = &quot;Treat&quot;
  ) +
  ggtitle(
    glue(&quot;{target}: Cross Correlations&quot;)
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_macs2_summary_files/figure-html/plot_correlation-1.png" alt="*Cross Correlaton between alignments up to 1kb apart. The dashed, grey, vertical line is the fragment length estimated by `macs2 callpeak` for each sample. For speed, only the first 5 chromosomes were used for sample-specific estimates.*" width="960" />
<p class="caption">
<em>Cross Correlaton between alignments up to 1kb apart. The dashed, grey, vertical line is the fragment length estimated by <code>macs2 callpeak</code> for each sample. For speed, only the first 5 chromosomes were used for sample-specific estimates.</em>
</p>
</div>
</div>
<div id="reads-in-peaks" class="section level3">
<h3>Reads In Peaks</h3>
<pre class="r"><code>samples$sample %&gt;%
  bplapply(
    function(x) {
      gr &lt;- individual_peaks[[x]]
      rip &lt;- 0
      if (length(gr) &gt; 0) {
        sbp &lt;- ScanBamParam(which = gr)
        rip &lt;- sum(countBam(bfl[[x]], param = sbp)$records)
      }
      tibble(
        name = x,
        reads_in_peaks = rip
      )
    }
  ) %&gt;%
  bind_rows() %&gt;%
  left_join(macs2_logs, by = &quot;name&quot;) %&gt;%
  mutate(
    frip = reads_in_peaks / n_tags_treatment
  ) %&gt;%
  ggplot(
    aes(label, frip, fill = qc)
  ) +
  geom_col() +
  geom_hline(yintercept = 0.01, colour = &quot;grey&quot;, linetype = 2) +
  facet_grid(~treat, scales = &quot;free_x&quot;, space = &quot;free&quot;) +
  scale_y_continuous(labels = percent, expand = expansion(c(0, 0.05))) +
  scale_fill_manual(values = colours$qc) +
  labs(
    x = &quot;Sample&quot;,
    y = &quot;Fraction of Reads In Peaks&quot;,
    fill = &quot;QC&quot;
  ) +
  ggtitle(
    glue(&quot;{target}: Fraction Of Reads In Peaks&quot;)
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_macs2_summary_files/figure-html/plot-frip-1.png" alt="*Fraction of Reads In Peaks for each sample. Higher values indicate more reads specifically associated with the ChIP target (ER). The minimum value for an acceptable sample (1%) is shown as a dashed horizontal line*" width="960" />
<p class="caption">
<em>Fraction of Reads In Peaks for each sample. Higher values indicate more reads specifically associated with the ChIP target (ER). The minimum value for an acceptable sample (1%) is shown as a dashed horizontal line</em>
</p>
</div>
</div>
</div>
<div id="results" class="section level2 tabset">
<h2 class="tabset">Results</h2>
<pre class="r"><code>oracle_peaks &lt;- treat_levels %&gt;%
  sapply(
    function(x) {
      gr &lt;- file.path(macs2_path, glue(&quot;{x}_merged_peaks.narrowPeak&quot;)) %&gt;%
        importPeaks(seqinfo = sq, blacklist = blacklist) %&gt;% 
        unlist()
      k &lt;- dplyr::filter(n_reps, treat == x)$n * config$peaks$qc$min_prop_reps
      if (k &gt; 0) {
        samp &lt;- dplyr::filter(macs2_logs, treat == x, qc != &quot;fail&quot;)$name
        gr$n_reps &lt;- countOverlaps(gr, individual_peaks[samp]) 
        gr$keep &lt;- gr$n_reps &gt; k
      } else {
        gr &lt;- GRanges(seqinfo = sq)
      }
      gr
    },
    simplify = FALSE
  ) %&gt;%
  GRangesList() 
consensus_peaks &lt;- oracle_peaks %&gt;%
  unlist() %&gt;%
  subset(keep) %&gt;%
  GenomicRanges::reduce() %&gt;%
  setNames(as.character(.)) %&gt;% 
  mutate(
    region = bestOverlap(
      ., 
      lapply(gene_regions, select, region) %&gt;% 
        GRangesList() %&gt;% 
        unlist(),
      var = &quot;region&quot;
    ) %&gt;% 
      factor(levels = regions)
  )
if (has_features) {
  consensus_peaks$Feature &lt;- bestOverlap(
    consensus_peaks, external_features, 
    var = &quot;feature&quot;, missing = &quot;no_feature&quot;
  ) %&gt;% 
    factor(levels = names(colours$features)) %&gt;% 
    fct_relabel(str_sep_to_title)
}
if (any_detected) {
  consensus_peaks$detected &lt;- bestOverlap(
    consensus_peaks, 
    gene_regions %&gt;% 
      lapply(select, any_of(&quot;detected&quot;)) %&gt;% 
      GRangesList() %&gt;% 
      unlist(), 
    var = &quot;detected&quot;
  ) %&gt;% 
    as.logical()
}</code></pre>
<p>A set of treatment-specific <code>target</code> <em>oracle peaks</em> was defined for each condition by taking the peaks defined by merging all samples from each treatment group. These were then compared to the individual replicates within each respective condition which passed the initial QC. Oracle peaks which overlapped a peak in more than 30% of the <em>individual replicates passing QC</em>, were retained.</p>
<pre class="r"><code>oracle_peaks %&gt;%
  lapply(as.data.frame) %&gt;%
  lapply(list) %&gt;%
  as_tibble() %&gt;%
  pivot_longer(everything(), names_to = &quot;treat&quot;, values_to = &quot;data&quot;) %&gt;%
  mutate(
    detected_peaks = map_int(data, nrow),
    retained = data %&gt;% lapply(dplyr::filter, keep) %&gt;% map_int(nrow),
    `% retained` = percent(retained / detected_peaks, 0.1)
  ) %&gt;%
  dplyr::select(treat, detected_peaks, retained, `% retained`) %&gt;%
  rename_all(str_sep_to_title) %&gt;% 
  pander(
    justify = &quot;lrrr&quot;,
    caption = glue(
      &quot;Oracle peaks detected by merging samples within each treatment group.&quot;,
      &quot;Peaks were only retained if detected in at least&quot;,
      &quot;{percent(config$peaks$qc$min_prop_reps)} of the retained samples for each&quot;, &quot;treatment group, as described above.&quot;,
      .sep = &quot; &quot;
    )
  )</code></pre>
<table style="width:65%;">
<caption>Oracle peaks detected by merging samples within each treatment group. Peaks were only retained if detected in at least 30% of the retained samples for each treatment group, as described above.</caption>
<colgroup>
<col width="11%" />
<col width="23%" />
<col width="15%" />
<col width="15%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Treat</th>
<th align="right">Detected Peaks</th>
<th align="right">Retained</th>
<th align="right">Retained</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Veh</td>
<td align="right">1,226</td>
<td align="right">1,225</td>
<td align="right">99.9%</td>
</tr>
<tr class="even">
<td align="left">DHT</td>
<td align="right">161</td>
<td align="right">161</td>
<td align="right">100.0%</td>
</tr>
</tbody>
</table>
<div id="consensus-peaks" class="section level3">
<h3>Consensus Peaks</h3>
<p>In addition to the oracle peaks, a set of 1,230 treatment-agnostic ER <strong>consensus peaks</strong> were defined. Ranges were the union of any overlapping ranges identified in one or more sets of oracle peaks.</p>
<pre class="r"><code>fig_path &lt;- here::here(&quot;docs&quot;, &quot;assets&quot;, target)
if (!dir.exists(fig_path)) dir.create(fig_path, recursive = TRUE)
fig_name &lt;- glue(&quot;{target}_common_peaks.png&quot;)
## An empty file to keep snakemake happy if &gt; 3 treatments
file.create(file.path(fig_path, fig_name)) 
if (length(treat_levels) &lt;= 3) {
  vd &lt;- oracle_peaks %&gt;%
    lapply(
      function(x) {
        names(
          subsetByOverlaps(
            consensus_peaks, 
            subset(x, keep)
          )
        )
      }
    ) %&gt;%
    setNames(treat_levels) %&gt;%
    venn.diagram(
      filename = file.path(fig_path, fig_name),
      imagetype = &quot;png&quot;,
      units = &quot;in&quot;,
      cat.cex = 1.4,
      height = 9, 
      width = 10,
      fill = colours$treat[treat_levels],
      alpha = 0.3
    )
  file.remove(list.files(fig_path, pattern = &quot;log$&quot;, full.names = TRUE))
}</code></pre>
<div class="figure">
<img src="assets/ER/ER_common_peaks.png" alt="" />
<p class="caption"><em>Number of ER consensus peaks which overlap oracle peaks defined within each condition.</em></p>
</div>
</div>
<div id="distance-to-tss" class="section level3">
<h3>Distance to TSS</h3>
<pre class="r"><code>a &lt;- consensus_peaks %&gt;% 
  join_nearest(
    mutate(tss, tss = start)
  ) %&gt;% 
  mutate(d = start + width/2 - tss) %&gt;% 
  mcols() %&gt;% 
  as.data.frame() %&gt;%
  ggplot(aes(d / 1e3)) +
  geom_density() +
  coord_cartesian(xlim = c(-100, 100)) +
  labs(
    x = &quot;Distance to TSS (kb)&quot;,
    y = &quot;Density&quot;
  )
b &lt;- consensus_peaks %&gt;% 
  join_nearest(
    mutate(tss, tss = start)
  ) %&gt;% 
  mutate(d = start + width/2 - tss) %&gt;%
  as.data.frame() %&gt;% 
  select(d) %&gt;% 
  arrange(abs(d)) %&gt;% 
  mutate(
    q = seq_along(d) / nrow(.)
  ) %&gt;% 
  ggplot(aes(abs(d) / 1e3, q)) +
  geom_line() +
  coord_cartesian(xlim = c(0, 100)) +
  labs(
    x = &quot;Distance to TSS (kb)&quot;,
    y = &quot;Quantile&quot;
  ) +
  scale_y_continuous(labels = percent, breaks = seq(0, 1, by = 0.2)) +
  scale_x_continuous(breaks = seq(0, 100, by = 20))
plot_grid(a, b, labels = c(&quot;A&quot;, &quot;B&quot;))</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_macs2_summary_files/figure-html/plot-dist-to-tss-hist-1.png" alt="*Distances from the centre of the ER consensus peak to the transcription start-site shown as A) a histogram, and B) as a cumulative distribution. 251 of the 1,230 ER consensus peaks (20.4%) directly overlapped a TSS.*" width="960" />
<p class="caption">
<em>Distances from the centre of the ER consensus peak to the transcription start-site shown as A) a histogram, and B) as a cumulative distribution. 251 of the 1,230 ER consensus peaks (20.4%) directly overlapped a TSS.</em>
</p>
</div>
</div>
<div id="gene-centric-regions" class="section level3">
<h3>Gene-Centric Regions</h3>
<pre class="r"><code>consensus_peaks %&gt;% 
  plotPie(fill = &quot;region&quot;) +
  scale_fill_manual(
    values = colours$regions %&gt;%
      setNames(regions[names(.)])
  ) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_macs2_summary_files/figure-html/plot-region-overlap-1.png" alt="*Proportions of ER consensus peaks which overlap gene-centric features.*" width="768" />
<p class="caption">
<em>Proportions of ER consensus peaks which overlap gene-centric features.</em>
</p>
</div>
</div>
<div id="detected-genes" class="section level3">
<h3>Detected Genes</h3>
<pre class="r"><code>consensus_peaks %&gt;% 
  mutate(
    region = fct_relabel(region, str_wrap, width = 20)
  ) %&gt;% 
  plotPie(
    fill = &quot;detected&quot;, x = &quot;region&quot;, scale_by = &quot;width&quot;, 
    label_alpha = 0.7, label_size = 4
  ) +
  scale_fill_manual(
    values = c(&quot;grey&quot;, &quot;red&quot;)
  ) +
  labs(
    x = &quot;Region&quot;, fill = &quot;Any Detected Genes&quot;
  ) +
  theme(
    panel.grid = element_blank(),
    panel.border = element_blank(),
    axis.line.x = element_line(),
    legend.position = &quot;top&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_macs2_summary_files/figure-html/plot-regions-expression-1.png" alt="*Proportion of ER peaks overlapping gene-centric features, incorporating detected expression from the RNA-Seq data. Note that intergenic regions &gt;10kb can never be near detected genes by simple definition. Values shown are the total genome occupancy in Kb*" width="768" />
<p class="caption">
<em>Proportion of ER peaks overlapping gene-centric features, incorporating detected expression from the RNA-Seq data. Note that intergenic regions &gt;10kb can never be near detected genes by simple definition. Values shown are the total genome occupancy in Kb</em>
</p>
</div>
</div>
<div id="external-features" class="section level3">
<h3>External Features</h3>
<pre class="r"><code>consensus_peaks %&gt;% 
  plotPie(fill = &quot;Feature&quot;, label_size = 4) +
  scale_fill_manual(
    values = colours$features %&gt;% setNames(str_sep_to_title(names(.)))
  ) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_macs2_summary_files/figure-html/plot-feature-overlap-1.png" alt="*The total number of ER consensus peaks which overlap external features provided in h3k27ac_features.gtf. If peaks map to multiple features, they are assigned to the one with the largest overlap.*" width="768" />
<p class="caption">
<em>The total number of ER consensus peaks which overlap external features provided in h3k27ac_features.gtf. If peaks map to multiple features, they are assigned to the one with the largest overlap.</em>
</p>
</div>
</div>
<div id="external-features-and-gene-centric-regions" class="section level3">
<h3>External Features And Gene-Centric Regions</h3>
<pre class="r"><code>consensus_peaks %&gt;% 
  plotPie(fill = &quot;region&quot;, x = &quot;Feature&quot;, label_size = 4, label_alpha = 0.7) +
  scale_fill_manual(
    values = colours$regions %&gt;%
      setNames(regions[names(.)])
  ) +
  labs(fill = &quot;Region&quot;) +
  theme(
    panel.grid = element_blank(),
    panel.border = element_blank(),
    axis.line.x = element_line(),
    legend.position = &quot;top&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_macs2_summary_files/figure-html/plot-feature-region-overlap-1.png" alt="*The total number of ER consensus peaks overlapping external features and gene-centric regions. If a peak overlaps multiple features or regions, it is assigned to be the one with the largest overlap.*" width="960" />
<p class="caption">
<em>The total number of ER consensus peaks overlapping external features and gene-centric regions. If a peak overlaps multiple features or regions, it is assigned to be the one with the largest overlap.</em>
</p>
</div>
</div>
<div id="detected-genes-external-features-and-gene-centric-regions" class="section level3">
<h3>Detected Genes, External Features And Gene-Centric Regions</h3>
<pre class="r"><code>consensus_peaks %&gt;% 
  plotPie(fill = &quot;region&quot;, x = &quot;Feature&quot;, y = &quot;detected&quot;, label_size = 4, label_alpha = 0.7) +
  scale_fill_manual(
    values = colours$regions %&gt;%
      setNames(regions[names(.)])
  ) +
  labs(y = &quot;Mapped To Detected Genes&quot;, fill = &quot;Region&quot;) +
  theme(
    panel.grid = element_blank(),
    axis.line.x = element_line(),
    legend.position = &quot;top&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_macs2_summary_files/figure-html/plot-pie-feature-region-detected-1.png" alt="*The total number of ER consensus peaks overlapping external features and gene-centric regions, separated by whether each peak is mapped to a detected gene. If a peak overlaps multiple features or regions, it is assigned to be the one with the largest overlap. In the absence of HiC an excess of Distal Intergenic regions would be expected to remain unmapped.*" width="960" />
<p class="caption">
<em>The total number of ER consensus peaks overlapping external features and gene-centric regions, separated by whether each peak is mapped to a detected gene. If a peak overlaps multiple features or regions, it is assigned to be the one with the largest overlap. In the absence of HiC an excess of Distal Intergenic regions would be expected to remain unmapped.</em>
</p>
</div>
</div>
</div>
<div id="data-export" class="section level2">
<h2>Data Export</h2>
<pre class="r"><code>all_out &lt;- list(
  consensus_peaks = file.path(out_path, &quot;consensus_peaks.bed&quot;),
  oracle_peaks = file.path(out_path, &quot;oracle_peaks.rds&quot;),
  renv = file.path(
    here::here(&quot;output/envs&quot;), 
    glue(&quot;{target}_macs2_summary.RData&quot;)
  )
) 
export(consensus_peaks, all_out$consensus_peaks)
oracle_peaks %&gt;%
  lapply(subset, keep) %&gt;%
  lapply(select, -keep) %&gt;%
  GRangesList() %&gt;%
  write_rds(all_out$oracle_peaks, compress = &quot;gz&quot;)
save.image(all_out$renv)</code></pre>
<p>During this workflow the following files were exported:</p>
<ul>
<li><strong>consensus_peaks</strong>: output/ER/consensus_peaks.bed</li>
<li><strong>oracle_peaks</strong>: output/ER/oracle_peaks.rds</li>
<li><strong>renv</strong>: output/envs/ER_macs2_summary.RData</li>
</ul>
<!-- end of list -->
<button type="button" class="btn btn-default btn-sessioninfo" data-toggle="collapse" data-target="#sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Session information
</button>
</p>
<div id="sessioninfo" class="collapse">
<p><strong>R version 4.1.1 (2021-08-10)</strong></p>
<p><strong>Platform:</strong> x86_64-conda-linux-gnu (64-bit)</p>
<p><strong>locale:</strong> <em>LC_CTYPE=en_AU.UTF-8</em>, <em>LC_NUMERIC=C</em>, <em>LC_TIME=en_AU.UTF-8</em>, <em>LC_COLLATE=en_AU.UTF-8</em>, <em>LC_MONETARY=en_AU.UTF-8</em>, <em>LC_MESSAGES=en_AU.UTF-8</em>, <em>LC_PAPER=en_AU.UTF-8</em>, <em>LC_NAME=C</em>, <em>LC_ADDRESS=C</em>, <em>LC_TELEPHONE=C</em>, <em>LC_MEASUREMENT=en_AU.UTF-8</em> and <em>LC_IDENTIFICATION=C</em></p>
<p><strong>attached base packages:</strong> <em>grid</em>, <em>stats4</em>, <em>stats</em>, <em>graphics</em>, <em>grDevices</em>, <em>utils</em>, <em>datasets</em>, <em>methods</em> and <em>base</em></p>
<p><strong>other attached packages:</strong> <em>extraChIPs(v.0.99.0)</em>, <em>SummarizedExperiment(v.1.24.0)</em>, <em>Biobase(v.2.54.0)</em>, <em>MatrixGenerics(v.1.6.0)</em>, <em>matrixStats(v.0.61.0)</em>, <em>ggside(v.0.2.0)</em>, <em>Rsamtools(v.2.10.0)</em>, <em>Biostrings(v.2.62.0)</em>, <em>XVector(v.0.34.0)</em>, <em>BiocParallel(v.1.28.3)</em>, <em>rlang(v.1.0.2)</em>, <em>cowplot(v.1.1.1)</em>, <em>VennDiagram(v.1.7.1)</em>, <em>futile.logger(v.1.4.3)</em>, <em>UpSetR(v.1.4.0)</em>, <em>ngsReports(v.1.10.0)</em>, <em>yaml(v.2.3.5)</em>, <em>plyranges(v.1.14.0)</em>, <em>scales(v.1.1.1)</em>, <em>pander(v.0.6.5)</em>, <em>glue(v.1.6.2)</em>, <em>rtracklayer(v.1.54.0)</em>, <em>GenomicRanges(v.1.46.1)</em>, <em>GenomeInfoDb(v.1.30.1)</em>, <em>IRanges(v.2.28.0)</em>, <em>S4Vectors(v.0.32.4)</em>, <em>BiocGenerics(v.0.40.0)</em>, <em>magrittr(v.2.0.3)</em>, <em>forcats(v.0.5.1)</em>, <em>stringr(v.1.4.0)</em>, <em>dplyr(v.1.0.8)</em>, <em>purrr(v.0.3.4)</em>, <em>readr(v.2.1.2)</em>, <em>tidyr(v.1.2.0)</em>, <em>tibble(v.3.1.6)</em>, <em>ggplot2(v.3.3.5)</em> and <em>tidyverse(v.1.3.1)</em></p>
<p><strong>loaded via a namespace (and not attached):</strong> <em>utf8(v.1.2.2)</em>, <em>tidyselect(v.1.1.2)</em>, <em>RSQLite(v.2.2.12)</em>, <em>AnnotationDbi(v.1.56.2)</em>, <em>htmlwidgets(v.1.5.4)</em>, <em>scatterpie(v.0.1.7)</em>, <em>munsell(v.0.5.0)</em>, <em>codetools(v.0.2-18)</em>, <em>DT(v.0.22)</em>, <em>withr(v.2.5.0)</em>, <em>colorspace(v.2.0-3)</em>, <em>filelock(v.1.0.2)</em>, <em>highr(v.0.9)</em>, <em>knitr(v.1.38)</em>, <em>rstudioapi(v.0.13)</em>, <em>labeling(v.0.4.2)</em>, <em>GenomeInfoDbData(v.1.2.7)</em>, <em>polyclip(v.1.10-0)</em>, <em>farver(v.2.1.0)</em>, <em>bit64(v.4.0.5)</em>, <em>rprojroot(v.2.0.3)</em>, <em>vctrs(v.0.4.0)</em>, <em>generics(v.0.1.2)</em>, <em>lambda.r(v.1.2.4)</em>, <em>xfun(v.0.30)</em>, <em>csaw(v.1.28.0)</em>, <em>biovizBase(v.1.42.0)</em>, <em>BiocFileCache(v.2.2.1)</em>, <em>R6(v.2.5.1)</em>, <em>doParallel(v.1.0.17)</em>, <em>clue(v.0.3-60)</em>, <em>locfit(v.1.5-9.5)</em>, <em>AnnotationFilter(v.1.18.0)</em>, <em>bitops(v.1.0-7)</em>, <em>cachem(v.1.0.6)</em>, <em>DelayedArray(v.0.20.0)</em>, <em>assertthat(v.0.2.1)</em>, <em>vroom(v.1.5.7)</em>, <em>BiocIO(v.1.4.0)</em>, <em>nnet(v.7.3-17)</em>, <em>gtable(v.0.3.0)</em>, <em>ensembldb(v.2.18.4)</em>, <em>GlobalOptions(v.0.1.2)</em>, <em>splines(v.4.1.1)</em>, <em>lazyeval(v.0.2.2)</em>, <em>dichromat(v.2.0-0)</em>, <em>broom(v.0.7.12)</em>, <em>checkmate(v.2.0.0)</em>, <em>modelr(v.0.1.8)</em>, <em>GenomicFeatures(v.1.46.5)</em>, <em>backports(v.1.4.1)</em>, <em>Hmisc(v.4.6-0)</em>, <em>EnrichedHeatmap(v.1.24.0)</em>, <em>tools(v.4.1.1)</em>, <em>ellipsis(v.0.3.2)</em>, <em>jquerylib(v.0.1.4)</em>, <em>RColorBrewer(v.1.1-2)</em>, <em>ggdendro(v.0.1.23)</em>, <em>Rcpp(v.1.0.8.3)</em>, <em>plyr(v.1.8.7)</em>, <em>base64enc(v.0.1-3)</em>, <em>progress(v.1.2.2)</em>, <em>zlibbioc(v.1.40.0)</em>, <em>RCurl(v.1.98-1.6)</em>, <em>prettyunits(v.1.1.1)</em>, <em>rpart(v.4.1.16)</em>, <em>GetoptLong(v.1.0.5)</em>, <em>zoo(v.1.8-9)</em>, <em>ggrepel(v.0.9.1)</em>, <em>haven(v.2.4.3)</em>, <em>cluster(v.2.1.3)</em>, <em>fs(v.1.5.2)</em>, <em>here(v.1.0.1)</em>, <em>data.table(v.1.14.2)</em>, <em>futile.options(v.1.0.1)</em>, <em>circlize(v.0.4.14)</em>, <em>reprex(v.2.0.1)</em>, <em>ProtGenerics(v.1.26.0)</em>, <em>hms(v.1.1.1)</em>, <em>evaluate(v.0.15)</em>, <em>XML(v.3.99-0.9)</em>, <em>jpeg(v.0.1-9)</em>, <em>readxl(v.1.4.0)</em>, <em>gridExtra(v.2.3)</em>, <em>shape(v.1.4.6)</em>, <em>compiler(v.4.1.1)</em>, <em>biomaRt(v.2.50.3)</em>, <em>crayon(v.1.5.1)</em>, <em>htmltools(v.0.5.2)</em>, <em>mgcv(v.1.8-40)</em>, <em>ggfun(v.0.0.6)</em>, <em>tzdb(v.0.3.0)</em>, <em>Formula(v.1.2-4)</em>, <em>lubridate(v.1.8.0)</em>, <em>DBI(v.1.1.2)</em>, <em>tweenr(v.1.0.2)</em>, <em>formatR(v.1.12)</em>, <em>dbplyr(v.2.1.1)</em>, <em>ComplexHeatmap(v.2.10.0)</em>, <em>MASS(v.7.3-56)</em>, <em>GenomicInteractions(v.1.28.0)</em>, <em>rappdirs(v.0.3.3)</em>, <em>Matrix(v.1.4-1)</em>, <em>cli(v.3.2.0)</em>, <em>metapod(v.1.2.0)</em>, <em>igraph(v.1.3.0)</em>, <em>parallel(v.4.1.1)</em>, <em>Gviz(v.1.38.3)</em>, <em>pkgconfig(v.2.0.3)</em>, <em>GenomicAlignments(v.1.30.0)</em>, <em>foreign(v.0.8-82)</em>, <em>plotly(v.4.10.0)</em>, <em>InteractionSet(v.1.22.0)</em>, <em>xml2(v.1.3.3)</em>, <em>foreach(v.1.5.2)</em>, <em>bslib(v.0.3.1)</em>, <em>rvest(v.1.0.2)</em>, <em>VariantAnnotation(v.1.40.0)</em>, <em>digest(v.0.6.29)</em>, <em>rmarkdown(v.2.13)</em>, <em>cellranger(v.1.1.0)</em>, <em>htmlTable(v.2.4.0)</em>, <em>edgeR(v.3.36.0)</em>, <em>restfulr(v.0.0.13)</em>, <em>curl(v.4.3.2)</em>, <em>rjson(v.0.2.21)</em>, <em>nlme(v.3.1-157)</em>, <em>lifecycle(v.1.0.1)</em>, <em>jsonlite(v.1.8.0)</em>, <em>limma(v.3.50.1)</em>, <em>viridisLite(v.0.4.0)</em>, <em>BSgenome(v.1.62.0)</em>, <em>fansi(v.1.0.3)</em>, <em>pillar(v.1.7.0)</em>, <em>lattice(v.0.20-45)</em>, <em>KEGGREST(v.1.34.0)</em>, <em>fastmap(v.1.1.0)</em>, <em>httr(v.1.4.2)</em>, <em>survival(v.3.3-1)</em>, <em>png(v.0.1-7)</em>, <em>iterators(v.1.0.14)</em>, <em>bit(v.4.0.4)</em>, <em>ggforce(v.0.3.3)</em>, <em>stringi(v.1.7.6)</em>, <em>sass(v.0.4.1)</em>, <em>blob(v.1.2.2)</em>, <em>latticeExtra(v.0.6-29)</em> and <em>memoise(v.2.0.1)</em></p>
</div>
</div>

<div class="footer">
  Workflow developed by:<br>
  Dr Stephen Pederson<br>
  Dame Roma Mitchell Cancer Research Laboratories<br>
  Adelaide Medical School, University of Adelaide, SA, Australia
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
