<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>ER Differential Binding: DHT Vs. Veh</title>

<script src="site_libs/header-attrs-2.13/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/core-js-2.5.3/shim.min.js"></script>
<script src="site_libs/react-17.0.0/react.min.js"></script>
<script src="site_libs/react-17.0.0/react-dom.min.js"></script>
<script src="site_libs/reactwidget-1.0.0/react-tools.js"></script>
<script src="site_libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<script src="site_libs/reactable-binding-0.2.3/reactable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GRAVI</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Annotations
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="annotation_description.html">Genome and Transcriptome Annotations</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    MACS2 Peak Calling
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="AR_macs2_summary.html">AR</a>
    </li>
    <li>
      <a href="ER_macs2_summary.html">ER</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Differential Binding
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">AR</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="AR_Veh_DHT_differential_binding.html">DHT Vs. Veh</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">ER</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="ER_Veh_DHT_differential_binding.html">DHT Vs. Veh</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pairwise Comparisons
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">AR-ER</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="AR_Veh_DHT_ER_Veh_DHT_pairwise_comparison.html">DHT Vs. Veh / DHT Vs. Veh</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steveped/GRAVI">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">ER Differential Binding: DHT Vs. Veh</h1>
<h4 class="date">05 April, 2022</h4>

</div>


<pre class="r"><code>target &lt;- &quot;ER&quot;
treat_levels &lt;- c(&quot;Veh&quot;, &quot;DHT&quot;)
threads &lt;- 6</code></pre>
<pre class="r"><code>library(tidyverse)
library(glue)
library(pander)
library(BiocParallel)
library(yaml)
library(plyranges)
library(rtracklayer)
library(GenomeInfoDb)
library(Rsamtools)
library(cowplot)
library(ngsReports)
library(csaw)
library(edgeR)
library(scales)
library(quantro)
library(qsmooth)
library(magrittr)
library(statmod)
library(IHW)
library(ggrepel)
library(rlang)
library(ggside)
library(InteractionSet)
library(GenomicInteractions)
library(reactable)
library(htmltools)
library(msigdbr)
library(goseq)</code></pre>
<pre class="r"><code>if (!&quot;extraChIPs&quot; %in% rownames(installed.packages()))
  BiocManager::install(&quot;steveped/extraChIPs&quot;)

stopifnot(library(extraChIPs, logical.return = TRUE))</code></pre>
<pre class="r"><code>panderOptions(&quot;big.mark&quot;, &quot;,&quot;)
panderOptions(&quot;missing&quot;, &quot;&quot;)
panderOptions(&quot;table.split.table&quot;, Inf)
theme_set(
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5)
  )
)
register(MulticoreParam(workers = threads))
source(here::here(&quot;workflow&quot;, &quot;scripts&quot;, &quot;custom_functions.R&quot;))</code></pre>
<pre class="r"><code>params &lt;- read_yaml(here::here(&quot;config&quot;, &quot;params.yml&quot;))
config &lt;- read_yaml(here::here(&quot;config&quot;, &quot;config.yml&quot;))
samples &lt;- here::here(&quot;output&quot;, target, &quot;qc_samples.tsv&quot;) %&gt;% 
  read_tsv() %&gt;% 
  dplyr::filter(
    qc == &quot;pass&quot;,
    treat %in% treat_levels
  ) %&gt;% 
  mutate(treat = factor(treat, levels = treat_levels)) %&gt;% 
  droplevels()
stopifnot(nrow(samples) &gt; 0)
rep_col &lt;- setdiff(
  colnames(samples), c(&quot;sample&quot;, &quot;treat&quot;, &quot;target&quot;, &quot;input&quot;, &quot;label&quot;, &quot;qc&quot;)
)
samples[[rep_col]] &lt;- as.factor(samples[[rep_col]])
fdr_alpha &lt;- config$comparisons$fdr</code></pre>
<pre class="r"><code>bam_path &lt;- here::here(config$paths$bam)
stopifnot(dir.exists(bam_path))
annotation_path &lt;- here::here(&quot;output&quot;, &quot;annotations&quot;)
stopifnot(dir.exists(annotation_path))
bw_path &lt;- here::here(config$paths$bigwig, target)
stopifnot(dir.exists(bw_path))
out_path &lt;- here::here(&quot;output&quot;, target)
if (!dir.exists(out_path)) dir.create(out_path)</code></pre>
<pre class="r"><code>sq &lt;- file.path(annotation_path, &quot;seqinfo.rds&quot;) %&gt;% 
  read_rds()
blacklist &lt;- here::here(annotation_path, &quot;blacklist.bed.gz&quot;) %&gt;%
  import.bed(seqinfo = sq) %&gt;%
  sort()

cb &lt;- config$genome$build %&gt;% 
  str_to_lower() %&gt;% 
  paste0(&quot;.cytobands&quot;) 
data(list = list(cb))
bands_df &lt;- mget(cb)[[cb]]

stopifnot(
  colnames(bands_df) == c(&quot;chrom&quot;, &quot;chromStart&quot;, &quot;chromEnd&quot;, &quot;name&quot;, &quot;gieStain&quot;)
)

genes_gr &lt;- file.path(annotation_path, &quot;all_gr.rds&quot;) %&gt;% 
  read_rds() %&gt;% 
  .[[&quot;gene&quot;]]
id2gene &lt;- structure(genes_gr$gene_name, names = genes_gr$gene_id)
trans_models &lt;- file.path(annotation_path, &quot;trans_models.rds&quot;) %&gt;% 
  read_rds() 

external_features &lt;- GRanges(seqinfo = sq)
if (!is.null(config$external$features)) {
  external_features &lt;- suppressWarnings(
    import.gff(here::here(config$external$features), genome = sq)
  )
  keep_cols &lt;- !vapply(
    mcols(external_features), function(x) all(is.na(x)), logical(1)
  )
  mcols(external_features) &lt;- mcols(external_features)[keep_cols]
}
has_features &lt;- length(external_features) &gt; 0
gene_regions &lt;- file.path(annotation_path, &quot;gene_regions.rds&quot;) %&gt;% 
  read_rds()
regions &lt;- vapply(gene_regions, function(x) unique(x$region), character(1))
any_detected &lt;- gene_regions %&gt;% 
  vapply(function(x) any(x$detected), logical(1)) %&gt;% 
  any()

rna_path &lt;- here::here(config$external$rnaseq)
rnaseq &lt;- tibble(gene_id = character())
if (length(rna_path) &gt; 0) {
  stopifnot(file.exists(rna_path))
  if (str_detect(rna_path, &quot;tsv$&quot;)) rnaseq &lt;- read_tsv(rna_path)
  if (str_detect(rna_path, &quot;csv$&quot;)) rnaseq &lt;- read_csv(rna_path)
  if (!&quot;gene_id&quot; %in% colnames(rnaseq)) stop(&quot;Supplied RNA-Seq data must contain the column &#39;gene_id&#39;&quot;)
}
tx_col &lt;- intersect(c(&quot;tx_id&quot;, &quot;transcript_id&quot;), colnames(rnaseq))
rna_gr_col &lt;- ifelse(length(tx_col) &gt; 0, &quot;transcript_id&quot;, &quot;gene_id&quot;)
rna_col &lt;- c(tx_col, &quot;gene_id&quot;)[[1]]</code></pre>
<pre class="r"><code>colours &lt;- read_rds(here::here(&quot;output&quot;, &quot;annotations&quot;, &quot;colours.rds&quot;)) %&gt;% 
  lapply(unlist)
region_colours &lt;- setNames(colours$regions, regions[names(colours$regions)])
if (has_features)
  feature_colours &lt;- setNames(
    colours$features, str_sep_to_title(names(colours$features))
  )
direction_colours &lt;- colours$direction %&gt;% 
  setNames(str_to_title(names(.))) %&gt;% 
  .[c(&quot;Up&quot;, &quot;Down&quot;, &quot;Unchanged&quot;)]
treat_colours &lt;- colours$treat[treat_levels]</code></pre>
<pre class="r"><code>hic &lt;- GInteractions()
hic_path &lt;- here::here(config$external$hic)
has_hic &lt;- FALSE
if (length(hic_path) &gt; 0)
  if (file.exists(hic_path)) {
    has_hic &lt;- TRUE
    hic &lt;- makeGenomicInteractionsFromFile(hic_path, type = &quot;bedpe&quot;)
    reg_combs &lt;- expand.grid(regions, regions) %&gt;% 
      as.matrix() %&gt;% 
      apply(
        MARGIN = 1, 
        function(x) {
          x &lt;- sort(factor(x, levels = regions))
          paste(as.character(x), collapse = &quot; - &quot;)
        }
      ) %&gt;% 
      unique()
    hic$regions &lt;- anchors(hic) %&gt;% 
      vapply(
        bestOverlap,
        y = GRangesList(lapply(gene_regions, granges)),
        character(length(hic))
      ) %&gt;% 
      apply(MARGIN = 2, function(x) regions[x]) %&gt;% 
      apply(
        MARGIN = 1, 
        function(x) {
          x &lt;- sort(factor(x, levels = regions))
          paste(as.character(x), collapse = &quot; - &quot;)
        }
      ) %&gt;% 
      factor(levels = reg_combs) %&gt;%
      fct_relabel(
        str_replace_all,
        pattern = &quot;Promoter \\([0-9kbp/\\+-]+\\)&quot;, replacement = &quot;Promoter&quot;
      )
    if (has_features) {
      feat_combs &lt;- expand.grid(names(feature_colours), names(feature_colours)) %&gt;% 
        as.matrix() %&gt;% 
        apply(
          MARGIN = 1, 
          function(x) {
            x &lt;- sort(factor(x, levels = names(feature_colours)))
            paste(as.character(x), collapse = &quot; - &quot;)
          }
        ) %&gt;% 
        unique()
      hic$features &lt;- vapply(
        anchors(hic),
        function(x) bestOverlap(
          x, external_features, var = &quot;feature&quot;, missing = &quot;no_feature&quot;
        ),
        character(length(hic))
      )  %&gt;% 
        apply(
          MARGIN = 1, 
          function(x) {
            x &lt;- sort(factor(x, levels = names(feature_colours)))
            paste(as.character(x), collapse = &quot; - &quot;)
          }
        ) %&gt;% 
        factor(levels = feat_combs) %&gt;%
        fct_relabel(str_sep_to_title, pattern = &quot;_&quot;)
    }
  }
stopifnot(is(hic, &quot;GInteractions&quot;))
seqlevels(hic) &lt;- seqlevels(sq)
seqinfo(hic) &lt;- sq
hic &lt;- hic[!overlapsAny(hic, blacklist)]</code></pre>
<pre class="r"><code>facet_labeller &lt;- as_labeller(
  c(
    c(
      `TRUE` = &quot;Consensus Peak&quot;,
      `FALSE` = &quot;No Consensus Peak&quot;
    ),
    structure(treat_levels, names = treat_levels),
    c(
      Up = glue(&quot;Increased {target} Binding&quot;),
      Down = glue(&quot;Decreased {target} Binding&quot;)
    ),
    c(
      AveExpr = &quot;logCPM&quot;, logFC = &quot;logFC&quot;
    )
  )
)</code></pre>
<pre class="r"><code>all_out &lt;- list(
  results = file.path(
    out_path, 
    glue(&quot;{treat_levels[[1]]}_{treat_levels[[2]]}_differential_binding.rds&quot;)
  ),
  csv = file.path(
    out_path, 
    glue(&quot;{target}_{treat_levels[[1]]}_{treat_levels[[2]]}_differential_binding.csv.gz&quot;)
  ),
  windows  = file.path(
    out_path,
    glue(&quot;{treat_levels[[1]]}_{treat_levels[[2]]}_filtered_windows.rds&quot;)
  ),
  up_regions = file.path(
    out_path, glue(&quot;{treat_levels[[1]]}_{treat_levels[[2]]}_up.bed&quot;)
  ),
  down_regions = file.path(
    out_path, glue(&quot;{treat_levels[[1]]}_{treat_levels[[2]]}_down.bed&quot;)
  ),
  renv = here::here(
    &quot;output/envs&quot;,
    glue(
      &quot;{target}_{treat_levels[[1]]}_{treat_levels[[2]]}_differential_binding.RData&quot;
    )
  )
)</code></pre>
<div id="data-preparation" class="section level1">
<h1>Data Preparation</h1>
<pre class="r"><code>consensus_peaks &lt;- here::here(&quot;output&quot;, target, &quot;consensus_peaks.bed&quot;) %&gt;%
    import.bed(seqinfo = sq)</code></pre>
<pre class="r"><code>bfl &lt;- bam_path %&gt;%
  file.path(target, glue(&quot;{samples$sample}.bam&quot;)) %&gt;%
  c(
    file.path(bam_path, &quot;Input&quot;, glue(&quot;{unique(samples$input)}.bam&quot;))
  ) %&gt;% 
  BamFileList() %&gt;%
  setNames(c(samples$sample, unique(samples$input)))</code></pre>
<pre class="r"><code>ys &lt;- 1000
bfl_summary &lt;- bfl %&gt;%
    bplapply(
        function(x) {
            yieldSize(x) &lt;- ys
            sbp &lt;- ScanBamParam(what = c(&quot;qwidth&quot;, &quot;mapq&quot;))
            open(x)
            vals &lt;- scanBam(x, param = sbp)[[1]]
            close(x)
            list(as_tibble(vals))
        }
    ) %&gt;% 
    as_tibble() %&gt;%
    pivot_longer(everything(), names_to = &quot;sample&quot;) %&gt;%
    unnest(value) %&gt;% 
    pivot_longer(
        cols = c(&quot;qwidth&quot;, &quot;mapq&quot;),
        names_to = &quot;stat&quot;,
        values_to = &quot;value&quot;
    ) %&gt;%
    left_join(samples, by = &quot;sample&quot;) %&gt;% 
    mutate(
        across(
            all_of(c(&quot;target&quot;, &quot;label&quot;, &quot;treat&quot;)),
            str_replace_na, replacement = &quot;Input&quot;
        ),
        treat = factor(treat, levels = c(&quot;Input&quot;, treat_levels)),
        label = fct_inorder(label)
    ) %&gt;% 
    split(f = .$stat)
anyDups &lt;- bplapply(
    bfl,
    function(x) {
        sbp &lt;- ScanBamParam(
            flag = scanBamFlag(isDuplicate = TRUE),
            which = GRanges(sq)[which.min(seqlengths(sq))],
            what = &quot;qname&quot;
            )
        length(scanBam(x, param = sbp)[[1]]$qname)  &gt; 0
    }
) %&gt;%
  unlist()
anyPE &lt;- bplapply(
    bfl,
    function(x){
        yieldSize(x) &lt;- ys
        open(x)
        flag &lt;- scanBam(x, param=ScanBamParam(what=&quot;flag&quot;))[[1]]$flag
        close(x)
        any(bamFlagTest(flag, &quot;isPaired&quot;))
    }
) %&gt;% 
  unlist()</code></pre>
<p>Taking the first 1,000 alignments, a brief inspection of the Bam Files revealed:</p>
<ul>
<li>Aligned read lengths ranged between 16 and 76. The median length was 76.</li>
<li>MAPQ scores of aligned reads ranged between 11 and 42. The median score was 37.</li>
<li>All bam files have been de-duplicated</li>
<li>All alignments were single-end</li>
</ul>
<div id="sliding-window-counts" class="section level2">
<h2>Sliding Window Counts</h2>
<pre class="r"><code>macs2_merged_logs &lt;- here::here(
    &quot;data&quot;, &quot;macs2&quot;, target, glue(&quot;{treat_levels}_merged_callpeak.log&quot;)
    ) %&gt;%
    importNgsLogs()
fl &lt;- max(macs2_merged_logs$fragment_length)
rp &lt;- readParam(
  pe = ifelse(any(anyPE), &quot;both&quot;, &quot;none&quot;),
  dedup = any(anyDups),
  restrict = seqnames(sq),
  discard = blacklist,
)
win_step &lt;- 10*(1 + fl %/% 30)
win_size &lt;- 3*win_step
window_counts &lt;- windowCounts(
  bam.files = bfl,
  spacing = win_step,
  width = win_size,
  ext = fl,
  filter = length(bfl) - 1,
  param = rp,
  BPPARAM = bpparam()
)
colData(window_counts) &lt;- colData(window_counts) %&gt;% 
  as_tibble(rownames = &quot;sample&quot;) %&gt;% 
  dplyr::select(all_of(c(&quot;sample&quot;, &quot;bam.files&quot;, &quot;totals&quot;, &quot;ext&quot;, &quot;rlen&quot;))) %&gt;% 
  left_join(samples, by = &quot;sample&quot;) %&gt;% 
  mutate(
    treat = fct_explicit_na(treat, &quot;Input&quot;), 
    target = str_replace_na(target, &quot;Input&quot;)
  ) %&gt;% 
  as.data.frame() %&gt;% 
  column_to_rownames(&quot;sample&quot;) %&gt;% 
  DataFrame()
window_counts &lt;- sortSeqlevels(window_counts)
seqinfo(window_counts) &lt;- sq</code></pre>
<p>Using the estimated fragment length of 299nt, a set of genomic sliding windows were defined using a window size of 300bp, sliding in increments of 100bp. With the exclusion of black-listed regions, all alignments within each window were counted for each ER-associated sample, and all relevant input samples. Any windows with fewer than 6 alignments across all samples were discarded, leaving a total of 1,270,602 sliding windows, covering 4% of the reference genome.</p>
<p>These windows were then filtered, discarding windows with low counts.</p>
<pre class="r"><code>filtered_counts &lt;- dualFilter(
  x = window_counts[, samples$sample],
  bg = window_counts[, samples$input],
  ref = consensus_peaks,
  keep.totals = TRUE,
  q = config$comparisons$filter_q
)
colData(filtered_counts) &lt;- droplevels(colData(filtered_counts))
n_max &lt;- min(5e4, nrow(filtered_counts))</code></pre>
<pre class="r"><code>list(
  `Genomic Windows` = window_counts,
  `Retained Windows` = filtered_counts,
  `Consensus Peaks` = consensus_peaks
) %&gt;% 
  lapply(granges) %&gt;% 
  lapply(
    function(x) {
      tibble(
        N = comma(length(x)),
        `Total Width (kb)` = comma(sum(width(reduce(x))) / 1e3),
        `Median Width (bp)` = median(width(x))
      )
    }
  ) %&gt;% 
  lapply(list) %&gt;% 
  as_tibble() %&gt;% 
  pivot_longer(everything(), names_to = &quot;Dataset&quot;) %&gt;% 
  unnest(everything()) %&gt;% 
  left_join(
    tibble(
      Dataset = c(&quot;Retained Windows&quot;, &quot;Consensus Peaks&quot;),
      `Unique (kb)` = c(
        round(sum(width(setdiff(granges(filtered_counts), consensus_peaks))) / 1e3, 1),
        round(sum(width(setdiff(consensus_peaks, granges(filtered_counts)))) / 1e3, 1)
      ),
      `% Unique` = percent(
        1e3*`Unique (kb)` / c(
          sum(width(reduce(granges(filtered_counts)))),
          sum(width(consensus_peaks))
        ),
        0.1
      )
    )
  ) %&gt;% 
  pander(
    justify = &quot;lrrrrr&quot;,
    caption = paste(
      &quot;A dual filtering strategy was used based on retaining&quot;, 
      percent(config$comparisons$filter_q, 0.1), 
      &quot;of genomic windows overlapping the consensus peaks identified by &quot;, 
      &quot;`macs2 callpeak` on merged samples. This approach combined both &quot;, 
      &quot;1) expression percentiles, and 2) signal strength in relation to the input sample.&quot;,
      &quot;The complete set of sliding windows covered the majority of the genome,&quot;,
      &quot;whilst those retained after filtering were focussed on strong binding&quot;,
      &quot;signal. Consensus peaks were as identified by `macs2 callpeak` in a&quot;, 
      &quot;previous step of the workflow.&quot;,
      &quot;Importantly, consensus peaks are non-overlapping, whilst the other two&quot;,
      &quot;datasets are derived from overlapping sliding windows.&quot;,
      &quot;This strategy of filtering the set of initial sliding windows retained&quot;,
      percent(
        1 - sum(width(setdiff(consensus_peaks, rowRanges(filtered_counts)))) / sum(width(consensus_peaks))
      ), 
      &quot;of the genomic regions covered by consensus peaks, with&quot;, 
      percent(
        sum(width(setdiff(rowRanges(filtered_counts), consensus_peaks))) / sum(width(reduce(granges(filtered_counts)))))
      , 
      &quot;of the retained windows being outside genomic regions covered by consensus peaks.&quot;,
      ifelse(
        n_max == 5e4, 
        &quot;For subsequent density and RLE plots, a subsample of 50,000 genomic regions will be used for speed.&quot;, 
        &quot;&quot;
      )
    )
  )</code></pre>
<table>
<caption>A dual filtering strategy was used based on retaining 60.0% of genomic windows overlapping the consensus peaks identified by <code>macs2 callpeak</code> on merged samples. This approach combined both 1) expression percentiles, and 2) signal strength in relation to the input sample. The complete set of sliding windows covered the majority of the genome, whilst those retained after filtering were focussed on strong binding signal. Consensus peaks were as identified by <code>macs2 callpeak</code> in a previous step of the workflow. Importantly, consensus peaks are non-overlapping, whilst the other two datasets are derived from overlapping sliding windows. This strategy of filtering the set of initial sliding windows retained 87% of the genomic regions covered by consensus peaks, with 37% of the retained windows being outside genomic regions covered by consensus peaks.</caption>
<colgroup>
<col width="20%" />
<col width="12%" />
<col width="20%" />
<col width="21%" />
<col width="14%" />
<col width="11%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Dataset</th>
<th align="right">N</th>
<th align="right">Total Width (kb)</th>
<th align="right">Median Width (bp)</th>
<th align="right">Unique (kb)</th>
<th align="right">% Unique</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Genomic Windows</td>
<td align="right">1,270,602</td>
<td align="right">127,461</td>
<td align="right">300</td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr class="even">
<td align="left">Retained Windows</td>
<td align="right">7,535</td>
<td align="right">1,013</td>
<td align="right">300</td>
<td align="right">370.1</td>
<td align="right">36.5%</td>
</tr>
<tr class="odd">
<td align="left">Consensus Peaks</td>
<td align="right">1,230</td>
<td align="right">735</td>
<td align="right">515.5</td>
<td align="right">92.6</td>
<td align="right">12.6%</td>
</tr>
</tbody>
</table>
</div>
<div id="normalisation" class="section level2 tabset">
<h2 class="tabset">Normalisation</h2>
<pre class="r"><code>qtest &lt;- quantro(
  assay(filtered_counts, &quot;logCPM&quot;),
  groupFactor = filtered_counts$treat
)
pander(
  anova(qtest), 
  caption = paste(
    &quot;*Results from qtest suggesting that the two treatment groups are drawn from&quot;,
    ifelse(
      qtest@anova$`Pr(&gt;F)`[[1]] &lt; 0.05,
      &quot;different distributions.*&quot;,
      &quot;the same distribution.*&quot;
    )
  )
)</code></pre>
<table style="width:86%;">
<caption><em>Results from qtest suggesting that the two treatment groups are drawn from different distributions.</em></caption>
<colgroup>
<col width="25%" />
<col width="6%" />
<col width="12%" />
<col width="13%" />
<col width="13%" />
<col width="13%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">Df</th>
<th align="center">Sum Sq</th>
<th align="center">Mean Sq</th>
<th align="center">F value</th>
<th align="center">Pr(&gt;F)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>groupFactor</strong></td>
<td align="center">1</td>
<td align="center">2.72</td>
<td align="center">2.72</td>
<td align="center">19.82</td>
<td align="center">0.01123</td>
</tr>
<tr class="even">
<td align="center"><strong>Residuals</strong></td>
<td align="center">4</td>
<td align="center">0.549</td>
<td align="center">0.1372</td>
<td align="center"></td>
<td align="center"></td>
</tr>
</tbody>
</table>
<pre class="r"><code>qs &lt;-qsmooth(
  assay(filtered_counts, &quot;logCPM&quot;), group_factor = filtered_counts$treat
)
assay(filtered_counts, &quot;qsmooth&quot;) &lt;- qsmoothData(qs)</code></pre>
<p>Smooth-Quantile normalisation was performed, grouping by treatment</p>
<div id="qsmooth-weights" class="section level3">
<h3>QSmooth Weights</h3>
<pre class="r"><code>qsmoothPlotWeights(qs)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-qs-weights-1.png" alt="*Quantile-specific weights used by the Smooth-Quantile normalisation. Low weights indicate expression quantiles which appear to be more specific within a group, whilst higher weights indicate similarity between groups.*" width="960" />
<p class="caption">
<em>Quantile-specific weights used by the Smooth-Quantile normalisation. Low weights indicate expression quantiles which appear to be more specific within a group, whilst higher weights indicate similarity between groups.</em>
</p>
</div>
</div>
<div id="density-plot-sliding-windows" class="section level3">
<h3>Density Plot: Sliding Windows</h3>
<pre class="r"><code>plotAssayDensities(
  window_counts, trans = &quot;log1p&quot;, n_max = n_max, colour = &quot;treat&quot;,
  linetype = &quot;target&quot;
) +
  scale_colour_manual(values = colours$treat) + 
  labs(
    x = &quot;log(counts + 1)&quot;, y = &quot;Density&quot;, colour = &quot;Treat&quot;, linetype = &quot;Target&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-density-windows-1.png" alt="*Density plot for all windows prior to selection of those more likely to contain true signal.*" width="960" />
<p class="caption">
<em>Density plot for all windows prior to selection of those more likely to contain true signal.</em>
</p>
</div>
</div>
<div id="density-plot-logcpm" class="section level3">
<h3>Density Plot: logCPM</h3>
<pre class="r"><code>filtered_counts %&gt;% 
  plotAssayDensities(&quot;logCPM&quot;, colour = &quot;treat&quot;, n_max = n_max) +
  scale_colour_manual(values = treat_colours) +
  labs(colour = &quot;Treat&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-logcpm-densities-1.png" alt="*logCPM distributions for all retained windows prior to smooth quantile normalisation*" width="960" />
<p class="caption">
<em>logCPM distributions for all retained windows prior to smooth quantile normalisation</em>
</p>
</div>
</div>
<div id="density-plot-normalised" class="section level3">
<h3>Density Plot: Normalised</h3>
<pre class="r"><code>filtered_counts %&gt;% 
  plotAssayDensities(&quot;qsmooth&quot;, colour = &quot;treat&quot;, n_max = n_max) +
  scale_colour_manual(values = treat_colours) +
  labs(
    x = &quot;Normalised logCPM&quot;, colour = &quot;Treat&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-qsmooth-densities-1.png" alt="*SQ-normalised logCPM distributions for all retained windows.*" width="960" />
<p class="caption">
<em>SQ-normalised logCPM distributions for all retained windows.</em>
</p>
</div>
</div>
<div id="rle-logcpm" class="section level3">
<h3>RLE: logCPM</h3>
<pre class="r"><code>filtered_counts %&gt;% 
  plotAssayRle(
  &quot;logCPM&quot;, fill = &quot;treat&quot;, rle_group = &quot;treat&quot;, x_col = &quot;label&quot;, n_max = n_max
) +
  geom_hline(yintercept = 0) + 
  facet_grid(.~treat, scales = &quot;free_x&quot;, space = &quot;free_x&quot;) +
  scale_fill_manual(values = treat_colours) +
  labs(x = &quot;Sample&quot;, fill = &quot;Treat&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-rle-logcpm-1.png" alt="*RLE plot showing logCPM values prior to normalisation. RLE values were calculated within each treatment group to account for the potentially different binding dynamics of ER.*" width="960" />
<p class="caption">
<em>RLE plot showing logCPM values prior to normalisation. RLE values were calculated within each treatment group to account for the potentially different binding dynamics of ER.</em>
</p>
</div>
</div>
<div id="rle-normalised" class="section level3">
<h3>RLE: Normalised</h3>
<pre class="r"><code>filtered_counts %&gt;% 
  plotAssayRle(
  &quot;qsmooth&quot;, fill = &quot;treat&quot;, rle_group = &quot;treat&quot;, x_col = &quot;label&quot;, n_max = n_max
) +
  geom_hline(yintercept = 0) + 
  facet_grid(.~treat, scales = &quot;free_x&quot;, space = &quot;free_x&quot;) +
  scale_fill_manual(values = treat_colours) +
  labs(x = &quot;Sample&quot;, fill = &quot;Treat&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-rle-normalised-1.png" alt="*RLE plot showing logCPM values after Smooth Quantile normalisation. RLE values were calculated within each treatment group to account for the potentially different binding dynamics of ER.*" width="960" />
<p class="caption">
<em>RLE plot showing logCPM values after Smooth Quantile normalisation. RLE values were calculated within each treatment group to account for the potentially different binding dynamics of ER.</em>
</p>
</div>
</div>
<div id="pca" class="section level3">
<h3>PCA</h3>
<pre class="r"><code>a &lt;- plotAssayPCA(filtered_counts, &quot;logCPM&quot;, colour = &quot;treat&quot;, label = &quot;label&quot;) +
  scale_colour_manual(values = treat_colours)
b &lt;- plotAssayPCA(filtered_counts, &quot;qsmooth&quot;, colour = &quot;treat&quot;, label = &quot;label&quot;) +
  scale_colour_manual(values = treat_colours) 
plot_grid(
  a + theme(legend.position = &quot;none&quot;), b + theme(legend.position = &quot;none&quot;), 
  get_legend(b + labs(colour = &quot;Treat&quot;)), rel_widths = c(1, 1, 0.2), 
  nrow = 1, labels = c(&quot;A&quot;, &quot;B&quot;)
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-pca-1.png" alt="*PCA plots for A) logCPM and B) SQ-Normalised logCPM*" width="960" />
<p class="caption">
<em>PCA plots for A) logCPM and B) SQ-Normalised logCPM</em>
</p>
</div>
</div>
</div>
</div>
<div id="differential-binding-analysis" class="section level1">
<h1>Differential Binding Analysis</h1>
<div id="primary-analysis" class="section level2 tabset">
<h2 class="tabset">Primary Analysis</h2>
<pre class="r"><code>X &lt;- model.matrix(~treat, data = colData(filtered_counts)) %&gt;%
  set_colnames(str_remove(colnames(.), &quot;treat&quot;))
colData(filtered_counts)$design &lt;- X
paired_cors &lt;- block &lt;- NULL
if (config$comparisons$paired) {
  block &lt;- colData(filtered_counts)[[rep_col]]
  set.seed(1e6)
  ind &lt;- sample.int(nrow(filtered_counts), n_max, replace = FALSE)
  paired_cors &lt;- duplicateCorrelation(
    object = assay(filtered_counts, &quot;qsmooth&quot;)[ind, ],
    design = X,
    block = block
  )$consensus.correlation
}
fc &lt;- ifelse(is.null(config$comparisons$fc), 1, config$comparisons$fc)
fit &lt;- assay(filtered_counts, &quot;qsmooth&quot;) %&gt;%
  lmFit(design = X, block = block, correlation = paired_cors) %&gt;% 
  treat(fc = fc, trend = TRUE, robust = FALSE)
res_cols &lt;- c(&quot;logFC&quot;, &quot;AveExpr&quot;, &quot;t&quot;, &quot;P.Value&quot;, &quot;fdr&quot;)
rowData(filtered_counts) &lt;- rowData(filtered_counts) %&gt;% 
  .[!colnames(.) %in% res_cols] %&gt;% 
  cbind(
    fit %&gt;% 
      topTable(sort.by = &quot;none&quot;, number = Inf) %&gt;% 
      DataFrame() %&gt;% 
      setNames(res_cols)
  )</code></pre>
<p>After SQ-normalisation of logCPM values, the <code>limma-trend</code> method was applied to all retained windows. A simple linear model was fitted taking Veh as the baseline and estimating the effects of DHT on ER binding within each sliding window, incorporating a trended prior-variance. After fitting, hypothesis testing was performed testing:</p>
<p><span class="math display">\[
H_0: -\lambda &lt; \mu &lt; \lambda
\]</span> against <span class="math display">\[
H_A: |\mu| &gt; \lambda
\]</span> where <span class="math inline">\(\mu\)</span> represents the true mean change in ER binding. The value <span class="math inline">\(\lambda =\)</span> 0.26 was chosen as this corresponds to a 20% change in detected signal on the log<sub>2</sub> scale. This is the <code>treat</code> methodology as developed by McCarthy &amp; Smyth, and p-values were obtained for each initial window, before merging adjoining windows.</p>
<pre class="r"><code>merged_results &lt;- mergeByCol(
  filtered_counts, col = &quot;AveExpr&quot;, merge_within = 2 * win_step
) %&gt;% 
  mutate(
    overlaps_ref = overlapsAny(., consensus_peaks),
    direction = case_when(
      n_up == 0 &amp; n_down == 0 ~ &quot;Unchanged&quot;,
      n_up &gt; n_down ~ &quot;Up&quot;,
      n_down &gt; n_up ~ &quot;Down&quot;,
      n_up == n_down ~ &quot;Ambiguous&quot;
    ) %&gt;% 
      factor(levels = c(&quot;Up&quot;, &quot;Down&quot;, &quot;Ambiguous&quot;, &quot;Unchanged&quot;)) %&gt;% 
      droplevels()
  ) </code></pre>
<p>After selection of the 7,535 sliding windows (300bp), these were merged into 1,232 genomic regions with size ranging from 300 to 3300bp, with the median width being 800bp. Representative estimates of differential binding (i.e. logFC) were taken from the sliding window with the <em>highest average signal</em> across all samples, within each set of windows to be merged. Similarly, p-values from the above tests were also selected from the same sliding window as representative of the merged region. The number of windows showing increased or decreased binding within each merged region were also included, by counting those within each set of windows with p-values lower than the selected window. FDR-adjustment was performed with 939 regions (76.2%) showing significance, based on an FDR-adjusted p-value alone.</p>
<div id="merged-windows" class="section level3">
<h3>Merged Windows</h3>
<pre class="r"><code>merged_results %&gt;% 
  mutate(kb = width / 1e3) %&gt;% 
  as_tibble() %&gt;% 
  mutate(
    overlaps_ref = unlist(facet_labeller(overlaps_ref)) %&gt;% 
      str_replace_all(&quot; Peak&quot;, &quot;\nPeak&quot;)
  ) %&gt;% 
  ggplot(aes(AveExpr, kb)) +
  geom_jitter(
    aes(colour = overlaps_ref), 
    alpha = 0.6, width = 0, height = 0.04
  ) +
  geom_xsideboxplot(
    aes(y = overlaps_ref, fill = overlaps_ref), orientation = &quot;y&quot;
  ) +
  geom_ysideboxplot(
    aes(x = overlaps_ref, fill = overlaps_ref), orientation = &quot;x&quot;
  ) +
  scale_xsidey_discrete() +
  scale_ysidex_discrete() +
  scale_colour_brewer(palette = &quot;Set1&quot;) +
  scale_fill_brewer(palette = &quot;Set1&quot;) +
  labs(
    x = &quot;logCPM&quot;, y = &quot;Merged Window Size (kb)&quot;,
    fill = glue(&quot;{target} Macs2 Peak&quot;), colour = glue(&quot;{target} Macs2 Peak&quot;)
  ) +
  theme(
    axis.title.x = element_text(hjust = 0.45),
    axis.title.y = element_text(hjust = 0.4),
    ggside.axis.text.y = element_text(hjust = 0.5),
    ggside.panel.scale.x = 0.3,
    ggside.panel.scale.y = 0.25,
    legend.position = c(0.9, 0.9)
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-merged-size-1.png" alt="*Comparison of ER signal intensity (logCPM) and merged window size (kb) for windows which overlap a macs2 consensus peak and those which don't. A mall amount of jitter has been added to the y-axis in the lower-left panel.*" width="960" />
<p class="caption">
<em>Comparison of ER signal intensity (logCPM) and merged window size (kb) for windows which overlap a macs2 consensus peak and those which don’t. A mall amount of jitter has been added to the y-axis in the lower-left panel.</em>
</p>
</div>
</div>
<div id="mean-variance-relationship" class="section level3">
<h3>Mean-Variance Relationship</h3>
<pre class="r"><code>tibble(A = fit$Amean, pre = fit$sigma, post = sqrt(fit$s2.post)) %&gt;% 
  sample_n(n_max) %&gt;% 
  pivot_longer(c(&quot;pre&quot;, &quot;post&quot;), names_to = &quot;step&quot;, values_to = &quot;sigma&quot;) %&gt;%
  mutate(step = factor(step, levels = c(&quot;pre&quot;, &quot;post&quot;))) %&gt;% 
  ggplot(aes(A, sigma, colour = step)) +
  geom_point(size = 1, alpha = 0.5) +
  geom_smooth(
    se = FALSE, method = &quot;loess&quot;, formula = y ~ x,
    data = . %&gt;% dplyr::filter(step == &quot;pre&quot;),
    colour = &quot;blue&quot;,
    show.legend = FALSE
  ) +
  labs(
    x = &quot;logCPM&quot;, y = expression(hat(sigma)), colour = &quot;Analytic\nStep&quot;
  ) +
  scale_colour_manual(values = c(pre = &quot;black&quot;, post = &quot;red&quot;)) +
  theme(axis.title.y = element_text(angle = 0, vjust = 0.5))</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-mean-var-1.png" alt="*Mean-variance for all retained windows. Black dots indicate the estimated standard deviation for each window as a function of signal. Red dots are the shrunken estimates after applying limma-trend, whilst the loess curve through the initial estimates is shown in blue.*" width="960" />
<p class="caption">
<em>Mean-variance for all retained windows. Black dots indicate the estimated standard deviation for each window as a function of signal. Red dots are the shrunken estimates after applying limma-trend, whilst the loess curve through the initial estimates is shown in blue.</em>
</p>
</div>
</div>
</div>
<div id="independent-hypothesis-weighting" class="section level2 tabset">
<h2 class="tabset">Independent Hypothesis Weighting</h2>
<pre class="r"><code>ihw_opts &lt;- c(&quot;regions&quot;, &quot;features&quot;, &quot;targets&quot;)
ihw_method &lt;- c(
  intersect(str_to_lower(config$comparisons$ihw), ihw_opts), &quot;none&quot;
)
ihw_method &lt;- ihw_method[[1]]
fdr_column &lt;- str_subset(names(mcols(merged_results)), &quot;fdr|FDR&quot;)
## Add status (direction) for the ihw = &quot;none&quot; possibility
merged_results &lt;- merged_results %&gt;% 
  mutate(
    status = case_when(
      !!sym(fdr_column) &lt; fdr_alpha ~ as.character(direction),
      !!sym(fdr_column) &gt;= fdr_alpha ~ &quot;Unchanged&quot;
    ) %&gt;% 
      factor(levels = str_to_title(names(colours$direction))) %&gt;% 
      fct_drop(only = &quot;Undetected&quot;)
  )</code></pre>
<pre class="r"><code>ihw_all &lt;- gene_regions %&gt;% 
  .[names(.) != &quot;tss&quot;] %&gt;% 
  lapply(
    function(x) propOverlap(merged_results, x)
  ) %&gt;% 
  as_tibble() %&gt;% 
  mutate(
    range = as.character(merged_results), id = seq_along(range)
  ) %&gt;% 
  pivot_longer(
    cols = any_of(names(gene_regions)),
    names_to = &quot;region&quot;,
    values_to = &quot;prop&quot;
  ) %&gt;% 
  group_by(id, range) %&gt;% 
  dplyr::filter(prop == max(prop)) %&gt;% 
  ungroup() %&gt;% 
  distinct(id, range, .keep_all = TRUE) %&gt;% 
  mutate(
    region = factor(region, levels = names(colours$region))
  ) %&gt;% 
  arrange(id) %&gt;% 
  .[[&quot;region&quot;]]
n_levels &lt;- sum(table(ihw_all) &gt; 1e3) - 1
merged_results$ihw_covariate &lt;- fct_lump_n(ihw_all, n = n_levels)
ihw_proceed &lt;- length(levels(merged_results$ihw_covariate)) &gt; 1
if (ihw_proceed)  fdr_column &lt;- &quot;fdr_ihw&quot;
ihw_cols &lt;- colours$regions
if (ihw_proceed) {
  ihw &lt;- ihw(
    pvalues = merged_results$P.Value,
    covariates = merged_results$ihw_covariate,
    alpha &lt;- fdr_alpha,
    covariate_type = &quot;nominal&quot;
  )
  merged_results &lt;- mutate(merged_results, fdr_ihw = adj_pvalues(ihw))
} 
merged_results &lt;- merged_results %&gt;% 
  mutate(
    status = case_when(
      !!sym(fdr_column) &lt; fdr_alpha ~ as.character(direction),
      !!sym(fdr_column) &gt;= fdr_alpha ~ &quot;Unchanged&quot;
    ) %&gt;% 
      factor(levels = str_to_title(names(colours$direction))) %&gt;% 
      fct_drop(only = &quot;Undetected&quot;)
  )</code></pre>
<p>Independent Hypothesis Weighting (IHW) was then used to partition the raw p-values for ER differential binding by genomic regions defined previously, and estimate weights to apply to each p-value as a divisor. This allows recalculation of the FDR using <em>weighted p-values</em> instead of <em>raw p-values</em>. In order for IHW to be a viable strategy, partitions should be greater than 1000. The provided genomic regions were used as initial partitions, merging the smallest groups below this size until all partitions were of a suitable size.</p>
<p>Windows were classified as overlapping a region if any section of the window overlapped the region Where multiple overlapping regions were found, the region with the <em>largest</em> overlap was selected.</p>
<p>Only one partition &gt; 1000 was able to be formed and IHW was not executed.</p>
<div id="initial-group-sizes" class="section level3">
<h3>Initial Group Sizes</h3>
<pre class="r"><code>tibble(ihw = ihw_all) %&gt;% 
  mutate(
    ihw_label = regions[as.character(ihw)] %&gt;% 
      fct_infreq()
  ) %&gt;% 
  ggplot(
    aes(ihw_label, fill = ihw)
  ) +
  geom_bar() +
  geom_hline(yintercept = 1e3, colour = &quot;blue&quot;) +
  geom_text(
    aes(y = n + 0.025*max(n), label = comma(n, 1)),
    data = . %&gt;%
      group_by(ihw_label, ihw) %&gt;%
      summarise(n = dplyr::n(), .groups = &quot;drop&quot;) 
  ) +
  scale_y_continuous(expand = expansion(c(0, 0.05)), labels = comma) +
  scale_fill_manual(values = ihw_cols) +
  labs(x = str_to_title(ihw_method), y = &quot;Total&quot;) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/ihw-partitions-raw-1.png" alt="*Breakdown of all windows which overlapped the regions previously defined. Any partitions with fewer than 1000 windows (indicated as the blue horizontal line) were combined into the next smallest partition consecutively, until all partitions contained &gt; 1000 windows.*" width="960" />
<p class="caption">
<em>Breakdown of all windows which overlapped the regions previously defined. Any partitions with fewer than 1000 windows (indicated as the blue horizontal line) were combined into the next smallest partition consecutively, until all partitions contained &gt; 1000 windows.</em>
</p>
</div>
</div>
<div id="p-value-distributions" class="section level3">
<h3>P-Value Distributions</h3>
<pre class="r"><code>lblr &lt;- as_labeller(c(regions, &quot;Other&quot; = &quot;Other&quot;))
y &lt;- ggplot_build(
  tibble(P.Value = merged_results$P.Value,) %&gt;% 
    ggplot(aes(P.Value, stat(density))) + 
    geom_histogram(bins = 100) 
) %&gt;% .[[&quot;data&quot;]] %&gt;% 
  .[[1]] %&gt;% 
  .[[&quot;y&quot;]] %&gt;% 
  max()
merged_results %&gt;% 
  select(P.Value, ihw_covariate) %&gt;% 
  mcols() %&gt;% 
  as_tibble() %&gt;% 
  ggplot(aes(P.Value, stat(density))) +
  geom_histogram(fill = &quot;grey&quot;, colour = &quot;black&quot;, bins = 100) +
  geom_text(
    aes(0.8, y, label = lab),
    data = . %&gt;% 
      group_by(ihw_covariate) %&gt;% 
      summarise(n = dplyr::n(), .groups = &quot;drop&quot;) %&gt;% 
      mutate(
        lab = glue(&quot;N = {comma(n, 1)}&quot;),
        y = y
      )
  ) +
  facet_wrap(~ihw_covariate, ncol = 3, labeller = lblr) +
  scale_y_continuous(expand = expansion(c(0, 0.05))) +
  labs(x = &quot;P Value&quot;, y = &quot;Density&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-ihw-pvals-1.png" alt="*P-Value distributions within all final data partitions. The size of each parition is given within each panel.*" width="960" />
<p class="caption">
<em>P-Value distributions within all final data partitions. The size of each parition is given within each panel.</em>
</p>
</div>
</div>
<div id="ihw-weights" class="section level3">
<h3>IHW Weights</h3>
<pre class="r"><code>plot(ihw) +
  geom_hline(yintercept = 1) +
  facet_wrap(~group, labeller =lblr) +
  scale_y_continuous(expand = expansion(c(0, 0.05))) +
  labs(x = &quot;Fold&quot;, y = &quot;Weight&quot;, fill = &quot;Fold&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-ihw-weights-1.png" alt="*Weights applied to p-values within each partition. 'Folds' represent random sub-partitions within each larger partition generated as part of the IHW process.*" width="960" />
<p class="caption">
<em>Weights applied to p-values within each partition. ‘Folds’ represent random sub-partitions within each larger partition generated as part of the IHW process.</em>
</p>
</div>
</div>
<div id="fdr-comparison" class="section level3">
<h3>FDR Comparison</h3>
<pre class="r"><code>ihw %&gt;%
  as.data.frame() %&gt;% 
  ggplot(
    aes(-log10(pvalue), -log10(weighted_pvalue), colour = fold)
  ) +
  geom_point(size = 0.4) +
  geom_hline(
    aes(yintercept = -log10(weighted_pvalue)),
    data = . %&gt;% 
      dplyr::select(-group) %&gt;% 
      dplyr::filter(adj_pvalue &lt; fdr_alpha) %&gt;% 
      dplyr::filter(weighted_pvalue == max(weighted_pvalue)),
    linetype = 2, colour = &quot;blue&quot;
  ) +
  geom_vline(
    aes(xintercept = -log10(pvalue)),
    data =. %&gt;% 
      dplyr::filter(p.adjust(pvalue, &quot;fdr&quot;) &lt; fdr_alpha) %&gt;% 
      dplyr::filter(pvalue == max(pvalue)) %&gt;% 
      dplyr::select(-group),
    linetype = 2, colour = &quot;blue&quot;
  ) +
  facet_wrap(~group, scales = &quot;free&quot;, labeller = lblr) +
  coord_cartesian(xlim = c(0, 5), ylim = c(0, 5)) +
  labs(
    x = expression(-log[10](p)),
    y = expression(-log[10](p[IHW])),
    colour = &quot;Fold&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-fdr-comparison-1.png" alt="*Comparison of raw and weighted p-values for each partition. Blue dashed lines indicate FDR = 0.05 for each set of p-values. Those in the lower-right qudrant would no longer be considered significant after IHW, whilst those in the upper-left quadrant would only be considered as significant after the IHW process. Those in the upper-right quadrant would be considered as significant regardless of the methodology.*" width="960" />
<p class="caption">
<em>Comparison of raw and weighted p-values for each partition. Blue dashed lines indicate FDR = 0.05 for each set of p-values. Those in the lower-right qudrant would no longer be considered significant after IHW, whilst those in the upper-left quadrant would only be considered as significant after the IHW process. Those in the upper-right quadrant would be considered as significant regardless of the methodology.</em>
</p>
</div>
</div>
<div id="summary-table" class="section level3">
<h3>Summary Table</h3>
<pre class="r"><code>ihw %&gt;%
  as.data.frame() %&gt;% 
  mutate(
    fdr = p.adjust(pvalue, &quot;fdr&quot;),
    status = case_when(
      fdr &lt; fdr_alpha &amp; adj_pvalue &gt; fdr_alpha ~ &quot;Lost&quot;,
      fdr &lt; fdr_alpha &amp; adj_pvalue &lt; fdr_alpha ~ &quot;Retained&quot;,
      fdr &gt; fdr_alpha &amp; adj_pvalue &lt; fdr_alpha ~ &quot;Gained&quot;,
      fdr &gt; fdr_alpha &amp; adj_pvalue &gt; fdr_alpha ~ &quot;Never Significant&quot;
    ) %&gt;% 
      factor(
        levels = c(&quot;Never Significant&quot;, &quot;Retained&quot;, &quot;Gained&quot;, &quot;Lost&quot;)
      )
  ) %&gt;% 
  as_tibble() %&gt;% 
  group_by(covariate, status) %&gt;% 
  tally() %&gt;% 
  ungroup() %&gt;% 
  complete(status = status, covariate = covariate, fill = list(n = 0)) %&gt;% 
  pivot_wider(
    names_from = status, values_from = n, values_fill = 0
  ) %&gt;% 
  mutate(
    covariate = unlist(lblr(covariate)),
    `Nett Change` = Gained - Lost
  ) %&gt;% 
  bind_rows(
    summarise_if(., is.numeric, sum)
  ) %&gt;% 
  mutate(
    covariate = str_replace_na(covariate, &quot;Total&quot;),
    Significant = Retained + Gained
  ) %&gt;% 
  dplyr::select(
    Covariate = covariate, ends_with(&quot;Significant&quot;), Retained, everything()
  ) %&gt;% 
  pander(
    justify = &quot;lrrrrrr&quot;,
    caption = paste(
      &quot;*Summary of changes introduced by IHW for windows considered as&quot;,
      paste0(&quot;differentially bound by &quot;, target, &quot;.&quot;), 
      &quot;This corresponds to a nett change of&quot;, 
        dplyr::slice(., nrow(.)) %&gt;% 
        mutate(p = `Nett Change` / (Significant - `Nett Change`)) %&gt;% 
        .[[&quot;p&quot;]] %&gt;% 
        percent(0.1),
      &quot;from the initial list.*&quot;
    ),
    emphasize.strong.rows = nrow(.)
  )</code></pre>
<table>
<caption><em>Summary of changes introduced by IHW for windows considered as differentially bound by ER. This corresponds to a nett change of 0.3% from the initial list.</em></caption>
<colgroup>
<col width="13%" />
<col width="22%" />
<col width="15%" />
<col width="12%" />
<col width="10%" />
<col width="9%" />
<col width="15%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Covariate</th>
<th align="right">Never Significant</th>
<th align="right">Significant</th>
<th align="right">Retained</th>
<th align="right">Gained</th>
<th align="right">Lost</th>
<th align="right">Nett Change</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Exon</td>
<td align="right">6</td>
<td align="right">30</td>
<td align="right">30</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">-1</td>
</tr>
<tr class="even">
<td align="left">Other</td>
<td align="right">283</td>
<td align="right">912</td>
<td align="right">908</td>
<td align="right">4</td>
<td align="right">0</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left"><strong>Total</strong></td>
<td align="right"><strong>289</strong></td>
<td align="right"><strong>942</strong></td>
<td align="right"><strong>938</strong></td>
<td align="right"><strong>4</strong></td>
<td align="right"><strong>1</strong></td>
<td align="right"><strong>3</strong></td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="mapping-windows" class="section level2">
<h2>Mapping Windows</h2>
<pre class="r"><code>merged_results &lt;- merged_results %&gt;% 
  mutate(
    region = bestOverlap(
      ., 
      gene_regions %&gt;% 
        lapply(select, region) %&gt;% 
        GRangesList() %&gt;% 
        unlist(),
      var = &quot;region&quot;
    ) %&gt;% 
      factor(levels = regions)
  )</code></pre>
<pre class="r"><code>merged_results &lt;- merged_results %&gt;% 
  mutate(
    feature = bestOverlap(
      ., external_features, &quot;feature&quot;, missing = &quot;no_feature&quot;
    ) %&gt;% 
      factor(levels = names(colours$features)) %&gt;% 
      fct_relabel(str_sep_to_title)
  )</code></pre>
<p>Merged windows were first mapped to the gene-centric region with the largest overlap. Windows were then mapped to the external features provided in the same manner.</p>
<pre class="r"><code>merged_results &lt;- merged_results %&gt;%
  mapByFeature(
    genes = genes_gr, 
    prom = GRangesList(
      ## Add any external features which appear to be promoters
      ifelse(
        &quot;feature&quot; %in% colnames(mcols(external_features)), 
        list(granges(subset(external_features, str_detect(feature, &quot;[Pp]rom&quot;)))), 
        list(GRanges())
      )[[1]], 
      granges(gene_regions$promoters)) %&gt;% 
      unlist() %&gt;% 
      reduce(),
    enh = ifelse(
      &quot;feature&quot; %in% colnames(mcols(external_features)), 
      list(granges(subset(external_features, str_detect(feature, &quot;[Ee]nhanc&quot;)))), 
      list(GRanges())
    )[[1]],
    gi = hic
  ) %&gt;% 
  mutate(
    detected = lapply(
      gene_id,
      function(x) {
        id2gene[intersect(x, rnaseq$gene_id)]
      }
    ) %&gt;% 
      as(&quot;CharacterList&quot;)
  )</code></pre>
<p>In addition to statistical analysis, the regions produced by merging windows were mapped to all annotated genes. Promoters were the union of all regions defined based on transcript-level annotation, and any external features which were defined as promoters. Enhancers were any regions defined in h3k27ac_features.gtf as enhancers. The set of 1338 HiC-interactions were also used to define long-range interactions.</p>
<p>These features were used to map merged windows to features using the process defined in the function <code>mapByFeature()</code>:</p>
<ol style="list-style-type: decimal">
<li>Ranges <em>overlapping a promoter</em> are assigned to genes for that specific promoter</li>
<li>Ranges <em>overlapping an enhancer</em> are assigned to all genes within 100kb</li>
<li>Ranges <em>overlapping a long-range interaction</em> are assigned to all genes connected by the interaction</li>
<li>Ranges with <em>no gene assignment from the previous steps</em> are assigned to all overlapping genes, or the nearest gene within 100kb</li>
</ol>
<p>Notably, genes are only passed to step 4 if no gene assignment has been made in steps 1, 2 or 3. For visualisation purposes, only genes which were considered as detected in any provided RNA-Seq data will be shown as the mapping targets.</p>
</div>
</div>
<div id="results" class="section level1">
<h1>Results</h1>
<div id="summary-plots" class="section level2 tabset">
<h2 class="tabset">Summary Plots</h2>
<pre class="r"><code>profile_width &lt;- 5e3
n_bins &lt;- 100
bwfl &lt;- file.path(bw_path, glue(&quot;{treat_levels}_merged_treat_pileup.bw&quot;)) %&gt;% 
  BigWigFileList() %&gt;% 
  setNames(treat_levels) 
sig_ranges &lt;- merged_results %&gt;% 
  filter(!!sym(fdr_column) &lt; fdr_alpha) %&gt;% 
  colToRanges(&quot;keyval_range&quot;) %&gt;% 
  splitAsList(f = .$direction) %&gt;% 
  .[vapply(., length, integer(1)) &gt; 0] 
sig_profiles &lt;- bplapply(
  sig_ranges, 
  function(x) getProfileData(
    bwfl, x, upstream = profile_width / 2, bins = n_bins
  ),
  BPPARAM = MulticoreParam(length(sig_ranges))
)
profile_heatmaps &lt;- sig_profiles %&gt;% 
  bplapply(
    plotProfileHeatmap,
    profileCol = &quot;profile_data&quot;, 
    facetY = &quot;overlaps_ref&quot;,
    colour = &quot;name&quot;,
    linetype = &quot;overlaps_ref&quot;,
    xLab = &quot;Distance from Centre (bp)&quot;,
    fillLab = &quot;CPM&quot;,
    labeller = as_labeller(
        c(
          setNames(treat_levels, treat_levels),
          `TRUE` = &quot;Consensus\nPeak&quot;,
          `FALSE` = &quot;No\nConsensus\nPeak&quot;
        )
    ),
    BPPARAM = MulticoreParam(length(sig_ranges))
  ) 
fill_range &lt;- profile_heatmaps %&gt;% 
  lapply(function(x) x$data[,&quot;score&quot;]) %&gt;% 
  unlist() %&gt;%
  range()
sidey_range &lt;- profile_heatmaps %&gt;% 
  lapply(function(x) x$layers[[3]]$data$y) %&gt;% 
  unlist() %&gt;% 
  range()
profile_heatmaps &lt;- profile_heatmaps %&gt;% 
  lapply(function(x) {
    x + 
      scale_fill_viridis_c(
        option = &quot;inferno&quot;, direction = -1, limits = fill_range
      ) +
      scale_xsidey_continuous(limits = sidey_range) +
      scale_colour_manual(values = treat_colours) +
      scale_linetype_manual(values = c(&quot;TRUE&quot; = 1, &quot;FALSE&quot; = 2)) +
      labs(
        x = &quot;Distance from Centre (bp)&quot;,
        fill = &quot;CPM&quot;, linetype = &quot;Consensus\nPeak\nOverlap&quot;,
        colour = &quot;Treat&quot;
      ) +
      theme(
        strip.text.y = element_text(angle = 0)
      )
  }
  )</code></pre>
<div id="ma-plot" class="section level3">
<h3>MA Plot</h3>
<pre class="r"><code>merged_results %&gt;% 
  as_tibble() %&gt;%
  ggplot(aes(AveExpr, logFC)) +
  geom_point(aes(colour = status), alpha = 0.5) +
  geom_smooth(se = FALSE, method = &quot;gam&quot;, formula = y ~ s(x, bs = &quot;cs&quot;)) +
  geom_hline(yintercept = 0) +
  geom_label_repel(
    aes(y = 1.1*logFC, label = annot, colour = status),
    data= . %&gt;%
      dplyr::filter(logFC == max(logFC) | logFC == min(logFC)) %&gt;%
      mutate(
        detected = detected %&gt;%
          lapply(paste, collapse = &quot;, &quot;) %&gt;%
          unlist() %&gt;%
          str_replace(&quot;^NA$&quot;, &quot;&quot;) %&gt;%
          str_wrap(width = 40),
        range = ifelse(detected == &quot;&quot;, range, paste0(range, &quot;\n&quot;)),
        annot = glue(&quot;{range}{detected}&quot;)
      ),
    fill = rgb(1, 1, 1, 0.2),
    size = 3,
    show.legend = FALSE
  ) +
  facet_grid(
    .~overlaps_ref, 
    scales = &quot;free_x&quot;, space = &quot;free_x&quot;,
    labeller = facet_labeller
  ) +
  scale_colour_manual(values = direction_colours) +
  labs(
    x = &quot;Ave Signal (logCPM)&quot;,
    colour = &quot;Status&quot;
  ) </code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-ma-1.png" alt="*MA plot showing the status of each window under consideration. The two most extreme regions are labelled by region and any associated genes, whilst the overall pattern of association between signal level (logCPM) and changed signal (logFC) is shown as the blue curve. Windows are plotted in separate panels based on those which overlap a `macs2` consensus peak.*" width="960" />
<p class="caption">
<em>MA plot showing the status of each window under consideration. The two most extreme regions are labelled by region and any associated genes, whilst the overall pattern of association between signal level (logCPM) and changed signal (logFC) is shown as the blue curve. Windows are plotted in separate panels based on those which overlap a <code>macs2</code> consensus peak.</em>
</p>
</div>
</div>
<div id="volcano-plot" class="section level3">
<h3>Volcano Plot</h3>
<pre class="r"><code>merged_results %&gt;% 
  as_tibble() %&gt;% 
  ggplot(aes(logFC, -log10(P.Value))) +
  geom_point(aes(colour = status), alpha = 0.6) +
  geom_label_repel(
    aes(label = annot, colour = status),
    data= . %&gt;%
      arrange(sign(logFC)*log10(P.Value)) %&gt;% 
      dplyr::slice(1:2) %&gt;% 
      dplyr::filter(logFC &gt; 1) %&gt;%
      mutate(
        detected = detected %&gt;%
          lapply(paste, collapse = &quot;, &quot;) %&gt;%
          unlist() %&gt;%
          str_replace(&quot;^NA$&quot;, &quot;&quot;) %&gt;%
          str_wrap(width = 40),
        range = ifelse(detected == &quot;&quot;, range, paste0(range, &quot;\n&quot;)),
        annot = glue(&quot;{range}{detected}&quot;)
      ),
    fill = rgb(1, 1, 1, 0.2),
    size = 3,
    show.legend = FALSE
  ) +
  geom_label_repel(
    aes(label = annot, colour = status),
    data= . %&gt;%
      arrange(-sign(logFC)*log10(P.Value)) %&gt;% 
      dplyr::slice(1:2) %&gt;% 
      dplyr::filter(logFC &lt; -1) %&gt;% 
      mutate(
        detected = detected %&gt;%
          lapply(paste, collapse = &quot;, &quot;) %&gt;%
          unlist() %&gt;%
          str_replace(&quot;^NA$&quot;, &quot;&quot;) %&gt;%
          str_wrap(width = 40),
        range = ifelse(detected == &quot;&quot;, range, paste0(range, &quot;\n&quot;)),
        annot = glue(&quot;{range}{detected}&quot;)
      ),
    fill = rgb(1, 1, 1, 0.2),
    size = 3,
    show.legend = FALSE
  ) +
  scale_colour_manual(values = direction_colours) +
  labs(
    x = &quot;logFC&quot;,
    y = expression(paste(-log[10], &quot;p&quot;)),
    colour = &quot;Status&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-volcano-1.png" alt="*Volcano plot showing regions with evidence of differential ER binding. Regions overlapping a consensus peak are coloured separately to those which are not. The most significant regions are labelled along with any genes these regions are mapped to.*" width="960" />
<p class="caption">
<em>Volcano plot showing regions with evidence of differential ER binding. Regions overlapping a consensus peak are coloured separately to those which are not. The most significant regions are labelled along with any genes these regions are mapped to.</em>
</p>
</div>
</div>
<div id="heatmaps-up" class="section level3">
<h3>Heatmaps: Up</h3>
<pre class="r"><code>profile_heatmaps$Up</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-up-heatmap-1.png" alt="*Heatmap and histogram for all regions considered to show evidence of increased ER binding in the DHT group, separated by those which did or did not overlap a consensus peak. A total of 4 regions were in this group.*" width="960" />
<p class="caption">
<em>Heatmap and histogram for all regions considered to show evidence of increased ER binding in the DHT group, separated by those which did or did not overlap a consensus peak. A total of 4 regions were in this group.</em>
</p>
</div>
</div>
<div id="heatmaps-down" class="section level3">
<h3>Heatmaps: Down</h3>
<pre class="r"><code>profile_heatmaps$Down</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-down-heatmap-1.png" alt="*Heatmap and histogram for all regions considered to show evidence of decreased ER binding in the DHT group, separated by those which did or did not overlap a consensus peak. A total of 938 regions were in this group.*" width="960" />
<p class="caption">
<em>Heatmap and histogram for all regions considered to show evidence of decreased ER binding in the DHT group, separated by those which did or did not overlap a consensus peak. A total of 938 regions were in this group.</em>
</p>
</div>
</div>
<div id="results-by-chromosome" class="section level3">
<h3>Results By Chromosome</h3>
<pre class="r"><code>merged_results %&gt;% 
  select(overlaps_ref, status) %&gt;% 
  as.data.frame() %&gt;% 
  mutate(status = str_remove(as.character(status), &quot; \\+ Ref Peak&quot;)) %&gt;% 
  ggplot(aes(seqnames, fill = status, alpha = overlaps_ref)) +
  geom_bar() +
  facet_grid(status~., scales = &quot;free_y&quot;) +
  scale_y_continuous(expand = expansion(c(0, 0.05)), labels = comma) +
  scale_fill_manual(values =  direction_colours) +
  scale_alpha_manual(values = c(0.5, 1)) +
  labs(
    x = &quot;Chromosome&quot;, y = &quot;Number of Windows&quot;, 
    fill = &quot;Status&quot;, alpha = &quot;Macs2 Peak&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-res-by-chromosome-1.png" alt="*Results for differential binding separated by chromosome*" width="960" />
<p class="caption">
<em>Results for differential binding separated by chromosome</em>
</p>
</div>
</div>
<div id="signal-by-region" class="section level3">
<h3>Signal By Region</h3>
<pre class="r"><code>merged_results %&gt;% 
  select(AveExpr, logFC, region) %&gt;% 
  as_tibble() %&gt;% 
  mutate(
    x = fct_relabel(
      region, str_replace_all, pattern = &quot; \\(&quot;, replacement = &quot;\n(&quot;
    ),
  ) %&gt;% 
  pivot_longer(cols = c(&quot;AveExpr&quot;, &quot;logFC&quot;)) %&gt;% 
  ggplot(aes(x, value, fill = region)) +
  geom_boxplot() +
  facet_grid(
    name~., scales = &quot;free&quot;, labeller = facet_labeller, switch = &quot;y&quot;
  ) +
  scale_fill_manual(values = region_colours) +
  labs(x = &quot;Region&quot;, y = c(), fill = &quot;Overlaps&quot;) </code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/boxplot-regions-1.png" alt="*Distributions of overall signal (logCPM) and changed signal (logFC) for each genomic region.*" width="960" />
<p class="caption">
<em>Distributions of overall signal (logCPM) and changed signal (logFC) for each genomic region.</em>
</p>
</div>
</div>
<div id="differential-binding-by-region" class="section level3">
<h3>Differential Binding By Region</h3>
<pre class="r"><code>merged_results %&gt;% 
  select(status, region) %&gt;% 
  as_tibble() %&gt;% 
  mutate(
    x = fct_relabel(
      region, str_replace_all, pattern = &quot; \\(&quot;, replacement = &quot;\n(&quot;
    )
  ) %&gt;% 
  group_by(x, region, status) %&gt;% 
  tally() %&gt;% 
  arrange(region, desc(status)) %&gt;% 
  mutate(y = cumsum(n)) %&gt;% 
  ggplot(
    aes(x, n, fill = status)
  ) +
  geom_col() +
  geom_label(
    aes(y = y - 0.5*n, label = comma(n, 1), colour = status),
    data = . %&gt;% dplyr::filter(n &gt; 0.2*max(y)),
    fill = &quot;white&quot;, alpha = 0.8,
    show.legend = FALSE
  ) +
  geom_text(
    aes(y = y1, label = comma(y, 1)),
    data = . %&gt;% 
      dplyr::filter(y == max(y)) %&gt;% 
      ungroup() %&gt;% 
      mutate(y1 = y + 0.03*max(y))
  ) +
  scale_y_continuous(expand = expansion(c(0, 0.5)), labels = comma) +
  scale_fill_manual(values = direction_colours) +
  scale_colour_manual(values = direction_colours) +
  labs(x = &quot;Region&quot;, y = &quot;Merged Windows&quot;, fill = &quot;Status&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-db-regions-1.png" alt="*Merged windows considered as showing differential ER binding across all genomic regions.*" width="960" />
<p class="caption">
<em>Merged windows considered as showing differential ER binding across all genomic regions.</em>
</p>
</div>
</div>
<div id="signal-by-feature" class="section level3">
<h3>Signal By Feature</h3>
<pre class="r"><code>merged_results %&gt;% 
  select(AveExpr, logFC, feature) %&gt;% 
  as_tibble() %&gt;% 
  pivot_longer(cols = c(&quot;AveExpr&quot;, &quot;logFC&quot;)) %&gt;% 
  ggplot(aes(feature, value, fill = feature)) +
  geom_boxplot() +
  facet_grid(
    name~., scales = &quot;free&quot;, labeller = facet_labeller, switch = &quot;y&quot;
  ) +
  scale_fill_manual(values = feature_colours) +
  labs(x = &quot;Region&quot;, y = c(), fill = &quot;Overlaps&quot;) </code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/boxplot-features-1.png" alt="*Distributions of overall signal (logCPM) and changed signal (logFC) for each external features.*" width="960" />
<p class="caption">
<em>Distributions of overall signal (logCPM) and changed signal (logFC) for each external features.</em>
</p>
</div>
</div>
<div id="differential-binding-by-feature" class="section level3">
<h3>Differential Binding By Feature</h3>
<pre class="r"><code>merged_results %&gt;% 
  select(status, feature) %&gt;% 
  as_tibble() %&gt;% 
  group_by(feature, status) %&gt;% 
  tally() %&gt;% 
  arrange(feature, desc(status)) %&gt;% 
  mutate(y = cumsum(n)) %&gt;% 
  ggplot(
    aes(feature, n, fill = status)
  ) +
  geom_col() +
  geom_label(
    aes(y = y - 0.5*n, label = comma(n, 1), colour = status),
    data = . %&gt;% dplyr::filter(n &gt; 0.2*max(y)),
    fill = &quot;white&quot;, alpha = 0.8,
    show.legend = FALSE
  ) +
  geom_text(
    aes(y = y1, label = comma(y, 1)),
    data = . %&gt;% 
      dplyr::filter(y == max(y)) %&gt;% 
      ungroup() %&gt;% 
      mutate(y1 = y + 0.03*max(y))
  ) +
  scale_y_continuous(expand = expansion(c(0, 0.5)), labels = comma) +
  scale_fill_manual(values = direction_colours) +
  scale_colour_manual(values = direction_colours) +
  labs(x = &quot;Feature&quot;, y = &quot;Merged Windows&quot;, fill = &quot;Status&quot;)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-db-features-1.png" alt="*Merged windows considered as showing differential ER binding across all external features.*" width="960" />
<p class="caption">
<em>Merged windows considered as showing differential ER binding across all external features.</em>
</p>
</div>
</div>
<div id="differential-binding-by-region-and-feature" class="section level3">
<h3>Differential Binding By Region And Feature</h3>
<pre class="r"><code>merged_results %&gt;% 
  mutate(region = fct_relabel(region, str_wrap, width = 15)) %&gt;% 
  plotPie(x = &quot;region&quot;, y = &quot;feature&quot;, fill = &quot;status&quot;) +
  scale_fill_manual(values = direction_colours) +
  labs(
    x = &quot;Region&quot;, y = &quot;Feature&quot;, fill = &quot;Status&quot;
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/scatterpie-region-feature-1.png" alt="*Comparison of changes in ER binding separated by gene-centric region and external features provided in h3k27ac_features.gtf*" width="960" />
<p class="caption">
<em>Comparison of changes in ER binding separated by gene-centric region and external features provided in h3k27ac_features.gtf</em>
</p>
</div>
</div>
</div>
<div id="result-tables" class="section level2 tabset">
<h2 class="tabset">Result Tables</h2>
<pre class="r"><code>up_col &lt;- function(x) {
  rgb(
    colorRamp(c(&quot;#ffffff&quot;, colours$direction[[&quot;up&quot;]]))(x), maxColorValue = 255
  )
}
down_col &lt;- function(x) {
  rgb(
    colorRamp(c(&quot;#ffffff&quot;, colours$direction[[&quot;down&quot;]]))(x), maxColorValue = 255
  )
}
unch_col &lt;- function(x) {
  rgb(
    colorRamp(c(&quot;#ffffff&quot;, colours$direction[[&quot;unchanged&quot;]]))(x), 
    maxColorValue = 255
  )
}
lfc_col &lt;- function(x){
  rgb(
    colorRamp(c(colours$direction[[&quot;down&quot;]], &quot;#ffffff&quot;, colours$direction[[&quot;up&quot;]]))(x), 
    maxColorValue = 255
  )
}
expr_col &lt;- function(x){
  rgb(colorRamp(hcl.colors(9, &quot;TealRose&quot;))(x), maxColorValue = 255)
}
bar_style &lt;- function(width = 1, fill = &quot;#e6e6e6&quot;, height = &quot;75%&quot;, align = c(&quot;left&quot;, &quot;right&quot;), color = NULL) {
  align &lt;- match.arg(align)
  if (align == &quot;left&quot;) {
    position &lt;- paste0(width * 100, &quot;%&quot;)
    image &lt;- sprintf(&quot;linear-gradient(90deg, %1$s %2$s, transparent %2$s)&quot;, fill, position)
  } else {
    position &lt;- paste0(100 - width * 100, &quot;%&quot;)
    image &lt;- sprintf(&quot;linear-gradient(90deg, transparent %1$s, %2$s %1$s)&quot;, position, fill)
  }
  list(
    backgroundImage = image,
    backgroundSize = paste(&quot;100%&quot;, height),
    backgroundRepeat = &quot;no-repeat&quot;,
    backgroundPosition = &quot;center&quot;,
    color = color
  )
}
with_tooltip &lt;- function(value, width = 30) {
  tags$span(title = value, str_trunc(value, width))
}</code></pre>
<div id="overall-results" class="section level3">
<h3>Overall Results</h3>
<pre class="r"><code>df &lt;- merged_results %&gt;% 
  as_tibble() %&gt;% 
  group_by(overlaps_ref, status) %&gt;% 
  summarise(n = dplyr::n(), .groups = &quot;drop&quot;) %&gt;% 
  complete(overlaps_ref, status, fill = list(n = 0)) %&gt;% 
  group_by(overlaps_ref) %&gt;% 
  mutate(Total = sum(n)) %&gt;%
  ungroup() %&gt;% 
  pivot_wider(names_from = &quot;status&quot;, values_from = &quot;n&quot;, values_fill = 0) %&gt;% 
  mutate(
    overlaps_ref = unlist(facet_labeller(overlaps_ref)),
    `% Changed` = 1 - Unchanged / Total
  ) %&gt;% 
  dplyr::rename(`Macs2 Peak` = overlaps_ref) 
cp &lt;- htmltools::em(
  glue(
    &quot;Overall results for changed {target} binding after {treat_levels[[2]]} &quot;,
    &quot;treament, broken down by whether each merged window overlaps a macs2-&quot;,
    &quot;derived consensus peak.&quot;
  )
)
tbl &lt;- df %&gt;%
  reactable(
    columns = list(
      `Macs2 Peak` = colDef(footer = htmltools::tags$b(&quot;Total&quot;)),
      Up = colDef(
        cell = function(value) comma(value, 1),
        style = function(value) {
          normalized &lt;- value / (length(merged_results) - sum(df$Unchanged))
          color &lt;- up_col(normalized)
          list(background = color)
        },
        footer = function(values) htmltools::tags$b(comma(sum(values)))
      ),
      Down = colDef(
        cell = function(value) comma(value, 1),
        style = function(value) {
          normalized &lt;- value / (length(merged_results) - sum(df$Unchanged))
          color &lt;- down_col(normalized)
          list(background = color)
        },
        footer = function(values) htmltools::tags$b(comma(sum(values)))
      ),
      Unchanged = colDef(
        cell = function(value) comma(value, 1),
        style = function(value) {
          normalized &lt;- value / length(merged_results)#sum(df$Unchanged)
          color &lt;- unch_col(normalized)
          list(background = color)
        },
        footer = function(values) htmltools::tags$b(comma(sum(values)))
      ),
      `% Changed` = colDef(
        cell = function(value) percent(value, 0.1),
        style = function(value) bar_style(width = value, fill = &quot;#B3B3B3&quot;),
        align = &quot;left&quot;,
        footer = htmltools::tags$b(
          percent(sum(1 - sum(df$Unchanged) / sum(df$Total)), 0.1)
        )
      ),
      Total = colDef(
        cell = function(value) comma(value, 1),
        style = function(value) {
          bar_style(width = value / length(merged_results), fill = &quot;#B3B3B3&quot;)
        },
        align = &quot;left&quot;,
        footer = function(values) htmltools::tags$b(comma(sum(values)))
      )
    )
  )
div(class = &quot;table&quot;,
  div(class = &quot;table-header&quot;,
      div(class = &quot;caption&quot;, cp),
      tbl
  )
)</code></pre>
<div class="table">
<div class="table-header">
<div class="caption">
<em>Overall results for changed ER binding after DHT treament, broken down by whether each merged window overlaps a macs2-derived consensus peak.</em>
</div>
<div id="htmlwidget-dd17598ddf691e9835d9" class="reactable html-widget" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-dd17598ddf691e9835d9">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"Macs2 Peak":["No Consensus Peak","Consensus Peak"],"Total":[204,1028],"Up":[2,2],"Down":[63,875],"Unchanged":[139,151],"% Changed":[0.318627450980392,0.853112840466926]},"columns":[{"accessor":"Macs2 Peak","name":"Macs2 Peak","type":"character","footer":{"name":"b","attribs":{},"children":["Total"]}},{"accessor":"Total","name":"Total","type":"numeric","cell":["204","1,028"],"footer":{"name":"b","attribs":{},"children":["1,232"]},"align":"left","style":[{"backgroundImage":"linear-gradient(90deg, #B3B3B3 16.5584415584416%, transparent 16.5584415584416%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null},{"backgroundImage":"linear-gradient(90deg, #B3B3B3 83.4415584415584%, transparent 83.4415584415584%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null}]},{"accessor":"Up","name":"Up","type":"numeric","cell":["2","2"],"footer":{"name":"b","attribs":{},"children":["4"]},"style":[{"background":"#FEFEFE"},{"background":"#FEFEFE"}]},{"accessor":"Down","name":"Down","type":"numeric","cell":["63","875"],"footer":{"name":"b","attribs":{},"children":["938"]},"style":[{"background":"#EEF5F9"},{"background":"#167BB5"}]},{"accessor":"Unchanged","name":"Unchanged","type":"numeric","cell":["139","151"],"footer":{"name":"b","attribs":{},"children":["290"]},"style":[{"background":"#F0F0F0"},{"background":"#EFEFEF"}]},{"accessor":"% Changed","name":"% Changed","type":"numeric","cell":["31.9%","85.3%"],"footer":{"name":"b","attribs":{},"children":["76.5%"]},"align":"left","style":[{"backgroundImage":"linear-gradient(90deg, #B3B3B3 31.8627450980392%, transparent 31.8627450980392%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null},{"backgroundImage":"linear-gradient(90deg, #B3B3B3 85.3112840466926%, transparent 85.3112840466926%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null}]}],"defaultPageSize":10,"paginationType":"numbers","showPageInfo":true,"minRows":1,"dataKey":"f3eb87fd9abc25f1ad2cba0c3d172b2d","key":"f3eb87fd9abc25f1ad2cba0c3d172b2d"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</div>
<div id="summary-by-region" class="section level3">
<h3>Summary By Region</h3>
<pre class="r"><code>df &lt;- merged_results %&gt;% 
  select(
    overlaps_ref, status, all_of(fdr_column), region
  ) %&gt;% 
  as_tibble() %&gt;% 
  group_by(region, status) %&gt;% 
  summarise(n = dplyr::n(), .groups = &quot;drop&quot;) %&gt;% 
  complete(region, status, fill = list(n = 0)) %&gt;% 
  pivot_wider(names_from = &quot;status&quot;, values_from = &quot;n&quot;) %&gt;% 
  ungroup() %&gt;% 
  mutate(
    Total = Up + Down + Unchanged,
    `% Of All Windows` = Total / length(merged_results),
    `% Changed` = 1 - Unchanged / Total
  ) %&gt;% 
  arrange(region) %&gt;% 
  dplyr::select(
    Region = region,
    any_of(str_to_title(names(colours$direction))),
    `% Changed`, Total, `% Of All Windows`
  )
cp &lt;- htmltools::em(
  glue(
    &quot;Overall results for changed {target} binding after {treat_levels[[2]]} &quot;,
    &quot;treament, broken down by which genomic region contains the largest &quot;, 
    &quot;overlap with each merged window.&quot;
  )
)
tbl &lt;- df %&gt;%
  reactable(
    columns = list(
      Region = colDef(footer = htmltools::tags$b(&quot;Total&quot;)),
      Up = colDef(
        cell = function(value) comma(value, 1),
        style = function(value) {
          normalized &lt;- value / (length(merged_results) - sum(df$Unchanged))
          color &lt;- up_col(normalized)
          list(background = color)
        },
        footer = function(values) htmltools::tags$b(comma(sum(values)))
      ),
      Down = colDef(
        cell = function(value) comma(value, 1),
        style = function(value) {
          normalized &lt;- value / (length(merged_results) - sum(df$Unchanged))
          color &lt;- down_col(normalized)
          list(background = color)
        },
        footer = function(values) htmltools::tags$b(comma(sum(values)))
      ),
      Unchanged = colDef(
        cell = function(value) comma(value, 1),
        style = function(value) {
          normalized &lt;- value / length(merged_results)#sum(df$Unchanged)
          color &lt;- unch_col(normalized)
          list(background = color)
        },
        footer = function(values) htmltools::tags$b(comma(sum(values)))
      ),
      `% Changed` = colDef(
        cell = function(value) percent(value, 0.1),
        footer = htmltools::tags$b(
          percent(sum(1 - sum(df$Unchanged) / sum(df$Total)), 0.1)
        )
      ),
      Total = colDef(
        cell = function(value) comma(value, 1),
        style = function(value) {
          bar_style(width = value / length(merged_results), fill = &quot;#B3B3B3&quot;)
        },
        footer = function(values) htmltools::tags$b(comma(sum(values)))
      ),
      `% Of All Windows` = colDef(
        cell = function(value) percent(value, 0.1),
        footer = htmltools::tags$b(&quot;100%&quot;)
      )
    )
  )
div(class = &quot;table&quot;,
  div(class = &quot;table-header&quot;,
      div(class = &quot;caption&quot;, cp),
      tbl
  )
)</code></pre>
<div class="table">
<div class="table-header">
<div class="caption">
<em>Overall results for changed ER binding after DHT treament, broken down by which genomic region contains the largest overlap with each merged window.</em>
</div>
<div id="htmlwidget-01c1d1f4506f536c1299" class="reactable html-widget" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-01c1d1f4506f536c1299">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"Region":["Promoter (-1500/+500bp)","Upstream Promoter (<5kb)","Exon","Intron","Intergenic (<10kb)","Intergenic (>10kb)"],"Up":[0,0,0,2,1,1],"Down":[246,59,30,379,78,146],"Unchanged":[88,17,7,118,24,36],"% Changed":[0.736526946107784,0.776315789473684,0.810810810810811,0.763527054108216,0.766990291262136,0.80327868852459],"Total":[334,76,37,499,103,183],"% Of All Windows":[0.271103896103896,0.0616883116883117,0.0300324675324675,0.405032467532468,0.0836038961038961,0.148538961038961]},"columns":[{"accessor":"Region","name":"Region","type":"factor","footer":{"name":"b","attribs":{},"children":["Total"]}},{"accessor":"Up","name":"Up","type":"numeric","cell":["0","0","0","2","1","1"],"footer":{"name":"b","attribs":{},"children":["4"]},"style":[{"background":"#FFFFFF"},{"background":"#FFFFFF"},{"background":"#FFFFFF"},{"background":"#FEFEFE"},{"background":"#FEFEFE"},{"background":"#FEFEFE"}]},{"accessor":"Down","name":"Down","type":"numeric","cell":["246","59","30","379","78","146"],"footer":{"name":"b","attribs":{},"children":["938"]},"style":[{"background":"#BDD9EA"},{"background":"#EFF6FA"},{"background":"#F7FAFC"},{"background":"#9AC5DF"},{"background":"#EAF3F8"},{"background":"#D8E8F2"}]},{"accessor":"Unchanged","name":"Unchanged","type":"numeric","cell":["88","17","7","118","24","36"],"footer":{"name":"b","attribs":{},"children":["290"]},"style":[{"background":"#F5F5F5"},{"background":"#FDFDFD"},{"background":"#FEFEFE"},{"background":"#F2F2F2"},{"background":"#FCFCFC"},{"background":"#FBFBFB"}]},{"accessor":"% Changed","name":"% Changed","type":"numeric","cell":["73.7%","77.6%","81.1%","76.4%","76.7%","80.3%"],"footer":{"name":"b","attribs":{},"children":["76.5%"]}},{"accessor":"Total","name":"Total","type":"numeric","cell":["334","76","37","499","103","183"],"footer":{"name":"b","attribs":{},"children":["1,232"]},"style":[{"backgroundImage":"linear-gradient(90deg, #B3B3B3 27.1103896103896%, transparent 27.1103896103896%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null},{"backgroundImage":"linear-gradient(90deg, #B3B3B3 6.16883116883117%, transparent 6.16883116883117%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null},{"backgroundImage":"linear-gradient(90deg, #B3B3B3 3.00324675324675%, transparent 3.00324675324675%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null},{"backgroundImage":"linear-gradient(90deg, #B3B3B3 40.5032467532467%, transparent 40.5032467532467%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null},{"backgroundImage":"linear-gradient(90deg, #B3B3B3 8.36038961038961%, transparent 8.36038961038961%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null},{"backgroundImage":"linear-gradient(90deg, #B3B3B3 14.8538961038961%, transparent 14.8538961038961%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null}]},{"accessor":"% Of All Windows","name":"% Of All Windows","type":"numeric","cell":["27.1%","6.2%","3.0%","40.5%","8.4%","14.9%"],"footer":{"name":"b","attribs":{},"children":["100%"]}}],"defaultPageSize":10,"paginationType":"numbers","showPageInfo":true,"minRows":1,"dataKey":"e8aa61709c270af2a783d30f3575a384","key":"e8aa61709c270af2a783d30f3575a384"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</div>
<div id="summary-by-feature" class="section level3">
<h3>Summary By Feature</h3>
<pre class="r"><code>df &lt;- merged_results %&gt;% 
  select(
    overlaps_ref, status, all_of(fdr_column), feature
  ) %&gt;% 
  as_tibble() %&gt;% 
  group_by(feature, status) %&gt;% 
  tally() %&gt;% 
  ungroup() %&gt;% 
  complete(feature, status, fill = list(n = 0)) %&gt;% 
  pivot_wider(names_from = &quot;status&quot;, values_from = &quot;n&quot;) %&gt;% 
  mutate(
    Total = Up + Down + Unchanged,
    `% Of All Windows` = Total / length(merged_results),
    `% Changed` = 1 - Unchanged / Total
  ) %&gt;% 
  arrange(feature) %&gt;% 
  dplyr::select(
    Feature = feature,
    any_of(str_to_title(names(colours$direction))),
    `% Changed`, Total, `% Of All Windows`
  )
cp &lt;- htmltools::em(
  glue(
    &quot;Overall results for changed {target} binding after {treat_levels[[2]]} &quot;,
    &quot;treament, broken down by which external feature contains the largest &quot;, 
    &quot;overlap with each merged window.&quot;
  )
)
tbl &lt;- df %&gt;%
  reactable(
    columns = list(
      Feature = colDef(footer = htmltools::tags$b(&quot;Total&quot;)),
      Up = colDef(
        cell = function(value) comma(value, 1),
        style = function(value) {
          normalized &lt;- value / (length(merged_results) - sum(df$Unchanged))
          color &lt;- up_col(normalized)
          list(background = color)
        },
        footer = function(values) htmltools::tags$b(comma(sum(values)))
      ),
      Down = colDef(
        cell = function(value) comma(value, 1),
        style = function(value) {
          normalized &lt;- value / (length(merged_results) - sum(df$Unchanged))
          color &lt;- down_col(normalized)
          list(background = color)
        },
        footer = function(values) htmltools::tags$b(comma(sum(values)))
      ),
      Unchanged = colDef(
        cell = function(value) comma(value, 1),
        style = function(value) {
          normalized &lt;- value / length(merged_results)#sum(df$Unchanged)
          color &lt;- unch_col(normalized)
          list(background = color)
        },
        footer = function(values) htmltools::tags$b(comma(sum(values)))
      ),
      `% Changed` = colDef(
        cell = function(value) percent(value, 0.1),
        footer = htmltools::tags$b(
          percent(sum(1 - sum(df$Unchanged) / sum(df$Total)), 0.1)
        )
      ),
      Total = colDef(
        cell = function(value) comma(value, 1),
        style = function(value) {
          bar_style(width = value / length(merged_results), fill = &quot;#B3B3B3&quot;)
        },
        align = &quot;left&quot;,
        footer = function(values) htmltools::tags$b(comma(sum(values)))
      ),
      `% Of All Windows` = colDef(
        cell = function(value) percent(value, 0.1),
        footer = htmltools::tags$b(&quot;100%&quot;)
      )
    )
  )
div(class = &quot;table&quot;,
  div(class = &quot;table-header&quot;,
      div(class = &quot;caption&quot;, cp),
      tbl
  )
)</code></pre>
<div class="table">
<div class="table-header">
<div class="caption">
<em>Overall results for changed ER binding after DHT treament, broken down by which external feature contains the largest overlap with each merged window.</em>
</div>
<div id="htmlwidget-b4938d4888017726fe01" class="reactable html-widget" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-b4938d4888017726fe01">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"Feature":["Promoter","Enhancer","Super Enhancer","No Feature"],"Up":[1,0,0,3],"Down":[278,380,115,165],"Unchanged":[95,99,34,62],"% Changed":[0.745989304812834,0.793319415448852,0.771812080536913,0.730434782608696],"Total":[374,479,149,230],"% Of All Windows":[0.303571428571429,0.388798701298701,0.120941558441558,0.186688311688312]},"columns":[{"accessor":"Feature","name":"Feature","type":"factor","footer":{"name":"b","attribs":{},"children":["Total"]}},{"accessor":"Up","name":"Up","type":"numeric","cell":["1","0","0","3"],"footer":{"name":"b","attribs":{},"children":["4"]},"style":[{"background":"#FEFEFE"},{"background":"#FFFFFF"},{"background":"#FFFFFF"},{"background":"#FEFEFE"}]},{"accessor":"Down","name":"Down","type":"numeric","cell":["278","380","115","165"],"footer":{"name":"b","attribs":{},"children":["938"]},"style":[{"background":"#B5D5E7"},{"background":"#9AC5DF"},{"background":"#E0EDF5"},{"background":"#D3E6F1"}]},{"accessor":"Unchanged","name":"Unchanged","type":"numeric","cell":["95","99","34","62"],"footer":{"name":"b","attribs":{},"children":["290"]},"style":[{"background":"#F5F5F5"},{"background":"#F4F4F4"},{"background":"#FBFBFB"},{"background":"#F8F8F8"}]},{"accessor":"% Changed","name":"% Changed","type":"numeric","cell":["74.6%","79.3%","77.2%","73.0%"],"footer":{"name":"b","attribs":{},"children":["76.5%"]}},{"accessor":"Total","name":"Total","type":"numeric","cell":["374","479","149","230"],"footer":{"name":"b","attribs":{},"children":["1,232"]},"align":"left","style":[{"backgroundImage":"linear-gradient(90deg, #B3B3B3 30.3571428571429%, transparent 30.3571428571429%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null},{"backgroundImage":"linear-gradient(90deg, #B3B3B3 38.8798701298701%, transparent 38.8798701298701%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null},{"backgroundImage":"linear-gradient(90deg, #B3B3B3 12.0941558441558%, transparent 12.0941558441558%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null},{"backgroundImage":"linear-gradient(90deg, #B3B3B3 18.6688311688312%, transparent 18.6688311688312%)","backgroundSize":"100% 75%","backgroundRepeat":"no-repeat","backgroundPosition":"center","color":null}]},{"accessor":"% Of All Windows","name":"% Of All Windows","type":"numeric","cell":["30.4%","38.9%","12.1%","18.7%"],"footer":{"name":"b","attribs":{},"children":["100%"]}}],"defaultPageSize":10,"paginationType":"numbers","showPageInfo":true,"minRows":1,"dataKey":"85d5f1bd38ff5dfcc2b537884e2aafaa","key":"85d5f1bd38ff5dfcc2b537884e2aafaa"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</div>
<div id="most-highly-ranked" class="section level3">
<h3>Most Highly Ranked</h3>
<pre class="r"><code>show_n &lt;- min(200, length(merged_results))
scaling_vals &lt;- list(
  logFC = c(-1, 1)*max(abs(merged_results$logFC)),
  AveExpr = range(merged_results$AveExpr),
  Width = max(width(merged_results))
)
cp &lt;- htmltools::em(
  glue(
    &quot;The {show_n} most highly-ranked merged windows by FDR. &quot;,
    &quot;Regions were assigned based on which genomic region showed the largest &quot;,
    &quot;overlap with the merged window. &quot;,
    ifelse(
      has_features, 
      glue(&quot;Features are as provided in the file {basename(config$external$features)}&quot;),
      glue(&quot;&quot;)
    ),
    &quot;For windows mapped to large numbers of genes, hovering your mouse over the &quot;, 
    &quot;cell will reveal the full set of genes. &quot;,
    ifelse(
      any_detected,
      &quot;Only genes considered as detected in the RNA-Seq data are shown. &quot;,
      &quot;&quot;
    ),
    &quot;The Macs2 Peak column is filterable using T/F values&quot;
  )
)
gene_col &lt;- ifelse(any_detected, &quot;detected&quot;, &quot;gene_name&quot;)
tbl &lt;- merged_results %&gt;%
  mutate(w = width) %&gt;% 
  arrange(!!sym(fdr_column)) %&gt;% 
  plyranges::slice(seq_len(show_n)) %&gt;% 
  select(
    w, AveExpr, logFC, FDR = !!sym(fdr_column), 
    overlaps_ref, Region = region, any_of(&quot;feature&quot;), Genes = !!sym(gene_col)
  ) %&gt;% 
  as_tibble() %&gt;% 
  dplyr::rename(
    Range = range, `Width (kb)` = w, `Macs2 Peak` = overlaps_ref
  ) %&gt;% 
  rename_all(str_replace_all, &quot;feature&quot;, &quot;Feature&quot;) %&gt;% 
  mutate(Genes = vapply(Genes, paste, character(1), collapse = &quot;; &quot;)) %&gt;% 
  reactable(
    filterable = TRUE,
    showPageSizeOptions = TRUE,
    pageSizeOptions = c(10, 20, 50, show_n), defaultPageSize = 10,
    borderless  = TRUE,
    columns = list(
      Range = colDef(
        minWidth = 150,
        cell = function(value) {
          str_replace_all(value, &quot;:&quot;, &quot;: &quot;)
        }
      ),
      `Width (kb)` = colDef(
        cell = function(value) round(value / 1e3, 1),
        style = function(value) {
          x &lt;- value / scaling_vals$Width
          colour &lt;- expr_col(x)
          list(
            background = colour, 
            borderRight = &quot;1px solid rgba(0, 0, 0, 0.1)&quot;
          )
        },
        maxWidth = 70
      ),
      AveExpr = colDef(
        cell = function(value) round(value, 2),
        style = function(value) {
          x &lt;- (value - min(scaling_vals$AveExpr)) / diff(scaling_vals$AveExpr)
          colour = expr_col(x)
          list(background = colour)
        },
        maxWidth = 80
      ),
      logFC = colDef(
        cell = function(value) round(value, 2),
        style = function(value) {
          x &lt;- (value - min(scaling_vals$logFC)) / diff(scaling_vals$logFC)
          colour &lt;- lfc_col(x)
          list(background = colour)
        },
        maxWidth = 70
      ),
      FDR = colDef(
        cell = function(value) sprintf(&quot;%.2e&quot;, value),
        style = list(borderRight = &quot;1px solid rgba(0, 0, 0, 0.1)&quot;),
        maxWidth = 80
      ),
      `Macs2 Peak` = colDef(
        cell = function(value) ifelse(value, &quot;\u2714 Yes&quot;, &quot;\u2716 No&quot;),
        style = function(value) {
          color &lt;- ifelse(value, &quot;#008000&quot;, &quot;#e00000&quot;)
          list(color = color, borderRight = &quot;1px solid rgba(0, 0, 0, 0.1)&quot;)
        },
        maxWidth = 70
      ),
      Region = colDef(maxWidth = 150),
      Genes = colDef(
        cell = function(value) with_tooltip(value),
        minWidth = 150
      )
    )
  )
div(class = &quot;table&quot;,
  div(class = &quot;table-header&quot;,
      div(class = &quot;caption&quot;, cp),
      tbl
  )
)</code></pre>
<div class="table">
<div class="table-header">
<div class="caption">
<em>The 200 most highly-ranked merged windows by FDR. Regions were assigned based on which genomic region showed the largest overlap with the merged window. Features are as provided in the file h3k27ac_features.gtfFor windows mapped to large numbers of genes, hovering your mouse over the cell will reveal the full set of genes. Only genes considered as detected in the RNA-Seq data are shown. The Macs2 Peak column is filterable using T/F values</em>
</div>
<div id="htmlwidget-06e530a484c17ef5796b" class="reactable html-widget" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-06e530a484c17ef5796b">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"Range":["chr10:44668201-44669700","chr10:62607601-62609300","chr10:104469901-104471300","chr10:3734201-3735400","chr10:30100801-30102000","chr10:63169201-63170700","chr10:43605701-43607200","chr10:93120101-93121500","chr10:5538001-5540400","chr10:5701501-5703000","chr10:106240001-106241500","chr10:48331501-48333200","chr10:8767601-8768700","chr10:3331301-3333100","chr10:43521301-43522700","chr10:406801-408000","chr10:44732001-44733000","chr10:45136901-45138300","chr10:135091901-135092900","chr10:63677501-63678600","chr10:80450501-80451700","chr10:562701-564400","chr10:28545401-28546600","chr10:70570901-70572000","chr10:71181401-71182400","chr10:99330601-99332400","chr10:44987801-44989000","chr10:27270701-27271900","chr10:32047401-32049600","chr10:63030901-63032000","chr10:95194301-95196100","chr10:824301-825500","chr10:33552801-33554100","chr10:1310701-1311800","chr10:43577001-43578400","chr10:32940201-32941200","chr10:21918801-21920100","chr10:71198701-71200000","chr10:22920801-22922800","chr10:52120501-52121600","chr10:5529801-5531300","chr10:30721901-30723500","chr10:79342001-79343000","chr10:103360901-103361800","chr10:6194801-6196100","chr10:30830701-30831900","chr10:30096801-30098500","chr10:82725601-82727900","chr10:73099301-73100500","chr10:5545101-5546200","chr10:71344001-71345100","chr10:51546901-51549700","chr10:49495601-49497100","chr10:100073701-100075000","chr10:29438701-29439700","chr10:29413401-29414500","chr10:80853301-80854700","chr10:92078901-92080100","chr10:43581501-43582500","chr10:12914301-12915700","chr10:61082301-61083800","chr10:102642801-102643900","chr10:384101-385600","chr10:129992801-129993700","chr10:5535201-5536500","chr10:22004501-22005400","chr10:76659301-76661700","chr10:95163501-95164400","chr10:76345001-76346300","chr10:80886501-80889800","chr10:99096701-99097700","chr10:43476501-43478000","chr10:43858601-43859700","chr10:30825201-30826200","chr10:104533801-104534800","chr10:85451401-85452600","chr10:29606801-29608000","chr10:29918901-29920000","chr10:481201-483500","chr10:1016001-1017100","chr10:8170001-8171100","chr10:51572001-51573000","chr10:80062601-80063700","chr10:120384601-120385800","chr10:103879101-103881500","chr10:93430801-93432000","chr10:73627701-73628600","chr10:23370101-23371000","chr10:125481001-125482100","chr10:82050301-82051500","chr10:112549301-112550300","chr10:126354001-126354800","chr10:33797901-33799400","chr10:43428101-43429100","chr10:100028601-100029700","chr10:18725801-18726700","chr10:117423801-117424900","chr10:90145001-90146100","chr10:104573501-104574600","chr10:8126701-8127600","chr10:95795701-95796800","chr10:63849301-63850200","chr10:71101601-71102600","chr10:72626301-72627400","chr10:126074501-126075300","chr10:30811401-30812300","chr10:61548301-61549400","chr10:50022801-50023500","chr10:9229801-9230900","chr10:127313001-127314100","chr10:43721001-43721900","chr10:79648801-79649900","chr10:3948601-3949600","chr10:9080001-9080900","chr10:100767001-100767900","chr10:129126901-129128200","chr10:79630801-79633800","chr10:112540701-112541600","chr10:117839801-117840700","chr10:115686801-115687700","chr10:95755101-95756000","chr10:3509801-3511300","chr10:32648801-32649800","chr10:410001-411300","chr10:37435301-37436200","chr10:43604201-43605100","chr10:103033301-103034300","chr10:102124401-102125400","chr10:17491801-17492900","chr10:29753401-29754300","chr10:49513901-49515000","chr10:126246001-126247000","chr10:29908001-29908700","chr10:76635501-76636100","chr10:84429401-84430200","chr10:93316201-93317300","chr10:112155201-112155900","chr10:115718101-115718900","chr10:71906101-71907200","chr10:68862301-68863300","chr10:13742201-13743000","chr10:76585101-76587300","chr10:6338201-6340900","chr10:78733401-78734300","chr10:100061201-100062100","chr10:100759901-100760900","chr10:102756501-102758800","chr10:80999001-81000500","chr10:51530801-51531800","chr10:30971001-30971800","chr10:79047301-79048200","chr10:6196301-6197500","chr10:50627101-50627500","chr10:571401-572200","chr10:51091801-51092500","chr10:76924201-76925100","chr10:566001-567100","chr10:32632301-32633700","chr10:75578901-75580000","chr10:33873101-33874100","chr10:99207501-99208600","chr10:80908501-80910500","chr10:34233901-34235100","chr10:73093701-73094500","chr10:89416801-89417800","chr10:112216201-112217100","chr10:107457001-107457900","chr10:69967901-69968900","chr10:90233901-90235100","chr10:5160101-5161100","chr10:62702401-62703100","chr10:48434101-48435000","chr10:19535001-19535800","chr10:5157901-5158900","chr10:19520901-19522100","chr10:102132801-102133800","chr10:105649501-105650500","chr10:71120301-71121000","chr10:126812001-126813100","chr10:73068401-73069600","chr10:76681301-76681900","chr10:9228801-9229200","chr10:74091101-74092400","chr10:77004501-77006000","chr10:88779501-88780400","chr10:65389601-65391400","chr10:125503101-125504200","chr10:27926201-27927200","chr10:22971801-22972900","chr10:69548101-69548900","chr10:30894001-30895300","chr10:25569101-25569700","chr10:34762401-34763500","chr10:34716001-34716900","chr10:32542101-32543000","chr10:104927601-104928900","chr10:69579401-69580200","chr10:30246401-30247300","chr10:79169501-79170600","chr10:103876801-103877900"],"Width (kb)":[1500,1700,1400,1200,1200,1500,1500,1400,2400,1500,1500,1700,1100,1800,1400,1200,1000,1400,1000,1100,1200,1700,1200,1100,1000,1800,1200,1200,2200,1100,1800,1200,1300,1100,1400,1000,1300,1300,2000,1100,1500,1600,1000,900,1300,1200,1700,2300,1200,1100,1100,2800,1500,1300,1000,1100,1400,1200,1000,1400,1500,1100,1500,900,1300,900,2400,900,1300,3300,1000,1500,1100,1000,1000,1200,1200,1100,2300,1100,1100,1000,1100,1200,2400,1200,900,900,1100,1200,1000,800,1500,1000,1100,900,1100,1100,1100,900,1100,900,1000,1100,800,900,1100,700,1100,1100,900,1100,1000,900,900,1300,3000,900,900,900,900,1500,1000,1300,900,900,1000,1000,1100,900,1100,1000,700,600,800,1100,700,800,1100,1000,800,2200,2700,900,900,1000,2300,1500,1000,800,900,1200,400,800,700,900,1100,1400,1100,1000,1100,2000,1200,800,1000,900,900,1000,1200,1000,700,900,800,1000,1200,1000,1000,700,1100,1200,600,400,1300,1500,900,1800,1100,1000,1100,800,1300,600,1100,900,900,1300,800,900,1100,1100],"AveExpr":[6.02064431056975,7.94959593697328,6.98028817575429,5.60879990455897,6.04794165019698,6.97417320835961,6.96442867423685,7.42677373019464,7.45775221537821,7.80148033672219,6.85876517311919,5.94340534656658,6.57621033050938,6.19796532829265,6.48390813019782,5.79100355789658,5.76849900677582,6.27336578524534,5.23007266090256,5.16373102565731,5.88356634121005,6.80099244018603,6.85608260439126,5.40280773051929,5.43908001421119,6.46636404794956,5.96564427720932,6.34571357682379,7.41676430549972,6.00584149914222,8.11564803972838,6.01144530092048,6.1755797445909,5.9303783549277,6.37843429790319,6.17133188341893,6.59302174035824,5.33119453958714,6.45228045367281,5.52240904713102,6.48698869028184,6.03198779967051,5.17546680092491,5.09763232657538,6.28323845237986,6.26880860087511,6.24593651269903,7.02655984447986,5.83450474551819,5.84934371911489,5.88108632432994,7.10743337204495,5.98076165560192,6.09398323293034,5.0437871938716,5.49659843547073,6.301590821286,6.32453226548536,5.19639919874481,5.99865638464416,6.58459185938321,5.91957016319598,6.65425988774445,4.78102813737183,6.36705576217432,5.44373283421634,5.9499621866073,5.19427864530474,5.40266210691775,7.80493948675659,5.40747027608121,4.91670413190765,5.955934801375,5.42998622347761,5.7644597315885,4.93849210558013,4.86317816303073,5.58760496551053,7.55249660277097,5.71493860131528,6.57196824328058,5.05387300722199,5.46963212175555,6.35410127401473,7.36563022088205,5.38174073293355,4.92087188369385,5.45261751527764,5.7161531304516,5.52210742235658,5.03448389114093,4.29942360352726,4.83628345752951,4.783508309226,5.3036290998655,5.21640637986373,6.86728886009805,5.86096866592317,5.88123817248954,5.37445778765695,5.90267575483294,4.84846168745247,4.9105462614903,4.95910534947317,4.28309881198575,5.41123165861139,5.55164298726628,4.14749621301993,5.67619185021101,6.16120062196913,4.47860770465308,6.23112233005916,5.28284961396165,5.11401853264715,4.5329865884427,7.51647786075537,6.81623932181061,4.68517018991579,4.67846066148775,5.09662165864876,4.50116192813487,5.11124694581129,6.16044571104926,6.31572201233876,4.85779894323897,4.25373576102411,4.93675417699332,5.60492748175264,5.76617052017116,4.43779707232838,5.0586114111256,5.96696343301311,4.98533229238405,4.51131775336099,4.66500381937385,5.93797200137004,4.15786866409587,4.56628134361708,5.0207425823759,4.93193062300422,4.92968312131171,7.09755515123347,7.12946735135428,4.82683201027878,4.54367907526813,5.05345062081348,5.40094262105446,5.44203091340267,4.7955391823986,5.16523232820427,4.78763663354883,5.72893455999298,3.63298071503041,4.48908663806941,4.01951118451817,4.28851849034393,6.09442080851153,6.93234545132971,6.04965238762141,4.74187325637627,5.48376487907053,5.17956522429597,5.10944950798979,5.01530485928446,5.07332291954764,4.24084232519108,4.87823763873446,6.35074676348239,6.83137954939851,5.02383434354068,4.30641702479925,4.5637742774038,4.81531681821757,5.07007390460655,6.08889146057847,5.18767742466179,5.35773561420314,4.40773396336934,5.67432329989005,6.02963911258172,4.12090430898691,4.65889542776434,5.97478135141569,5.39029921572805,4.41919707340276,6.24261288794582,5.50250170397612,5.13651599063436,5.56770828892393,4.63865852870931,5.07543948066003,4.82650243652985,4.71446911206755,4.65139556018384,4.83133832656089,6.98179801956975,4.32159310016001,4.68805133760965,5.35740772156717,5.28910353453098],"logFC":[-2.57454704301321,-1.50622807108302,-2.02190205005877,-2.86298365735142,-2.48772792067924,-1.91410962001902,-1.95492931766496,-1.61566100719646,-1.80766863027008,-1.50152140775048,-2.11322527986073,-2.18122468147304,-2.03925449564214,-2.11002402736541,-2.18653856585798,-2.21364029940976,-2.18850371931719,-2.07789646646673,-2.68095862296553,-2.4981922695588,-2.1121669214915,-2.012314623368,-1.7051771905542,-2.35498500855946,-2.44663030032111,-1.84207628493012,-2.01897851178402,-1.84763233403226,-1.41654655047877,-1.97609085916686,-1.18451162588288,-2.17865209763171,-1.93179553988708,-2.09353906875851,-1.95546979489938,-1.90652265406901,-1.95668651233009,-2.61734691952554,-1.78412379708105,-2.03822662073842,-1.70196980500973,-1.85758344199528,-2.19192323654994,-2.52212880197202,-1.69145958611256,-1.91817641734855,-1.71640871473017,-1.42585426480307,-2.01692845442446,-1.82499759788239,-1.85243302649766,-1.49421546542088,-1.83448580297311,-1.87764271295018,-2.08740763802227,-1.91727571569021,-1.68758802507796,-1.60478246083692,-2.34644585538138,-1.75797500384511,-1.68850724078879,-2.07626421302043,-1.48124703105677,-2.22199814882887,-1.66967113798785,-1.82762990326571,-1.85167047476594,-2.13751820276918,-1.84005714383105,-1.07028569274609,-1.88513472189783,-2.06962953458543,-1.67745710058697,-1.75863349453845,-1.77493107805113,-2.06002318187918,-1.98451296642984,-1.76294274303202,-1.16925157455749,-1.79699223739946,-1.43447046754701,-1.92023230968511,-1.73185226596149,-1.47684011632763,-1.15520451613473,-1.76788443901288,-2.21503906589323,-1.7449479068567,-1.7035213151877,-1.74153953953492,-1.90524952199677,-2.02245984178991,-1.87563043489027,-2.18442155617903,-1.98962093365242,-1.8702639009596,-1.36753256820957,-1.85942887386476,-1.72143859173461,-1.9884342093767,-1.78751760503308,-2.09453804385051,-2.10035412877131,-1.86928540275801,-2.10046783490784,-1.78381687520171,-1.85665036719482,-2.15207832355855,-1.70304487728102,-1.52656296028573,-1.96328255568359,-1.53203524221272,-1.75474425318735,-1.86184848367498,-2.32073255512251,-1.08382923731965,-1.39778258193666,-2.2437370383917,-1.97283938424377,-1.85545884272223,-2.11629518605853,-1.76270717734244,-1.54114180796761,-1.46972730429854,-1.97314000728175,-1.9701978100807,-1.82745842362192,-1.66680592848302,-1.74998229447156,-1.98421426496817,-1.79743295863745,-1.56710669156617,-1.79795920724498,-1.87494311836617,-1.8743354552006,-1.51826219835984,-1.93582103961296,-1.8005959722236,-1.7476736113328,-1.76631374079949,-1.71429217590067,-1.13321192440143,-1.20216080667012,-1.95118761880062,-1.96050972895352,-1.94112230525588,-1.59875589140435,-1.62845150077334,-1.90306518498542,-1.6385551697719,-1.71735245223247,-1.54541565074121,-2.36065137490176,-1.78213536226905,-2.10130399499669,-2.03976326872034,-1.45156874781456,-1.16160087594868,-1.44363938632391,-1.76636563040579,-1.59283868629312,-1.58828018108531,-1.66550048821374,-1.8325963941985,-1.79060785426608,-1.96384021608325,-1.72175173870941,-1.41597079620417,-1.18932586667634,-1.62325888659355,-1.86432782412976,-1.75862175799549,-1.69972411649801,-1.62433081171113,-1.4201952717066,-1.62790395683345,-1.58013614875076,-1.81790799656322,-1.48586151792115,-1.41370720018232,-1.80625836344202,-1.64657986131803,-1.39731797420106,-1.60127846059238,-1.72824212416363,-1.54940236590615,-1.51643671926952,-1.66150996203826,-1.51307552581091,-1.92758710898006,-1.67574783412787,-1.63592588334797,-1.64864191946567,-1.75098893585774,-1.61688012467227,-1.19393853657838,-1.73424921301429,-1.6840073467392,-1.62013908403174,-1.64018145044525],"FDR":[2.63448136176932e-08,2.63448136176932e-08,3.61156999370621e-08,3.72404129622557e-08,3.72404129622557e-08,3.87366525228245e-08,4.69012697785226e-08,4.69012697785226e-08,4.70675301227671e-08,5.17805117756333e-08,5.9166847664717e-08,9.09132836929384e-08,1.28915749033607e-07,1.50090095154036e-07,1.52371562105568e-07,1.69620970555808e-07,1.69620970555808e-07,1.69620970555808e-07,1.69620970555808e-07,1.79258038748266e-07,1.79258038748266e-07,1.90486090878064e-07,1.90486090878064e-07,1.90486090878064e-07,2.06164267655775e-07,2.06808700694952e-07,2.47265271304986e-07,2.71060420298417e-07,2.84836782812516e-07,3.26232818515528e-07,3.26232818515528e-07,5.30800688380227e-07,5.74507616461309e-07,6.11953343572105e-07,6.11953343572105e-07,6.92212970971299e-07,7.68560200408928e-07,7.68560200408928e-07,7.83737919741949e-07,8.69904688343183e-07,1.02053851188189e-06,1.13750944640209e-06,1.14766740945632e-06,1.19309975221528e-06,1.30542316045482e-06,1.38670589510236e-06,1.46916974461982e-06,1.46916974461982e-06,1.52891298561893e-06,1.66943066348365e-06,1.66943066348365e-06,1.72105627331922e-06,1.73613931190134e-06,1.9923909408895e-06,2.10767616954903e-06,2.12297457503421e-06,2.15334620815887e-06,2.37866069040008e-06,2.57708312242681e-06,2.62467893133507e-06,3.05951813168516e-06,3.22147795120618e-06,3.24883500474128e-06,3.24883500474128e-06,3.83728646562104e-06,4.54951039390236e-06,5.08788558673318e-06,5.86845946747183e-06,6.61515853929394e-06,6.62714874408032e-06,7.37191600476075e-06,7.6970134696548e-06,7.94075807721672e-06,8.06902746992594e-06,8.10305562306569e-06,8.23355967672853e-06,8.46960262977778e-06,8.46960262977778e-06,8.74502431083616e-06,8.74502431083616e-06,8.74502431083616e-06,8.74502431083616e-06,8.74502431083616e-06,8.74502431083616e-06,9.37198111303562e-06,9.65106937250829e-06,9.79701533840962e-06,1.00588738686076e-05,1.01239152160716e-05,1.04702646180645e-05,1.04702646180645e-05,1.04702646180645e-05,1.06383032808295e-05,1.06383032808295e-05,1.06383032808295e-05,1.10723975008231e-05,1.10723975008231e-05,1.11596689407131e-05,1.11596689407131e-05,1.12513621481281e-05,1.14334475315449e-05,1.21093179367054e-05,1.21093179367054e-05,1.21093179367054e-05,1.21093179367054e-05,1.2455998441193e-05,1.2455998441193e-05,1.27548102370418e-05,1.31253240511378e-05,1.31253240511378e-05,1.39971587064989e-05,1.41546280852722e-05,1.4347555216212e-05,1.46474272276203e-05,1.47951401445751e-05,1.47951401445751e-05,1.53621799580101e-05,1.53621799580101e-05,1.55673827410518e-05,1.59543436356892e-05,1.61468630792298e-05,1.63882778907857e-05,1.65626544430519e-05,1.66416850169026e-05,1.76392342180484e-05,1.76392342180484e-05,1.76392342180484e-05,1.90025037344799e-05,1.90676466025569e-05,2.21363436063569e-05,2.21363436063569e-05,2.22813184864353e-05,2.28296008029813e-05,2.37971993483983e-05,2.37971993483983e-05,2.37971993483983e-05,2.37971993483983e-05,2.55294499452287e-05,2.67240397842813e-05,2.73468125824306e-05,2.83980823607515e-05,2.96675590270379e-05,2.98318038203342e-05,3.00995925249807e-05,3.00995925249807e-05,3.00995925249807e-05,3.00995925249807e-05,3.16702030885104e-05,3.26693814298254e-05,3.38932106247906e-05,3.47724160166534e-05,3.54786769528435e-05,3.54786769528435e-05,3.76175575340992e-05,3.76175575340992e-05,3.76175575340992e-05,3.82345456400351e-05,3.84045079481944e-05,3.96593895841229e-05,4.37102876879288e-05,4.37102876879288e-05,4.53735892480728e-05,4.6669964163064e-05,4.6669964163064e-05,4.6669964163064e-05,4.6669964163064e-05,4.75946909514328e-05,4.76511135296071e-05,4.81257187329605e-05,4.88546058220858e-05,4.88546058220858e-05,5.27537116527817e-05,5.37872680417687e-05,5.69787303761301e-05,5.69787303761301e-05,5.71621168939579e-05,5.72970266191786e-05,5.79490522746514e-05,6.19798643449133e-05,6.22977911508196e-05,6.22977911508196e-05,6.25568971286408e-05,6.25568971286408e-05,6.25568971286408e-05,6.37469394448345e-05,6.49904721602264e-05,6.89257600111582e-05,7.0464425529704e-05,7.05980951558195e-05,7.05980951558195e-05,7.09743915964027e-05,7.26380718585467e-05,7.36438589213101e-05,7.44519433687604e-05,7.64391019919698e-05,7.64391019919698e-05,7.96819949313437e-05,8.4823170218354e-05,8.4823170218354e-05,8.4823170218354e-05],"Macs2 Peak":[true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true],"Region":["Intergenic (>10kb)","Intergenic (>10kb)","Intron","Intergenic (>10kb)","Intergenic (>10kb)","Exon","Intron","Intergenic (>10kb)","Intergenic (<10kb)","Intron","Promoter (-1500/+500bp)","Promoter (-1500/+500bp)","Intergenic (<10kb)","Intergenic (<10kb)","Intergenic (>10kb)","Intron","Intergenic (>10kb)","Intergenic (>10kb)","Exon","Intron","Intergenic (<10kb)","Intron","Intron","Intergenic (>10kb)","Intergenic (<10kb)","Promoter (-1500/+500bp)","Upstream Promoter (<5kb)","Intergenic (<10kb)","Upstream Promoter (<5kb)","Intergenic (>10kb)","Intron","Intergenic (>10kb)","Promoter (-1500/+500bp)","Intron","Intron","Intron","Intron","Intergenic (<10kb)","Intron","Intron","Intergenic (>10kb)","Promoter (-1500/+500bp)","Intron","Promoter (-1500/+500bp)","Intron","Intron","Intergenic (>10kb)","Intergenic (>10kb)","Intron","Upstream Promoter (<5kb)","Intergenic (<10kb)","Promoter (-1500/+500bp)","Upstream Promoter (<5kb)","Intergenic (>10kb)","Intergenic (>10kb)","Intergenic (>10kb)","Intron","Intergenic (>10kb)","Intron","Upstream Promoter (<5kb)","Intron","Intergenic (>10kb)","Intron","Intron","Intergenic (<10kb)","Intron","Intron","Intron","Intron","Intron","Intron","Promoter (-1500/+500bp)","Intergenic (<10kb)","Intron","Upstream Promoter (<5kb)","Intergenic (>10kb)","Intergenic (<10kb)","Intron","Promoter (-1500/+500bp)","Intergenic (<10kb)","Intergenic (>10kb)","Promoter (-1500/+500bp)","Intron","Intergenic (<10kb)","Promoter (-1500/+500bp)","Upstream Promoter (<5kb)","Intergenic (<10kb)","Intron","Intron","Promoter (-1500/+500bp)","Intron","Intron","Intergenic (>10kb)","Intergenic (>10kb)","Promoter (-1500/+500bp)","Intron","Intron","Intron","Exon","Intergenic (<10kb)","Intron","Intron","Intron","Upstream Promoter (<5kb)","Intergenic (>10kb)","Upstream Promoter (<5kb)","Exon","Intron","Intergenic (>10kb)","Intron","Intron","Promoter (-1500/+500bp)","Upstream Promoter (<5kb)","Intergenic (>10kb)","Intron","Intron","Upstream Promoter (<5kb)","Exon","Intron","Intergenic (>10kb)","Intergenic (>10kb)","Intron","Intron","Intron","Intron","Intron","Intergenic (<10kb)","Intergenic (<10kb)","Intron","Intron","Promoter (-1500/+500bp)","Intron","Intron","Intron","Intron","Intergenic (>10kb)","Intergenic (>10kb)","Intergenic (>10kb)","Promoter (-1500/+500bp)","Intron","Intron","Promoter (-1500/+500bp)","Intron","Promoter (-1500/+500bp)","Intergenic (>10kb)","Intron","Promoter (-1500/+500bp)","Upstream Promoter (<5kb)","Upstream Promoter (<5kb)","Intron","Intron","Intron","Promoter (-1500/+500bp)","Intron","Intron","Intron","Intron","Intron","Upstream Promoter (<5kb)","Intron","Upstream Promoter (<5kb)","Intron","Intergenic (>10kb)","Intron","Intron","Intron","Intergenic (>10kb)","Intron","Intron","Intergenic (<10kb)","Intron","Intron","Intron","Intergenic (<10kb)","Intron","Promoter (-1500/+500bp)","Intron","Intron","Intron","Intergenic (<10kb)","Intron","Intergenic (>10kb)","Intergenic (<10kb)","Intron","Promoter (-1500/+500bp)","Promoter (-1500/+500bp)","Intron","Intergenic (>10kb)","Intron","Intergenic (<10kb)","Intergenic (<10kb)","Intron","Intron","Promoter (-1500/+500bp)","Intergenic (>10kb)","Intron","Intron","Intergenic (>10kb)","Intron","Upstream Promoter (<5kb)"],"Feature":["Super Enhancer","Super Enhancer","Promoter","Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","Promoter","Super Enhancer","Enhancer","Enhancer","Enhancer","Super Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","Promoter","Enhancer","Enhancer","Enhancer","Super Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","Super Enhancer","Enhancer","Super Enhancer","Enhancer","Promoter","No Feature","Promoter","Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","Promoter","Promoter","Enhancer","Promoter","Enhancer","Super Enhancer","Enhancer","Super Enhancer","Super Enhancer","Promoter","Enhancer","Promoter","Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","Promoter","No Feature","Enhancer","Enhancer","Super Enhancer","Enhancer","Promoter","Enhancer","Enhancer","Enhancer","Super Enhancer","Enhancer","Promoter","Enhancer","Enhancer","Super Enhancer","Promoter","Enhancer","Enhancer","Promoter","Enhancer","Super Enhancer","No Feature","Promoter","Promoter","Enhancer","Promoter","Enhancer","Enhancer","Super Enhancer","Enhancer","Promoter","Enhancer","Enhancer","Super Enhancer","Enhancer","Promoter","Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","Super Enhancer","Enhancer","Enhancer","Promoter","Enhancer","Enhancer","Enhancer","No Feature","Enhancer","Enhancer","Enhancer","Promoter","Enhancer","Super Enhancer","Enhancer","Enhancer","Promoter","No Feature","No Feature","Enhancer","Enhancer","Enhancer","No Feature","No Feature","Enhancer","Enhancer","Enhancer","Enhancer","Promoter","No Feature","Promoter","Enhancer","Super Enhancer","Enhancer","No Feature","Super Enhancer","Enhancer","Enhancer","Promoter","No Feature","Promoter","Promoter","Super Enhancer","No Feature","No Feature","No Feature","Promoter","Promoter","Enhancer","No Feature","Enhancer","No Feature","Enhancer","Enhancer","No Feature","Enhancer","Enhancer","Promoter","Promoter","No Feature","Promoter","Enhancer","Enhancer","Super Enhancer","Promoter","Enhancer","Enhancer","No Feature","Enhancer","Enhancer","Promoter","Enhancer","Enhancer","Enhancer","Enhancer","Promoter","Super Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","Promoter","Enhancer","Super Enhancer","Super Enhancer","Super Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","No Feature","Super Enhancer","Enhancer","Enhancer","Enhancer","Enhancer","Promoter"],"Genes":["","CDK1; RHOBTB1; ARID5B","SFXN2; ARL3; DDX21; KIFBP; FAM149B1; DNAJC9","KLF6","SVIL","TMEM26","RET; CSGALNACT2; RASGEF1A","PCGF5; HECTD2","CALML5; CALML3-AS1","TASOR2; AL732437.3; ASB13","","ANXA8; ZNF488","","","RET","DIP2C","","","ADAM8; TUBGCP2","ARID5B; RHOBTB1; RTKN2","","DIP2C; LARP4B","MPP7","CCAR1; STOX1; DDX50; FAM149B1; DNAJC9; WBP1L","HK1; TSPAN15; TACR2","UBTD1; HOGA1; PI4K2A; MMS19; MORN4","","ANKRD26","ARHGAP12","","CEP55; MYOF","DIP2C; LARP4B","NRP1","PFKP","RET","CCDC7","MLLT10; HNRNPRP1","HK1; TSPAN15; TACR2","PIP4K2A","DYNC1I2P1; ASAH2; SGMS1","CALML5; CALML3-AS1","MAP3K8","KCNMA1","DPCD","RBM17; PFKFB3","MAP3K8","SVIL","","UNC5B; SLC29A3; AL359183.1","CALML5; CALML3-AS1","TSPAN15; FAM241B","MSMB; CXCL12; RASSF4; C10orf25; ZNF22; MARCHF8; ZFAND4; AL356056.2; SYT15; FO681492.1; AGAP14P","MAPK8; FRMPD2","R3HCC1L; LOXL4; PYROXD2; GOT1","","","ZMIZ1; ZMIZ1-AS1; LINC00595","RPP30","RET","","","PAX2; SLF2; SEMA4G; MRPL43; BLOC1S2; SCD; SEC31B; NDUFB8; HIF1AN","ZMYND11; DIP2C","MKI67","CALML5; CALML3-AS1","MLLT10; HNRNPRP1; DNAJC1","KAT6B","CEP55; MYOF","ADK","ZMIZ1; ZMIZ1-AS1","FRAT2","RET","ZNF487; RASGEF1A; HNRNPF","MAP3K8","WBP1L; STOX1; DDX50; DDX21; KIFBP; VPS26A; FAM149B1; DNAJC9","","SVIL-AS1","SVIL","DIP2C","GTPBP4; IDI2-AS1; WDR37; LARP4B; AL359878.1; IDI1; PFKP","","NCOA4; MSMB; TIMM23; CXCL12; RASSF4; C10orf25; ZNF22; ALOX5; MARCHF8; ZFAND4; AL356056.2; SYT15; FO681492.1; AGAP14P","LINC00595; ZMIZ1","CACUL1","LDB1; PPRC1; NOLC1; GBF1; NFKB2","PPP1R3C","VSIR; PSAP","ARMC3; MSRB2; YWHAZP3","CPXM2","MAT1A","RBM20; PDCD4","LHPP; FAM53B; EEF1AKMT2","","BMS1","LOXL4; R3HCC1L; GOT1","CACNB2; AL450384.2","ATRNL1","RNLS; ATAD1; KLLN; PTEN","SFXN2; WBP1L; BORCS7; ARL3; STOX1; DDX50; DDX21; KIFBP; FAM149B1; DNAJC9","TAF3; GATA3","SLC35G1; PLCE1; HDAC1P1; PIPSL","ARID5B; RTKN2","HK1; TACR2","SGPL1","LHPP; OAT","MAP3K8","SLC16A9; CCDC6; BICC1","WDFY4","","EDRF1; ALDOAP2","RET; CSGALNACT2; RASGEF1A","AL391421.1; DLG5","","TAF3","","DOCK1","AL391421.1; DLG5","RBM20","GFRA1","NHLRC2; DCLRE1A","SLC35G1; PLCE1; PIPSL","","EPC1","DIP2C","ANKRD30A; AL157387.1; LINC00993","RET; CSGALNACT2; RASGEF1A","BTRC","SCD; OLMALINC; CWF19L1; BLOC1S2","ST8SIA6-AS1; ST8SIA6; CCNY","SVIL-AS1; SVIL","MAPK8; FRMPD2","LHPP; FAM53B","SVIL","KAT6B; ADK","","HECTD2; PPP1R3C","SMNDC1","NHLRC2","TYSND1; SAR1A","CTNNA3","FRMD4A","KAT6B","PFKFB3","KCNMA1","GOT1","","LZTS2","ZMIZ1; LINC00595","MSMB; NCOA4; AGAP7P; TIMM23; CXCL12; RASSF4; C10orf25; ZNF22; MARCHF8; ZFAND4; AL356056.2; SYT15; FO681492.1; AGAP14P","SVIL2P; ARHGAP12","KCNMA1","PFKFB3","ERCC6","DIP2C; LARP4B","PARG","SAMD8; VDAC2; DUSP13; COMTD1","DIP2C; LARP4B","EPC1","CAMK2G","","ZDHHC16; EXOSC1","ZMIZ1; ZMIZ1-AS1","CACNB2","UNC5B; SLC29A3; AL359183.1","PAPSS2","DUSP5","YWHAZP5","SIRT1; MYPN","RNLS; ATAD1; KLLN; PTEN","AKR1C3; AKR1C4; AKR1C2","RHOBTB1","ZNF488","MALRD1","AKR1C3; AKR1C4; AKR1C2","MALRD1","OLMALINC","SLK; SH3PXD2A; STN1","HK1; TSPAN15; TACR2","CTBP2","UNC5B; SLC29A3; AL359183.1","KAT6B; ADK","","DNAJB12","SAMD8; VDAC2; COMTD1","BMPR1A; SNCG; SHLD2; MMRN2; ADIRF-AS1; GLUD1","REEP3; JMJD1C; TATDN1P1","CPXM2","RAB18; MKX","PIP4K2A","AKR1B10P1; SIRT1; CTNNA3; DNAJC12","SVIL2P","","PARD3","PARD3","PPIAP31; EPC1","CNNM2; RPEL1; NT5C2","AKR1B10P1; SIRT1; DNAJC12","","KCNMA1","LDB1"]},"columns":[{"accessor":"Range","name":"Range","type":"character","cell":["chr10: 44668201-44669700","chr10: 62607601-62609300","chr10: 104469901-104471300","chr10: 3734201-3735400","chr10: 30100801-30102000","chr10: 63169201-63170700","chr10: 43605701-43607200","chr10: 93120101-93121500","chr10: 5538001-5540400","chr10: 5701501-5703000","chr10: 106240001-106241500","chr10: 48331501-48333200","chr10: 8767601-8768700","chr10: 3331301-3333100","chr10: 43521301-43522700","chr10: 406801-408000","chr10: 44732001-44733000","chr10: 45136901-45138300","chr10: 135091901-135092900","chr10: 63677501-63678600","chr10: 80450501-80451700","chr10: 562701-564400","chr10: 28545401-28546600","chr10: 70570901-70572000","chr10: 71181401-71182400","chr10: 99330601-99332400","chr10: 44987801-44989000","chr10: 27270701-27271900","chr10: 32047401-32049600","chr10: 63030901-63032000","chr10: 95194301-95196100","chr10: 824301-825500","chr10: 33552801-33554100","chr10: 1310701-1311800","chr10: 43577001-43578400","chr10: 32940201-32941200","chr10: 21918801-21920100","chr10: 71198701-71200000","chr10: 22920801-22922800","chr10: 52120501-52121600","chr10: 5529801-5531300","chr10: 30721901-30723500","chr10: 79342001-79343000","chr10: 103360901-103361800","chr10: 6194801-6196100","chr10: 30830701-30831900","chr10: 30096801-30098500","chr10: 82725601-82727900","chr10: 73099301-73100500","chr10: 5545101-5546200","chr10: 71344001-71345100","chr10: 51546901-51549700","chr10: 49495601-49497100","chr10: 100073701-100075000","chr10: 29438701-29439700","chr10: 29413401-29414500","chr10: 80853301-80854700","chr10: 92078901-92080100","chr10: 43581501-43582500","chr10: 12914301-12915700","chr10: 61082301-61083800","chr10: 102642801-102643900","chr10: 384101-385600","chr10: 129992801-129993700","chr10: 5535201-5536500","chr10: 22004501-22005400","chr10: 76659301-76661700","chr10: 95163501-95164400","chr10: 76345001-76346300","chr10: 80886501-80889800","chr10: 99096701-99097700","chr10: 43476501-43478000","chr10: 43858601-43859700","chr10: 30825201-30826200","chr10: 104533801-104534800","chr10: 85451401-85452600","chr10: 29606801-29608000","chr10: 29918901-29920000","chr10: 481201-483500","chr10: 1016001-1017100","chr10: 8170001-8171100","chr10: 51572001-51573000","chr10: 80062601-80063700","chr10: 120384601-120385800","chr10: 103879101-103881500","chr10: 93430801-93432000","chr10: 73627701-73628600","chr10: 23370101-23371000","chr10: 125481001-125482100","chr10: 82050301-82051500","chr10: 112549301-112550300","chr10: 126354001-126354800","chr10: 33797901-33799400","chr10: 43428101-43429100","chr10: 100028601-100029700","chr10: 18725801-18726700","chr10: 117423801-117424900","chr10: 90145001-90146100","chr10: 104573501-104574600","chr10: 8126701-8127600","chr10: 95795701-95796800","chr10: 63849301-63850200","chr10: 71101601-71102600","chr10: 72626301-72627400","chr10: 126074501-126075300","chr10: 30811401-30812300","chr10: 61548301-61549400","chr10: 50022801-50023500","chr10: 9229801-9230900","chr10: 127313001-127314100","chr10: 43721001-43721900","chr10: 79648801-79649900","chr10: 3948601-3949600","chr10: 9080001-9080900","chr10: 100767001-100767900","chr10: 129126901-129128200","chr10: 79630801-79633800","chr10: 112540701-112541600","chr10: 117839801-117840700","chr10: 115686801-115687700","chr10: 95755101-95756000","chr10: 3509801-3511300","chr10: 32648801-32649800","chr10: 410001-411300","chr10: 37435301-37436200","chr10: 43604201-43605100","chr10: 103033301-103034300","chr10: 102124401-102125400","chr10: 17491801-17492900","chr10: 29753401-29754300","chr10: 49513901-49515000","chr10: 126246001-126247000","chr10: 29908001-29908700","chr10: 76635501-76636100","chr10: 84429401-84430200","chr10: 93316201-93317300","chr10: 112155201-112155900","chr10: 115718101-115718900","chr10: 71906101-71907200","chr10: 68862301-68863300","chr10: 13742201-13743000","chr10: 76585101-76587300","chr10: 6338201-6340900","chr10: 78733401-78734300","chr10: 100061201-100062100","chr10: 100759901-100760900","chr10: 102756501-102758800","chr10: 80999001-81000500","chr10: 51530801-51531800","chr10: 30971001-30971800","chr10: 79047301-79048200","chr10: 6196301-6197500","chr10: 50627101-50627500","chr10: 571401-572200","chr10: 51091801-51092500","chr10: 76924201-76925100","chr10: 566001-567100","chr10: 32632301-32633700","chr10: 75578901-75580000","chr10: 33873101-33874100","chr10: 99207501-99208600","chr10: 80908501-80910500","chr10: 34233901-34235100","chr10: 73093701-73094500","chr10: 89416801-89417800","chr10: 112216201-112217100","chr10: 107457001-107457900","chr10: 69967901-69968900","chr10: 90233901-90235100","chr10: 5160101-5161100","chr10: 62702401-62703100","chr10: 48434101-48435000","chr10: 19535001-19535800","chr10: 5157901-5158900","chr10: 19520901-19522100","chr10: 102132801-102133800","chr10: 105649501-105650500","chr10: 71120301-71121000","chr10: 126812001-126813100","chr10: 73068401-73069600","chr10: 76681301-76681900","chr10: 9228801-9229200","chr10: 74091101-74092400","chr10: 77004501-77006000","chr10: 88779501-88780400","chr10: 65389601-65391400","chr10: 125503101-125504200","chr10: 27926201-27927200","chr10: 22971801-22972900","chr10: 69548101-69548900","chr10: 30894001-30895300","chr10: 25569101-25569700","chr10: 34762401-34763500","chr10: 34716001-34716900","chr10: 32542101-32543000","chr10: 104927601-104928900","chr10: 69579401-69580200","chr10: 30246401-30247300","chr10: 79169501-79170600","chr10: 103876801-103877900"],"minWidth":150},{"accessor":"Width (kb)","name":"Width (kb)","type":"numeric","cell":["1.5","1.7","1.4","1.2","1.2","1.5","1.5","1.4","2.4","1.5","1.5","1.7","1.1","1.8","1.4","1.2","1","1.4","1","1.1","1.2","1.7","1.2","1.1","1","1.8","1.2","1.2","2.2","1.1","1.8","1.2","1.3","1.1","1.4","1","1.3","1.3","2","1.1","1.5","1.6","1","0.9","1.3","1.2","1.7","2.3","1.2","1.1","1.1","2.8","1.5","1.3","1","1.1","1.4","1.2","1","1.4","1.5","1.1","1.5","0.9","1.3","0.9","2.4","0.9","1.3","3.3","1","1.5","1.1","1","1","1.2","1.2","1.1","2.3","1.1","1.1","1","1.1","1.2","2.4","1.2","0.9","0.9","1.1","1.2","1","0.8","1.5","1","1.1","0.9","1.1","1.1","1.1","0.9","1.1","0.9","1","1.1","0.8","0.9","1.1","0.7","1.1","1.1","0.9","1.1","1","0.9","0.9","1.3","3","0.9","0.9","0.9","0.9","1.5","1","1.3","0.9","0.9","1","1","1.1","0.9","1.1","1","0.7","0.6","0.8","1.1","0.7","0.8","1.1","1","0.8","2.2","2.7","0.9","0.9","1","2.3","1.5","1","0.8","0.9","1.2","0.4","0.8","0.7","0.9","1.1","1.4","1.1","1","1.1","2","1.2","0.8","1","0.9","0.9","1","1.2","1","0.7","0.9","0.8","1","1.2","1","1","0.7","1.1","1.2","0.6","0.4","1.3","1.5","0.9","1.8","1.1","1","1.1","0.8","1.3","0.6","1.1","0.9","0.9","1.3","0.8","0.9","1.1","1.1"],"maxWidth":70,"style":[{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#EDE5C5","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#D6DBBE","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#D6DBBE","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#DAB095","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#EDE5C5","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E9DEBE","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#D6DBBE","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#D6DBBE","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#EDE5C5","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E9DEBE","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#DEC0A3","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E9DEBE","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#CDD5BA","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#D6DBBE","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#CDD5BA","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#CDD5BA","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E3CFB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#EAE6C6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#CDD5BA","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#EDE5C5","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#DCB89C","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#D58E80","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#CDD5BA","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#D6DBBE","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#D6DBBE","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#CDD5BA","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#DAB095","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#CDD5BA","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#D35C79","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#DCB89C","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#DAB095","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#9BBAA2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#9BBAA2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#8CB69E","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#CDD5BA","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#D47B7B","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#CDD5BA","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#8CB69E","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#7EB19A","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#9BBAA2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#8CB69E","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#9BBAA2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#9BBAA2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#DEC0A3","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#D69785","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#DCB89C","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#9BBAA2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#60A893","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#9BBAA2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#8CB69E","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#D6DBBE","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E3CFB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#9BBAA2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#8CB69E","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#9BBAA2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#8CB69E","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#C3D0B6","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#7EB19A","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#60A893","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#CDD5BA","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E0E0C2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#E9DEBE","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#AFC5AC","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#9BBAA2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#CDD5BA","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#7EB19A","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#CDD5BA","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#9BBAA2","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#A6C0A7","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"background":"#B9CAB1","borderRight":"1px solid rgba(0, 0, 0, 0.1)"}]},{"accessor":"AveExpr","name":"AveExpr","type":"numeric","cell":["6.02","7.95","6.98","5.61","6.05","6.97","6.96","7.43","7.46","7.8","6.86","5.94","6.58","6.2","6.48","5.79","5.77","6.27","5.23","5.16","5.88","6.8","6.86","5.4","5.44","6.47","5.97","6.35","7.42","6.01","8.12","6.01","6.18","5.93","6.38","6.17","6.59","5.33","6.45","5.52","6.49","6.03","5.18","5.1","6.28","6.27","6.25","7.03","5.83","5.85","5.88","7.11","5.98","6.09","5.04","5.5","6.3","6.32","5.2","6","6.58","5.92","6.65","4.78","6.37","5.44","5.95","5.19","5.4","7.8","5.41","4.92","5.96","5.43","5.76","4.94","4.86","5.59","7.55","5.71","6.57","5.05","5.47","6.35","7.37","5.38","4.92","5.45","5.72","5.52","5.03","4.3","4.84","4.78","5.3","5.22","6.87","5.86","5.88","5.37","5.9","4.85","4.91","4.96","4.28","5.41","5.55","4.15","5.68","6.16","4.48","6.23","5.28","5.11","4.53","7.52","6.82","4.69","4.68","5.1","4.5","5.11","6.16","6.32","4.86","4.25","4.94","5.6","5.77","4.44","5.06","5.97","4.99","4.51","4.67","5.94","4.16","4.57","5.02","4.93","4.93","7.1","7.13","4.83","4.54","5.05","5.4","5.44","4.8","5.17","4.79","5.73","3.63","4.49","4.02","4.29","6.09","6.93","6.05","4.74","5.48","5.18","5.11","5.02","5.07","4.24","4.88","6.35","6.83","5.02","4.31","4.56","4.82","5.07","6.09","5.19","5.36","4.41","5.67","6.03","4.12","4.66","5.97","5.39","4.42","6.24","5.5","5.14","5.57","4.64","5.08","4.83","4.71","4.65","4.83","6.98","4.32","4.69","5.36","5.29"],"maxWidth":80,"style":[{"background":"#ECE3C3"},{"background":"#D4727A"},{"background":"#DAB095"},{"background":"#DADDC0"},{"background":"#EBE1C2"},{"background":"#DAB095"},{"background":"#DAB195"},{"background":"#D69584"},{"background":"#D69483"},{"background":"#D47D7B"},{"background":"#DCB79B"},{"background":"#EEE7C7"},{"background":"#DFC6A9"},{"background":"#E8DABB"},{"background":"#E1CBAD"},{"background":"#E7E4C5"},{"background":"#E5E3C4"},{"background":"#E6D6B7"},{"background":"#C0CEB4"},{"background":"#BCCCB2"},{"background":"#EDE8C8"},{"background":"#DCBA9D"},{"background":"#DCB79B"},{"background":"#CCD5BA"},{"background":"#CFD6BB"},{"background":"#E1CCAE"},{"background":"#EDE6C6"},{"background":"#E4D2B4"},{"background":"#D69684"},{"background":"#ECE4C4"},{"background":"#D36679"},{"background":"#ECE3C4"},{"background":"#E8DBBC"},{"background":"#EEE7C7"},{"background":"#E3D1B2"},{"background":"#E8DBBC"},{"background":"#DFC5A8"},{"background":"#C7D2B8"},{"background":"#E2CDAF"},{"background":"#D4D9BD"},{"background":"#E1CBAD"},{"background":"#ECE2C3"},{"background":"#BDCCB3"},{"background":"#B7C9B0"},{"background":"#E6D5B7"},{"background":"#E6D6B7"},{"background":"#E6D7B8"},{"background":"#D9AE92"},{"background":"#E9E6C6"},{"background":"#EAE6C7"},{"background":"#EDE7C8"},{"background":"#D8A98F"},{"background":"#EDE5C5"},{"background":"#EADFC0"},{"background":"#B4C7AE"},{"background":"#D2D8BD"},{"background":"#E5D4B6"},{"background":"#E5D3B5"},{"background":"#BECDB3"},{"background":"#ECE4C4"},{"background":"#DFC6A8"},{"background":"#EEE8C8"},{"background":"#DEC2A5"},{"background":"#A2BDA5"},{"background":"#E4D1B3"},{"background":"#CFD6BB"},{"background":"#EEE6C7"},{"background":"#BECDB3"},{"background":"#CCD5BA"},{"background":"#D47D7B"},{"background":"#CCD5BA"},{"background":"#ABC2AA"},{"background":"#EDE6C6"},{"background":"#CED6BB"},{"background":"#E5E3C4"},{"background":"#ADC3AB"},{"background":"#A7C0A8"},{"background":"#D9DCBF"},{"background":"#D58E80"},{"background":"#E1E1C3"},{"background":"#DFC6A9"},{"background":"#B4C8AE"},{"background":"#D1D7BC"},{"background":"#E4D2B3"},{"background":"#D79986"},{"background":"#CBD4B9"},{"background":"#ABC3AA"},{"background":"#CFD7BB"},{"background":"#E1E1C3"},{"background":"#D4D9BD"},{"background":"#B3C7AE"},{"background":"#72AD97"},{"background":"#A6BFA7"},{"background":"#A2BDA5"},{"background":"#C5D1B7"},{"background":"#BFCEB4"},{"background":"#DBB69A"},{"background":"#EBE7C7"},{"background":"#EDE7C8"},{"background":"#CAD4B9"},{"background":"#EEE8C8"},{"background":"#A6C0A7"},{"background":"#ABC2AA"},{"background":"#AEC4AB"},{"background":"#71AD96"},{"background":"#CDD5BA"},{"background":"#D6DBBE"},{"background":"#63A993"},{"background":"#DFDFC2"},{"background":"#E8DCBC"},{"background":"#85B39C"},{"background":"#E7D8B9"},{"background":"#C4D0B6"},{"background":"#B8CAB0"},{"background":"#8AB59E"},{"background":"#D69081"},{"background":"#DCB99D"},{"background":"#9ABAA2"},{"background":"#99BAA2"},{"background":"#B7C9B0"},{"background":"#87B49D"},{"background":"#B8CAB0"},{"background":"#E8DCBC"},{"background":"#E5D4B5"},{"background":"#A7C0A8"},{"background":"#6EAC96"},{"background":"#ACC3AA"},{"background":"#DADDC0"},{"background":"#E5E3C4"},{"background":"#80B29B"},{"background":"#B5C8AF"},{"background":"#EDE6C6"},{"background":"#B0C5AC"},{"background":"#88B49D"},{"background":"#98B9A2"},{"background":"#EEE7C7"},{"background":"#64A993"},{"background":"#8EB69F"},{"background":"#B2C6AD"},{"background":"#ACC3AA"},{"background":"#ACC3AA"},{"background":"#D8AA8F"},{"background":"#D8A88E"},{"background":"#A5BFA7"},{"background":"#8BB59E"},{"background":"#B4C8AE"},{"background":"#CCD5BA"},{"background":"#CFD6BB"},{"background":"#A3BEA6"},{"background":"#BCCCB2"},{"background":"#A2BEA5"},{"background":"#E2E1C3"},{"background":"#0C9793"},{"background":"#86B49C"},{"background":"#4DA493"},{"background":"#71AD97"},{"background":"#EADFC0"},{"background":"#DBB397"},{"background":"#EBE1C2"},{"background":"#9FBCA4"},{"background":"#D2D8BC"},{"background":"#BDCCB3"},{"background":"#B8CAB0"},{"background":"#B2C6AD"},{"background":"#B6C8AF"},{"background":"#6CAC95"},{"background":"#A8C1A8"},{"background":"#E4D2B3"},{"background":"#DCB89C"},{"background":"#B2C6AD"},{"background":"#73AE97"},{"background":"#8DB69F"},{"background":"#A4BFA6"},{"background":"#B5C8AF"},{"background":"#EADFC0"},{"background":"#BDCDB3"},{"background":"#C9D3B9"},{"background":"#7DB19A"},{"background":"#DFDFC2"},{"background":"#ECE2C3"},{"background":"#5EA893"},{"background":"#97B9A1"},{"background":"#EDE5C5"},{"background":"#CBD4BA"},{"background":"#7EB19A"},{"background":"#E7D7B9"},{"background":"#D3D9BD"},{"background":"#BACBB1"},{"background":"#D7DBBF"},{"background":"#95B8A1"},{"background":"#B6C8AF"},{"background":"#A5BFA7"},{"background":"#9DBBA3"},{"background":"#96B9A1"},{"background":"#A5BFA7"},{"background":"#DAB095"},{"background":"#75AE98"},{"background":"#9ABAA2"},{"background":"#C9D3B9"},{"background":"#C4D0B6"}]},{"accessor":"logFC","name":"logFC","type":"numeric","cell":["-2.57","-1.51","-2.02","-2.86","-2.49","-1.91","-1.95","-1.62","-1.81","-1.5","-2.11","-2.18","-2.04","-2.11","-2.19","-2.21","-2.19","-2.08","-2.68","-2.5","-2.11","-2.01","-1.71","-2.35","-2.45","-1.84","-2.02","-1.85","-1.42","-1.98","-1.18","-2.18","-1.93","-2.09","-1.96","-1.91","-1.96","-2.62","-1.78","-2.04","-1.7","-1.86","-2.19","-2.52","-1.69","-1.92","-1.72","-1.43","-2.02","-1.82","-1.85","-1.49","-1.83","-1.88","-2.09","-1.92","-1.69","-1.6","-2.35","-1.76","-1.69","-2.08","-1.48","-2.22","-1.67","-1.83","-1.85","-2.14","-1.84","-1.07","-1.89","-2.07","-1.68","-1.76","-1.77","-2.06","-1.98","-1.76","-1.17","-1.8","-1.43","-1.92","-1.73","-1.48","-1.16","-1.77","-2.22","-1.74","-1.7","-1.74","-1.91","-2.02","-1.88","-2.18","-1.99","-1.87","-1.37","-1.86","-1.72","-1.99","-1.79","-2.09","-2.1","-1.87","-2.1","-1.78","-1.86","-2.15","-1.7","-1.53","-1.96","-1.53","-1.75","-1.86","-2.32","-1.08","-1.4","-2.24","-1.97","-1.86","-2.12","-1.76","-1.54","-1.47","-1.97","-1.97","-1.83","-1.67","-1.75","-1.98","-1.8","-1.57","-1.8","-1.87","-1.87","-1.52","-1.94","-1.8","-1.75","-1.77","-1.71","-1.13","-1.2","-1.95","-1.96","-1.94","-1.6","-1.63","-1.9","-1.64","-1.72","-1.55","-2.36","-1.78","-2.1","-2.04","-1.45","-1.16","-1.44","-1.77","-1.59","-1.59","-1.67","-1.83","-1.79","-1.96","-1.72","-1.42","-1.19","-1.62","-1.86","-1.76","-1.7","-1.62","-1.42","-1.63","-1.58","-1.82","-1.49","-1.41","-1.81","-1.65","-1.4","-1.6","-1.73","-1.55","-1.52","-1.66","-1.51","-1.93","-1.68","-1.64","-1.65","-1.75","-1.62","-1.19","-1.73","-1.68","-1.62","-1.64"],"maxWidth":70,"style":[{"background":"#1E7FB7"},{"background":"#7BB4D5"},{"background":"#4E9AC7"},{"background":"#0571B0"},{"background":"#2583BA"},{"background":"#57A0CA"},{"background":"#549EC9"},{"background":"#71AED2"},{"background":"#61A5CD"},{"background":"#7BB4D5"},{"background":"#4696C4"},{"background":"#4092C2"},{"background":"#4C99C6"},{"background":"#4696C4"},{"background":"#4092C2"},{"background":"#3D91C1"},{"background":"#3F92C2"},{"background":"#4997C5"},{"background":"#147AB5"},{"background":"#2483BA"},{"background":"#4696C4"},{"background":"#4F9BC7"},{"background":"#6AAACF"},{"background":"#318ABE"},{"background":"#2985BB"},{"background":"#5EA3CC"},{"background":"#4E9AC7"},{"background":"#5DA3CC"},{"background":"#83B8D7"},{"background":"#529CC8"},{"background":"#97C4DE"},{"background":"#4092C2"},{"background":"#569FC9"},{"background":"#4897C5"},{"background":"#549EC9"},{"background":"#58A0CA"},{"background":"#549DC9"},{"background":"#1A7DB6"},{"background":"#63A6CD"},{"background":"#4D99C6"},{"background":"#6AAAD0"},{"background":"#5CA2CB"},{"background":"#3F92C2"},{"background":"#2281B9"},{"background":"#6BABD0"},{"background":"#579FCA"},{"background":"#69A9CF"},{"background":"#82B8D7"},{"background":"#4E9AC7"},{"background":"#5FA4CC"},{"background":"#5DA3CB"},{"background":"#7CB4D5"},{"background":"#5EA4CC"},{"background":"#5BA1CB"},{"background":"#4897C5"},{"background":"#579FCA"},{"background":"#6BABD0"},{"background":"#72AFD2"},{"background":"#328ABE"},{"background":"#65A7CE"},{"background":"#6BABD0"},{"background":"#4998C5"},{"background":"#7DB5D6"},{"background":"#3C90C1"},{"background":"#6DACD0"},{"background":"#5FA4CC"},{"background":"#5DA3CB"},{"background":"#4494C4"},{"background":"#5EA3CC"},{"background":"#A1C9E1"},{"background":"#5AA1CA"},{"background":"#4A98C5"},{"background":"#6CABD0"},{"background":"#65A7CE"},{"background":"#64A6CE"},{"background":"#4B98C6"},{"background":"#519CC8"},{"background":"#65A7CE"},{"background":"#98C5DE"},{"background":"#62A5CD"},{"background":"#81B7D7"},{"background":"#579FCA"},{"background":"#67A9CF"},{"background":"#7EB5D6"},{"background":"#9AC5DF"},{"background":"#64A7CE"},{"background":"#3D91C1"},{"background":"#66A8CE"},{"background":"#6AAACF"},{"background":"#66A8CE"},{"background":"#58A0CA"},{"background":"#4E9AC7"},{"background":"#5BA1CB"},{"background":"#4092C2"},{"background":"#519CC8"},{"background":"#5BA2CB"},{"background":"#87BBD9"},{"background":"#5CA2CB"},{"background":"#68A9CF"},{"background":"#519CC8"},{"background":"#62A6CD"},{"background":"#4897C5"},{"background":"#4796C5"},{"background":"#5BA2CB"},{"background":"#4796C5"},{"background":"#63A6CD"},{"background":"#5CA2CB"},{"background":"#4394C3"},{"background":"#6AAAD0"},{"background":"#79B3D4"},{"background":"#539DC8"},{"background":"#79B3D4"},{"background":"#65A7CE"},{"background":"#5CA2CB"},{"background":"#348BBE"},{"background":"#A0C9E1"},{"background":"#84B9D8"},{"background":"#3B8FC1"},{"background":"#529DC8"},{"background":"#5CA2CB"},{"background":"#4696C4"},{"background":"#65A7CE"},{"background":"#78B2D4"},{"background":"#7EB6D6"},{"background":"#529DC8"},{"background":"#529DC8"},{"background":"#5FA4CC"},{"background":"#6DACD1"},{"background":"#66A8CE"},{"background":"#519CC8"},{"background":"#62A5CD"},{"background":"#76B1D3"},{"background":"#61A5CD"},{"background":"#5BA2CB"},{"background":"#5BA2CB"},{"background":"#7AB3D5"},{"background":"#559EC9"},{"background":"#61A5CD"},{"background":"#66A8CE"},{"background":"#64A7CE"},{"background":"#69A9CF"},{"background":"#9CC6DF"},{"background":"#96C3DD"},{"background":"#549EC9"},{"background":"#539DC8"},{"background":"#559EC9"},{"background":"#73AFD2"},{"background":"#70AED2"},{"background":"#58A0CA"},{"background":"#6FADD1"},{"background":"#69A9CF"},{"background":"#78B2D4"},{"background":"#3089BD"},{"background":"#63A6CD"},{"background":"#4796C5"},{"background":"#4C99C6"},{"background":"#80B7D6"},{"background":"#99C5DE"},{"background":"#80B7D7"},{"background":"#64A7CE"},{"background":"#73AFD3"},{"background":"#74B0D3"},{"background":"#6DACD1"},{"background":"#5EA4CC"},{"background":"#62A6CD"},{"background":"#539DC8"},{"background":"#68A9CF"},{"background":"#83B8D7"},{"background":"#97C4DE"},{"background":"#71AED2"},{"background":"#5CA2CB"},{"background":"#65A7CE"},{"background":"#6AAAD0"},{"background":"#71AED2"},{"background":"#82B8D7"},{"background":"#70AED2"},{"background":"#75B0D3"},{"background":"#60A4CC"},{"background":"#7DB5D5"},{"background":"#83B8D7"},{"background":"#61A5CD"},{"background":"#6FADD1"},{"background":"#84B9D8"},{"background":"#73AFD2"},{"background":"#68A9CF"},{"background":"#77B2D4"},{"background":"#7AB3D5"},{"background":"#6DACD1"},{"background":"#7AB3D5"},{"background":"#569FC9"},{"background":"#6CABD0"},{"background":"#70ADD1"},{"background":"#6FADD1"},{"background":"#66A8CE"},{"background":"#71AED2"},{"background":"#96C3DE"},{"background":"#67A8CF"},{"background":"#6BABD0"},{"background":"#71AED2"},{"background":"#6FADD1"}]},{"accessor":"FDR","name":"FDR","type":"numeric","cell":["2.63e-08","2.63e-08","3.61e-08","3.72e-08","3.72e-08","3.87e-08","4.69e-08","4.69e-08","4.71e-08","5.18e-08","5.92e-08","9.09e-08","1.29e-07","1.50e-07","1.52e-07","1.70e-07","1.70e-07","1.70e-07","1.70e-07","1.79e-07","1.79e-07","1.90e-07","1.90e-07","1.90e-07","2.06e-07","2.07e-07","2.47e-07","2.71e-07","2.85e-07","3.26e-07","3.26e-07","5.31e-07","5.75e-07","6.12e-07","6.12e-07","6.92e-07","7.69e-07","7.69e-07","7.84e-07","8.70e-07","1.02e-06","1.14e-06","1.15e-06","1.19e-06","1.31e-06","1.39e-06","1.47e-06","1.47e-06","1.53e-06","1.67e-06","1.67e-06","1.72e-06","1.74e-06","1.99e-06","2.11e-06","2.12e-06","2.15e-06","2.38e-06","2.58e-06","2.62e-06","3.06e-06","3.22e-06","3.25e-06","3.25e-06","3.84e-06","4.55e-06","5.09e-06","5.87e-06","6.62e-06","6.63e-06","7.37e-06","7.70e-06","7.94e-06","8.07e-06","8.10e-06","8.23e-06","8.47e-06","8.47e-06","8.75e-06","8.75e-06","8.75e-06","8.75e-06","8.75e-06","8.75e-06","9.37e-06","9.65e-06","9.80e-06","1.01e-05","1.01e-05","1.05e-05","1.05e-05","1.05e-05","1.06e-05","1.06e-05","1.06e-05","1.11e-05","1.11e-05","1.12e-05","1.12e-05","1.13e-05","1.14e-05","1.21e-05","1.21e-05","1.21e-05","1.21e-05","1.25e-05","1.25e-05","1.28e-05","1.31e-05","1.31e-05","1.40e-05","1.42e-05","1.43e-05","1.46e-05","1.48e-05","1.48e-05","1.54e-05","1.54e-05","1.56e-05","1.60e-05","1.61e-05","1.64e-05","1.66e-05","1.66e-05","1.76e-05","1.76e-05","1.76e-05","1.90e-05","1.91e-05","2.21e-05","2.21e-05","2.23e-05","2.28e-05","2.38e-05","2.38e-05","2.38e-05","2.38e-05","2.55e-05","2.67e-05","2.73e-05","2.84e-05","2.97e-05","2.98e-05","3.01e-05","3.01e-05","3.01e-05","3.01e-05","3.17e-05","3.27e-05","3.39e-05","3.48e-05","3.55e-05","3.55e-05","3.76e-05","3.76e-05","3.76e-05","3.82e-05","3.84e-05","3.97e-05","4.37e-05","4.37e-05","4.54e-05","4.67e-05","4.67e-05","4.67e-05","4.67e-05","4.76e-05","4.77e-05","4.81e-05","4.89e-05","4.89e-05","5.28e-05","5.38e-05","5.70e-05","5.70e-05","5.72e-05","5.73e-05","5.79e-05","6.20e-05","6.23e-05","6.23e-05","6.26e-05","6.26e-05","6.26e-05","6.37e-05","6.50e-05","6.89e-05","7.05e-05","7.06e-05","7.06e-05","7.10e-05","7.26e-05","7.36e-05","7.45e-05","7.64e-05","7.64e-05","7.97e-05","8.48e-05","8.48e-05","8.48e-05"],"maxWidth":80,"style":{"borderRight":"1px solid rgba(0, 0, 0, 0.1)"}},{"accessor":"Macs2 Peak","name":"Macs2 Peak","type":"logical","cell":["✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes","✔ Yes"],"maxWidth":70,"style":[{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"},{"color":"#008000","borderRight":"1px solid rgba(0, 0, 0, 0.1)"}]},{"accessor":"Region","name":"Region","type":"factor","maxWidth":150},{"accessor":"Feature","name":"Feature","type":"factor"},{"accessor":"Genes","name":"Genes","type":"character","cell":[{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"CDK1; RHOBTB1; ARID5B"},"children":["CDK1; RHOBTB1; ARID5B"]},{"name":"span","attribs":{"title":"SFXN2; ARL3; DDX21; KIFBP; FAM149B1; DNAJC9"},"children":["SFXN2; ARL3; DDX21; KIFBP; ..."]},{"name":"span","attribs":{"title":"KLF6"},"children":["KLF6"]},{"name":"span","attribs":{"title":"SVIL"},"children":["SVIL"]},{"name":"span","attribs":{"title":"TMEM26"},"children":["TMEM26"]},{"name":"span","attribs":{"title":"RET; CSGALNACT2; RASGEF1A"},"children":["RET; CSGALNACT2; RASGEF1A"]},{"name":"span","attribs":{"title":"PCGF5; HECTD2"},"children":["PCGF5; HECTD2"]},{"name":"span","attribs":{"title":"CALML5; CALML3-AS1"},"children":["CALML5; CALML3-AS1"]},{"name":"span","attribs":{"title":"TASOR2; AL732437.3; ASB13"},"children":["TASOR2; AL732437.3; ASB13"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"ANXA8; ZNF488"},"children":["ANXA8; ZNF488"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"RET"},"children":["RET"]},{"name":"span","attribs":{"title":"DIP2C"},"children":["DIP2C"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"ADAM8; TUBGCP2"},"children":["ADAM8; TUBGCP2"]},{"name":"span","attribs":{"title":"ARID5B; RHOBTB1; RTKN2"},"children":["ARID5B; RHOBTB1; RTKN2"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"DIP2C; LARP4B"},"children":["DIP2C; LARP4B"]},{"name":"span","attribs":{"title":"MPP7"},"children":["MPP7"]},{"name":"span","attribs":{"title":"CCAR1; STOX1; DDX50; FAM149B1; DNAJC9; WBP1L"},"children":["CCAR1; STOX1; DDX50; FAM149..."]},{"name":"span","attribs":{"title":"HK1; TSPAN15; TACR2"},"children":["HK1; TSPAN15; TACR2"]},{"name":"span","attribs":{"title":"UBTD1; HOGA1; PI4K2A; MMS19; MORN4"},"children":["UBTD1; HOGA1; PI4K2A; MMS19..."]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"ANKRD26"},"children":["ANKRD26"]},{"name":"span","attribs":{"title":"ARHGAP12"},"children":["ARHGAP12"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"CEP55; MYOF"},"children":["CEP55; MYOF"]},{"name":"span","attribs":{"title":"DIP2C; LARP4B"},"children":["DIP2C; LARP4B"]},{"name":"span","attribs":{"title":"NRP1"},"children":["NRP1"]},{"name":"span","attribs":{"title":"PFKP"},"children":["PFKP"]},{"name":"span","attribs":{"title":"RET"},"children":["RET"]},{"name":"span","attribs":{"title":"CCDC7"},"children":["CCDC7"]},{"name":"span","attribs":{"title":"MLLT10; HNRNPRP1"},"children":["MLLT10; HNRNPRP1"]},{"name":"span","attribs":{"title":"HK1; TSPAN15; TACR2"},"children":["HK1; TSPAN15; TACR2"]},{"name":"span","attribs":{"title":"PIP4K2A"},"children":["PIP4K2A"]},{"name":"span","attribs":{"title":"DYNC1I2P1; ASAH2; SGMS1"},"children":["DYNC1I2P1; ASAH2; SGMS1"]},{"name":"span","attribs":{"title":"CALML5; CALML3-AS1"},"children":["CALML5; CALML3-AS1"]},{"name":"span","attribs":{"title":"MAP3K8"},"children":["MAP3K8"]},{"name":"span","attribs":{"title":"KCNMA1"},"children":["KCNMA1"]},{"name":"span","attribs":{"title":"DPCD"},"children":["DPCD"]},{"name":"span","attribs":{"title":"RBM17; PFKFB3"},"children":["RBM17; PFKFB3"]},{"name":"span","attribs":{"title":"MAP3K8"},"children":["MAP3K8"]},{"name":"span","attribs":{"title":"SVIL"},"children":["SVIL"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"UNC5B; SLC29A3; AL359183.1"},"children":["UNC5B; SLC29A3; AL359183.1"]},{"name":"span","attribs":{"title":"CALML5; CALML3-AS1"},"children":["CALML5; CALML3-AS1"]},{"name":"span","attribs":{"title":"TSPAN15; FAM241B"},"children":["TSPAN15; FAM241B"]},{"name":"span","attribs":{"title":"MSMB; CXCL12; RASSF4; C10orf25; ZNF22; MARCHF8; ZFAND4; AL356056.2; SYT15; FO681492.1; AGAP14P"},"children":["MSMB; CXCL12; RASSF4; C10or..."]},{"name":"span","attribs":{"title":"MAPK8; FRMPD2"},"children":["MAPK8; FRMPD2"]},{"name":"span","attribs":{"title":"R3HCC1L; LOXL4; PYROXD2; GOT1"},"children":["R3HCC1L; LOXL4; PYROXD2; GOT1"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"ZMIZ1; ZMIZ1-AS1; LINC00595"},"children":["ZMIZ1; ZMIZ1-AS1; LINC00595"]},{"name":"span","attribs":{"title":"RPP30"},"children":["RPP30"]},{"name":"span","attribs":{"title":"RET"},"children":["RET"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"PAX2; SLF2; SEMA4G; MRPL43; BLOC1S2; SCD; SEC31B; NDUFB8; HIF1AN"},"children":["PAX2; SLF2; SEMA4G; MRPL43;..."]},{"name":"span","attribs":{"title":"ZMYND11; DIP2C"},"children":["ZMYND11; DIP2C"]},{"name":"span","attribs":{"title":"MKI67"},"children":["MKI67"]},{"name":"span","attribs":{"title":"CALML5; CALML3-AS1"},"children":["CALML5; CALML3-AS1"]},{"name":"span","attribs":{"title":"MLLT10; HNRNPRP1; DNAJC1"},"children":["MLLT10; HNRNPRP1; DNAJC1"]},{"name":"span","attribs":{"title":"KAT6B"},"children":["KAT6B"]},{"name":"span","attribs":{"title":"CEP55; MYOF"},"children":["CEP55; MYOF"]},{"name":"span","attribs":{"title":"ADK"},"children":["ADK"]},{"name":"span","attribs":{"title":"ZMIZ1; ZMIZ1-AS1"},"children":["ZMIZ1; ZMIZ1-AS1"]},{"name":"span","attribs":{"title":"FRAT2"},"children":["FRAT2"]},{"name":"span","attribs":{"title":"RET"},"children":["RET"]},{"name":"span","attribs":{"title":"ZNF487; RASGEF1A; HNRNPF"},"children":["ZNF487; RASGEF1A; HNRNPF"]},{"name":"span","attribs":{"title":"MAP3K8"},"children":["MAP3K8"]},{"name":"span","attribs":{"title":"WBP1L; STOX1; DDX50; DDX21; KIFBP; VPS26A; FAM149B1; DNAJC9"},"children":["WBP1L; STOX1; DDX50; DDX21;..."]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"SVIL-AS1"},"children":["SVIL-AS1"]},{"name":"span","attribs":{"title":"SVIL"},"children":["SVIL"]},{"name":"span","attribs":{"title":"DIP2C"},"children":["DIP2C"]},{"name":"span","attribs":{"title":"GTPBP4; IDI2-AS1; WDR37; LARP4B; AL359878.1; IDI1; PFKP"},"children":["GTPBP4; IDI2-AS1; WDR37; LA..."]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"NCOA4; MSMB; TIMM23; CXCL12; RASSF4; C10orf25; ZNF22; ALOX5; MARCHF8; ZFAND4; AL356056.2; SYT15; FO681492.1; AGAP14P"},"children":["NCOA4; MSMB; TIMM23; CXCL12..."]},{"name":"span","attribs":{"title":"LINC00595; ZMIZ1"},"children":["LINC00595; ZMIZ1"]},{"name":"span","attribs":{"title":"CACUL1"},"children":["CACUL1"]},{"name":"span","attribs":{"title":"LDB1; PPRC1; NOLC1; GBF1; NFKB2"},"children":["LDB1; PPRC1; NOLC1; GBF1; N..."]},{"name":"span","attribs":{"title":"PPP1R3C"},"children":["PPP1R3C"]},{"name":"span","attribs":{"title":"VSIR; PSAP"},"children":["VSIR; PSAP"]},{"name":"span","attribs":{"title":"ARMC3; MSRB2; YWHAZP3"},"children":["ARMC3; MSRB2; YWHAZP3"]},{"name":"span","attribs":{"title":"CPXM2"},"children":["CPXM2"]},{"name":"span","attribs":{"title":"MAT1A"},"children":["MAT1A"]},{"name":"span","attribs":{"title":"RBM20; PDCD4"},"children":["RBM20; PDCD4"]},{"name":"span","attribs":{"title":"LHPP; FAM53B; EEF1AKMT2"},"children":["LHPP; FAM53B; EEF1AKMT2"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"BMS1"},"children":["BMS1"]},{"name":"span","attribs":{"title":"LOXL4; R3HCC1L; GOT1"},"children":["LOXL4; R3HCC1L; GOT1"]},{"name":"span","attribs":{"title":"CACNB2; AL450384.2"},"children":["CACNB2; AL450384.2"]},{"name":"span","attribs":{"title":"ATRNL1"},"children":["ATRNL1"]},{"name":"span","attribs":{"title":"RNLS; ATAD1; KLLN; PTEN"},"children":["RNLS; ATAD1; KLLN; PTEN"]},{"name":"span","attribs":{"title":"SFXN2; WBP1L; BORCS7; ARL3; STOX1; DDX50; DDX21; KIFBP; FAM149B1; DNAJC9"},"children":["SFXN2; WBP1L; BORCS7; ARL3;..."]},{"name":"span","attribs":{"title":"TAF3; GATA3"},"children":["TAF3; GATA3"]},{"name":"span","attribs":{"title":"SLC35G1; PLCE1; HDAC1P1; PIPSL"},"children":["SLC35G1; PLCE1; HDAC1P1; PIPSL"]},{"name":"span","attribs":{"title":"ARID5B; RTKN2"},"children":["ARID5B; RTKN2"]},{"name":"span","attribs":{"title":"HK1; TACR2"},"children":["HK1; TACR2"]},{"name":"span","attribs":{"title":"SGPL1"},"children":["SGPL1"]},{"name":"span","attribs":{"title":"LHPP; OAT"},"children":["LHPP; OAT"]},{"name":"span","attribs":{"title":"MAP3K8"},"children":["MAP3K8"]},{"name":"span","attribs":{"title":"SLC16A9; CCDC6; BICC1"},"children":["SLC16A9; CCDC6; BICC1"]},{"name":"span","attribs":{"title":"WDFY4"},"children":["WDFY4"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"EDRF1; ALDOAP2"},"children":["EDRF1; ALDOAP2"]},{"name":"span","attribs":{"title":"RET; CSGALNACT2; RASGEF1A"},"children":["RET; CSGALNACT2; RASGEF1A"]},{"name":"span","attribs":{"title":"AL391421.1; DLG5"},"children":["AL391421.1; DLG5"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"TAF3"},"children":["TAF3"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"DOCK1"},"children":["DOCK1"]},{"name":"span","attribs":{"title":"AL391421.1; DLG5"},"children":["AL391421.1; DLG5"]},{"name":"span","attribs":{"title":"RBM20"},"children":["RBM20"]},{"name":"span","attribs":{"title":"GFRA1"},"children":["GFRA1"]},{"name":"span","attribs":{"title":"NHLRC2; DCLRE1A"},"children":["NHLRC2; DCLRE1A"]},{"name":"span","attribs":{"title":"SLC35G1; PLCE1; PIPSL"},"children":["SLC35G1; PLCE1; PIPSL"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"EPC1"},"children":["EPC1"]},{"name":"span","attribs":{"title":"DIP2C"},"children":["DIP2C"]},{"name":"span","attribs":{"title":"ANKRD30A; AL157387.1; LINC00993"},"children":["ANKRD30A; AL157387.1; LINC0..."]},{"name":"span","attribs":{"title":"RET; CSGALNACT2; RASGEF1A"},"children":["RET; CSGALNACT2; RASGEF1A"]},{"name":"span","attribs":{"title":"BTRC"},"children":["BTRC"]},{"name":"span","attribs":{"title":"SCD; OLMALINC; CWF19L1; BLOC1S2"},"children":["SCD; OLMALINC; CWF19L1; BLO..."]},{"name":"span","attribs":{"title":"ST8SIA6-AS1; ST8SIA6; CCNY"},"children":["ST8SIA6-AS1; ST8SIA6; CCNY"]},{"name":"span","attribs":{"title":"SVIL-AS1; SVIL"},"children":["SVIL-AS1; SVIL"]},{"name":"span","attribs":{"title":"MAPK8; FRMPD2"},"children":["MAPK8; FRMPD2"]},{"name":"span","attribs":{"title":"LHPP; FAM53B"},"children":["LHPP; FAM53B"]},{"name":"span","attribs":{"title":"SVIL"},"children":["SVIL"]},{"name":"span","attribs":{"title":"KAT6B; ADK"},"children":["KAT6B; ADK"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"HECTD2; PPP1R3C"},"children":["HECTD2; PPP1R3C"]},{"name":"span","attribs":{"title":"SMNDC1"},"children":["SMNDC1"]},{"name":"span","attribs":{"title":"NHLRC2"},"children":["NHLRC2"]},{"name":"span","attribs":{"title":"TYSND1; SAR1A"},"children":["TYSND1; SAR1A"]},{"name":"span","attribs":{"title":"CTNNA3"},"children":["CTNNA3"]},{"name":"span","attribs":{"title":"FRMD4A"},"children":["FRMD4A"]},{"name":"span","attribs":{"title":"KAT6B"},"children":["KAT6B"]},{"name":"span","attribs":{"title":"PFKFB3"},"children":["PFKFB3"]},{"name":"span","attribs":{"title":"KCNMA1"},"children":["KCNMA1"]},{"name":"span","attribs":{"title":"GOT1"},"children":["GOT1"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"LZTS2"},"children":["LZTS2"]},{"name":"span","attribs":{"title":"ZMIZ1; LINC00595"},"children":["ZMIZ1; LINC00595"]},{"name":"span","attribs":{"title":"MSMB; NCOA4; AGAP7P; TIMM23; CXCL12; RASSF4; C10orf25; ZNF22; MARCHF8; ZFAND4; AL356056.2; SYT15; FO681492.1; AGAP14P"},"children":["MSMB; NCOA4; AGAP7P; TIMM23..."]},{"name":"span","attribs":{"title":"SVIL2P; ARHGAP12"},"children":["SVIL2P; ARHGAP12"]},{"name":"span","attribs":{"title":"KCNMA1"},"children":["KCNMA1"]},{"name":"span","attribs":{"title":"PFKFB3"},"children":["PFKFB3"]},{"name":"span","attribs":{"title":"ERCC6"},"children":["ERCC6"]},{"name":"span","attribs":{"title":"DIP2C; LARP4B"},"children":["DIP2C; LARP4B"]},{"name":"span","attribs":{"title":"PARG"},"children":["PARG"]},{"name":"span","attribs":{"title":"SAMD8; VDAC2; DUSP13; COMTD1"},"children":["SAMD8; VDAC2; DUSP13; COMTD1"]},{"name":"span","attribs":{"title":"DIP2C; LARP4B"},"children":["DIP2C; LARP4B"]},{"name":"span","attribs":{"title":"EPC1"},"children":["EPC1"]},{"name":"span","attribs":{"title":"CAMK2G"},"children":["CAMK2G"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"ZDHHC16; EXOSC1"},"children":["ZDHHC16; EXOSC1"]},{"name":"span","attribs":{"title":"ZMIZ1; ZMIZ1-AS1"},"children":["ZMIZ1; ZMIZ1-AS1"]},{"name":"span","attribs":{"title":"CACNB2"},"children":["CACNB2"]},{"name":"span","attribs":{"title":"UNC5B; SLC29A3; AL359183.1"},"children":["UNC5B; SLC29A3; AL359183.1"]},{"name":"span","attribs":{"title":"PAPSS2"},"children":["PAPSS2"]},{"name":"span","attribs":{"title":"DUSP5"},"children":["DUSP5"]},{"name":"span","attribs":{"title":"YWHAZP5"},"children":["YWHAZP5"]},{"name":"span","attribs":{"title":"SIRT1; MYPN"},"children":["SIRT1; MYPN"]},{"name":"span","attribs":{"title":"RNLS; ATAD1; KLLN; PTEN"},"children":["RNLS; ATAD1; KLLN; PTEN"]},{"name":"span","attribs":{"title":"AKR1C3; AKR1C4; AKR1C2"},"children":["AKR1C3; AKR1C4; AKR1C2"]},{"name":"span","attribs":{"title":"RHOBTB1"},"children":["RHOBTB1"]},{"name":"span","attribs":{"title":"ZNF488"},"children":["ZNF488"]},{"name":"span","attribs":{"title":"MALRD1"},"children":["MALRD1"]},{"name":"span","attribs":{"title":"AKR1C3; AKR1C4; AKR1C2"},"children":["AKR1C3; AKR1C4; AKR1C2"]},{"name":"span","attribs":{"title":"MALRD1"},"children":["MALRD1"]},{"name":"span","attribs":{"title":"OLMALINC"},"children":["OLMALINC"]},{"name":"span","attribs":{"title":"SLK; SH3PXD2A; STN1"},"children":["SLK; SH3PXD2A; STN1"]},{"name":"span","attribs":{"title":"HK1; TSPAN15; TACR2"},"children":["HK1; TSPAN15; TACR2"]},{"name":"span","attribs":{"title":"CTBP2"},"children":["CTBP2"]},{"name":"span","attribs":{"title":"UNC5B; SLC29A3; AL359183.1"},"children":["UNC5B; SLC29A3; AL359183.1"]},{"name":"span","attribs":{"title":"KAT6B; ADK"},"children":["KAT6B; ADK"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"DNAJB12"},"children":["DNAJB12"]},{"name":"span","attribs":{"title":"SAMD8; VDAC2; COMTD1"},"children":["SAMD8; VDAC2; COMTD1"]},{"name":"span","attribs":{"title":"BMPR1A; SNCG; SHLD2; MMRN2; ADIRF-AS1; GLUD1"},"children":["BMPR1A; SNCG; SHLD2; MMRN2;..."]},{"name":"span","attribs":{"title":"REEP3; JMJD1C; TATDN1P1"},"children":["REEP3; JMJD1C; TATDN1P1"]},{"name":"span","attribs":{"title":"CPXM2"},"children":["CPXM2"]},{"name":"span","attribs":{"title":"RAB18; MKX"},"children":["RAB18; MKX"]},{"name":"span","attribs":{"title":"PIP4K2A"},"children":["PIP4K2A"]},{"name":"span","attribs":{"title":"AKR1B10P1; SIRT1; CTNNA3; DNAJC12"},"children":["AKR1B10P1; SIRT1; CTNNA3; D..."]},{"name":"span","attribs":{"title":"SVIL2P"},"children":["SVIL2P"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"PARD3"},"children":["PARD3"]},{"name":"span","attribs":{"title":"PARD3"},"children":["PARD3"]},{"name":"span","attribs":{"title":"PPIAP31; EPC1"},"children":["PPIAP31; EPC1"]},{"name":"span","attribs":{"title":"CNNM2; RPEL1; NT5C2"},"children":["CNNM2; RPEL1; NT5C2"]},{"name":"span","attribs":{"title":"AKR1B10P1; SIRT1; DNAJC12"},"children":["AKR1B10P1; SIRT1; DNAJC12"]},{"name":"span","attribs":{"title":""},"children":[""]},{"name":"span","attribs":{"title":"KCNMA1"},"children":["KCNMA1"]},{"name":"span","attribs":{"title":"LDB1"},"children":["LDB1"]}],"minWidth":150}],"filterable":true,"defaultPageSize":10,"showPageSizeOptions":true,"pageSizeOptions":[10,20,50,200],"paginationType":"numbers","showPageInfo":true,"minRows":1,"borderless":true,"dataKey":"75bd49988e50d3902a5fe236cba443f1","key":"75bd49988e50d3902a5fe236cba443f1"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</div>
</div>
<div id="inspection-of-top-ranked-regions" class="section level2 tabset">
<h2 class="tabset">Inspection of Top-Ranked Regions</h2>
<pre class="r"><code>## Define a GRL with the key ranges.
## Then we can step through it &amp; make all of the requisite plots
grl_to_plot &lt;- GRangesList(
  top_up_by_fdr = merged_results %&gt;% 
    filter(logFC &gt; 0) %&gt;% 
    filter(!!sym(fdr_column) == min(!!sym(fdr_column))),
  top_up_by_lfc = merged_results %&gt;% 
    filter(logFC == max(logFC)),
  top_up_by_expr = merged_results %&gt;% 
    filter(logFC &gt; 0, !!sym(fdr_column) &lt; fdr_alpha) %&gt;% 
    filter(AveExpr == max(AveExpr)),
  top_up_no_macs2 = merged_results %&gt;% 
    filter(logFC &gt; 0, !!sym(fdr_column) &lt; fdr_alpha, !overlaps_ref) %&gt;% 
    arrange(!!sym(fdr_column)),
  top_down_by_fdr = merged_results %&gt;% 
    filter(logFC &lt; 0) %&gt;% 
    filter(!!sym(fdr_column) == min(!!sym(fdr_column))),
  top_down_by_lfc = merged_results %&gt;% 
    filter(logFC == min(logFC)),
  top_down_by_expr = merged_results %&gt;% 
    filter(logFC &lt; 0, !!sym(fdr_column) &lt; fdr_alpha) %&gt;% 
    filter(AveExpr == max(AveExpr)),
  top_up_no_macs2 = merged_results %&gt;% 
    filter(logFC &lt; 0, !!sym(fdr_column) &lt; fdr_alpha, !overlaps_ref) %&gt;% 
    arrange(!!sym(fdr_column)) 
) %&gt;% 
  unlist() %&gt;% 
  .[!duplicated(.)] %&gt;% 
  splitAsList(names(.)) %&gt;% 
  lapply(setNames, c()) %&gt;% 
  lapply(function(x) x[1]) %&gt;% 
  GRangesList()</code></pre>
<pre class="r"><code>## The coverage
bwfl &lt;- list2(
  &quot;{target}&quot; := file.path(
    bw_path, glue(&quot;{treat_levels}_merged_treat_pileup.bw&quot;)
  ) %&gt;% 
    BigWigFileList() %&gt;% 
    setNames(treat_levels)
)
line_col &lt;- list2(&quot;{target}&quot; := treat_colours)
## Coverage annotations
annot &lt;- merged_results %&gt;% 
  splitAsList(.$status) %&gt;% 
  .[vapply(., length, integer(1)) &gt; 0] %&gt;% 
  endoapply(granges) %&gt;% 
  list() %&gt;% 
  setNames(target)
## Coverage y-limits
ind &lt;- filtered_counts %&gt;% 
  assay(&quot;counts&quot;) %&gt;% 
  apply(MARGIN = 2, which.max) %&gt;%
  unique()
max_ranges &lt;- rowRanges(filtered_counts[ind])
y_lim &lt;- bwfl[[target]] %&gt;%
  lapply(import.bw, which = max_ranges) %&gt;%
  lapply(function(x) c(0, max(x$score))) %&gt;% 
  unlist() %&gt;% 
  range() %&gt;% 
  list() %&gt;% 
  setNames(target)

## The features track
feat_gr &lt;- gene_regions %&gt;% 
  lapply(granges) %&gt;% 
  GRangesList()
feature_colours &lt;- colours$regions
if (has_features) {
  feat_gr &lt;- feat_gr %&gt;% 
    c(
      splitAsList(external_features, external_features$feature)
    )
  feature_colours &lt;- c(feature_colours, colours$features)
}

## The genes track
hfgc_genes &lt;- trans_models
gene_col &lt;- &quot;grey&quot;
if (any_detected){
  rna_lfc_col &lt;- colnames(rnaseq)[str_detect(str_to_lower(colnames(rnaseq)), &quot;logfc&quot;)][1]
  rna_fdr_col &lt;- colnames(rnaseq)[str_detect(str_to_lower(colnames(rnaseq)), &quot;fdr|adjp&quot;)][1]
  if (!is.na(rna_lfc_col) &amp; !is.na(rna_fdr_col)) {
    hfgc_genes &lt;- trans_models %&gt;% 
      mutate(
        status = case_when(
          !gene %in% rnaseq$gene_id ~ &quot;Undetected&quot;,
          gene %in% dplyr::filter(
            rnaseq, !!sym(rna_lfc_col) &gt; 0, !!sym(rna_fdr_col) &lt; fdr_alpha
          )$gene_id ~ &quot;Up&quot;,
          gene %in% dplyr::filter(
            rnaseq, !!sym(rna_lfc_col) &lt; 0, !!sym(rna_fdr_col) &lt; fdr_alpha
          )$gene_id ~ &quot;Down&quot;,
          gene %in% dplyr::filter(
            rnaseq, !!sym(rna_fdr_col) &gt;= fdr_alpha
          )$gene_id ~ &quot;Unchanged&quot;,
        )
      ) %&gt;% 
      splitAsList(.$status) %&gt;% 
      lapply(select, -status) %&gt;% 
      GRangesList()
    gene_col &lt;- colours$direction %&gt;% 
      setNames(str_to_title(names(.)))
  }
}

## External Coverage (Optional)
if (!is.null(config$external$coverage)) {
  ext_cov_path &lt;- config$external$coverage %&gt;% 
    lapply(unlist) %&gt;% 
    lapply(function(x) setNames(here::here(x), names(x)))
  bwfl &lt;- c(
    bwfl[target],
    lapply(ext_cov_path, function(x) BigWigFileList(x) %&gt;% setNames(names(x)))
  )
  line_col &lt;- c(
    line_col[target],
    ext_cov_path %&gt;% 
      lapply(
        function(x) {
          missing &lt;- setdiff(names(x), names(treat_colours))
          cmn &lt;- intersect(names(x), names(treat_colours))
          col &lt;- setNames(character(length(names(x))), names(x))
          if (length(cmn) &gt; 0) col[cmn] &lt;- treat_colours[cmn]
          if (length(missing) &gt; 0) 
            col[missing] &lt;- hcl.colors(
              max(5, length(missing)), &quot;Zissou 1&quot;)[seq_along(missing)]
          col
        }
      )
  )
  
  y_ranges &lt;- grl_to_plot %&gt;% 
    unlist() %&gt;% 
    granges() %&gt;% 
    resize(w = 10 * width(.), fix = &#39;center&#39;)
  
  y_lim &lt;- c(
    y_lim[target],
    bwfl[names(bwfl) != target] %&gt;% 
      lapply(
        function(x) {
          GRangesList(lapply(x, import.bw, which = y_ranges)) %&gt;% 
            unlist() %&gt;% 
            filter(score == max(score)) %&gt;% 
            mcols() %&gt;% 
            .[[&quot;score&quot;]] %&gt;% 
            c(0) %&gt;% 
            range()
        }
      )
  )
  
}</code></pre>
<pre class="r"><code>makeCaption &lt;- function(.gr) {
  if (is.null(.gr)) return(NULL)
  dir &lt;- ifelse(.gr$direction == &quot;Down&quot;, &#39;decreased&#39;, &#39;increased&#39;)
  reg &lt;- case_when(
    str_detect(.gr$region, &quot;Inter&quot;) ~ paste(&quot;an&quot;, .gr$region, &quot;region&quot;),
    str_detect(.gr$region, &quot;Upstream&quot;) ~ paste(&quot;an&quot;, .gr$region),
    str_detect(.gr$region, &quot;(Ex|Intr)on&quot;) ~ paste(&quot;an&quot;, .gr$region),
    str_detect(.gr$region, &quot;^Prom&quot;) ~ paste(&quot;a&quot;, .gr$region)
  )
  feat &lt;- c()
  if (has_features) feat &lt;- case_when(
    str_detect(.gr$feature, &quot;^[AEIOU]&quot;) ~ paste(&quot;an&quot;, .gr$feature),
    !str_detect(.gr$feature, &quot;^[AEIOU]&quot;) ~ paste(&quot;a&quot;, .gr$feature)
  )
  gn &lt;- unlist(.gr$gene_name)
  cp &lt;- c(
    glue(
      &quot;*The {width(.gr)}bp region showing {dir} {target} binding in response to &quot;, 
      &quot;{treat_levels[[2]]} treatment (FDR = {sprintf(&#39;%.2e&#39;, mcols(.gr)[[fdr_column]])}). &quot;,
      &quot;The range mostly overlapped with {reg}, with all &quot;,
      &quot;defined regions shown as a contiguous bar. &quot;,
      ifelse(
        has_features,
        glue(
          &quot;Using the features supplied in {basename(config$external$features)}, &quot;,
          &quot;this mostly overlapped {feat}, shown as a separate block &quot;,
          &quot;with the gene-centric regions. &quot;
        ),
        glue(&quot;&quot;)
      ),
    ),
    ifelse(
      .gr$overlaps_ref,
      paste(
        &quot;A consensus peak overlapping this region was identified by&quot;,
        &quot;`macs2 callpeak` when using merged samples.&quot;
      ),
      &quot;No consensus peak was identified using `macs2 callpeak`.&quot;
    ),
    ifelse(
      length(gn) &gt; 0,
      paste0(
        &quot;Using the above mapping strategy, this range is likely to regulate &quot;,
        collapseGenes(gn, format = &quot;&quot;)
      ),
      &quot;No genes were able to be assigned to this region.&quot;
    ),
    paste(
      &quot;For each sample, the y-axis limits represent the values from the window&quot;, 
      &quot;with the highest signal.*&quot;
    )
  )
  paste(cp, collapse = &quot; &quot;)
}</code></pre>
<div id="normalised-sliding-windows" class="section level3">
<h3>Normalised Sliding Windows</h3>
<pre class="r"><code>grl_to_plot %&gt;% 
  lapply(
    function(x){
      filtered_counts %&gt;% 
        subsetByOverlaps(x) %&gt;% 
        assay(&quot;qsmooth&quot;) %&gt;% 
        as_tibble(rownames = &quot;window&quot;) %&gt;% 
        pivot_longer(
          cols = all_of(colnames(filtered_counts)),
          names_to = &quot;sample&quot;,
          values_to = &quot;logCPM&quot;
        ) %&gt;% 
        mutate(range = as.character(x)) %&gt;% 
        list()
    }
  ) %&gt;% 
  as_tibble() %&gt;% 
  pivot_longer(everything()) %&gt;% 
  unnest(everything()) %&gt;% 
  mutate(
    direction = str_extract(name, &quot;down|up&quot;),
    by = str_extract(name, &quot;by_.+&quot;),
    across(
      all_of(c(&quot;name&quot;, &quot;direction&quot;, &quot;by&quot;)), str_sep_to_title
    ),
    name = str_remove_all(name, &quot;Top &quot;) %&gt;% 
      str_replace_all(&quot;Lfc&quot;, &quot;logFC&quot;) %&gt;% 
      str_replace_all(&quot;Fdr&quot;, &quot;FDR&quot;) %&gt;% 
      str_replace_all(&quot;Expr&quot;, &quot;Highest Signal&quot;) %&gt;% 
      str_replace_all(&quot;By (.+)&quot;, &quot;(\\1)&quot;) %&gt;% 
      str_replace_all(&quot;No Macs2&quot;, &quot;(No Macs2 Peak)&quot;),
    window = as.integer(window)
  ) %&gt;% 
  left_join(samples, by = &quot;sample&quot;) %&gt;% 
  ggplot(
    aes(window, logCPM, colour = treat, linetype = !!sym(rep_col))
  ) +
  geom_line(
    data = . %&gt;% 
      group_by(name) %&gt;% 
      dplyr::filter(max(window) &gt; 1)
  ) + 
  geom_point(
    data = . %&gt;% 
      group_by(name) %&gt;% 
      dplyr::filter(max(window) == 1),
    show.legend = FALSE
  ) + 
  facet_wrap(~name + range, scales = &quot;free_x&quot;) +
  scale_colour_manual(values = treat_colours) + 
  labs(
    x = &quot;Sliding Window\n(Unmerged)&quot;, y = &quot;SQ-Normalised logCPM&quot;, colour = &quot;Treat&quot;,
    linetype = str_sep_to_title(rep_col)
  )</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-qsmooth-top-ranges-1.png" alt="*Most highly ranked ranges for both gained and decreased ER binding in repsonse to DHT treatment. The smooth-quantile normalised values are shown across the initial set of sliding windows before merging. Ranges were chosen as the most extreme for FDR, Binding Strength (Signal) and logFC. Windows are shown free of the genomic context.*" width="960" />
<p class="caption">
<em>Most highly ranked ranges for both gained and decreased ER binding in repsonse to DHT treatment. The smooth-quantile normalised values are shown across the initial set of sliding windows before merging. Ranges were chosen as the most extreme for FDR, Binding Strength (Signal) and logFC. Windows are shown free of the genomic context.</em>
</p>
</div>
</div>
<div id="decreased-binding-strongest-signal" class="section level3">
<h3>Decreased Binding (Strongest Signal)</h3>
<pre class="r"><code>plotHFGC(
  grl_to_plot$top_down_by_expr,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = colours$direction %&gt;% 
    setNames(str_to_title(names(.))) %&gt;% 
    .[unique(unlist(lapply(annot, names)))],
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10,
  max = 10 * width(grl_to_plot$top_down_by_expr),
  ylim = y_lim,
  featsize = 2,
  col.title = &quot;black&quot;, background.title = &quot;white&quot;, showAxis = FALSE,
  rotation.title = 90
)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-top-down-by-expr-1.png" alt="*The 2300bp region showing decreased ER binding in response to DHT treatment (FDR = 1.11e-03). The range mostly overlapped with an Intron, with all defined regions shown as a contiguous bar. Using the features supplied in h3k27ac_features.gtf, this mostly overlapped a Promoter, shown as a separate block with the gene-centric regions.  A consensus peak overlapping this region was identified by `macs2 callpeak` when using merged samples. Using the above mapping strategy, this range is likely to regulate SEC61A2 For each sample, the y-axis limits represent the values from the window with the highest signal.*" width="960" />
<p class="caption">
<em>The 2300bp region showing decreased ER binding in response to DHT treatment (FDR = 1.11e-03). The range mostly overlapped with an Intron, with all defined regions shown as a contiguous bar. Using the features supplied in h3k27ac_features.gtf, this mostly overlapped a Promoter, shown as a separate block with the gene-centric regions. A consensus peak overlapping this region was identified by <code>macs2 callpeak</code> when using merged samples. Using the above mapping strategy, this range is likely to regulate SEC61A2 For each sample, the y-axis limits represent the values from the window with the highest signal.</em>
</p>
</div>
</div>
<div id="decreased-binding-lowest-fdr" class="section level3">
<h3>Decreased Binding (Lowest FDR)</h3>
<pre class="r"><code>plotHFGC(
  grl_to_plot$top_down_by_fdr,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = colours$direction %&gt;% 
    setNames(str_to_title(names(.))) %&gt;% 
    .[unique(unlist(lapply(annot, names)))],
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10,
  max = 10 * width(grl_to_plot$top_down_by_fdr),
  ylim = y_lim,
  featsize = 2,
  col.title = &quot;black&quot;, background.title = &quot;white&quot;, showAxis = FALSE,
  rotation.title = 90
)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-top_down_by_fdr-1.png" alt="*The 1500bp region showing decreased ER binding in response to DHT treatment (FDR = 2.63e-08). The range mostly overlapped with an Intergenic (&gt;10kb) region, with all defined regions shown as a contiguous bar. Using the features supplied in h3k27ac_features.gtf, this mostly overlapped a Super Enhancer, shown as a separate block with the gene-centric regions.  A consensus peak overlapping this region was identified by `macs2 callpeak` when using merged samples. Using the above mapping strategy, this range is likely to regulate AL137026.2 For each sample, the y-axis limits represent the values from the window with the highest signal.*" width="960" />
<p class="caption">
<em>The 1500bp region showing decreased ER binding in response to DHT treatment (FDR = 2.63e-08). The range mostly overlapped with an Intergenic (&gt;10kb) region, with all defined regions shown as a contiguous bar. Using the features supplied in h3k27ac_features.gtf, this mostly overlapped a Super Enhancer, shown as a separate block with the gene-centric regions. A consensus peak overlapping this region was identified by <code>macs2 callpeak</code> when using merged samples. Using the above mapping strategy, this range is likely to regulate AL137026.2 For each sample, the y-axis limits represent the values from the window with the highest signal.</em>
</p>
</div>
</div>
<div id="decreased-binding-most-extreme" class="section level3">
<h3>Decreased Binding (Most Extreme)</h3>
<pre class="r"><code>plotHFGC(
  grl_to_plot$top_down_by_lfc,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = direction_colours,
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10, 
  max = 10 * width(grl_to_plot$top_down_by_lfc),
  ylim = y_lim,
  featsize = 2,
  col.title = &quot;black&quot;, background.title = &quot;white&quot;, showAxis = FALSE,
  rotation.title = 90
)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-top_down_by_lfc-1.png" alt="*The 1200bp region showing decreased ER binding in response to DHT treatment (FDR = 3.72e-08). The range mostly overlapped with an Intergenic (&gt;10kb) region, with all defined regions shown as a contiguous bar. Using the features supplied in h3k27ac_features.gtf, this mostly overlapped an Enhancer, shown as a separate block with the gene-centric regions.  A consensus peak overlapping this region was identified by `macs2 callpeak` when using merged samples. Using the above mapping strategy, this range is likely to regulate AL450322.1, AL450322.2 and KLF6 For each sample, the y-axis limits represent the values from the window with the highest signal.*" width="960" />
<p class="caption">
<em>The 1200bp region showing decreased ER binding in response to DHT treatment (FDR = 3.72e-08). The range mostly overlapped with an Intergenic (&gt;10kb) region, with all defined regions shown as a contiguous bar. Using the features supplied in h3k27ac_features.gtf, this mostly overlapped an Enhancer, shown as a separate block with the gene-centric regions. A consensus peak overlapping this region was identified by <code>macs2 callpeak</code> when using merged samples. Using the above mapping strategy, this range is likely to regulate AL450322.1, AL450322.2 and KLF6 For each sample, the y-axis limits represent the values from the window with the highest signal.</em>
</p>
</div>
</div>
<div id="increased-binding-strongest-signal" class="section level3">
<h3>Increased Binding (Strongest Signal)</h3>
<pre class="r"><code>plotHFGC(
  grl_to_plot$top_up_by_expr,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = colours$direction %&gt;% 
    setNames(str_to_title(names(.))) %&gt;% 
    .[unique(unlist(lapply(annot, names)))],
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10,
  max = 10 * width(grl_to_plot$top_up_by_expr),
  ylim = y_lim,
  featsize = 2,
  col.title = &quot;black&quot;, background.title = &quot;white&quot;, showAxis = FALSE,
  rotation.title = 90
)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-top-up-by-expr-1.png" alt="*The 600bp region showing increased ER binding in response to DHT treatment (FDR = 4.46e-02). The range mostly overlapped with an Intergenic (&lt;10kb) region, with all defined regions shown as a contiguous bar. Using the features supplied in h3k27ac_features.gtf, this mostly overlapped a No Feature, shown as a separate block with the gene-centric regions.  No consensus peak was identified using `macs2 callpeak`. Using the above mapping strategy, this range is likely to regulate CNNM2, LINC02655, NT5C2, SH2D4B and WARS2P1 For each sample, the y-axis limits represent the values from the window with the highest signal.*" width="960" />
<p class="caption">
<em>The 600bp region showing increased ER binding in response to DHT treatment (FDR = 4.46e-02). The range mostly overlapped with an Intergenic (&lt;10kb) region, with all defined regions shown as a contiguous bar. Using the features supplied in h3k27ac_features.gtf, this mostly overlapped a No Feature, shown as a separate block with the gene-centric regions. No consensus peak was identified using <code>macs2 callpeak</code>. Using the above mapping strategy, this range is likely to regulate CNNM2, LINC02655, NT5C2, SH2D4B and WARS2P1 For each sample, the y-axis limits represent the values from the window with the highest signal.</em>
</p>
</div>
</div>
<div id="increased-binding-lowest-fdr" class="section level3">
<h3>Increased Binding (Lowest FDR)</h3>
<pre class="r"><code>plotHFGC(
  grl_to_plot$top_up_by_fdr,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = colours$direction %&gt;% 
    setNames(str_to_title(names(.))) %&gt;% 
    .[unique(unlist(lapply(annot, names)))],
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10,
  max = 10 * width(grl_to_plot$top_up_by_fdr),
  ylim = y_lim,
  featsize = 2,
  col.title = &quot;black&quot;, background.title = &quot;white&quot;, showAxis = FALSE,
  rotation.title = 90
)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-top_up_by_fdr-1.png" alt="*The 500bp region showing increased ER binding in response to DHT treatment (FDR = 9.88e-03). The range mostly overlapped with an Intergenic (&gt;10kb) region, with all defined regions shown as a contiguous bar. Using the features supplied in h3k27ac_features.gtf, this mostly overlapped a No Feature, shown as a separate block with the gene-centric regions.  A consensus peak overlapping this region was identified by `macs2 callpeak` when using merged samples. Using the above mapping strategy, this range is likely to regulate AC025946.1, AC025946.2, AL138773.1 and LINC00709 For each sample, the y-axis limits represent the values from the window with the highest signal.*" width="960" />
<p class="caption">
<em>The 500bp region showing increased ER binding in response to DHT treatment (FDR = 9.88e-03). The range mostly overlapped with an Intergenic (&gt;10kb) region, with all defined regions shown as a contiguous bar. Using the features supplied in h3k27ac_features.gtf, this mostly overlapped a No Feature, shown as a separate block with the gene-centric regions. A consensus peak overlapping this region was identified by <code>macs2 callpeak</code> when using merged samples. Using the above mapping strategy, this range is likely to regulate AC025946.1, AC025946.2, AL138773.1 and LINC00709 For each sample, the y-axis limits represent the values from the window with the highest signal.</em>
</p>
</div>
</div>
<div id="significant-increased-binding-no-consensus-peak" class="section level3">
<h3>Significant Increased Binding (No Consensus Peak)</h3>
<pre class="r"><code>plotHFGC(
  grl_to_plot$top_up_no_macs2,
  hic = hic,
  features = feat_gr,featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col,
  coverage = bwfl, linecol = line_col,
  annotation = annot, 
  annotcol = colours$direction %&gt;% 
    setNames(str_to_title(names(.))) %&gt;% 
    .[unique(unlist(lapply(annot, names)))],
  cytobands = bands_df,
  collapseTranscripts = FALSE,
  zoom = 10,
  max = 10 * width(grl_to_plot$top_up_no_macs2),
  ylim = y_lim,
  featsize = 2,
  col.title = &quot;black&quot;, background.title = &quot;white&quot;, showAxis = FALSE,
  rotation.title = 90
)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-top_up_no_macs2-1.png" alt="*The 600bp region showing increased ER binding in response to DHT treatment (FDR = 1.92e-02). The range mostly overlapped with an Intron, with all defined regions shown as a contiguous bar. Using the features supplied in h3k27ac_features.gtf, this mostly overlapped a No Feature, shown as a separate block with the gene-centric regions.  No consensus peak was identified using `macs2 callpeak`. Using the above mapping strategy, this range is likely to regulate AC026396.1, ADARB2, AL731533.1, AL731533.2, AL731533.3 and PFKP For each sample, the y-axis limits represent the values from the window with the highest signal.*" width="960" />
<p class="caption">
<em>The 600bp region showing increased ER binding in response to DHT treatment (FDR = 1.92e-02). The range mostly overlapped with an Intron, with all defined regions shown as a contiguous bar. Using the features supplied in h3k27ac_features.gtf, this mostly overlapped a No Feature, shown as a separate block with the gene-centric regions. No consensus peak was identified using <code>macs2 callpeak</code>. Using the above mapping strategy, this range is likely to regulate AC026396.1, ADARB2, AL731533.1, AL731533.2, AL731533.3 and PFKP For each sample, the y-axis limits represent the values from the window with the highest signal.</em>
</p>
</div>
</div>
</div>
</div>
<div id="enrichment-testing" class="section level1 tabset">
<h1 class="tabset">Enrichment Testing</h1>
<pre class="r"><code>msigdb &lt;- msigdbr(species = &quot;Homo sapiens&quot;) %&gt;% 
  dplyr::filter(
    gs_cat %in% unlist(params$msigdb$gs_cat) |
      gs_subcat %in% unlist(params$msigdb$gs_subcat),
    str_detect(ensembl_gene, &quot;^E&quot;)
  ) %&gt;% 
  dplyr::rename(gene_id = ensembl_gene, gene_name = gene_symbol) %&gt;% # For easier integration 
  dplyr::select(-starts_with(&quot;human&quot;), -contains(&quot;entrez&quot;)) %&gt;% 
  dplyr::filter(gene_id %in% genes_gr$gene_id) %&gt;% 
  mutate(gs_url = str_extract(gs_url, &quot;^[^|]+&quot;))
gs_by_gsid &lt;- msigdb %&gt;% 
  split(.$gs_name) %&gt;% 
  lapply(pull, &quot;gene_id&quot;) %&gt;% 
  lapply(unique)
gs_by_geneid &lt;- msigdb %&gt;% 
  split(.$gene_id) %&gt;% 
  lapply(pull, &quot;gs_name&quot;) %&gt;% 
  lapply(unique)</code></pre>
<p>As an initial exploration, the retained windows were compared to pre-defined gene-sets. These were taken from the <code>MSigDB</code> database and the gene-sets used for enrichment testing were:</p>
<pre class="r"><code>msigdb %&gt;% 
  distinct(gs_cat, gs_subcat, gs_name) %&gt;%
  group_by(gs_cat, gs_subcat) %&gt;% 
  summarise(`Gene Sets` = dplyr::n(),.groups = &quot;drop&quot;) %&gt;% 
  dplyr::rename(Category = gs_cat, `Sub-Category` = gs_subcat) %&gt;% 
  mutate(
    Description = case_when(
      Category == &quot;H&quot; ~ &quot;Hallmark Gene-Sets&quot;,
      Category == &quot;C1&quot; ~ &quot;Positional Gene Sets&quot;,
      Category == &quot;C2&quot; ~ str_replace_all(`Sub-Category`, &quot;CP:(.+)&quot;, &quot;Curated Gene-Sets: \\1&quot;),
      str_detect(`Sub-Category`, &quot;TFT&quot;) ~ str_replace_all(`Sub-Category`, &quot;TFT:(.+)&quot;, &quot;Transcription Factor Target Prediction Gene-Sets: \\1&quot;),
      str_detect(`Sub-Category`, &quot;MIR&quot;) ~ str_replace_all(`Sub-Category`, &quot;MIR:(.+)&quot;, &quot;microRNA Target Gene-Sets: \\1&quot;),
      str_detect(`Sub-Category`, &quot;CGN&quot;) ~ &quot;Cancer Gene Neighbourhoods&quot;,
      str_detect(`Sub-Category`, &quot;CM&quot;) ~ &quot;Cancer Modules&quot;,
      str_detect(`Sub-Category`, &quot;GO:BP&quot;) ~ &quot;Gene Ontology: Bological Process&quot;,
      str_detect(`Sub-Category`, &quot;GO:MF&quot;) ~ &quot;Gene Ontology: Molecular Function&quot;,
      str_detect(`Sub-Category`, &quot;GO:CC&quot;) ~ &quot;Gene Ontology: Cellular Component&quot;,
      str_detect(`Sub-Category`, &quot;HPO&quot;) ~ &quot;Human Phenotype Ontology&quot;,
      Category == &quot;C6&quot; ~ &quot;Oncogenic Signature Gene Sets&quot;,
      `Sub-Category` == &quot;IMMUNESIGDB&quot; ~ &quot;ImmuneSigDB Gene Sets&quot;,
      `Sub-Category` == &quot;VAX&quot; ~ &quot;Vaccine Response Gene Sets&quot;,
      Category == &quot;C8&quot; ~ &quot;Cell Type Signature Gene Sets&quot;
    ),
    Category = factor(Category, levels = c(&quot;H&quot;, paste0(&quot;C&quot;, 1:8)))
  ) %&gt;% 
  droplevels() %&gt;% 
  arrange(Category) %&gt;% 
  pander(
    justify = &quot;llrl&quot;, 
    caption = &quot;*Summary of Gene-Sets used for Enrichment Testing*&quot;
  )</code></pre>
<table style="width:99%;">
<caption><em>Summary of Gene-Sets used for Enrichment Testing</em></caption>
<colgroup>
<col width="15%" />
<col width="25%" />
<col width="16%" />
<col width="41%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Category</th>
<th align="left">Sub-Category</th>
<th align="right">Gene Sets</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">H</td>
<td align="left"></td>
<td align="right">50</td>
<td align="left">Hallmark Gene-Sets</td>
</tr>
<tr class="even">
<td align="left">C2</td>
<td align="left">CP:KEGG</td>
<td align="right">186</td>
<td align="left">Curated Gene-Sets: KEGG</td>
</tr>
<tr class="odd">
<td align="left">C2</td>
<td align="left">CP:REACTOME</td>
<td align="right">1,615</td>
<td align="left">Curated Gene-Sets: REACTOME</td>
</tr>
<tr class="even">
<td align="left">C2</td>
<td align="left">CP:WIKIPATHWAYS</td>
<td align="right">664</td>
<td align="left">Curated Gene-Sets: WIKIPATHWAYS</td>
</tr>
<tr class="odd">
<td align="left">C3</td>
<td align="left">TFT:GTRD</td>
<td align="right">517</td>
<td align="left">Transcription Factor Target Prediction Gene-Sets: GTRD</td>
</tr>
</tbody>
</table>
<p>Retained windows were tested for enrichment of these gene-sets using:</p>
<ol style="list-style-type: decimal">
<li>Any window mapped to a gene from these gene-sets</li>
<li>Any window with evidence of differential binding mapped to a gene from these gene-sets</li>
</ol>
<p>In the first case, mapped genes were tested for enrichment against all annotated genes. In the second case, i.e. Differentially Bound Windows, the control set of genes were those mapped to a window (i.e. those test set from step 1). All enrichment testing ws performed using <code>goseq</code> with gene-length as the bias term to offset biased sampling of genes</p>
<div id="windows-with-detected-er" class="section level2">
<h2>Windows With Detected ER</h2>
<pre class="r"><code>pwf_mapped &lt;- genes_gr %&gt;% 
  mutate(w = width(.)) %&gt;% 
  as_tibble() %&gt;% 
  dplyr::select(gene_id, w) %&gt;% 
  arrange(desc(w)) %&gt;% 
  distinct(gene_id, .keep_all = TRUE) %&gt;% 
  mutate(mapped = gene_id %in% unlist(merged_results$gene_id)) %&gt;% 
  with(
    nullp(
      structure(mapped, names = gene_id),
      genome = config$genome$build,
      bias.data = log10(w),
      plot.fit = FALSE
    )
  )
goseq_mapped &lt;- goseq(
  pwf_mapped, config$genome$build, gene2cat = gs_by_geneid
) %&gt;% 
  dplyr::rename(gs_name = category, pval = over_represented_pvalue) %&gt;% 
  dplyr::select(-starts_with(&quot;under&quot;)) %&gt;% 
  as_tibble() %&gt;% 
  arrange(pval) %&gt;% 
  dplyr::filter(numDEInCat &gt; 0) %&gt;% 
  mutate(fdr = p.adjust(pval, &quot;fdr&quot;))</code></pre>
<pre class="r"><code>cp &lt;- htmltools::em(
  glue(
    &quot;All {sum(goseq_mapped$fdr &lt; fdr_alpha)} gene sets considered as &quot;, 
    &quot;significantly enriched (FDR &lt; {fdr_alpha}) amongst the merged windows &quot;, 
    &quot;with detectable {target} binding. Genes mapped to a merged window were &quot;,
    &quot;compared to those not mapped to any merged windows.&quot;
  )
)
tbl &lt;- goseq_mapped %&gt;%
  dplyr::filter(fdr &lt; fdr_alpha) %&gt;% 
  left_join(
    dplyr::select(msigdb, gs_name, gene_name, gene_id, gs_url, gs_description),
    by = &quot;gs_name&quot;
  ) %&gt;% 
  dplyr::filter(gene_id %in% unlist(merged_results$gene_id)) %&gt;% 
  dplyr::select(-gene_id) %&gt;% 
  chop(gene_name) %&gt;% 
  mutate(
    gene_name = vapply(gene_name, paste, character(1), collapse = &quot;; &quot;),
    `%` = numDEInCat / numInCat
  ) %&gt;% 
  nest(gs_name = all_of(c(&quot;gs_name&quot;, &quot;gs_url&quot;))) %&gt;% 
  dplyr::select(
    `Gene Set` = gs_name, Description = gs_description,
    `%`, numDEInCat, numInCat, 
    p = pval, FDR = fdr, 
    `Mapped Genes` = gene_name
  ) %&gt;% 
  reactable(
    filterable = TRUE,
    columns = list(
      &quot;Gene Set&quot; = colDef(
        minWidth = 150,
        cell = function(value) htmltools::tags$a(
          href = value$gs_url, target = &quot;_blank&quot;, 
          str_replace_all(value$gs_name, &quot;_&quot;, &quot; &quot;)
        ),
        html = TRUE
      ),
      Description = colDef(
        minWidth = 150,
        cell = function(value) with_tooltip(value, width = 100)
      ),
      numDEInCat = colDef(name = &quot;Total Mapped&quot;, maxWidth = 80),
      numInCat = colDef(
        name = &quot;Gene Set Size&quot;,
        maxWidth = 80,
        cell = function(value) comma(value, 1)
      ),
      &quot;%&quot; = colDef(
        name = &quot;% Mapped&quot;,
        cell = function(value) percent(value, 0.1),
        maxWidth = 80
      ),
      p = colDef(
        cell = function(value) {
          ifelse(
            value &lt; 0.001, sprintf(&quot;%.2e&quot;, value), sprintf(&quot;%.3f&quot;, value)
          )
        },
        maxWidth = 80
      ),
      FDR = colDef(
        cell = function(value) {
          ifelse(
            value &lt; 0.001, sprintf(&quot;%.2e&quot;, value), sprintf(&quot;%.3f&quot;, value)
          )
        },
        maxWidth = 80
      ),
      &quot;Mapped Genes&quot; = colDef(
        cell = function(value) with_tooltip(value, width = 100),
        minWidth = 150
      )
    )
  )
div(class = &quot;table&quot;,
  div(class = &quot;table-header&quot;,
      div(class = &quot;caption&quot;, cp),
      tbl
  )
)</code></pre>
<div class="table">
<div class="table-header">
<div class="caption">
<em>All 4 gene sets considered as significantly enriched (FDR &lt; 0.05) amongst the merged windows with detectable ER binding. Genes mapped to a merged window were compared to those not mapped to any merged windows.</em>
</div>
<div id="htmlwidget-10342757fa1a68d016d0" class="reactable html-widget" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-10342757fa1a68d016d0">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"Gene Set":[{"gs_name":["AUTS2_TARGET_GENES"],"gs_url":["http://gtrd.biouml.org/"]},{"gs_name":["RYBP_TARGET_GENES"],"gs_url":["http://gtrd.biouml.org/"]},{"gs_name":["NOTCH3_TARGET_GENES"],"gs_url":["http://gtrd.biouml.org/"]},{"gs_name":["WP_BENZOAPYRENE_METABOLISM"],"gs_url":["http://www.wikipathways.org/instance/WP696_r106311"]}],"Description":["Genes containing one or more binding sites for UniProt:Q8WXX7 (AUTS2) in their promoter regions (TSS -1000,+100 bp) as identified by GTRD version 20.06 ChIP-seq harmonization.","Genes containing one or more binding sites for UniProt:Q8N488 (RYBP) in their promoter regions (TSS -1000,+100 bp) as identified by GTRD version 20.06 ChIP-seq harmonization.","Genes containing one or more binding sites for UniProt:Q9UM47 (NOTCH3) in their promoter regions (TSS -1000,+100 bp) as identified by GTRD version 20.06 ChIP-seq harmonization.","Benzo(a)pyrene metabolism"],"%":[0.0949227373068433,0.0495156081808396,0.0548672566371681,0.444444444444444],"numDEInCat":[43,92,62,4],"numInCat":[453,1858,1130,9],"p":[9.65148668666196e-11,1.08808219510523e-05,2.24285236261417e-05,0.000108236489810404],"FDR":[1.76815236099647e-07,0.00996683290716393,0.0136963517610305,0.0495723123331649],"Mapped Genes":["ARHGAP12; ARHGAP21; ASCC1; ATE1; ATE1-AS1; BMI1; CACUL1; COMMD3; CTBP2; DDIT4; DDX21; DNAJB12; EXOSC1; FAM241B; GATA3-AS1; HECTD2; HERC4; HNRNPF; IDE; KIFBP; LCOR; MASTL; MIR1915; MLLT10; MRPL43; MXI1; NRP1; OAT; ODAD2; PARD3; PARD3-AS1; PFKP; PIP4K2A; POLL; STN1; SUV39H2; SVIL; TWNK; ZDHHC16; ZFAND4; ZMIZ1; ZNF503; ZNF503-AS2","ACTR1A; ADK; ADRB1; ANK3; ANK3-DT; ARHGAP22; ARMC3; ASCC1; ATAD1; ATE1-AS1; BMI1; BMPR1A; BMS1; C10orf88; CALY; CAMK1D; CC2D2B; CCDC3; CFL1P1; CISD1; CXCL12; CYP26A1; DCLRE1A; DCLRE1C; DIP2C; DNAJC9-AS1; EDRF1; EDRF1-DT; EIF3A; ENKUR; ENTPD1-AS1; ERCC6; FBXL15; FRA10AC1; GATA3; GATA3-AS1; GBF1; HELLS; HIF1AN; HMX2; IKZF5; IPMK; JMJD1C; KIF20B; LBX1-AS1; LIPA; LRMDA; LRRC20; MARVELD1; MASTL; MKX; MRPS16; NCOA4; NEBL-AS1; NFKB2; NHLRC2; NKX2-3; NOC3L; NOLC1; NRP1; OPTN; PDZD7; PFKP; PHYHIPL; PIP4K2A; PRKG1; PSD; PTEN; RAB18; RET; SCD; SFMBT2; SGMS1; SKIDA1; SLC29A3; SLC35G1; SORBS1; SPAG6; STOX1; SUFU; SUV39H2; TASOR2; TCTN3; TLL2; TMEM72-AS1; UPF2; VDAC2; WAC; WAC-AS1; YME1L1; ZNF248; ZNF503-AS1","ABLIM1; ADIRF; ADIRF-AS1; ARL5B; ASCC1; BAG3; BLOC1S2; BORCS7; BORCS7-ASMT; CACNB2; CASP7; CDC123; CDNF; CNNM1; CTBP2; DDX50; DUSP5; ENTPD1-AS1; FAM107B; FRAT2; FUOM; GNG10P1; GRID1; GSTO2; HNRNPF; HSPA14; KAT6B; KIAA1217; KIF20B; KIFBP; KLLN; LINC00705; LINC00865; LINC02621; LIPA; MAP3K8; MSMB; MTPAP; NSUN6; NUDT5; OLMALINC; PCBD1; PDSS1; PIK3AP1; PTEN; PTER; RBM17; RHOBTB1; SELENOOLP; SORBS1; STAM; STAM-DT; STN1; TAF3; TIMM23; TNKS2-AS1; WAC; WAC-AS1; WDFY4; ZFAND4; ZNF32; ZNF365","AKR1C1; AKR1C2; AKR1C3; AKR1C4"]},"columns":[{"accessor":"Gene Set","name":"Gene Set","type":"list","cell":[{"name":"a","attribs":{"href":"http://gtrd.biouml.org/","target":"_blank"},"children":["AUTS2 TARGET GENES"]},{"name":"a","attribs":{"href":"http://gtrd.biouml.org/","target":"_blank"},"children":["RYBP TARGET GENES"]},{"name":"a","attribs":{"href":"http://gtrd.biouml.org/","target":"_blank"},"children":["NOTCH3 TARGET GENES"]},{"name":"a","attribs":{"href":"http://www.wikipathways.org/instance/WP696_r106311","target":"_blank"},"children":["WP BENZOAPYRENE METABOLISM"]}],"html":true,"minWidth":150},{"accessor":"Description","name":"Description","type":"character","cell":[{"name":"span","attribs":{"title":"Genes containing one or more binding sites for UniProt:Q8WXX7 (AUTS2) in their promoter regions (TSS -1000,+100 bp) as identified by GTRD version 20.06 ChIP-seq harmonization."},"children":["Genes containing one or more binding sites for UniProt:Q8WXX7 (AUTS2) in their promoter regions (..."]},{"name":"span","attribs":{"title":"Genes containing one or more binding sites for UniProt:Q8N488 (RYBP) in their promoter regions (TSS -1000,+100 bp) as identified by GTRD version 20.06 ChIP-seq harmonization."},"children":["Genes containing one or more binding sites for UniProt:Q8N488 (RYBP) in their promoter regions (T..."]},{"name":"span","attribs":{"title":"Genes containing one or more binding sites for UniProt:Q9UM47 (NOTCH3) in their promoter regions (TSS -1000,+100 bp) as identified by GTRD version 20.06 ChIP-seq harmonization."},"children":["Genes containing one or more binding sites for UniProt:Q9UM47 (NOTCH3) in their promoter regions ..."]},{"name":"span","attribs":{"title":"Benzo(a)pyrene metabolism"},"children":["Benzo(a)pyrene metabolism"]}],"minWidth":150},{"accessor":"%","name":"% Mapped","type":"numeric","cell":["9.5%","5.0%","5.5%","44.4%"],"maxWidth":80},{"accessor":"numDEInCat","name":"Total Mapped","type":"numeric","maxWidth":80},{"accessor":"numInCat","name":"Gene Set Size","type":"numeric","cell":["453","1,858","1,130","9"],"maxWidth":80},{"accessor":"p","name":"p","type":"numeric","cell":["9.65e-11","1.09e-05","2.24e-05","1.08e-04"],"maxWidth":80},{"accessor":"FDR","name":"FDR","type":"numeric","cell":["1.77e-07","0.010","0.014","0.050"],"maxWidth":80},{"accessor":"Mapped Genes","name":"Mapped Genes","type":"character","cell":[{"name":"span","attribs":{"title":"ARHGAP12; ARHGAP21; ASCC1; ATE1; ATE1-AS1; BMI1; CACUL1; COMMD3; CTBP2; DDIT4; DDX21; DNAJB12; EXOSC1; FAM241B; GATA3-AS1; HECTD2; HERC4; HNRNPF; IDE; KIFBP; LCOR; MASTL; MIR1915; MLLT10; MRPL43; MXI1; NRP1; OAT; ODAD2; PARD3; PARD3-AS1; PFKP; PIP4K2A; POLL; STN1; SUV39H2; SVIL; TWNK; ZDHHC16; ZFAND4; ZMIZ1; ZNF503; ZNF503-AS2"},"children":["ARHGAP12; ARHGAP21; ASCC1; ATE1; ATE1-AS1; BMI1; CACUL1; COMMD3; CTBP2; DDIT4; DDX21; DNAJB12; EX..."]},{"name":"span","attribs":{"title":"ACTR1A; ADK; ADRB1; ANK3; ANK3-DT; ARHGAP22; ARMC3; ASCC1; ATAD1; ATE1-AS1; BMI1; BMPR1A; BMS1; C10orf88; CALY; CAMK1D; CC2D2B; CCDC3; CFL1P1; CISD1; CXCL12; CYP26A1; DCLRE1A; DCLRE1C; DIP2C; DNAJC9-AS1; EDRF1; EDRF1-DT; EIF3A; ENKUR; ENTPD1-AS1; ERCC6; FBXL15; FRA10AC1; GATA3; GATA3-AS1; GBF1; HELLS; HIF1AN; HMX2; IKZF5; IPMK; JMJD1C; KIF20B; LBX1-AS1; LIPA; LRMDA; LRRC20; MARVELD1; MASTL; MKX; MRPS16; NCOA4; NEBL-AS1; NFKB2; NHLRC2; NKX2-3; NOC3L; NOLC1; NRP1; OPTN; PDZD7; PFKP; PHYHIPL; PIP4K2A; PRKG1; PSD; PTEN; RAB18; RET; SCD; SFMBT2; SGMS1; SKIDA1; SLC29A3; SLC35G1; SORBS1; SPAG6; STOX1; SUFU; SUV39H2; TASOR2; TCTN3; TLL2; TMEM72-AS1; UPF2; VDAC2; WAC; WAC-AS1; YME1L1; ZNF248; ZNF503-AS1"},"children":["ACTR1A; ADK; ADRB1; ANK3; ANK3-DT; ARHGAP22; ARMC3; ASCC1; ATAD1; ATE1-AS1; BMI1; BMPR1A; BMS1; C..."]},{"name":"span","attribs":{"title":"ABLIM1; ADIRF; ADIRF-AS1; ARL5B; ASCC1; BAG3; BLOC1S2; BORCS7; BORCS7-ASMT; CACNB2; CASP7; CDC123; CDNF; CNNM1; CTBP2; DDX50; DUSP5; ENTPD1-AS1; FAM107B; FRAT2; FUOM; GNG10P1; GRID1; GSTO2; HNRNPF; HSPA14; KAT6B; KIAA1217; KIF20B; KIFBP; KLLN; LINC00705; LINC00865; LINC02621; LIPA; MAP3K8; MSMB; MTPAP; NSUN6; NUDT5; OLMALINC; PCBD1; PDSS1; PIK3AP1; PTEN; PTER; RBM17; RHOBTB1; SELENOOLP; SORBS1; STAM; STAM-DT; STN1; TAF3; TIMM23; TNKS2-AS1; WAC; WAC-AS1; WDFY4; ZFAND4; ZNF32; ZNF365"},"children":["ABLIM1; ADIRF; ADIRF-AS1; ARL5B; ASCC1; BAG3; BLOC1S2; BORCS7; BORCS7-ASMT; CACNB2; CASP7; CDC123..."]},{"name":"span","attribs":{"title":"AKR1C1; AKR1C2; AKR1C3; AKR1C4"},"children":["AKR1C1; AKR1C2; AKR1C3; AKR1C4"]}],"minWidth":150}],"filterable":true,"defaultPageSize":10,"paginationType":"numbers","showPageInfo":true,"minRows":1,"dataKey":"4ead554ce7b9b2b191fba8dc2de6708b","key":"4ead554ce7b9b2b191fba8dc2de6708b"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</div>
<div id="windows-with-differentially-bound-er" class="section level2">
<h2>Windows With Differentially Bound ER</h2>
<pre class="r"><code>pwf_diff &lt;- genes_gr %&gt;% 
  mutate(w = width(.)) %&gt;% 
  as_tibble() %&gt;% 
  dplyr::select(gene_id, w) %&gt;% 
  dplyr::filter(gene_id %in% unlist(merged_results$gene_id)) %&gt;% 
  arrange(desc(w)) %&gt;% 
  distinct(gene_id, .keep_all = TRUE) %&gt;% 
  mutate(
    diff = gene_id %in% unlist(
      subset(merged_results, status != &quot;Unchanged&quot;)$gene_id
    )
  ) %&gt;% 
  with(
    nullp(
      structure(diff, names = gene_id),
      genome = config$genome$build,
      bias.data = log10(w),
      plot.fit = FALSE
    )
  )
goseq_diff &lt;- goseq(
  pwf_diff, config$genome$build, gene2cat = gs_by_geneid
) %&gt;% 
  dplyr::rename(gs_name = category, pval = over_represented_pvalue) %&gt;% 
  dplyr::select(-starts_with(&quot;under&quot;)) %&gt;% 
  as_tibble() %&gt;% 
  arrange(pval) %&gt;% 
  dplyr::filter(numDEInCat &gt; 0) %&gt;% 
  mutate(fdr = p.adjust(pval, &quot;fdr&quot;))</code></pre>
<p>No enrichment was found amongst genes mapped to sites showing any differential binding.</p>
</div>
<div id="windows-with-increased-er-binding" class="section level2">
<h2>Windows With Increased ER Binding</h2>
<pre class="r"><code>pwf_up &lt;- genes_gr %&gt;% 
  mutate(w = width(.)) %&gt;% 
  as_tibble() %&gt;% 
  dplyr::select(gene_id, w) %&gt;% 
  dplyr::filter(gene_id %in% unlist(merged_results$gene_id)) %&gt;% 
  arrange(desc(w)) %&gt;% 
  distinct(gene_id, .keep_all = TRUE) %&gt;% 
  mutate(
    diff = gene_id %in% unlist(
      subset(merged_results, status == &quot;Up&quot;)$gene_id
    )
  ) %&gt;% 
  with(
    nullp(
      structure(diff, names = gene_id),
      genome = config$genome$build,
      bias.data = log10(w),
      plot.fit = FALSE
    )
  )
goseq_up &lt;- goseq(
  pwf_up, config$genome$build, gene2cat = gs_by_geneid
) %&gt;% 
  dplyr::rename(gs_name = category, pval = over_represented_pvalue) %&gt;% 
  dplyr::select(-starts_with(&quot;under&quot;)) %&gt;% 
  as_tibble() %&gt;% 
  arrange(pval) %&gt;% 
  dplyr::filter(numDEInCat &gt; 0) %&gt;% 
  mutate(fdr = p.adjust(pval, &quot;fdr&quot;))</code></pre>
<p>No enrichment was found amongst genes mapped to sites with increased binding</p>
</div>
<div id="windows-with-decreased-er-binding" class="section level2">
<h2>Windows With Decreased ER Binding</h2>
<pre class="r"><code>pwf_down &lt;- genes_gr %&gt;% 
  mutate(w = width(.)) %&gt;% 
  as_tibble() %&gt;% 
  dplyr::select(gene_id, w) %&gt;% 
  dplyr::filter(gene_id %in% unlist(merged_results$gene_id)) %&gt;% 
  arrange(desc(w)) %&gt;% 
  distinct(gene_id, .keep_all = TRUE) %&gt;% 
  mutate(
    diff = gene_id %in% unlist(
      subset(merged_results, status == &quot;Down&quot;)$gene_id
    )
  ) %&gt;% 
  with(
    nullp(
      structure(diff, names = gene_id),
      genome = config$genome$build,
      bias.data = log10(w),
      plot.fit = FALSE
    )
  )
goseq_down &lt;- goseq(
  pwf_down, config$genome$build, gene2cat = gs_by_geneid
) %&gt;% 
  dplyr::rename(gs_name = category, pval = over_represented_pvalue) %&gt;% 
  dplyr::select(-starts_with(&quot;under&quot;)) %&gt;% 
  as_tibble() %&gt;% 
  arrange(pval) %&gt;% 
  dplyr::filter(numDEInCat &gt; 0) %&gt;% 
  mutate(fdr = p.adjust(pval, &quot;fdr&quot;))</code></pre>
<p>No enrichment was found amongst genes mapped to sites with decreased binding</p>
</div>
</div>
<div id="comparison-with-rna-seq" class="section level1">
<h1>Comparison With RNA Seq</h1>
<pre class="r"><code>library(fgsea)
rna_status &lt;- rnaseq %&gt;% 
  mutate(
    de_status = case_when(
      !!sym(rna_fdr_col) &lt; fdr_alpha &amp; !!sym(rna_lfc_col) &lt; 0 ~ &quot;Down&quot;,
      !!sym(rna_fdr_col) &lt; fdr_alpha &amp; !!sym(rna_lfc_col) &gt; 0 ~ &quot;Up&quot;,
      !!sym(rna_fdr_col) &gt;= fdr_alpha ~ &quot;Unchanged&quot;
    )
  ) %&gt;% 
  dplyr::select(gene_id, gene_name, de_status) %&gt;% 
  bind_rows(
    genes_gr %&gt;% 
      subset(!gene_id %in% rnaseq$gene_id) %&gt;% 
      select(gene_id, gene_name) %&gt;% 
      mcols() %&gt;% 
      as_tibble() %&gt;% 
      mutate(de_status = &quot;Undetected&quot;)
  ) %&gt;% 
  mutate(
    de_status = factor(
      de_status, levels = str_to_title(names(colours$direction))
    )
  ) %&gt;% 
  arrange(gene_id)
rna_p_col &lt;- colnames(rnaseq)[
  str_detect(str_to_lower(colnames(rnaseq)), &quot;pval|p\\.val&quot;)
][1]
stopifnot(length(rna_p_col) == 1)
diff_windows &lt;- merged_results %&gt;% 
  splitAsList(.$status) %&gt;% 
  lapply(select, gene_id) %&gt;% 
  lapply(mcols) %&gt;% 
  lapply(as_tibble) %&gt;% 
  lapply(pull, &quot;gene_id&quot;) %&gt;% 
  lapply(unlist) %&gt;% 
  lapply(unique)</code></pre>
<div id="relationship-to-detected-genes" class="section level2 tabset">
<h2 class="tabset">Relationship To Detected Genes</h2>
<div id="detected-genes-by-er-binding-region" class="section level3">
<h3>Detected Genes By ER Binding Region</h3>
<p>Firstly, the distribution of merged windows, where ER was considered as present, was compared to genes within the RNA-Seq dataset. Genes were simply considered as detected if present in the RNA-Seq data, or undetected if not present. Gene-centric regions were as defined previously.</p>
<pre class="r"><code>w &lt;- 0.8
min_p &lt;- 0.05
merged_results %&gt;% 
  mutate(
    any_detected = ifelse(
      vapply(detected, length, integer(1)) &gt; 1,
      &quot;Mapped to Detected Genes&quot;,
      &quot;Mapped to Undetected Genes&quot;
    )
  ) %&gt;% 
  plotPie(fill = &quot;region&quot;, x = &quot;any_detected&quot;, label_size = 4, width = w) +
  geom_text(
    aes(xlab, ylab, label =lab),
    data = . %&gt;% 
      mutate(
        xlab = x + w*(r + 0.1)*sin(label_radians),
        ylab = 1 + w*(r + 0.1)*cos(label_radians),
        lab = glue(&quot;{region}\n{percent(p, 0.1)}&quot;) %&gt;% 
          str_wrap(10)
      ) %&gt;% 
      dplyr::filter(p * N &gt; min_p * max(N)),
    size = 3
  ) +
  geom_segment(
    aes(x_inner, y_inner, xend = x_outer, yend = y_outer),
     data = . %&gt;% 
      mutate(
        x_inner = x + w*r*sin(label_radians),
        y_inner = 1 + w*r*cos(label_radians),
        x_outer = x + w*(r + 0.01)*sin(label_radians),
        y_outer = 1 + w*(r + 0.01)*cos(label_radians),
      ) %&gt;% 
      dplyr::filter(p * N &gt; min_p * max(N))
  ) +
  scale_fill_manual(
    values = colours$regions %&gt;% 
      setNames(regions[names(.)]) 
  ) +
  labs(x = &quot;&quot;, fill = &quot;Region&quot;) +
  theme(panel.grid = element_blank())</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-region-detected-1.png" alt="*Distribution of ER-bound windows by region, according to whether the window is mapped to a detected gene in the RNA-Seq dataset.*" width="960" />
<p class="caption">
<em>Distribution of ER-bound windows by region, according to whether the window is mapped to a detected gene in the RNA-Seq dataset.</em>
</p>
</div>
</div>
<div id="detected-genes-by-feature" class="section level3">
<h3>Detected Genes By Feature</h3>
<pre class="r"><code>merged_results %&gt;% 
  mutate(
    any_detected = ifelse(
      vapply(detected, length, integer(1)) &gt; 1,
      &quot;Mapped to Detected Genes&quot;,
      &quot;Mapped to Undetected Genes&quot;
    )
  ) %&gt;% 
  plotPie(fill = &quot;feature&quot;, x = &quot;any_detected&quot;, label_size = 4, width = w) +
  geom_text(
    aes(xlab, ylab, label =lab),
    data = . %&gt;% 
      mutate(
        xlab = x + w*(r + 0.1)*sin(label_radians),
        ylab = 1 + w*(r + 0.1)*cos(label_radians),
        lab = glue(&quot;{feature}\n{percent(p, 0.1)}&quot;) %&gt;% 
          str_wrap(10)
      ) %&gt;% 
      dplyr::filter(p * N &gt; min_p * max(N)),
    size = 3
  ) +
  geom_segment(
    aes(x_inner, y_inner, xend = x_outer, yend = y_outer),
     data = . %&gt;% 
      mutate(
        x_inner = x + w*r*sin(label_radians),
        y_inner = 1 + w*r*cos(label_radians),
        x_outer = x + w*(r + 0.01)*sin(label_radians),
        y_outer = 1 + w*(r + 0.01)*cos(label_radians),
      ) %&gt;% 
      dplyr::filter(p * N &gt; min_p * max(N))
  ) +
  scale_fill_manual(
    values = colours$features %&gt;% 
      setNames(str_sep_to_title(names(.)))
  ) +
  labs(x = &quot;&quot;, fill = &quot;External Feature&quot;) +
  theme(panel.grid = element_blank())</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-feature-detected-1.png" alt="*Distribution of ER-bound windows by externally-defined features from h3k27ac_features.gtf, according to whether the window is mapped to a detected gene in the RNA-Seq dataset.*" width="960" />
<p class="caption">
<em>Distribution of ER-bound windows by externally-defined features from h3k27ac_features.gtf, according to whether the window is mapped to a detected gene in the RNA-Seq dataset.</em>
</p>
</div>
</div>
</div>
<div id="relationship-to-differentially-expressed-genes" class="section level2 tabset">
<h2 class="tabset">Relationship To Differentially Expressed Genes</h2>
<pre class="r"><code>rna_dir_ranks &lt;- rnaseq %&gt;% 
  mutate(
    ranking_stat = -sign(!!sym(rna_lfc_col))*log10(!!sym(rna_p_col)) %&gt;% 
      setNames(gene_id)
  ) %&gt;% 
  arrange(ranking_stat) %&gt;% 
  dplyr::filter(!!sym(rna_p_col) &lt; 1) %&gt;% 
  pull(&quot;ranking_stat&quot;) 
rna_sig_ranks &lt;- rna_dir_ranks %&gt;% 
  abs() %&gt;% 
  sort()
dir_gs &lt;- merged_results %&gt;% 
  select(status, starts_with(&quot;gene&quot;)) %&gt;% 
  as_tibble() %&gt;% 
  mutate(group = glue(&quot;{target} {status}&quot;)) %&gt;% 
  split(.$group)%&gt;% 
  lapply(pull, &quot;gene_id&quot;) %&gt;% 
  lapply(unlist) %&gt;% 
  lapply(unique) %&gt;% 
  .[vapply(., length, integer(1)) &gt; 20]
reg_dir_gs &lt;- merged_results %&gt;% 
  select(
    overlaps_ref, status, any_of(c(&quot;region&quot;, &quot;feature&quot;)), starts_with(&quot;gene&quot;)
  ) %&gt;% 
  as_tibble() %&gt;% 
  mutate(group = glue(&quot;{region} / {target} {status}&quot;)) %&gt;% 
  split(.$group)%&gt;% 
  lapply(pull, &quot;gene_id&quot;) %&gt;% 
  lapply(unlist) %&gt;% 
  lapply(unique) %&gt;% 
  .[vapply(., length, integer(1)) &gt; 20]
if (has_features) {
  feat_dir_gs &lt;- merged_results %&gt;% 
  select(
    overlaps_ref, status, any_of(c(&quot;feature&quot;)), starts_with(&quot;gene&quot;)
  ) %&gt;% 
  as_tibble() %&gt;% 
  mutate(group = glue(&quot;{feature} / {target} {status}&quot;)) %&gt;% 
  split(.$group)%&gt;% 
  lapply(pull, &quot;gene_id&quot;) %&gt;% 
  lapply(unlist) %&gt;% 
  lapply(unique) %&gt;% 
  .[vapply(., length, integer(1)) &gt; 20]
}</code></pre>
<div id="p-values-by-er-status" class="section level3">
<h3>P-Values By ER Status</h3>
<pre class="r"><code>y &lt;- ggplot_build(
  tibble(p = rnaseq[[rna_p_col]]) %&gt;% 
    ggplot(aes(p, stat(density))) + 
    geom_histogram(bins = 100) 
) %&gt;% .[[&quot;data&quot;]] %&gt;% 
  .[[1]] %&gt;% 
  .[[&quot;y&quot;]] %&gt;% 
  max()
rnaseq %&gt;% 
  mutate(
    mapped = case_when(
      gene_id %in% unlist(diff_windows[names(diff_windows) %in% c(&quot;Up&quot;, &quot;Down&quot;)]) ~ paste(target, &quot;Differentially Bound&quot;),
      gene_id %in% unlist(diff_windows[names(diff_windows) %in% c(&quot;Unchanged&quot;)]) ~ paste(target, &quot;Unchanged&quot;),
      TRUE ~ paste(&quot;No&quot;, target, &quot;Detected&quot;)
    )
  ) %&gt;% 
  ggplot(
    aes(!!sym(rna_p_col), stat(density))
  ) +
  geom_histogram(
    colour = &quot;black&quot;, fill = &quot;grey&quot;, bins = 100
  ) +
  geom_text(
    aes(0.5, y, label = lab),
    data = . %&gt;% 
      group_by(mapped) %&gt;% 
      tally() %&gt;% 
      mutate(lab = glue(&quot;N = {comma(n, 1)}&quot;))
  ) +
  facet_wrap(~mapped, nrow = 1) +
  scale_y_continuous(expand = expansion(c(0, 0.05)))</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-p-target-1.png" alt="*P-Values for differential expression partitioned by ER ChIP peak status. Genes were considered as being mapped to a differentially bound peak if one or more peaks was considered as differentially bound.*" width="960" />
<p class="caption">
<em>P-Values for differential expression partitioned by ER ChIP peak status. Genes were considered as being mapped to a differentially bound peak if one or more peaks was considered as differentially bound.</em>
</p>
</div>
</div>
</div>
<div id="directional-gsea" class="section level2 tabset">
<h2 class="tabset">Directional GSEA</h2>
<div id="all-windows" class="section level3">
<h3>All Windows</h3>
<pre class="r"><code>gsea_dir &lt;- fgsea(dir_gs, rna_dir_ranks)
gsea_dir_sig &lt;- gsea_dir %&gt;% 
  arrange(pval) %&gt;% 
  dplyr::filter(padj &lt; fdr_alpha) %&gt;% 
  as_tibble() 
p &lt;- gsea_dir_sig %&gt;% 
  dplyr::slice(1:9) %&gt;% 
  pull(&quot;pathway&quot;) %&gt;% 
  lapply(
    function(x) {
      plotEnrichment(dir_gs[[x]], rna_dir_ranks) +
        ggtitle(x) +
        theme(
          plot.title = element_text(hjust = 0.5, size = 10)
        )
    }
  )</code></pre>
<p>No association was found between ER binding and <em>directional</em> significance in differentially expressed genes.</p>
</div>
<div id="windows-by-region" class="section level3">
<h3>Windows By Region</h3>
<pre class="r"><code>gsea_reg_dir &lt;- fgsea(reg_dir_gs, rna_dir_ranks)
gsea_reg_dir_sig &lt;- gsea_reg_dir %&gt;% 
  arrange(pval) %&gt;% 
  dplyr::filter(padj &lt; fdr_alpha) %&gt;% 
  as_tibble() 
p &lt;- gsea_reg_dir_sig %&gt;% 
  dplyr::slice(1:9) %&gt;% 
  pull(&quot;pathway&quot;) %&gt;% 
  lapply(
    function(x) {
      plotEnrichment(reg_dir_gs[[x]], rna_dir_ranks) +
        ggtitle(x) +
        theme(
          plot.title = element_text(hjust = 0.5, size = 10)
        )
    }
  )</code></pre>
<p>No association was found between ER binding and <em>directional</em> significance in differentially expressed genes, when incorporating genomic region.</p>
</div>
<div id="windows-by-feature" class="section level3">
<h3>Windows By Feature</h3>
<pre class="r"><code>gsea_feat_dir &lt;- fgsea(feat_dir_gs, rna_dir_ranks)
gsea_feat_dir_sig &lt;- gsea_feat_dir %&gt;% 
  arrange(pval) %&gt;% 
  dplyr::filter(padj &lt; fdr_alpha) %&gt;% 
  as_tibble() 
p &lt;- gsea_feat_dir_sig %&gt;% 
  dplyr::slice(1:9) %&gt;% 
  pull(&quot;pathway&quot;) %&gt;% 
  lapply(
    function(x) {
      plotEnrichment(feat_dir_gs[[x]], rna_dir_ranks) +
        ggtitle(x) +
        theme(
          plot.title = element_text(hjust = 0.5, size = 10)
        )
    }
  )</code></pre>
<pre class="r"><code>cp &lt;-  htmltools::em(
  glue(
    &quot;Merged windows were mapped to genes and their position amongst the &quot;,
    &quot;RNA-Seq results was assessed. Windows were classified based on which &quot;,
    &quot;feature was the best overlap, and the direction of {target} binding with &quot;,
    &quot;{treat_levels[[2]]} treatment. {nrow(gsea_feat_dir_sig)} sets of windows were &quot;,
    &quot;associated with changes in gene expression, using the sign of &quot;, 
    &quot;fold-change and ranking statistic to initially rank the &quot;, 
    &quot;{comma(nrow(rnaseq), 1)} genes considered as detected.&quot;
  )
)
tbl &lt;- gsea_feat_dir_sig %&gt;%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %&gt;% 
      vapply(paste, character(1), collapse = &quot;; &quot;),
    Direction = ifelse(NES &gt; 0, &quot;\u21E7 Up-regulated&quot;, &quot;\u21E9 Down-regulated&quot;),
    &quot;{target}&quot; := str_extract(pathway, &quot;(Up|Down|Unchanged)$&quot;),
    pathway = str_remove_all(pathway, &quot; / [^/]+$&quot;)
  ) %&gt;% 
  dplyr::select(
    pathway, !!sym(target), Windows = size, Direction, 
    p = pval, FDR = padj, `Edge Size`, `Leading Edge` = leadingEdge
  ) %&gt;% 
  reactable(
    filterable = TRUE,
    columns = list2(
      pathway = colDef(
        minWidth = 100, name = &quot;Region&quot;
      ),
      &quot;{target}&quot; := colDef(
        maxWidth = 90,
        cell = function(value) {
          html_symbol &lt;- &quot;&quot;
          if (str_detect(value, &quot;Up&quot;)) html_symbol &lt;- &quot;\u21E7&quot;
          if (str_detect(value, &quot;Down&quot;)) html_symbol &lt;- &quot;\u21E9&quot;
          paste(html_symbol, value)
        },
        style = function(value) {
          colour &lt;- case_when(
            str_detect(value, &quot;Up&quot;) ~ colours$direction[[&quot;up&quot;]],
            str_detect(value, &quot;Down&quot;) ~ colours$direction[[&quot;down&quot;]],
            TRUE ~ colours$direction[[&quot;unchanged&quot;]]
          )
          list(color = colour)
        }
      ),
      Windows = colDef(maxWidth = 80),
      Direction = colDef(
        name = &quot;Gene Direction&quot;,
        maxWidth = 120,
        style = function(value) {
          colour &lt;- ifelse(
            str_detect(value, &quot;Up&quot;), 
            colours$direction[[&quot;up&quot;]], 
            colours$direction[[&quot;down&quot;]]
          )
          list(color = colour)
        },
      ),
      p = colDef(
        cell = function(value) sprintf(&quot;%.2e&quot;, value), maxWidth = 80
      ),
      FDR = colDef(
        cell = function(value) sprintf(&quot;%.2e&quot;, value), maxWidth = 80
      ),
      &quot;Edge Size&quot; = colDef(maxWidth = 80),
     &quot;Leading Edge&quot; = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = &quot;table&quot;,
  div(class = &quot;table-header&quot;,
      div(class = &quot;caption&quot;, cp),
      tbl
  )
)</code></pre>
<div class="table">
<div class="table-header">
<div class="caption">
<em>Merged windows were mapped to genes and their position amongst the RNA-Seq results was assessed. Windows were classified based on which feature was the best overlap, and the direction of ER binding with DHT treatment. 1 sets of windows were associated with changes in gene expression, using the sign of fold-change and ranking statistic to initially rank the 15,388 genes considered as detected.</em>
</div>
<div id="htmlwidget-e6e2ec4422ac4786afeb" class="reactable html-widget" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-e6e2ec4422ac4786afeb">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"pathway":["Super Enhancer"],"ER":["Unchanged"],"Windows":[71],"Direction":["⇧ Up-regulated"],"p":[0.00562065345381048],"FDR":[0.0449652276304838],"Edge Size":[15],"Leading Edge":["KCNMA1; AKR1C3; OLMALINC; AKR1C2; SCD; AKR1C1; MMRN2; FAM107B; SNCG; DDIT4; GLUD1; ADIRF-AS1; HHEX; STN1; SLF2"]},"columns":[{"accessor":"pathway","name":"Region","type":"character","minWidth":100},{"accessor":"ER","name":"ER","type":"character","cell":[" Unchanged"],"maxWidth":90,"style":[{"color":"#7F7F7F"}]},{"accessor":"Windows","name":"Windows","type":"numeric","maxWidth":80},{"accessor":"Direction","name":"Gene Direction","type":"character","maxWidth":120,"style":[{"color":"#CA0020"}]},{"accessor":"p","name":"p","type":"numeric","cell":["5.62e-03"],"maxWidth":80},{"accessor":"FDR","name":"FDR","type":"numeric","cell":["4.50e-02"],"maxWidth":80},{"accessor":"Edge Size","name":"Edge Size","type":"numeric","maxWidth":80},{"accessor":"Leading Edge","name":"Leading Edge","type":"character","cell":[{"name":"span","attribs":{"title":"KCNMA1; AKR1C3; OLMALINC; AKR1C2; SCD; AKR1C1; MMRN2; FAM107B; SNCG; DDIT4; GLUD1; ADIRF-AS1; HHEX; STN1; SLF2"},"children":["KCNMA1; AKR1C3; OLMALINC; AKR1C2; SCD; AKR1C1; ..."]}],"minWidth":150}],"filterable":true,"defaultPageSize":10,"paginationType":"numbers","showPageInfo":true,"minRows":1,"dataKey":"3349e454ee73049af0d417c4042ed5e0","key":"3349e454ee73049af0d417c4042ed5e0"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<pre class="r"><code>plot_grid(plotlist = p, nrow = ceiling(length(p) / 3))</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/barcode-gsea-feat-dir-sig-1.png" alt="*Barcode plots for the top 1 sets of windows associated with __directional__ changes in gene expression.*" width="960" />
<p class="caption">
<em>Barcode plots for the top 1 sets of windows associated with <strong>directional</strong> changes in gene expression.</em>
</p>
</div>
</div>
</div>
<div id="non-directional-gsea" class="section level2 tabset">
<h2 class="tabset">Non-Directional GSEA</h2>
<div id="all-windows-1" class="section level3">
<h3>All Windows</h3>
<pre class="r"><code>gsea_nondir &lt;- fgsea(dir_gs, rna_sig_ranks, scoreType = &quot;pos&quot;)
gsea_nondir_sig &lt;- gsea_nondir %&gt;% 
  arrange(pval) %&gt;% 
  dplyr::filter(padj &lt; fdr_alpha) %&gt;% 
  as_tibble() 
p &lt;- gsea_nondir_sig %&gt;% 
  dplyr::slice(1:9) %&gt;% 
  pull(&quot;pathway&quot;) %&gt;% 
  lapply(
    function(x) {
      plotEnrichment(dir_gs[[x]], rna_sig_ranks) +
        ggtitle(x) +
        theme(plot.title = element_text(hjust = 0.5, size = 10))
    }
  )</code></pre>
<pre class="r"><code>cp &lt;-  htmltools::em(
  glue(
    &quot;Merged windows were mapped to genes and their position amongst the &quot;,
    &quot;RNA-Seq results was assessed. Windows were classified based on  &quot;,
    &quot;the direction of {target} binding with &quot;,
    &quot;{treat_levels[[2]]} treatment. {nrow(gsea_nondir_sig)} sets of windows were &quot;,
    &quot;associated with changes in gene expression, using only the p-value to &quot;, 
    &quot;rank the {comma(nrow(rnaseq), 1)} genes considered as detected.&quot;
  )
)
tbl &lt;- gsea_nondir_sig %&gt;%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %&gt;% 
      vapply(paste, character(1), collapse = &quot;; &quot;),
    &quot;{target}&quot; := str_extract(pathway, &quot;(Up|Down|Unchanged)$&quot;),
    pathway = str_remove_all(pathway, &quot; / [^/]+$&quot;)
  ) %&gt;% 
  dplyr::select(
    !!sym(target), Windows = size, p = pval, FDR = padj, 
    `Edge Size`, `Leading Edge` = leadingEdge
  ) %&gt;% 
  reactable(
    filterable = TRUE,
    columns = list2(
      &quot;{target}&quot; := colDef(
        maxWidth = 120,
        cell = function(value) {
          html_symbol &lt;- &quot;&quot;
          if (str_detect(value, &quot;Up&quot;)) html_symbol &lt;- &quot;\u21E7&quot;
          if (str_detect(value, &quot;Down&quot;)) html_symbol &lt;- &quot;\u21E9&quot;
          paste(html_symbol, value)
        },
        style = function(value) {
           colour &lt;- case_when(
            str_detect(value, &quot;Up&quot;) ~ colours$direction[[&quot;up&quot;]],
            str_detect(value, &quot;Down&quot;) ~ colours$direction[[&quot;down&quot;]],
            TRUE ~ colours$direction[[&quot;unchanged&quot;]]
          )
          list(color = colour)
        }
      ),
      Windows = colDef(maxWidth = 80),
      p = colDef(
        cell = function(value) sprintf(&quot;%.2e&quot;, value), maxWidth = 80
      ),
      FDR = colDef(
        cell = function(value) sprintf(&quot;%.2e&quot;, value), maxWidth = 80
      ),
      &quot;Edge Size&quot; = colDef(maxWidth = 80),
     &quot;Leading Edge&quot; = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = &quot;table&quot;,
  div(class = &quot;table-header&quot;,
      div(class = &quot;caption&quot;, cp),
      tbl
  )
)</code></pre>
<div class="table">
<div class="table-header">
<div class="caption">
<em>Merged windows were mapped to genes and their position amongst the RNA-Seq results was assessed. Windows were classified based on  the direction of ER binding with DHT treatment. 1 sets of windows were associated with changes in gene expression, using only the p-value to rank the 15,388 genes considered as detected.</em>
</div>
<div id="htmlwidget-6d2a351af8f41b1a8046" class="reactable html-widget" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-6d2a351af8f41b1a8046">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"ER":["Unchanged"],"Windows":[319],"p":[0.00420146779888697],"FDR":[0.00840293559777394],"Edge Size":[91],"Leading Edge":["ADRA2A; KCNMA1; PPIF; MYPN; ZWINT; SVIL; SFTPA2; CEP55; KIF20B; NEBL; ASB13; MASTL; CALML5; AKR1C3; KIF11; OLMALINC; ALOX5; AKR1C2; CDK1; SCD; UNC5B; AKR1C1; MMRN2; FAM107B; ANK3; ITGB1; NOLC1; NPM3; USP54; LINC00595; SNCG; HECTD2; ZMYND11; GATA3; PPRC1; ARL3; AL391421.1; DDIT4; TASOR2; RBM20; GLUD1; SVIL2P; RPS24; MAP3K8; ADIRF-AS1; USP6NL; ZNF488; MSS51; STOX1; HHEX; STN1; MALRD1; FRMPD2; SLF2; SFXN2; SLC16A12; CACNB2; SRP9P1; ANXA11; HERC4; SEMA4G; ASCC1; LHPP; PDCD11; ZMIZ1-AS1; AC022400.6; GSTO1; PDZD7; KIF5B; ADAM8; DCLRE1C; CNNM2; AKR1C4; PPP3CB-AS1; TWNK; DNAJC1; SLC29A3; PRKG1; MPP7; AFAP1L2; SYT15; SFMBT2; PSD; ABCC2; SLC25A16; RPL12P29; KAZALD1; DNAJC9; AL732437.3; HNRNPF; PRR26"]},"columns":[{"accessor":"ER","name":"ER","type":"character","cell":[" Unchanged"],"maxWidth":120,"style":[{"color":"#7F7F7F"}]},{"accessor":"Windows","name":"Windows","type":"numeric","maxWidth":80},{"accessor":"p","name":"p","type":"numeric","cell":["4.20e-03"],"maxWidth":80},{"accessor":"FDR","name":"FDR","type":"numeric","cell":["8.40e-03"],"maxWidth":80},{"accessor":"Edge Size","name":"Edge Size","type":"numeric","maxWidth":80},{"accessor":"Leading Edge","name":"Leading Edge","type":"character","cell":[{"name":"span","attribs":{"title":"ADRA2A; KCNMA1; PPIF; MYPN; ZWINT; SVIL; SFTPA2; CEP55; KIF20B; NEBL; ASB13; MASTL; CALML5; AKR1C3; KIF11; OLMALINC; ALOX5; AKR1C2; CDK1; SCD; UNC5B; AKR1C1; MMRN2; FAM107B; ANK3; ITGB1; NOLC1; NPM3; USP54; LINC00595; SNCG; HECTD2; ZMYND11; GATA3; PPRC1; ARL3; AL391421.1; DDIT4; TASOR2; RBM20; GLUD1; SVIL2P; RPS24; MAP3K8; ADIRF-AS1; USP6NL; ZNF488; MSS51; STOX1; HHEX; STN1; MALRD1; FRMPD2; SLF2; SFXN2; SLC16A12; CACNB2; SRP9P1; ANXA11; HERC4; SEMA4G; ASCC1; LHPP; PDCD11; ZMIZ1-AS1; AC022400.6; GSTO1; PDZD7; KIF5B; ADAM8; DCLRE1C; CNNM2; AKR1C4; PPP3CB-AS1; TWNK; DNAJC1; SLC29A3; PRKG1; MPP7; AFAP1L2; SYT15; SFMBT2; PSD; ABCC2; SLC25A16; RPL12P29; KAZALD1; DNAJC9; AL732437.3; HNRNPF; PRR26"},"children":["ADRA2A; KCNMA1; PPIF; MYPN; ZWINT; SVIL; SFTPA2..."]}],"minWidth":150}],"filterable":true,"defaultPageSize":10,"paginationType":"numbers","showPageInfo":true,"minRows":1,"dataKey":"caee799b83183e76aad762b28318974f","key":"caee799b83183e76aad762b28318974f"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<pre class="r"><code>plot_grid(plotlist = p, nrow = ceiling(length(p) / 3))</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/barcode-gsea-nondir-sig-1.png" alt="*Barcode plots for the top 1 sets of windows associated with __non-directional__ changes in gene expression.*" width="960" />
<p class="caption">
<em>Barcode plots for the top 1 sets of windows associated with <strong>non-directional</strong> changes in gene expression.</em>
</p>
</div>
</div>
<div id="windows-by-region-1" class="section level3">
<h3>Windows By Region</h3>
<pre class="r"><code>gsea_reg_nondir &lt;- fgsea(reg_dir_gs, rna_sig_ranks, scoreType = &quot;pos&quot;)
gsea_reg_nondir_sig &lt;- gsea_reg_nondir %&gt;% 
  arrange(pval) %&gt;% 
  dplyr::filter(padj &lt; fdr_alpha) %&gt;% 
  as_tibble() 
p &lt;- gsea_reg_nondir_sig %&gt;% 
  dplyr::slice(1:9) %&gt;% 
  pull(&quot;pathway&quot;) %&gt;% 
  lapply(
    function(x) {
      plotEnrichment(reg_dir_gs[[x]], rna_sig_ranks) +
        ggtitle(x) +
        theme(plot.title = element_text(hjust = 0.5, size = 10))
    }
  )</code></pre>
<p>No association was found between ER binding and <em>overall</em> significance in differentially expressed genes.</p>
</div>
<div id="windows-by-feature-1" class="section level3">
<h3>Windows By Feature</h3>
<pre class="r"><code>gsea_feat_nondir &lt;- fgsea(feat_dir_gs, rna_sig_ranks, scoreType = &quot;pos&quot;)
gsea_feat_nondir_sig &lt;- gsea_feat_nondir %&gt;% 
  arrange(pval) %&gt;% 
  dplyr::filter(padj &lt; fdr_alpha) %&gt;% 
  as_tibble() 
p &lt;- gsea_feat_nondir_sig %&gt;% 
  dplyr::slice(1:9) %&gt;% 
  pull(&quot;pathway&quot;) %&gt;% 
  lapply(
    function(x) {
      plotEnrichment(feat_dir_gs[[x]], rna_sig_ranks) +
        ggtitle(x) +
        theme(
          plot.title = element_text(hjust = 0.5, size = 10)
        )
    }
  )</code></pre>
<pre class="r"><code>cp &lt;-  htmltools::em(
  glue(
    &quot;Merged windows were mapped to genes and their position amongst the &quot;,
    &quot;RNA-Seq results was assessed. Windows were classified based on which &quot;,
    &quot;feature was the best overlap, and the direction of {target} binding with &quot;,
    &quot;{treat_levels[[2]]} treatment. {nrow(gsea_feat_nondir_sig)} sets of windows were &quot;,
    &quot;associated with overall changes in gene expression, using only the &quot;, 
    &quot;test statistic to rank the &quot;, 
    &quot;{comma(nrow(rnaseq), 1)} genes considered as detected.&quot;
  )
)
tbl &lt;- gsea_feat_nondir_sig %&gt;%
  mutate(
    `Edge Size` = vapply(leadingEdge, length, integer(1)),
    leadingEdge = lapply(leadingEdge, function(x) id2gene[x]) %&gt;% 
      vapply(paste, character(1), collapse = &quot;; &quot;),
    Direction = ifelse(NES &gt; 0, &quot;\u21E7 Up-regulated&quot;, &quot;\u21E9 Down-regulated&quot;),
    &quot;{target}&quot; := str_extract(pathway, &quot;(Up|Down|Unchanged)$&quot;),
    pathway = str_remove_all(pathway, &quot; / [^/]+$&quot;)
  ) %&gt;% 
  dplyr::select(
    pathway, !!sym(target), Windows = size, Direction, 
    p = pval, FDR = padj, `Edge Size`, `Leading Edge` = leadingEdge
  ) %&gt;% 
  reactable(
    filterable = TRUE,
    columns = list2(
      pathway = colDef(
        minWidth = 100, name = &quot;Region&quot;
      ),
      &quot;{target}&quot; := colDef(
        maxWidth = 90,
        cell = function(value) {
          html_symbol &lt;- &quot;&quot;
          if (str_detect(value, &quot;Up&quot;)) html_symbol &lt;- &quot;\u21E7&quot;
          if (str_detect(value, &quot;Down&quot;)) html_symbol &lt;- &quot;\u21E9&quot;
          paste(html_symbol, value)
        },
        style = function(value) {
          colour &lt;- case_when(
            str_detect(value, &quot;Up&quot;) ~ colours$direction[[&quot;up&quot;]],
            str_detect(value, &quot;Down&quot;) ~ colours$direction[[&quot;down&quot;]],
            TRUE ~ colours$direction[[&quot;unchanged&quot;]]
          )
          list(color = colour)
        }
      ),
      Windows = colDef(maxWidth = 80),
      Direction = colDef(
        name = &quot;Gene Direction&quot;,
        maxWidth = 120,
        style = function(value) {
          colour &lt;- ifelse(
            str_detect(value, &quot;Up&quot;), 
            colours$direction[[&quot;up&quot;]], 
            colours$direction[[&quot;down&quot;]]
          )
          list(color = colour)
        },
      ),
      p = colDef(
        cell = function(value) sprintf(&quot;%.2e&quot;, value), maxWidth = 80
      ),
      FDR = colDef(
        cell = function(value) sprintf(&quot;%.2e&quot;, value), maxWidth = 80
      ),
      &quot;Edge Size&quot; = colDef(maxWidth = 80),
     &quot;Leading Edge&quot; = colDef(
       minWidth = 150,
       cell = function(value) with_tooltip(value, width = 50)
     ) 
    )
  )
div(class = &quot;table&quot;,
  div(class = &quot;table-header&quot;,
      div(class = &quot;caption&quot;, cp),
      tbl
  )
)</code></pre>
<div class="table">
<div class="table-header">
<div class="caption">
<em>Merged windows were mapped to genes and their position amongst the RNA-Seq results was assessed. Windows were classified based on which feature was the best overlap, and the direction of ER binding with DHT treatment. 4 sets of windows were associated with overall changes in gene expression, using only the test statistic to rank the 15,388 genes considered as detected.</em>
</div>
<div id="htmlwidget-704564cd723715c7f8fc" class="reactable html-widget" style="width:auto;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-704564cd723715c7f8fc">{"x":{"tag":{"name":"Reactable","attribs":{"data":{"pathway":["Enhancer","Super Enhancer","Super Enhancer","Promoter"],"ER":["Unchanged","Down","Unchanged","Down"],"Windows":[177,169,71,270],"Direction":["⇧ Up-regulated","⇧ Up-regulated","⇧ Up-regulated","⇧ Up-regulated"],"p":[0.000636542635145292,0.00390402681091231,0.00561431853612597,0.0133764707121745],"FDR":[0.00509234108116234,0.0149715160963359,0.0149715160963359,0.026752941424349],"Edge Size":[56,63,22,85],"Leading Edge":["ADRA2A; KCNMA1; PPIF; MYPN; CEP55; NEBL; ASB13; AKR1C3; OLMALINC; CDK1; SCD; UNC5B; MMRN2; FAM107B; ANK3; ITGB1; NPM3; USP54; LINC00595; SNCG; HECTD2; ZMYND11; GATA3; ARL3; DDIT4; TASOR2; GLUD1; SVIL2P; MAP3K8; ADIRF-AS1; USP6NL; STOX1; STN1; SLF2; SFXN2; SLC16A12; CACNB2; HERC4; SEMA4G; ASCC1; ZMIZ1-AS1; PDZD7; KIF5B; ADAM8; CNNM2; AKR1C4; PPP3CB-AS1; TWNK; SLC29A3; PRKG1; AFAP1L2; SFMBT2; RPL12P29; KAZALD1; DNAJC9; AL732437.3","KCNMA1; PPIF; HELLS; SVIL; CEP55; ASB13; RHOBTB1; RET; OLMALINC; CDK1; LINC02635; SCD; UNC5B; MMRN2; FAM107B; RASGEF1A; ITGB1; NOLC1; LINC00595; SNCG; HECTD2; ZMYND11; PPRC1; ARL3; AL391421.1; DDIT4; TASOR2; GLUD1; SVIL2P; MAP3K8; ADIRF-AS1; USP6NL; STN1; IDI1; NOC3L; SLF2; SFXN2; GTPBP4; SRP9P1; ASCC1; CPXM2; NUTM2A; ZMIZ1-AS1; IDI2-AS1; DCLRE1C; PLCE1; SLC29A3; SUV39H2; MPP7; SYT15; SFMBT2; PSD; DNAJC9; AL732437.3; HNRNPF; PRR26; RPS3AP5; AL356056.2; WDR37; ST8SIA6; ZNF438; BLOC1S2; DDX21","KCNMA1; CEP55; AKR1C3; KIF11; OLMALINC; AKR1C2; SCD; AKR1C1; MMRN2; FAM107B; SNCG; HECTD2; DDIT4; GLUD1; ADIRF-AS1; MSS51; HHEX; STN1; SLF2; CACNB2; SRP9P1; ASCC1","PPIF; SYNPO2L; SVIL; CEP55; NEBL; ASB13; RHOBTB1; CALML5; RET; OLMALINC; ALOX5; SCD; ARHGAP19; MMRN2; FAM107B; ITGB1; DHTKD1; NOLC1; USP54; LINC00595; SNCG; ZMYND11; GATA3; PPRC1; ARL3; AL391421.1; DDIT4; TASOR2; RBM20; GLUD1; PLEKHS1; CASP7; MAP3K8; ADIRF-AS1; USP6NL; ZNF488; DEPP1; STOX1; HHEX; STN1; BAG3; TMEM26; FRMPD2; IDI1; SLF2; SFXN2; BAMBI; P4HA1; CACNB2; GTPBP4; ZNF239; ANXA11; HERC4; FAM241B; DNAJC12; ZMIZ1-AS1; GSTO1; IDI2-AS1; ADAM8; CNNM2; ACBD5; PPP3CB-AS1; DNAJC1; TSPAN15; MPP7; SYT15; SLC25A16; THNSL1; DNAJC9; DUSP5; AL732437.3; SEPHS1; CAP1P2; AC016394.4; MAT1A; MXI1; AL356056.2; ZNF487; WDR37; C10orf25; ST8SIA6; ZNF438; BLOC1S2; MAPK8; DDX21"]},"columns":[{"accessor":"pathway","name":"Region","type":"character","minWidth":100},{"accessor":"ER","name":"ER","type":"character","cell":[" Unchanged","⇩ Down"," Unchanged","⇩ Down"],"maxWidth":90,"style":[{"color":"#7F7F7F"},{"color":"#0571B0"},{"color":"#7F7F7F"},{"color":"#0571B0"}]},{"accessor":"Windows","name":"Windows","type":"numeric","maxWidth":80},{"accessor":"Direction","name":"Gene Direction","type":"character","maxWidth":120,"style":[{"color":"#CA0020"},{"color":"#CA0020"},{"color":"#CA0020"},{"color":"#CA0020"}]},{"accessor":"p","name":"p","type":"numeric","cell":["6.37e-04","3.90e-03","5.61e-03","1.34e-02"],"maxWidth":80},{"accessor":"FDR","name":"FDR","type":"numeric","cell":["5.09e-03","1.50e-02","1.50e-02","2.68e-02"],"maxWidth":80},{"accessor":"Edge Size","name":"Edge Size","type":"numeric","maxWidth":80},{"accessor":"Leading Edge","name":"Leading Edge","type":"character","cell":[{"name":"span","attribs":{"title":"ADRA2A; KCNMA1; PPIF; MYPN; CEP55; NEBL; ASB13; AKR1C3; OLMALINC; CDK1; SCD; UNC5B; MMRN2; FAM107B; ANK3; ITGB1; NPM3; USP54; LINC00595; SNCG; HECTD2; ZMYND11; GATA3; ARL3; DDIT4; TASOR2; GLUD1; SVIL2P; MAP3K8; ADIRF-AS1; USP6NL; STOX1; STN1; SLF2; SFXN2; SLC16A12; CACNB2; HERC4; SEMA4G; ASCC1; ZMIZ1-AS1; PDZD7; KIF5B; ADAM8; CNNM2; AKR1C4; PPP3CB-AS1; TWNK; SLC29A3; PRKG1; AFAP1L2; SFMBT2; RPL12P29; KAZALD1; DNAJC9; AL732437.3"},"children":["ADRA2A; KCNMA1; PPIF; MYPN; CEP55; NEBL; ASB13;..."]},{"name":"span","attribs":{"title":"KCNMA1; PPIF; HELLS; SVIL; CEP55; ASB13; RHOBTB1; RET; OLMALINC; CDK1; LINC02635; SCD; UNC5B; MMRN2; FAM107B; RASGEF1A; ITGB1; NOLC1; LINC00595; SNCG; HECTD2; ZMYND11; PPRC1; ARL3; AL391421.1; DDIT4; TASOR2; GLUD1; SVIL2P; MAP3K8; ADIRF-AS1; USP6NL; STN1; IDI1; NOC3L; SLF2; SFXN2; GTPBP4; SRP9P1; ASCC1; CPXM2; NUTM2A; ZMIZ1-AS1; IDI2-AS1; DCLRE1C; PLCE1; SLC29A3; SUV39H2; MPP7; SYT15; SFMBT2; PSD; DNAJC9; AL732437.3; HNRNPF; PRR26; RPS3AP5; AL356056.2; WDR37; ST8SIA6; ZNF438; BLOC1S2; DDX21"},"children":["KCNMA1; PPIF; HELLS; SVIL; CEP55; ASB13; RHOBTB..."]},{"name":"span","attribs":{"title":"KCNMA1; CEP55; AKR1C3; KIF11; OLMALINC; AKR1C2; SCD; AKR1C1; MMRN2; FAM107B; SNCG; HECTD2; DDIT4; GLUD1; ADIRF-AS1; MSS51; HHEX; STN1; SLF2; CACNB2; SRP9P1; ASCC1"},"children":["KCNMA1; CEP55; AKR1C3; KIF11; OLMALINC; AKR1C2;..."]},{"name":"span","attribs":{"title":"PPIF; SYNPO2L; SVIL; CEP55; NEBL; ASB13; RHOBTB1; CALML5; RET; OLMALINC; ALOX5; SCD; ARHGAP19; MMRN2; FAM107B; ITGB1; DHTKD1; NOLC1; USP54; LINC00595; SNCG; ZMYND11; GATA3; PPRC1; ARL3; AL391421.1; DDIT4; TASOR2; RBM20; GLUD1; PLEKHS1; CASP7; MAP3K8; ADIRF-AS1; USP6NL; ZNF488; DEPP1; STOX1; HHEX; STN1; BAG3; TMEM26; FRMPD2; IDI1; SLF2; SFXN2; BAMBI; P4HA1; CACNB2; GTPBP4; ZNF239; ANXA11; HERC4; FAM241B; DNAJC12; ZMIZ1-AS1; GSTO1; IDI2-AS1; ADAM8; CNNM2; ACBD5; PPP3CB-AS1; DNAJC1; TSPAN15; MPP7; SYT15; SLC25A16; THNSL1; DNAJC9; DUSP5; AL732437.3; SEPHS1; CAP1P2; AC016394.4; MAT1A; MXI1; AL356056.2; ZNF487; WDR37; C10orf25; ST8SIA6; ZNF438; BLOC1S2; MAPK8; DDX21"},"children":["PPIF; SYNPO2L; SVIL; CEP55; NEBL; ASB13; RHOBTB..."]}],"minWidth":150}],"filterable":true,"defaultPageSize":10,"paginationType":"numbers","showPageInfo":true,"minRows":1,"dataKey":"b70ac741ee4699f0779018edc0454ff5","key":"b70ac741ee4699f0779018edc0454ff5"},"children":[]},"class":"reactR_markup"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<pre class="r"><code>plot_grid(plotlist = p, nrow = ceiling(length(p) / 3))</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/barcode-gsea-feat-nondir-sig-1.png" alt="*Barcode plots for the top 4 sets of windows associated with __nondirectional__ changes in gene expression.*" width="960" />
<p class="caption">
<em>Barcode plots for the top 4 sets of windows associated with <strong>nondirectional</strong> changes in gene expression.</em>
</p>
</div>
<!-- Could be done regardless of changes in expression -->
</div>
</div>
<div id="highly-ranked-genes" class="section level2 tabset">
<h2 class="tabset">Highly Ranked Genes</h2>
<pre class="r"><code>all_ids &lt;- unlist(merged_results$gene_id)
db_ids &lt;- unlist(filter(merged_results, !!sym(fdr_column) &lt; fdr_alpha)$gene_id)
top_diff_win &lt;- rnaseq %&gt;% 
  mutate(
    n_windows = vapply(
      gene_id,
      function(x) sum(str_detect(all_ids, x)),
      integer(1)
    ),
    n_db_windows = vapply(
      gene_id,
      function(x) sum(str_detect(db_ids, x)),
      integer(1)
    )
  ) %&gt;% 
  arrange(!!sym(rna_p_col)) %&gt;% 
  dplyr::filter(n_windows &gt; 1) %&gt;% 
  dplyr::slice(1:5)
grl_to_plot &lt;- top_diff_win$gene_id %&gt;% 
  sapply(
    function(x) {
      subset(
        merged_results, 
        vapply(gene_id, function(id) x %in% id, logical(1))
      ) %&gt;% 
        granges() %&gt;% 
        ## Restrict this to only sites within 5Mb of a gene
        subsetByOverlaps(
          resize(subset(genes_gr, gene_id == x), width = 1e7, fix = &#39;center&#39;)
        ) %&gt;% 
        c(
          granges(subset(genes_gr, gene_id == x))
        ) %&gt;% 
        unstrand() %&gt;% 
        sort()
    },
    simplify = FALSE
  ) %&gt;% 
  lapply(range) %&gt;% 
  GRangesList()
hfgc_genes &lt;- hfgc_genes[names(hfgc_genes) != &quot;Undetected&quot;]
if (!is.null(config$external$coverage)) {
  y_ranges &lt;- granges(unlist(grl_to_plot))
  y_lim &lt;- c(
    y_lim[target],
    bwfl[names(bwfl) != target] %&gt;% 
      lapply(
        function(x) {
          GRangesList(lapply(x, import.bw, which = y_ranges)) %&gt;% 
            unlist() %&gt;% 
            filter(score == max(score)) %&gt;% 
            mcols() %&gt;% 
            .[[&quot;score&quot;]] %&gt;% 
            c(0) %&gt;% 
            range()
        }
      )
  )
}
i &lt;- 1</code></pre>
<p>Using the ranked list of gens from the RNA-Seq data, the 5 genes most highly ranked for differential expression, with more than one ER-bound window mapped to it were selected for visualisation.</p>
<div id="kcnma1" class="section level3">
<h3>KCNMA1</h3>
<pre class="r"><code>ct &lt;- trans_models %&gt;% 
  subsetByOverlaps(grl_to_plot[[i]]) %&gt;% 
  as_tibble() %&gt;% 
  pull(&quot;transcript&quot;) %&gt;% 
  unique() %&gt;% 
  length() %&gt;% 
  is_greater_than(10) %&gt;% 
  ifelse(&quot;meta&quot;, FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic %&gt;% 
     subsetByOverlaps(
       subset(genes_gr, gene_id == names(grl_to_plot)[[i]])
     ) %&gt;% 
    subsetByOverlaps(
      subsetByOverlaps(merged_results, grl_to_plot[[i]])
    ),
  features = feat_gr, featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %&gt;%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  fontsize = 10,
  max = width(grl_to_plot[[i]]),
  ylim = y_lim,
  highlight = c(),
  col.title = &quot;black&quot;, background.title = &quot;white&quot;, showAxis = FALSE,
  rotation.title = 90
)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-gene1-1.png" alt="*Region showing all merged ER-bound windows mapped to PPIF. Windows considered to be unchanged for ER are annotated in grey, with other colours indicating ER gain or loss. Mapped windows are only shown if within 5Mb of PPIF. HiC Interactions are only shown which directly connect ER to PPIF. Gene-centric regions and external features are shown in the top panel. The estimated logFC for PPIF is 2.52 with an FDR of 1.26e-05. Undetected and unchanged genes are shown on separate tracks.*" width="960" />
<p class="caption">
<em>Region showing all merged ER-bound windows mapped to PPIF. Windows considered to be unchanged for ER are annotated in grey, with other colours indicating ER gain or loss. Mapped windows are only shown if within 5Mb of PPIF. HiC Interactions are only shown which directly connect ER to PPIF. Gene-centric regions and external features are shown in the top panel. The estimated logFC for PPIF is 2.52 with an FDR of 1.26e-05. Undetected and unchanged genes are shown on separate tracks.</em>
</p>
</div>
<pre class="r"><code>i &lt;- i + 1</code></pre>
</div>
<div id="ppif" class="section level3">
<h3>PPIF</h3>
<pre class="r"><code>ct &lt;- trans_models %&gt;% 
  subsetByOverlaps(grl_to_plot[[i]]) %&gt;% 
  as_tibble() %&gt;% 
  pull(&quot;transcript&quot;) %&gt;% 
  unique() %&gt;% 
  length() %&gt;% 
  is_greater_than(10) %&gt;% 
  ifelse(&quot;meta&quot;, FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic %&gt;% 
     subsetByOverlaps(
       subset(genes_gr, gene_id == names(grl_to_plot)[[i]])
     ) %&gt;% 
    subsetByOverlaps(
      subsetByOverlaps(merged_results, grl_to_plot[[i]])
    ),
  features = feat_gr, featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %&gt;%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  max = width(grl_to_plot[[i]]),
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  col.title = &quot;black&quot;, background.title = &quot;white&quot;, showAxis = FALSE,
  rotation.title = 90
)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-gene2-1.png" alt="*Region showing all merged ER-bound windows mapped to MYPN. Windows considered to be unchanged for ER are annotated in grey, with other colours indicating ER gain or loss. Mapped windows are only shown if within 5Mb of MYPN. HiC Interactions are only shown which directly connect ER to MYPN. Gene-centric regions and external features are shown in the top panel. The estimated logFC for MYPN is 4.09 with an FDR of 5.75e-05. Undetected and unchanged genes are shown on separate tracks.*" width="960" />
<p class="caption">
<em>Region showing all merged ER-bound windows mapped to MYPN. Windows considered to be unchanged for ER are annotated in grey, with other colours indicating ER gain or loss. Mapped windows are only shown if within 5Mb of MYPN. HiC Interactions are only shown which directly connect ER to MYPN. Gene-centric regions and external features are shown in the top panel. The estimated logFC for MYPN is 4.09 with an FDR of 5.75e-05. Undetected and unchanged genes are shown on separate tracks.</em>
</p>
</div>
<pre class="r"><code>i &lt;- i + 1</code></pre>
</div>
<div id="mypn" class="section level3">
<h3>MYPN</h3>
<pre class="r"><code>ct &lt;- trans_models %&gt;% 
  subsetByOverlaps(grl_to_plot[[i]]) %&gt;% 
  as_tibble() %&gt;% 
  pull(&quot;transcript&quot;) %&gt;% 
  unique() %&gt;% 
  length() %&gt;% 
  is_greater_than(10) %&gt;% 
  ifelse(&quot;meta&quot;, FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic %&gt;% 
     subsetByOverlaps(
       subset(genes_gr, gene_id == names(grl_to_plot)[[i]])
     ) %&gt;% 
    subsetByOverlaps(
      subsetByOverlaps(merged_results, grl_to_plot[[i]])
    ),
  features = feat_gr, featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %&gt;%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  max = width(grl_to_plot[[i]]),
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  col.title = &quot;black&quot;, background.title = &quot;white&quot;, showAxis = FALSE,
  rotation.title = 90
)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-gene3-1.png" alt="*Region showing all merged ER-bound windows mapped to SYNPO2L. Windows considered to be unchanged for ER are annotated in grey, with other colours indicating ER gain or loss. Mapped windows are only shown if within 5Mb of SYNPO2L. HiC Interactions are only shown which directly connect ER to SYNPO2L. Gene-centric regions and external features are shown in the top panel. The estimated logFC for SYNPO2L is 2.86 with an FDR of 9.67e-05. Undetected and unchanged genes are shown on separate tracks.*" width="960" />
<p class="caption">
<em>Region showing all merged ER-bound windows mapped to SYNPO2L. Windows considered to be unchanged for ER are annotated in grey, with other colours indicating ER gain or loss. Mapped windows are only shown if within 5Mb of SYNPO2L. HiC Interactions are only shown which directly connect ER to SYNPO2L. Gene-centric regions and external features are shown in the top panel. The estimated logFC for SYNPO2L is 2.86 with an FDR of 9.67e-05. Undetected and unchanged genes are shown on separate tracks.</em>
</p>
</div>
<pre class="r"><code>i &lt;- i + 1</code></pre>
</div>
<div id="synpo2l" class="section level3">
<h3>SYNPO2L</h3>
<pre class="r"><code>ct &lt;- trans_models %&gt;% 
  subsetByOverlaps(grl_to_plot[[i]]) %&gt;% 
  as_tibble() %&gt;% 
  pull(&quot;transcript&quot;) %&gt;% 
  unique() %&gt;% 
  length() %&gt;% 
  is_greater_than(10) %&gt;% 
  ifelse(&quot;meta&quot;, FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic %&gt;% 
     subsetByOverlaps(
       subset(genes_gr, gene_id == names(grl_to_plot)[[i]])
     ) %&gt;% 
    subsetByOverlaps(
      subsetByOverlaps(merged_results, grl_to_plot[[i]])
    ),
  features = feat_gr, featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %&gt;%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  max = width(grl_to_plot[[i]]),
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  col.title = &quot;black&quot;, background.title = &quot;white&quot;, showAxis = FALSE,
  rotation.title = 90
)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-gene4-1.png" alt="*Region showing all merged ER-bound windows mapped to SVIL. Windows considered to be unchanged for ER are annotated in grey, with other colours indicating ER gain or loss. Mapped windows are only shown if within 5Mb of SVIL. HiC Interactions are only shown which directly connect ER to SVIL. Gene-centric regions and external features are shown in the top panel. The estimated logFC for SVIL is -1.14 with an FDR of 3.51e-04. Undetected and unchanged genes are shown on separate tracks.*" width="960" />
<p class="caption">
<em>Region showing all merged ER-bound windows mapped to SVIL. Windows considered to be unchanged for ER are annotated in grey, with other colours indicating ER gain or loss. Mapped windows are only shown if within 5Mb of SVIL. HiC Interactions are only shown which directly connect ER to SVIL. Gene-centric regions and external features are shown in the top panel. The estimated logFC for SVIL is -1.14 with an FDR of 3.51e-04. Undetected and unchanged genes are shown on separate tracks.</em>
</p>
</div>
<pre class="r"><code>i &lt;- i + 1</code></pre>
</div>
<div id="svil" class="section level3">
<h3>SVIL</h3>
<pre class="r"><code>ct &lt;- trans_models %&gt;% 
  subsetByOverlaps(grl_to_plot[[i]]) %&gt;% 
  as_tibble() %&gt;% 
  pull(&quot;transcript&quot;) %&gt;% 
  unique() %&gt;% 
  length() %&gt;% 
  is_greater_than(10) %&gt;% 
  ifelse(&quot;meta&quot;, FALSE)
plotHFGC(
  grl_to_plot[[i]],
  hic = hic %&gt;% 
     subsetByOverlaps(
       subset(genes_gr, gene_id == names(grl_to_plot)[[i]])
     ) %&gt;% 
    subsetByOverlaps(
      subsetByOverlaps(merged_results, grl_to_plot[[i]])
    ),
  features = feat_gr, featcol = feature_colours,
  genes = hfgc_genes, genecol = gene_col, genesize = 0.7,
  coverage = bwfl,
  linecol = line_col,
  annotation = annot,
  annotcol = colours$direction %&gt;%
    setNames(str_to_title(names(.))),
  cytobands = bands_df,
  collapseTranscripts = ct, # Set this as a function of n_transcripts
  zoom = 1.1,
  max = width(grl_to_plot[[i]]),
  fontsize = 10,
  ylim = y_lim,
  highlight = c(),
  col.title = &quot;black&quot;, background.title = &quot;white&quot;, showAxis = FALSE,
  rotation.title = 90
)</code></pre>
<div class="figure" style="text-align: center">
<img src="ER_Veh_DHT_differential_binding_files/figure-html/plot-gene5-1.png" alt="*Region showing all merged ER-bound windows mapped to SVIL. Windows considered to be unchanged for ER are annotated in grey, with other colours indicating ER gain or loss. Mapped windows are only shown if within 5Mb of SVIL. HiC Interactions are only shown which directly connect ER to SVIL. Gene-centric regions and external features are shown in the top panel. The estimated logFC for SVIL is -1.14 with an FDR of 3.51e-04. Undetected and unchanged genes are shown on separate tracks.*" width="960" />
<p class="caption">
<em>Region showing all merged ER-bound windows mapped to SVIL. Windows considered to be unchanged for ER are annotated in grey, with other colours indicating ER gain or loss. Mapped windows are only shown if within 5Mb of SVIL. HiC Interactions are only shown which directly connect ER to SVIL. Gene-centric regions and external features are shown in the top panel. The estimated logFC for SVIL is -1.14 with an FDR of 3.51e-04. Undetected and unchanged genes are shown on separate tracks.</em>
</p>
</div>
</div>
</div>
<div id="integration-with-gene-expression-gsea" class="section level2 tabset">
<h2 class="tabset">Integration With Gene Expression GSEA</h2>
<pre class="r"><code>rnaseq_gsea &lt;- fgsea(gs_by_gsid, rna_dir_ranks)</code></pre>
<p>The enrichment testing results from the ER differential binding analysis were then compared to GSEA results obtained from the RNA-Seq data set</p>
<div id="differentially-bound-windows" class="section level3">
<h3>Differentially Bound Windows</h3>
<pre class="r"><code>cmn_gs &lt;- intersect(
  dplyr::filter(goseq_diff, fdr &lt; fdr_alpha)$gs_name,
  dplyr::filter(rnaseq_gsea, padj &lt; fdr_alpha)$pathway
)</code></pre>
<p>No gene sets were shared between all ER differentially bound windows and the RNA-Seq data</p>
</div>
<div id="windows-with-increased-er" class="section level3">
<h3>Windows With Increased ER</h3>
<pre class="r"><code>cmn_gs &lt;- intersect(
  dplyr::filter(goseq_up, fdr &lt; fdr_alpha)$gs_name,
  dplyr::filter(rnaseq_gsea, padj &lt; fdr_alpha)$pathway
)</code></pre>
<p>No gene sets were shared between ER gained windows and the RNA-Seq data</p>
</div>
<div id="windows-with-decreased-er" class="section level3">
<h3>Windows With Decreased ER</h3>
<pre class="r"><code>cmn_gs &lt;- intersect(
  dplyr::filter(goseq_down, fdr &lt; fdr_alpha)$gs_name,
  dplyr::filter(rnaseq_gsea, padj &lt; fdr_alpha)$pathway
)</code></pre>
<p>No gene sets were shared between ER decreased windows and the RNA-Seq data</p>
</div>
</div>
</div>
<div id="data-export" class="section level1">
<h1>Data Export</h1>
<pre class="r"><code>merged_results %&gt;% 
  as_tibble() %&gt;%
  unnest(all_of(&quot;gene_id&quot;)) %&gt;%
  mutate(gene_name = id2gene[gene_id]) %&gt;% 
  dplyr::select(
    gene_id, gene_name, range, AveExpr, logFC, P.Value, FDR = !!sym(fdr_column), 
    status, any_of(c(&quot;region&quot;, &quot;feature&quot;)), macs2_peak = overlaps_ref,
  ) %&gt;% 
  mutate(
    distance_to_gene = distance(
      GRanges(range),
      setNames(genes_gr, genes_gr$gene_id)[gene_id]
    )
  ) %&gt;% 
  write_csv(
    gzfile(all_out$csv)
  )</code></pre>
<pre class="r"><code>write_rds(merged_results, all_out$results, compress = &quot;gz&quot;)
write_rds(filtered_counts, all_out$windows, compress = &quot;gz&quot;)

file.create(all_out$up_regions)
if (length(sig_ranges$Up) &gt; 0) 
  export.bed(granges(sig_ranges$Up), all_out$up_regions)

file.create(all_out$down_regions)
if (length(sig_ranges$Down) &gt; 0) 
  export.bed(granges(sig_ranges$Down), all_out$down_regions)

save.image(all_out$renv)</code></pre>
<p>During this workflow, the following files were exported:</p>
<ul>
<li><strong>results</strong>: output/ER/Veh_DHT_differential_binding.rds</li>
<li><strong>csv</strong>: output/ER/ER_Veh_DHT_differential_binding.csv.gz</li>
<li><strong>windows</strong>: output/ER/Veh_DHT_filtered_windows.rds</li>
<li><strong>up_regions</strong>: output/ER/Veh_DHT_up.bed</li>
<li><strong>down_regions</strong>: output/ER/Veh_DHT_down.bed</li>
<li><strong>renv</strong>: output/envs/ER_Veh_DHT_differential_binding.RData</li>
</ul>
<!-- end of list -->
<button type="button" class="btn btn-default btn-sessioninfo" data-toggle="collapse" data-target="#sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Session information
</button>
</p>
<div id="sessioninfo" class="collapse">
<p><strong>R version 4.1.1 (2021-08-10)</strong></p>
<p><strong>Platform:</strong> x86_64-conda-linux-gnu (64-bit)</p>
<p><strong>locale:</strong> <em>LC_CTYPE=en_AU.UTF-8</em>, <em>LC_NUMERIC=C</em>, <em>LC_TIME=en_AU.UTF-8</em>, <em>LC_COLLATE=en_AU.UTF-8</em>, <em>LC_MONETARY=en_AU.UTF-8</em>, <em>LC_MESSAGES=en_AU.UTF-8</em>, <em>LC_PAPER=en_AU.UTF-8</em>, <em>LC_NAME=C</em>, <em>LC_ADDRESS=C</em>, <em>LC_TELEPHONE=C</em>, <em>LC_MEASUREMENT=en_AU.UTF-8</em> and <em>LC_IDENTIFICATION=C</em></p>
<p><strong>attached base packages:</strong> <em>stats4</em>, <em>stats</em>, <em>graphics</em>, <em>grDevices</em>, <em>utils</em>, <em>datasets</em>, <em>methods</em> and <em>base</em></p>
<p><strong>other attached packages:</strong> <em>fgsea(v.1.20.0)</em>, <em>extraChIPs(v.0.99.0)</em>, <em>goseq(v.1.46.0)</em>, <em>geneLenDataBase(v.1.30.0)</em>, <em>BiasedUrn(v.1.07)</em>, <em>msigdbr(v.7.5.1)</em>, <em>htmltools(v.0.5.2)</em>, <em>reactable(v.0.2.3)</em>, <em>GenomicInteractions(v.1.28.0)</em>, <em>InteractionSet(v.1.22.0)</em>, <em>ggside(v.0.2.0)</em>, <em>rlang(v.1.0.2)</em>, <em>ggrepel(v.0.9.1)</em>, <em>IHW(v.1.22.0)</em>, <em>statmod(v.1.4.36)</em>, <em>magrittr(v.2.0.3)</em>, <em>qsmooth(v.1.10.0)</em>, <em>quantro(v.1.28.0)</em>, <em>scales(v.1.1.1)</em>, <em>edgeR(v.3.36.0)</em>, <em>limma(v.3.50.1)</em>, <em>csaw(v.1.28.0)</em>, <em>SummarizedExperiment(v.1.24.0)</em>, <em>Biobase(v.2.54.0)</em>, <em>MatrixGenerics(v.1.6.0)</em>, <em>matrixStats(v.0.61.0)</em>, <em>ngsReports(v.1.10.0)</em>, <em>cowplot(v.1.1.1)</em>, <em>Rsamtools(v.2.10.0)</em>, <em>Biostrings(v.2.62.0)</em>, <em>XVector(v.0.34.0)</em>, <em>rtracklayer(v.1.54.0)</em>, <em>plyranges(v.1.14.0)</em>, <em>GenomicRanges(v.1.46.1)</em>, <em>GenomeInfoDb(v.1.30.1)</em>, <em>IRanges(v.2.28.0)</em>, <em>S4Vectors(v.0.32.4)</em>, <em>BiocGenerics(v.0.40.0)</em>, <em>yaml(v.2.3.5)</em>, <em>BiocParallel(v.1.28.3)</em>, <em>pander(v.0.6.5)</em>, <em>glue(v.1.6.2)</em>, <em>forcats(v.0.5.1)</em>, <em>stringr(v.1.4.0)</em>, <em>dplyr(v.1.0.8)</em>, <em>purrr(v.0.3.4)</em>, <em>readr(v.2.1.2)</em>, <em>tidyr(v.1.2.0)</em>, <em>tibble(v.3.1.6)</em>, <em>ggplot2(v.3.3.5)</em> and <em>tidyverse(v.1.3.1)</em></p>
<p><strong>loaded via a namespace (and not attached):</strong> <em>rappdirs(v.0.3.3)</em>, <em>bumphunter(v.1.36.0)</em>, <em>minfi(v.1.40.0)</em>, <em>bit64(v.4.0.5)</em>, <em>knitr(v.1.38)</em>, <em>DelayedArray(v.0.20.0)</em>, <em>data.table(v.1.14.2)</em>, <em>rpart(v.4.1.16)</em>, <em>KEGGREST(v.1.34.0)</em>, <em>RCurl(v.1.98-1.6)</em>, <em>GEOquery(v.2.62.2)</em>, <em>AnnotationFilter(v.1.18.0)</em>, <em>doParallel(v.1.0.17)</em>, <em>generics(v.0.1.2)</em>, <em>GenomicFeatures(v.1.46.5)</em>, <em>preprocessCore(v.1.56.0)</em>, <em>EnrichedHeatmap(v.1.24.0)</em>, <em>RSQLite(v.2.2.12)</em>, <em>bit(v.4.0.4)</em>, <em>tzdb(v.0.3.0)</em>, <em>xml2(v.1.3.3)</em>, <em>lubridate(v.1.8.0)</em>, <em>assertthat(v.0.2.1)</em>, <em>xfun(v.0.30)</em>, <em>hms(v.1.1.1)</em>, <em>jquerylib(v.0.1.4)</em>, <em>babelgene(v.22.3)</em>, <em>evaluate(v.0.15)</em>, <em>fansi(v.1.0.3)</em>, <em>restfulr(v.0.0.13)</em>, <em>scrime(v.1.3.5)</em>, <em>progress(v.1.2.2)</em>, <em>dbplyr(v.2.1.1)</em>, <em>readxl(v.1.4.0)</em>, <em>igraph(v.1.3.0)</em>, <em>DBI(v.1.1.2)</em>, <em>htmlwidgets(v.1.5.4)</em>, <em>reshape(v.0.8.8)</em>, <em>ellipsis(v.0.3.2)</em>, <em>crosstalk(v.1.2.0)</em>, <em>backports(v.1.4.1)</em>, <em>annotate(v.1.72.0)</em>, <em>biomaRt(v.2.50.3)</em>, <em>sparseMatrixStats(v.1.6.0)</em>, <em>vctrs(v.0.4.0)</em>, <em>here(v.1.0.1)</em>, <em>ensembldb(v.2.18.4)</em>, <em>cachem(v.1.0.6)</em>, <em>withr(v.2.5.0)</em>, <em>ggforce(v.0.3.3)</em>, <em>Gviz(v.1.38.3)</em>, <em>BSgenome(v.1.62.0)</em>, <em>vroom(v.1.5.7)</em>, <em>checkmate(v.2.0.0)</em>, <em>GenomicAlignments(v.1.30.0)</em>, <em>fdrtool(v.1.2.17)</em>, <em>prettyunits(v.1.1.1)</em>, <em>mclust(v.5.4.9)</em>, <em>cluster(v.2.1.3)</em>, <em>lazyeval(v.0.2.2)</em>, <em>crayon(v.1.5.1)</em>, <em>genefilter(v.1.76.0)</em>, <em>labeling(v.0.4.2)</em>, <em>pkgconfig(v.2.0.3)</em>, <em>slam(v.0.1-50)</em>, <em>tweenr(v.1.0.2)</em>, <em>nlme(v.3.1-157)</em>, <em>ProtGenerics(v.1.26.0)</em>, <em>nnet(v.7.3-17)</em>, <em>lifecycle(v.1.0.1)</em>, <em>filelock(v.1.0.2)</em>, <em>BiocFileCache(v.2.2.1)</em>, <em>modelr(v.0.1.8)</em>, <em>dichromat(v.2.0-0)</em>, <em>polyclip(v.1.10-0)</em>, <em>cellranger(v.1.1.0)</em>, <em>rprojroot(v.2.0.3)</em>, <em>reactR(v.0.4.4)</em>, <em>rngtools(v.1.5.2)</em>, <em>base64(v.2.0)</em>, <em>Matrix(v.1.4-1)</em>, <em>lpsymphony(v.1.22.0)</em>, <em>Rhdf5lib(v.1.16.0)</em>, <em>zoo(v.1.8-9)</em>, <em>reprex(v.2.0.1)</em>, <em>base64enc(v.0.1-3)</em>, <em>GlobalOptions(v.0.1.2)</em>, <em>png(v.0.1-7)</em>, <em>viridisLite(v.0.4.0)</em>, <em>rjson(v.0.2.21)</em>, <em>bitops(v.1.0-7)</em>, <em>rhdf5filters(v.1.6.0)</em>, <em>blob(v.1.2.2)</em>, <em>DelayedMatrixStats(v.1.16.0)</em>, <em>shape(v.1.4.6)</em>, <em>doRNG(v.1.8.2)</em>, <em>nor1mix(v.1.3-0)</em>, <em>jpeg(v.0.1-9)</em>, <em>memoise(v.2.0.1)</em>, <em>plyr(v.1.8.7)</em>, <em>zlibbioc(v.1.40.0)</em>, <em>scatterpie(v.0.1.7)</em>, <em>compiler(v.4.1.1)</em>, <em>BiocIO(v.1.4.0)</em>, <em>RColorBrewer(v.1.1-2)</em>, <em>illuminaio(v.0.36.0)</em>, <em>clue(v.0.3-60)</em>, <em>cli(v.3.2.0)</em>, <em>htmlTable(v.2.4.0)</em>, <em>Formula(v.1.2-4)</em>, <em>MASS(v.7.3-56)</em>, <em>mgcv(v.1.8-40)</em>, <em>tidyselect(v.1.1.2)</em>, <em>stringi(v.1.7.6)</em>, <em>highr(v.0.9)</em>, <em>askpass(v.1.1)</em>, <em>locfit(v.1.5-9.5)</em>, <em>latticeExtra(v.0.6-29)</em>, <em>grid(v.4.1.1)</em>, <em>sass(v.0.4.1)</em>, <em>VariantAnnotation(v.1.40.0)</em>, <em>fastmatch(v.1.1-3)</em>, <em>tools(v.4.1.1)</em>, <em>parallel(v.4.1.1)</em>, <em>circlize(v.0.4.14)</em>, <em>rstudioapi(v.0.13)</em>, <em>foreach(v.1.5.2)</em>, <em>foreign(v.0.8-82)</em>, <em>metapod(v.1.2.0)</em>, <em>gridExtra(v.2.3)</em>, <em>farver(v.2.1.0)</em>, <em>digest(v.0.6.29)</em>, <em>quadprog(v.1.5-8)</em>, <em>Rcpp(v.1.0.8.3)</em>, <em>siggenes(v.1.68.0)</em>, <em>broom(v.0.7.12)</em>, <em>httr(v.1.4.2)</em>, <em>ggdendro(v.0.1.23)</em>, <em>AnnotationDbi(v.1.56.2)</em>, <em>biovizBase(v.1.42.0)</em>, <em>ComplexHeatmap(v.2.10.0)</em>, <em>colorspace(v.2.0-3)</em>, <em>rvest(v.1.0.2)</em>, <em>XML(v.3.99-0.9)</em>, <em>fs(v.1.5.2)</em>, <em>splines(v.4.1.1)</em>, <em>multtest(v.2.50.0)</em>, <em>plotly(v.4.10.0)</em>, <em>xtable(v.1.8-4)</em>, <em>jsonlite(v.1.8.0)</em>, <em>ggfun(v.0.0.6)</em>, <em>R6(v.2.5.1)</em>, <em>Hmisc(v.4.6-0)</em>, <em>pillar(v.1.7.0)</em>, <em>fastmap(v.1.1.0)</em>, <em>DT(v.0.22)</em>, <em>beanplot(v.1.2)</em>, <em>codetools(v.0.2-18)</em>, <em>utf8(v.1.2.2)</em>, <em>lattice(v.0.20-45)</em>, <em>bslib(v.0.3.1)</em>, <em>sva(v.3.42.0)</em>, <em>curl(v.4.3.2)</em>, <em>GO.db(v.3.14.0)</em>, <em>openssl(v.2.0.0)</em>, <em>survival(v.3.3-1)</em>, <em>rmarkdown(v.2.13)</em>, <em>munsell(v.0.5.0)</em>, <em>GetoptLong(v.1.0.5)</em>, <em>rhdf5(v.2.38.1)</em>, <em>GenomeInfoDbData(v.1.2.7)</em>, <em>iterators(v.1.0.14)</em>, <em>HDF5Array(v.1.22.1)</em>, <em>haven(v.2.4.3)</em> and <em>gtable(v.0.3.0)</em></p>
</div>
</div>

<div class="footer">
  Workflow developed by:<br>
  Dr Stephen Pederson<br>
  Dame Roma Mitchell Cancer Research Laboratories<br>
  Adelaide Medical School, University of Adelaide, SA, Australia
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
