<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Outline Of Analysis</title>

<script src="site_libs/header-attrs-2.13/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<script src="site_libs/viz-1.8.2/viz.js"></script>
<link href="site_libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet" />
<script src="site_libs/grViz-binding-1.0.9/grViz.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GRAVI</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Annotations
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="annotation_description.html">Genome and Transcriptome Annotations</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    MACS2 Peak Calling
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="AR_macs2_summary.html">AR</a>
    </li>
    <li>
      <a href="ER_macs2_summary.html">ER</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Differential Binding
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">AR</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="AR_Veh_DHT_differential_binding.html">DHT Vs. Veh</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">ER</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="ER_Veh_DHT_differential_binding.html">DHT Vs. Veh</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pairwise Comparisons
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">AR-ER</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="AR_Veh_DHT_ER_Veh_DHT_pairwise_comparison.html">DHT Vs. Veh / DHT Vs. Veh</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steveped/GRAVI">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Outline Of Analysis</h1>

</div>


<pre class="r"><code>library(tidyverse)
library(DiagrammeR)
library(pander)
library(yaml)
library(scales)
source(
  here::here(&quot;workflow/scripts/plot_dag_functions.R&quot;)
)</code></pre>
<pre class="r"><code>config &lt;- here::here(&quot;config/config.yml&quot;) %&gt;%
  read_yaml()
samples &lt;- config$samples$file %&gt;%
  here::here() %&gt;%
  read_tsv()</code></pre>
<div id="description" class="section level2">
<h2>Description</h2>
<p>This is a standardised workflow for beginning the comparison between two or ChIP targets, using BAM files as the starting point. Treatment groups and targets are specified using <code>config/config.yml</code>.</p>
</div>
<div id="workflow" class="section level2">
<h2>Workflow</h2>
<pre class="r"><code>here::here(&quot;workflow&quot;, &quot;rules&quot;, &quot;rulegraph.dot&quot;) %&gt;%
  readLines() %&gt;%
  rm_dot_node(node = &quot;\&quot;all\&quot;&quot;) %&gt;%
  add_input_node(node = &quot;Alignments\n(Bam Files)&quot;, col = &quot;red&quot;, ignore = &quot;(download|define|macs2|create|install)&quot;) %&gt;%
  change_node_colour(&quot;(setup|index|macs2|bedgraph|identify|binding|install)&quot;, &quot;red&quot;) %&gt;%
  change_node_colour(&quot;(download|write|copy|install)&quot;, &quot;blue&quot;) %&gt;%
  change_node_colour(&quot;(site|create|build|differential|pairwise)&quot;, &quot;forestgreen&quot;) %&gt;%
  str_replace_all(&quot;_&quot;, &quot;\n&quot;) %&gt;%
  str_replace_all(&quot;snakemake\ndag&quot;, &quot;snakemake_dag&quot;) %&gt;%
  str_replace_all(&quot;fontsize=[0-9]+&quot;, &quot;fontsize=16&quot;) %&gt;%
  str_replace_all(&quot;(.+graph.+)(\\];)&quot;, &quot;\\1, rankdir = LR\\2&quot;) %&gt;%
  grViz()</code></pre>
<div class="figure" style="text-align: center">
<div id="htmlwidget-5f39655aac5ad23f5845" style="width:960px;height:960px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-5f39655aac5ad23f5845">{"x":{"diagram":"digraph snakemake_dag {\n    graph[bgcolor=white, margin=0, rankdir = LR];\n    node[shape=box, style=rounded, fontname=sans,                 fontsize=16, penwidth=2];\n    edge[penwidth=2, color=grey];\n\t1[label = \"create\nannotations\", color = \"#228B22\", style=\"rounded\"];\n\t2[label = \"download\ngtf\", color = \"#0000FF\", style=\"rounded\"];\n\t3[label = \"install\npackages\", color = \"#0000FF\", style=\"rounded\"];\n\t4[label = \"build\nannotations\nrmd\", color = \"#228B22\", style=\"rounded\"];\n\t5[label = \"download\nblacklist\", color = \"#0000FF\", style=\"rounded\"];\n\t6[label = \"create\nsetup\nchunk\", color = \"#228B22\", style=\"rounded\"];\n\t7[label = \"create\nsite\nyaml\", color = \"#228B22\", style=\"rounded\"];\n\t8[label = \"build\nmacs2\nsummary\", color = \"#228B22\", style=\"rounded\"];\n\t9[label = \"index\nbam\", color = \"#FF0000\", style=\"rounded\"];\n\t10[label = \"macs2\nindividual\", color = \"#FF0000\", style=\"rounded\"];\n\t11[label = \"macs2\nmerged\", color = \"#FF0000\", style=\"rounded\"];\n\t12[label = \"macs2\nqc\", color = \"#FF0000\", style=\"rounded\"];\n\t13[label = \"differential\nbinding\", color = \"#228B22\", style=\"rounded\"];\n\t14[label = \"bedgraph\nto\nbigwig\", color = \"#FF0000\", style=\"rounded\"];\n\t15[label = \"pairwise\ncomparisons\", color = \"#228B22\", style=\"rounded\"];\n\t16[label = \"build\nsite\nindex\", color = \"#228B22\", style=\"rounded\"];\n\t17[label = \"Alignments\n(Bam Files)\", color = \"#FF0000\", style=\"rectangle\"];\n\t2 -> 1\n\t3 -> 1\n\t7 -> 4\n\t6 -> 4\n\t5 -> 4\n\t1 -> 4\n\t4 -> 8\n\t10 -> 8\n\t1 -> 8\n\t11 -> 8\n\t7 -> 8\n\t12 -> 8\n\t6 -> 8\n\t5 -> 8\n\t9 -> 8\n\t3 -> 8\n\t9 -> 10\n\t12 -> 11\n\t9 -> 11\n\t1 -> 12\n\t10 -> 12\n\t5 -> 12\n\t9 -> 12\n\t3 -> 12\n\t4 -> 13\n\t1 -> 13\n\t11 -> 13\n\t7 -> 13\n\t12 -> 13\n\t6 -> 13\n\t9 -> 13\n\t14 -> 13\n\t3 -> 13\n\t8 -> 13\n\t11 -> 14\n\t10 -> 14\n\t1 -> 14\n\t4 -> 15\n\t1 -> 15\n\t7 -> 15\n\t6 -> 15\n\t13 -> 15\n\t3 -> 15\n\t4 -> 16\n\t7 -> 16\n\t6 -> 16\n\t13 -> 16\n\t8 -> 16\n\t15 -> 16\n\t17 -> 9\n}            ","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>
<p class="caption">
<em>Summary of processing workflow. The primary data pipeline is shown in red, preparatory steps are shown in blue whilst collation of final output is in green.</em>
</p>
</div>
</div>
<div id="targets" class="section level2">
<h2>Targets</h2>
<pre class="r"><code>samples %&gt;% 
  group_by(target, treat) %&gt;%
  summarise(n = dplyr::n(), .groups = &quot;drop&quot;) %&gt;% 
  rename_all(str_to_title) %&gt;% 
  pander(
    justify = &quot;llr&quot;,
    caption = &quot;Summary of supplied ChIP targets and treatment groups&quot;
  )</code></pre>
<table style="width:29%;">
<caption>Summary of supplied ChIP targets and treatment groups</caption>
<colgroup>
<col width="12%" />
<col width="11%" />
<col width="5%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Target</th>
<th align="left">Treat</th>
<th align="right">N</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">AR</td>
<td align="left">DHT</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">AR</td>
<td align="left">Veh</td>
<td align="right">3</td>
</tr>
<tr class="odd">
<td align="left">ER</td>
<td align="left">DHT</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">ER</td>
<td align="left">Veh</td>
<td align="right">3</td>
</tr>
</tbody>
</table>
</div>
<div id="differential-binding" class="section level2">
<h2>Differential Binding</h2>
<p>The specified comparisons are given below, where ‘Vs.’ can be interpreted as a ‘-’ sign. This effectively makes the first listed group as the treatment and the second listed as the reference or control. Differential binding analysis will be performed for every ChIP target where both treatment groups are present.</p>
<ul>
<li>DHT Vs. Veh</li>
</ul>
<!-- end of list -->
</div>
<div id="pairwise-comparisons" class="section level2">
<h2>Pairwise Comparisons</h2>
<p>All combinations of the comparisons defined above will will be used to perform pairwise comparisions.</p>
</div>
<div id="settings" class="section level2">
<h2>Settings</h2>
<div id="genome" class="section level3">
<h3>Genome</h3>
<ul>
<li>The genome build (GRCh37) was sourced from Gencode release 33</li>
<li>Blacklisted sites were sourced from <a href="ENCODE">https://github.com/boyle-lab/Blacklist/</a> for this specific genome build</li>
</ul>
</div>
<div id="external-data" class="section level3">
<h3>External Data</h3>
<p>The external data was:</p>
<ul>
<li><strong>rnaseq</strong>: PDX_HCI-005_Veh_DHT_Sarm_RNASeq.tsv</li>
<li><strong>features</strong>: h3k27ac_features.gtf</li>
<li><strong>hic</strong>: encode_hic.bedpe</li>
<li><strong>coverage</strong>: <em>H3K27Ac_Veh.bw</em> and <em>H3K27Ac_DHT.bw</em></li>
</ul>
<!-- end of list -->
</div>
<div id="peak-detection" class="section level3">
<h3>Peak Detection</h3>
<p>Macs2 callpeak settings were default with the following exceptions:</p>
<ul>
<li><strong>gsize</strong>: hs</li>
<li><strong>fdr</strong>: <em>0.05</em></li>
<li><strong>keep_duplicates</strong>: all</li>
</ul>
<!-- end of list -->
<p>Two additional QC settings were incorporated for peak filtering.</p>
<ul>
<li>For a sample to pass initial QC, at least 10% of the peaks identified in the sample identifying the highest number of peaks need to be returned. Any many instances, no peaks will be returned in any condition and whilst in this situation all samples will be retained. However, samples significantly diverging from the ‘best’ sample will be excluded.</li>
<li>For a peak to be included in the set of “Oracle Peaks”, it needs to overlap an identified peak in at least 30% of samples passing the above QC steps</li>
</ul>
<button type="button" class="btn btn-default btn-sessioninfo" data-toggle="collapse" data-target="#sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Session information
</button>
</p>
<div id="sessioninfo" class="collapse">
<p><strong>R version 4.1.1 (2021-08-10)</strong></p>
<p><strong>Platform:</strong> x86_64-conda-linux-gnu (64-bit)</p>
<p><strong>locale:</strong> <em>LC_CTYPE=en_AU.UTF-8</em>, <em>LC_NUMERIC=C</em>, <em>LC_TIME=en_AU.UTF-8</em>, <em>LC_COLLATE=en_AU.UTF-8</em>, <em>LC_MONETARY=en_AU.UTF-8</em>, <em>LC_MESSAGES=en_AU.UTF-8</em>, <em>LC_PAPER=en_AU.UTF-8</em>, <em>LC_NAME=C</em>, <em>LC_ADDRESS=C</em>, <em>LC_TELEPHONE=C</em>, <em>LC_MEASUREMENT=en_AU.UTF-8</em> and <em>LC_IDENTIFICATION=C</em></p>
<p><strong>attached base packages:</strong> <em>stats</em>, <em>graphics</em>, <em>grDevices</em>, <em>utils</em>, <em>datasets</em>, <em>methods</em> and <em>base</em></p>
<p><strong>other attached packages:</strong> <em>scales(v.1.1.1)</em>, <em>yaml(v.2.3.5)</em>, <em>pander(v.0.6.5)</em>, <em>DiagrammeR(v.1.0.9)</em>, <em>forcats(v.0.5.1)</em>, <em>stringr(v.1.4.0)</em>, <em>dplyr(v.1.0.8)</em>, <em>purrr(v.0.3.4)</em>, <em>readr(v.2.1.2)</em>, <em>tidyr(v.1.2.0)</em>, <em>tibble(v.3.1.6)</em>, <em>ggplot2(v.3.3.5)</em> and <em>tidyverse(v.1.3.1)</em></p>
<p><strong>loaded via a namespace (and not attached):</strong> <em>Rcpp(v.1.0.8.3)</em>, <em>lubridate(v.1.8.0)</em>, <em>here(v.1.0.1)</em>, <em>visNetwork(v.2.1.0)</em>, <em>assertthat(v.0.2.1)</em>, <em>rprojroot(v.2.0.3)</em>, <em>digest(v.0.6.29)</em>, <em>utf8(v.1.2.2)</em>, <em>R6(v.2.5.1)</em>, <em>cellranger(v.1.1.0)</em>, <em>backports(v.1.4.1)</em>, <em>reprex(v.2.0.1)</em>, <em>evaluate(v.0.15)</em>, <em>httr(v.1.4.2)</em>, <em>pillar(v.1.7.0)</em>, <em>rlang(v.1.0.2)</em>, <em>readxl(v.1.4.0)</em>, <em>rstudioapi(v.0.13)</em>, <em>jquerylib(v.0.1.4)</em>, <em>rmarkdown(v.2.13)</em>, <em>htmlwidgets(v.1.5.4)</em>, <em>bit(v.4.0.4)</em>, <em>munsell(v.0.5.0)</em>, <em>broom(v.0.7.12)</em>, <em>compiler(v.4.1.1)</em>, <em>modelr(v.0.1.8)</em>, <em>xfun(v.0.30)</em>, <em>pkgconfig(v.2.0.3)</em>, <em>htmltools(v.0.5.2)</em>, <em>tidyselect(v.1.1.2)</em>, <em>fansi(v.1.0.3)</em>, <em>crayon(v.1.5.1)</em>, <em>tzdb(v.0.3.0)</em>, <em>dbplyr(v.2.1.1)</em>, <em>withr(v.2.5.0)</em>, <em>grid(v.4.1.1)</em>, <em>jsonlite(v.1.8.0)</em>, <em>gtable(v.0.3.0)</em>, <em>lifecycle(v.1.0.1)</em>, <em>DBI(v.1.1.2)</em>, <em>magrittr(v.2.0.3)</em>, <em>vroom(v.1.5.7)</em>, <em>cli(v.3.2.0)</em>, <em>stringi(v.1.7.6)</em>, <em>fs(v.1.5.2)</em>, <em>xml2(v.1.3.3)</em>, <em>bslib(v.0.3.1)</em>, <em>ellipsis(v.0.3.2)</em>, <em>generics(v.0.1.2)</em>, <em>vctrs(v.0.4.0)</em>, <em>RColorBrewer(v.1.1-3)</em>, <em>tools(v.4.1.1)</em>, <em>bit64(v.4.0.5)</em>, <em>glue(v.1.6.2)</em>, <em>hms(v.1.1.1)</em>, <em>parallel(v.4.1.1)</em>, <em>fastmap(v.1.1.0)</em>, <em>colorspace(v.2.0-3)</em>, <em>rvest(v.1.0.2)</em>, <em>knitr(v.1.38)</em>, <em>haven(v.2.4.3)</em> and <em>sass(v.0.4.1)</em></p>
</div>
</div>
</div>

<div class="footer">
  Workflow developed by:<br>
  Dr Stephen Pederson<br>
  Dame Roma Mitchell Cancer Research Laboratories<br>
  Adelaide Medical School, University of Adelaide, SA, Australia
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
