---
title: "Test Pairwise"
---

```{r set-knitr-opts, echo=FALSE, child = here::here('workflow/modules/setup_chunk.Rmd')}
```


```{r example, echo = FALSE, eval=TRUE}
## This module requires the following to be set during preparation
# threads <- 6
pairs <- list(AR = c("Veh", "DHT"), ER = c("Veh", "DHT"))
```



```{r packages}
library(tidyverse)
library(glue)
library(pander)
library(yaml)
library(reactable)
library(plyranges)
library(rtracklayer)
library(VennDiagram)
library(UpSetR)
library(magrittr)
library(scales)
library(ggrepel)
library(rlang)
library(ggside)
library(msigdbr)
library(htmltools)
library(goseq)
library(parallel)
# library(BiocParallel)
# library(cowplot)
# library(InteractionSet)
```

```{r remotes, results = 'hide'}
# if (!"extraChIPs" %in% rownames(installed.packages()))
#   BiocManager::install("steveped/extraChIPs", ask = FALSE)

stopifnot(library(extraChIPs, logical.return = TRUE))
source(here::here("workflow", "scripts", "custom_functions.R"))
```

```{r options}
panderOptions("big.mark", ",")
panderOptions("missing", "")
panderOptions("table.split.table", Inf)
theme_set(
  theme_bw() +
  theme(
    plot.title = element_text(hjust = 0.5)
  )
)
```

```{r set-bpparam}
# register(MulticoreParam(workers = threads))
```

```{r samples}
targets <- names(pairs)
samples <- targets %>% 
  lapply(
    function(x) {
      fl <- here::here("output", x, "qc_samples.tsv")
      read_tsv(fl) %>% 
        dplyr::filter(treat %in% pairs[[x]])
    }
  ) %>% 
  bind_rows() %>% 
  dplyr::filter(qc == "pass") %>% 
  unite(label, target, label, remove = FALSE) %>% 
  mutate(
    target = factor(target, levels = targets),
    treat = factor(treat, levels = unique(unlist(pairs)))
  ) %>% 
  dplyr::select(-qc) %>% 
  droplevels()
stopifnot(nrow(samples) > 0)
rep_col <- setdiff(
  colnames(samples), c("sample", "treat", "target", "input", "label", "qc")
)
samples[[rep_col]] <- as.factor(samples[[rep_col]])
```

```{r load-config}
params <- read_yaml(here::here("config", "params.yml"))
config <- read_yaml(here::here("config", "config.yml"))
fdr_alpha <- config$comparisons$fdr
colours <- read_rds(here::here("output", "annotations", "colours.rds"))
treat_cols <- unlist(colours$treat[levels(samples$treat)])
```

```{r load-annotations}
annotation_path <- here::here("output", "annotations")
sq <- read_rds(file.path(annotation_path, "seqinfo.rds"))
all_gr <- file.path(annotation_path, "all_gr.rds") %>% 
  read_rds()
external_features <- c()
has_features <- FALSE
if (!is.null(config$external$features)) {
  external_features <- suppressWarnings(
    import.gff(here::here(config$external$features), genome = sq)
  )
  keep_cols <- !vapply(
    mcols(external_features), function(x) all(is.na(x)), logical(1)
  )
  mcols(external_features) <- mcols(external_features)[keep_cols]
  has_features <- TRUE
  feature_colours <- colours$features %>% 
    setNames(str_sep_to_title(names(.)))
}
gene_regions <- read_rds(file.path(annotation_path, "gene_regions.rds"))
regions <- vapply(gene_regions, function(x) unique(x$region), character(1))
region_colours <-  unlist(colours$regions) %>% setNames(regions[names(.)])
any_detected <- gene_regions %>% 
  vapply(function(x) any(x$detected), logical(1)) %>% 
  any()
```

```{r tbl-funs}
up_col <- function(x) {
  rgb(
    colorRamp(c("#ffffff", colours$direction[["up"]]))(x), maxColorValue = 255
  )
}
down_col <- function(x) {
  rgb(
    colorRamp(c("#ffffff", colours$direction[["down"]]))(x), maxColorValue = 255
  )
}
unch_col <- function(x) {
  rgb(
    colorRamp(c("#ffffff", colours$direction[["unchanged"]]))(x), 
    maxColorValue = 255
  )
}
lfc_col <- function(x){
  rgb(
    colorRamp(c(colours$direction[["down"]], "#ffffff", colours$direction[["up"]]))(x), 
    maxColorValue = 255
  )
}
expr_col <- function(x){
  rgb(colorRamp(hcl.colors(9, "TealRose"))(x), maxColorValue = 255)
}
bar_style <- function(width = 1, fill = "#e6e6e6", height = "75%", align = c("left", "right"), color = NULL) {
  align <- match.arg(align)
  if (align == "left") {
    position <- paste0(width * 100, "%")
    image <- sprintf("linear-gradient(90deg, %1$s %2$s, transparent %2$s)", fill, position)
  } else {
    position <- paste0(100 - width * 100, "%")
    image <- sprintf("linear-gradient(90deg, transparent %1$s, %2$s %1$s)", position, fill)
  }
  list(
    backgroundImage = image,
    backgroundSize = paste("100%", height),
    backgroundRepeat = "no-repeat",
    backgroundPosition = "center",
    color = color
  )
}
with_tooltip <- function(value, width = 30) {
  htmltools::tags$span(title = value, str_trunc(value, width))
}
```


```{r load-data}
oracle_peaks <- targets %>% 
  lapply(
    function(x) {
      read_rds(
        here::here("output", x, "oracle_peaks.rds")
      )
    }
  ) %>% 
  setNames(targets)
consensus_peaks <- targets %>% 
  lapply(
    function(x) {
      import.bed(
        here::here("output", x, "consensus_peaks.bed"), seqinfo = sq
      )
    }
  ) %>% 
  setNames(targets)
all_consensus <- consensus_peaks %>% 
  GRangesList() %>% 
  unlist() %>% 
  reduce() %>% 
  sort() 
db_results <- targets %>% 
  lapply(
    function(x) {
      here::here(
        "output", x, glue("{pairs[[x]][[1]]}_{pairs[[x]][[2]]}_differential_binding.rds")
      )
    }
  ) %>% 
  lapply(read_rds) %>% 
  setNames(targets)
fdr_column <- ifelse(
  "fdr_ihw" %in% colnames(mcols(db_results[[1]])), 
  "fdr_ihw", 
  str_subset("fdr", colnames(mcols(db_results[[1]])))
)
```

# Peak Comparison {.tabset}

The sets of macs2-derived peaks for both `r targets[[1]]` and `r targets[[2]]` were first compared using the treatment-agnostic peaks for each target.
The sets of treatment-specific *oracle-peaks* were then compared across the combined treatment groups and targets.

## Consensus Peaks

```{r plot-consensus, results = 'hide'}
fig_path <- here::here("docs", "assets", paste(targets, collapse = "_"))
if (!dir.exists(fig_path)) dir.create(fig_path, recursive = TRUE)
fig_name <- targets %>% 
  paste(collapse = "_") %>% 
  paste0("_common_peaks.png")
lapply(
  consensus_peaks,
  function(x) as.character(subsetByOverlaps(all_consensus, x))
) %>% 
  venn.diagram(
    file.path(fig_path, fig_name),
    imagetype = "png",
    units = "in",
    cat.cex = 1.4,
    height = 9, 
    width = 10,
    fill = hcl.colors(2, "Zissou 1", rev = TRUE), # Might need to fix later
    alpha = 0.3
  )
 file.remove(list.files(fig_path, pattern = "log$", full.names = TRUE))
```

`r glue("![*Union of peaks found in {targets[[1]]} and {targets[[2]]}, and which of the two targets each peak is found in. Peaks are based on the union of both sets of consensus peaks, which are themselves independent of treatment group.*](assets/{targets[[1]]}_{targets[[2]]}/{fig_name})")`

## Treatment Specific Peaks

```{r oracle-peaks-upset, fig.height = 7, fig.cap = glue("*Macs2 peaks detected for {targets[[1]]} and {targets[[2]]} in the treatment groups ", glue_collapse(unique(sapply(pairs, glue_collapse, sep = " & ")), sep = " or "), ".*")}
samples %>% 
  distinct(target, treat) %>% 
  mutate_all(as.character) %>% 
  unite(group, target, treat, remove = FALSE) %>% 
  split(.$group) %>% 
  lapply(
    function(x) {
      gr <- subsetByOverlaps(all_consensus, oracle_peaks[[x$target]][[x$treat]])
      as.character(gr)
    }
  ) %>% 
  setNames(
    str_replace_all(names(.), "(.+)_(.+)", "\\1 (\\2)")
  ) %>% 
  fromList() %>% 
  upset(
    sets = targets %>% 
      vapply(function(x) paste(x, pairs[[x]], sep = "_"), character(2)) %>% 
      as.character() %>% 
      rev() %>% 
      str_replace_all("(.+)_(.+)", "\\1 (\\2)"),
    keep.order = TRUE,
    order.by = "freq",
    set_size.show = TRUE, 
    set_size.scale_max = oracle_peaks %>% 
      lapply(function(x) lapply(x, length)) %>% 
      unlist() %>% 
      max() %>% 
      multiply_by(1.15),
    sets.bar.color = pairs %>% 
      lapply(function(x) treat_cols[x]) %>% 
      unlist() %>% 
      as.character() %>% 
      rev()
  )
```

# Differentially Bound Windows {.tabset}

```{r all-windows}
all_windows <- db_results %>%
  lapply(
    select, any_of(c("gene_id", "gene_name", "detected"))
  ) %>% 
  GRangesList() %>% 
  unlist() %>% 
  sort() %>% 
  reduceMC() %>% 
  mutate(
    region = regions[
      bestOverlap(., GRangesList(lapply(gene_regions, granges)))
    ]
  ) %>%
  join_overlap_left(
    db_results[[targets[[1]]]] %>% 
      mutate(status = as.character(status)) %>% 
      select(
        !!sym(glue("{targets[[1]]}_AveExpr")) := AveExpr, 
        !!sym(glue("{targets[[1]]}_logFC")) := logFC, 
        !!sym(glue("{targets[[1]]}_status")) := status, 
        !!sym(glue("{targets[[1]]}_fdr")) := !!sym(fdr_column)
      )
  ) %>%
  join_overlap_left(
    db_results[[targets[[2]]]] %>% 
      mutate(status = as.character(status)) %>% 
      select(
        !!sym(glue("{targets[[2]]}_AveExpr")) := AveExpr, 
        !!sym(glue("{targets[[2]]}_logFC")) := logFC, 
        !!sym(glue("{targets[[2]]}_status")) := status,
        !!sym(glue("{targets[[2]]}_fdr")) := !!sym(fdr_column)
      )
  ) %>% 
  mutate(
    
    ## Change the classification for unchanged sites to DB using 3 * fdr_alpha 
    ## if the other target is significantly DB
    "{targets[[1]]}_status" := case_when(
      
      (!!sym(glue("{targets[[2]]}_fdr")) < fdr_alpha) &
        (!!sym(glue("{targets[[1]]}_fdr")) < fdr_alpha * 3) &
        (!!sym(glue("{targets[[1]]}_logFC")) > 0) ~ "Up",
      
      (!!sym(glue("{targets[[2]]}_fdr")) < fdr_alpha) &
        (!!sym(glue("{targets[[1]]}_fdr")) < fdr_alpha * 3) &
        (!!sym(glue("{targets[[1]]}_logFC")) < 0) ~ "Down",
      
      TRUE ~ !!sym(glue("{targets[[1]]}_status"))
    ) %>% 
      factor(levels = str_sep_to_title(names(colours$direction))) %>% 
      fct_explicit_na("Undetected") %>% 
      fct_relabel(function(x) paste(targets[[1]], x)),
    
    "{targets[[2]]}_status" := case_when(
      
      (!!sym(glue("{targets[[1]]}_fdr")) < fdr_alpha) &
        (!!sym(glue("{targets[[2]]}_fdr")) < fdr_alpha * 3) &
        (!!sym(glue("{targets[[2]]}_logFC")) > 0) ~ "Up",
      
      (!!sym(glue("{targets[[1]]}_fdr")) < fdr_alpha) &
        (!!sym(glue("{targets[[2]]}_fdr")) < fdr_alpha * 3) &
        (!!sym(glue("{targets[[2]]}_logFC")) < 0) ~ "Down",
      
      TRUE ~ !!sym(glue("{targets[[2]]}_status"))
    ) %>% 
      factor(levels = str_sep_to_title(names(colours$direction)))%>% 
      fct_explicit_na("Undetected") %>% 
      fct_relabel(function(x) paste(targets[[2]], x)),
    
    status = fct_cross(
      !!sym(glue("{targets[[1]]}_status")), 
      !!sym(glue("{targets[[2]]}_status")),
      sep = " - "
    )
  )
if (has_features) 
  all_windows$feature <- bestOverlap(
    all_windows, external_features, var = "feature", missing = "no_feature"
  ) %>% 
  str_sep_to_title()
all_windows <- select(all_windows, any_of(c("region", "feature")), everything())
```


```{r set-direction-colours}
direction_colours <- names(colours$direction) %>% 
  rep(each = length(.)) %>% 
  cbind(rep(unique(.), times = 4)) %>% 
  set_colnames(targets) %>% 
  as_tibble() %>% 
  mutate(
    c1 = lapply(
      !!sym(targets[[1]]), function(x) t(col2rgb(colours$direction[[x]]))
      ), 
    c2 = lapply(
      !!sym(targets[[2]]), function(x) t(col2rgb(colours$direction[[x]]))
    )
  ) %>% 
  split(f = seq_len(nrow(.))) %>% 
  lapply(unnest, everything()) %>% 
  lapply(
    function(x) {
      vals <- rbind(x$c1, x$c2) %>% 
        colMeans()
      rgb <- rgb(
        vals["red"], vals["green"], vals["blue"], 
        maxColorValue = colours$direction %>%
          unlist() %>%
          col2rgb() %>% 
          max()
      )
      select(x, all_of(targets)) %>% 
        mutate(rgb = rgb)
    }
  ) %>% 
  bind_rows() %>% 
  mutate(
    !!sym(targets[[1]]) := paste(targets[[1]], str_to_title(!!sym(targets[[1]]))),
    !!sym(targets[[2]]) := paste(targets[[2]], str_to_title(!!sym(targets[[2]])))
  ) %>% 
  unite(status, !!!syms(targets), sep = " - ") %>% 
  mutate(rgb = setNames(rgb, status)) %>% 
  pull("rgb")
```

A set of common windows was formed by obtaining the union of windows produced during the analysis of `r targets[[1]]` and `r targets[[2]]` individually.
These common windows were formed using any overlap between target-specific windows.
The window-to-gene mappings from previous steps were maintained and the best overlap with genomic regions was also reformed.
`r if (has_features) "The best overlap with external features was also reformed."`

Merged windows were compared across both targets and in order to obtain the best classification for each window, the initial values of an FDR-adjusted p-value < `r fdr_alpha` was used.
A secondary step was then introduced for each window such that if one target was significant and the other target had received an FDR-adjusted p-value < `r 3 * fdr_alpha`, these windows were then considered significant in the second comparison.
This was to help *reduce the number of windows incorrectly classified as unchanged* whilst the alternative ChIP target was defined as changed.

```{r tbl-summary}
cp <- htmltools::tags$em(
  glue(
    "Summary of combined binding windows for ",
    glue_collapse(targets, sep = " and "),
    ". Windows were classified as described above. ",
    "The number of windows showing each combined pattern are given, along ",
    "with the number of unique genes mapped to each set of combined windows, ",
    "noting that some genes will be mapped to multiple sets of windows. ",
    "The % of all genes with a window mapped from either target is given in ",
    "the final column."
  )
)
tbl <- all_windows %>% 
  as_tibble() %>% 
  group_by(
    !!sym(glue("{targets[[1]]}_status")), !!sym(glue("{targets[[2]]}_status"))
  ) %>% 
  summarise(
    windows = dplyr::n(),
    mapped = length(unique(unlist(gene_id))),
    .groups = "drop"
  ) %>% 
  mutate(
    `% genes` = mapped / length(unique(unlist(all_windows$gene_id)))
  ) %>% 
  reactable(
    filterable = TRUE,
    columns = list2(
      "{targets[[1]]}_status" := colDef(
        name = paste(targets[[1]], "Status"),
        cell = function(value) {
          if (str_detect(value, "Up")) value <- paste(value, "\u21E7")
          if (str_detect(value, "Down")) value <- paste(value, "\u21E9")
          value
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            str_detect(value, "Unchanged") ~ colours$direction[["unchanged"]],
            str_detect(value, "Undetected") ~ colours$direction[["undetected"]]
          )
          list(color = colour)
        },
        footer = htmltools::tags$b("Total")
      ),
      "{targets[[2]]}_status" := colDef(
        name = paste(targets[[2]], "Status"),
        cell = function(value) {
          if (str_detect(value, "Up")) value <- paste(value, "\u21E7")
          if (str_detect(value, "Down")) value <- paste(value, "\u21E9")
          value
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            str_detect(value, "Unchanged") ~ colours$direction[["unchanged"]],
            str_detect(value, "Undetected") ~ colours$direction[["undetected"]]
          )
          list(color = colour)
        }
      ),
      windows = colDef(
        name = "Number of Windows", 
        cell = function(value) comma(value, 1),
        style = function(value) {
          bar_style(width = 0.9*value / length(all_windows), fill = "#B3B3B3")
        },
        footer = htmltools::tags$b(comma(length(all_windows), 1))
      ),
      mapped = colDef(
        name = "Uniquely Mapped Genes", 
        cell = function(value) comma(value, 1),
        style = function(value) {
          bar_style(width = 0.9*value /length(unique(unlist(all_windows$gene_id))), fill = "#B3B3B3")
        },
        footer = htmltools::tags$b(comma(length(unique(unlist(all_windows$gene_id))), 1))
      ),
      `% genes` = colDef(
        name = "% Of All Mapped Genes",
        cell = function(x) percent(x, 0.1),
        style = function(value) {
          bar_style(width = 0.9*value, fill = "#B3B3B3")
        }
      )
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```

Results across all windows are shown below for both `r glue_collapse(targets, sep = " and ")` separately, then together.

## `r targets[[1]]` By `r targets[[2]]` Status

```{r target1-by2, fig.cap = glue("*Differential binding patterns of {targets[[1]]} by {targets[[2]]} status. {targets[[1]]} values are coloured by their individual classification as Up, Down or Unchanged, with the distributions of these shown by {targets[[2]]} status in the side panels.*")}
all_windows %>%
  as_tibble() %>% 
  dplyr::filter(
    !!sym(glue("{targets[[1]]}_status")) != paste(targets[[1]], "Undetected")
  ) %>% 
  droplevels() %>% 
  ggplot(
    aes(
      !!sym(glue("{targets[[1]]}_AveExpr")), 
      !!sym(glue("{targets[[1]]}_logFC")),
      colour = !!sym(glue("{targets[[1]]}_status")), 
      fill = !!sym(glue("{targets[[2]]}_status"))
    )
  ) +
  geom_point() +
  geom_xsideboxplot(
    aes(y = !!sym(glue("{targets[[2]]}_status"))),
    orientation = "y", colour = "black"
  ) +
  geom_ysideboxplot(
    aes(x = !!sym(glue("{targets[[2]]}_status"))),
    orientation = "x", colour = "black"
  ) +
  geom_ysideline(
    aes(x, y),
    data = tibble(x = seq(0, 5), y = 0),
    inherit.aes = FALSE
  ) +
  geom_hline(yintercept = 0) +
  scale_xsidey_discrete() +
  scale_ysidex_discrete(guide = guide_axis(angle = 90)) +
  scale_colour_manual(
    values = colours$direction %>% 
      setNames(str_to_title(names(.))) %>% 
      setNames(paste(targets[[1]], names(.))) %>% 
      .[!str_detect(names(.), "Undetected")]
  ) +
  scale_fill_manual(
    values = colours$direction %>% 
      setNames(str_to_title(names(.))) %>% 
      setNames(paste(targets[[2]], names(.))) 
  ) +
  labs(
    x = paste(targets[[1]], "logCPM"),
    y = paste(targets[[1]], "logFC"),
    colour = paste(targets[[1]], "Status"),
    fill = paste(targets[[2]], "Status")
  ) +
  theme(
    ggside.panel.scale.y = .2,
    ggside.panel.scale.x = .25
  )
```

## `r targets[[2]]` By `r targets[[1]]` Status

```{r target2-by1, fig.cap = glue("*Differential binding patterns of {targets[[2]]} by {targets[[1]]} status. {targets[[2]]} values are coloured by their individual classification as Up, Down or Unchanged, with the distributions of these shown by {targets[[1]]} status in the side panels.*")}
all_windows %>%
  as_tibble() %>% 
  dplyr::filter(
    !!sym(glue("{targets[[2]]}_status")) != paste(targets[[2]], "Undetected")
  ) %>% 
  droplevels() %>% 
  ggplot(
    aes(
      !!sym(glue("{targets[[2]]}_AveExpr")), 
      !!sym(glue("{targets[[2]]}_logFC")),
      colour = !!sym(glue("{targets[[2]]}_status")), 
      fill = !!sym(glue("{targets[[1]]}_status"))
    )
  ) +
  geom_point() +
  geom_xsideboxplot(
    aes(y = !!sym(glue("{targets[[1]]}_status"))),
    orientation = "y", colour = "black"
  ) +
  geom_ysideboxplot(
    aes(x = !!sym(glue("{targets[[1]]}_status"))),
    orientation = "x", colour = "black"
  ) +
  geom_ysideline(
    aes(x, y),
    data = tibble(x = seq(0, 5), y = 0),
    inherit.aes = FALSE
  ) +
  geom_hline(yintercept = 0) +
  scale_xsidey_discrete() +
  scale_ysidex_discrete(guide = guide_axis(angle = 90)) +
  scale_colour_manual(
    values = colours$direction %>% 
      setNames(str_to_title(names(.))) %>% 
      setNames(paste(targets[[2]], names(.))) %>% 
      .[!str_detect(names(.), "Undetected")]
  ) +
  scale_fill_manual(
    values = colours$direction %>% 
      setNames(str_to_title(names(.))) %>% 
      setNames(paste(targets[[1]], names(.)))
  ) +
  labs(
    x = paste(targets[[2]], "logCPM"),
    y = paste(targets[[2]], "logFC"),
    colour = paste(targets[[2]], "Status"),
    fill = paste(targets[[1]], "Status")
  ) +
  theme(
    ggside.panel.scale.y = .2,
    ggside.panel.scale.x = .25
  )
```



## Windows By Differential Binding


```{r plot-dbwin, fig.cap = glue("*Comparative changes in both ", glue_collapse(targets, sep = " and "), " across ", glue_collapse(unique(vapply(pairs, paste, character(1), collapse = " & ")), sep = " or "), ". The window with most extreme combined change in binding is shown with any mapped genes labelled. The range around zero used for range-based hypothesis testing is indicated with grey dashed lines.*")}
all_windows %>% 
  as_tibble() %>% 
  dplyr::filter(!str_detect(status, "Undetected")) %>% 
  droplevels() %>% 
  ggplot(
    aes(
      !!sym(glue("{targets[[1]]}_logFC")), !!sym(glue("{targets[[2]]}_logFC")),
      colour = status
    )
  ) +
  geom_point() +
  geom_hline(yintercept = 0) +
  geom_hline(
    yintercept = c(1, -1)*log2(config$comparisons$fc), 
    linetype = 2, colour = "grey"
  ) +
  geom_vline(xintercept = 0) +
  geom_vline(
    xintercept = c(1, -1)*log2(config$comparisons$fc), 
    linetype = 2, colour = "grey"
  ) +
  geom_label_repel(
    aes(label = detected),
    data = . %>% 
      dplyr::filter(
        !str_detect(status, "Unchanged.+Unchanged"),
        vapply(detected, length, integer(1)) > 0
      ) %>% 
      mutate(
        total_lfc = abs(!!sym(glue("{targets[[1]]}_logFC"))) + abs(!!sym(glue("{targets[[2]]}_logFC"))) 
      ) %>% 
      group_by(status) %>% 
      dplyr::filter(total_lfc == max(total_lfc)) %>% 
      mutate(
        detected = vapply(detected, paste, character(1), collapse = "; ") %>% 
          str_wrap(15)
      ),
    size = 3, alpha = 0.6,
    show.legend = FALSE
  ) +
  geom_text(
    aes(x, y, label = lab),
    data = . %>%
      mutate(
        x = max((!!sym(glue("{targets[[1]]}_logFC")))),
        y = min(!!sym(glue("{targets[[2]]}_logFC")))
      ) %>% 
      # group_by(
      #   !!sym(glue("{targets[[1]]}_status")), !!sym(glue("{targets[[2]]}_status"))
      # ) %>%
      summarise(
        x = unique(x), 
        y = unique(y),
        n = dplyr::n(),
        lab = paste("n =", comma(n, 1)),
        .groups = "drop"
      ),
    inherit.aes = FALSE
  ) +
  # stat_ellipse(level = 0.9) +
  # facet_grid(
  #   as.formula(
  #     glue("{targets[[2]]}_status ~ fct_rev({targets[[1]]}_status)")
  #   )
  # ) +
  scale_colour_manual(
    values = direction_colours[
      filter(all_windows, !str_detect(status, "Undetected"))$status %>%
        droplevels() %>%
        levels()
    ]
  ) +
  labs(
    x = paste(targets[[1]], "logFC"),
    y = paste(targets[[2]], "logFC"),
    colour = "Status"
  ) +
  theme(legend.position = "top")
```


## Peaks By Region

```{r plot-dbwin-by-region, fig.cap = glue("*Comparative changes in both ", glue_collapse(targets, sep = " and "), " across ", glue_collapse(unique(vapply(pairs, paste, character(1), collapse = " & ")), sep = " or "), ". The window with most extreme combined change in binding for each region is shown with any mapped genes labelled. The range around zero used for range-based hypothesis testing is indicated with grey dashed lines.*")}
all_windows %>% 
  as_tibble() %>% 
  dplyr::filter(!str_detect(status, "Undetected")) %>% 
  droplevels() %>% 
  ggplot(
    aes(
      !!sym(glue("{targets[[1]]}_logFC")), !!sym(glue("{targets[[2]]}_logFC")),
      colour = status
    )
  ) +
  geom_point() +
  geom_hline(yintercept = 0) +
  geom_hline(
    yintercept = c(1, -1)*log2(config$comparisons$fc), 
    linetype = 2, colour = "grey"
  ) +
  geom_vline(xintercept = 0) +
  geom_vline(
    xintercept = c(1, -1)*log2(config$comparisons$fc), 
    linetype = 2, colour = "grey"
  ) +
  geom_label_repel(
    aes(label = label),
    data = . %>% 
      dplyr::filter(
        !str_detect(status, "Unchanged.+Unchanged"),
        vapply(detected, length, integer(1)) > 0
      ) %>% 
      mutate(
        total_lfc = abs(!!sym(glue("{targets[[1]]}_logFC"))) + abs(!!sym(glue("{targets[[2]]}_logFC"))) 
      ) %>% 
      group_by(region) %>% 
      dplyr::filter(total_lfc == max(total_lfc)) %>% 
      mutate(
        detected = vapply(detected, paste, character(1), collapse = "; "),
        label = paste(status, detected, sep = ":\n") %>% 
          str_wrap(25)
      ),
    size = 3, alpha = 0.6,
    show.legend = FALSE
  ) +
  geom_text(
    aes(x, y, label = lab),
    data = . %>%
      mutate(
        x = min(!!sym(glue("{targets[[1]]}_logFC"))),
        y = min(!!sym(glue("{targets[[2]]}_logFC")))
      ) %>% 
      group_by(region) %>%
      summarise(
        x = unique(x), 
        y = unique(y),
        n = dplyr::n(),
        lab = paste("n =", comma(n, 1)),
        .groups = "drop"
      ),
    inherit.aes = FALSE
  ) +
  facet_wrap(~region) +
  scale_colour_manual(
    values = direction_colours[
      filter(all_windows, !str_detect(status, "Undetected"))$status %>% 
        droplevels() %>% 
        levels()
    ]
  ) +
  labs(
    x = paste(targets[[1]], "logFC"),
    y = paste(targets[[2]], "logFC"),
    colour = "Status"
  ) +
  theme(legend.position = "top")
```

`r if (has_features) "## Peaks By Feature"`

```{r plot-dbwin-by-feature, eval = has_features, echo = has_features, fig.cap = glue("*Comparative changes in both ", glue_collapse(targets, sep = " and "), " across ", glue_collapse(unique(vapply(pairs, paste, character(1), collapse = " & ")), sep = " or "), ". The window with most extreme combined change in binding for each feature is shown with any mapped genes labelled. The range around zero used for range-based hypothesis testing is indicated with grey dashed lines.*")}
all_windows %>% 
  as_tibble() %>% 
  dplyr::filter(!str_detect(status, "Undetected")) %>% 
  droplevels() %>% 
  ggplot(
    aes(
      !!sym(glue("{targets[[1]]}_logFC")), !!sym(glue("{targets[[2]]}_logFC")),
      colour = status
    )
  ) +
  geom_point() +
  geom_hline(yintercept = 0) +
  geom_hline(
    yintercept = c(1, -1)*log2(config$comparisons$fc), 
    linetype = 2, colour = "grey"
  ) +
  geom_vline(xintercept = 0) +
  geom_vline(
    xintercept = c(1, -1)*log2(config$comparisons$fc), 
    linetype = 2, colour = "grey"
  ) +
  geom_label_repel(
    aes(label = label),
    data = . %>% 
      dplyr::filter(!str_detect(status, "Unchanged.+Unchanged")) %>% 
      mutate(
        total_lfc = abs(!!sym(glue("{targets[[1]]}_logFC"))) + abs(!!sym(glue("{targets[[2]]}_logFC"))) 
      ) %>% 
      group_by(feature) %>% 
      dplyr::filter(total_lfc == max(total_lfc)) %>% 
      mutate(
        detected = vapply(detected, paste, character(1), collapse = "; "),
        label = paste(status, detected, sep = ":\n") %>% 
          str_wrap(25)
      ) %>% 
      dplyr::filter(detected != ""),
    size = 3, alpha = 0.6,
    show.legend = FALSE
  ) +
  geom_text(
    aes(x, y, label = lab),
    data = . %>%
      mutate(
        x = min(!!sym(glue("{targets[[1]]}_logFC"))),
        y = min(!!sym(glue("{targets[[2]]}_logFC")))
      ) %>% 
      group_by(feature) %>%
      summarise(
        x = unique(x), 
        y = unique(y),
        n = dplyr::n(),
        lab = paste("n =", comma(n, 1)),
        .groups = "drop"
      ),
    inherit.aes = FALSE
  ) +
  facet_wrap(~feature) +
  scale_colour_manual(
    values = direction_colours[
      filter(all_windows, !str_detect(status, "Undetected"))$status %>% 
        droplevels() %>% 
        levels()
    ]
  ) +
  labs(
    x = paste(targets[[1]], "logFC"),
    y = paste(targets[[2]], "logFC"),
    colour = "Status"
  ) +
  theme(legend.position = "top")
```


# Enrichment Analysis


```{r msigdb}
all_ids <- all_gr$gene$gene_id
msigdb <- msigdbr(species = "Homo sapiens") %>% 
  dplyr::filter(
    gs_cat %in% unlist(params$msigdb$gs_cat) |
      gs_subcat %in% unlist(params$msigdb$gs_subcat),
    str_detect(ensembl_gene, "^E")
  ) %>% 
  dplyr::rename(gene_id = ensembl_gene, gene_name = gene_symbol) %>% # For easier integration 
  dplyr::select(-starts_with("human"), -contains("entrez")) %>% 
  dplyr::filter(gene_id %in% all_ids)
gs_by_gsid <- msigdb %>% 
  split(.$gs_name) %>% 
  lapply(pull, "gene_id") %>% 
  lapply(unique)
gs_by_geneid <- msigdb %>% 
  split(.$gene_id) %>% 
  lapply(pull, "gs_name") %>% 
  lapply(unique)
```


```{r goseq-groups}
mapped_ids <- all_windows$gene_id %>%
  unlist() %>% 
  unique() %>% 
  sort()
gene_lengths <- structure(width(all_gr$gene), names = all_gr$gene$gene_id)[mapped_ids]
group_ids <- all_windows %>% 
  split(f = .$status) %>% 
  lapply(function(x) unique(unlist(x$gene_id))
  ) 
goseq_group_res <- group_ids %>% 
  mclapply(
    function(x) {
      vec <- structure(mapped_ids %in% x, names = mapped_ids)
      pwf <- nullp(vec, bias.data = log10(gene_lengths))
      res <- as_tibble(goseq(pwf, gene2cat = gs_by_geneid))
      res$fdr <- p.adjust(res$over_represented_pvalue, "fdr")
      res <- dplyr::select(
        res,
        gs_name = category, pval = over_represented_pvalue, fdr, ends_with("InCat")
      )
      res %>% 
        # dplyr::filter(fdr < fdr_alpha) %>% 
        dplyr::slice(1) %>% 
        left_join(
          dplyr::filter(msigdb, gene_id %in% x)[c("gs_name", "gene_name", "gs_url")],
          by = "gs_name"
        ) %>% 
        chop(gene_name) %>% 
        mutate(gene_name = vapply(gene_name, paste, character(1), collapse = "; "))
    },
    mc.cores = 6
  ) %>% 
  lapply(list) %>% 
  as_tibble() %>% 
  pivot_longer(everything(), names_to = "group") %>% 
  unnest(everything())
```

The same MSigDB gene sets were used for associating any pathways with the combined changes in both `r glue_collapse(targets, sep = " and ")`.
The 'universe' of genes was defined as all `r comma(length(mapped_ids))` gene ids with a binding window mapped from either target.
`r if(nrow(goseq_group_res) == 0) "No enrichment was found amongst any group of windows"`


```{r goseq-group-res, eval = nrow(goseq_group_res) > 0, echo =  nrow(goseq_group_res) > 0}
cp <- htmltools::tags$em(
  glue(
    "Pathways enriched amongst genes mapped to windows, using both ",
    glue_collapse(targets, sep = " and "),
    " status to define groups of genes."
  )
)
tbl <- goseq_group_res %>% 
  nest(gs_name = all_of(c("gs_name", "gs_url"))) %>% 
  separate(
    group, into = targets, sep = " - "
  ) %>% 
  dplyr::select(all_of(targets), gs_name, everything()) %>% 
  reactable(
    filterable = TRUE,
    columns = list2(
      "{targets[[1]]}" := colDef(
        name = paste(targets[[1]], "Status"),
        cell = function(value) {
          if (str_detect(value, "Up")) value <- paste(value, "\u21E7")
          if (str_detect(value, "Down")) value <- paste(value, "\u21E9")
          value
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            str_detect(value, "Unchanged") ~ colours$direction[["unchanged"]],
            str_detect(value, "Undetected") ~ colours$direction[["undetected"]]
          )
          list(color = colour)
        },
        maxWidth = 150
      ),
      "{targets[[2]]}" := colDef(
        name = paste(targets[[2]], "Status"),
        cell = function(value) {
          if (str_detect(value, "Up")) value <- paste(value, "\u21E7")
          if (str_detect(value, "Down")) value <- paste(value, "\u21E9")
          value
        },
        style = function(value) {
          colour <- case_when(
            str_detect(value, "Up") ~ colours$direction[["up"]],
            str_detect(value, "Down") ~ colours$direction[["down"]],
            str_detect(value, "Unchanged") ~ colours$direction[["unchanged"]],
            str_detect(value, "Undetected") ~ colours$direction[["undetected"]]
          )
          list(color = colour)
        },
        maxWidth = 150
      ),
      gs_name = colDef(
        name = "Gene Set",
        cell = function(value) {
          htmltools::tags$a(
            href = value$gs_url, target = "_blank",
            str_replace_all(value$gs_name, "_", " ")
          )
        },
        html = TRUE
      ),
      pval = colDef(
        name = "P-Value",
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ),
        maxWidth = 80
      ),
      fdr = colDef(
        name = "FDR",
        cell = function(value) ifelse(
          value < 0.001,
          sprintf("%.2e", value),
          sprintf("%.3f", value)
        ),
        maxWidth = 80
      ),
      numDEInCat = colDef(
        name = "Genes In Group",
        cell = function(value) comma(value, 1),
        maxWidth = 80
      ),
      numInCat = colDef(
        name = "All Genes",
        cell = function(value) comma(value, 1),
        maxWidth = 80
      ),
      gene_name = colDef(
        name = "Genes",
        cell = function(value) with_tooltip(value, width = 30)
      )
    )
  )
div(class = "table",
  div(class = "table-header",
      div(class = "caption", cp),
      tbl
  )
)
```





<!-- # RNA-Seq -->

<!-- Add another RNA-Seq module here -->


